// @matter/protocol/dist/cjs/dcl/DclConfig.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DclConfig_exports = {};
__export(DclConfig_exports, {
  DclConfig: () => DclConfig,
  DclGithubConfig: () => DclGithubConfig
});
module.exports = __toCommonJS(DclConfig_exports);
/**
 * @license
 * Copyright 2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DclConfig;
((DclConfig2) => {
  DclConfig2.production = {
    url: "https://on.dcl.csa-iot.org"
  };
  DclConfig2.test = {
    url: "https://on.test-net.dcl.csa-iot.org"
  };
})(DclConfig || (DclConfig = {}));
var DclGithubConfig;
((DclGithubConfig2) => {
  DclGithubConfig2.defaults = {
    owner: "project-chip",
    repo: "connectedhomeip",
    branch: "master",
    certPath: "credentials/development/paa-root-certs"
  };
})(DclGithubConfig || (DclGithubConfig = {}));
