// @matter/protocol/dist/cjs/dcl/DclRestApiTypes.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DclRestApiTypes_exports = {};
__export(DclRestApiTypes_exports, {
  DclErrorCodes: () => DclErrorCodes
});
module.exports = __toCommonJS(DclRestApiTypes_exports);
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DclErrorCodes = /* @__PURE__ */ ((DclErrorCodes2) => {
  DclErrorCodes2[DclErrorCodes2["Ok"] = 0] = "Ok";
  DclErrorCodes2[DclErrorCodes2["Canceled"] = 1] = "Canceled";
  DclErrorCodes2[DclErrorCodes2["Unknown"] = 2] = "Unknown";
  DclErrorCodes2[DclErrorCodes2["InvalidArgument"] = 3] = "InvalidArgument";
  DclErrorCodes2[DclErrorCodes2["DeadlineExceeded"] = 4] = "DeadlineExceeded";
  DclErrorCodes2[DclErrorCodes2["NotFound"] = 5] = "NotFound";
  DclErrorCodes2[DclErrorCodes2["AlreadyExists"] = 6] = "AlreadyExists";
  DclErrorCodes2[DclErrorCodes2["PermissionDenied"] = 7] = "PermissionDenied";
  DclErrorCodes2[DclErrorCodes2["ResourceExhausted"] = 8] = "ResourceExhausted";
  DclErrorCodes2[DclErrorCodes2["FailedPrecondition"] = 9] = "FailedPrecondition";
  DclErrorCodes2[DclErrorCodes2["Aborted"] = 10] = "Aborted";
  DclErrorCodes2[DclErrorCodes2["OutOfRange"] = 11] = "OutOfRange";
  DclErrorCodes2[DclErrorCodes2["Unimplemented"] = 12] = "Unimplemented";
  DclErrorCodes2[DclErrorCodes2["Internal"] = 13] = "Internal";
  DclErrorCodes2[DclErrorCodes2["Unavailable"] = 14] = "Unavailable";
  DclErrorCodes2[DclErrorCodes2["DataLoss"] = 15] = "DataLoss";
  DclErrorCodes2[DclErrorCodes2["Unauthenticated"] = 16] = "Unauthenticated";
  return DclErrorCodes2;
})(DclErrorCodes || {});
