// @matter/protocol/dist/cjs/dcl/DclVendorInfoService.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DclVendorInfoService_exports = {};
__export(DclVendorInfoService_exports, {
  DclVendorInfoService: () => DclVendorInfoService
});
module.exports = __toCommonJS(DclVendorInfoService_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_PairingHintBitmap = require("../advertisement/PairingHintBitmap.js");
var import_DclClient = require("./DclClient.js");
var import_DclConfig = require("./DclConfig.js");
var import_DclRestApiTypes = require("./DclRestApiTypes.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("DclVendorInfoService");
class DclVendorInfoService {
  #construction;
  #storageManager;
  #storage;
  #vendorIndex = /* @__PURE__ */ new Map();
  #updateTimer;
  #closed = false;
  #options;
  #fetchPromise;
  constructor(environment, options = {}) {
    environment.root.set(DclVendorInfoService, this);
    this.#options = options;
    logger.info(
      "Initialize VendorInfoService",
      import_general.Diagnostic.dict({
        source: options.dclConfig?.url ?? import_DclConfig.DclConfig.production.url,
        interval: options.updateInterval === null ? "disabled" : import_general.Duration.format(options.updateInterval ?? import_general.Days.one)
      })
    );
    this.#construction = (0, import_general.Construction)(this, async () => {
      this.#storageManager = await environment.get(import_general.StorageService).open("vendors");
      this.#storage = this.#storageManager.createContext("info");
      await this.#loadVendors(this.#storage);
      await this.update();
      if (options.updateInterval !== null) {
        const updateInterval = options.updateInterval ?? import_general.Days.one;
        this.#updateTimer = import_general.Time.getPeriodicTimer(
          "DCL Vendor Info Update",
          updateInterval,
          () => this.update()
        ).start();
      }
    });
  }
  get construction() {
    return this.#construction;
  }
  /**
   * Get vendor information by vendor ID. Returns undefined if not found.
   */
  infoFor(vendorId) {
    this.construction.assert();
    return this.#vendorIndex.get(vendorId);
  }
  /**
   * Get all vendor information as a map indexed by vendor ID.
   */
  get vendors() {
    this.construction.assert();
    return this.#vendorIndex;
  }
  /**
   * Update vendor information from DCL.
   */
  async update() {
    if (this.#closed || !this.#storage) {
      return;
    }
    if (this.#fetchPromise !== void 0) {
      await this.#fetchPromise;
      return;
    }
    try {
      this.#fetchPromise = this.#fetchVendorsFromDcl(this.#storage).finally(() => {
        this.#fetchPromise = void 0;
      });
      await this.#fetchPromise;
    } catch (error) {
      logger.info("Error updating vendor information", error);
    }
  }
  /**
   * Fetch product details for a given vendor ID and product ID from DCL.
   * Returns undefined if the product is not found or the service is closed.
   */
  async productInfoFor(vendorId, productId) {
    if (this.#closed) {
      return void 0;
    }
    try {
      const dclClient = new import_DclClient.DclClient(this.#options.dclConfig ?? import_DclConfig.DclConfig.production);
      const model = await dclClient.fetchModelByVidPid(vendorId, productId, this.#options);
      return this.#normalizeProductModel(model);
    } catch (error) {
      if (error instanceof import_DclClient.MatterDclResponseError && error.response.code === import_DclRestApiTypes.DclErrorCodes.NotFound) {
        logger.debug("Product not found in DCL", import_general.Diagnostic.dict({ vendorId, productId }));
        return void 0;
      }
      throw error;
    }
  }
  #normalizeProductModel(model) {
    const {
      deviceTypeID,
      productName,
      productLabel,
      partNumber,
      discoveryCapabilitiesBitmask,
      commissioningCustomFlow,
      commissioningCustomFlowUrl,
      commissioningModeInitialStepsHint,
      commissioningModeInitialStepsInstruction,
      commissioningModeSecondaryStepsHint,
      commissioningModeSecondaryStepsInstruction,
      commissioningFallbackUrl,
      userManualUrl,
      supportUrl,
      productUrl,
      lsfUrl,
      lsfRevision,
      enhancedSetupFlowOptions,
      enhancedSetupFlowTCUrl,
      enhancedSetupFlowTCRevision,
      enhancedSetupFlowMaintenanceUrl
    } = model;
    try {
      return {
        deviceTypeID,
        productName,
        productLabel,
        partNumber,
        discoveryCapabilities: import_types.DiscoveryCapabilitiesSchema.decode(discoveryCapabilitiesBitmask),
        commissioningFlow: commissioningCustomFlow,
        commissioningCustomFlowUrl,
        commissioningModeInitialStepsHint: import_PairingHintBitmap.PairingHintBitmapSchema.decode(commissioningModeInitialStepsHint),
        commissioningModeInitialStepsInstruction,
        commissioningModeSecondaryStepsHint: import_PairingHintBitmap.PairingHintBitmapSchema.decode(
          commissioningModeSecondaryStepsHint
        ),
        commissioningModeSecondaryStepsInstruction,
        commissioningFallbackUrl,
        userManualUrl,
        supportUrl,
        productUrl,
        lsfUrl,
        lsfRevision,
        enhancedSetupFlowOptions: enhancedSetupFlowOptions !== void 0 ? import_types.EnhancedSetupFlowOptionsSchema.decode(enhancedSetupFlowOptions) : void 0,
        enhancedSetupFlowTCUrl,
        enhancedSetupFlowTCRevision,
        enhancedSetupFlowMaintenanceUrl
      };
    } catch (error) {
      logger.warn(
        "Invalid DCL product entry",
        import_general.Diagnostic.dict({ vendorId: model.vid, productId: model.pid }),
        error
      );
      return void 0;
    }
  }
  /**
   * Load vendor information from storage.
   */
  async #loadVendors(storage) {
    const storedVendors = await storage.get("vendors", []);
    for (const vendor of storedVendors) {
      this.#vendorIndex.set(vendor.vendorId, vendor);
    }
    if (storedVendors.length > 0) {
      logger.info(`Loaded ${this.#vendorIndex.size} vendors from storage`);
    }
  }
  /**
   * Fetch vendor information from DCL and store it.
   */
  async #fetchVendorsFromDcl(storage) {
    logger.info("Fetching vendor information from DCL");
    const dclClient = new import_DclClient.DclClient(this.#options.dclConfig ?? import_DclConfig.DclConfig.production);
    const vendors = await dclClient.fetchAllVendors(this.#options);
    logger.info(`Fetched ${vendors.length} vendors from DCL`);
    const hardcodedVendors = [
      ...[65521, 65522, 65523, 65524].map((vendorID) => ({
        vendorID,
        vendorName: "Test Vendor",
        companyLegalName: "Test Vendor Inc.",
        companyPreferredName: "Test Vendor",
        vendorLandingPageURL: "https://test.example.com",
        creator: "System"
      })),
      {
        vendorID: 4939,
        vendorName: "Nabu Casa",
        companyLegalName: "Nabu Casa, Inc.",
        companyPreferredName: "Nabu Casa",
        vendorLandingPageURL: "https://www.nabucasa.com",
        creator: "System"
      }
    ];
    for (const hardcoded of hardcodedVendors) {
      if (!vendors.find((v) => v.vendorID === hardcoded.vendorID)) {
        vendors.push(hardcoded);
      }
    }
    const vendorsToStore = [];
    for (const vendor of vendors) {
      const normalizedVendor = this.#normalizeDclVendor(vendor);
      this.#vendorIndex.set(normalizedVendor.vendorId, normalizedVendor);
      vendorsToStore.push(normalizedVendor);
    }
    await storage.set("vendors", vendorsToStore);
    logger.info(`Stored ${this.#vendorIndex.size} vendors`);
  }
  /**
   * Normalize DCL vendor format to internal format.
   */
  #normalizeDclVendor(dclVendor) {
    return {
      vendorId: dclVendor.vendorID,
      vendorName: dclVendor.vendorName,
      companyLegalName: dclVendor.companyLegalName,
      companyPreferredName: dclVendor.companyPreferredName,
      vendorLandingPageUrl: dclVendor.vendorLandingPageURL,
      creator: dclVendor.creator
    };
  }
  /**
   * Close the service and stop periodic updates.
   */
  async close() {
    if (this.#closed) return;
    this.#closed = true;
    await this.#construction.close(async () => {
      await this.#fetchPromise;
      this.#updateTimer?.stop();
      this.#updateTimer = void 0;
      await this.#storageManager?.close();
    });
  }
}
