// homebridge/dist/matter/server/AccessoryQuery.js
import { Logger } from '../../logger.js';
const log = Logger.withPrefix('Matter/Server');
export class AccessoryQuery {
    accessories;
    getAccessoryCache;
    constructor(accessories, getAccessoryCache) {
        this.accessories = accessories;
        this.getAccessoryCache = getAccessoryCache;
    }
    getAccessories() {
        return Array.from(this.accessories.values(), (acc) => {
            const { endpoint, registered, ...publicAccessory } = acc;
            return publicAccessory;
        });
    }
    getAccessory(uuid) {
        const accessory = this.accessories.get(uuid);
        if (!accessory) {
            return undefined;
        }
        const { endpoint, registered, ...publicAccessory } = accessory;
        return publicAccessory;
    }
    getAllCachedAccessories() {
        const cache = this.getAccessoryCache();
        if (!cache) {
            log.debug('getAllCachedAccessories: No cache available');
            return [];
        }
        const cached = [...cache.getAllCached().values()];
        log.debug(`getAllCachedAccessories: Returning ${cached.length} accessories`);
        return cached;
    }
    getDeviceTypeName(deviceType) {
        return deviceType.name || 'Unknown';
    }
    getCurrentState(uuid, partId) {
        const accessory = this.accessories.get(uuid);
        if (!accessory) {
            return {};
        }
        if (partId) {
            const part = accessory._parts?.find(p => p.id === partId);
            return part?.clusters || {};
        }
        return accessory.clusters || {};
    }
    collectAccessories(bridgeUsername, bridgeType, bridgeName) {
        const accessories = [];
        for (const [uuid, accessory] of this.accessories.entries()) {
            const transformed = {
                uuid,
                displayName: accessory.displayName,
                deviceType: this.getDeviceTypeName(accessory.deviceType),
                clusters: this.getCurrentState(uuid),
                parts: accessory._parts?.map(part => ({
                    id: part.id,
                    displayName: part.displayName,
                    deviceType: this.getDeviceTypeName(part.deviceType),
                    clusters: this.getCurrentState(uuid, part.id),
                })),
                bridge: {
                    username: bridgeUsername,
                    type: bridgeType,
                    name: bridgeName,
                },
            };
            accessories.push(transformed);
        }
        return accessories;
    }
    getAccessoryInfo(uuid) {
        const accessory = this.accessories.get(uuid);
        if (!accessory) {
            return undefined;
        }
        return {
            uuid,
            displayName: accessory.displayName,
            deviceType: this.getDeviceTypeName(accessory.deviceType),
            clusters: this.getCurrentState(uuid),
            parts: accessory._parts?.map(part => ({
                id: part.id,
                displayName: part.displayName,
                deviceType: this.getDeviceTypeName(part.deviceType),
                clusters: this.getCurrentState(uuid, part.id),
            })),
        };
    }
}
