// homebridge/dist/matter/server/BehaviorMap.js
import { HomebridgeAirQualityServer, HomebridgeCarbonMonoxideConcentrationMeasurementServer, HomebridgeColorControlServer, HomebridgeDoorLockServer, HomebridgeFanControlServer, HomebridgeIdentifyServer, HomebridgeLevelControlServer, HomebridgeNitrogenDioxideConcentrationMeasurementServer, HomebridgeOnOffServer, HomebridgeOzoneConcentrationMeasurementServer, HomebridgePm10ConcentrationMeasurementServer, HomebridgePm25ConcentrationMeasurementServer, HomebridgeRvcCleanModeServer, HomebridgeRvcOperationalStateServer, HomebridgeRvcRunModeServer, HomebridgeServiceAreaServer, HomebridgeThermostatServer, HomebridgeValveConfigurationAndControlServer, HomebridgeWindowCoveringServer, } from '../behaviors/index.js';
export const CORE_CLUSTER_BEHAVIOR_MAP = {
    onOff: HomebridgeOnOffServer,
    levelControl: HomebridgeLevelControlServer,
    colorControl: HomebridgeColorControlServer,
    windowCovering: HomebridgeWindowCoveringServer,
    doorLock: HomebridgeDoorLockServer,
    fanControl: HomebridgeFanControlServer,
    thermostat: HomebridgeThermostatServer,
    rvcOperationalState: HomebridgeRvcOperationalStateServer,
    rvcRunMode: HomebridgeRvcRunModeServer,
    rvcCleanMode: HomebridgeRvcCleanModeServer,
    serviceArea: HomebridgeServiceAreaServer,
    identify: HomebridgeIdentifyServer,
    valveConfigurationAndControl: HomebridgeValveConfigurationAndControlServer,
};
export const FULL_CLUSTER_BEHAVIOR_MAP = {
    ...CORE_CLUSTER_BEHAVIOR_MAP,
    airQuality: HomebridgeAirQualityServer,
    carbonMonoxideConcentrationMeasurement: HomebridgeCarbonMonoxideConcentrationMeasurementServer,
    nitrogenDioxideConcentrationMeasurement: HomebridgeNitrogenDioxideConcentrationMeasurementServer,
    ozoneConcentrationMeasurement: HomebridgeOzoneConcentrationMeasurementServer,
    pm10ConcentrationMeasurement: HomebridgePm10ConcentrationMeasurementServer,
    pm25ConcentrationMeasurement: HomebridgePm25ConcentrationMeasurementServer,
};
