// homebridge/dist/matter/server/CommissioningManager.js
import { randomBytes } from 'node:crypto';
import { writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import { ManualPairingCodeCodec, QrCode, QrPairingCodeCodec } from '@matter/types/schema';
import { Logger } from '../../logger.js';
import { DEFAULT_PRODUCT_ID, DEFAULT_VENDOR_ID, MAX_PASSCODE_ATTEMPTS } from './ServerConfig.js';
const log = Logger.withPrefix('Matter/Server');
export class CommissioningManager {
    passcode = 0;
    discriminator = 0;
    vendorId;
    productId;
    commissioningInfo = {};
    constructor() {
        this.vendorId = DEFAULT_VENDOR_ID;
        this.productId = DEFAULT_PRODUCT_ID;
    }
    generateSecurePasscode() {
        let passcode;
        const maxAttempts = MAX_PASSCODE_ATTEMPTS;
        let attempts = 0;
        const invalidPasscodes = [
            0,
            11111111,
            22222222,
            33333333,
            44444444,
            55555555,
            66666666,
            77777777,
            88888888,
            99999999,
            12345678,
            87654321,
        ];
        do {
            const maxRange = 99999998;
            const maxUint32 = 0xFFFFFFFF;
            const limit = Math.floor(maxUint32 / maxRange) * maxRange;
            let randomValue;
            do {
                randomValue = randomBytes(4).readUInt32BE(0);
            } while (randomValue >= limit);
            passcode = (randomValue % maxRange) + 1;
            attempts++;
            if (attempts > maxAttempts) {
                throw new Error('Failed to generate secure passcode after maximum attempts');
            }
        } while (invalidPasscodes.includes(passcode)
            || !this.isValidPasscode(passcode));
        return passcode;
    }
    isValidPasscode(passcode) {
        if (passcode < 1 || passcode > 99999998) {
            return false;
        }
        const passcodeStr = passcode.toString().padStart(8, '0');
        let isSequential = true;
        for (let i = 1; i < passcodeStr.length; i++) {
            if (Number.parseInt(passcodeStr[i]) !== Number.parseInt(passcodeStr[i - 1]) + 1) {
                isSequential = false;
                break;
            }
        }
        if (isSequential) {
            return false;
        }
        let isReverseSequential = true;
        for (let i = 1; i < passcodeStr.length; i++) {
            if (Number.parseInt(passcodeStr[i]) !== Number.parseInt(passcodeStr[i - 1]) - 1) {
                isReverseSequential = false;
                break;
            }
        }
        if (isReverseSequential) {
            return false;
        }
        const digitCounts = new Map();
        for (const digit of passcodeStr) {
            digitCounts.set(digit, (digitCounts.get(digit) || 0) + 1);
            const count = digitCounts.get(digit);
            if (count !== undefined && count > 3) {
                return false;
            }
        }
        return true;
    }
    generateRandomDiscriminator() {
        const discriminator = randomBytes(2).readUInt16BE(0) & 0x0FFF;
        if (discriminator < 0 || discriminator > 4095) {
            throw new Error(`Invalid discriminator generated: ${discriminator}`);
        }
        return discriminator;
    }
    async loadOrGenerateCredentials(matterStoragePath) {
        const credentialsPath = join(matterStoragePath, 'credentials.json');
        try {
            const { readFile } = await import('node:fs/promises');
            const data = JSON.parse(await readFile(credentialsPath, 'utf-8'));
            if (data.passcode && data.discriminator) {
                log.info('Loading existing commissioning credentials from storage');
                this.passcode = data.passcode;
                this.discriminator = data.discriminator;
                return;
            }
        }
        catch {
        }
        log.info('Generating new commissioning credentials');
        this.passcode = this.generateSecurePasscode();
        this.discriminator = this.generateRandomDiscriminator();
        await writeFile(credentialsPath, JSON.stringify({ passcode: this.passcode, discriminator: this.discriminator }, null, 2), 'utf-8');
        log.info('Commissioning credentials saved to storage');
    }
    async generateCommissioningInfo(deps) {
        const passcode = this.passcode.toString().padStart(8, '0');
        const discriminator = this.discriminator;
        const vendorId = this.vendorId;
        const productId = this.productId;
        const manualCode = ManualPairingCodeCodec.encode({
            discriminator,
            passcode: this.passcode,
        });
        const manualPairingCode = `${manualCode.slice(0, 4)}-${manualCode.slice(4, 7)}-${manualCode.slice(7, 11)}`;
        log.info(`Encoding QR code with: passcode=${this.passcode}, discriminator=${discriminator}, vendorId=${vendorId}, productId=${productId}`);
        const qrCodePayload = QrPairingCodeCodec.encode([{
                version: 0,
                vendorId,
                productId,
                flowType: 0,
                discoveryCapabilities: 4,
                discriminator,
                passcode: this.passcode,
            }]);
        log.info(`Generated QR code: ${qrCodePayload}`);
        log.info(`Generated manual code: ${manualPairingCode}`);
        this.commissioningInfo = {
            qrCode: qrCodePayload,
            manualPairingCode,
        };
        try {
            if (!deps.matterStoragePath) {
                throw new Error('Matter storage path not initialized');
            }
            const commissioningFilePath = join(deps.matterStoragePath, 'commissioning.json');
            const commissioningData = {
                qrCode: qrCodePayload,
                manualPairingCode,
                serialNumber: deps.serialNumber,
                passcode: this.passcode,
                discriminator: this.discriminator,
                commissioned: deps.fabricManager.isCommissioned(),
            };
            await writeFile(commissioningFilePath, JSON.stringify(commissioningData, null, 2), 'utf-8');
            log.debug(`Saved commissioning info to ${commissioningFilePath}`);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            log.warn(`Failed to save commissioning info to disk: ${errorMessage}`);
        }
        log.info(`${'='.repeat(60)}`);
        log.info('📱 MATTER COMMISSIONING INFORMATION');
        log.info('='.repeat(60));
        log.info(`Manual Pairing Code: ${manualPairingCode}`);
        log.info(`Passcode: ${passcode}`);
        log.info(`Discriminator: ${discriminator}`);
        log.info('QR Code for commissioning:');
        const qrCodeString = QrCode.get(qrCodePayload);
        log.info(`\n${qrCodeString}`);
        log.info(`${'='.repeat(60)}`);
    }
    setupCommissioningEventListeners(deps) {
        if (!deps.serverNode) {
            log.warn('Cannot set up commissioning event listeners - serverNode not initialized');
            return;
        }
        log.debug('Setting up commissioning event listeners');
        try {
            deps.serverNode.events.commissioning.fabricsChanged.on((fabricIndex, action) => {
                log.info(`Fabric ${action}: index ${fabricIndex}`);
                this.updateCommissioningFile(deps).catch((error) => {
                    log.warn('Failed to update commissioning file after fabric change:', error);
                });
                const commissioned = deps.fabricManager.isCommissioned();
                const fabricCount = deps.fabricManager.getCommissionedFabricCount();
                deps.emitter.emit('commissioning-status-changed', commissioned, fabricCount);
            });
            deps.serverNode.events.commissioning.commissioned.on(() => {
                log.info('Bridge commissioned');
                this.updateCommissioningFile(deps).catch((error) => {
                    log.warn('Failed to update commissioning file after commissioning:', error);
                });
                const fabricCount = deps.fabricManager.getCommissionedFabricCount();
                deps.emitter.emit('commissioning-status-changed', true, fabricCount);
            });
            deps.serverNode.events.commissioning.decommissioned.on(() => {
                log.info('Bridge decommissioned');
                this.updateCommissioningFile(deps).catch((error) => {
                    log.warn('Failed to update commissioning file after decommissioning:', error);
                });
                deps.emitter.emit('commissioning-status-changed', false, 0);
            });
            log.debug('Commissioning event listeners registered successfully');
        }
        catch (error) {
            log.error('Failed to set up commissioning event listeners:', error);
        }
    }
    async updateCommissioningFile(deps) {
        try {
            if (!deps.matterStoragePath) {
                return;
            }
            const commissioningFilePath = join(deps.matterStoragePath, 'commissioning.json');
            const commissioningData = {
                qrCode: this.commissioningInfo.qrCode,
                manualPairingCode: this.commissioningInfo.manualPairingCode,
                serialNumber: deps.serialNumber,
                passcode: this.passcode,
                discriminator: this.discriminator,
                commissioned: deps.fabricManager.isCommissioned(),
                fabricCount: deps.fabricManager.getCommissionedFabricCount(),
                fabrics: deps.fabricManager.getFabricInfo(),
            };
            await writeFile(commissioningFilePath, JSON.stringify(commissioningData, null, 2), 'utf-8');
            log.debug('Updated commissioning info file');
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            log.debug(`Failed to update commissioning info file: ${errorMessage}`);
        }
    }
}
