// homebridge/dist/matter/server/FabricManager.js
import fs from 'node:fs';
import path from 'node:path';
import { Logger } from '../../logger.js';
import { MatterDeviceError } from '../types.js';
const log = Logger.withPrefix('Matter/Server');
export class FabricManager {
    getServerNode;
    getMatterStoragePath;
    constructor(getServerNode, getMatterStoragePath) {
        this.getServerNode = getServerNode;
        this.getMatterStoragePath = getMatterStoragePath;
    }
    getFabricInfo() {
        try {
            const serverNode = this.getServerNode();
            if (!serverNode) {
                return [];
            }
            const env = serverNode.env;
            if (!env) {
                return [];
            }
            try {
                const serverState = serverNode;
                const fabrics = serverState?.state?.operationalCredentials?.fabrics;
                if (Array.isArray(fabrics) && fabrics.length > 0) {
                    return fabrics.map((fabric) => ({
                        fabricIndex: fabric.fabricIndex || 0,
                        fabricId: fabric.fabricId?.toString() || '',
                        nodeId: fabric.nodeId?.toString() || '',
                        rootVendorId: fabric.rootVendorId || 0,
                        label: fabric.label || '',
                    }));
                }
            }
            catch {
            }
            return this.readFabricsFromStorage();
        }
        catch (error) {
            log.debug('Failed to get fabric info:', error);
            return [];
        }
    }
    readFabricsFromStorage() {
        try {
            const storagePath = this.getMatterStoragePath();
            if (!storagePath) {
                return [];
            }
            const files = fs.readdirSync(storagePath).filter((f) => f.endsWith('.json'));
            for (const file of files) {
                try {
                    const data = JSON.parse(fs.readFileSync(path.join(storagePath, file), 'utf-8'));
                    const fabricsData = data?.['fabrics.fabrics'] || data?.fabrics?.fabrics;
                    if (Array.isArray(fabricsData) && fabricsData.length > 0) {
                        return fabricsData.map((fabric) => ({
                            fabricIndex: fabric.fabricIndex || 0,
                            fabricId: fabric.fabricId?.value?.toString() || fabric.fabricId?.toString() || '',
                            nodeId: fabric.nodeId?.value?.toString() || fabric.nodeId?.toString() || '',
                            rootVendorId: fabric.rootVendorId || 0,
                            label: fabric.label || '',
                        }));
                    }
                }
                catch {
                }
            }
            return [];
        }
        catch {
            return [];
        }
    }
    isCommissioned() {
        try {
            const serverNode = this.getServerNode();
            if (serverNode) {
                try {
                    const serverState = serverNode;
                    const commissioned = serverState?.state?.commissioning?.commissioned;
                    if (commissioned === true) {
                        return true;
                    }
                }
                catch {
                }
            }
            const fabrics = this.getFabricInfo();
            return fabrics.length > 0;
        }
        catch (error) {
            log.debug('Failed to check commissioned status:', error);
            return false;
        }
    }
    getCommissionedFabricCount() {
        return this.getFabricInfo().length;
    }
    async removeFabric(fabricIndex) {
        const serverNode = this.getServerNode();
        if (!serverNode) {
            throw new MatterDeviceError('Matter server not started');
        }
        try {
            log.info(`Removing fabric ${fabricIndex}...`);
            const serverState = serverNode;
            const removeFabric = serverState?.state?.commissioning?.removeFabric;
            if (typeof removeFabric !== 'function') {
                throw new MatterDeviceError('Fabric removal not supported by Matter.js version');
            }
            await removeFabric(fabricIndex);
            log.info(`Fabric ${fabricIndex} removed successfully`);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            log.error(`Failed to remove fabric ${fabricIndex}:`, error);
            throw new MatterDeviceError(`Failed to remove fabric: ${errorMessage}`, {
                originalError: error instanceof Error ? error : undefined,
            });
        }
    }
    hasFabric(fabricIndex) {
        const fabrics = this.getFabricInfo();
        return fabrics.some(f => f.fabricIndex === fabricIndex);
    }
}
