// homebridge/dist/matter/server/ServerConfig.js
import { resolve } from 'node:path';
import { sanitizeUniqueId, truncateString, validatePort } from '../configValidator.js';
import { MatterDeviceError } from '../types.js';
export const DEFAULT_MATTER_PORT = 5540;
export const DEFAULT_VENDOR_ID = 0xFFF1;
export const DEFAULT_PRODUCT_ID = 0x8001;
export const MAX_DEVICES_PER_BRIDGE = 1000;
export const SERVER_READY_TIMEOUT_MS = 5000;
export const SERVER_READY_POLL_INTERVAL_MS = 100;
export const SERVER_INIT_DELAY_MS = 200;
export const MAX_PASSCODE_ATTEMPTS = 100;
export function validateAndSanitizeConfig(config) {
    const errors = [];
    const port = config.port || DEFAULT_MATTER_PORT;
    const portValidation = validatePort(port, false);
    if (!portValidation.valid) {
        errors.push(`Invalid port: ${portValidation.error}`);
    }
    if (!config.uniqueId) {
        errors.push('uniqueId is required for Matter server configuration');
    }
    const rawUniqueId = config.uniqueId || '';
    const uniqueIdResult = sanitizeUniqueId(rawUniqueId);
    const uniqueId = uniqueIdResult.value;
    if (uniqueId.length === 0) {
        errors.push('Invalid uniqueId: must be a non-empty string');
    }
    let storagePath = config.storagePath;
    if (storagePath !== undefined) {
        storagePath = resolve(storagePath);
    }
    let manufacturer = config.manufacturer;
    if (manufacturer !== undefined) {
        manufacturer = truncateString(manufacturer, 32, 'Manufacturer name').value;
    }
    let model = config.model;
    if (model !== undefined) {
        model = truncateString(model, 32, 'Model name').value;
    }
    let firmwareRevision = config.firmwareRevision;
    if (firmwareRevision !== undefined) {
        firmwareRevision = truncateString(firmwareRevision, 64, 'Firmware revision').value;
    }
    let serialNumber = config.serialNumber;
    if (serialNumber !== undefined) {
        serialNumber = truncateString(serialNumber, 32, 'Serial number').value;
    }
    const debugModeEnabled = config.debugModeEnabled || false;
    const externalAccessory = config.externalAccessory || false;
    if (errors.length > 0) {
        throw new MatterDeviceError(`Matter configuration validation failed:\n${errors.map(e => `  - ${e}`).join('\n')}`);
    }
    return {
        port,
        uniqueId,
        storagePath,
        displayName: config.displayName,
        manufacturer,
        model,
        firmwareRevision,
        serialNumber,
        debugModeEnabled,
        externalAccessory,
        networkInterfaces: config.networkInterfaces,
    };
}
