// homebridge/dist/matter/server/ServerLifecycle.js
import { constants } from 'node:fs';
import { access, mkdir, rm, stat } from 'node:fs/promises';
import { homedir, release } from 'node:os';
import { join, normalize, resolve } from 'node:path';
import process from 'node:process';
import { Filesystem } from '@matter/general';
import { Endpoint, Environment, ServerNode as MatterServerNode, VendorId, } from '@matter/main';
import { AggregatorEndpoint as AggregatorEndpointType } from '@matter/main/endpoints';
import { NodeJsFilesystem } from '@matter/nodejs';
import { DEFAULT_BRIDGE_DEFAULTS } from '../../bridgeService.js';
import { Logger } from '../../logger.js';
import getVersion from '../../version.js';
import { errorHandler } from '../errorHandler.js';
import { MatterDeviceError } from '../types.js';
import { SERVER_INIT_DELAY_MS, SERVER_READY_POLL_INTERVAL_MS, SERVER_READY_TIMEOUT_MS, } from './ServerConfig.js';
const log = Logger.withPrefix('Matter/Server');
export class ServerLifecycle {
    matterStoragePath;
    async createServerNodeWithRecovery(nodeOptions, sanitizedId) {
        try {
            return await MatterServerNode.create(nodeOptions);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : '';
            const causeMessage = error instanceof Error && error.cause instanceof Error ? error.cause.message : '';
            const isStorageError = errorMessage.includes('Invalid public key encoding')
                || errorMessage.includes('FabricManager unavailable')
                || errorMessage.includes('key-input')
                || causeMessage.includes('Invalid public key encoding');
            if (!isStorageError) {
                throw error;
            }
            log.warn('Detected corrupted Matter storage, attempting automatic recovery...');
            const environment = Environment.default;
            const filesystem = environment.get(Filesystem);
            const storageLocation = filesystem.path;
            if (!storageLocation) {
                throw new Error('Storage location not set, cannot recover from corrupted storage');
            }
            const serverNodeStorePath = join(storageLocation, sanitizedId);
            const serverNodeStoreJsonFile = `${serverNodeStorePath}.json`;
            try {
                let removedSomething = false;
                try {
                    await stat(serverNodeStorePath);
                    log.info(`Removing corrupted ServerNodeStore directory: ${serverNodeStorePath}`);
                    await rm(serverNodeStorePath, { recursive: true, force: true });
                    removedSomething = true;
                }
                catch (err) {
                    const code = err instanceof Error && 'code' in err ? err.code : undefined;
                    if (code !== 'ENOENT') {
                        throw err;
                    }
                }
                try {
                    await stat(serverNodeStoreJsonFile);
                    log.info(`Removing corrupted ServerNodeStore JSON file: ${serverNodeStoreJsonFile}`);
                    await rm(serverNodeStoreJsonFile, { force: true });
                    removedSomething = true;
                }
                catch (err) {
                    const code = err instanceof Error && 'code' in err ? err.code : undefined;
                    if (code !== 'ENOENT') {
                        throw err;
                    }
                }
                if (removedSomething) {
                    log.info('Corrupted storage removed, retrying ServerNode creation...');
                }
                else {
                    log.warn('No corrupted storage files found, corruption may be elsewhere');
                }
                const serverNode = await MatterServerNode.create(nodeOptions);
                log.info('Successfully recovered from corrupted Matter storage');
                return serverNode;
            }
            catch (retryError) {
                log.error('Failed to recover from corrupted storage:', retryError);
                log.error('Original error:', error);
                throw new Error('Matter storage is corrupted and automatic recovery failed. '
                    + `Please manually delete: ${serverNodeStorePath}`);
            }
        }
    }
    async setupStorage(config) {
        if (!config.storagePath) {
            throw new Error('Storage path is required for Matter server');
        }
        const storagePath = resolve(config.storagePath);
        const normalizedPath = normalize(storagePath);
        const allowedBasePaths = [
            resolve(homedir(), '.homebridge'),
            resolve(process.cwd()),
            '/var/lib/homebridge',
        ];
        const isAllowed = allowedBasePaths.some(basePath => normalizedPath.startsWith(basePath));
        if (!isAllowed || normalizedPath.includes('..')) {
            throw new Error(`Storage path not allowed: ${normalizedPath}. Must be within homebridge directories.`);
        }
        try {
            await mkdir(normalizedPath, { recursive: true });
            await access(normalizedPath, constants.R_OK | constants.W_OK);
        }
        catch (error) {
            throw new Error(`Storage path not accessible: ${error}`);
        }
        const bridgeId = config.uniqueId || 'default';
        this.matterStoragePath = join(normalizedPath, bridgeId);
        await mkdir(this.matterStoragePath, { recursive: true });
        const environment = Environment.default;
        environment.set(Filesystem, new NodeJsFilesystem(this.matterStoragePath));
        const { MatterAccessoryCache } = await import('../accessoryCache.js');
        const cache = new MatterAccessoryCache(normalizedPath, bridgeId);
        log.info(`Matter storage initialized at: ${this.matterStoragePath}`);
        return cache;
    }
    async start(deps) {
        if (deps.getIsRunning()) {
            log.warn('Matter server is already running');
            return;
        }
        try {
            log.info('Starting Matter.js server...');
            const cache = await this.setupStorage(deps.config);
            deps.setAccessoryCache(cache);
            await deps.commissioningManager.loadOrGenerateCredentials(this.matterStoragePath);
            log.info(`Configuration: Port=${deps.config.port}, Passcode=${deps.commissioningManager.passcode}, Discriminator=${deps.commissioningManager.discriminator}`);
            if (deps.config.networkInterfaces && deps.config.networkInterfaces.length > 0) {
                const environment = Environment.default;
                const interfaceConfig = {};
                for (const interfaceName of deps.config.networkInterfaces) {
                    interfaceConfig[interfaceName] = { type: 2 };
                }
                environment.vars.set('network.interface', interfaceConfig);
                log.info(`Configured Matter server to use network interfaces: ${deps.config.networkInterfaces.join(', ')}`);
            }
            else {
                log.debug('No network interfaces specified, using all available interfaces');
            }
            const commissioningOptions = {
                passcode: deps.commissioningManager.passcode,
                discriminator: deps.commissioningManager.discriminator,
            };
            log.info(`Using commissioning credentials: passcode=${deps.commissioningManager.passcode}, discriminator=${deps.commissioningManager.discriminator}`);
            const displayName = deps.config.displayName || 'Matter Device';
            const sanitizedId = deps.config.uniqueId;
            const nodeOptions = {
                id: sanitizedId,
                network: {
                    port: deps.config.port,
                    ipv4: true,
                },
                commissioning: commissioningOptions,
                basicInformation: {
                    nodeLabel: displayName.slice(0, 32),
                    vendorId: VendorId(deps.commissioningManager.vendorId),
                    vendorName: DEFAULT_BRIDGE_DEFAULTS.vendorName,
                    productId: deps.commissioningManager.productId,
                    productName: displayName.slice(0, 32),
                    productLabel: displayName.slice(0, 64),
                    serialNumber: deps.config.serialNumber || deps.config.uniqueId,
                    hardwareVersion: 1,
                    hardwareVersionString: release(),
                    softwareVersion: 1,
                    softwareVersionString: deps.config.firmwareRevision || getVersion(),
                    reachable: true,
                },
            };
            if (!deps.config.externalAccessory) {
                nodeOptions.productDescription = {
                    name: displayName,
                    deviceType: AggregatorEndpointType.deviceType,
                };
            }
            const serverNode = await this.createServerNodeWithRecovery(nodeOptions, sanitizedId);
            deps.setServerNode(serverNode);
            deps.commissioningManager.setupCommissioningEventListeners(deps.getCommissioningDeps());
            if (!deps.config.externalAccessory) {
                const aggregator = new Endpoint(AggregatorEndpointType, {
                    id: 'homebridge-aggregator',
                });
                await serverNode.add(aggregator);
                deps.setAggregator(aggregator);
                log.debug('Created aggregator endpoint for bridged mode');
            }
            else {
                log.debug('External accessory mode - skipping aggregator creation');
            }
            await deps.commissioningManager.generateCommissioningInfo(deps.getCommissioningDeps());
            const shutdownHandler = async () => {
                log.info('Shutting down Matter server...');
                await deps.onStop();
            };
            deps.setShutdownHandler(shutdownHandler);
            process.on('SIGINT', shutdownHandler);
            process.on('SIGTERM', shutdownHandler);
            if (!deps.config.externalAccessory) {
                await this.startServerNode(serverNode, deps);
            }
            else {
                log.debug('Deferred start mode - server prepared but not running yet (will start after device registration)');
            }
            log.info(`Matter server started successfully on port ${deps.config.port}`);
            log.info('Plugins can now register Matter accessories via the API');
        }
        catch (error) {
            log.error('Failed to start Matter server:', error);
            await this.cleanup(deps);
            throw error;
        }
    }
    async runServer(deps) {
        const serverNode = deps.getServerNode();
        if (!serverNode) {
            throw new MatterDeviceError('Server node not initialized - call start() first');
        }
        if (deps.getIsRunning()) {
            log.warn('Matter server is already running');
            return;
        }
        if (!deps.config.externalAccessory) {
            throw new MatterDeviceError('runServer() should only be called when externalAccessory mode is enabled');
        }
        log.debug('Running deferred server with device(s) already attached');
        await this.startServerNode(serverNode, deps);
        log.info('Matter server is now running');
    }
    async startServerNode(serverNode, deps) {
        serverNode.run().then(() => {
            log.info('Matter server stopped normally');
        }, (error) => {
            log.error('Matter server stopped with error:', error);
            errorHandler.handleError(error, 'server-runtime');
        });
        await this.waitForServerReady(deps);
        const cache = deps.getAccessoryCache();
        if (cache) {
            const loaded = await cache.load();
            log.debug(`Matter cache loaded: ${loaded.size} accessories`);
        }
        else {
            log.debug('No accessory cache available');
        }
        deps.commissioningManager.updateCommissioningFile(deps.getCommissioningDeps()).catch((error) => {
            log.warn('Failed to update commissioning file on startup:', error);
        });
        deps.setIsRunning(true);
    }
    async waitForServerReady(deps, maxWaitTime = SERVER_READY_TIMEOUT_MS) {
        const startTime = Date.now();
        while (!deps.getServerNode() || (!deps.config.externalAccessory && !deps.getAggregator())) {
            if (Date.now() - startTime > maxWaitTime) {
                throw new Error('Server failed to become ready within timeout');
            }
            await new Promise(resolve => setTimeout(resolve, SERVER_READY_POLL_INTERVAL_MS));
        }
        await new Promise(resolve => setTimeout(resolve, SERVER_INIT_DELAY_MS));
    }
    async stop(deps, accessories) {
        if (!deps.getIsRunning()) {
            log.debug('Matter server is not running');
            return;
        }
        deps.setIsRunning(false);
        try {
            const cache = deps.getAccessoryCache();
            if (cache && accessories.size > 0) {
                await cache.save(accessories);
                log.debug('Saved accessory cache before shutdown');
            }
            const serverNode = deps.getServerNode();
            if (serverNode) {
                await serverNode.close();
                log.debug('ServerNode closed (all endpoints cleaned up)');
            }
            accessories.clear();
            await this.cleanup(deps);
            log.info('Matter server stopped');
        }
        catch (error) {
            log.error('Error stopping Matter server:', error);
            await errorHandler.handleError(error, 'server-stop');
            throw error;
        }
        finally {
            deps.setIsRunning(false);
        }
    }
    async cleanup(deps) {
        if (deps.shutdownHandler) {
            process.off('SIGINT', deps.shutdownHandler);
            process.off('SIGTERM', deps.shutdownHandler);
            deps.setShutdownHandler(null);
        }
        for (const handler of deps.cleanupHandlers) {
            try {
                await handler();
            }
            catch (error) {
                log.debug('Error during cleanup handler:', error);
            }
        }
        deps.cleanupHandlers.length = 0;
        deps.setServerNode(null);
        deps.setAggregator(null);
        deps.setIsRunning(false);
    }
}
