// @matter/general/dist/esm/storage/wal/WalCommit.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { fromJson, toJson } from "../StringifyTools.js";
const MAX_SEGMENT_LINES = 65535;
function encodeContextKey(contexts) {
  return contexts.map((s) => s.replace(/%/g, "%25").replace(/\//g, "%2F")).join("/");
}
function decodeContextKey(key) {
  return key.split("/").map((s) => s.replace(/%2F/gi, "/").replace(/%25/g, "%"));
}
function serializeCommit(commit) {
  return toJson({ ts: commit.ts, ops: commit.ops });
}
function deserializeCommit(line) {
  const parsed = fromJson(line);
  if (Array.isArray(parsed)) {
    return { ts: 0, ops: parsed };
  }
  return parsed;
}
function segmentFilename(segment) {
  return segment.toString(16).padStart(8, "0") + ".jsonl";
}
function compressedSegmentFilename(segment) {
  return segment.toString(16).padStart(8, "0") + ".jsonl.gz";
}
function isCompressedSegmentFile(name) {
  return /^[0-9a-f]{8}\.jsonl\.gz$/i.test(name);
}
function parseSegmentFilename(name) {
  const match = name.match(/^([0-9a-f]{8})\.jsonl(\.gz)?$/i);
  if (!match) {
    return void 0;
  }
  return parseInt(match[1], 16);
}
function commitIdToNumber(id) {
  return id.segment * 65536 + id.offset;
}
function compareCommitIds(a, b) {
  if (a.segment !== b.segment) {
    return a.segment - b.segment;
  }
  return a.offset - b.offset;
}
export {
  MAX_SEGMENT_LINES,
  commitIdToNumber,
  compareCommitIds,
  compressedSegmentFilename,
  decodeContextKey,
  deserializeCommit,
  encodeContextKey,
  isCompressedSegmentFile,
  parseSegmentFilename,
  segmentFilename,
  serializeCommit
};
