// @matter/general/dist/esm/storage/wal/WalCommit.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { StorageError } from "../StorageDriver.js";
import { fromJson, toJson } from "../StringifyTools.js";
const MAX_SEGMENT_LINES = 65535;
function encodeContextKey(contexts) {
  return contexts.map((s) => s.replace(/%/g, "%25").replace(/\//g, "%2F")).join("/");
}
function decodeContextKey(key) {
  return key.split("/").map((s) => s.replace(/%2F/gi, "/").replace(/%25/g, "%"));
}
function serializeCommit(commit) {
  return toJson({ ts: commit.ts, ops: commit.ops });
}
function deserializeCommit(line) {
  const parsed = fromJson(line);
  if (Array.isArray(parsed)) {
    parsed.forEach(validateOp);
    return { ts: 0, ops: parsed };
  }
  if (!isPlainRecord(parsed)) {
    throw new StorageError("WAL commit is not an object");
  }
  const { ts, ops } = parsed;
  if (typeof ts !== "number") {
    throw new StorageError("WAL commit `ts` is missing or not a number");
  }
  if (!Array.isArray(ops)) {
    throw new StorageError("WAL commit `ops` is missing or not an array");
  }
  ops.forEach(validateOp);
  return { ts, ops };
}
function isPlainRecord(v) {
  return v !== null && typeof v === "object" && !Array.isArray(v);
}
function validateOp(op, index) {
  if (!isPlainRecord(op)) {
    throw new StorageError(`WAL op[${index}] is not an object`);
  }
  const { op: kind, key, values } = op;
  if (typeof key !== "string") {
    throw new StorageError(`WAL op[${index}] \`key\` is not a string`);
  }
  if (kind === "upd") {
    if (!isPlainRecord(values)) {
      throw new StorageError(`WAL op[${index}] upd \`values\` is not an object`);
    }
    return;
  }
  if (kind === "del") {
    if (values === void 0) {
      return;
    }
    if (!Array.isArray(values) || !values.every((v) => typeof v === "string")) {
      throw new StorageError(`WAL op[${index}] del \`values\` is not a string array`);
    }
    return;
  }
  throw new StorageError(`WAL op[${index}] has unknown op kind \`${String(kind)}\``);
}
function segmentFilename(segment) {
  return segment.toString(16).padStart(8, "0") + ".jsonl";
}
function compressedSegmentFilename(segment) {
  return segment.toString(16).padStart(8, "0") + ".jsonl.gz";
}
function isCompressedSegmentFile(name) {
  return /^[0-9a-f]{8}\.jsonl\.gz$/i.test(name);
}
function parseSegmentFilename(name) {
  const match = name.match(/^([0-9a-f]{8})\.jsonl(\.gz)?$/i);
  if (!match) {
    return void 0;
  }
  return parseInt(match[1], 16);
}
function commitIdToNumber(id) {
  return id.segment * 65536 + id.offset;
}
function compareCommitIds(a, b) {
  if (a.segment !== b.segment) {
    return a.segment - b.segment;
  }
  return a.offset - b.offset;
}
export {
  MAX_SEGMENT_LINES,
  commitIdToNumber,
  compareCommitIds,
  compressedSegmentFilename,
  decodeContextKey,
  deserializeCommit,
  encodeContextKey,
  isCompressedSegmentFile,
  parseSegmentFilename,
  segmentFilename,
  serializeCommit
};
