// @matter/general/dist/cjs/crypto/StandardCrypto.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StandardCrypto_exports = {};
__export(StandardCrypto_exports, {
  StandardCrypto: () => StandardCrypto
});
module.exports = __toCommonJS(StandardCrypto_exports);
var import_Environment = require("#environment/Environment.js");
var import_MatterError = require("#MatterError.js");
var import_Bytes = require("#util/Bytes.js");
var import_Entropy = require("#util/Entropy.js");
var import_String = require("#util/String.js");
var import_Logger = require("../log/Logger.js");
var import_Ccm = require("./aes/Ccm.js");
var import_Crypto = require("./Crypto.js");
var import_CryptoError = require("./CryptoError.js");
var import_EcdsaSignature = require("./EcdsaSignature.js");
var import_Key = require("./Key.js");
/**
 * @license
 *
 * Portions copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("StandardCrypto");
const SIGNATURE_ALGORITHM = {
  name: "ECDSA",
  namedCurve: "P-256",
  hash: { name: "SHA-256" }
};
const requiredCryptoMethods = ["getRandomValues"];
const requiredSubtleMethods = [
  "digest",
  "deriveBits",
  "sign",
  "verify",
  "generateKey",
  "exportKey",
  "importKey"
];
class StandardCrypto extends import_Crypto.Crypto {
  implementationName = "JS";
  #crypto;
  #subtle;
  constructor(crypto = globalThis.crypto) {
    const { subtle } = crypto;
    assertInterface("crypto", crypto, requiredCryptoMethods);
    assertInterface("crypto.subtle", subtle, requiredSubtleMethods);
    super();
    this.#crypto = crypto;
    this.#subtle = subtle;
  }
  get subtle() {
    return this.#subtle;
  }
  static provider() {
    return new StandardCrypto();
  }
  randomBytes(length) {
    const result = new Uint8Array(length);
    this.#crypto.getRandomValues(result);
    return result;
  }
  encrypt(key, data, nonce, associatedData) {
    const ccm = (0, import_Ccm.Ccm)(key);
    return ccm.encrypt({
      pt: import_Bytes.Bytes.of(data),
      nonce: import_Bytes.Bytes.of(nonce),
      adata: associatedData !== void 0 ? import_Bytes.Bytes.of(associatedData) : void 0
    });
  }
  decrypt(key, data, nonce, associatedData) {
    const ccm = (0, import_Ccm.Ccm)(key);
    return ccm.decrypt({
      ct: import_Bytes.Bytes.of(data),
      nonce: import_Bytes.Bytes.of(nonce),
      adata: associatedData !== void 0 ? import_Bytes.Bytes.of(associatedData) : void 0
    });
  }
  computeHash(buffer, algorithm = "SHA-256") {
    if (Array.isArray(buffer)) {
      buffer = import_Bytes.Bytes.concat(...buffer);
    }
    if (!import_Bytes.Bytes.isBytes(buffer)) {
      const chunks = [];
      const iterator = Symbol.asyncIterator in buffer ? buffer[Symbol.asyncIterator]() : buffer;
      const collectAndHash = async () => {
        while (true) {
          const result = await iterator.next();
          if (result.done) break;
          const chunk = result.value instanceof Uint8Array ? result.value : new Uint8Array(result.value);
          chunks.push(chunk);
        }
        const combined = import_Bytes.Bytes.concat(...chunks);
        if (combined.byteLength > 1e5) {
          logger.info(
            `Streamed hash computation used with StandardCrypto for ${algorithm} and ${Math.floor(combined.byteLength / 1024)}kB. Consider alternatives that do not load all data into memory.`
          );
        }
        return await this.#subtle.digest(algorithm, import_Bytes.Bytes.exclusive(combined));
      };
      return collectAndHash();
    }
    return this.#subtle.digest(algorithm, import_Bytes.Bytes.exclusive(buffer));
  }
  async createPbkdf2Key(secret, salt, iteration, keyLength) {
    const key = await this.importKey("raw", secret, "PBKDF2", false, ["deriveBits"]);
    return this.#subtle.deriveBits(
      {
        name: "PBKDF2",
        hash: "SHA-256",
        salt: import_Bytes.Bytes.exclusive(salt),
        iterations: iteration
      },
      key,
      keyLength * 8
    );
  }
  async createHkdfKey(secret, salt, info, length = import_Crypto.CRYPTO_SYMMETRIC_KEY_LENGTH) {
    const key = await this.importKey("raw", secret, "HKDF", false, ["deriveBits"]);
    return this.#subtle.deriveBits(
      {
        name: "HKDF",
        hash: "SHA-256",
        salt: import_Bytes.Bytes.exclusive(salt),
        info: import_Bytes.Bytes.exclusive(info)
      },
      key,
      8 * length
    );
  }
  signHmac(secret, data) {
    return this.importKey("raw", secret, { name: "HMAC", hash: "SHA-256" }, false, ["sign"]).then(
      (key) => this.#subtle.sign("HMAC", key, import_Bytes.Bytes.exclusive(data))
    );
  }
  async signEcdsa(key, data) {
    if (Array.isArray(data)) {
      data = import_Bytes.Bytes.concat(...data);
    }
    const { crv, kty, d, x, y } = key;
    key = {
      kty,
      crv,
      d,
      x,
      y,
      ext: true,
      // Required by some subtle implementations to sign
      key_ops: ["sign"]
    };
    const subtleKey = await this.importKey("jwk", key, SIGNATURE_ALGORITHM, false, ["sign"]);
    const ieeeP1363 = import_Bytes.Bytes.of(await this.#subtle.sign(SIGNATURE_ALGORITHM, subtleKey, import_Bytes.Bytes.exclusive(data)));
    return new import_EcdsaSignature.EcdsaSignature(ieeeP1363);
  }
  async verifyEcdsa(key, data, signature) {
    const { crv, kty, x, y } = key;
    key = { crv, kty, x, y };
    const subtleKey = await this.importKey("jwk", key, SIGNATURE_ALGORITHM, false, ["verify"]);
    const verified = await this.#subtle.verify(
      SIGNATURE_ALGORITHM,
      subtleKey,
      import_Bytes.Bytes.exclusive(signature.bytes),
      import_Bytes.Bytes.exclusive(data)
    );
    if (!verified) {
      throw new import_CryptoError.CryptoVerifyError("Signature verification failed");
    }
  }
  async createKeyPair() {
    const key = await this.generateJwk();
    return (0, import_Key.Key)({
      kty: import_Key.KeyType.EC,
      crv: import_Key.CurveType.p256,
      d: key.d,
      x: key.x,
      y: key.y
    });
  }
  async generateJwk() {
    const subtleKey = await this.#subtle.generateKey(
      {
        // We must specify either ECDH or ECDSA to get an EC key but we may use the key for either (but not for
        // both)
        name: "ECDH",
        namedCurve: "P-256"
      },
      true,
      // We must also specify usage but will drop this on export
      ["deriveKey"]
    );
    return await this.#subtle.exportKey("jwk", subtleKey.privateKey);
  }
  async generateDhSecret(key, peerKey) {
    const subtleKey = await this.importKey(
      "jwk",
      key,
      {
        name: "ECDH",
        namedCurve: "P-256"
      },
      false,
      ["deriveBits"]
    );
    const subtlePeerKey = await this.importKey(
      "jwk",
      peerKey,
      {
        name: "ECDH",
        namedCurve: "P-256"
      },
      false,
      []
    );
    return this.#subtle.deriveBits(
      {
        name: "ECDH",
        public: subtlePeerKey
      },
      subtleKey,
      256
    );
  }
  ecMultiply(point, scalar) {
    return import_Crypto.ec.p256.Point.fromBytes(import_Bytes.Bytes.of(point)).multiply(import_Bytes.Bytes.asBigInt(import_Bytes.Bytes.of(scalar))).toBytes(false);
  }
  async importKey(format, keyData, algorithm, extractable, keyUsages) {
    try {
      return await this.#subtle.importKey(format, keyData, algorithm, extractable, keyUsages);
    } catch (cause) {
      throw new import_CryptoError.KeyInputError("Invalid key", { cause });
    }
  }
}
function assertInterface(name, object, requiredMethods) {
  if (typeof object !== "object" || object === null) {
    throw new import_MatterError.ImplementationError(
      `The ${name} implementation passed to StandardCrypto is invalid (received ${typeof object})`
    );
  }
  const missingMethods = requiredMethods.filter((name2) => typeof object[name2] !== "function");
  if (missingMethods.length) {
    throw new import_MatterError.ImplementationError(
      `The ${name} implementation passed to StandardCrypto is missing required method${missingMethods.length === 1 ? "" : "s"} ${(0, import_String.describeList)("and", ...missingMethods)}`
    );
  }
}
if ("crypto" in globalThis && globalThis.crypto?.subtle && !import_Environment.Environment.default.has(import_Crypto.Crypto)) {
  const crypto = new StandardCrypto();
  import_Environment.Environment.default.set(import_Entropy.Entropy, crypto);
  import_Environment.Environment.default.set(import_Crypto.Crypto, crypto);
}
