// @matter/general/dist/esm/net/http/HttpSharedEndpoint.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "#log/Logger.js";
const logger = Logger.get("HttpSharedEntpoint");
class HttpSharedEndpoint {
  #create;
  #target;
  #instances = /* @__PURE__ */ new Set();
  #isTls;
  constructor(isTls, create) {
    this.#isTls = isTls;
    this.#create = create;
  }
  get isTls() {
    return this.#isTls;
  }
  async use() {
    if (this.#target === void 0) {
      this.#target = this.#createTarget();
    }
    await this.#target;
    const instance = {
      http: void 0,
      ws: void 0,
      close: async () => {
        this.#instances.delete(instance);
        if (!this.#instances.size) {
          try {
            await (await this.#target)?.close();
          } catch (e) {
            logger.error("Error closing HTTP endpoint", e);
          } finally {
            this.#target = void 0;
          }
        }
      }
    };
    this.#instances.add(instance);
    return instance;
  }
  async #createTarget() {
    const target = await this.#create();
    target.http = async (request) => {
      for (const instance of this.#instances) {
        const response = await instance.http?.(request);
        if (response) {
          return response;
        }
      }
    };
    target.ws = async (request, connect) => {
      let connected = false;
      for (const instance of this.#instances) {
        await instance.ws?.(request, async () => {
          connected = true;
          return connect();
        });
        if (connected) {
          break;
        }
      }
    };
    return target;
  }
}
export {
  HttpSharedEndpoint
};
