// @matter/general/dist/esm/net/http/MockHttpEndpointFactory.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { HttpEndpointFactory } from "./HttpEndpointFactory.js";
import { MockWsConnection } from "./MockWsConnection.js";
class MockHttpEndpointFactory extends HttpEndpointFactory {
  #target;
  async create(_options) {
    const endpoint = {
      http: void 0,
      ws: void 0,
      async close() {
      }
    };
    this.#target = endpoint;
    return endpoint;
  }
  /**
   * Create a new mock WebSocket connection pair.  Invokes the captured WS handler with the server side and returns
   * the client side.
   */
  async connect() {
    const target = this.#target;
    if (!target?.ws) {
      throw new Error("No WebSocket handler registered; is WebSocketServer started?");
    }
    const pair = MockWsConnection();
    target.ws(new Request(MockHttpEndpointFactory.WS_URL), () => Promise.resolve(pair.server));
    return pair.client;
  }
}
((MockHttpEndpointFactory2) => {
  MockHttpEndpointFactory2.WS_URL = "ws+unix://matter.sock/";
})(MockHttpEndpointFactory || (MockHttpEndpointFactory = {}));
export {
  MockHttpEndpointFactory
};
