// @matter/general/dist/esm/net/http/WebSocketClient.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Environmental } from "#environment/Environmental.js";
import { NoProviderError } from "#MatterError.js";
import { Gate } from "#util/Promises.js";
class WebSocketClient {
  #Ctor;
  constructor(Ctor) {
    this.#Ctor = Ctor;
  }
  /**
   * Open a WebSocket connection and return it as a {@link HttpEndpoint.WsConnection}.
   */
  async connect(url) {
    const Ctor = this.#Ctor ?? globalThis.WebSocket;
    if (!Ctor) {
      throw new NoProviderError("No WebSocket implementation available");
    }
    const ws = new Ctor(url);
    ws.binaryType = "arraybuffer";
    await new Promise((resolve, reject) => {
      ws.addEventListener("open", onOpen);
      ws.addEventListener("error", onError);
      function off() {
        ws.removeEventListener("open", onOpen);
        ws.removeEventListener("error", onError);
      }
      function onOpen() {
        off();
        resolve();
      }
      function onError() {
        off();
        reject(new Error(`WebSocket connection to ${url} failed`));
      }
    });
    return WebSocketClient.streams(ws);
  }
  static [Environmental.create](env) {
    const instance = new WebSocketClient();
    env.set(WebSocketClient, instance);
    return instance;
  }
  /**
   * Wrap an open standard WebSocket into a {@link HttpEndpoint.WsConnection}.
   */
  static streams(ws) {
    return {
      readable: createReadable(ws),
      writable: createWritable(ws)
    };
  }
}
function createReadable(ws) {
  const queue = Array();
  const ready = new Gate();
  let isClosed = false;
  ws.addEventListener("message", onMessage);
  ws.addEventListener("error", onError);
  ws.addEventListener("close", onClose);
  return new ReadableStream({
    async pull(controller) {
      if (!isClosed && !queue.length) {
        await ready;
      }
      const next = queue.shift();
      if (!queue.length && !isClosed) {
        ready.close();
      }
      if (next === void 0) {
        controller.close();
        return;
      }
      if (next instanceof Error) {
        controller.error(next);
        return;
      }
      controller.enqueue(next);
    },
    cancel() {
      if (!isClosed) {
        ws.close();
        isClosed = true;
        queue.length = 0;
      }
    }
  });
  function onMessage(event) {
    let { data } = event;
    if (data instanceof ArrayBuffer) {
      data = new Uint8Array(data);
    }
    queue.push(data);
    ready.open();
  }
  function onError() {
    isClosed = true;
    queue.push(new Error("WebSocket error"));
    ready.open();
  }
  function onClose() {
    isClosed = true;
    ready.open();
  }
}
function createWritable(ws) {
  return new WritableStream({
    write(chunk) {
      ws.send(chunk);
    },
    close() {
      if (ws.readyState < 2) {
        ws.close(1e3);
      }
    },
    abort() {
      if (ws.readyState < 2) {
        ws.close();
      }
    }
  });
}
export {
  WebSocketClient
};
