// @matter/general/dist/cjs/net/mock/MockUdpChannel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MockUdpChannel_exports = {};
__export(MockUdpChannel_exports, {
  MockUdpChannel: () => MockUdpChannel
});
module.exports = __toCommonJS(MockUdpChannel_exports);
var import_Channel = require("#net/Channel.js");
var import_Time = require("#time/Time.js");
var import_Ip = require("#util/Ip.js");
var import_UdpChannel = require("../udp/UdpChannel.js");
var import_MockRouter = require("./MockRouter.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MockUdpChannel {
  #host;
  #router;
  #listeningIp;
  #listeningPort;
  #type;
  maxPayloadSize = import_UdpChannel.MAX_UDP_MESSAGE_SIZE;
  constructor(network, { listeningAddress, listeningPort, type }, interceptor) {
    this.#router = (0, import_MockRouter.MockRouter)();
    if (interceptor) {
      this.#router.intercept(interceptor);
    }
    if (listeningAddress !== "*") {
      this.#listeningIp = listeningAddress;
    }
    this.#type = type;
    this.#host = network;
    this.#listeningPort = listeningPort ?? 1024 + Math.floor(Math.random() * 64511);
    network.router.add(this.#router);
  }
  onData(listener) {
    const router = (packet) => {
      if (packet.kind !== "udp") {
        return;
      }
      if (!this.#host.shouldReceive(packet.destAddress)) {
        return;
      }
      if (this.#listeningIp && packet.destAddress !== this.#listeningIp && !this.#host.isMemberOf(packet.destAddress)) {
        return;
      }
      switch (this.#type) {
        case "udp4":
          if (!(0, import_Ip.isIPv4)(packet.destAddress)) {
            return;
          }
          break;
        case "udp6":
          if (!(0, import_Ip.isIPv6)(packet.destAddress)) {
            return;
          }
          break;
      }
      if (packet.destPort !== this.#listeningPort) {
        return;
      }
      listener("fake0", packet.sourceAddress, packet.sourcePort, packet.payload);
    };
    this.#router.add(router);
    return {
      close: async () => {
        this.#router.delete(router);
      }
    };
  }
  async send(host, port, payload) {
    await import_Time.Time.macrotask;
    this.#host.simulator.router({
      kind: "udp",
      sourceAddress: this.#host.defaultRoute,
      sourcePort: this.#listeningPort,
      destAddress: host,
      destPort: port,
      payload
    });
  }
  async close() {
    this.#host.router.delete(this.#router);
  }
  async [Symbol.asyncDispose]() {
    return this.close();
  }
  get port() {
    return this.#listeningPort;
  }
  supports(type, _address) {
    return type === import_Channel.ChannelType.UDP;
  }
  addMembership(address) {
    this.#host.addMembership(address);
  }
  dropMembership(address) {
    this.#host.dropMembership(address);
  }
}
