// @matter/general/dist/cjs/net/mock/MockNetwork.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MockNetwork_exports = {};
__export(MockNetwork_exports, {
  MockNetwork: () => MockNetwork
});
module.exports = __toCommonJS(MockNetwork_exports);
var import_Channel = require("#net/Channel.js");
var import_Ip = require("../../util/Ip.js");
var import_Network = require("../Network.js");
var import_MockRouter = require("./MockRouter.js");
var import_MockTcpListener = require("./MockTcpListener.js");
var import_MockUdpSocket = require("./MockUdpSocket.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MockNetwork extends import_Network.Network {
  router = (0, import_MockRouter.MockRouter)();
  #simulator;
  #intf;
  #defaultRoute;
  #ips;
  #multicastIps = /* @__PURE__ */ new Set();
  #tcpServers = /* @__PURE__ */ new Map();
  constructor(simulator, mac, ips) {
    super();
    this.#simulator = simulator;
    this.#ips = new Set(ips);
    const self = this;
    this.#intf = {
      mac,
      get ipV4() {
        return [...self.#ips].filter((ip) => (0, import_Ip.isIPv4)(ip));
      },
      get ipV6() {
        return [...self.#ips].filter((ip) => !(0, import_Ip.isIPv4)(ip));
      }
    };
    this.#simulator.router.add(this.router);
  }
  get simulator() {
    return this.#simulator;
  }
  addAddr(...ips) {
    for (const ip of ips) {
      this.#ips.add(ip);
    }
  }
  deleteAddr(...ips) {
    for (const ip of ips) {
      this.#ips.delete(ip);
    }
  }
  addMembership(...multicastIps) {
    for (const ip of multicastIps) {
      this.#multicastIps.add(ip);
    }
  }
  dropMembership(...multicastIps) {
    for (const ip of multicastIps) {
      this.#multicastIps.delete(ip);
    }
  }
  isMemberOf(multicastIp) {
    return this.#multicastIps.has(multicastIp);
  }
  shouldReceive(ip) {
    return this.#ips.has(ip) || this.#multicastIps.has(ip);
  }
  set defaultRoute(ip) {
    this.#defaultRoute = ip;
  }
  get defaultRoute() {
    const ip = this.#defaultRoute ?? this.#ips[Symbol.iterator]().next().value;
    if (ip === void 0) {
      throw new Error("No default route");
    }
    return ip;
  }
  getNetInterfaces() {
    return [{ name: "fake0" }];
  }
  getIpMac(name) {
    if (name !== "fake0") {
      throw new import_Network.NetworkError(`No such interface ${name}`);
    }
    return this.#intf;
  }
  createUdpSocket(options) {
    return Promise.resolve(new import_MockUdpSocket.MockUdpSocket(this, options));
  }
  createTcpListener(options) {
    return Promise.resolve(new import_MockTcpListener.MockTcpListener(this, options));
  }
  async connectTcp(host, port) {
    const targetNetwork = this.#simulator.findNetwork(host);
    if (!targetNetwork) {
      throw new import_Network.NetworkError(`No mock network hosts address ${host}`);
    }
    const server = targetNetwork.#tcpServers.get(port);
    if (!server) {
      throw new import_Network.NetworkError(`No TCP server listening on ${host}:${port}`);
    }
    const clientPort = 1024 + Math.floor(Math.random() * 64511);
    return server.accept(this.defaultRoute, clientPort);
  }
  registerTcpListener(server) {
    this.#tcpServers.set(server.port, server);
  }
  unregisterTcpListener(port) {
    this.#tcpServers.delete(port);
  }
  supports(type, _address) {
    return type === import_Channel.ChannelType.UDP || type === import_Channel.ChannelType.TCP;
  }
  async close() {
    this.#simulator.router.delete(this.router);
  }
}
