// @matter/general/dist/esm/net/mock/MockNetwork.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ChannelType } from "#net/Channel.js";
import { isIPv4 } from "../../util/Ip.js";
import { Network, NetworkError } from "../Network.js";
import { MockRouter } from "./MockRouter.js";
import { MockTcpListener } from "./MockTcpListener.js";
import { MockUdpSocket } from "./MockUdpSocket.js";
class MockNetwork extends Network {
  router = MockRouter();
  #simulator;
  #intf;
  #defaultRoute;
  #ips;
  #multicastIps = /* @__PURE__ */ new Set();
  #tcpServers = /* @__PURE__ */ new Map();
  constructor(simulator, mac, ips) {
    super();
    this.#simulator = simulator;
    this.#ips = new Set(ips);
    const self = this;
    this.#intf = {
      mac,
      get ipV4() {
        return [...self.#ips].filter((ip) => isIPv4(ip));
      },
      get ipV6() {
        return [...self.#ips].filter((ip) => !isIPv4(ip));
      }
    };
    this.#simulator.router.add(this.router);
  }
  get simulator() {
    return this.#simulator;
  }
  addAddr(...ips) {
    for (const ip of ips) {
      this.#ips.add(ip);
    }
  }
  deleteAddr(...ips) {
    for (const ip of ips) {
      this.#ips.delete(ip);
    }
  }
  addMembership(...multicastIps) {
    for (const ip of multicastIps) {
      this.#multicastIps.add(ip);
    }
  }
  dropMembership(...multicastIps) {
    for (const ip of multicastIps) {
      this.#multicastIps.delete(ip);
    }
  }
  isMemberOf(multicastIp) {
    return this.#multicastIps.has(multicastIp);
  }
  shouldReceive(ip) {
    return this.#ips.has(ip) || this.#multicastIps.has(ip);
  }
  set defaultRoute(ip) {
    this.#defaultRoute = ip;
  }
  get defaultRoute() {
    const ip = this.#defaultRoute ?? this.#ips[Symbol.iterator]().next().value;
    if (ip === void 0) {
      throw new Error("No default route");
    }
    return ip;
  }
  getNetInterfaces() {
    return [{ name: "fake0" }];
  }
  getIpMac(name) {
    if (name !== "fake0") {
      throw new NetworkError(`No such interface ${name}`);
    }
    return this.#intf;
  }
  createUdpSocket(options) {
    return Promise.resolve(new MockUdpSocket(this, options));
  }
  createTcpListener(options) {
    return Promise.resolve(new MockTcpListener(this, options));
  }
  async connectTcp(host, port) {
    const targetNetwork = this.#simulator.findNetwork(host);
    if (!targetNetwork) {
      throw new NetworkError(`No mock network hosts address ${host}`);
    }
    const server = targetNetwork.#tcpServers.get(port);
    if (!server) {
      throw new NetworkError(`No TCP server listening on ${host}:${port}`);
    }
    const clientPort = 1024 + Math.floor(Math.random() * 64511);
    return server.accept(this.defaultRoute, clientPort);
  }
  registerTcpListener(server) {
    this.#tcpServers.set(server.port, server);
  }
  unregisterTcpListener(port) {
    this.#tcpServers.delete(port);
  }
  supports(type, _address) {
    return type === ChannelType.UDP || type === ChannelType.TCP;
  }
  async close() {
    this.#simulator.router.delete(this.router);
  }
}
export {
  MockNetwork
};
