// @matter/model/dist/esm/logic/ModelIndex.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Callable, camelize, ImplementationError } from "@matter/general";
class MutableModelIndex extends Callable {
  #items;
  #nameIndex;
  #idIndex;
  constructor(items = []) {
    let self;
    super(function(key) {
      return self.for(key);
    });
    self = this;
    delete this.length;
    this.#items = items;
  }
  /**
   * Add items.  Used during construction only.
   */
  push(...items) {
    this.#items.push(...items);
    this.#nameIndex = void 0;
    this.#idIndex = void 0;
  }
  get length() {
    return this.#items.length;
  }
  at(index) {
    return this.#items.at(index);
  }
  [Symbol.iterator]() {
    return this.#items[Symbol.iterator]();
  }
  forEach(fn) {
    this.#items.forEach(fn);
  }
  map(fn) {
    return this.#items.map(fn);
  }
  filter(fn) {
    return this.#items.filter(fn);
  }
  find(fn) {
    return this.#items.find(fn);
  }
  some(fn) {
    return this.#items.some(fn);
  }
  every(fn) {
    return this.#items.every(fn);
  }
  reduce(fn, initialValue) {
    return this.#items.reduce(fn, initialValue);
  }
  for(key, type) {
    if (key === void 0) {
      return void 0;
    }
    let untyped;
    if (typeof key === "number") {
      untyped = this.#ids.get(key);
    } else {
      untyped = this.#names.get(camelize(key));
    }
    if (untyped === void 0) {
      return void 0;
    }
    if (Array.isArray(untyped)) {
      if (type) {
        return untyped.find((m) => m instanceof type);
      }
      return untyped[0];
    }
    if (type && !(untyped instanceof type)) {
      return void 0;
    }
    return untyped;
  }
  require(key, type) {
    if (key === void 0) {
      return void 0;
    }
    const result = this.for(key, type);
    if (result === void 0) {
      throw new ImplementationError(`Required member "${key}" not found`);
    }
    return result;
  }
  extend(key, properties, ...children) {
    return this.require(key).extend(properties, ...children);
  }
  ofType(type) {
    return this.#items.filter((m) => m instanceof type);
  }
  get #ids() {
    if (this.#idIndex) {
      return this.#idIndex;
    }
    const index = this.#idIndex = /* @__PURE__ */ new Map();
    for (const model of this.#items) {
      if (model.id === void 0) {
        continue;
      }
      const already = index.get(model.id);
      if (already) {
        if (Array.isArray(already)) {
          already.push(model);
        } else {
          index.set(model.id, [already, model]);
        }
      } else {
        index.set(model.id, model);
      }
    }
    return this.#idIndex;
  }
  get #names() {
    if (this.#nameIndex) {
      return this.#nameIndex;
    }
    const index = this.#nameIndex = /* @__PURE__ */ new Map();
    for (const model of this.#items) {
      const name = model.propertyName;
      const already = index.get(name);
      if (already) {
        if (Array.isArray(already)) {
          already.push(model);
        } else {
          index.set(name, [already, model]);
        }
      } else {
        index.set(name, model);
      }
    }
    return this.#nameIndex;
  }
}
export {
  MutableModelIndex
};
