// @matter/general/dist/cjs/log/DiagnosticPresentation.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DiagnosticPresentation_exports = {};
__export(DiagnosticPresentation_exports, {
  DiagnosticPresentation: () => DiagnosticPresentation
});
module.exports = __toCommonJS(DiagnosticPresentation_exports);
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DiagnosticPresentation = /* @__PURE__ */ ((DiagnosticPresentation2) => {
  DiagnosticPresentation2["Message"] = "message";
  DiagnosticPresentation2["List"] = "list";
  DiagnosticPresentation2["Squash"] = "squash";
  DiagnosticPresentation2["Strong"] = "strong";
  DiagnosticPresentation2["Weak"] = "weak";
  DiagnosticPresentation2["Flag"] = "flag";
  DiagnosticPresentation2["Error"] = "error";
  DiagnosticPresentation2["Dictionary"] = "dictionary";
  DiagnosticPresentation2["Via"] = "via";
  DiagnosticPresentation2["Added"] = "added";
  DiagnosticPresentation2["Deleted"] = "deleted";
  return DiagnosticPresentation2;
})(DiagnosticPresentation || {});
((DiagnosticPresentation2) => {
  DiagnosticPresentation2.presentation = /* @__PURE__ */ Symbol.for("matter.diagnostic.presentation");
  DiagnosticPresentation2.value = /* @__PURE__ */ Symbol.for("matter.diagnostic.value");
  DiagnosticPresentation2.name = /* @__PURE__ */ Symbol.for("matter.diagnostic.name");
})(DiagnosticPresentation || (DiagnosticPresentation = {}));
