// @matter/general/dist/esm/crypto/X509.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  ContextTagged,
  ContextTaggedBytes,
  DatatypeOverride,
  DerBitString,
  DerCodec,
  DerError,
  DerObject,
  DerType,
  ObjectId,
  RawBytes
} from "#codec/DerCodec.js";
import { ImplementationError } from "#MatterError.js";
import { Bytes } from "#util/Bytes.js";
import { CertificateError } from "./CryptoError.js";
import { Pem } from "./Pem.js";
import { X962 } from "./X962.js";
var X509;
((X5092) => {
  async function sign(crypto, key, cert) {
    return {
      ...cert,
      signature: (await crypto.signEcdsa(key, certificateToDer(cert))).der
    };
  }
  X5092.sign = sign;
  function certificateToPem(cert) {
    return Pem.encode(certificateToDer(cert));
  }
  X5092.certificateToPem = certificateToPem;
  function certificateToDer(cert) {
    const { serialNumber, signatureAlgorithm, issuer, validity, subject, publicKey, extensions } = cert;
    const { basicConstraints } = extensions;
    if (basicConstraints && !basicConstraints.isCa && basicConstraints.pathLen !== void 0) {
      throw new CertificateError("Path length must be undefined for non-CA certificates.");
    }
    let ast = {
      version: ContextTagged(0, 2),
      // v3
      serialNumber: DatatypeOverride(DerType.Integer, serialNumber),
      signatureAlgorithm: signatureAlgorithm ?? X962.EcdsaWithSHA256,
      issuer,
      validity,
      subject,
      publicKeyInfo: publicKey,
      extensions: ContextTagged(3, extensionsToAst(extensions))
    };
    if ("signature" in cert) {
      ast = {
        ast,
        signatureAlgorithm: cert.signatureAlgorithm,
        signature: DerBitString(cert.signature)
      };
    }
    return DerCodec.encode(ast);
  }
  X5092.certificateToDer = certificateToDer;
  X5092.KeyUsageFlags = [
    "digitalSignature",
    "nonRepudiation",
    "keyEncipherment",
    "dataEncipherment",
    "keyAgreement",
    "keyCertSign",
    "cRLSign",
    "encipherOnly",
    "decipherOnly"
  ];
  let Extensions;
  ((Extensions2) => {
    Extensions2.BASIC_CONSTRAINTS = 0x551d13n;
    Extensions2.KEY_USAGE = 0x551d0fn;
    Extensions2.EXTENDED_KEY_USAGE = 0x551d25n;
    Extensions2.SUBJECT_KEY_IDENTIFIER = 0x551d0en;
    Extensions2.AUTHORITY_KEY_IDENTIFIER = 0x551d23n;
  })(Extensions = X5092.Extensions || (X5092.Extensions = {}));
  function SubjectKeyIdentifier(identifier) {
    return DerObject(Extensions.SUBJECT_KEY_IDENTIFIER, { value: DerCodec.encode(identifier) });
  }
  X5092.SubjectKeyIdentifier = SubjectKeyIdentifier;
  function AuthorityKeyIdentifier(identifier) {
    return DerObject(Extensions.AUTHORITY_KEY_IDENTIFIER, {
      value: DerCodec.encode({ id: ContextTaggedBytes(0, identifier) })
    });
  }
  X5092.AuthorityKeyIdentifier = AuthorityKeyIdentifier;
  function BasicConstraints(constraints) {
    const toEncode = { ...constraints };
    if (toEncode?.isCa === false) {
      delete toEncode.isCa;
    }
    return DerObject(Extensions.BASIC_CONSTRAINTS, { critical: true, value: DerCodec.encode(toEncode) });
  }
  X5092.BasicConstraints = BasicConstraints;
  function ExtendedKeyUsage(values) {
    if (values === void 0) {
      return;
    }
    const data = {};
    values.forEach((value) => {
      switch (value) {
        case 1:
          data.serverAuth = ObjectId(ExtendedKeyUsage.SERVER_AUTH);
          break;
        case 2:
          data.clientAuth = ObjectId(ExtendedKeyUsage.CLIENT_AUTH);
          break;
        case 3:
          data.codeSigning = ObjectId(ExtendedKeyUsage.CODE_SIGNING);
          break;
        case 4:
          data.emailProtection = ObjectId(ExtendedKeyUsage.EMAIL_PROTECTION);
          break;
        case 5:
          data.timeStamping = ObjectId(ExtendedKeyUsage.TIME_STAMPING);
          break;
        case 6:
          data.ocspSigning = ObjectId(ExtendedKeyUsage.OCSP_SIGNING);
          break;
        default:
          throw new DerError(`Unsupported extended key usage value ${value}`);
      }
    });
    return DerObject(Extensions.EXTENDED_KEY_USAGE, {
      critical: true,
      value: DerCodec.encode(data)
    });
  }
  X5092.ExtendedKeyUsage = ExtendedKeyUsage;
  ((ExtendedKeyUsage2) => {
    ExtendedKeyUsage2.SERVER_AUTH = 0x2b06010505070301n;
    ExtendedKeyUsage2.CLIENT_AUTH = 0x2b06010505070302n;
    ExtendedKeyUsage2.CODE_SIGNING = 0x2b06010505070303n;
    ExtendedKeyUsage2.EMAIL_PROTECTION = 0x2b06010505070304n;
    ExtendedKeyUsage2.TIME_STAMPING = 0x2b06010505070308n;
    ExtendedKeyUsage2.OCSP_SIGNING = 0x2b06010505070309n;
  })(ExtendedKeyUsage = X5092.ExtendedKeyUsage || (X5092.ExtendedKeyUsage = {}));
  function KeyUsage(value) {
    if (typeof value !== "number") {
      value = encodeKeyUsage(value);
    }
    return DerObject(Extensions.KEY_USAGE, {
      critical: true,
      value: DerCodec.encode(DatatypeOverride(DerType.BitString, value))
    });
  }
  X5092.KeyUsage = KeyUsage;
})(X509 || (X509 = {}));
const ExtensionEncoders = {
  basicConstraints: X509.BasicConstraints,
  keyUsage: X509.KeyUsage,
  extendedKeyUsage: X509.ExtendedKeyUsage,
  subjectKeyIdentifier: X509.SubjectKeyIdentifier,
  authorityKeyIdentifier: X509.AuthorityKeyIdentifier,
  futureExtension: (future) => RawBytes(Bytes.concat(...future ?? []))
};
function extensionsToAst(extensions) {
  const ast = {};
  for (const key in extensions) {
    const value = extensions[key];
    if (value === void 0) {
      continue;
    }
    const encoder = ExtensionEncoders[key];
    if (encoder === void 0) {
      throw new ImplementationError(`Unsupported X.509 certificate extension ${key}`);
    }
    ast[key] = encoder(value);
  }
  return ast;
}
function encodeKeyUsage(flags) {
  let bit = 1;
  let encoded = 0;
  for (const flag of X509.KeyUsageFlags) {
    if (flags[flag]) {
      encoded |= bit;
    }
    bit <<= 1;
  }
  return encoded;
}
export {
  X509
};
