// @matter/node/dist/esm/tags/ClosureTag.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SemanticNamespace } from "../endpoint/type/SemanticNamespace.js";
const ClosureTag = SemanticNamespace({
  id: 68,
  tags: {
    /**
     * Any material or treatment used to cover a window for privacy, light control, insulation, or decoration.
     */
    Covering: { id: 0, label: "Covering" },
    /**
     * An opening in a wall, door, or roof fitted with glass or another transparent material to allow light, air,
     * and views while providing insulation and security.
     */
    Window: { id: 1, label: "Window" },
    /**
     * A physical or conceptual obstruction that prevents movement, access, or progress that serve purposes such as
     * security, safety, privacy, control, etc.
     */
    Barrier: { id: 2, label: "Barrier" },
    /**
     * A storage unit with shelves, drawers, or doors, used for organizing and protecting items.
     */
    Cabinet: { id: 3, label: "Cabinet" },
    /**
     * A movable barrier that controls entry or exit through an opening in a fence, wall, or enclosure.
     */
    Gate: { id: 4, label: "Gate" },
    /**
     * A large, movable barrier that provides access to a garage.
     */
    GarageDoor: { id: 5, label: "GarageDoor" },
    /**
     * A movable barrier that allows entry and exit between spaces while providing security, privacy, and
     * insulation.
     */
    Door: { id: 6, label: "Door" }
  }
});
export {
  ClosureTag
};
