// @matter/protocol/dist/esm/certificate/kinds/Certificate.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  Bytes,
  CertificateError,
  ContextTagged,
  DerBitString,
  DerCodec,
  DerType,
  EcdsaSignature,
  PublicKey,
  RawBytes,
  X509,
  X520,
  X962
} from "@matter/general";
import {
  CaseAuthenticatedTag,
  MATTER_EPOCH_OFFSET_S
} from "@matter/types";
import { assertCertificateDerSize } from "./common.js";
import {
  FabricId_Matter,
  FirmwareSigningId_Matter,
  IcacId_Matter,
  MATTER_OID_TO_FIELD_MAP,
  matterToJsDate,
  NocCat_Matter,
  NodeId_Matter,
  ProductId_Matter,
  RcacId_Matter,
  VendorId_Matter,
  VvsId_Matter
} from "./definitions/asn.js";
import { ExtensionKeyUsageSchema } from "./definitions/base.js";
class Certificate {
  #signature;
  #cert;
  constructor(cert) {
    this.#cert = cert;
    if ("signature" in cert) {
      this.#signature = new EcdsaSignature(cert.signature);
    }
  }
  get cert() {
    return this.#cert;
  }
  get isSigned() {
    return this.#signature !== void 0;
  }
  /**
   * Get the signature of the certificate.
   * If the certificate is not signed, it throws a CertificateError.
   */
  get signature() {
    if (this.#signature === void 0) {
      throw new CertificateError("Certificate is not signed");
    }
    return this.#signature;
  }
  /**
   * Set the signature of the certificate.
   * If the certificate is already signed, it throws a CertificateError.
   */
  set signature(signature) {
    if (this.isSigned) {
      throw new CertificateError("Certificate is already signed");
    }
    this.#signature = signature;
  }
  /**
   * Sign the certificate using the provided crypto and key.
   * It throws a CertificateError if the certificate is already signed.
   */
  async sign(crypto, key) {
    this.signature = await crypto.signEcdsa(key, this.asUnsignedDer());
  }
  /**
   * Serialize as signed DER.
   *
   * If no signature is present, throws an error.
   */
  asSignedDer() {
    const certBytes = X509.certificateToDer({
      ...matterToX509(this.cert),
      signatureAlgorithm: X962.EcdsaWithSHA256,
      signature: this.signature.der
    });
    assertCertificateDerSize(certBytes);
    return certBytes;
  }
  /**
   * Serialize as DER without signature.
   */
  asUnsignedDer() {
    const certBytes = X509.certificateToDer(matterToX509(this.cert));
    assertCertificateDerSize(certBytes);
    return certBytes;
  }
  /**
   * Convert the extensions of the certificate to ASN.1 DER format.
   */
  extensionsToAsn1(extensions) {
    const asn = {};
    Object.entries(extensions).forEach(([key, value]) => {
      if (value === void 0) {
        return;
      }
      switch (key) {
        case "basicConstraints":
          asn.basicConstraints = X509.BasicConstraints(value);
          break;
        case "keyUsage":
          asn.keyUsage = X509.KeyUsage(
            ExtensionKeyUsageSchema.encode(
              value
            )
          );
          break;
        case "extendedKeyUsage":
          asn.extendedKeyUsage = X509.ExtendedKeyUsage(value);
          break;
        case "subjectKeyIdentifier":
          asn.subjectKeyIdentifier = X509.SubjectKeyIdentifier(value);
          break;
        case "authorityKeyIdentifier":
          asn.authorityKeyIdentifier = X509.AuthorityKeyIdentifier(value);
          break;
        case "futureExtension":
          asn.futureExtension = RawBytes(Bytes.concat(...value ?? []));
          break;
      }
    });
    return asn;
  }
}
((Certificate2) => {
  async function createCertificateSigningRequest(crypto, key) {
    const request = {
      version: 0,
      subject: { organization: X520.OrganisationName("CSR") },
      publicKey: X962.PublicKeyEcPrime256v1(key.publicKey),
      endSignedBytes: ContextTagged(0)
    };
    return DerCodec.encode({
      request,
      signAlgorithm: X962.EcdsaWithSHA256,
      signature: DerBitString((await crypto.signEcdsa(key, DerCodec.encode(request))).der)
    });
  }
  Certificate2.createCertificateSigningRequest = createCertificateSigningRequest;
  function oidToSubjectField(oid) {
    const oidHex = Bytes.toHex(oid);
    const field = X520.OID_TO_FIELD_MAP[oidHex] ?? MATTER_OID_TO_FIELD_MAP[oidHex];
    if (field !== void 0) {
      return { field, isPrintable: false };
    }
  }
  function parseSubjectOrIssuer(node) {
    const result = {};
    const { _elements: rdnSequence } = node;
    if (!rdnSequence) {
      throw new CertificateError("Invalid subject/issuer structure");
    }
    for (const rdnSet of rdnSequence) {
      const { _elements: attributeSets } = rdnSet;
      if (!attributeSets) continue;
      for (const attributeSet of attributeSets) {
        const { _elements: attrElements } = attributeSet;
        if (!attrElements || attrElements.length !== 2) continue;
        const [oidNode, valueNode] = attrElements;
        const oid = oidNode._bytes;
        const fieldInfo = oidToSubjectField(oid);
        if (fieldInfo === void 0) continue;
        let { field } = fieldInfo;
        let value;
        const valueBytes = Bytes.of(valueNode._bytes);
        const valueTag = valueNode._tag;
        switch (field) {
          case "nodeId":
          case "fabricId": {
            const hexString = Bytes.toString(valueBytes);
            value = BigInt("0x" + hexString);
            break;
          }
          case "icacId":
          case "rcacId":
          case "vvsId": {
            const hexString = Bytes.toString(valueBytes);
            const bigIntValue = BigInt("0x" + hexString);
            value = bigIntValue <= BigInt(Number.MAX_SAFE_INTEGER) ? Number(bigIntValue) : bigIntValue;
            break;
          }
          case "firmwareSigningId":
          case "productId":
          case "vendorId": {
            const hexString = Bytes.toString(valueBytes);
            value = parseInt(hexString, 16);
            break;
          }
          case "caseAuthenticatedTag": {
            const hexString = Bytes.toString(valueBytes);
            const catValue = parseInt(hexString, 16);
            if (result.caseAuthenticatedTags !== void 0) {
              result.caseAuthenticatedTags.push(catValue);
              continue;
            }
            field = "caseAuthenticatedTags";
            value = [catValue];
            break;
          }
          default: {
            value = Bytes.toString(valueBytes);
            if (valueTag === DerType.PrintableString) {
              field += "Ps";
            }
          }
        }
        result[field] = value;
      }
    }
    return result;
  }
  Certificate2.REQUIRED_PAA_EXTENSIONS = ["basicConstraints", "keyUsage", "subjectKeyIdentifier"];
  Certificate2.REQUIRED_EXTENSIONS = [...Certificate2.REQUIRED_PAA_EXTENSIONS, "authorityKeyIdentifier"];
  function parseExtensions(extensionsNode, requiredExtensions) {
    const result = {
      basicConstraints: { isCa: false }
    };
    const { _elements: extensions } = extensionsNode;
    if (!extensions) {
      throw new CertificateError("Invalid extensions structure");
    }
    for (const ext of extensions) {
      const { _elements: extElements } = ext;
      if (!extElements || extElements.length < 2) continue;
      const oid = extElements[0]._bytes;
      const oidValue = Bytes.asBigInt(oid);
      let valueIndex = 1;
      if (extElements.length > 2 && extElements[1]._tag === DerType.Boolean) {
        valueIndex = 2;
      }
      const valueOctetString = extElements[valueIndex]._bytes;
      const valueNode = DerCodec.decode(valueOctetString);
      switch (oidValue) {
        case X509.Extensions.BASIC_CONSTRAINTS:
          {
            const { _elements: bcElements } = valueNode;
            if (bcElements && bcElements.length > 0) {
              if (bcElements[0]._tag === DerType.Boolean) {
                const bcBytes = Bytes.of(bcElements[0]._bytes);
                result.basicConstraints.isCa = bcBytes[0] !== 0;
              }
              if (bcElements.length > 1 && bcElements[1]._tag === DerType.Integer) {
                const pathLenBytes = Bytes.of(bcElements[1]._bytes);
                result.basicConstraints.pathLen = pathLenBytes[0];
              }
            }
          }
          break;
        case X509.Extensions.KEY_USAGE:
          {
            const bitString = Bytes.of(valueNode._bytes);
            if (bitString.length >= 1) {
              const usageByte = bitString[0];
              result.keyUsage = {
                digitalSignature: (usageByte & 128) !== 0,
                nonRepudiation: (usageByte & 64) !== 0,
                keyEncipherment: (usageByte & 32) !== 0,
                dataEncipherment: (usageByte & 16) !== 0,
                keyAgreement: (usageByte & 8) !== 0,
                keyCertSign: (usageByte & 4) !== 0,
                cRLSign: (usageByte & 2) !== 0,
                encipherOnly: (usageByte & 1) !== 0,
                decipherOnly: bitString.length > 1 ? (bitString[1] & 128) !== 0 : false
              };
            }
          }
          break;
        case X509.Extensions.EXTENDED_KEY_USAGE:
          {
            const { _elements: ekuElements } = valueNode;
            if (ekuElements) {
              const ekuValues = [];
              for (const eku of ekuElements) {
                const ekuOidValue = Bytes.asBigInt(eku._bytes);
                switch (ekuOidValue) {
                  case X509.ExtendedKeyUsage.SERVER_AUTH:
                    ekuValues.push(1);
                    break;
                  case X509.ExtendedKeyUsage.CLIENT_AUTH:
                    ekuValues.push(2);
                    break;
                  case X509.ExtendedKeyUsage.CODE_SIGNING:
                    ekuValues.push(3);
                    break;
                  case X509.ExtendedKeyUsage.EMAIL_PROTECTION:
                    ekuValues.push(4);
                    break;
                  case X509.ExtendedKeyUsage.TIME_STAMPING:
                    ekuValues.push(5);
                    break;
                  case X509.ExtendedKeyUsage.OCSP_SIGNING:
                    ekuValues.push(6);
                    break;
                }
              }
              if (ekuValues.length > 0) {
                result.extendedKeyUsage = ekuValues;
              }
            }
          }
          break;
        case X509.Extensions.SUBJECT_KEY_IDENTIFIER:
          result.subjectKeyIdentifier = valueNode._bytes;
          break;
        case X509.Extensions.AUTHORITY_KEY_IDENTIFIER:
          {
            const { _elements: akiElements } = valueNode;
            if (akiElements && akiElements.length > 0) {
              result.authorityKeyIdentifier = akiElements[0]._bytes;
            }
          }
          break;
      }
    }
    if (requiredExtensions.some((ext) => result[ext] === void 0)) {
      throw new CertificateError("Missing required extensions in certificate");
    }
    return result;
  }
  function parseDate(node) {
    const dateBytes = node._bytes;
    const dateString = Bytes.toString(dateBytes);
    const tag = node._tag;
    let year, month, day, hour, minute, second;
    if (tag === DerType.UtcDate) {
      year = parseInt(dateString.substring(0, 2));
      year += year >= 50 ? 1900 : 2e3;
      month = parseInt(dateString.substring(2, 4));
      day = parseInt(dateString.substring(4, 6));
      hour = parseInt(dateString.substring(6, 8));
      minute = parseInt(dateString.substring(8, 10));
      second = parseInt(dateString.substring(10, 12));
    } else if (tag === DerType.GeneralizedTime) {
      year = parseInt(dateString.substring(0, 4));
      month = parseInt(dateString.substring(4, 6));
      day = parseInt(dateString.substring(6, 8));
      hour = parseInt(dateString.substring(8, 10));
      minute = parseInt(dateString.substring(10, 12));
      second = parseInt(dateString.substring(12, 14));
    } else {
      throw new CertificateError(`Unsupported date type: ${tag}`);
    }
    const date = new Date(Date.UTC(year, month - 1, day, hour, minute, second));
    if (date.getTime() === X520.NON_WELL_DEFINED_DATE.getTime()) {
      return 0;
    }
    return Math.floor(date.getTime() / 1e3) - MATTER_EPOCH_OFFSET_S;
  }
  function parseAsn1Certificate(encodedCert, requiredExtensions = Certificate2.REQUIRED_EXTENSIONS) {
    const { _elements: rootElements } = DerCodec.decode(encodedCert);
    if (!rootElements || rootElements.length !== 3) {
      throw new CertificateError(
        `Invalid certificate structure - expected 3 root elements, got ${rootElements?.length ?? 0}`
      );
    }
    const [certificateNode, , signatureNode] = rootElements;
    const { _elements: certElements } = certificateNode;
    if (!certElements || certElements.length < 7) {
      throw new CertificateError("Invalid TBSCertificate structure");
    }
    let idx = 0;
    if (certElements[idx]._tag === 160) {
      idx++;
    }
    const serialNumber = certElements[idx++]._bytes;
    const signatureAlgorithmOid = certElements[idx]._elements?.[0]?._bytes;
    if (!signatureAlgorithmOid) {
      throw new CertificateError("Invalid signature algorithm structure");
    }
    const signatureAlgorithm = Bytes.toHex(signatureAlgorithmOid) === "2a8648ce3d040302" ? 1 : 0;
    idx++;
    const issuer = parseSubjectOrIssuer(certElements[idx++]);
    const { _elements: validityElements } = certElements[idx++];
    if (!validityElements || validityElements.length !== 2) {
      throw new CertificateError("Invalid validity structure");
    }
    const notBefore = parseDate(validityElements[0]);
    const notAfter = parseDate(validityElements[1]);
    const subject = parseSubjectOrIssuer(certElements[idx++]);
    const { _elements: publicKeyElements } = certElements[idx++];
    if (!publicKeyElements || publicKeyElements.length !== 2) {
      throw new CertificateError("Invalid public key structure");
    }
    const { _elements: algorithmElements } = publicKeyElements[0];
    if (!algorithmElements || algorithmElements.length !== 2) {
      throw new CertificateError("Invalid public key algorithm structure");
    }
    const publicKeyAlgorithmOid = Bytes.toHex(algorithmElements[0]._bytes);
    const publicKeyAlgorithm = publicKeyAlgorithmOid === "2a8648ce3d0201" ? 1 : 0;
    const ellipticCurveOid = Bytes.toHex(algorithmElements[1]._bytes);
    const ellipticCurveIdentifier = ellipticCurveOid === "2a8648ce3d030107" ? 1 : 0;
    const ellipticCurvePublicKey = Bytes.of(publicKeyElements[1]._bytes);
    if (idx >= certElements.length || certElements[idx]._tag !== 163) {
      throw new CertificateError("Missing required extensions in certificate");
    }
    const extensionsBytes = certElements[idx]._bytes;
    const extensionsSequence = DerCodec.decode(extensionsBytes);
    const extensions = parseExtensions(extensionsSequence, requiredExtensions);
    const signature = new EcdsaSignature(Bytes.of(signatureNode._bytes), "der").bytes;
    return {
      serialNumber,
      signatureAlgorithm,
      issuer,
      notBefore,
      notAfter,
      subject,
      publicKeyAlgorithm,
      ellipticCurveIdentifier,
      ellipticCurvePublicKey,
      extensions,
      signature
    };
  }
  Certificate2.parseAsn1Certificate = parseAsn1Certificate;
  async function getPublicKeyFromCsr(crypto, encodedCsr) {
    const { _elements: rootElements } = DerCodec.decode(encodedCsr);
    if (rootElements?.length !== 3) {
      throw new CertificateError("Invalid CSR data");
    }
    const [requestNode, signAlgorithmNode, signatureNode] = rootElements;
    const { _elements: requestElements } = requestNode;
    if (requestElements?.length !== 4) {
      throw new CertificateError("Invalid CSR data");
    }
    const [versionNode, subjectNode, publicKeyNode] = requestElements;
    const requestVersionBytes = Bytes.of(versionNode._bytes);
    if (requestVersionBytes.length !== 1 || requestVersionBytes[0] !== 0) {
      throw new CertificateError(`Unsupported CSR version ${requestVersionBytes[0]}`);
    }
    if (!subjectNode._elements?.length) {
      throw new CertificateError("Missing subject in CSR data");
    }
    const { _elements: publicKeyElements } = publicKeyNode;
    if (publicKeyElements?.length !== 2) {
      throw new CertificateError("Invalid CSR data");
    }
    const [publicKeyTypeNode, publicKeyBytesNode] = publicKeyElements;
    const { _elements: publicKeyTypeNodeElements } = publicKeyTypeNode;
    if (publicKeyTypeNodeElements?.length !== 2) {
      throw new CertificateError("Invalid public key type in CSR");
    }
    if (!Bytes.areEqual(publicKeyTypeNodeElements[0]._bytes, X962.PublicKeyAlgorithmEcPublicKey._bytes)) {
      throw new CertificateError("Unsupported public key algorithm in CSR");
    }
    if (!Bytes.areEqual(publicKeyTypeNodeElements[1]._bytes, X962.PublicKeyAlgorithmEcPublicKeyP256._bytes)) {
      throw new CertificateError("Unsupported public key curve in CSR");
    }
    const publicKey = publicKeyBytesNode._bytes;
    const signatureAlgorithmBytes = signAlgorithmNode._elements?.[0]?._bytes;
    if (signatureAlgorithmBytes === void 0 || !Bytes.areEqual(X962.EcdsaWithSHA256._objectId._bytes, signatureAlgorithmBytes)) {
      throw new CertificateError("Unsupported signature algorithm in CSR");
    }
    await crypto.verifyEcdsa(
      PublicKey(publicKey),
      DerCodec.encode(requestNode),
      new EcdsaSignature(signatureNode._bytes, "der")
    );
    return publicKey;
  }
  Certificate2.getPublicKeyFromCsr = getPublicKeyFromCsr;
})(Certificate || (Certificate = {}));
function matterToX509(cert) {
  const { serialNumber, notBefore, notAfter, issuer, subject, ellipticCurvePublicKey, extensions } = cert;
  return {
    serialNumber,
    validity: {
      notBefore: matterToJsDate(notBefore),
      notAfter: matterToJsDate(notAfter)
    },
    issuer: astOfDistinguishedName(issuer),
    subject: astOfDistinguishedName(subject),
    extensions,
    signatureAlgorithm: X962.EcdsaWithSHA256,
    publicKey: X962.PublicKeyEcPrime256v1(ellipticCurvePublicKey)
  };
}
function astOfDistinguishedName(data) {
  const ast = {};
  Object.entries(data).forEach(([key, value]) => {
    if (value === void 0) {
      return;
    }
    switch (key) {
      case "commonName":
        ast.commonName = X520.CommonName(value);
        break;
      case "surName":
        ast.surName = X520.SurName(value);
        break;
      case "serialNum":
        ast.serialNum = X520.SerialNumber(value);
        break;
      case "countryName":
        ast.countryName = X520.CountryName(value);
        break;
      case "localityName":
        ast.localityName = X520.LocalityName(value);
        break;
      case "stateOrProvinceName":
        ast.stateOrProvinceName = X520.StateOrProvinceName(value);
        break;
      case "orgName":
        ast.orgName = X520.OrganisationName(value);
        break;
      case "orgUnitName":
        ast.orgUnitName = X520.OrganizationalUnitName(value);
        break;
      case "title":
        ast.title = X520.Title(value);
        break;
      case "name":
        ast.name = X520.Name(value);
        break;
      case "givenName":
        ast.givenName = X520.GivenName(value);
        break;
      case "initials":
        ast.initials = X520.Initials(value);
        break;
      case "genQualifier":
        ast.genQualifier = X520.GenerationQualifier(value);
        break;
      case "dnQualifier":
        ast.dnQualifier = X520.DnQualifier(value);
        break;
      case "pseudonym":
        ast.pseudonym = X520.Pseudonym(value);
        break;
      case "domainComponent":
        ast.domainComponent = X520.DomainComponent(value);
        break;
      case "nodeId":
        ast.nodeId = NodeId_Matter(value);
        break;
      case "firmwareSigningId":
        ast.firmwareSigningId = FirmwareSigningId_Matter(value);
        break;
      case "icacId":
        ast.icacId = IcacId_Matter(value);
        break;
      case "rcacId":
        ast.rcacId = RcacId_Matter(value);
        break;
      case "vvsId":
        ast.vvsId = VvsId_Matter(value);
        break;
      case "fabricId":
        ast.fabricId = FabricId_Matter(value);
        break;
      case "caseAuthenticatedTags":
        const caseAuthenticatedTags = value;
        CaseAuthenticatedTag.validateNocTagList(caseAuthenticatedTags);
        const cat0 = caseAuthenticatedTags[0];
        const cat1 = caseAuthenticatedTags[1];
        const cat2 = caseAuthenticatedTags[2];
        if (cat0 !== void 0) {
          ast.caseAuthenticatedTag0 = NocCat_Matter(cat0);
        }
        if (cat1 !== void 0) {
          ast.caseAuthenticatedTag1 = NocCat_Matter(cat1);
        }
        if (cat2 !== void 0) {
          ast.caseAuthenticatedTag2 = NocCat_Matter(cat2);
        }
        break;
      case "vendorId":
        ast.vendorId = VendorId_Matter(value);
        break;
      case "productId":
        ast.productId = ProductId_Matter(value);
        break;
      case "commonNamePs":
        ast.commonNamePs = X520.CommonName(value, true);
        break;
      case "surNamePs":
        ast.surNamePs = X520.SurName(value, true);
        break;
      case "serialNumPs":
        ast.serialNumPs = X520.SerialNumber(value, true);
        break;
      case "countryNamePs":
        ast.countryNamePs = X520.CountryName(value, true);
        break;
      case "localityNamePs":
        ast.localityNamePs = X520.LocalityName(value, true);
        break;
      case "stateOrProvinceNamePs":
        ast.stateOrProvinceNamePs = X520.StateOrProvinceName(value, true);
        break;
      case "orgNamePs":
        ast.orgNamePs = X520.OrganisationName(value, true);
        break;
      case "orgUnitNamePs":
        ast.orgUnitNamePs = X520.OrganizationalUnitName(value, true);
        break;
      case "titlePs":
        ast.titlePs = X520.Title(value, true);
        break;
      case "namePs":
        ast.namePs = X520.Name(value, true);
        break;
      case "givenNamePs":
        ast.givenNamePs = X520.GivenName(value, true);
        break;
      case "initialsPs":
        ast.initialsPs = X520.Initials(value, true);
        break;
      case "genQualifierPs":
        ast.genQualifierPs = X520.GenerationQualifier(value, true);
        break;
      case "dnQualifierPs":
        ast.dnQualifierPs = X520.DnQualifier(value, true);
        break;
      case "pseudonymPs":
        ast.pseudonymPs = X520.Pseudonym(value, true);
        break;
    }
  });
  return ast;
}
export {
  Certificate
};
