// @matter/general/dist/cjs/storage/DatafileRoot.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DatafileRoot_exports = {};
__export(DatafileRoot_exports, {
  DatafileRoot: () => DatafileRoot
});
module.exports = __toCommonJS(DatafileRoot_exports);
var import_DataNamespace = require("./DataNamespace.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class DatafileRoot extends import_DataNamespace.DataNamespace {
  #directory;
  #release;
  #refs = 0;
  #lockPromise;
  constructor(directory) {
    super(directory.name);
    this.#directory = directory;
  }
  get directory() {
    return this.#directory;
  }
  get path() {
    return this.#directory.path;
  }
  get isLocked() {
    return this.#refs > 0;
  }
  /**
   * Acquire a reference-counted lock.  The physical directory lock is acquired on the first call and released when
   * the last {@link DatafileRoot.Lock} closes.
   */
  async lock() {
    if (this.#lockPromise) {
      await this.#lockPromise;
    }
    const promise = this.#acquireRef();
    this.#lockPromise = promise;
    try {
      return await promise;
    } finally {
      if (this.#lockPromise === promise) {
        this.#lockPromise = void 0;
      }
    }
  }
  async #acquireRef() {
    if (this.#refs === 0) {
      this.#release = await this.#directory.lock();
    }
    this.#refs++;
    return new DatafileRoot.Lock(this.#directory, () => this.#releaseRef());
  }
  async #releaseRef() {
    this.#refs--;
    if (this.#refs === 0) {
      const release = this.#release;
      this.#release = void 0;
      await release?.();
    }
  }
}
((DatafileRoot2) => {
  class Lock {
    #directory;
    #release;
    constructor(directory, release) {
      this.#directory = directory;
      this.#release = release;
    }
    get directory() {
      return this.#directory;
    }
    get path() {
      return this.#directory.path;
    }
    async close() {
      const release = this.#release;
      if (release) {
        this.#release = void 0;
        await release();
      }
    }
    async [Symbol.asyncDispose]() {
      await this.close();
    }
  }
  DatafileRoot2.Lock = Lock;
})(DatafileRoot || (DatafileRoot = {}));
