// @matter/general/dist/cjs/storage/MemoryBlobStorageDriver.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MemoryBlobStorageDriver_exports = {};
__export(MemoryBlobStorageDriver_exports, {
  MemoryBlobStorageDriver: () => MemoryBlobStorageDriver
});
module.exports = __toCommonJS(MemoryBlobStorageDriver_exports);
var import_Bytes = require("#util/Bytes.js");
var import_BlobStorageDriver = require("./BlobStorageDriver.js");
var import_StorageDriver = require("./StorageDriver.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MemoryBlobStorageDriver extends import_BlobStorageDriver.BlobStorageDriver {
  static id = "memory-blob";
  #store = /* @__PURE__ */ new Map();
  get initialized() {
    return true;
  }
  initialize() {
  }
  close() {
    this.#store.clear();
  }
  #createContextKey(contexts) {
    for (const ctx of contexts) {
      if (!ctx.length || ctx.includes(".")) {
        throw new import_StorageDriver.StorageError("Context must not contain empty segments or leading or trailing dots.");
      }
    }
    return contexts.join(".");
  }
  #contextMap(contexts) {
    return this.#store.get(this.#createContextKey(contexts));
  }
  #ensureContextMap(contexts) {
    const key = this.#createContextKey(contexts);
    let map = this.#store.get(key);
    if (map === void 0) {
      map = /* @__PURE__ */ new Map();
      this.#store.set(key, map);
    }
    return map;
  }
  openBlob(contexts, key) {
    const map = this.#contextMap(contexts);
    if (map === void 0) {
      return new Blob([]);
    }
    const value = map.get(key);
    if (value === void 0) {
      return new Blob([]);
    }
    if (!import_Bytes.Bytes.isBytes(value)) {
      throw new import_StorageDriver.StorageError("Value must be Bytes to read as blob.");
    }
    return new Blob([import_Bytes.Bytes.exclusive(value)]);
  }
  async writeBlobFromStream(contexts, key, stream) {
    const bytes = new Uint8Array(await new Response(stream).arrayBuffer());
    this.#ensureContextMap(contexts).set(key, bytes);
  }
  /**
   * Synchronously set bytes for a key, bypassing stream-based write.
   * Useful in tests where async stream operations may not work with MockTime.
   */
  setBytes(contexts, key, data) {
    this.#ensureContextMap(contexts).set(key, import_Bytes.Bytes.of(data));
  }
  /**
   * Synchronously get bytes for a key.
   * Useful in tests where async blob reading may not work with MockTime.
   */
  getBytes(contexts, key) {
    return this.#contextMap(contexts)?.get(key);
  }
  delete(contexts, key) {
    this.#contextMap(contexts)?.delete(key);
  }
  has(contexts, key) {
    return this.#contextMap(contexts)?.has(key) ?? false;
  }
  keys(contexts) {
    const map = this.#contextMap(contexts);
    return map ? [...map.keys()] : [];
  }
  contexts(contexts) {
    const contextKey = contexts.length ? this.#createContextKey(contexts) : "";
    const startContextKey = contextKey.length ? `${contextKey}.` : "";
    const foundContexts = new Array();
    for (const key of this.#store.keys()) {
      if (key.startsWith(startContextKey)) {
        const subKeys = key.substring(startContextKey.length).split(".");
        if (subKeys.length < 1) continue;
        const context = subKeys[0];
        if (context.length && !foundContexts.includes(context)) {
          foundContexts.push(context);
        }
      }
    }
    return foundContexts;
  }
  clearAll(contexts) {
    const contextKey = this.#createContextKey(contexts);
    this.#store.delete(contextKey);
    const startContextKey = contextKey.length ? `${contextKey}.` : "";
    for (const key of [...this.#store.keys()]) {
      if (key.startsWith(startContextKey)) {
        this.#store.delete(key);
      }
    }
  }
}
