// @matter/general/dist/cjs/storage/StorageDriverHandle.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StorageDriverHandle_exports = {};
__export(StorageDriverHandle_exports, {
  StorageDriverHandle: () => StorageDriverHandle
});
module.exports = __toCommonJS(StorageDriverHandle_exports);
var import_StorageDriver = require("./StorageDriver.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class StorageDriverHandle extends import_StorageDriver.StorageDriver {
  #driver;
  #release;
  clone;
  constructor(driver, release) {
    super();
    this.#driver = driver;
    this.#release = release;
    if (import_StorageDriver.CloneableStorage.is(driver)) {
      this.clone = () => driver.clone();
    }
  }
  get id() {
    return this.#driver.id;
  }
  get initialized() {
    return this.#driver.initialized;
  }
  initialize() {
  }
  close() {
    return this.#release();
  }
  get(contexts, key) {
    return this.#driver.get(contexts, key);
  }
  set(contexts, keyOrValues, value) {
    if (typeof keyOrValues === "string") {
      return this.#driver.set(contexts, keyOrValues, value);
    }
    return this.#driver.set(contexts, keyOrValues);
  }
  delete(contexts, key) {
    return this.#driver.delete(contexts, key);
  }
  keys(contexts) {
    return this.#driver.keys(contexts);
  }
  values(contexts) {
    return this.#driver.values(contexts);
  }
  contexts(contexts) {
    return this.#driver.contexts(contexts);
  }
  clearAll(contexts) {
    return this.#driver.clearAll(contexts);
  }
  has(contexts, key) {
    return this.#driver.has(contexts, key);
  }
  begin() {
    return this.#driver.begin();
  }
}
