// @matter/protocol/dist/cjs/mdns/CommissionableMdnsScanner.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CommissionableMdnsScanner_exports = {};
__export(CommissionableMdnsScanner_exports, {
  CommissionableMdnsScanner: () => CommissionableMdnsScanner
});
module.exports = __toCommonJS(CommissionableMdnsScanner_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_Scanner = require("../common/Scanner.js");
var import_MdnsConsts = require("./MdnsConsts.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("CommissionableMdnsScanner");
const SPECULATIVE_TARGET_TTL = (0, import_general.Hours)(1);
const SPECULATIVE_CLEANUP_INTERVAL = (0, import_general.Minutes)(30);
const SPECULATIVE_TARGET_MAX = 50;
const COMMISSIONING_RETRY_INTERVAL = (0, import_general.Seconds)(30);
class CommissionableMdnsScanner {
  type = import_general.ChannelType.UDP;
  #names;
  #filter;
  #observers = new import_general.ObserverGroup();
  #cache = /* @__PURE__ */ new Map();
  #waiters = /* @__PURE__ */ new Set();
  // Targets materialized by PTR-follow that never received SRV/TXT.  Each target gets a child abort driving
  // its own SRV/TXT discover() loop and a null observer keeping the name alive until we decide to drop it.
  #speculativeTargets = /* @__PURE__ */ new Map();
  #speculativeObserver = () => void 0;
  #speculativeCleanupTimer;
  #scanAbort = new import_general.Abort();
  constructor(names) {
    this.#names = names;
    const suffix1 = `._matterc._udp.local`;
    const base1 = `_matterc._udp.local`;
    const suffix2 = `._matterd._udp.local`;
    const base2 = `_matterd._udp.local`;
    this.#filter = (record) => {
      const lower = record.name.toLowerCase();
      return lower === base1 || lower.endsWith(suffix1) || lower === base2 || lower.endsWith(suffix2);
    };
    names.filters.add(this.#filter);
    this.#observers.on(names.discovered, this.#onDiscovered.bind(this));
    this.#speculativeCleanupTimer = import_general.Time.getPeriodicTimer(
      "Speculative PTR target cleanup",
      SPECULATIVE_CLEANUP_INTERVAL,
      this.#pruneSpeculativeTargets.bind(this)
    );
    this.#speculativeCleanupTimer.utility = true;
    this.#speculativeCleanupTimer.start();
  }
  async close() {
    for (const waiter of this.#waiters) {
      waiter.cancel();
    }
    this.#speculativeCleanupTimer.stop();
    this.#scanAbort();
    for (const qname of this.#speculativeTargets.keys()) {
      this.#names.maybeGet(qname)?.off(this.#speculativeObserver);
    }
    this.#speculativeTargets.clear();
    this.#names.filters.delete(this.#filter);
    this.#observers.close();
    await import_general.MatterAggregateError.allSettled(
      [...this.#cache.values()].map(({ ipService }) => ipService.close()),
      "Error closing cached IpServices"
    );
    this.#cache.clear();
  }
  #pruneSpeculativeTargets() {
    if (this.#speculativeTargets.size === 0) {
      return;
    }
    const cutoff = import_general.Time.nowMs - SPECULATIVE_TARGET_TTL;
    for (const [qname, { abort, createdAt }] of this.#speculativeTargets) {
      if (createdAt > cutoff) {
        continue;
      }
      this.#speculativeTargets.delete(qname);
      abort();
      this.#names.maybeGet(qname)?.off(this.#speculativeObserver);
    }
  }
  #clearSpeculativeTarget(qname, name) {
    const tracking = this.#speculativeTargets.get(qname);
    if (tracking === void 0) {
      return;
    }
    this.#speculativeTargets.delete(qname);
    tracking.abort();
    name.off(this.#speculativeObserver);
  }
  async findCommissionableDevicesContinuously(identifier, callback, timeout, cancelSignal) {
    for (const name of this.#names.discoveredNames) {
      this.#onDiscovered(name);
    }
    const seen = /* @__PURE__ */ new Set();
    const result = [];
    let cancelResolve;
    const internalCancel = new Promise((resolve) => cancelResolve = resolve);
    const waiter = {
      identifier,
      cancel: cancelResolve,
      notify: (device) => {
        if (matchesIdentifier(device, identifier) && !seen.has(device.deviceIdentifier)) {
          seen.add(device.deviceIdentifier);
          result.push(device);
          callback(device);
        }
      }
    };
    this.#waiters.add(waiter);
    let callbackInvoked = false;
    for (const cached of this.#cache.values()) {
      const device = refreshAddresses(cached);
      if (matchesIdentifier(device, identifier) && device.addresses.length > 0) {
        seen.add(device.deviceIdentifier);
        result.push(device);
        callbackInvoked = true;
        callback(device);
      }
    }
    const sleepTimer = timeout !== void 0 ? import_general.Time.sleep("commissionable scanner timeout", timeout) : void 0;
    const signals = [internalCancel];
    if (sleepTimer) signals.push(sleepTimer);
    if (cancelSignal) signals.push(cancelSignal);
    if (callbackInvoked) {
      const proceed = /* @__PURE__ */ Symbol();
      const settle = import_general.Time.sleep("cancel settlement", 0).then(() => proceed);
      const winner = await Promise.race([
        ...signals.map(
          (s) => s.then(
            () => void 0,
            () => void 0
          )
        ),
        settle
      ]);
      if (winner !== proceed) {
        sleepTimer?.cancel();
        this.#waiters.delete(waiter);
        return result;
      }
    }
    const queryAbort = new import_general.Abort();
    const discoveries = this.#startDiscovery(identifier, queryAbort);
    try {
      await Promise.race([...signals, discoveries]);
      sleepTimer?.cancel();
    } finally {
      queryAbort();
      await discoveries;
      this.#waiters.delete(waiter);
    }
    return result;
  }
  getDiscoveredCommissionableDevices(identifier) {
    return [...this.#cache.values()].map((cached) => refreshAddresses(cached)).filter((device) => matchesIdentifier(device, identifier) && device.addresses.length > 0);
  }
  cancelCommissionableDeviceDiscovery(identifier) {
    for (const waiter of this.#waiters) {
      if (matchesWaiterIdentifier(waiter.identifier, identifier)) {
        waiter.cancel();
      }
    }
  }
  #onDiscovered(name) {
    const lower = name.qname.toLowerCase();
    const matterc = lower === "_matterc._udp.local" || lower.endsWith("._matterc._udp.local");
    const matterd = lower === "_matterd._udp.local" || lower.endsWith("._matterd._udp.local");
    if (!matterc && !matterd) {
      return;
    }
    if (lower.startsWith("_")) {
      for (const record of name.records) {
        if (record.recordType !== import_general.DnsRecordType.PTR) {
          continue;
        }
        if (record.value.startsWith("_")) {
          continue;
        }
        const target = this.#names.get(record.value);
        let hasSrv2 = false;
        let hasTxt = false;
        for (const r of target.records) {
          if (r.recordType === import_general.DnsRecordType.SRV) {
            hasSrv2 = true;
          } else if (r.recordType === import_general.DnsRecordType.TXT) {
            hasTxt = true;
          }
          if (hasSrv2 && hasTxt) {
            break;
          }
        }
        const recordTypes2 = new Array();
        if (!hasSrv2) {
          recordTypes2.push(import_general.DnsRecordType.SRV);
        }
        if (!hasTxt) {
          recordTypes2.push(import_general.DnsRecordType.TXT);
        }
        if (recordTypes2.length) {
          const targetKey = record.value.toLowerCase();
          if (!this.#speculativeTargets.has(targetKey) && this.#speculativeTargets.size < SPECULATIVE_TARGET_MAX) {
            target.on(this.#speculativeObserver);
            const abort = new import_general.Abort({ abort: this.#scanAbort });
            this.#speculativeTargets.set(targetKey, { abort, createdAt: import_general.Time.nowMs });
            void this.#names.solicitor.discover({
              name: target,
              recordTypes: recordTypes2,
              abort: abort.signal,
              // Match the main PTR discovery cap so retries stay dense within commissioning windows
              retries: { maximumInterval: COMMISSIONING_RETRY_INTERVAL }
            }).catch((error) => {
              this.#speculativeTargets.delete(targetKey);
              target.off(this.#speculativeObserver);
              if (!(error instanceof import_general.AbortedError)) {
                logger.error(`Speculative discovery for ${target.qname} failed:`, error);
              }
            });
          }
        }
      }
      return;
    }
    this.#clearSpeculativeTarget(lower, name);
    if (this.#cache.has(lower)) {
      return;
    }
    const ipService = new import_general.IpService(name.qname, import_general.Diagnostic.via("commissionable-scanner"), this.#names);
    const device = buildCommissionableDevice(name);
    if (device !== void 0) {
      this.#cacheDevice(lower, device, ipService, name);
      return;
    }
    const hasSrv = [...name.records].some((r) => r.recordType === import_general.DnsRecordType.SRV);
    const recordTypes = hasSrv ? [import_general.DnsRecordType.TXT] : [import_general.DnsRecordType.SRV, import_general.DnsRecordType.TXT];
    this.#names.solicitor.solicit({ name, recordTypes });
    const pendingObserver = ({ name: changedName }) => {
      if (!changedName.isDiscovered) {
        this.#observers.off(name, pendingObserver);
        void ipService.close();
        return;
      }
      const built = buildCommissionableDevice(changedName);
      if (built !== void 0) {
        this.#observers.off(name, pendingObserver);
        this.#cacheDevice(lower, built, ipService, name);
      }
    };
    this.#observers.on(name, pendingObserver);
  }
  #cacheDevice(lower, device, ipService, name) {
    const observer = ({ name: changedName }) => {
      if (!changedName.isDiscovered) {
        const cached2 = this.#cache.get(lower);
        if (cached2) {
          this.#cache.delete(lower);
          this.#observers.off(name, cached2.observer);
          if (cached2.onAddresses) {
            this.#observers.off(ipService.changed, cached2.onAddresses);
            cached2.onAddresses = void 0;
          }
          void cached2.ipService.close();
        }
      }
    };
    const cached = { device, ipService, name, observer };
    this.#cache.set(lower, cached);
    this.#observers.on(name, observer);
    if (!this.#deliverDeviceIfResolved(cached)) {
      for (const record of name.records) {
        if (record.recordType !== import_general.DnsRecordType.SRV) {
          continue;
        }
        const hostname = this.#names.get(record.value.target);
        this.#names.solicitor.solicit({
          name: hostname,
          recordTypes: [import_general.DnsRecordType.A, import_general.DnsRecordType.AAAA]
        });
      }
      const onAddresses = () => {
        if (this.#deliverDeviceIfResolved(cached)) {
          this.#observers.off(ipService.changed, onAddresses);
          cached.onAddresses = void 0;
        }
      };
      cached.onAddresses = onAddresses;
      this.#observers.on(ipService.changed, onAddresses);
    }
  }
  #deliverDeviceIfResolved(cached) {
    const device = refreshAddresses(cached);
    if (device.addresses.length === 0) {
      return false;
    }
    for (const waiter of this.#waiters) {
      waiter.notify(device);
    }
    return true;
  }
  async #startDiscovery(identifier, abort) {
    const solicitor = this.#names.solicitor;
    const qnames = getQueryQnames(identifier);
    const discoveries = qnames.map(
      (qname) => solicitor.discover({
        name: this.#names.get(qname),
        recordTypes: [import_general.DnsRecordType.PTR],
        abort,
        // Cap backoff for commissioning: short-lived discovery shouldn't drift toward the 1h default
        retries: { maximumInterval: COMMISSIONING_RETRY_INTERVAL }
      })
    );
    await import_general.MatterAggregateError.allSettled(discoveries);
  }
}
function buildCommissionableDevice(name) {
  const params = name.parameters;
  const D = Number(params.get("D"));
  const CM = Number(params.get("CM"));
  if (!isFinite(D) || !isFinite(CM)) {
    return void 0;
  }
  const instanceId = name.qname.split(".")[0];
  const dd = (0, import_Scanner.DiscoveryData)(params);
  dd.T ??= 0;
  dd.ICD ??= 0;
  return {
    ...dd,
    deviceIdentifier: instanceId,
    D,
    CM,
    addresses: []
  };
}
function refreshAddresses(cached) {
  cached.device.addresses = [...cached.ipService.addresses];
  return cached.device;
}
function matchesIdentifier(device, identifier) {
  if ("instanceId" in identifier) {
    return device.deviceIdentifier === identifier.instanceId;
  }
  if ("longDiscriminator" in identifier) {
    return device.D === identifier.longDiscriminator;
  }
  if ("shortDiscriminator" in identifier) {
    return (device.D >> 8 & 15) === identifier.shortDiscriminator;
  }
  if ("vendorId" in identifier) {
    const vp = device.VP?.split("+");
    if (!vp) return false;
    const vendorMatch = (0, import_types.VendorId)(Number(vp[0])) === identifier.vendorId;
    if ("productId" in identifier && identifier.productId !== void 0) {
      return vendorMatch && Number(vp[1]) === identifier.productId;
    }
    return vendorMatch;
  }
  if ("deviceType" in identifier) {
    return device.DT === identifier.deviceType;
  }
  if ("productId" in identifier) {
    const vp = device.VP?.split("+");
    return vp ? Number(vp[1]) === identifier.productId : false;
  }
  return device.CM === 1 || device.CM === 2;
}
function matchesWaiterIdentifier(a, b) {
  const aKeys = Object.keys(a);
  const bKeys = Object.keys(b);
  if (aKeys.length !== bKeys.length) return false;
  for (const key of aKeys) {
    if (a[key] !== b[key]) return false;
  }
  return true;
}
function getQueryQnames(identifier) {
  const base = [import_MdnsConsts.MATTER_COMMISSION_SERVICE_QNAME];
  if ("instanceId" in identifier) {
    return [...base, (0, import_MdnsConsts.getCommissionableDeviceQname)(identifier.instanceId)];
  }
  if ("longDiscriminator" in identifier) {
    return [...base, (0, import_MdnsConsts.getLongDiscriminatorQname)(identifier.longDiscriminator)];
  }
  if ("shortDiscriminator" in identifier) {
    return [...base, (0, import_MdnsConsts.getShortDiscriminatorQname)(identifier.shortDiscriminator)];
  }
  if ("vendorId" in identifier) {
    return [...base, (0, import_MdnsConsts.getVendorQname)(identifier.vendorId)];
  }
  if ("deviceType" in identifier) {
    return [...base, (0, import_MdnsConsts.getDeviceTypeQname)(identifier.deviceType)];
  }
  return [...base, (0, import_MdnsConsts.getCommissioningModeQname)()];
}
