// @matter/protocol/dist/cjs/mdns/MdnsService.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MdnsService_exports = {};
__export(MdnsService_exports, {
  MdnsService: () => MdnsService
});
module.exports = __toCommonJS(MdnsService_exports);
var import_general = require("@matter/general");
var import_MdnsServer = require("../mdns/MdnsServer.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("MDNS");
class MdnsService {
  #entropy;
  #construction;
  #enableIpv4;
  limitedToNetInterface;
  #socket;
  #server;
  #names;
  get enableIpv4() {
    return this.#enableIpv4;
  }
  constructor(environment, options) {
    this.#entropy = environment.get(import_general.Entropy);
    const network = environment.get(import_general.Network);
    const rootEnvironment = environment.root;
    rootEnvironment.set(MdnsService, this);
    rootEnvironment.runtime.add(this);
    const vars = environment.get(import_general.VariableService);
    this.#enableIpv4 = vars.boolean("mdns.ipv4") ?? options?.ipv4 ?? true;
    this.limitedToNetInterface = vars.get("mdns.networkInterface", options?.networkInterface);
    this.#construction = (0, import_general.Construction)(this, async () => {
      this.#socket = await import_general.MdnsSocket.create(network, {
        lifetime: this.#construction,
        enableIpv4: this.enableIpv4,
        netInterface: this.limitedToNetInterface
      });
      this.#server = new import_MdnsServer.MdnsServer(this.#socket, this.#construction);
    });
  }
  static [import_general.Environmental.create](environment) {
    return new this(environment);
  }
  get server() {
    return this.#construction.assert("MDNS service", this.#server);
  }
  get names() {
    if (this.#names === void 0) {
      this.#names = new import_general.DnssdNames({
        socket: this.#construction.assert("MDNS socket", this.#socket),
        lifetime: this.#construction,
        entropy: this.#entropy,
        filter: ({ name }) => {
          const lower = name.toLowerCase();
          return lower.endsWith("._matter._tcp.local") || lower.endsWith("._matterc._udp.local") || lower.endsWith("._matterd._udp.local");
        }
      });
    }
    return this.#names;
  }
  get [import_general.Diagnostic.value]() {
    return "MDNS";
  }
  get construction() {
    return this.#construction;
  }
  async close() {
    await this.#construction.close(async () => {
      try {
        await import_general.MatterAggregateError.allSettled(
          [this.#server, this.#names].map((svc) => svc?.close()),
          "Error disposing MDNS services"
        );
      } catch (e) {
        logger.error(e);
      }
      if (this.#socket) {
        await this.#socket?.close();
      }
      this.#server = this.#names = void 0;
    });
  }
}
