// @matter/nodejs/dist/esm/storage/sqlite/platform/BunSqlite.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "@matter/general";
import { constants, Database } from "bun:sqlite";
const logger = Logger.get("BunSqlite");
function createBunDatabase(path) {
  const db = new Database(path, {
    strict: true,
    create: true
  });
  if (path === ":memory:") {
    return db;
  }
  db.exec("PRAGMA journal_mode = WAL");
  db.exec("PRAGMA synchronous = NORMAL");
  return {
    prepare: db.prepare.bind(db),
    exec: db.exec.bind(db),
    close: () => {
      try {
        db.fileControl(constants.SQLITE_FCNTL_PERSIST_WAL, 0);
      } catch (error) {
        logger.warn("Failed to clear WAL persistence flag:", error);
      }
      try {
        db.exec("PRAGMA wal_checkpoint(TRUNCATE)");
      } catch (error) {
        logger.warn("WAL checkpoint failed, WAL will be replayed on next open:", error);
      } finally {
        db.close();
      }
    }
  };
}
export {
  createBunDatabase
};
