// @matter/protocol/dist/esm/advertisement/mdns/CommissionerMdnsAdvertisement.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  getCommissionerDeviceQname,
  MATTER_COMMISSIONER_SERVICE_QNAME,
  SERVICE_DISCOVERY_QNAME
} from "#mdns/MdnsConsts.js";
import { PtrRecord } from "@matter/general";
import { MdnsAdvertisement } from "./MdnsAdvertisement.js";
class CommissionerMdnsAdvertisement extends MdnsAdvertisement {
  instanceId;
  constructor(advertiser, description) {
    const instanceId = advertiser.createInstanceId();
    const qname = getCommissionerDeviceQname(instanceId);
    super(advertiser, qname, description);
    this.instanceId = instanceId;
  }
  get ptrRecords() {
    const { deviceType, vendorId } = this.description;
    const vendorQname = `_V${vendorId}._sub.${MATTER_COMMISSIONER_SERVICE_QNAME}`;
    const records = [
      PtrRecord(SERVICE_DISCOVERY_QNAME, MATTER_COMMISSIONER_SERVICE_QNAME),
      PtrRecord(MATTER_COMMISSIONER_SERVICE_QNAME, vendorQname),
      PtrRecord(vendorQname, this.qname)
    ];
    if (deviceType !== void 0) {
      const deviceTypeQname = `_T${deviceType}._sub.${MATTER_COMMISSIONER_SERVICE_QNAME}`;
      records.push(PtrRecord(SERVICE_DISCOVERY_QNAME, deviceTypeQname));
      records.push(PtrRecord(deviceTypeQname, this.qname));
    }
    return records;
  }
  get txtValues() {
    const { vendorId, productId, deviceType, name: deviceName } = this.description;
    const values = {
      DN: deviceName,
      DT: deviceType
    };
    if (!this.isPrivacyMasked) {
      values.VP = `${vendorId}+${productId}`;
    }
    return values;
  }
  isDuplicate(other) {
    return other.isCommissioning();
  }
}
export {
  CommissionerMdnsAdvertisement
};
