// @matter/protocol/dist/esm/advertisement/mdns/MdnsAdvertisement.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Advertisement } from "#advertisement/Advertisement.js";
import { SupportedTransportsSchema } from "#common/SupportedTransportsBitmap.js";
import { SessionIntervals } from "#session/SessionIntervals.js";
import {
  AAAARecord,
  ARecord,
  Duration,
  Logger,
  SrvRecord,
  Time,
  Timestamp,
  TxtRecord
} from "@matter/general";
const logger = Logger.get("MdnsAdvertisement");
class MdnsAdvertisement extends Advertisement {
  /**
   * The device qname.
   */
  qname;
  #isPrivacyMasked;
  #stopAt;
  #broadcastPromise;
  constructor(advertiser, qname, description) {
    description = {
      ...description,
      ...SessionIntervals(description)
    };
    super(advertiser, `mdns:${qname}`, description, { omitPrivateDetails: advertiser.omitPrivateDetails });
    this.qname = qname;
    this.#isPrivacyMasked = this.isPrivacyMasked;
  }
  get txtValues() {
    return {};
  }
  async run(context, event = "startup") {
    this.#stopAt = void 0;
    let number = 0;
    for (const retryInterval of this.advertiser.broadcastScheduleFor(this, event)) {
      if (this.#stopAt !== void 0 && this.#stopAt <= Time.nowMs) {
        break;
      }
      number++;
      logger.debug("Broadcast", this.dict({ number, next: Duration.format(retryInterval) }));
      await this.broadcast();
      await context.sleep("MDNS repeat", retryInterval);
      if (context.cancelled) {
        break;
      }
    }
  }
  /**
   * Broadcast a single announcement immediately.
   */
  async broadcast() {
    if (this.#broadcastPromise !== void 0) {
      return this.#broadcastPromise;
    }
    this.#broadcastPromise = this.#broadcast().then(() => {
      this.#broadcastPromise = void 0;
    });
    return this.#broadcastPromise;
  }
  async #broadcast() {
    if (!this.#isPrivacyMasked && this.isPrivacyMasked) {
      this.#isPrivacyMasked = true;
      await this.advertiser.server.setRecordsGenerator(this.service, this.#recordsGenerator);
    }
    await this.advertiser.server.broadcast(this.service);
  }
  /**
   * Broadcast expiration announcement immediately.
   */
  async expire() {
    if (this.#broadcastPromise !== void 0) {
      await this.#broadcastPromise;
    }
    logger.info("Unpublishing", this.dict({ time: Duration.format(this.duration) }));
    await this.advertiser.server.expireAnnouncements(this.service);
  }
  serviceConnected() {
    const { broadcastAfterConnection } = this.advertiser.broadcastConfigFor(this);
    if (broadcastAfterConnection === void 0) {
      return;
    }
    if (broadcastAfterConnection <= 0) {
      this.stop();
      return;
    }
    this.#stopAt = Timestamp(Time.nowMs + broadcastAfterConnection);
  }
  serviceDisconnected() {
    this.start("reconnect");
  }
  async onCreate() {
    await Promise.resolve().then(() => {
      logger.info("Publishing", this.dict());
      return this.advertiser.server.setRecordsGenerator(this.service, this.#recordsGenerator);
    });
  }
  async onClose() {
    await this.expire();
  }
  get #recordsGenerator() {
    return (_intf, addrs) => this.#recordsFor(addrs);
  }
  #recordsFor(addrs) {
    const hostname = addrs.mac.replace(/:/g, "").toUpperCase() + "0000.local";
    const records = [
      ...this.ptrRecords,
      SrvRecord(this.qname, { priority: 0, weight: 0, port: this.advertiser.port, target: hostname }),
      TxtRecord(
        this.qname,
        Object.entries(this.#txtValues).filter(([, v]) => v !== void 0).map(([k, v]) => `${k}=${v}`)
      )
    ];
    for (const addr of addrs.ipV6) {
      records.push(AAAARecord(hostname, addr));
    }
    if (this.advertiser.server.supportsIpv4) {
      for (const addr of addrs.ipV4) {
        records.push(ARecord(hostname, addr));
      }
    }
    return records;
  }
  get #txtValues() {
    const values = {
      SII: this.description.idleInterval,
      SAI: this.description.activeInterval,
      SAT: this.description.activeThreshold,
      ...this.txtValues
    };
    if (this.description.tcp !== void 0) {
      values.T = SupportedTransportsSchema.encode(this.description.tcp);
    }
    if (this.description.icd !== void 0) {
      values.ICD = this.description.icd;
    }
    return values;
  }
}
export {
  MdnsAdvertisement
};
