// @matter/protocol/dist/esm/advertisement/mdns/MdnsAdvertiser.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Advertiser } from "#advertisement/Advertiser.js";
import {
  Bytes,
  ImplementationError,
  Instant,
  InternalError,
  RetrySchedule,
  Seconds,
  STANDARD_MATTER_PORT
} from "@matter/general";
import { DatatypeModel, FieldElement } from "@matter/model";
import { MAXIMUM_COMMISSIONING_TIMEOUT } from "@matter/types";
import { CommissionableMdnsAdvertisement } from "./CommissionableMdnsAdvertisement.js";
import { CommissionerMdnsAdvertisement } from "./CommissionerMdnsAdvertisement.js";
import { OperationalMdnsAdvertisement } from "./OperationalMdnsAdvertisement.js";
class MdnsAdvertiser extends Advertiser {
  constructor(crypto, server, options) {
    super(options?.lifetime);
    this.crypto = crypto;
    this.server = server;
    this.port = options?.port ?? STANDARD_MATTER_PORT;
    this.omitPrivateDetails = options?.omitPrivateDetails ?? false;
    let hasDefaultSchedule = false;
    let hasRetransmitSchedule = false;
    if (options?.schedules) {
      for (const schedule of options.schedules) {
        if (schedule.serviceKind === void 0) {
          if (schedule.event === void 0) {
            hasDefaultSchedule = true;
          }
          if (schedule.event === "retransmit") {
            hasRetransmitSchedule = true;
          }
        }
        this.#schedules.push({ config: schedule });
      }
    }
    if (!hasRetransmitSchedule) {
      this.#schedules.push({ config: MdnsAdvertiser.RetransmissionBroadcastSchedule });
    }
    if (!hasDefaultSchedule) {
      this.#schedules.push({ config: MdnsAdvertiser.DefaultBroadcastSchedule });
    }
  }
  crypto;
  server;
  port;
  omitPrivateDetails;
  #schedules = new Array();
  /**
   * Create an advertisement for the specific service.
   */
  getAdvertisement(description) {
    switch (description.kind) {
      case "operational":
        return this.advertisementForFabric(description.fabric) ?? new OperationalMdnsAdvertisement(this, description);
      case "commissionable":
        return new CommissionableMdnsAdvertisement(this, description);
      case "commissioner":
        return new CommissionerMdnsAdvertisement(this, description);
      default:
        throw new ImplementationError(
          `Unsupported service description kind "${description.kind}`
        );
    }
  }
  /**
   * Generate a random 8-byte Matter instance identifier.
   */
  createInstanceId() {
    return Bytes.toHex(this.crypto.randomBytes(8)).toUpperCase();
  }
  /**
   * Retrieve the broadcast schedule for a specific advertisement.
   */
  broadcastScheduleFor(advertisement, event) {
    const entry = this.#scheduleFor(advertisement, event);
    if (entry.schedule === void 0) {
      entry.schedule = new RetrySchedule(
        this.crypto,
        RetrySchedule.Configuration(MdnsAdvertiser.DefaultBroadcastSchedule, entry.config)
      );
    }
    return entry.schedule;
  }
  /**
   * Retrieve broadcast configuration for a specific advertisement.
   */
  broadcastConfigFor(advertisement, event) {
    return this.#scheduleFor(advertisement, event).config;
  }
  #scheduleFor(advertisement, event) {
    if (!event) {
      event = "startup";
    }
    for (const entry of this.#schedules) {
      const { serviceKind, event: event2 } = entry.config;
      if (serviceKind !== void 0 && serviceKind !== advertisement.description.kind) {
        continue;
      }
      if (event2 !== void 0 && event2 !== event) {
        continue;
      }
      return entry;
    }
    throw new InternalError("Default retry schedule not present");
  }
}
((MdnsAdvertiser2) => {
  MdnsAdvertiser2.DefaultBroadcastSchedule = {
    // Mandated by MDNS specification
    initialInterval: Seconds.one,
    // Maximum commissioning timeout per Matter specification 5.4.2.3.1, although
    timeout: MAXIMUM_COMMISSIONING_TIMEOUT,
    // Minimum per MDNS specification
    backoffFactor: 2,
    // Technically this may result in us emitting more than 8 packets which is the maximum per the MDNS
    // specification but extra packets will only come every 90 seconds
    maximumInterval: Seconds(90),
    // Not in any specification AFAIK but common sense to reduce thundering herd
    jitterFactor: 0.25,
    // Generally operational broadcast is unnecessary once an operational connection is established
    broadcastAfterConnection: Instant
  };
  MdnsAdvertiser2.RetransmissionBroadcastSchedule = {
    event: "retransmit",
    // Retransmission events occur for every packet retry so limit count to 1
    maximumCount: 1
  };
  MdnsAdvertiser2.OptionsSchema = new DatatypeModel(
    { name: "MdnsAdvertiserOptions", type: "struct" },
    FieldElement({ name: "omitPrivateDetails", type: "bool" }),
    FieldElement(
      { name: "schedules", type: "list" },
      FieldElement(
        { name: "entry", type: "struct" },
        FieldElement({ name: "initialInterval", type: "uint32" }),
        FieldElement({ name: "timeout", type: "uint32" }),
        FieldElement({ name: "backoffFactor", type: "uint8" }),
        FieldElement({ name: "maximumInterval", type: "uint32" }),
        FieldElement({ name: "jitterFactor", type: "single" }),
        FieldElement({ name: "serviceKind", type: "string" }),
        FieldElement({ name: "event", type: "string" })
      )
    )
  );
})(MdnsAdvertiser || (MdnsAdvertiser = {}));
export {
  MdnsAdvertiser
};
