// @matter/protocol/dist/esm/ble/Ble.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ChannelType, Observable } from "@matter/general";
import { PeerCommunicationError } from "../peer/PeerCommunicationError.js";
import { MatterBle } from "./BleConsts.js";
class BleError extends PeerCommunicationError {
}
class BleDisconnectedError extends BleError {
}
class BleChannelClosedError extends BleDisconnectedError {
}
class Ble {
}
class BleChannel {
  maxPayloadSize = MatterBle.MAX_MATTER_PAYLOAD_SIZE;
  isReliable = true;
  supportsLargeMessages = false;
  type = ChannelType.BLE;
  #closed = Observable();
  #closedFired = false;
  /**
   * Emitted exactly once when the channel is lost (peripheral disconnect, BTP session close,
   * or explicit {@link close}).  Consumers that hold a secure session over this channel use
   * this to force-close the session and reject pending exchanges without waiting for MRP
   * timeouts.
   *
   * Subclasses must call {@link emitClosed} from each of their termination paths.  The
   * latch guarantees exactly-once semantics regardless of which triggers fire.
   */
  get closed() {
    return this.#closed;
  }
  emitClosed() {
    if (this.#closedFired) {
      return;
    }
    this.#closedFired = true;
    this.#closed.emit();
  }
}
export {
  Ble,
  BleChannel,
  BleChannelClosedError,
  BleDisconnectedError,
  BleError
};
