// @matter/protocol/dist/cjs/certificate/kinds/AttestationCertificates.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AttestationCertificates_exports = {};
__export(AttestationCertificates_exports, {
  AttestationBaseCertificate: () => AttestationBaseCertificate,
  Dac: () => Dac,
  Paa: () => Paa,
  Pai: () => Pai
});
module.exports = __toCommonJS(AttestationCertificates_exports);
var import_Certificate = require("./Certificate.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class AttestationBaseCertificate extends import_Certificate.Certificate {
  /**
   * Sign the certificate using the provided crypto and key.
   * If the certificate is already signed, it throws a CertificateError.
   */
  async sign(crypto, key) {
    this.signature = await crypto.signEcdsa(key, this.asUnsignedDer());
  }
}
class Paa extends AttestationBaseCertificate {
  /** Construct the class from an ASN.1/DER encoded certificate */
  static fromAsn1(asn1) {
    const cert = import_Certificate.Certificate.parseAsn1Certificate(asn1, import_Certificate.Certificate.REQUIRED_PAA_EXTENSIONS);
    return new Paa(cert);
  }
}
class Pai extends AttestationBaseCertificate {
}
class Dac extends AttestationBaseCertificate {
}
