// @matter/nodejs/dist/esm/fs/fs-utils.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  FileNotFoundError,
  FileTypeError
} from "@matter/general";
import { createWriteStream } from "node:fs";
import { cp, readdir, rm, stat, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { finished } from "node:stream/promises";
async function nodeExists(path) {
  try {
    await stat(path);
    return true;
  } catch (e) {
    if (e.code === "ENOENT") {
      return false;
    }
    throw e;
  }
}
async function nodeStat(path) {
  let s;
  try {
    s = await stat(path);
  } catch (e) {
    if (e.code === "ENOENT") {
      throw new FileNotFoundError(`Not found: ${path}`);
    }
    throw e;
  }
  return {
    size: s.size,
    mtime: s.mtime,
    type: s.isDirectory() ? "directory" : "file"
  };
}
async function* nodeEntries(fs, dirPath, FileClass, DirClass) {
  const dirents = await readdir(dirPath, { withFileTypes: true });
  for (const dirent of dirents) {
    const fullPath = resolve(dirPath, dirent.name);
    const s = await stat(fullPath);
    const cached = {
      size: s.size,
      mtime: s.mtime,
      type: dirent.isDirectory() ? "directory" : "file"
    };
    if (dirent.isDirectory()) {
      yield new DirClass(fs, fullPath, dirent.name, cached);
    } else {
      yield new FileClass(fs, fullPath, dirent.name, cached);
    }
  }
}
function resolveCopyArg(basePath, arg) {
  if (typeof arg === "string") {
    return resolve(basePath, arg);
  }
  if ("path" in arg && typeof arg.path === "string") {
    return arg.path;
  }
  throw new FileTypeError("Cannot resolve path for copy argument");
}
async function nodeCopy(basePath, source, target) {
  const srcPath = resolveCopyArg(basePath, source);
  const dstPath = resolveCopyArg(basePath, target);
  await cp(srcPath, dstPath, { recursive: true });
}
async function nodeDelete(path) {
  await rm(path, { recursive: true, force: true });
}
function isBytes(value) {
  return ArrayBuffer.isView(value) || value instanceof ArrayBuffer || value instanceof SharedArrayBuffer;
}
function toBytes(value) {
  return value instanceof Uint8Array ? value : new Uint8Array(ArrayBuffer.isView(value) ? value.buffer : value);
}
async function streamWrite(ws, chunk) {
  if (!ws.write(chunk)) {
    await new Promise((resolve2) => ws.once("drain", resolve2));
  }
}
async function writeData(path, data) {
  if (typeof data === "string") {
    await writeFile(path, data, "utf8");
    return;
  }
  if (isBytes(data)) {
    await writeFile(path, toBytes(data));
    return;
  }
  const iter = Symbol.asyncIterator in data ? data[Symbol.asyncIterator]() : data[Symbol.iterator]();
  const first = await iter.next();
  if (first.done) {
    await writeFile(path, new Uint8Array(0));
    return;
  }
  const isText = typeof first.value === "string";
  const ws = createWriteStream(path, isText ? { encoding: "utf8" } : void 0);
  try {
    if (isText) {
      await streamWrite(ws, first.value);
      while (true) {
        const next = await iter.next();
        if (next.done) break;
        await streamWrite(ws, "\n" + next.value);
      }
    } else {
      await streamWrite(ws, toBytes(first.value));
      while (true) {
        const next = await iter.next();
        if (next.done) break;
        await streamWrite(ws, toBytes(next.value));
      }
    }
    ws.end();
    await finished(ws);
  } catch (e) {
    ws.destroy();
    throw e;
  }
}
export {
  isBytes,
  nodeCopy,
  nodeDelete,
  nodeEntries,
  nodeExists,
  nodeStat,
  resolveCopyArg,
  toBytes,
  writeData
};
