// @matter/nodejs/dist/esm/fs/lock-utils.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger, StorageLockError } from "@matter/general";
import { randomBytes } from "node:crypto";
import { open as fsOpen, readFile, stat, unlink, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
const logger = Logger.get("NodeJsDirectoryLock");
const LOCK_FILE = "matter.lock";
const PID_FILE = "matter.pid";
const PROCESS_TOKEN = randomBytes(8).toString("hex");
async function acquireDirectoryLock(dirPath, dirName) {
  const lockPath = resolve(dirPath, LOCK_FILE);
  const pidPath = resolve(dirPath, PID_FILE);
  if (!await dirExists(dirPath)) {
    return async () => {
    };
  }
  await acquireLock(lockPath, pidPath);
  await writeFile(pidPath, `${process.pid} ${PROCESS_TOKEN}`);
  logger.debug("Acquired storage lock for", dirName, "pid", process.pid);
  return async () => {
    await safeUnlink(pidPath);
    await safeUnlink(lockPath);
    logger.debug("Released storage lock for", dirName);
  };
}
async function acquireLock(lockPath, pidPath) {
  try {
    const fd = await fsOpen(lockPath, "wx");
    await fd.close();
  } catch (error) {
    if (error.code !== "EEXIST") {
      throw error;
    }
    const info = await readLockInfo(pidPath);
    if (isStale(info)) {
      logger.info("Cleaning stale storage lock");
      await safeUnlink(pidPath);
      await safeUnlink(lockPath);
      try {
        const fd = await fsOpen(lockPath, "wx");
        await fd.close();
      } catch (retryError) {
        if (retryError.code === "EEXIST") {
          throw new StorageLockError("Storage is locked by another process (lock reclaimed during retry)");
        }
        throw retryError;
      }
    } else if (info?.pid === process.pid) {
      throw new StorageLockError("Storage is already locked by this process");
    } else {
      throw new StorageLockError(`Storage is locked by another process (pid ${info?.pid})`);
    }
  }
}
function isStale(info) {
  if (info === void 0) {
    return true;
  }
  if (info.pid === process.pid) {
    return info.token !== PROCESS_TOKEN;
  }
  try {
    process.kill(info.pid, 0);
    return false;
  } catch (error) {
    if (error.code === "ESRCH") {
      return true;
    }
    return false;
  }
}
async function readLockInfo(pidPath) {
  try {
    const content = await readFile(pidPath, "utf-8");
    const parts = content.trim().split(/\s+/);
    const pid = parseInt(parts[0], 10);
    if (!Number.isFinite(pid) || pid <= 0) {
      return void 0;
    }
    return { pid, token: parts[1] };
  } catch {
    return void 0;
  }
}
async function dirExists(path) {
  try {
    const s = await stat(path);
    return s.isDirectory();
  } catch (e) {
    if (e.code === "ENOENT") {
      return false;
    }
    throw e;
  }
}
async function safeUnlink(path) {
  try {
    await unlink(path);
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }
}
export {
  acquireDirectoryLock
};
