// @matter/nodejs/dist/esm/fs/NodeJsDirectory.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Directory } from "@matter/general";
import { rename as fsRename, mkdir, rm, stat } from "node:fs/promises";
import { basename, dirname, resolve } from "node:path";
import { nodeCopy, nodeEntries, nodeStat } from "./fs-utils.js";
import { acquireDirectoryLock } from "./lock-utils.js";
import { NodeJsFile } from "./NodeJsFile.js";
class NodeJsDirectory extends Directory {
  #fs;
  #path;
  #name;
  #cachedStat;
  constructor(fs, path, name, cachedStat) {
    super();
    this.#fs = fs;
    this.#path = path;
    this.#name = name;
    this.#cachedStat = cachedStat;
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#name;
  }
  get path() {
    return this.#path;
  }
  async exists() {
    try {
      const s = await stat(this.#path);
      return s.isDirectory();
    } catch (e) {
      if (e.code === "ENOENT") {
        return false;
      }
      throw e;
    }
  }
  stat() {
    if (this.#cachedStat) {
      return this.#cachedStat;
    }
    return nodeStat(this.#path);
  }
  async rename(newName) {
    const newPath = resolve(dirname(this.#path), newName);
    await fsRename(this.#path, newPath);
    this.#path = newPath;
    this.#name = basename(newPath);
  }
  async delete() {
    await rm(this.#path, { recursive: true, force: true });
  }
  async *entries() {
    yield* nodeEntries(this.#fs, this.#path, NodeJsFile, NodeJsDirectory);
  }
  file(name) {
    return new NodeJsFile(this.#fs, resolve(this.#path, name), name);
  }
  directory(name) {
    return new NodeJsDirectory(this.#fs, resolve(this.#path, name), name);
  }
  async mkdir() {
    await mkdir(this.#path, { recursive: true });
  }
  async copy(source, target) {
    await nodeCopy(this.#path, source, target);
  }
  async lock() {
    await this.mkdir();
    return acquireDirectoryLock(this.#path, this.#name);
  }
}
export {
  NodeJsDirectory
};
