// @matter/nodejs/dist/esm/fs/NodeJsFilesystem.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { FileTypeError, Filesystem } from "@matter/general";
import { mkdir, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { resolve } from "node:path";
import { NodeJsDirectory } from "./NodeJsDirectory.js";
import { NodeJsFile } from "./NodeJsFile.js";
import { nodeCopy, nodeEntries, nodeExists, nodeStat } from "./fs-utils.js";
class NodeJsFilesystem extends Filesystem {
  #rootPath;
  #tempCounter = 0;
  constructor(workingDirectory) {
    super();
    this.#rootPath = workingDirectory;
  }
  get name() {
    return "";
  }
  get path() {
    return typeof this.#rootPath === "function" ? this.#rootPath() : this.#rootPath;
  }
  async exists() {
    return nodeExists(this.path);
  }
  stat() {
    return nodeStat(this.path);
  }
  rename() {
    throw new FileTypeError("Cannot rename root");
  }
  async delete() {
    await rm(this.path, { recursive: true, force: true });
  }
  async *entries() {
    yield* nodeEntries(this, this.path, NodeJsFile, NodeJsDirectory);
  }
  file(name) {
    return new NodeJsFile(this, resolve(this.path, name), name);
  }
  directory(name) {
    return new NodeJsDirectory(this, resolve(this.path, name), name);
  }
  async mkdir() {
    await mkdir(this.path, { recursive: true });
  }
  async copy(source, target) {
    await nodeCopy(this.path, source, target);
  }
  tempFilename() {
    return resolve(tmpdir(), `matter-${process.pid}-${Date.now()}-${this.#tempCounter++}`);
  }
  tempDirectory() {
    return new NodeJsDirectory(this, this.tempFilename(), "");
  }
}
export {
  NodeJsFilesystem
};
