// @matter/nodejs/dist/esm/fs/NodeJsFile.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  File,
  FileHandleTracker,
  FileTypeError
} from "@matter/general";
import { createReadStream } from "node:fs";
import { open as fsOpen, rename as fsRename, mkdir, rm } from "node:fs/promises";
import { basename, dirname, resolve } from "node:path";
import { nodeExists, nodeStat, toBytes, writeData } from "./fs-utils.js";
class NodeJsFile extends File {
  #fs;
  #path;
  #name;
  #cachedStat;
  constructor(fs, path, name, cachedStat) {
    super();
    this.#fs = fs;
    this.#path = path;
    this.#name = name;
    this.#cachedStat = cachedStat;
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#name;
  }
  get path() {
    return this.#path;
  }
  async exists() {
    return nodeExists(this.#path);
  }
  stat() {
    if (this.#cachedStat) {
      return this.#cachedStat;
    }
    return nodeStat(this.#path);
  }
  async *readBytes() {
    const s = await nodeStat(this.#path);
    if (s.type === "directory") {
      throw new FileTypeError("Cannot read bytes from a directory");
    }
    const stream = createReadStream(this.#path);
    try {
      for await (const chunk of stream) {
        yield chunk;
      }
    } finally {
      stream.destroy();
    }
  }
  async *readText(options) {
    const s = await nodeStat(this.#path);
    if (s.type === "directory") {
      throw new FileTypeError("Cannot read text from a directory");
    }
    if (options?.lines) {
      const stream = createReadStream(this.#path, { encoding: "utf8" });
      try {
        let carry = "";
        for await (const chunk of stream) {
          const parts = (carry + chunk).split("\n");
          carry = parts.pop();
          yield* parts;
        }
        yield carry;
      } finally {
        stream.destroy();
      }
    } else {
      const stream = createReadStream(this.#path, { encoding: "utf8" });
      try {
        for await (const chunk of stream) {
          yield chunk;
        }
      } finally {
        stream.destroy();
      }
    }
  }
  async write(data) {
    await writeData(this.#path, data);
  }
  async open(purpose, mode) {
    const m = mode ?? "r";
    const flags = m === "w" ? "w" : m === "a" ? "a" : "r";
    if (flags === "w" || flags === "a") {
      await mkdir(dirname(this.#path), { recursive: true });
    }
    const fh = await fsOpen(this.#path, flags);
    return new NodeJsFileHandle({ fs: this.#fs, path: this.#path, name: this.#name, purpose, fh });
  }
  async rename(newName) {
    const newPath = resolve(dirname(this.#path), newName);
    await fsRename(this.#path, newPath);
    this.#path = newPath;
    this.#name = basename(newPath);
  }
  async delete() {
    await rm(this.#path, { recursive: true, force: true });
  }
}
class NodeJsFileHandle extends File.Handle {
  #fs;
  #path;
  #name;
  purpose;
  #fh;
  constructor(options) {
    super();
    this.#fs = options.fs;
    this.#path = options.path;
    this.#name = options.name;
    this.purpose = options.purpose;
    this.#fh = options.fh;
    FileHandleTracker.register(this);
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#name;
  }
  get path() {
    return this.#path;
  }
  async exists() {
    return nodeExists(this.#path);
  }
  stat() {
    return nodeStat(this.#path);
  }
  async *readBytes() {
    const stream = createReadStream(this.#path);
    try {
      for await (const chunk of stream) {
        yield chunk;
      }
    } finally {
      stream.destroy();
    }
  }
  async *readText(options) {
    if (options?.lines) {
      const stream = createReadStream(this.#path, { encoding: "utf8" });
      try {
        let carry = "";
        for await (const chunk of stream) {
          const parts = (carry + chunk).split("\n");
          carry = parts.pop();
          yield* parts;
        }
        yield carry;
      } finally {
        stream.destroy();
      }
    } else {
      const stream = createReadStream(this.#path, { encoding: "utf8" });
      try {
        for await (const chunk of stream) {
          yield chunk;
        }
      } finally {
        stream.destroy();
      }
    }
  }
  async write(data) {
    await writeData(this.#path, data);
  }
  async open(_purpose) {
    return this;
  }
  async writeHandle(data) {
    if (typeof data === "string") {
      await this.#fh.write(data, void 0, "utf8");
    } else {
      await this.#fh.write(toBytes(data));
    }
  }
  cursor(max, bufferSize) {
    return new NodeJsCursor(this.#fh, max, bufferSize);
  }
  async truncate(size) {
    await this.#fh.truncate(size ?? 0);
  }
  async fsync() {
    await this.#fh.datasync();
  }
  async close() {
    FileHandleTracker.unregister(this);
    await this.#fh.close();
  }
  async rename(newName) {
    const newPath = resolve(dirname(this.#path), newName);
    await fsRename(this.#path, newPath);
    this.#path = newPath;
    this.#name = basename(newPath);
  }
  async delete() {
    await rm(this.#path, { recursive: true, force: true });
  }
}
class NodeJsCursor extends File.Cursor {
  #fh;
  #shared;
  constructor(fh, max, bufferSize) {
    super(max);
    this.#fh = fh;
    this.#shared = Buffer.alloc(bufferSize ?? 8192);
  }
  async readAt(position, length, copy) {
    if (copy || length > this.#shared.length) {
      const buf = Buffer.alloc(length);
      const { bytesRead: bytesRead2 } = await this.#fh.read(buf, 0, length, position);
      return bytesRead2 < length ? buf.subarray(0, bytesRead2) : buf;
    }
    const { bytesRead } = await this.#fh.read(this.#shared, 0, length, position);
    return this.#shared.subarray(0, bytesRead);
  }
}
export {
  NodeJsFile
};
