// @matter/nodejs/dist/esm/net/NodeJsHttpEndpoint.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { AppAddress, asError, HttpEndpointFactory, Logger, NetworkError } from "@matter/general";
import { existsSync, ReadStream, rmSync, statSync } from "node:fs";
import { createServer as createHttpServer, IncomingMessage } from "node:http";
import { createServer as createHttpsServer } from "node:https";
import { normalize } from "node:path";
import { WsAdapter } from "./WsAdapter.js";
const logger = new Logger("NodeJsHttpEndpoint");
class NodeJsHttpEndpoint {
  #server;
  #ready;
  #http;
  #httpListener;
  #ws;
  #wsListener;
  #notFound;
  #wsAdapter;
  #wsAdapterFactory;
  static async create(config) {
    const endpoint = new NodeJsHttpEndpoint(config);
    await endpoint.ready;
    return endpoint;
  }
  /**
   * Create a new endpoint.
   *
   * You may pass an existing {@link Server} or pass {@link HttpEndpoint.Configuration} to create a server dedicated
   * to this endpoint.
   */
  constructor(config) {
    let close, ready, server, notFound;
    if ("on" in config) {
      ({ close, ready, server, notFound } = this.#bindToServer(config));
    } else {
      ({ close, ready, server, notFound } = this.#createDedicatedServer(config));
    }
    this.#server = server;
    this.#ready = ready;
    this.close = close;
    this.#notFound = notFound;
  }
  get server() {
    return this.#server;
  }
  #bindToServer(server) {
    return {
      server,
      ready: Promise.resolve(),
      close: async () => {
        this.http = void 0;
        this.ws = void 0;
      },
      notFound: () => void 0
    };
  }
  #createDedicatedServer(options) {
    const address = AppAddress.for(options.address);
    let server;
    if (address.isTls) {
      if (options.certificate === void 0 || options.key === void 0) {
        throw new NetworkError(
          `Cannot create HTTPS endpoint for ${address} because no certificate and/or key is present`
        );
      }
      server = createHttpsServer({
        keepAlive: true,
        cert: options.certificate,
        key: options.key
      });
    } else {
      server = createHttpServer({ keepAlive: true });
    }
    const opts = {};
    const { transport } = address;
    switch (transport.kind) {
      case "ip":
        if (!address.isWildcardHost) {
          opts.host = address.hostname;
        }
        if (!address.isWildcardPort) {
          opts.port = address.portNum;
        }
        break;
      case "unix":
        opts.path = normalize(decodeURIComponent(address.hostname));
        if (existsSync(opts.path)) {
          if (statSync(opts.path).isSocket()) {
            try {
              rmSync(opts.path);
            } catch (e) {
              throw new NetworkError(
                `Error deleting previous socket at ${opts.path}: ${asError(e).message}`
              );
            }
          } else {
            throw new NetworkError(`UNIX socket path ${opts.path} exists and is not a socket`);
          }
        }
        break;
      default:
        throw new NetworkError(
          `Unsupported address type "${options.address?.type}" for HTTP endpoint`
        );
    }
    server.listen(opts);
    return {
      server,
      ready: new Promise((resolve, reject) => {
        let settled = false;
        server.once("listening", () => {
          if (settled) {
            return;
          }
          settled = true;
          resolve();
        });
        server.on("error", (error) => {
          if (settled) {
            logger.warn("HTTP server error:", error.message);
            return;
          }
          settled = true;
          reject(error);
        });
      }),
      close: async () => {
        return new Promise((resolve, reject) => {
          server.close((err) => {
            if (err) {
              reject(err);
              return;
            }
            resolve();
          });
        });
      },
      notFound: (res) => respondError(res, 404)
    };
  }
  get ready() {
    return this.#ready;
  }
  set http(handler) {
    this.#http = handler;
    if (!this.#http) {
      if (this.#httpListener) {
        this.#server.off("request", this.#httpListener);
      }
      return;
    }
    if (this.#httpListener) {
      return;
    }
    this.#httpListener = (req, res) => {
      this.#handleHttp(req, res).catch((error) => {
        logger.error("Unhandled error in HTTP endpoint handler", error);
        respondError(res, 500);
      });
    };
    this.#server.on("request", this.#httpListener);
  }
  set ws(handler) {
    this.#ws = handler;
    if (!this.#ws) {
      if (this.#wsListener) {
        this.#server.off("upgrade", this.#wsListener);
      }
      return;
    }
    let adapter = this.#wsAdapter;
    if (!adapter) {
      const factory = this.#wsAdapterFactory ?? WsAdapter.defaultFactory;
      if (!factory) {
        logger.warn(
          "WebSocket support disabled because no adapter is installed; please import @matter/nodejs-ws or equivalent"
        );
        return;
      }
      adapter = this.#wsAdapter = factory();
    }
    this.#wsListener = (req, socket, head) => {
      this.#handleUpgrade(adapter, req, socket, head).catch((error) => {
        logger.error("Unhandled error WebSocket endpoint", error);
      });
    };
    this.#server.on("upgrade", this.#wsListener);
  }
  close;
  async #handleHttp(req, res) {
    if (!this.#http) {
      return;
    }
    const request = new NodeJsHttpRequest(req);
    const response = await this.#http(request);
    if (!response) {
      this.#notFound(res);
      return;
    }
    res.statusCode = response.status;
    res.statusMessage = response.statusText;
    response.headers.forEach(([name, value]) => res.appendHeader(name, value));
    if (response.body === null) {
      res.end();
      return;
    }
    const nodeBodyStream = ReadStream.fromWeb(response.body);
    nodeBodyStream.on("error", (error) => {
      logger.error("Error transmitting HTTP body", error);
      respondError(res, 500);
    });
    nodeBodyStream.pipe(res);
  }
  async #handleUpgrade(adapter, req, socket, head) {
    if (req.headers.upgrade !== "websocket") {
      socket.destroy();
      return;
    }
    if (!this.#ws) {
      socket.destroy();
      return;
    }
    const request = new NodeJsHttpRequest(req);
    try {
      await this.#ws(request, async () => {
        return adapter.handle(req, socket, head);
      });
    } finally {
      if (!socket.destroyed) {
        socket.destroy();
      }
    }
  }
}
class NodeJsHttpRequest extends Request {
  constructor(message) {
    const { method, rawHeaders } = message;
    const url = `http://${message.headers.host ?? "unknown"}${message.url ?? "/"}`;
    const headers = new Headers();
    for (let i = 0; i < message.rawHeaders.length; i += 2) {
      headers.append(rawHeaders[i], rawHeaders[i + 1]);
    }
    const init = {
      method,
      headers,
      duplex: "half"
      // Not in RequestInit type but required by node
    };
    if (method !== "GET" && method !== "HEAD") {
      init.body = IncomingMessage.toWeb(message);
    }
    super(url, init);
  }
}
function respondError(res, code) {
  if (res.closed) {
    return;
  }
  try {
    if (!res.headersSent) {
      res.statusCode = code;
      res.setHeader("Content-Type", "text/plain");
      res.end(`HTTP error ${code}
`);
    } else {
      res.end();
    }
  } catch (e) {
    logger.warn(`Error conveying ${code} error:`, asError(e).message);
  }
}
((NodeJsHttpEndpoint2) => {
  class Factory extends HttpEndpointFactory {
    async create(config) {
      return NodeJsHttpEndpoint2.create(config);
    }
  }
  NodeJsHttpEndpoint2.Factory = Factory;
})(NodeJsHttpEndpoint || (NodeJsHttpEndpoint = {}));
export {
  NodeJsHttpEndpoint
};
