// @matter/nodejs/dist/esm/net/NodeJsNetwork.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  Cache,
  InterfaceType,
  isIPv6,
  Logger,
  Minutes,
  Network,
  NetworkError,
  onSameNetwork,
  TCP_CONNECTION_TIMEOUT_MS
} from "@matter/general";
import { createConnection } from "node:net";
import { networkInterfaces } from "node:os";
import { NodeJsTcpConnection } from "./NodeJsTcpConnection.js";
import { NodeJsTcpListener } from "./NodeJsTcpListener.js";
import { NodeJsUdpSocket } from "./NodeJsUdpSocket.js";
const logger = Logger.get("NetworkNode");
class NodeJsNetwork extends Network {
  static getMulticastInterfaceIpv4(netInterface) {
    const netInterfaceInfo = networkInterfaces()[netInterface];
    if (netInterfaceInfo === void 0) throw new NetworkError(`Unknown interface: ${netInterface}`);
    for (const info of netInterfaceInfo) {
      if (familyIs(4, info)) {
        return info.address;
      }
    }
    return void 0;
  }
  static getMembershipMulticastInterfaces(netInterfaceOrZone, ipv4) {
    if (ipv4) {
      return [void 0];
    } else {
      const multicastInterfaces = Object.entries(networkInterfaces()).flatMap(([netIf, netIfInfo]) => {
        if (netIfInfo === void 0) {
          return [];
        }
        const zone = this.getNetInterfaceZoneIpv6Internal(netIf, netIfInfo);
        if (zone === void 0) {
          return [];
        }
        if (netInterfaceOrZone !== void 0 && netInterfaceOrZone !== zone) {
          return [];
        }
        return [`::%${zone}`];
      });
      if (multicastInterfaces.length === 0) {
        logger.warn(
          `No IPv6 multicast interface found${netInterfaceOrZone !== void 0 ? ` for interface ${netInterfaceOrZone}` : ""}.`
        );
      }
      return multicastInterfaces;
    }
  }
  static getNetInterfaceZoneIpv6(netInterface) {
    const netInterfaceInfo = networkInterfaces()[netInterface];
    if (netInterfaceInfo === void 0) throw new NetworkError(`Unknown interface: ${netInterface}`);
    return this.getNetInterfaceZoneIpv6Internal(netInterface, netInterfaceInfo);
  }
  /**
   * Returns the zone-scoped name of the first non-internal IPv6 interface.
   * Used as a fallback for multicast sends when no interface has been observed
   * from inbound traffic (e.g. when all sessions run over TCP).
   */
  static getDefaultNetInterface() {
    const interfaces = networkInterfaces();
    for (const name in interfaces) {
      const infos = interfaces[name];
      if (infos.some((info) => familyIs(6, info) && !info.internal)) {
        return this.getNetInterfaceZoneIpv6Internal(name, infos);
      }
    }
    return void 0;
  }
  static getNetInterfaceForIp(ip) {
    return this.netInterfaces.get(ip);
  }
  static netInterfaces = new Cache(
    "Network interface",
    (ip) => this.getNetInterfaceForRemoteAddress(ip),
    Minutes(5)
  );
  async close() {
    await NodeJsNetwork.netInterfaces.close();
  }
  static getNetInterfaceForRemoteAddress(ip) {
    if (ip.includes("%")) {
      return ip.split("%")[1];
    } else {
      const interfaces = networkInterfaces();
      for (const name in interfaces) {
        const netInterfaces = interfaces[name];
        for (const { address, netmask } of netInterfaces) {
          if (onSameNetwork(ip, address, netmask)) {
            return this.getNetInterfaceZoneIpv6Internal(name, netInterfaces);
          }
        }
      }
      if (isIPv6(ip)) {
        if (ip.startsWith("fd")) {
          return "";
        }
      }
      return void 0;
    }
  }
  static getNetInterfaceZoneIpv6Internal(netInterface, netInterfaceInfos) {
    if (process.platform !== "win32") {
      return netInterface;
    }
    if (netInterfaceInfos === void 0) return void 0;
    return netInterfaceInfos.find((info) => familyIs(6, info) && info.address.startsWith("fe80::"))?.scopeid?.toString();
  }
  /**
   * Get all network interfaces.
   * The optional configuration parameter allows to map interface names to types if this mapping is known.
   * Each network interface which has no mapped type is returned as Ethernet for now.
   *
   * @param configuration - An array of objects with the name and type properties.
   */
  getNetInterfaces(configuration = []) {
    const result = new Array();
    const interfaces = networkInterfaces();
    for (const name in interfaces) {
      const netInterfaces = interfaces[name];
      if (netInterfaces.length === 0) continue;
      if (netInterfaces[0].internal) continue;
      let type = InterfaceType.Ethernet;
      if (configuration.length > 0) {
        const nameType = configuration.find(({ name: mapName }) => name === mapName);
        if (nameType !== void 0 && nameType.type !== void 0) {
          type = nameType.type;
        }
      }
      result.push({ name, type });
    }
    return result;
  }
  getIpMac(netInterface) {
    const netInterfaceInfo = networkInterfaces()[netInterface];
    if (netInterfaceInfo === void 0) return void 0;
    const ipV4 = netInterfaceInfo.filter((info) => familyIs(4, info)).map(({ address }) => address);
    const ipV6 = netInterfaceInfo.filter((info) => familyIs(6, info)).map(({ address }) => address);
    return { mac: netInterfaceInfo[0].mac, ipV4, ipV6 };
  }
  createUdpSocket(options) {
    return NodeJsUdpSocket.create(options);
  }
  createTcpListener(options) {
    return NodeJsTcpListener.create(options);
  }
  async connectTcp(host, port, options) {
    return new Promise((resolve, reject) => {
      let settled = false;
      const settle = (fn) => {
        if (!settled) {
          settled = true;
          fn();
        }
      };
      const socket = createConnection({ host, port, noDelay: true }, () => {
        socket.setTimeout(0);
        socket.off("error", rejectOnce);
        settle(() => resolve(new NodeJsTcpConnection(socket)));
      });
      const rejectOnce = (error) => settle(() => {
        socket.destroy();
        reject(error);
      });
      socket.setTimeout(options?.timeout ?? TCP_CONNECTION_TIMEOUT_MS);
      socket.once("timeout", () => {
        socket.destroy();
        settle(() => reject(new NetworkError("TCP connection timeout")));
      });
      socket.on("error", rejectOnce);
    });
  }
}
function familyIs(version, { family }) {
  return family === `IPv${version}`;
}
export {
  NodeJsNetwork
};
