// @matter/nodejs/dist/esm/net/NodeJsUdpChannel.js
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  AddressInUseError,
  AddressUnreachableError,
  BindError,
  Bytes,
  ChannelType,
  createPromise,
  Diagnostic,
  ImplementationError,
  isIPv4,
  isIPv6,
  Lifetime,
  Logger,
  MAX_UDP_MESSAGE_SIZE,
  Millis,
  NetworkError,
  NetworkUnreachableError,
  NoAddressAvailableError,
  repackErrorAs,
  Seconds,
  Time
} from "@matter/general";
import * as dgram from "node:dgram";
import { NodeJsNetwork } from "./NodeJsNetwork.js";
const logger = Logger.get("NodejsChannel");
const UDP_SEND_TIMEOUT_CHECK_INTERVAL = Seconds.one;
function createDgramSocket(host, port, options) {
  const socket = dgram.createSocket(options);
  return new Promise((resolve, reject) => {
    const handleBindError = (error) => {
      try {
        socket.close();
      } catch (error2) {
        logger.debug("Error closing socket:", error2);
      }
      const code = error?.code;
      let desc = `${host ? host : options.type === "udp4" ? "0.0.0.0" : "{::}"}`;
      if (port !== void 0) {
        desc = `${desc}:${port}`;
      }
      switch (code) {
        case "EADDRINUSE":
          error = new AddressInUseError(`Cannot bind ${desc} because port is already in use`);
          break;
        default:
          error = new BindError(`Cannot bind to ${desc} (code ${code})`);
          break;
      }
      reject(error);
    };
    socket.on("error", handleBindError);
    socket.bind(port, host, () => {
      const { address: localHost, port: localPort } = socket.address();
      logger.debug(
        "Socket created and bound ",
        Diagnostic.dict({
          remoteAddress: `${host}:${port}`,
          localAddress: `${localHost}:${localPort}`
        })
      );
      socket.removeListener("error", handleBindError);
      socket.on("error", (error) => logger.error(error));
      resolve(socket);
    });
  });
}
class NodeJsUdpChannel {
  #lifetime;
  #type;
  #socket;
  #netInterface;
  #observedInterface;
  static async create({
    lifetime: lifetimeOwner,
    listeningPort,
    type,
    listeningAddress,
    netInterface,
    reuseAddress
  }) {
    var _stack2 = [];
    try {
      const name = `${listeningAddress?.includes(":") ? `[${listeningAddress}]` : listeningAddress ?? "*"}:${listeningPort}`;
      const lifetime = __using(_stack2, (lifetimeOwner ?? Lifetime.process).join("socket", Diagnostic.strong(name)));
      lifetime.details.intf = netInterface;
      let dgramType;
      switch (type) {
        case "udp":
        case "udp6":
          dgramType = "udp6";
          break;
        case "udp4":
          dgramType = "udp4";
          break;
        default:
          throw new ImplementationError(`Unrecognized UDP socket type ${type}`);
      }
      const socketOptions = { type: dgramType };
      if (type === "udp6") {
        socketOptions.ipv6Only = true;
      }
      if (reuseAddress) {
        socketOptions.reuseAddr = true;
      }
      let socket;
      {
        var _stack = [];
        try {
          const _creating = __using(_stack, lifetime.join("creating"));
          socket = await createDgramSocket(listeningAddress, listeningPort, socketOptions);
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          __callDispose(_stack, _error, _hasError);
        }
      }
      socket.setBroadcast(true);
      let netInterfaceZone;
      if (netInterface === void 0 && listeningAddress !== void 0) {
        netInterfaceZone = NodeJsNetwork.getNetInterfaceForIp(listeningAddress) || void 0;
      }
      if (netInterface !== void 0) {
        netInterfaceZone = NodeJsNetwork.getNetInterfaceZoneIpv6(netInterface);
        let multicastInterface;
        if (type === "udp4") {
          multicastInterface = NodeJsNetwork.getMulticastInterfaceIpv4(netInterface);
          if (multicastInterface === void 0) {
            throw new NoAddressAvailableError(`No IPv4 addresses on interface "${netInterface}"`);
          }
        } else {
          if (netInterfaceZone === void 0) {
            throw new NoAddressAvailableError(`No IPv6 addresses on interface "${netInterface}"`);
          }
          multicastInterface = `::%${netInterfaceZone}`;
        }
        logger.debug(
          "Initialize multicast",
          Diagnostic.dict({
            address: `${multicastInterface}:${listeningPort}`,
            interface: netInterface,
            type
          })
        );
        socket.setMulticastInterface(multicastInterface);
      }
      return new NodeJsUdpChannel(lifetime, type, socket, netInterfaceZone);
    } catch (_2) {
      var _error2 = _2, _hasError2 = true;
    } finally {
      __callDispose(_stack2, _error2, _hasError2);
    }
  }
  maxPayloadSize = MAX_UDP_MESSAGE_SIZE;
  /**
   * Timer for a maximum interval to check for dangling send calls that are not completed.
   * The way it is implemented we ensure that any "send" is rejected latest after < 2s
   */
  #sendTimer = Time.getTimer(
    "UDPChannel.send timeout check",
    UDP_SEND_TIMEOUT_CHECK_INTERVAL,
    () => this.#rejectDanglingSends()
  );
  #sendsInProgress = /* @__PURE__ */ new Map();
  constructor(lifetime, type, socket, netInterface) {
    this.#lifetime = lifetime;
    this.#type = type;
    this.#socket = socket;
    this.#netInterface = netInterface;
  }
  addMembership(membershipAddress) {
    const multicastInterfaces = NodeJsNetwork.getMembershipMulticastInterfaces(
      this.#netInterface,
      this.#type === "udp4"
    );
    for (const multicastInterface of multicastInterfaces) {
      try {
        this.#socket.addMembership(membershipAddress, multicastInterface);
      } catch (error) {
        logger.warn(
          `Error adding membership for address ${membershipAddress}${multicastInterface ? ` with interface ${multicastInterface}` : ""}: ${error}`
        );
      }
    }
  }
  dropMembership(membershipAddress) {
    const multicastInterfaces = NodeJsNetwork.getMembershipMulticastInterfaces(
      this.#netInterface,
      this.#type === "udp4"
    );
    for (const multicastInterface of multicastInterfaces) {
      try {
        this.#socket.dropMembership(membershipAddress, multicastInterface);
      } catch (error) {
        logger.warn(
          `Error removing membership for address ${membershipAddress}${multicastInterface ? ` with interface ${multicastInterface}` : ""}: ${error}`
        );
      }
    }
  }
  onData(listener) {
    const messageListener = (data, { address, port }) => {
      const netInterface = this.#netInterface ?? NodeJsNetwork.getNetInterfaceForIp(address);
      if (netInterface && this.#observedInterface === void 0) {
        this.#observedInterface = netInterface;
      }
      listener(netInterface, address, port, data);
    };
    this.#socket.on("message", messageListener);
    return {
      close: async () => {
        this.#socket.removeListener("message", messageListener);
      }
    };
  }
  /**
   * At minimum once every second we check for dangling sends that are not completed. That means that a dangling send
   * is removed very latest after <2s.
   */
  #rejectDanglingSends() {
    if (this.#sendsInProgress.size === 0) {
      return;
    }
    const now = Time.nowMs;
    for (const [promise, { sendMs, rejecter }] of this.#sendsInProgress) {
      const elapsed = Millis(now - sendMs);
      if (elapsed >= UDP_SEND_TIMEOUT_CHECK_INTERVAL) {
        this.#sendsInProgress.delete(promise);
        rejecter(new NetworkError("UDP send timeout"));
      }
    }
    if (this.#sendsInProgress.size > 0) {
      this.#sendTimer.start();
    }
  }
  async send(host, port, data) {
    const { promise, resolver, rejecter } = createPromise();
    const rejectOrResolve = (error) => {
      if (!this.#sendsInProgress.has(promise)) {
        return;
      }
      this.#sendsInProgress.delete(promise);
      if (!error) {
        resolver();
      } else {
        let repackAs;
        if ("code" in error) {
          switch (error.code) {
            case "EHOSTUNREACH":
              repackAs = AddressUnreachableError;
              break;
            case "ENETUNREACH":
              repackAs = NetworkUnreachableError;
              break;
            default:
              repackAs = NetworkError;
              break;
          }
        } else {
          repackAs = NetworkError;
        }
        rejecter(repackErrorAs(error, repackAs));
      }
    };
    let sendHost = host;
    if (host.startsWith("ff") && !host.includes("%")) {
      const effectiveInterface = this.#netInterface ?? this.#observedInterface;
      if (effectiveInterface) {
        sendHost = `${host}%${effectiveInterface}`;
      } else {
        throw new NetworkError(
          `Cannot send to multicast address ${host} without a known network interface. Configure network.listeningAddressIpv6 to bind to a specific interface.`
        );
      }
    }
    this.#sendsInProgress.set(promise, { sendMs: Time.nowMs, rejecter });
    if (!this.#sendTimer.isRunning) {
      this.#sendTimer.start();
    }
    try {
      this.#socket.send(Bytes.of(data), port, sendHost, (error) => rejectOrResolve(error));
    } catch (error) {
      rejectOrResolve(repackErrorAs(error, NetworkError));
    }
    return promise;
  }
  async close() {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      this.#sendTimer.stop();
      try {
        await new Promise((resolve) => this.#socket.close(resolve));
      } catch (error) {
        if (!(error instanceof Error) || error.message !== "Not running") {
          logger.debug("Error on closing socket", error);
        }
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  get port() {
    return this.#socket.address().port;
  }
  supports(type, address) {
    if (type !== ChannelType.UDP) {
      return false;
    }
    if (address === void 0) {
      return true;
    }
    if (this.#type === "udp4") {
      return isIPv4(address);
    }
    return isIPv6(address);
  }
}
export {
  NodeJsUdpChannel
};
