// @matter/general/dist/cjs/time/TimeUnit.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeUnit_exports = {};
__export(TimeUnit_exports, {
  Days: () => Days,
  Forever: () => Forever,
  Hours: () => Hours,
  Instant: () => Instant,
  Microseconds: () => Microseconds,
  Millis: () => Millis,
  Minutes: () => Minutes,
  Seconds: () => Seconds,
  TimeUnit: () => TimeUnit
});
module.exports = __toCommonJS(TimeUnit_exports);
var import_NodeJsStyleInspectable = require("#log/NodeJsStyleInspectable.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function TimeUnit(kind, abbrev, one, props = {}) {
  const unit = {
    [kind]: (scale) => {
      if (scale === void 0) {
        return void 0;
      }
      if (typeof scale === "bigint") {
        scale = Number(scale);
      }
      return scale * one;
    }
  }[kind];
  Object.assign(unit, {
    ...props,
    kind,
    abbrev,
    one,
    of,
    fractionalOf,
    ceil,
    floor,
    round,
    toString: kindOf
  });
  (0, import_NodeJsStyleInspectable.NodeJsStyleInspectable)(unit, kindOf);
  return unit;
}
function of(duration) {
  return Math.floor(duration / this.one);
}
function fractionalOf(duration) {
  return duration / this.one;
}
function ceil(duration) {
  return Math.ceil(duration / this.one) * this.one;
}
function floor(duration) {
  return Math.floor(duration / this.one) * this.one;
}
function round(duration) {
  return Math.round(duration / this.one) * this.one;
}
function kindOf() {
  return this.kind;
}
const Microseconds = TimeUnit("microsecond", "\u03BCs", 1e-3);
const Millis = TimeUnit("millisecond", "ms", 1);
const Seconds = TimeUnit("second", "s", 1e3);
const Minutes = TimeUnit("minute", "m", 6e4);
const Hours = TimeUnit("hour", "h", 36e5);
const Days = TimeUnit("day", "d", 864e5);
const Instant = Millis(0);
const Forever = Millis(Infinity);
