// @matter/general/dist/cjs/time/Timestamp.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Timestamp_exports = {};
__export(Timestamp_exports, {
  Timestamp: () => Timestamp,
  TimestampFormatError: () => TimestampFormatError
});
module.exports = __toCommonJS(Timestamp_exports);
var import_MatterError = require("#MatterError.js");
var import_Time = require("./Time.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function Timestamp(source) {
  if (source === void 0) {
    return import_Time.Time.nowMs;
  }
  if (typeof source === "string") {
    const date = Date.parse(source);
    if (!Number.isFinite(date.valueOf())) {
      throw new TimestampFormatError("Invalid timestamp format");
    }
  } else if (typeof source === "object" && source !== null && source instanceof Date) {
    if (!Number.isFinite(source.valueOf())) {
      throw new TimestampFormatError("Cannot convert date to timestamp because it is invalid");
    }
    source = source.getTime();
  } else if (typeof source === "bigint") {
    source = Number(source);
  }
  if (typeof source === "number") {
    if (!Number.isFinite(source)) {
      throw new TimestampFormatError("A timestamp must be a finite number");
    }
  } else {
    throw new TimestampFormatError(`Timestamp is not a number (received ${typeof source})`);
  }
  return source;
}
class TimestampFormatError extends import_MatterError.UnexpectedDataError {
}
((Timestamp2) => {
  function fromSeconds(source) {
    return Math.floor(source * 1e3);
  }
  Timestamp2.fromSeconds = fromSeconds;
  function fromMicroseconds(source) {
    if (typeof source === "bigint") {
      source = Number(source);
    }
    return Timestamp2(source / 1e3);
  }
  Timestamp2.fromMicroseconds = fromMicroseconds;
  Timestamp2.zero = Timestamp2(0);
  function dateOf(timestamp) {
    return new Date(timestamp);
  }
  Timestamp2.dateOf = dateOf;
  function delta(from, to = import_Time.Time.nowMs) {
    return to - from;
  }
  Timestamp2.delta = delta;
})(Timestamp || (Timestamp = {}));
