// @matter/general/dist/cjs/time/Timespan.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Timespan_exports = {};
__export(Timespan_exports, {
  Timespan: () => Timespan
});
module.exports = __toCommonJS(Timespan_exports);
var import_Timestamp = require("./Timestamp.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function Timespan(start, stop) {
  return Object.create(prototype, {
    start: { value: (0, import_Timestamp.Timestamp)(start), enumerable: true },
    stop: { value: (0, import_Timestamp.Timestamp)(stop), enumerable: true }
  });
}
const prototype = {
  start: 0,
  stop: 0,
  get duration() {
    return this.stop - this.start;
  },
  from(newStart) {
    return Timespan(newStart, this.stop);
  },
  to(newStop) {
    return Timespan(this.start, newStop);
  },
  valueOf() {
    return { start: this.start, stop: this.stop };
  }
};
