// @matter/protocol/dist/esm/protocol/DeviceAdvertiser.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Advertisement } from "#advertisement/Advertisement.js";
import { ServiceDescription } from "#advertisement/ServiceDescription.js";
import { FabricManager } from "#fabric/FabricManager.js";
import { SessionManager } from "#session/SessionManager.js";
import { Environmental, Logger, MatterAggregateError, ObserverGroup } from "@matter/general";
const logger = Logger.get("DeviceAdvertiser");
class DeviceAdvertiser {
  #context;
  #advertisers = /* @__PURE__ */ new Set();
  #observers = new ObserverGroup();
  #isOperational = false;
  #isClosing = false;
  #commissioningService;
  constructor(context) {
    this.#context = context;
    const { fabrics, sessions } = this.#context;
    this.#observers.on(fabrics.events.added, (fabric) => {
      if (!this.#isOperational) {
        return;
      }
      this.#advertiseFabric(fabric, "startup");
    });
    this.#observers.on(fabrics.events.replaced, async (fabric) => {
      if (!this.#isOperational) {
        return;
      }
      const fabricIndexAdvertisements = this.#advertisements(
        (ad) => ad.isOperational() && ad.description.fabric.fabricIndex === fabric.fabricIndex
      );
      if (fabricIndexAdvertisements.every(
        (ad) => ad.isOperational() && ad.description.fabric.globalId === fabric.globalId && ad.description.fabric.nodeId === fabric.nodeId
      )) {
        return;
      }
      for (const ad of fabricIndexAdvertisements) {
        await ad.close();
      }
      this.#advertiseFabric(fabric, "startup");
    });
    this.#observers.on(fabrics.events.deleting, (fabric) => {
      Advertisement.cancelAll(this.#advertisements((ad) => ad.isOperational() && ad.description.fabric === fabric));
    });
    this.#observers.on(sessions.sessions.added, (session) => {
      const fabricIndex = session.fabric?.fabricIndex;
      const fabric = fabricIndex ? fabrics.maybeFor(fabricIndex) : void 0;
      if (!fabric) {
        return;
      }
      for (const advertiser of this.#advertisers) {
        const ad = advertiser.advertisementForFabric(fabric);
        if (ad) {
          ad.serviceConnected();
        }
      }
    });
    this.#observers.on(sessions.sessions.deleted, (session) => {
      const fabricIndex = session.fabric?.fabricIndex;
      const fabric = fabricIndex ? fabrics.maybeFor(fabricIndex) : void 0;
      if (fabric) {
        if (fabric.hasSessionForPeer(session.peerNodeId)) {
          return;
        }
        for (const advertiser of this.#advertisers) {
          const ad = advertiser.advertisementForFabric(fabric);
          if (ad) {
            ad.serviceDisconnected();
          }
        }
        return;
      }
      if (!session.isPase) {
        return;
      }
      this.#startCommissioningAdvertisement();
    });
    this.#observers.on(sessions.retry, (session) => {
      if (!session.isSecure || !session.fabric) {
        return;
      }
      this.#advertiseFabric(session.associatedFabric, "retransmit");
    });
    this.#observers.on(this.#context.sessions.subscriptionsChanged, (session, subscription) => {
      if (this.#isOperational && subscription.isCanceledByPeer && session.fabric !== void 0) {
        logger.debug(`Subscription canceled by peer, re-announce`);
        this.#advertiseFabric(session.fabric, "reconnect");
      }
    });
  }
  /** Set the transport support bitmap for operational advertisements. */
  set supportedTransports(value) {
    this.#context.supportedTransports = value;
  }
  toString() {
    return "DeviceAdvertiser";
  }
  static [Environmental.create](env) {
    const instance = new DeviceAdvertiser({
      fabrics: env.get(FabricManager),
      sessions: env.get(SessionManager)
    });
    env.set(DeviceAdvertiser, instance);
    return instance;
  }
  /**
   * Advertise the device as commissionable.
   */
  enterCommissioningMode(description) {
    this.#commissioningService = description;
    this.#startCommissioningAdvertisement();
  }
  /**
   * Begin automatic broadcast for commissioning if in commissionable mode.
   */
  #startCommissioningAdvertisement() {
    if (this.#isClosing) {
      return;
    }
    if (this.#commissioningService === void 0) {
      return;
    }
    for (const advertiser of this.#advertisers) {
      advertiser.advertise(this.#commissioningService, "startup");
    }
  }
  /**
   * Cease advertising the device as commissionable.
   */
  exitCommissioningMode() {
    if (!this.#commissioningService) {
      return;
    }
    this.#commissioningService = void 0;
    return Advertisement.closeAll(this.#advertisements((ad) => ad.isCommissioning()));
  }
  /**
   * Advertise the device as operational.
   */
  enterOperationalMode() {
    if (this.#isOperational || this.#isClosing) {
      return;
    }
    this.#isOperational = true;
    this.#startOperationalAdvertisement();
  }
  /**
   * Begin automatic broadcast for fabrics if in operational mode.
   */
  #startOperationalAdvertisement() {
    if (!this.#isOperational || this.#isClosing) {
      return;
    }
    const fabricsAdvertised = new Set(
      this.#advertisements((ad) => ad.isOperational()).map(
        (ad) => ad.description.fabric
      )
    );
    for (const fabric of this.#context.fabrics) {
      if (!fabricsAdvertised.has(fabric)) {
        this.#advertiseFabric(fabric, "startup");
      }
    }
  }
  /**
   * Cease advertising the device as operational.
   */
  exitOperationalMode() {
    if (!this.#isOperational) {
      return;
    }
    this.#isOperational = false;
    Advertisement.cancelAll(this.#advertisements((ad) => ad.isOperational()));
  }
  /**
   * Reset automatic advertisement.
   *
   * This is useful when there is a network change.
   */
  restartAdvertisement() {
    this.#startCommissioningAdvertisement();
    this.#startOperationalAdvertisement();
  }
  async close() {
    this.#isClosing = true;
    this.#observers.close();
    await this.clearAdvertisers();
  }
  hasAdvertiser(advertiser) {
    return this.#advertisers.has(advertiser);
  }
  addAdvertiser(advertiser) {
    this.#advertisers.add(advertiser);
  }
  async deleteAdvertiser(advertiser) {
    this.#advertisers.delete(advertiser);
    await advertiser.close();
  }
  async clearAdvertisers() {
    const advertisers = [...this.#advertisers];
    this.#advertisers.clear();
    await MatterAggregateError.allSettled(advertisers.map((advertiser) => advertiser.close()));
  }
  #advertiseFabric(fabric, event) {
    if (this.#isClosing) {
      return;
    }
    for (const advertiser of this.#advertisers) {
      if (event === "startup") {
        if (this.#context.sessions.forFabric(fabric).some((session) => session.subscriptions.size > 0)) {
          event = "retransmit";
        }
      }
      advertiser.advertise(
        ServiceDescription.Operational({ fabric, tcp: this.#context.supportedTransports }),
        event
      );
    }
  }
  #advertisements(predicate) {
    return [...this.#advertisers].flatMap(
      (advertiser) => predicate ? advertiser.filter(predicate) : [...advertiser.advertisements]
    );
  }
}
export {
  DeviceAdvertiser
};
