// @matter/protocol/dist/esm/protocol/MessageExchange.js
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Message, SessionType } from "#codec/MessageCodec.js";
import { Mark } from "#common/Mark.js";
import { PeerUnresponsiveError, TransientPeerCommunicationError } from "#peer/PeerCommunicationError.js";
import { GroupSession } from "#session/GroupSession.js";
import { Session } from "#session/Session.js";
import {
  Abort,
  AbortedError,
  asError,
  AsyncObservableValue,
  causedBy,
  ClosedError,
  createPromise,
  CRYPTO_AEAD_MIC_LENGTH_BYTES,
  DataReadQueue,
  Diagnostic,
  Duration,
  Forever,
  hex,
  Instant,
  InternalError,
  Logger,
  MatterFlowError,
  Millis,
  NetworkError,
  ServerAddress,
  Time,
  TimeoutError,
  Timestamp
} from "@matter/general";
import {
  GroupId,
  SECURE_CHANNEL_PROTOCOL_ID,
  SecureMessageType,
  StatusCode,
  StatusResponseError
} from "@matter/types";
import { SessionClosedError, UnexpectedMessageError } from "./errors.js";
import { MRP } from "./MRP.js";
const logger = Logger.get("MessageExchange");
const MATTER_MESSAGE_OVERHEAD = 26 + 12 + CRYPTO_AEAD_MIC_LENGTH_BYTES;
class MessageExchange {
  static fromInitialMessage(context, initialMessage, options) {
    const { session } = context;
    return new MessageExchange({
      context,
      isInitiator: false,
      peerSessionId: session.id,
      nodeId: initialMessage.packetHeader.destNodeId,
      peerNodeId: initialMessage.packetHeader.sourceNodeId,
      exchangeId: initialMessage.payloadHeader.exchangeId,
      protocolId: initialMessage.payloadHeader.protocolId,
      ...options
    });
  }
  static initiate(context, exchangeId, protocolId, options) {
    const { session } = context;
    return new MessageExchange({
      context,
      isInitiator: true,
      peerSessionId: session.peerSessionId,
      nodeId: session.nodeId,
      peerNodeId: session.peerNodeId,
      exchangeId,
      protocolId,
      ...options
    });
  }
  #context;
  #isInitiator;
  #messagesQueue = new DataReadQueue();
  #lifetime;
  #onSend;
  #onReceive;
  #addressOverride;
  #receivedMessageToAck;
  #receivedMessageAckTimer = Time.getTimer("ack receipt timeout", MRP.STANDALONE_ACK_TIMEOUT, () => {
    if (this.#receivedMessageToAck !== void 0) {
      const messageToAck = this.#receivedMessageToAck;
      this.#receivedMessageToAck = void 0;
      this.sendStandaloneAckForMessage(messageToAck).catch(
        (error) => logger.error("An error happened when sending a standalone ack", error)
      );
    }
  });
  #closeTimer;
  #closeCause;
  #isDestroyed = false;
  #timedInteractionTimer;
  #used;
  #peerSessionId;
  #nodeId;
  #peerNodeId;
  #exchangeId;
  #protocolId;
  #closed = AsyncObservableValue();
  #closing = AsyncObservableValue();
  #channel;
  // TODO - following are associated with current active transmission and should maybe go in a closure
  #sentMessageToAck;
  #sentMessageAckSuccess;
  #sentMessageAckFailure;
  #sendOptions = {};
  #retransmissionCounter = 0;
  #totalRetransmissionCounter = 0;
  // counter for all messages within this exchange
  #messageSendCounter = 0;
  #messageReceivedCounter = 0;
  #retransmissionTimer;
  constructor(config) {
    const {
      context,
      isInitiator,
      peerSessionId,
      nodeId,
      peerNodeId,
      exchangeId,
      protocolId,
      onSend,
      onReceive,
      network,
      addressOverride
    } = config;
    this.#context = context;
    this.#isInitiator = isInitiator;
    this.#peerSessionId = peerSessionId;
    this.#nodeId = nodeId;
    this.#peerNodeId = peerNodeId;
    this.#exchangeId = exchangeId;
    this.#protocolId = protocolId;
    this.#onSend = onSend;
    this.#onReceive = onReceive;
    this.#addressOverride = addressOverride;
    const { activeThreshold, activeInterval, idleInterval } = this.session.parameters;
    this.#used = !isInitiator;
    const { session } = context;
    logger.debug(
      "New exchange",
      isInitiator ? Mark.OUTBOUND : Mark.INBOUND,
      this.via,
      Diagnostic.dict({
        protocol: this.#protocolId,
        peerSess: Session.idStrOf(this.#peerSessionId),
        address: addressOverride ? ServerAddress.urlFor(addressOverride) : void 0,
        SAT: activeThreshold !== void 0 ? Duration.format(activeThreshold) : void 0,
        SAI: activeInterval !== void 0 ? Duration.format(activeInterval) : void 0,
        SII: idleInterval !== void 0 ? Duration.format(idleInterval) : void 0,
        maxTrans: MRP.MAX_TRANSMISSIONS,
        exchangeFlags: Diagnostic.asFlags({
          MRP: this.session.usesMrp,
          I: this.isInitiator,
          ...network ? { [network.id]: true } : {}
        })
      })
    );
    session.addExchange(this);
    this.#lifetime = this.#context.session.join("exchange", Diagnostic.via(hex.word(this.id)));
  }
  get context() {
    return this.#context;
  }
  get isInitiator() {
    return this.#isInitiator;
  }
  /** Emits when the exchange is actually closed. This happens after all Retries and Communication are done. */
  get closed() {
    return this.#closed;
  }
  get considerClosed() {
    return this.#closed.value || this.#isInitiator && this.#closing.value;
  }
  /**
   * Emit when the exchange is closing, but not yet closed. We only wait for acks and retries to happen, but the
   * actual interaction logic is already done.
   */
  get closing() {
    return this.#closing;
  }
  get id() {
    return this.#exchangeId;
  }
  get idStr() {
    return hex.word(this.#exchangeId);
  }
  get session() {
    return this.context.session;
  }
  /** Number of retransmissions of the current outstanding message (resets on ack or new send). */
  get retransmissionCount() {
    return this.#retransmissionCounter;
  }
  get channel() {
    if (this.#channel === void 0) {
      this.#channel = this.session.channel;
    }
    return this.#channel;
  }
  /**
   * Max Payload size of the exchange which bases on the maximum payload size of the channel reduced by Matter
   * protocol overhead.
   */
  get maxPayloadSize() {
    return this.channel.maxPayloadSize - MATTER_MESSAGE_OVERHEAD;
  }
  join(...name) {
    return this.#lifetime.join(...name);
  }
  async onMessageReceived(message, duplicate = false) {
    logger.debug("Message", Mark.INBOUND, Message.diagnosticsOf(this, message, { duplicate }));
    this.#messageReceivedCounter++;
    if (message.payloadHeader.requiresAck && !this.session.usesMrp) {
      logger.debug("Ignoring ack-required flag because MRP is not used for this exchange");
      message.payloadHeader.requiresAck = false;
    }
    const {
      packetHeader: { messageId },
      payloadHeader: { requiresAck, ackedMessageId, protocolId, messageType }
    } = message;
    const isStandaloneAck = SecureMessageType.isStandaloneAck(protocolId, messageType);
    if (protocolId !== this.#protocolId && !isStandaloneAck) {
      throw new MatterFlowError(
        `Drop received a message for an unexpected protocol. Expected: ${this.#protocolId}, received: ${protocolId}`
      );
    }
    this.session.notifyActivity(true);
    this.#onReceive?.(message, duplicate);
    if (duplicate) {
      if (requiresAck) {
        await this.sendStandaloneAckForMessage(message);
      }
      return;
    }
    if (messageId === this.#sentMessageToAck?.payloadHeader.ackedMessageId) {
      var _stack = [];
      try {
        const _acking = __using(_stack, this.join("resending ack"));
        this.#messageSendCounter++;
        await this.channel.send(this.#sentMessageToAck, { exchange: this, addressOverride: this.#addressOverride });
        return;
      } catch (_) {
        var _error = _, _hasError = true;
      } finally {
        __callDispose(_stack, _error, _hasError);
      }
    }
    const sentMessageIdToAck = this.#sentMessageToAck?.packetHeader.messageId;
    if (sentMessageIdToAck !== void 0) {
      if (ackedMessageId === void 0) {
        throw new MatterFlowError("Previous message ack is missing");
      } else if (ackedMessageId !== sentMessageIdToAck) {
        if (isStandaloneAck) {
        } else {
          throw new MatterFlowError(
            `Incorrect ack received. Expected ${hex.fixed(sentMessageIdToAck, 8)}, received: ${hex.fixed(ackedMessageId, 8)}`
          );
        }
      } else {
        this.#retransmissionTimer?.stop();
        this.#retransmissionCounter = 0;
        this.#sentMessageAckSuccess?.(message);
        this.#sentMessageAckSuccess = void 0;
        this.#sentMessageAckFailure = void 0;
        this.#sentMessageToAck = void 0;
        if (isStandaloneAck && this.#closeTimer !== void 0) {
          return this.#close();
        }
      }
    }
    if (isStandaloneAck) {
      return;
    }
    if (requiresAck) {
      if (this.#receivedMessageToAck !== void 0) {
        this.#receivedMessageAckTimer.stop();
        await this.sendStandaloneAckForMessage(this.#receivedMessageToAck);
      }
      this.#receivedMessageToAck = message;
      this.#receivedMessageAckTimer.start();
    }
    this.#messagesQueue.write(message);
  }
  async send(messageType, payload, options = {}) {
    if (this.#closeCause) {
      throw new ClosedError("Exchange is closed", { cause: this.#closeCause });
    }
    this.#sendOptions = options;
    this.#retransmissionCounter = 0;
    try {
      await this.#send(messageType, payload);
    } catch (e) {
      if (!options.suppressPeerLoss && this.#messageReceivedCounter === 0 && causedBy(e, TransientPeerCommunicationError, TimeoutError, NetworkError)) {
        await this.#context.peerLost(this, asError(e));
      }
      throw e;
    }
  }
  async #send(messageType, payload, standaloneAckMessageId) {
    var _stack2 = [];
    try {
      const {
        expectAckOnly = false,
        disableMrpLogic,
        expectedProcessingTime = MRP.DEFAULT_EXPECTED_PROCESSING_TIME,
        logContext = {}
      } = this.#sendOptions;
      const abort = __using(_stack2, new Abort(this.#sendOptions));
      const isStandaloneAck = standaloneAckMessageId !== void 0;
      if (standaloneAckMessageId !== void 0) {
        if (!this.session.usesMrp) {
          throw new InternalError("Cannot include an acknowledge message ID when MRP is not used");
        }
        if (messageType !== SecureMessageType.StandaloneAck) {
          throw new InternalError("Cannot override an acknowledge message ID for non-standalone acks");
        }
      }
      let { requiresAck } = this.#sendOptions;
      if (requiresAck && !(this.session.usesMrp || this.session.isPeerLost)) {
        requiresAck = false;
      }
      const protocolId = isStandaloneAck ? SECURE_CHANNEL_PROTOCOL_ID : this.#protocolId;
      if (isStandaloneAck) {
        if (!this.session.usesMrp) {
          return;
        }
        if (requiresAck) {
          throw new MatterFlowError("A standalone ack may not require acknowledgement");
        }
      }
      if (this.#sentMessageToAck !== void 0 && !isStandaloneAck) {
        throw new MatterFlowError(
          `The previous message ${Message.via(this, this.#sentMessageToAck)} has not been acked yet, cannot send a new message`
        );
      }
      this.#used = true;
      this.#messageSendCounter++;
      this.session.notifyActivity(false);
      let ackedMessageId = standaloneAckMessageId;
      if (ackedMessageId === void 0 && this.session.usesMrp) {
        ackedMessageId = this.#receivedMessageToAck?.packetHeader.messageId;
        if (ackedMessageId !== void 0) {
          this.#receivedMessageAckTimer.stop();
          this.#receivedMessageToAck = void 0;
        }
      }
      let packetHeader;
      if (this.session.type === SessionType.Unicast) {
        const messageId = isStandaloneAck ? await this.session.getIncrementedMessageCounter() : await abort.attempt(this.session.getIncrementedMessageCounter());
        if (messageId === void 0) {
          return;
        }
        packetHeader = {
          sessionId: this.#peerSessionId,
          sessionType: SessionType.Unicast,
          messageId,
          destNodeId: this.#peerNodeId,
          sourceNodeId: this.#nodeId,
          hasPrivacyEnhancements: false,
          isControlMessage: false,
          hasMessageExtensions: false
        };
      } else if (this.session.type === SessionType.Group) {
        const session = this.session;
        if (!GroupSession.is(session)) {
          throw new InternalError("Session is not a GroupSession, but session type is Group.");
        }
        const destGroupId = GroupId.fromNodeId(this.#peerNodeId);
        if (destGroupId === 0) {
          throw new InternalError(`Invalid GroupId extracted from NodeId ${this.#peerNodeId}`);
        }
        const messageId = await abort.attempt(this.session.getIncrementedMessageCounter());
        if (messageId === void 0) {
          return;
        }
        packetHeader = {
          sessionId: this.#peerSessionId,
          sessionType: SessionType.Group,
          messageId,
          destGroupId,
          sourceNodeId: this.#nodeId,
          // We are the source node, so use our NodeId
          hasPrivacyEnhancements: false,
          isControlMessage: false,
          hasMessageExtensions: false
        };
      } else {
        throw new InternalError(`Unknown session type: ${this.session.type}`);
      }
      const message = {
        packetHeader,
        payloadHeader: {
          exchangeId: this.#exchangeId,
          protocolId,
          messageType,
          isInitiatorMessage: this.isInitiator,
          requiresAck: requiresAck ?? (this.session.usesMrp && !isStandaloneAck),
          ackedMessageId,
          hasSecuredExtension: false
        },
        payload
      };
      let ackPromise;
      if (this.session.usesMrp && message.payloadHeader.requiresAck && !disableMrpLogic) {
        this.#sentMessageToAck = message;
        const backOff = this.#mrpResubmissionBackOffTime;
        logContext.backOff = Duration.format(backOff);
        this.#retransmissionTimer = Time.getTimer(
          `retransmitting ${Message.via(this, message)}`,
          backOff,
          () => this.#retransmitMessage(message, expectedProcessingTime)
        );
        const { promise, resolver } = createPromise();
        ackPromise = promise;
        this.#sentMessageAckSuccess = resolver;
        const startedWaitingAt = Time.nowMs;
        this.#sentMessageAckFailure = () => {
          abort.abort(new PeerUnresponsiveError(Timestamp.delta(startedWaitingAt)));
        };
      }
      this.#onSend?.(message, 0);
      const sending = __using(_stack2, this.join("sending", Diagnostic.strong(Message.via(this, message))));
      if (isStandaloneAck) {
        await this.channel.send(message, { exchange: this, addressOverride: this.#addressOverride });
      } else {
        await abort.attempt(
          this.channel.send(message, { exchange: this, logContext, addressOverride: this.#addressOverride })
        );
      }
      if (abort.aborted) {
        return;
      }
      if (ackPromise !== void 0) {
        var _stack = [];
        try {
          this.#retransmissionCounter = 0;
          this.#retransmissionTimer?.start();
          const _waiting = __using(_stack, sending.join("waiting for ack"));
          const responseMessage = await abort.race(ackPromise);
          if (abort.aborted) {
            this.#retransmissionTimer?.stop();
            if (!causedBy(abort.reason, AbortedError)) {
              throw abort.reason;
            }
            return;
          }
          this.#sentMessageAckSuccess = void 0;
          this.#sentMessageAckFailure = void 0;
          if (responseMessage) {
            const {
              payloadHeader: { protocolId: protocolId2, messageType: messageType2 }
            } = responseMessage;
            if (expectAckOnly && !SecureMessageType.isStandaloneAck(protocolId2, messageType2)) {
              throw new UnexpectedMessageError("Expected ack only", this.session, responseMessage);
            }
          }
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          __callDispose(_stack, _error, _hasError);
        }
      }
    } catch (_2) {
      var _error2 = _2, _hasError2 = true;
    } finally {
      __callDispose(_stack2, _error2, _hasError2);
    }
  }
  async nextMessage(options) {
    try {
      return await this.#nextMessage(options);
    } catch (e) {
      if (!options?.suppressPeerLoss && this.#messageReceivedCounter === 0 && causedBy(e, TransientPeerCommunicationError)) {
        await this.#context.peerLost(this, asError(e));
      }
      throw e;
    }
  }
  async #nextMessage(options) {
    var _stack = [];
    try {
      let timeout;
      if (options?.timeout !== void 0) {
        timeout = options.timeout;
      } else if (!this.session.isClosed) {
        timeout = this.channel.calculateMaximumPeerResponseTime(
          this.session.parameters,
          this.context.localSessionParameters,
          options?.expectedProcessingTime
        );
      }
      const localAbort = __using(_stack, new Abort({
        timeout,
        abort: options?.abort,
        timeoutHandler: () => {
          throw new PeerUnresponsiveError(timeout);
        }
      }));
      return await this.#messagesQueue.read(localAbort);
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async sendStandaloneAckForMessage(message) {
    const {
      packetHeader: { messageId },
      payloadHeader: { requiresAck }
    } = message;
    if (!requiresAck || !this.session.usesMrp) {
      return;
    }
    await this.#send(SecureMessageType.StandaloneAck, new Uint8Array(0), messageId);
  }
  #retransmitMessage(message, expectedProcessingTime) {
    this.#retransmissionCounter++;
    this.#totalRetransmissionCounter++;
    if (this.considerClosed || this.#retransmissionCounter >= (this.#sendOptions.maxRetransmissions ?? MRP.MAX_TRANSMISSIONS)) {
      if (expectedProcessingTime && !this.considerClosed) {
        const finalWaitTime = Millis(
          this.channel.calculateMaximumPeerResponseTime(
            this.session.parameters,
            this.context.localSessionParameters,
            expectedProcessingTime
          ) - (this.#retransmissionTimer?.interval ?? Instant)
        );
        if (finalWaitTime > 0) {
          this.#retransmissionCounter--;
          this.#totalRetransmissionCounter--;
          logger.debug(
            `Message ${Message.via(this, message)}: Wait additional ${Duration.format(finalWaitTime)} for processing time and peer resubmissions after all our resubmissions`
          );
          this.#retransmissionTimer = Time.getTimer(
            `waiting after resubmissions for ${Message.via(this, message)}`,
            finalWaitTime,
            () => this.#retransmitMessage(message)
          ).start();
          return;
        }
      }
      if (this.#sentMessageToAck !== void 0 && this.#sentMessageAckFailure !== void 0) {
        this.#receivedMessageToAck = void 0;
        this.#sentMessageAckFailure();
        this.#sentMessageAckFailure = void 0;
        this.#sentMessageAckSuccess = void 0;
      }
      if (this.#closeTimer !== void 0) {
        this.#close().catch((error) => logger.error("Error closing exchange", error));
      }
      return;
    }
    this.#messageSendCounter++;
    this.session.notifyActivity(false);
    this.context.retry(this.#retransmissionCounter);
    const resubmissionBackoffTime = this.#mrpResubmissionBackOffTime;
    this.#onSend?.(message, this.#retransmissionCounter);
    this.channel.send(message, {
      exchange: this,
      logContext: {
        "retrans#": this.#retransmissionCounter,
        backoff: Duration.format(resubmissionBackoffTime)
      },
      addressOverride: this.#addressOverride
    }).then(() => this.#initializeResubmission(message, resubmissionBackoffTime, expectedProcessingTime)).catch((error) => {
      logger.error(`Error retransmitting ${Message.via(this, message)}:`, error);
      if (error instanceof SessionClosedError) {
        this.#close().catch((error2) => logger.error("An error happened when closing the exchange", error2));
      } else {
        this.#initializeResubmission(message, resubmissionBackoffTime, expectedProcessingTime);
      }
    });
  }
  #initializeResubmission(message, resubmissionBackoffTime, expectedProcessingTimeMs) {
    this.#retransmissionTimer = Time.getTimer(
      "Message retransmission",
      resubmissionBackoffTime,
      () => this.#retransmitMessage(message, expectedProcessingTimeMs)
    ).start();
  }
  [Symbol.asyncDispose]() {
    return this.destroy();
  }
  async destroy() {
    if (this.#isDestroyed) {
      return;
    }
    this.#isDestroyed = true;
    if (this.#closeTimer === void 0 && this.#receivedMessageToAck !== void 0) {
      this.#receivedMessageAckTimer.stop();
      const messageToAck = this.#receivedMessageToAck;
      this.#receivedMessageToAck = void 0;
      try {
        await this.sendStandaloneAckForMessage(messageToAck);
      } catch (error) {
        logger.error("An error happened when closing the exchange", error);
      }
    }
    await this.#close();
  }
  startTimedInteraction(timeout) {
    if (this.#timedInteractionTimer !== void 0 && this.#timedInteractionTimer.isRunning) {
      this.#timedInteractionTimer.stop();
      throw new StatusResponseError(
        "Timed interaction already running for this exchange",
        StatusCode.InvalidAction
      );
    }
    logger.debug(
      "Starting timed interaction",
      Mark.INBOUND,
      this.channel.name,
      Diagnostic.dict({ exId: this.#exchangeId, timeout: Duration.format(timeout) })
    );
    this.#timedInteractionTimer = Time.getTimer("Timed interaction", timeout, () => {
      logger.debug(
        "Timed interaction timeout!",
        Diagnostic.dict({ exId: this.#exchangeId, via: this.channel.name })
      );
    }).start();
  }
  clearTimedInteraction() {
    if (this.#timedInteractionTimer !== void 0) {
      logger.debug(
        "Clearing timed interaction",
        Diagnostic.dict({ exId: this.#exchangeId, via: this.channel.name })
      );
      this.#timedInteractionTimer.stop();
      this.#timedInteractionTimer = void 0;
    }
  }
  hasTimedInteraction() {
    return this.#timedInteractionTimer !== void 0;
  }
  hasActiveTimedInteraction() {
    return this.#timedInteractionTimer !== void 0 && this.#timedInteractionTimer.isRunning;
  }
  hasExpiredTimedInteraction() {
    return this.#timedInteractionTimer !== void 0 && !this.#timedInteractionTimer.isRunning;
  }
  /**
   * Closes the exchange.
   *
   * If cause is defined, the exchange is closed immediately, even if there are still messages to send.  Further
   * attempts at communicating via the exchange will throw this error.
   *
   * If cause is undefined, the exchange is closed only after all messages have been sent.
   */
  async close(cause) {
    var _stack3 = [];
    try {
      if (this.#isDestroyed) {
        return;
      }
      const closing = __using(_stack3, this.#lifetime.closing());
      if (this.#closeTimer !== void 0) {
        if (cause) {
          this.#closeTimer.stop();
          return this.#close(cause);
        }
        return;
      }
      if (!this.#used) {
        logger.info(this.via, `Exchange never used, closing directly`);
        return this.#close(cause);
      }
      {
        var _stack = [];
        try {
          const _emitting = __using(_stack, closing.join("emitting"));
          await this.#closing.emit(true);
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          __callDispose(_stack, _error, _hasError);
        }
      }
      if (this.#closed.value) {
        return;
      }
      if (this.#receivedMessageToAck !== void 0) {
        this.#receivedMessageAckTimer.stop();
        const messageToAck = this.#receivedMessageToAck;
        this.#receivedMessageToAck = void 0;
        try {
          var _stack2 = [];
          try {
            const _acking = __using(_stack2, closing.join("acking"));
            await this.sendStandaloneAckForMessage(messageToAck);
          } catch (_2) {
            var _error2 = _2, _hasError2 = true;
          } finally {
            __callDispose(_stack2, _error2, _hasError2);
          }
        } catch (error) {
          logger.error(this.via, `Unhandled error closing exchange`, error);
        }
        if (cause || this.#sentMessageToAck === void 0) {
          await this.#close(cause);
          return;
        }
      } else if (this.#sentMessageToAck === void 0 || cause) {
        await this.#close(cause);
        return;
      }
      let maxResubmissionTime = Instant;
      for (let i = this.#retransmissionCounter; i <= MRP.MAX_TRANSMISSIONS; i++) {
        maxResubmissionTime = Millis(
          maxResubmissionTime + this.channel.getMrpResubmissionBackOffTime(i, void 0, true)
        );
      }
      this.#closeTimer = Time.getTimer(
        `Exchange ${this.via} close`,
        maxResubmissionTime,
        async () => await this.#close(cause)
      ).start();
    } catch (_3) {
      var _error3 = _3, _hasError3 = true;
    } finally {
      __callDispose(_stack3, _error3, _hasError3);
    }
  }
  async #close(cause) {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      if (this.#closeCause === void 0) {
        this.#closeCause = cause;
      }
      try {
        this.#retransmissionTimer?.stop();
        this.#receivedMessageAckTimer.stop();
        this.#sentMessageAckSuccess?.(void 0);
        this.#closeTimer?.stop();
        this.#timedInteractionTimer?.stop();
        this.#messagesQueue.close(this.#closeCause);
      } finally {
        await this.#closed.emit(true);
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  get via() {
    if (this.session === void 0) {
      return Diagnostic.via(`${Mark.EXCHANGE}${this.idStr}`);
    }
    return Diagnostic.via(`${this.session.via}${Mark.EXCHANGE}${this.idStr}`);
  }
  /**
   * Expose some diagnostics for logging.
   * For simply one message exchanges without retransmissions nothing will be exposed
   */
  get diagnostics() {
    if (this.#totalRetransmissionCounter === 0 && this.#messageSendCounter === 1 && this.#messageReceivedCounter === 1) {
      return "";
    }
    return `${this.#messageReceivedCounter}${Mark.TRANSFERRED}${this.#messageSendCounter}${this.#totalRetransmissionCounter > 0 ? `+${this.#totalRetransmissionCounter}` : ""}`;
  }
  get #mrpResubmissionBackOffTime() {
    let backOff = this.channel.getMrpResubmissionBackOffTime(this.#retransmissionCounter);
    if (this.#sendOptions.initialRetransmissionTime !== void 0) {
      backOff = Millis(backOff + this.#sendOptions.initialRetransmissionTime);
    }
    return Duration.min(backOff, this.#sendOptions.maxRetransmissionTime ?? Forever);
  }
}
export {
  MATTER_MESSAGE_OVERHEAD,
  MessageExchange
};
