// @matter/protocol/dist/cjs/advertisement/Advertiser.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var Advertiser_exports = {};
__export(Advertiser_exports, {
  Advertiser: () => Advertiser
});
module.exports = __toCommonJS(Advertiser_exports);
var import_general = require("@matter/general");
var import_Advertisement = require("./Advertisement.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Advertiser {
  #lifetime;
  #advertisements = /* @__PURE__ */ new Set();
  #isClosed = false;
  constructor(lifetime = import_general.Lifetime.process) {
    this.#lifetime = lifetime.join("advertiser");
  }
  /**
   * Begin advertising on configured schedule.
   *
   * Note that so long as the advertisement is installed the device will continue to respond to MDNS queries even if
   * the advertisement schedule is elapsed.
   *
   * Returns undefined if the advertiser does not support this type of advertisement.
   */
  advertise(description, event) {
    if (this.#isClosed) {
      return;
    }
    const ad = this.getAdvertisement(description);
    if (ad) {
      ad.start(event);
    }
    return ad;
  }
  /**
   * Destroy the instance.
   */
  async close() {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      this.#isClosed = true;
      await import_Advertisement.Advertisement.closeAll(this.#advertisements);
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  /**
   * The set of advertisements active for this advertiser.
   */
  get advertisements() {
    return this.#advertisements;
  }
  /**
   * Obtain the active advertisement for {@link fabric}, if any.
   */
  advertisementForFabric(fabric) {
    for (const ad of this.#advertisements) {
      if (ad.isOperational() && ad.description.fabric === fabric) {
        return ad;
      }
    }
  }
  /**
   * Select advertisements using a predicate.
   */
  filter(predicate) {
    return [...this.#advertisements].filter(predicate);
  }
  join(...name) {
    return this.#lifetime.join(...name);
  }
}
