// @matter/protocol/dist/cjs/advertisement/ServiceDescription.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ServiceDescription_exports = {};
__export(ServiceDescription_exports, {
  ServiceDescription: () => ServiceDescription
});
module.exports = __toCommonJS(ServiceDescription_exports);
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ServiceDescription;
((ServiceDescription2) => {
  function isCommissioning(description) {
    return description.kind === "commissionable" || description.kind === "commissioner";
  }
  ServiceDescription2.isCommissioning = isCommissioning;
  function isOperational(description) {
    return description.kind === "operational";
  }
  ServiceDescription2.isOperational = isOperational;
  function Commissionable(definition) {
    return {
      ...definition,
      kind: "commissionable"
    };
  }
  ServiceDescription2.Commissionable = Commissionable;
  function Operational(definition) {
    return {
      ...definition,
      kind: "operational"
    };
  }
  ServiceDescription2.Operational = Operational;
  function Commissioner(definition) {
    return {
      ...definition,
      kind: "commissioner"
    };
  }
  ServiceDescription2.Commissioner = Commissioner;
})(ServiceDescription || (ServiceDescription = {}));
