// @matter/protocol/dist/cjs/securechannel/SecureChannelMessenger.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SecureChannelMessenger_exports = {};
__export(SecureChannelMessenger_exports, {
  ChannelStatusResponseError: () => ChannelStatusResponseError,
  DEFAULT_NORMAL_PROCESSING_TIME: () => DEFAULT_NORMAL_PROCESSING_TIME,
  EXPECTED_CRYPTO_PROCESSING_TIME: () => EXPECTED_CRYPTO_PROCESSING_TIME,
  SecureChannelMessenger: () => SecureChannelMessenger
});
module.exports = __toCommonJS(SecureChannelMessenger_exports);
var import_PeerCommunicationError = require("#peer/PeerCommunicationError.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_SecureChannelStatusMessageSchema = require("./SecureChannelStatusMessageSchema.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ChannelStatusResponseError extends import_PeerCommunicationError.TransientPeerCommunicationError {
  constructor(message, generalStatusCode, protocolStatusCode, vendorId, protocolData) {
    super(
      `(${import_types.GeneralStatusCode[generalStatusCode]} (${generalStatusCode}) / ${import_types.SecureChannelStatusCode[protocolStatusCode]} (${protocolStatusCode})) ${message}`
    );
    this.generalStatusCode = generalStatusCode;
    this.protocolStatusCode = protocolStatusCode;
    this.vendorId = vendorId;
    this.protocolData = protocolData;
    if (generalStatusCode === import_types.GeneralStatusCode.Busy && protocolStatusCode === import_types.SecureChannelStatusCode.Busy && protocolData?.byteLength === 2) {
      this.busyDelay = (0, import_general.Millis)(new import_general.DataReader(protocolData, import_general.Endian.Little).readUInt16());
    }
  }
  generalStatusCode;
  protocolStatusCode;
  vendorId;
  protocolData;
  busyDelay;
}
const EXPECTED_CRYPTO_PROCESSING_TIME = (0, import_general.Seconds)(35);
const DEFAULT_NORMAL_PROCESSING_TIME = (0, import_general.Seconds)(2);
class SecureChannelMessenger {
  constructor(exchange, defaultExpectedProcessingTime = EXPECTED_CRYPTO_PROCESSING_TIME) {
    this.exchange = exchange;
    this.#defaultExpectedProcessingTime = defaultExpectedProcessingTime;
  }
  exchange;
  #defaultExpectedProcessingTime;
  get channel() {
    return this.exchange.channel;
  }
  get via() {
    return this.exchange.via;
  }
  /**
   * Waits for the next message and returns it.
   *
   * {@link expectedProcessingTime} defaults to {@link EXPECTED_CRYPTO_PROCESSING_TIME}.
   */
  async nextMessage({
    type,
    expectedProcessingTime = this.#defaultExpectedProcessingTime,
    description,
    abort
  } = {}) {
    const message = await this.exchange.nextMessage({ expectedProcessingTime, abort });
    const messageType = message.payloadHeader.messageType;
    if (type !== void 0 && description === void 0) {
      description = import_types.SecureMessageType[type];
    }
    this.throwIfErrorStatusReport(message, description);
    if (type !== void 0 && messageType !== type)
      throw new import_general.UnexpectedDataError(
        `Received unexpected message type: ${messageType}, expected: ${type} (${description})`
      );
    return message;
  }
  /**
   * Waits for the next message and decodes it.
   *
   * When no expectedProcessingTimeMs is provided, the default value of EXPECTED_CRYPTO_PROCESSING_TIME is used.
   */
  async nextMessageDecoded(schema, options) {
    return schema.decode((await this.nextMessage(options)).payload);
  }
  /**
   * Waits for the next message and returns it.
   *
   * When no expectedProcessingTimeMs is provided, the default value of EXPECTED_CRYPTO_PROCESSING_TIME is used.
   */
  async waitForSuccess(options) {
    await this.nextMessage({
      ...options,
      type: import_types.SecureMessageType.StatusReport
    });
  }
  /**
   * Sends a message of the given type with the given payload.
   *
   * If no ExchangeSendOptions are provided, the expectedProcessingTimeMs will be set to
   * EXPECTED_CRYPTO_PROCESSING_TIME.
   */
  async send(message, type, schema, options) {
    options = {
      ...options,
      expectedProcessingTime: options?.expectedProcessingTime ?? this.#defaultExpectedProcessingTime
    };
    const payload = schema.encode(message);
    await this.exchange.send(type, payload, options);
    return payload;
  }
  sendError(code, abort) {
    return this.#sendStatusReport(import_types.GeneralStatusCode.Failure, code, abort);
  }
  sendSuccess(abort) {
    return this.#sendStatusReport(import_types.GeneralStatusCode.Success, import_types.SecureChannelStatusCode.Success, abort);
  }
  sendBusy(minimumRetryInterval, abort) {
    if (minimumRetryInterval <= 0) {
      throw new Error("Busy minimum retry interval must be greater than 0ms");
    }
    const writer = new import_general.DataWriter(import_general.Endian.Little);
    writer.writeUInt16(Math.min(minimumRetryInterval, 65535));
    return this.#sendStatusReport(
      import_types.GeneralStatusCode.Busy,
      import_types.SecureChannelStatusCode.Busy,
      abort,
      void 0,
      writer.toByteArray()
    );
  }
  sendCloseSession(abort) {
    return this.#sendStatusReport(import_types.GeneralStatusCode.Success, import_types.SecureChannelStatusCode.CloseSession, abort, false);
  }
  get channelName() {
    return this.exchange.channel.channel.name;
  }
  [Symbol.asyncDispose]() {
    return this.close();
  }
  async close() {
    await this.exchange.close();
  }
  async #sendStatusReport(generalStatus, protocolStatus, abort, requiresAck, protocolData) {
    await this.exchange.send(
      import_types.SecureMessageType.StatusReport,
      import_SecureChannelStatusMessageSchema.SecureChannelStatusMessage.encode({
        generalStatus,
        protocolStatus,
        protocolData
      }),
      {
        requiresAck,
        logContext: {
          generalStatus: import_types.GeneralStatusCode[generalStatus] ?? import_general.Diagnostic.hex(generalStatus),
          protocolStatus: import_types.SecureChannelStatusCode[protocolStatus] ?? import_general.Diagnostic.hex(protocolStatus)
        },
        abort
      }
    );
  }
  throwIfErrorStatusReport(message, logHint) {
    const {
      payloadHeader: { messageType },
      payload
    } = message;
    if (messageType !== import_types.SecureMessageType.StatusReport) return;
    const { generalStatus, protocolId, protocolStatus, vendorId, protocolData } = import_SecureChannelStatusMessageSchema.SecureChannelStatusMessage.decode(payload);
    if (generalStatus !== import_types.GeneralStatusCode.Success) {
      throw new ChannelStatusResponseError(
        `Received general error status for protocol ${protocolId}${logHint ? ` (${logHint})` : ""}`,
        generalStatus,
        protocolStatus,
        vendorId,
        protocolData
      );
    }
    if (protocolStatus !== import_types.SecureChannelStatusCode.Success) {
      throw new ChannelStatusResponseError(
        `Received general success status, but protocol status is not Success${logHint ? ` (${logHint})` : ""}`,
        generalStatus,
        protocolStatus
      );
    }
  }
}
