// @matter/protocol/dist/esm/common/Scanner.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  BasicSet,
  ChannelType,
  Diagnostic,
  Duration,
  Environmental,
  Millis
} from "@matter/general";
import { SupportedTransportsSchema } from "./SupportedTransportsBitmap.js";
function DiscoveryData(kvs) {
  const dd = {};
  for (const key of kvs.keys()) {
    switch (key) {
      case "VP":
      case "DN":
      case "RI":
      case "PI":
        dd[key] = `${kvs.get(key)}`;
        break;
      case "DT":
      case "PH":
      case "ICD": {
        const num = Number(kvs.get(key));
        if (isFinite(num)) {
          dd[key] = num;
        }
        break;
      }
      case "T": {
        const num = Number(kvs.get(key));
        if (isFinite(num)) {
          dd.T = SupportedTransportsSchema.decode(num);
        }
        break;
      }
      case "SII":
      case "SAI":
      case "SAT": {
        const num = Number(kvs.get(key));
        if (isFinite(num)) {
          dd[key] = Millis(num);
        }
        break;
      }
    }
  }
  return dd;
}
function DiscoveryDataDiagnostics(data, kind) {
  return Diagnostic.dict({
    kind,
    DN: data.DN,
    SII: data.SII !== void 0 ? Duration.format(data.SII) : void 0,
    SAI: data.SAI !== void 0 ? Duration.format(data.SAI) : void 0,
    SAT: data.SAT !== void 0 ? Duration.format(data.SAT) : void 0,
    T: data.T !== void 0 ? Diagnostic.asFlags(data.T) : void 0,
    DT: data.DT,
    PH: data.PH,
    ICD: data.ICD,
    VP: data.VP,
    RI: data.RI,
    PI: data.PI
  });
}
class ScannerSet extends BasicSet {
  scannerFor(type) {
    return this.find((scanner) => scanner.type === type);
  }
  hasScannerFor(type) {
    return this.scannerFor(type) !== void 0;
  }
  /**
   * Select a set of scanners based on discovery capabilities.
   */
  select(discoveryCapabilities) {
    return this.filter(
      (scanner) => scanner.type === ChannelType.UDP || discoveryCapabilities?.ble && scanner.type === ChannelType.BLE
    );
  }
  static [Environmental.create](env) {
    const instance = new ScannerSet();
    env.set(ScannerSet, instance);
    return instance;
  }
}
export {
  DiscoveryData,
  DiscoveryDataDiagnostics,
  ScannerSet
};
