// @matter/types/dist/cjs/dcl/device.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var device_exports = {};
__export(device_exports, {
  EnhancedSetupFlowOptionsBitmap: () => EnhancedSetupFlowOptionsBitmap,
  EnhancedSetupFlowOptionsSchema: () => EnhancedSetupFlowOptionsSchema
});
module.exports = __toCommonJS(device_exports);
var import_BitmapSchema = require("../schema/BitmapSchema.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const EnhancedSetupFlowOptionsBitmap = {
  /**
   * A commissioner is required to show Terms and Conditions content if this is the initial
   * commissioning for the device out of a factory-reset state.
   */
  tcOnInitialCommission: (0, import_BitmapSchema.BitFlag)(0),
  /**
   * The commissioner SHALL NOT reuse user's responses to the required Terms and Conditions
   * from commissioning a previous product (same VID) with the same T&C.
   */
  disallowTcReuseForRequiredTc: (0, import_BitmapSchema.BitFlag)(1),
  /**
   * The commissioner SHALL NOT reuse user's responses to the non-required Terms and Conditions
   * from commissioning a previous product (same VID) with the same T&C.
   */
  disallowTcReuseForNonRequiredTc: (0, import_BitmapSchema.BitFlag)(2),
  /**
   * The commissioner SHALL NOT reuse user's responses to the Terms and Conditions from
   * commissioning a previous product (same VID, other PID) with the same T&C.
   */
  disallowTcReuseBetweenPids: (0, import_BitmapSchema.BitFlag)(3)
};
const EnhancedSetupFlowOptionsSchema = (0, import_BitmapSchema.BitmapSchema)(EnhancedSetupFlowOptionsBitmap);
