// @matter/protocol/dist/esm/bdx/FileDesignator.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Bytes, Diagnostic, ImplementationError } from "@matter/general";
import { NodeId } from "@matter/types";
class FileDesignator {
  #fd;
  /**
   * Initialize a FileDesignator from a BDX URI string. Also returns the source node ID of the owner of the file.
   */
  static fromBdxUri(uri) {
    const bdxParts = uri.match(/^bdx:\/\/([0-9A-F]{16})\/(\S+)$/);
    if (!bdxParts) {
      throw new ImplementationError(`Invalid OTA URI "${uri}"`);
    }
    return {
      fileDesignator: new FileDesignator(bdxParts[2]),
      sourceNodeId: NodeId(BigInt(`0x${bdxParts[1]}`))
    };
  }
  /** Create a FileDesignator from a string or bytes object. */
  constructor(fd) {
    if (typeof fd === "string") {
      this.#fd = Bytes.fromString(fd);
    } else {
      this.#fd = fd;
    }
  }
  /** Return the bytes representation of the FileDesignator. */
  get bytes() {
    return this.#fd;
  }
  /** Return the textual/string representation of the FileDesignator. */
  get text() {
    const fileDesignatorData = Bytes.of(this.#fd);
    const isValidName = fileDesignatorData.every(
      (byte) => byte >= 65 && byte <= 90 || // A-Z
      byte >= 97 && byte <= 122 || // a-z
      byte >= 48 && byte <= 57 || // 0..9
      byte === 46 || // "."
      byte === 45 || // "-"
      byte === 95 || // "_"
      byte === 47
      // "/"
    );
    if (isValidName) {
      return fileDesignatorData.reduce((name, byte) => name + String.fromCharCode(byte), "");
    } else {
      return `0x${Bytes.toHex(fileDesignatorData)}`;
    }
  }
  /**
   * Return the FileDesignator as a BDX URI string. The source node ID of the owner of the file is required to
   * generate a valid URI.
   */
  asBdxUri(sourceNode) {
    return `bdx://${NodeId.strOf(sourceNode).toUpperCase()}/${this.text}`;
  }
  get [Diagnostic.value]() {
    return this.text;
  }
}
export {
  FileDesignator
};
