// @matter/node/dist/esm/devices/cook-surface.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { OnOffServer as BaseOnOffServer } from "../behaviors/on-off/OnOffServer.js";
import {
  TemperatureControlServer as BaseTemperatureControlServer
} from "../behaviors/temperature-control/TemperatureControlServer.js";
import {
  TemperatureMeasurementServer as BaseTemperatureMeasurementServer
} from "../behaviors/temperature-measurement/TemperatureMeasurementServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var CookSurfaceRequirements;
((CookSurfaceRequirements2) => {
  CookSurfaceRequirements2.OnOffServer = BaseOnOffServer.with("OffOnly");
  CookSurfaceRequirements2.TemperatureControlServer = BaseTemperatureControlServer.with("TemperatureLevel");
  CookSurfaceRequirements2.TemperatureMeasurementServer = BaseTemperatureMeasurementServer;
  CookSurfaceRequirements2.server = {
    optional: {
      OnOff: CookSurfaceRequirements2.OnOffServer,
      TemperatureControl: CookSurfaceRequirements2.TemperatureControlServer,
      TemperatureMeasurement: CookSurfaceRequirements2.TemperatureMeasurementServer
    },
    mandatory: {}
  };
})(CookSurfaceRequirements || (CookSurfaceRequirements = {}));
const CookSurfaceDeviceDefinition = MutableEndpoint({
  name: "CookSurface",
  deviceType: 119,
  deviceRevision: 2,
  requirements: CookSurfaceRequirements,
  behaviors: SupportedBehaviors()
});
Object.freeze(CookSurfaceDeviceDefinition);
const CookSurfaceDevice = CookSurfaceDeviceDefinition;
export {
  CookSurfaceDevice,
  CookSurfaceDeviceDefinition,
  CookSurfaceRequirements
};
