// @matter/node/dist/esm/devices/flow-sensor.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { FlowMeasurementServer as BaseFlowMeasurementServer } from "../behaviors/flow-measurement/FlowMeasurementServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var FlowSensorRequirements;
((FlowSensorRequirements2) => {
  FlowSensorRequirements2.IdentifyServer = BaseIdentifyServer;
  FlowSensorRequirements2.FlowMeasurementServer = BaseFlowMeasurementServer;
  FlowSensorRequirements2.server = { mandatory: { Identify: FlowSensorRequirements2.IdentifyServer, FlowMeasurement: FlowSensorRequirements2.FlowMeasurementServer } };
})(FlowSensorRequirements || (FlowSensorRequirements = {}));
const FlowSensorDeviceDefinition = MutableEndpoint({
  name: "FlowSensor",
  deviceType: 774,
  deviceRevision: 3,
  requirements: FlowSensorRequirements,
  behaviors: SupportedBehaviors(
    FlowSensorRequirements.server.mandatory.Identify,
    FlowSensorRequirements.server.mandatory.FlowMeasurement
  )
});
Object.freeze(FlowSensorDeviceDefinition);
const FlowSensorDevice = FlowSensorDeviceDefinition;
export {
  FlowSensorDevice,
  FlowSensorDeviceDefinition,
  FlowSensorRequirements
};
