// @matter/node/dist/esm/devices/pressure-sensor.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import {
  PressureMeasurementServer as BasePressureMeasurementServer
} from "../behaviors/pressure-measurement/PressureMeasurementServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var PressureSensorRequirements;
((PressureSensorRequirements2) => {
  PressureSensorRequirements2.IdentifyServer = BaseIdentifyServer;
  PressureSensorRequirements2.PressureMeasurementServer = BasePressureMeasurementServer;
  PressureSensorRequirements2.server = { mandatory: { Identify: PressureSensorRequirements2.IdentifyServer, PressureMeasurement: PressureSensorRequirements2.PressureMeasurementServer } };
})(PressureSensorRequirements || (PressureSensorRequirements = {}));
const PressureSensorDeviceDefinition = MutableEndpoint({
  name: "PressureSensor",
  deviceType: 773,
  deviceRevision: 3,
  requirements: PressureSensorRequirements,
  behaviors: SupportedBehaviors(
    PressureSensorRequirements.server.mandatory.Identify,
    PressureSensorRequirements.server.mandatory.PressureMeasurement
  )
});
Object.freeze(PressureSensorDeviceDefinition);
const PressureSensorDevice = PressureSensorDeviceDefinition;
export {
  PressureSensorDevice,
  PressureSensorDeviceDefinition,
  PressureSensorRequirements
};
