// @matter/node/dist/esm/devices/variance-test.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { LevelControlServer as BaseLevelControlServer } from "../behaviors/level-control/LevelControlServer.js";
import { ColorControlServer as BaseColorControlServer } from "../behaviors/color-control/ColorControlServer.js";
import { OnOffServer as BaseOnOffServer } from "../behaviors/on-off/OnOffServer.js";
import { ThermostatServer as BaseThermostatServer } from "../behaviors/thermostat/ThermostatServer.js";
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
var TestLevelControl;
((TestLevelControl2) => {
  TestLevelControl2.LevelControlServer = BaseLevelControlServer.with("Lighting", "OnOff").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 }
    }
  });
})(TestLevelControl || (TestLevelControl = {}));
var TestColorControl;
((TestColorControl2) => {
  TestColorControl2.ColorControlServer = BaseColorControlServer.with("ColorTemperature").alter({
    attributes: {
      startUpColorTemperatureMireds: { default: 0 }
    }
  });
})(TestColorControl || (TestColorControl = {}));
var TestOnOff;
((TestOnOff2) => {
  TestOnOff2.OnOffServer = BaseOnOffServer.with("Lighting").alter({ attributes: {} });
})(TestOnOff || (TestOnOff = {}));
var TestThermostat;
((TestThermostat2) => {
  TestThermostat2.ThermostatServer = BaseThermostatServer.with("Heating").alter({ attributes: {} });
})(TestThermostat || (TestThermostat = {}));
var TestIdentify;
((TestIdentify2) => {
  TestIdentify2.IdentifyServer = BaseIdentifyServer.with().alter({ attributes: {} });
})(TestIdentify || (TestIdentify = {}));
var TestDoubleWith;
((TestDoubleWith2) => {
  TestDoubleWith2.LevelControlServer = BaseLevelControlServer.with("Lighting", "OnOff").with();
})(TestDoubleWith || (TestDoubleWith = {}));
export {
  TestColorControl,
  TestDoubleWith,
  TestIdentify,
  TestLevelControl,
  TestOnOff,
  TestThermostat
};
