// @matter/protocol/dist/esm/action/client/subscription/ClientSubscription.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Abort, decamelize, Diagnostic, Logger } from "@matter/general";
const logger = Logger.get("ClientSubscription");
class ClientSubscription {
  request;
  peer;
  /**
   * If the subscription has an async worker, this is the promise associated with the worker.
   */
  done;
  #lifetime;
  #closed;
  #abort;
  #isClosed = false;
  #id = -1;
  constructor({ request, peer, closed, abort, lifetime }) {
    this.#lifetime = lifetime.join(this.kind);
    this.request = request;
    this.peer = peer;
    this.#closed = closed;
    this.#abort = new Abort({
      abort,
      handler: this.close.bind(this)
    });
  }
  get kind() {
    return decamelize(this.constructor.name.replace(/Subscription$/, ""), " ");
  }
  get subscriptionId() {
    return this.#id;
  }
  set subscriptionId(id) {
    this.#id = id;
    this.#lifetime.name = [this.kind, Diagnostic.strong(id)];
  }
  close() {
    if (this.#isClosed) {
      return;
    }
    const closing = this.#lifetime.closing();
    this.#isClosed = true;
    this.#abort();
    this.#closed();
    const unhandledError = (e) => {
      this.logger.error("Unhandled error in subscription to", Diagnostic.strong(this.peer.toString()), e);
    };
    if (this.done) {
      this.done.finally(() => {
        this.request.closed?.();
        closing[Symbol.dispose]();
      }).catch(unhandledError);
    } else {
      try {
        this.request.closed?.();
      } catch (e) {
        unhandledError(e);
      } finally {
        closing[Symbol.dispose]();
      }
    }
  }
  get abort() {
    return this.#abort;
  }
  get logger() {
    return logger;
  }
}
((ClientSubscription2) => {
  ClientSubscription2.NO_SUBSCRIPTION = -1;
})(ClientSubscription || (ClientSubscription = {}));
export {
  ClientSubscription
};
