// @matter/protocol/dist/esm/action/client/subscription/PeerSubscription.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Subscription } from "#interaction/Subscription.js";
import { Diagnostic, Duration, Millis, Seconds } from "@matter/general";
import { ClientSubscription } from "./ClientSubscription.js";
class PeerSubscription extends ClientSubscription {
  interactionModelRevision;
  maxInterval;
  #maxPeerResponseTime;
  isReading = false;
  timeoutAt;
  constructor(config) {
    const {
      maxPeerResponseTime,
      response: { subscriptionId, interactionModelRevision, maxInterval }
    } = config;
    super(config);
    this.subscriptionId = subscriptionId;
    this.interactionModelRevision = interactionModelRevision;
    this.maxInterval = maxInterval;
    this.#maxPeerResponseTime = maxPeerResponseTime;
  }
  get timeout() {
    return Millis(Seconds(this.maxInterval) + this.#maxPeerResponseTime * 2);
  }
  timedOut() {
    this.logger.info(
      "Subscription",
      Diagnostic.strong(Subscription.idStrOf(this.subscriptionId)),
      "to peer",
      this.peer.toString(),
      "timed out after",
      Diagnostic.strong(Duration.format(this.timeout))
    );
    this.close();
  }
}
export {
  PeerSubscription
};
