// @matter/general/dist/cjs/fs/MockFilesystem.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MockFilesystem_exports = {};
__export(MockFilesystem_exports, {
  MockFilesystem: () => MockFilesystem
});
module.exports = __toCommonJS(MockFilesystem_exports);
var import_Bytes = require("../util/Bytes.js");
var import_Directory = require("./Directory.js");
var import_File = require("./File.js");
var import_Filesystem = require("./Filesystem.js");
var import_FilesystemError = require("./FilesystemError.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function createDirNode() {
  return { type: "directory", mtime: /* @__PURE__ */ new Date(), children: /* @__PURE__ */ new Map() };
}
function createFileNode(content) {
  return { type: "file", mtime: /* @__PURE__ */ new Date(), content };
}
class MockFilesystem extends import_Filesystem.Filesystem {
  #root;
  #tempCounter = 0;
  constructor() {
    super();
    this.#root = createDirNode();
  }
  get name() {
    return "";
  }
  get path() {
    return "";
  }
  exists() {
    return Promise.resolve(true);
  }
  stat() {
    return {
      size: 0,
      mtime: this.#root.mtime,
      type: "directory"
    };
  }
  rename() {
    throw new import_File.FileTypeError("Cannot rename root");
  }
  async delete() {
    this.#root.children?.clear();
  }
  async *entries() {
    if (!this.#root.children) {
      return;
    }
    for (const [name, child] of this.#root.children) {
      if (child.type === "directory") {
        yield new MockDirectory(this, [name], this.#root, statOf(child));
      } else {
        yield new MockFile(this, [name], this.#root, statOf(child));
      }
    }
  }
  file(name) {
    return new MockFile(this, [name], this.#root);
  }
  directory(name) {
    return new MockDirectory(this, [name], this.#root);
  }
  async mkdir() {
  }
  async copy(source, target) {
    mockCopy(this.#root, source, target);
  }
  tempFilename() {
    return `tmp-${Date.now()}-${this.#tempCounter++}`;
  }
  tempDirectory() {
    return this.directory(this.tempFilename());
  }
}
function statOf(node) {
  return {
    size: node.type === "file" ? node.content?.length ?? 0 : 0,
    mtime: node.mtime,
    type: node.type
  };
}
function resolvePath(segments, root) {
  let current = root;
  for (const segment of segments) {
    if (current.type !== "directory" || !current.children) {
      return void 0;
    }
    const child = current.children.get(segment);
    if (!child) {
      return void 0;
    }
    current = child;
  }
  return current;
}
function mkdirp(segments, root) {
  let current = root;
  for (const segment of segments) {
    if (!current.children) {
      current.children = /* @__PURE__ */ new Map();
    }
    let child = current.children.get(segment);
    if (!child) {
      child = createDirNode();
      current.children.set(segment, child);
    }
    if (child.type !== "directory") {
      throw new import_File.FileTypeError(`"${segment}" is not a directory`);
    }
    current = child;
  }
  return current;
}
function isBytes(value) {
  return import_Bytes.Bytes.isBytes(value);
}
async function textToBytes(data) {
  let text;
  if (typeof data === "string") {
    text = data;
  } else {
    const lines = Array();
    for await (const line of data) {
      lines.push(line);
    }
    text = lines.join("\n");
  }
  return new TextEncoder().encode(text);
}
class MockFile extends import_File.File {
  #fs;
  #segments;
  #root;
  #cachedStat;
  constructor(fs, segments, root, cachedStat) {
    super();
    this.#fs = fs;
    this.#segments = segments;
    this.#root = root;
    this.#cachedStat = cachedStat;
    mockSegmentsMap.set(this, segments);
  }
  get fs() {
    return this.#fs;
  }
  get path() {
    return this.#segments.join("/");
  }
  async open(purpose, mode) {
    const m = mode ?? "r";
    if (m === "w" || m === "a") {
      mkdirp(this.#segments.slice(0, -1), this.#root);
      const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
      let node = resolvePath(this.#segments, this.#root);
      if (!node) {
        node = createFileNode(new Uint8Array(0));
        parent.children.set(this.name, node);
      }
      if (m === "w") {
        node.content = new Uint8Array(0);
      }
    } else {
      const node = resolvePath(this.#segments, this.#root);
      if (!node) {
        throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
      }
    }
    return new MockFileHandle(this.#fs, this.#segments, this.#root, purpose);
  }
  get name() {
    return this.#segments[this.#segments.length - 1];
  }
  async exists() {
    return resolvePath(this.#segments, this.#root) !== void 0;
  }
  stat() {
    if (this.#cachedStat) {
      return this.#cachedStat;
    }
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    return statOf(node);
  }
  async *readBytes() {
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    if (node.type === "directory") {
      throw new import_File.FileTypeError("Cannot read bytes from a directory");
    }
    if (node.content && node.content.length > 0) {
      yield node.content;
    }
  }
  async *readText(options) {
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    if (node.type === "directory") {
      throw new import_File.FileTypeError("Cannot read text from a directory");
    }
    const text = new TextDecoder().decode(node.content ?? new Uint8Array());
    if (options?.lines) {
      for (const line of text.split("\n")) {
        yield line;
      }
    } else {
      yield text;
    }
  }
  async write(data) {
    let content;
    if (typeof data === "string") {
      content = await textToBytes(data);
    } else if (isBytes(data)) {
      content = import_Bytes.Bytes.of(data);
    } else {
      const iter = Symbol.asyncIterator in data ? data[Symbol.asyncIterator]() : data[Symbol.iterator]();
      const first = await iter.next();
      if (first.done) {
        content = new Uint8Array(0);
      } else if (typeof first.value === "string") {
        const lines = [first.value];
        while (true) {
          const next = await iter.next();
          if (next.done) break;
          lines.push(next.value);
        }
        content = new TextEncoder().encode(lines.join("\n"));
      } else {
        const firstBytes = import_Bytes.Bytes.of(first.value);
        const chunks = [firstBytes];
        let totalLength = firstBytes.length;
        while (true) {
          const next = await iter.next();
          if (next.done) break;
          const bytes = import_Bytes.Bytes.of(next.value);
          chunks.push(bytes);
          totalLength += bytes.length;
        }
        if (chunks.length === 1) {
          content = chunks[0];
        } else {
          content = new Uint8Array(totalLength);
          let offset = 0;
          for (const chunk of chunks) {
            content.set(chunk, offset);
            offset += chunk.length;
          }
        }
      }
    }
    const parentNode = mkdirp(this.#segments.slice(0, -1), this.#root);
    const existing = parentNode.children?.get(this.name);
    if (existing && existing.type === "directory") {
      throw new import_File.FileTypeError(`"${this.name}" is a directory`);
    }
    parentNode.children.set(this.name, createFileNode(content));
  }
  async rename(newName) {
    const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
    if (!parent?.children) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    const node = parent.children.get(this.name);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    parent.children.delete(this.name);
    parent.children.set(newName, node);
    this.#segments[this.#segments.length - 1] = newName;
  }
  async delete() {
    const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
    if (parent?.children) {
      parent.children.delete(this.name);
    }
  }
}
class MockDirectory extends import_Directory.Directory {
  #fs;
  #segments;
  #root;
  #cachedStat;
  constructor(fs, segments, root, cachedStat) {
    super();
    this.#fs = fs;
    this.#segments = segments;
    this.#root = root;
    this.#cachedStat = cachedStat;
    mockSegmentsMap.set(this, segments);
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#segments[this.#segments.length - 1];
  }
  get path() {
    return this.#segments.join("/");
  }
  async exists() {
    const node = resolvePath(this.#segments, this.#root);
    return node !== void 0 && node.type === "directory";
  }
  stat() {
    if (this.#cachedStat) {
      return this.#cachedStat;
    }
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`Directory not found: ${this.#segments.join("/")}`);
    }
    return statOf(node);
  }
  rename(newName) {
    const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
    if (!parent?.children) {
      return Promise.reject(new import_FilesystemError.FileNotFoundError(`Directory not found: ${this.#segments.join("/")}`));
    }
    const node = parent.children.get(this.name);
    if (!node) {
      return Promise.reject(new import_FilesystemError.FileNotFoundError(`Directory not found: ${this.#segments.join("/")}`));
    }
    parent.children.delete(this.name);
    parent.children.set(newName, node);
    this.#segments[this.#segments.length - 1] = newName;
    return Promise.resolve();
  }
  async delete() {
    const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
    if (parent?.children) {
      parent.children.delete(this.name);
    }
  }
  async *entries() {
    const node = resolvePath(this.#segments, this.#root);
    if (!node || node.type !== "directory") {
      throw new import_FilesystemError.FileNotFoundError(`Directory not found: ${this.#segments.join("/")}`);
    }
    if (!node.children) {
      return;
    }
    for (const [name, child] of node.children) {
      if (child.type === "directory") {
        yield new MockDirectory(this.#fs, [...this.#segments, name], this.#root, statOf(child));
      } else {
        yield new MockFile(this.#fs, [...this.#segments, name], this.#root, statOf(child));
      }
    }
  }
  file(name) {
    return new MockFile(this.#fs, [...this.#segments, name], this.#root);
  }
  directory(name) {
    return new MockDirectory(this.#fs, [...this.#segments, name], this.#root);
  }
  async mkdir() {
    mkdirp(this.#segments, this.#root);
  }
  async copy(source, target) {
    mockCopy(this.#root, source, target);
  }
}
const mockSegmentsMap = /* @__PURE__ */ new WeakMap();
function deepCloneNode(node) {
  if (node.type === "file") {
    return {
      type: "file",
      mtime: new Date(node.mtime.getTime()),
      content: node.content ? new Uint8Array(node.content) : void 0
    };
  }
  const children = /* @__PURE__ */ new Map();
  if (node.children) {
    for (const [name, child] of node.children) {
      children.set(name, deepCloneNode(child));
    }
  }
  return { type: "directory", mtime: new Date(node.mtime.getTime()), children };
}
function resolveArg(arg) {
  if (typeof arg === "string") {
    return arg.split("/").filter((s) => s !== "");
  }
  const segments = mockSegmentsMap.get(arg);
  if (segments) {
    return segments;
  }
  if (arg.name === "") {
    return [];
  }
  throw new import_File.FileTypeError("Cannot resolve non-mock filesystem node in MockFilesystem.copy()");
}
function mockCopy(root, source, target) {
  const sourceSegments = resolveArg(source);
  const targetSegments = resolveArg(target);
  let sourceNode;
  if (sourceSegments.length === 0) {
    sourceNode = root;
  } else {
    sourceNode = resolvePath(sourceSegments, root);
  }
  if (!sourceNode) {
    throw new import_FilesystemError.FileNotFoundError(`Source not found: ${sourceSegments.join("/")}`);
  }
  const cloned = deepCloneNode(sourceNode);
  if (targetSegments.length === 0) {
    if (cloned.children) {
      if (!root.children) {
        root.children = /* @__PURE__ */ new Map();
      }
      for (const [name, child] of cloned.children) {
        root.children.set(name, child);
      }
    }
    return;
  }
  const parentSegments = targetSegments.slice(0, -1);
  const parent = parentSegments.length > 0 ? mkdirp(parentSegments, root) : root;
  const targetName = targetSegments[targetSegments.length - 1];
  if (!parent.children) {
    parent.children = /* @__PURE__ */ new Map();
  }
  parent.children.set(targetName, cloned);
}
class MockFileHandle extends import_File.File.Handle {
  #fs;
  #segments;
  #root;
  purpose;
  #closed = false;
  constructor(fs, segments, root, purpose) {
    super();
    this.#fs = fs;
    this.#segments = segments;
    this.#root = root;
    this.purpose = purpose;
    mockSegmentsMap.set(this, segments);
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#segments[this.#segments.length - 1];
  }
  get path() {
    return this.#segments.join("/");
  }
  async exists() {
    return resolvePath(this.#segments, this.#root) !== void 0;
  }
  stat() {
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    return statOf(node);
  }
  async *readBytes() {
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    if (node.content && node.content.length > 0) {
      yield node.content;
    }
  }
  async *readText(options) {
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    const text = new TextDecoder().decode(node.content ?? new Uint8Array());
    if (options?.lines) {
      for (const line of text.split("\n")) {
        yield line;
      }
    } else {
      yield text;
    }
  }
  async write(data) {
    const file = new MockFile(this.#fs, this.#segments, this.#root);
    return file.write(data);
  }
  async open(_purpose) {
    return this;
  }
  async writeHandle(data) {
    if (this.#closed) {
      throw new import_File.FileTypeError("File handle is closed");
    }
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    const newBytes = typeof data === "string" ? new TextEncoder().encode(data) : data instanceof Uint8Array ? data : import_Bytes.Bytes.of(data);
    const existing = node.content ?? new Uint8Array(0);
    const combined = new Uint8Array(existing.length + newBytes.length);
    combined.set(existing);
    combined.set(newBytes, existing.length);
    node.content = combined;
    node.mtime = /* @__PURE__ */ new Date();
  }
  cursor(max) {
    return new MockCursor(this.#segments, this.#root, max);
  }
  async truncate(size) {
    if (this.#closed) {
      throw new import_File.FileTypeError("File handle is closed");
    }
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    const s = size ?? 0;
    if (s === 0) {
      node.content = new Uint8Array(0);
    } else {
      const content = node.content ?? new Uint8Array(0);
      node.content = content.slice(0, s);
    }
    node.mtime = /* @__PURE__ */ new Date();
  }
  async fsync() {
    if (this.#closed) {
      throw new import_File.FileTypeError("File handle is closed");
    }
  }
  async close() {
    this.#closed = true;
  }
  async rename(newName) {
    const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
    if (!parent?.children) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    const node = parent.children.get(this.name);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    parent.children.delete(this.name);
    parent.children.set(newName, node);
    this.#segments[this.#segments.length - 1] = newName;
  }
  async delete() {
    const parent = resolvePath(this.#segments.slice(0, -1), this.#root);
    if (parent?.children) {
      parent.children.delete(this.name);
    }
    this.#closed = true;
  }
}
class MockCursor extends import_File.File.Cursor {
  #segments;
  #root;
  constructor(segments, root, max) {
    super(max);
    this.#segments = segments;
    this.#root = root;
  }
  async readAt(position, length, copy) {
    const node = resolvePath(this.#segments, this.#root);
    if (!node) {
      throw new import_FilesystemError.FileNotFoundError(`File not found: ${this.#segments.join("/")}`);
    }
    const content = node.content ?? new Uint8Array(0);
    const start = Math.min(position, content.length);
    const end = Math.min(start + length, content.length);
    return copy ? content.slice(start, end) : content.subarray(start, end);
  }
}
