// @matter/node/dist/esm/storage/client/RemoteWriteParticipant.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "@matter/general";
const logger = Logger.get("RemoteWriteParticipant");
class RemoteWriteParticipant {
  #request = [];
  #writer;
  #snapshots = /* @__PURE__ */ new Map();
  /**
   * There is one participant for each transaction/writer pair.  We use the writer function itself as the dedup key.
   */
  get role() {
    return this.#writer;
  }
  /**
   * Add an attribute update to the write request.
   *
   * The optional {@link snapshot} captures the pre-write values for the affected attributes plus a
   * {@link RemoteWriteParticipant.Compensator} the participant invokes if the remote rejects the write.
   */
  set(endpointNumber, behaviorId, values, snapshot) {
    this.#request.push({
      number: endpointNumber,
      behaviorId,
      values
    });
    if (snapshot) {
      const key = snapshotKey(endpointNumber, behaviorId);
      const existing = this.#snapshots.get(key);
      if (existing) {
        Object.assign(existing.writtenValues, values);
        for (const [k, v] of Object.entries(snapshot.previousValues)) {
          if (!(k in existing.previousValues)) {
            existing.previousValues[k] = v;
          }
        }
      } else {
        this.#snapshots.set(key, {
          compensator: snapshot.compensator,
          previousValues: { ...snapshot.previousValues },
          writtenValues: { ...values }
        });
      }
    }
  }
  async commit2() {
    if (!this.#request.length) {
      return;
    }
    const request = this.#request;
    const snapshots = this.#snapshots;
    this.#request = [];
    this.#snapshots = /* @__PURE__ */ new Map();
    await this.#writer(request, (failures) => this.#compensate(snapshots, failures));
  }
  async #compensate(snapshots, failures) {
    if (!snapshots.size) {
      return;
    }
    const failedByCluster = /* @__PURE__ */ new Map();
    for (const f of failures) {
      const key = snapshotKey(f.path.endpointId, String(f.path.clusterId));
      let ids = failedByCluster.get(key);
      if (!ids) {
        ids = /* @__PURE__ */ new Set();
        failedByCluster.set(key, ids);
      }
      ids.add(String(f.path.attributeId));
    }
    for (const [key, snapshot] of snapshots) {
      const failedIds = failedByCluster.get(key);
      if (!failedIds?.size) {
        continue;
      }
      try {
        await snapshot.compensator.compensate(failedIds, snapshot.previousValues, snapshot.writtenValues);
      } catch (compensateError) {
        logger.warn(`Failed to restore local state for ${key} after remote write decline:`, compensateError);
      }
    }
  }
  rollback() {
    this.#request = [];
    this.#snapshots = /* @__PURE__ */ new Map();
  }
  toString() {
    return `remote-writer`;
  }
  constructor(writer) {
    this.#writer = writer;
  }
}
function snapshotKey(endpointNumber, behaviorId) {
  return `${endpointNumber}|${behaviorId}`;
}
export {
  RemoteWriteParticipant
};
