// @matter/general/dist/esm/net/dns-sd/DnssdName.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DnsRecordType } from "#codec/DnsCodec.js";
import { Logger } from "#log/Logger.js";
import { Time } from "#time/Time.js";
import { Millis } from "#time/TimeUnit.js";
import { Bytes } from "#util/Bytes.js";
import { BasicObservable } from "#util/Observable.js";
import { DnssdParameters } from "./DnssdParameters.js";
const logger = Logger.get("DnssdName");
const DEFAULT_TTL_GRACE_FACTOR = 1.05;
class DnssdName extends BasicObservable {
  constructor(qname, context) {
    super((e) => logger.error(`Unhandled error in observer for DNS name "${qname}":`, e));
    this.qname = qname;
    this.#context = context;
  }
  qname;
  #context;
  #records = /* @__PURE__ */ new Map();
  #recordCount = 0;
  #changes;
  #notified;
  #maybeDeleting;
  #parameters;
  #dependencies;
  #nullObserver;
  off(observer) {
    super.off(observer);
    this.#deleteIfUnused();
  }
  async close() {
    if (this.#notified) {
      await this.#notified;
    }
    if (this.#maybeDeleting) {
      await this.#maybeDeleting;
    }
  }
  get records() {
    return this.#records.values();
  }
  get parameters() {
    if (this.#parameters === void 0) {
      const raw = /* @__PURE__ */ new Map();
      const txtRecords = new Array();
      for (const record of this.#records.values()) {
        if (record.recordType === DnsRecordType.TXT) {
          txtRecords.push(record);
        }
      }
      txtRecords.sort((a, b) => b.installedAt - a.installedAt);
      for (const record of txtRecords) {
        for (const entry of record.value) {
          const bytes = Bytes.of(entry);
          if (bytes.byteLength === 0) {
            continue;
          }
          const eqIndex = bytes.indexOf(61);
          if (eqIndex === 0) {
            continue;
          }
          const key = eqIndex === -1 ? Bytes.toString(bytes) : Bytes.toString(bytes.subarray(0, eqIndex));
          if (raw.has(key)) {
            continue;
          }
          raw.set(key, eqIndex === -1 ? new Uint8Array(0) : bytes.subarray(eqIndex + 1));
        }
      }
      this.#parameters = new DnssdParameters(raw);
    }
    return this.#parameters;
  }
  get isDiscovered() {
    return !!this.#recordCount;
  }
  installRecord(record, options) {
    const key = keyOf(record);
    if (key === void 0) {
      this.#deleteIfUnused();
      return false;
    }
    const oldRecord = this.#records.get(key);
    if (oldRecord) {
      this.#context.unregisterForExpiration(oldRecord);
    } else {
      this.#recordCount++;
    }
    const at = options?.installedAt ?? Time.nowMs;
    const isHostRecord = record.recordType === DnsRecordType.A || record.recordType === DnsRecordType.AAAA;
    const recordWithExpire = {
      ...record,
      installedAt: at,
      expiresAt: at + Millis(Math.round(record.ttl * this.#context.ttlGraceFactor)),
      ...isHostRecord ? { sourceIntf: record.recordType === DnsRecordType.AAAA ? options?.sourceIntf : void 0 } : {}
    };
    this.#records.set(key, recordWithExpire);
    if (record.recordType === DnsRecordType.TXT) {
      this.#parameters = void 0;
    }
    this.#context.registerForExpiration(recordWithExpire);
    if (record.recordType === DnsRecordType.SRV && !this.#dependencies?.has(key)) {
      const dependency = this.#context.get(record.value.target);
      dependency.on(this.#nullObserver ??= () => void 0);
      (this.#dependencies ??= /* @__PURE__ */ new Map()).set(key, dependency);
    }
    this.#notify("update", key, recordWithExpire);
  }
  deleteRecord(record, ifOlderThan) {
    const key = keyOf(record);
    if (key === void 0) {
      this.#deleteIfUnused();
      return;
    }
    const recordWithExpire = this.#records?.get(key);
    if (!recordWithExpire) {
      this.#deleteIfUnused();
      return;
    }
    if (ifOlderThan !== void 0 && recordWithExpire.installedAt >= ifOlderThan) {
      return;
    }
    this.#records.delete(key);
    this.#recordCount--;
    if (record.recordType === DnsRecordType.TXT) {
      this.#parameters = void 0;
    }
    const dependency = this.#dependencies?.get(key);
    if (dependency) {
      this.#dependencies.delete(key);
      dependency.off(this.#nullObserver);
    }
    this.#context.unregisterForExpiration(recordWithExpire);
    if (this.#deleteIfUnused()) {
      return;
    }
    this.#notify("delete", key, recordWithExpire);
  }
  /**
   * Delete if unused.
   *
   * This is async so we assess whether deletion is appropriate after a batch of updates.
   */
  #deleteIfUnused() {
    if (this.isObserved || this.isDiscovered) {
      return false;
    }
    if (this.#maybeDeleting) {
      return true;
    }
    const maybeDelete = async () => {
      this.#maybeDeleting = void 0;
      if (this.isObserved || this.isDiscovered) {
        return;
      }
      this.#context.delete(this);
    };
    this.#maybeDeleting = maybeDelete();
    return true;
  }
  /**
   * Notification of observers.
   *
   * This is async so we coalesce changes into a single notification.
   */
  #notify(kind, key, record) {
    if (this.#changes === void 0) {
      this.#changes = /* @__PURE__ */ new Map();
    }
    this.#changes.set(key, { kind, record });
    if (this.#notified) {
      return;
    }
    const notify = async () => {
      while (this.#changes?.size) {
        const changes = { name: this };
        for (const { kind: kind2, record: record2 } of this.#changes.values()) {
          const key2 = `${kind2}d`;
          const list = changes[key2];
          if (list === void 0) {
            changes[key2] = [record2];
          } else {
            list.push(record2);
          }
        }
        this.#changes.clear();
        await this.emit(changes);
      }
      this.#notified = void 0;
    };
    this.#notified = notify();
  }
}
function keyOf(record) {
  switch (record.recordType) {
    case DnsRecordType.A:
    case DnsRecordType.AAAA:
    case DnsRecordType.PTR:
      if (typeof record.value === "string") {
        return `${record.recordType} ${record.value}`;
      }
      break;
    case DnsRecordType.SRV:
      if (typeof record.value === "object") {
        const srv = record.value;
        return `${record.recordType} ${srv.target}:${srv.port}`;
      }
      break;
    case DnsRecordType.TXT:
      if (Array.isArray(record.value)) {
        const keys = record.value.map((entry) => Bytes.toHex(entry));
        keys.sort();
        return `${record.recordType} ${keys.join(" ")}`;
      }
      break;
  }
}
export {
  DEFAULT_TTL_GRACE_FACTOR,
  DnssdName
};
