// @matter/protocol/dist/esm/action/request/Invoke.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SessionParameters } from "#session/SessionParameters.js";
import { Diagnostic, isObject } from "@matter/general";
import {
  FabricIndex,
  ObjectSchema,
  TlvOfModel,
  TlvVoid
} from "@matter/types";
import { MalformedRequestError } from "./MalformedRequestError.js";
import { resolvePathForSpecifier, Specifier } from "./Specifier.js";
function Invoke(optionsOrData, ...commands) {
  if (optionsOrData === void 0) {
    throw new MalformedRequestError(`Invocation requires at least one command`);
  }
  let options;
  if ("commands" in optionsOrData) {
    options = optionsOrData;
    commands = [...optionsOrData.commands, ...commands];
  } else {
    commands = [optionsOrData, ...commands];
    options = {};
  }
  const {
    interactionModelRevision = SessionParameters.fallbacks.interactionModelRevision,
    suppressResponse = false,
    timeout,
    expectedProcessingTime,
    useExtendedFailSafeMessageResponseTimeout = false,
    skipValidation = false
  } = options;
  let timedRequest = !!options.timed || !!timeout;
  if (!commands?.length) {
    throw new MalformedRequestError(`Invocation requires at least one command`);
  }
  if (commands.length > 1) {
    const commandRefs = /* @__PURE__ */ new Set();
    for (const { commandRef } of commands) {
      if (commandRef === void 0) {
        throw new MalformedRequestError(`CommandRef required when invoking multiple commands`);
      }
      if (commandRefs.has(commandRef)) {
        throw new MalformedRequestError(`Duplicate commandRef ${commandRef} in multiple command invoke`);
      }
      commandRefs.add(commandRef);
    }
  }
  const commandMap = /* @__PURE__ */ new Map();
  const invokeRequests = commands.map((cmd) => {
    const cmdData = Invoke.Command(cmd, skipValidation);
    if (options.timed !== false) {
      timedRequest ||= !!cmdData.timed;
    }
    commandMap.set(cmdData.commandRef, cmd);
    return cmdData;
  });
  return {
    timedRequest,
    timeout,
    invokeRequests,
    interactionModelRevision,
    suppressResponse,
    expectedProcessingTime,
    useExtendedFailSafeMessageResponseTimeout,
    // Additional meta-data for client side processing
    skipValidation,
    commands: commandMap,
    [Diagnostic.value]: () => Diagnostic.list(
      commands.map((cmd) => {
        const { commandRef } = cmd;
        const fields = "fields" in cmd ? cmd.fields : void 0;
        return [
          Diagnostic.strong(Invoke.isLegacy(cmd) ? "(legacy)" : resolvePathForSpecifier(cmd)),
          "with",
          isObject(fields) ? Diagnostic.dict(fields) : "(no payload)",
          commandRef !== void 0 ? `(ref ${commandRef})` : ""
        ];
      })
    )
  };
}
((Invoke2) => {
  function Command(request, skipValidation = false) {
    let requestSchema;
    let requestId;
    let timed;
    if (typeof request.command !== "string" && "requestId" in request.command) {
      const legacy = request.command;
      requestSchema = legacy.requestSchema;
      requestId = legacy.requestId;
      timed = legacy.timed;
    } else {
      const command = Invoke2.commandOf(request);
      requestSchema = TlvOfModel(command.schema) ?? TlvVoid;
      requestId = command.id;
      timed = command.schema.effectiveAccess.timed === true;
    }
    const { commandRef } = request;
    let fields = "fields" in request ? request.fields : void 0;
    if (requestSchema instanceof ObjectSchema) {
      if (fields === void 0) {
        fields = {};
      }
      if (requestSchema.isFabricScoped && fields.fabricIndex === void 0) {
        fields.fabricIndex = FabricIndex.NO_FABRIC;
      }
    }
    if (!skipValidation) {
      requestSchema.validate(fields);
    }
    const commandFields = requestSchema.encodeTlv(fields);
    const result = {
      commandPath: {
        // Endpoint id is added below if not wildcard
        clusterId: Specifier.clusterFor(request.cluster).id,
        commandId: requestId
      },
      commandFields,
      timed: timed ?? false,
      commandRef
    };
    const endpointId = Specifier.endpointIdOf(request);
    if (endpointId !== void 0) {
      result.commandPath.endpointId = endpointId;
    }
    return result;
  }
  Invoke2.Command = Command;
  function isLegacy(request) {
    return typeof request.command !== "string" && "requestId" in request.command;
  }
  Invoke2.isLegacy = isLegacy;
  function ConcreteCommandRequest(data) {
    if (data.endpoint === void 0) {
      throw new MalformedRequestError(`ConcreteCommandRequest requires an endpoint`);
    }
    return data;
  }
  Invoke2.ConcreteCommandRequest = ConcreteCommandRequest;
  function WildcardCommandRequest(data) {
    if ("endpoint" in data && data.endpoint !== void 0) {
      throw new MalformedRequestError(`ConcreteCommandRequest must not have an endpoint`);
    }
    return data;
  }
  Invoke2.WildcardCommandRequest = WildcardCommandRequest;
  function commandOf(request) {
    if (typeof request.command === "string") {
      const cluster = Specifier.clusterFor(request.cluster);
      const command = cluster.commands?.[request.command];
      if (command === void 0) {
        throw new MalformedRequestError(`Cluster ${cluster.name} does not define command ${request.command}`);
      }
      return command;
    }
    return request.command;
  }
  Invoke2.commandOf = commandOf;
})(Invoke || (Invoke = {}));
export {
  Invoke
};
