// @matter/protocol/dist/esm/action/request/Specifier.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { camelize } from "@matter/general";
import { DataModelPath } from "@matter/model";
import { Global } from "@matter/types";
import { MalformedRequestError } from "./MalformedRequestError.js";
var Specifier;
((Specifier2) => {
  function clusterFor(specifier) {
    if (specifier === void 0) {
      throw new MalformedRequestError(`Cannot designate command "${specifier}" without a cluster`);
    }
    if ("cluster" in specifier) {
      return specifier["cluster"];
    }
    return specifier;
  }
  Specifier2.clusterFor = clusterFor;
  function attributeFor(cluster, specifier) {
    if (typeof specifier === "object") {
      return specifier;
    }
    if (cluster === void 0) {
      const attr2 = Global.attributes[specifier];
      if (attr2 === void 0) {
        throw new MalformedRequestError(`Cannot designate attribute "${specifier}" without a cluster`);
      }
      return attr2;
    }
    const attr = cluster.attributes?.[specifier] ?? Global.attributes[specifier];
    if (attr === void 0) {
      throw new MalformedRequestError(`Cluster ${cluster.name} does not define attribute ${specifier}`);
    }
    return attr;
  }
  Specifier2.attributeFor = attributeFor;
  function commandFor(cluster, specifier) {
    if (typeof specifier !== "string") {
      return specifier;
    }
    if (cluster === void 0) {
      throw new MalformedRequestError(`Cannot designate command "${specifier}" without a cluster`);
    }
    const command = cluster.commands?.[specifier];
    if (command === void 0) {
      throw new MalformedRequestError(`Cluster ${cluster.name} does not define command ${specifier}`);
    }
    return command;
  }
  Specifier2.commandFor = commandFor;
  function eventFor(cluster, specifier) {
    if (typeof specifier === "object") {
      return specifier;
    }
    if (cluster === void 0) {
      throw new MalformedRequestError(`Cannot designate event "${specifier}" without a cluster`);
    }
    const event = cluster.events?.[specifier];
    if (event === void 0) {
      throw new MalformedRequestError(`Cluster ${cluster.name} does not define event ${specifier}`);
    }
    return event;
  }
  Specifier2.eventFor = eventFor;
  function clusterOf(request) {
    if (request.cluster) {
      return Specifier2.clusterFor(request.cluster);
    }
    return void 0;
  }
  Specifier2.clusterOf = clusterOf;
  function endpointIdOf(request) {
    if (typeof request.endpoint === "number") {
      return request.endpoint;
    }
    return request.endpoint?.number;
  }
  Specifier2.endpointIdOf = endpointIdOf;
})(Specifier || (Specifier = {}));
function resolvePathForSpecifier(location) {
  const endpointId = Specifier.endpointIdOf(location);
  const cluster = location.cluster ? Specifier.clusterFor(location.cluster) : void 0;
  let dmPath = new DataModelPath(endpointId ?? "*", "endpoint");
  dmPath = dmPath.at(cluster ? camelize(cluster.name) : "*", "cluster");
  if (location.attribute !== void 0) {
    const name = nameForElement(location.attribute, cluster && Specifier.attributeFor(cluster, location.attribute));
    dmPath = dmPath.at("state").at(name);
    if (location.listIndex === null) {
      dmPath = dmPath.at("[ADD]", "marker");
    }
    return dmPath;
  }
  if (location.event !== void 0) {
    const name = nameForElement(location.event, cluster && Specifier.eventFor(cluster, location.event));
    dmPath = dmPath.at("events").at(name);
    if (location.isUrgent) {
      dmPath = dmPath.at("!", "marker");
    }
    return dmPath;
  }
  if (location.command !== void 0) {
    const name = nameForElement(location.command, cluster && Specifier.commandFor(cluster, location.command));
    return dmPath.at(name);
  }
  return dmPath.at("*", "element");
  function nameForElement(specifier, resolved) {
    if (typeof specifier === "string") {
      return camelize(specifier);
    }
    if (resolved?.name !== void 0) {
      return camelize(resolved.name);
    }
    if (resolved?.id !== void 0) {
      return `0x${resolved.id.toString(16)}`;
    }
    return "?";
  }
}
export {
  Specifier,
  resolvePathForSpecifier
};
