// @matter/protocol/dist/cjs/protocol/DeviceCommissioner.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DeviceCommissioner_exports = {};
__export(DeviceCommissioner_exports, {
  CommissioningConfigProvider: () => CommissioningConfigProvider,
  DeviceCommissioner: () => DeviceCommissioner
});
module.exports = __toCommonJS(DeviceCommissioner_exports);
var import_CommissioningMode = require("#advertisement/CommissioningMode.js");
var import_ServiceDescription = require("#advertisement/ServiceDescription.js");
var import_FabricManager = require("#fabric/FabricManager.js");
var import_SecureChannelProtocol = require("#securechannel/SecureChannelProtocol.js");
var import_PaseServer = require("#session/pase/PaseServer.js");
var import_SessionManager = require("#session/SessionManager.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_administrator_commissioning = require("@matter/types/clusters/administrator-commissioning");
var import_DeviceAdvertiser = require("./DeviceAdvertiser.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("DeviceCommissioner");
class CommissioningConfigProvider {
}
class DeviceCommissioner {
  #context;
  #failsafeContext;
  #windowStatus = import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
  #activeDiscriminator;
  #activeCommissioningEndCallback;
  #observers = new import_general.ObserverGroup(this);
  #commissioningTimeout;
  constructor(context) {
    this.#context = context;
    this.#observers.on(this.#context.secureChannelProtocol.tooManyPaseErrors, async () => {
      logger.info("Maximum number of PASE pairing errors reached, canceling commissioning");
      await this.endCommissioning();
    });
  }
  static [import_general.Environmental.create](env) {
    const instance = new DeviceCommissioner({
      fabrics: env.get(import_FabricManager.FabricManager),
      sessions: env.get(import_SessionManager.SessionManager),
      advertiser: env.get(import_DeviceAdvertiser.DeviceAdvertiser),
      secureChannelProtocol: env.get(import_SecureChannelProtocol.SecureChannelProtocol),
      commissioningConfig: env.get(CommissioningConfigProvider)
    });
    env.set(DeviceCommissioner, instance);
    return instance;
  }
  get failsafeContext() {
    this.assertFailsafeArmed();
    return this.#failsafeContext;
  }
  async allowEnhancedCommissioning(discriminator, paseServer, commissioningEndCallback) {
    if (this.#windowStatus === import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.BasicWindowOpen) {
      throw new import_general.MatterFlowError(
        "Basic commissioning window is already open! Cannot set enhanced commissioning mode."
      );
    }
    this.#context.secureChannelProtocol.setPaseCommissioner(paseServer);
    this.#becomeCommissionable(
      import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen,
      commissioningEndCallback,
      discriminator
    );
  }
  async allowBasicCommissioning(commissioningEndCallback) {
    if (this.#windowStatus === import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen) {
      throw new import_general.MatterFlowError(
        "Enhanced commissioning window is already open! Cannot set basic commissioning mode."
      );
    }
    this.#context.secureChannelProtocol.setPaseCommissioner(
      await import_PaseServer.PaseServer.fromPin(this.#context.sessions, this.#context.commissioningConfig.values.passcode, {
        iterations: 1e3,
        salt: this.#context.fabrics.crypto.randomBytes(32)
      })
    );
    this.#becomeCommissionable(
      import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.BasicWindowOpen,
      commissioningEndCallback
    );
  }
  beginTimed(failsafeContext) {
    this.#failsafeContext = failsafeContext;
    failsafeContext.commissioned.on(async () => await this.endCommissioning());
    failsafeContext.construction.change.on((status) => {
      if (status === import_general.Lifecycle.Status.Destroyed) {
        this.#failsafeContext = void 0;
      }
    });
  }
  get isFailsafeArmed() {
    return this.#failsafeContext !== void 0;
  }
  assertFailsafeArmed(message) {
    if (this.isFailsafeArmed) return;
    throw new import_types.StatusResponseError(
      message ?? "Failsafe timer needs to be armed to execute this action.",
      import_types.StatusCode.FailsafeRequired
    );
  }
  #enterCommissioningMode(windowStatus, discriminator) {
    this.#cancelTimeout();
    this.#windowStatus = windowStatus;
    const commissioningConfig = this.#context.commissioningConfig.values;
    const advertisementWindowS = commissioningConfig.advertisementWindow ?? import_types.STANDARD_COMMISSIONING_TIMEOUT;
    const mode = windowStatus === import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen ? import_CommissioningMode.CommissioningMode.Enhanced : import_CommissioningMode.CommissioningMode.Basic;
    this.#context.advertiser.enterCommissioningMode(
      import_ServiceDescription.ServiceDescription.Commissionable({
        ...commissioningConfig.productDescription,
        mode,
        discriminator: discriminator ?? commissioningConfig.discriminator
      })
    );
    this.#commissioningTimeout = import_general.Time.getTimer(
      "Commissioning timeout",
      advertisementWindowS,
      this.endCommissioning.bind(this)
    );
  }
  #becomeCommissionable(windowStatus, activeCommissioningEndCallback, discriminator) {
    if (this.#windowStatus !== import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      if (this.#windowStatus !== windowStatus) {
        throw new import_general.InternalError(
          `Commissioning mode ${windowStatus} request but already in mode ${this.#windowStatus}`
        );
      }
      if (this.#activeCommissioningEndCallback && this.#activeCommissioningEndCallback !== activeCommissioningEndCallback) {
        throw new import_general.InternalError(`Already in commissioning mode with a different callback`);
      }
    }
    if (this.#windowStatus === windowStatus && (discriminator === void 0 || discriminator === this.#activeDiscriminator)) {
      this.#enterCommissioningMode(this.#windowStatus, this.#activeDiscriminator);
      return;
    }
    if (this.#windowStatus !== import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      throw new import_general.InternalError(`Commissioning window already open with different mode (${this.#windowStatus})!`);
    }
    if (this.#activeCommissioningEndCallback !== void 0) {
      throw new import_general.InternalError("Commissioning window already open with different callback!");
    }
    this.#activeCommissioningEndCallback = activeCommissioningEndCallback;
    this.#activeDiscriminator = discriminator;
    this.#enterCommissioningMode(windowStatus, discriminator);
  }
  async endCommissioning() {
    if (this.#windowStatus === import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      return;
    }
    this.#cancelTimeout();
    await this.#context.secureChannelProtocol.removePaseCommissioner();
    this.#windowStatus = import_administrator_commissioning.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
    if (this.#activeCommissioningEndCallback !== void 0) {
      const activeCommissioningEndCallback = this.#activeCommissioningEndCallback;
      this.#activeCommissioningEndCallback = void 0;
      await activeCommissioningEndCallback();
    }
    await this.#context.advertiser.exitCommissioningMode();
    logger.debug("No longer commissioning");
  }
  async close() {
    this.#cancelTimeout();
    this.#observers.close();
    await this.endCommissioning();
    if (this.#failsafeContext) {
      await this.#failsafeContext.close();
      this.#failsafeContext = void 0;
    }
  }
  #cancelTimeout() {
    if (!this.#commissioningTimeout) {
      return;
    }
    this.#commissioningTimeout.stop();
    this.#commissioningTimeout = void 0;
  }
}
