// @matter/protocol/dist/cjs/protocol/ExchangeProvider.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ExchangeProvider_exports = {};
__export(ExchangeProvider_exports, {
  DedicatedChannelExchangeProvider: () => DedicatedChannelExchangeProvider,
  ExchangeProvider: () => ExchangeProvider
});
module.exports = __toCommonJS(ExchangeProvider_exports);
var import_NodeSession = require("#session/NodeSession.js");
var import_types = require("@matter/types");
var import_MRP = require("./MRP.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ExchangeProvider {
  constructor(exchangeManager) {
    this.exchangeManager = exchangeManager;
  }
  exchangeManager;
  /**
   * Dedicated secure session backing this provider, if any.
   *
   * Providers that multiplex sessions (e.g. {@link PeerExchangeProvider}) return `undefined`.
   * Consumers can use this to react to session-lifecycle events without casting.
   */
  get session() {
    return void 0;
  }
  /**
   * Ensure the peer is reachable without creating an exchange.
   *
   * The default implementation is a no-op (already connected).
   */
  async connect(_options) {
  }
}
class DedicatedChannelExchangeProvider extends ExchangeProvider {
  #session;
  constructor(exchangeManager, session) {
    super(exchangeManager);
    this.#session = session;
  }
  get maxPathsPerInvoke() {
    return this.#session.parameters.maxPathsPerInvoke;
  }
  async initiateExchange() {
    return this.exchangeManager.initiateExchangeForSession(this.#session, import_types.INTERACTION_PROTOCOL_ID);
  }
  get channelType() {
    return this.#session.channel.transportChannel.type;
  }
  get session() {
    return this.#session;
  }
  maximumPeerResponseTime(expectedProcessingTime = import_MRP.MRP.DEFAULT_EXPECTED_PROCESSING_TIME, includeMaximumSendingTime) {
    return this.exchangeManager.calculateMaximumPeerResponseTimeMsFor(
      this.#session,
      expectedProcessingTime,
      includeMaximumSendingTime
    );
  }
  get peerAddress() {
    if (import_NodeSession.NodeSession.is(this.#session)) {
      return this.#session.peerAddress;
    }
  }
}
