// @matter/protocol/dist/cjs/protocol/MessageExchange.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var MessageExchange_exports = {};
__export(MessageExchange_exports, {
  MATTER_MESSAGE_OVERHEAD: () => MATTER_MESSAGE_OVERHEAD,
  MessageExchange: () => MessageExchange
});
module.exports = __toCommonJS(MessageExchange_exports);
var import_MessageCodec = require("#codec/MessageCodec.js");
var import_Mark = require("#common/Mark.js");
var import_PeerCommunicationError = require("#peer/PeerCommunicationError.js");
var import_GroupSession = require("#session/GroupSession.js");
var import_Session = require("#session/Session.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_errors = require("./errors.js");
var import_MRP = require("./MRP.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("MessageExchange");
const MATTER_MESSAGE_OVERHEAD = 26 + 12 + import_general.CRYPTO_AEAD_MIC_LENGTH_BYTES;
class MessageExchange {
  static fromInitialMessage(context, initialMessage, options) {
    const { session } = context;
    return new MessageExchange({
      context,
      isInitiator: false,
      peerSessionId: session.id,
      nodeId: initialMessage.packetHeader.destNodeId,
      peerNodeId: initialMessage.packetHeader.sourceNodeId,
      exchangeId: initialMessage.payloadHeader.exchangeId,
      protocolId: initialMessage.payloadHeader.protocolId,
      ...options
    });
  }
  static initiate(context, exchangeId, protocolId, options) {
    const { session } = context;
    return new MessageExchange({
      context,
      isInitiator: true,
      peerSessionId: session.peerSessionId,
      nodeId: session.nodeId,
      peerNodeId: session.peerNodeId,
      exchangeId,
      protocolId,
      ...options
    });
  }
  #context;
  #isInitiator;
  #messagesQueue = new import_general.DataReadQueue();
  #lifetime;
  #onSend;
  #onReceive;
  #addressOverride;
  #receivedMessageToAck;
  #receivedMessageAckTimer = import_general.Time.getTimer("ack receipt timeout", import_MRP.MRP.STANDALONE_ACK_TIMEOUT, () => {
    if (this.#receivedMessageToAck !== void 0) {
      const messageToAck = this.#receivedMessageToAck;
      this.#receivedMessageToAck = void 0;
      this.sendStandaloneAckForMessage(messageToAck).catch(
        (error) => logger.error("An error happened when sending a standalone ack", error)
      );
    }
  });
  #closeTimer;
  #closeCause;
  #isDestroyed = false;
  #timedInteractionTimer;
  #used;
  #peerSessionId;
  #nodeId;
  #peerNodeId;
  #exchangeId;
  #protocolId;
  #closed = (0, import_general.AsyncObservableValue)();
  #closing = (0, import_general.AsyncObservableValue)();
  #channel;
  // TODO - following are associated with current active transmission and should maybe go in a closure
  #sentMessageToAck;
  #sentMessageAckSuccess;
  #sentMessageAckFailure;
  #sendOptions = {};
  #retransmissionCounter = 0;
  #totalRetransmissionCounter = 0;
  // counter for all messages within this exchange
  #messageSendCounter = 0;
  #messageReceivedCounter = 0;
  #retransmissionTimer;
  constructor(config) {
    const {
      context,
      isInitiator,
      peerSessionId,
      nodeId,
      peerNodeId,
      exchangeId,
      protocolId,
      onSend,
      onReceive,
      network,
      addressOverride
    } = config;
    this.#context = context;
    this.#isInitiator = isInitiator;
    this.#peerSessionId = peerSessionId;
    this.#nodeId = nodeId;
    this.#peerNodeId = peerNodeId;
    this.#exchangeId = exchangeId;
    this.#protocolId = protocolId;
    this.#onSend = onSend;
    this.#onReceive = onReceive;
    this.#addressOverride = addressOverride;
    const { activeThreshold, activeInterval, idleInterval } = this.session.parameters;
    this.#used = !isInitiator;
    const { session } = context;
    logger.debug(
      "New exchange",
      isInitiator ? import_Mark.Mark.OUTBOUND : import_Mark.Mark.INBOUND,
      this.via,
      import_general.Diagnostic.dict({
        protocol: this.#protocolId,
        peerSess: import_Session.Session.idStrOf(this.#peerSessionId),
        address: addressOverride ? import_general.ServerAddress.urlFor(addressOverride) : void 0,
        SAT: activeThreshold !== void 0 ? import_general.Duration.format(activeThreshold) : void 0,
        SAI: activeInterval !== void 0 ? import_general.Duration.format(activeInterval) : void 0,
        SII: idleInterval !== void 0 ? import_general.Duration.format(idleInterval) : void 0,
        maxTrans: import_MRP.MRP.MAX_TRANSMISSIONS,
        exchangeFlags: import_general.Diagnostic.asFlags({
          MRP: this.session.usesMrp,
          I: this.isInitiator,
          ...network ? { [network.id]: true } : {}
        })
      })
    );
    session.addExchange(this);
    this.#lifetime = this.#context.session.join("exchange", import_general.Diagnostic.via(import_general.hex.word(this.id)));
  }
  get context() {
    return this.#context;
  }
  get isInitiator() {
    return this.#isInitiator;
  }
  /** Emits when the exchange is actually closed. This happens after all Retries and Communication are done. */
  get closed() {
    return this.#closed;
  }
  get considerClosed() {
    return this.#closed.value || this.#isInitiator && this.#closing.value;
  }
  /**
   * Emit when the exchange is closing, but not yet closed. We only wait for acks and retries to happen, but the
   * actual interaction logic is already done.
   */
  get closing() {
    return this.#closing;
  }
  get id() {
    return this.#exchangeId;
  }
  get idStr() {
    return import_general.hex.word(this.#exchangeId);
  }
  get session() {
    return this.context.session;
  }
  /** Number of retransmissions of the current outstanding message (resets on ack or new send). */
  get retransmissionCount() {
    return this.#retransmissionCounter;
  }
  get channel() {
    if (this.#channel === void 0) {
      this.#channel = this.session.channel;
    }
    return this.#channel;
  }
  /**
   * Max Payload size of the exchange which bases on the maximum payload size of the channel reduced by Matter
   * protocol overhead.
   */
  get maxPayloadSize() {
    return this.channel.maxPayloadSize - MATTER_MESSAGE_OVERHEAD;
  }
  join(...name) {
    return this.#lifetime.join(...name);
  }
  async onMessageReceived(message, duplicate = false) {
    logger.debug("Message", import_Mark.Mark.INBOUND, import_MessageCodec.Message.diagnosticsOf(this, message, { duplicate }));
    this.#messageReceivedCounter++;
    if (message.payloadHeader.requiresAck && !this.session.usesMrp) {
      logger.debug("Ignoring ack-required flag because MRP is not used for this exchange");
      message.payloadHeader.requiresAck = false;
    }
    const {
      packetHeader: { messageId },
      payloadHeader: { requiresAck, ackedMessageId, protocolId, messageType }
    } = message;
    const isStandaloneAck = import_types.SecureMessageType.isStandaloneAck(protocolId, messageType);
    if (protocolId !== this.#protocolId && !isStandaloneAck) {
      throw new import_general.MatterFlowError(
        `Drop received a message for an unexpected protocol. Expected: ${this.#protocolId}, received: ${protocolId}`
      );
    }
    this.session.notifyActivity(true);
    this.#onReceive?.(message, duplicate);
    if (duplicate) {
      if (requiresAck) {
        await this.sendStandaloneAckForMessage(message);
      }
      return;
    }
    if (messageId === this.#sentMessageToAck?.payloadHeader.ackedMessageId) {
      var _stack = [];
      try {
        const _acking = __using(_stack, this.join("resending ack"));
        this.#messageSendCounter++;
        await this.channel.send(this.#sentMessageToAck, { exchange: this, addressOverride: this.#addressOverride });
        return;
      } catch (_) {
        var _error = _, _hasError = true;
      } finally {
        __callDispose(_stack, _error, _hasError);
      }
    }
    const sentMessageIdToAck = this.#sentMessageToAck?.packetHeader.messageId;
    if (sentMessageIdToAck !== void 0) {
      if (ackedMessageId === void 0) {
        throw new import_general.MatterFlowError("Previous message ack is missing");
      } else if (ackedMessageId !== sentMessageIdToAck) {
        if (isStandaloneAck) {
        } else {
          throw new import_general.MatterFlowError(
            `Incorrect ack received. Expected ${import_general.hex.fixed(sentMessageIdToAck, 8)}, received: ${import_general.hex.fixed(ackedMessageId, 8)}`
          );
        }
      } else {
        this.#retransmissionTimer?.stop();
        this.#retransmissionCounter = 0;
        this.#sentMessageAckSuccess?.(message);
        this.#sentMessageAckSuccess = void 0;
        this.#sentMessageAckFailure = void 0;
        this.#sentMessageToAck = void 0;
        if (isStandaloneAck && this.#closeTimer !== void 0) {
          return this.#close();
        }
      }
    }
    if (isStandaloneAck) {
      return;
    }
    if (requiresAck) {
      if (this.#receivedMessageToAck !== void 0) {
        this.#receivedMessageAckTimer.stop();
        await this.sendStandaloneAckForMessage(this.#receivedMessageToAck);
      }
      this.#receivedMessageToAck = message;
      this.#receivedMessageAckTimer.start();
    }
    this.#messagesQueue.write(message);
  }
  async send(messageType, payload, options = {}) {
    if (this.#closeCause) {
      throw new import_general.ClosedError("Exchange is closed", { cause: this.#closeCause });
    }
    this.#sendOptions = options;
    this.#retransmissionCounter = 0;
    try {
      await this.#send(messageType, payload);
    } catch (e) {
      if (!options.suppressPeerLoss && this.#messageReceivedCounter === 0 && (0, import_general.causedBy)(e, import_PeerCommunicationError.TransientPeerCommunicationError, import_general.TimeoutError, import_general.NetworkError)) {
        await this.#context.peerLost(this, (0, import_general.asError)(e));
      }
      throw e;
    }
  }
  async #send(messageType, payload, standaloneAckMessageId) {
    var _stack2 = [];
    try {
      const {
        expectAckOnly = false,
        disableMrpLogic,
        expectedProcessingTime = import_MRP.MRP.DEFAULT_EXPECTED_PROCESSING_TIME,
        logContext = {}
      } = this.#sendOptions;
      const abort = __using(_stack2, new import_general.Abort(this.#sendOptions));
      const isStandaloneAck = standaloneAckMessageId !== void 0;
      if (standaloneAckMessageId !== void 0) {
        if (!this.session.usesMrp) {
          throw new import_general.InternalError("Cannot include an acknowledge message ID when MRP is not used");
        }
        if (messageType !== import_types.SecureMessageType.StandaloneAck) {
          throw new import_general.InternalError("Cannot override an acknowledge message ID for non-standalone acks");
        }
      }
      let { requiresAck } = this.#sendOptions;
      if (requiresAck && !(this.session.usesMrp || this.session.isPeerLost)) {
        requiresAck = false;
      }
      const protocolId = isStandaloneAck ? import_types.SECURE_CHANNEL_PROTOCOL_ID : this.#protocolId;
      if (isStandaloneAck) {
        if (!this.session.usesMrp) {
          return;
        }
        if (requiresAck) {
          throw new import_general.MatterFlowError("A standalone ack may not require acknowledgement");
        }
      }
      if (this.#sentMessageToAck !== void 0 && !isStandaloneAck) {
        throw new import_general.MatterFlowError(
          `The previous message ${import_MessageCodec.Message.via(this, this.#sentMessageToAck)} has not been acked yet, cannot send a new message`
        );
      }
      this.#used = true;
      this.#messageSendCounter++;
      this.session.notifyActivity(false);
      let ackedMessageId = standaloneAckMessageId;
      if (ackedMessageId === void 0 && this.session.usesMrp) {
        ackedMessageId = this.#receivedMessageToAck?.packetHeader.messageId;
        if (ackedMessageId !== void 0) {
          this.#receivedMessageAckTimer.stop();
          this.#receivedMessageToAck = void 0;
        }
      }
      let packetHeader;
      if (this.session.type === import_MessageCodec.SessionType.Unicast) {
        const messageId = isStandaloneAck ? await this.session.getIncrementedMessageCounter() : await abort.attempt(this.session.getIncrementedMessageCounter());
        if (messageId === void 0) {
          return;
        }
        packetHeader = {
          sessionId: this.#peerSessionId,
          sessionType: import_MessageCodec.SessionType.Unicast,
          messageId,
          destNodeId: this.#peerNodeId,
          sourceNodeId: this.#nodeId,
          hasPrivacyEnhancements: false,
          isControlMessage: false,
          hasMessageExtensions: false
        };
      } else if (this.session.type === import_MessageCodec.SessionType.Group) {
        const session = this.session;
        if (!import_GroupSession.GroupSession.is(session)) {
          throw new import_general.InternalError("Session is not a GroupSession, but session type is Group.");
        }
        const destGroupId = import_types.GroupId.fromNodeId(this.#peerNodeId);
        if (destGroupId === 0) {
          throw new import_general.InternalError(`Invalid GroupId extracted from NodeId ${this.#peerNodeId}`);
        }
        const messageId = await abort.attempt(this.session.getIncrementedMessageCounter());
        if (messageId === void 0) {
          return;
        }
        packetHeader = {
          sessionId: this.#peerSessionId,
          sessionType: import_MessageCodec.SessionType.Group,
          messageId,
          destGroupId,
          sourceNodeId: this.#nodeId,
          // We are the source node, so use our NodeId
          hasPrivacyEnhancements: false,
          isControlMessage: false,
          hasMessageExtensions: false
        };
      } else {
        throw new import_general.InternalError(`Unknown session type: ${this.session.type}`);
      }
      const message = {
        packetHeader,
        payloadHeader: {
          exchangeId: this.#exchangeId,
          protocolId,
          messageType,
          isInitiatorMessage: this.isInitiator,
          requiresAck: requiresAck ?? (this.session.usesMrp && !isStandaloneAck),
          ackedMessageId,
          hasSecuredExtension: false
        },
        payload
      };
      let ackPromise;
      if (this.session.usesMrp && message.payloadHeader.requiresAck && !disableMrpLogic) {
        this.#sentMessageToAck = message;
        const backOff = this.#mrpResubmissionBackOffTime;
        logContext.backOff = import_general.Duration.format(backOff);
        this.#retransmissionTimer = import_general.Time.getTimer(
          `retransmitting ${import_MessageCodec.Message.via(this, message)}`,
          backOff,
          () => this.#retransmitMessage(message, expectedProcessingTime)
        );
        const { promise, resolver } = (0, import_general.createPromise)();
        ackPromise = promise;
        this.#sentMessageAckSuccess = resolver;
        const startedWaitingAt = import_general.Time.nowMs;
        this.#sentMessageAckFailure = () => {
          abort.abort(new import_PeerCommunicationError.PeerUnresponsiveError(import_general.Timestamp.delta(startedWaitingAt)));
        };
      }
      this.#onSend?.(message, 0);
      const sending = __using(_stack2, this.join("sending", import_general.Diagnostic.strong(import_MessageCodec.Message.via(this, message))));
      if (isStandaloneAck) {
        await this.channel.send(message, { exchange: this, addressOverride: this.#addressOverride });
      } else {
        await abort.attempt(
          this.channel.send(message, { exchange: this, logContext, addressOverride: this.#addressOverride })
        );
      }
      if (abort.aborted) {
        return;
      }
      if (ackPromise !== void 0) {
        var _stack = [];
        try {
          this.#retransmissionCounter = 0;
          this.#retransmissionTimer?.start();
          const _waiting = __using(_stack, sending.join("waiting for ack"));
          const responseMessage = await abort.race(ackPromise);
          if (abort.aborted) {
            this.#retransmissionTimer?.stop();
            if (!(0, import_general.causedBy)(abort.reason, import_general.AbortedError)) {
              throw abort.reason;
            }
            return;
          }
          this.#sentMessageAckSuccess = void 0;
          this.#sentMessageAckFailure = void 0;
          if (responseMessage) {
            const {
              payloadHeader: { protocolId: protocolId2, messageType: messageType2 }
            } = responseMessage;
            if (expectAckOnly && !import_types.SecureMessageType.isStandaloneAck(protocolId2, messageType2)) {
              throw new import_errors.UnexpectedMessageError("Expected ack only", this.session, responseMessage);
            }
          }
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          __callDispose(_stack, _error, _hasError);
        }
      }
    } catch (_2) {
      var _error2 = _2, _hasError2 = true;
    } finally {
      __callDispose(_stack2, _error2, _hasError2);
    }
  }
  async nextMessage(options) {
    try {
      return await this.#nextMessage(options);
    } catch (e) {
      if (!options?.suppressPeerLoss && this.#messageReceivedCounter === 0 && (0, import_general.causedBy)(e, import_PeerCommunicationError.TransientPeerCommunicationError)) {
        await this.#context.peerLost(this, (0, import_general.asError)(e));
      }
      throw e;
    }
  }
  async #nextMessage(options) {
    var _stack = [];
    try {
      let timeout;
      if (options?.timeout !== void 0) {
        timeout = options.timeout;
      } else if (!this.session.isClosed) {
        timeout = this.channel.calculateMaximumPeerResponseTime(
          this.session.parameters,
          this.context.localSessionParameters,
          options?.expectedProcessingTime
        );
      }
      const localAbort = __using(_stack, new import_general.Abort({
        timeout,
        abort: options?.abort,
        timeoutHandler: () => {
          throw new import_PeerCommunicationError.PeerUnresponsiveError(timeout);
        }
      }));
      return await this.#messagesQueue.read(localAbort);
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async sendStandaloneAckForMessage(message) {
    const {
      packetHeader: { messageId },
      payloadHeader: { requiresAck }
    } = message;
    if (!requiresAck || !this.session.usesMrp) {
      return;
    }
    await this.#send(import_types.SecureMessageType.StandaloneAck, new Uint8Array(0), messageId);
  }
  #retransmitMessage(message, expectedProcessingTime) {
    this.#retransmissionCounter++;
    this.#totalRetransmissionCounter++;
    if (this.considerClosed || this.#retransmissionCounter >= (this.#sendOptions.maxRetransmissions ?? import_MRP.MRP.MAX_TRANSMISSIONS)) {
      if (expectedProcessingTime && !this.considerClosed) {
        const finalWaitTime = (0, import_general.Millis)(
          this.channel.calculateMaximumPeerResponseTime(
            this.session.parameters,
            this.context.localSessionParameters,
            expectedProcessingTime
          ) - (this.#retransmissionTimer?.interval ?? import_general.Instant)
        );
        if (finalWaitTime > 0) {
          this.#retransmissionCounter--;
          this.#totalRetransmissionCounter--;
          logger.debug(
            `Message ${import_MessageCodec.Message.via(this, message)}: Wait additional ${import_general.Duration.format(finalWaitTime)} for processing time and peer resubmissions after all our resubmissions`
          );
          this.#retransmissionTimer = import_general.Time.getTimer(
            `waiting after resubmissions for ${import_MessageCodec.Message.via(this, message)}`,
            finalWaitTime,
            () => this.#retransmitMessage(message)
          ).start();
          return;
        }
      }
      if (this.#sentMessageToAck !== void 0 && this.#sentMessageAckFailure !== void 0) {
        this.#receivedMessageToAck = void 0;
        this.#sentMessageAckFailure();
        this.#sentMessageAckFailure = void 0;
        this.#sentMessageAckSuccess = void 0;
      }
      if (this.#closeTimer !== void 0) {
        this.#close().catch((error) => logger.error("Error closing exchange", error));
      }
      return;
    }
    this.#messageSendCounter++;
    this.session.notifyActivity(false);
    this.context.retry(this.#retransmissionCounter);
    const resubmissionBackoffTime = this.#mrpResubmissionBackOffTime;
    this.#onSend?.(message, this.#retransmissionCounter);
    this.channel.send(message, {
      exchange: this,
      logContext: {
        "retrans#": this.#retransmissionCounter,
        backoff: import_general.Duration.format(resubmissionBackoffTime)
      },
      addressOverride: this.#addressOverride
    }).then(() => this.#initializeResubmission(message, resubmissionBackoffTime, expectedProcessingTime)).catch((error) => {
      logger.error(`Error retransmitting ${import_MessageCodec.Message.via(this, message)}:`, error);
      if (error instanceof import_errors.SessionClosedError) {
        this.#close().catch((error2) => logger.error("An error happened when closing the exchange", error2));
      } else {
        this.#initializeResubmission(message, resubmissionBackoffTime, expectedProcessingTime);
      }
    });
  }
  #initializeResubmission(message, resubmissionBackoffTime, expectedProcessingTimeMs) {
    this.#retransmissionTimer = import_general.Time.getTimer(
      "Message retransmission",
      resubmissionBackoffTime,
      () => this.#retransmitMessage(message, expectedProcessingTimeMs)
    ).start();
  }
  [Symbol.asyncDispose]() {
    return this.destroy();
  }
  async destroy() {
    if (this.#isDestroyed) {
      return;
    }
    this.#isDestroyed = true;
    if (this.#closeTimer === void 0 && this.#receivedMessageToAck !== void 0) {
      this.#receivedMessageAckTimer.stop();
      const messageToAck = this.#receivedMessageToAck;
      this.#receivedMessageToAck = void 0;
      try {
        await this.sendStandaloneAckForMessage(messageToAck);
      } catch (error) {
        logger.error("An error happened when closing the exchange", error);
      }
    }
    await this.#close();
  }
  startTimedInteraction(timeout) {
    if (this.#timedInteractionTimer !== void 0 && this.#timedInteractionTimer.isRunning) {
      this.#timedInteractionTimer.stop();
      throw new import_types.StatusResponseError(
        "Timed interaction already running for this exchange",
        import_types.StatusCode.InvalidAction
      );
    }
    logger.debug(
      "Starting timed interaction",
      import_Mark.Mark.INBOUND,
      this.channel.name,
      import_general.Diagnostic.dict({ exId: this.#exchangeId, timeout: import_general.Duration.format(timeout) })
    );
    this.#timedInteractionTimer = import_general.Time.getTimer("Timed interaction", timeout, () => {
      logger.debug(
        "Timed interaction timeout!",
        import_general.Diagnostic.dict({ exId: this.#exchangeId, via: this.channel.name })
      );
    }).start();
  }
  clearTimedInteraction() {
    if (this.#timedInteractionTimer !== void 0) {
      logger.debug(
        "Clearing timed interaction",
        import_general.Diagnostic.dict({ exId: this.#exchangeId, via: this.channel.name })
      );
      this.#timedInteractionTimer.stop();
      this.#timedInteractionTimer = void 0;
    }
  }
  hasTimedInteraction() {
    return this.#timedInteractionTimer !== void 0;
  }
  hasActiveTimedInteraction() {
    return this.#timedInteractionTimer !== void 0 && this.#timedInteractionTimer.isRunning;
  }
  hasExpiredTimedInteraction() {
    return this.#timedInteractionTimer !== void 0 && !this.#timedInteractionTimer.isRunning;
  }
  /**
   * Closes the exchange.
   *
   * If cause is defined, the exchange is closed immediately, even if there are still messages to send.  Further
   * attempts at communicating via the exchange will throw this error.
   *
   * If cause is undefined, the exchange is closed only after all messages have been sent.
   */
  async close(cause) {
    var _stack3 = [];
    try {
      if (this.#isDestroyed) {
        return;
      }
      const closing = __using(_stack3, this.#lifetime.closing());
      if (this.#closeTimer !== void 0) {
        if (cause) {
          this.#closeTimer.stop();
          return this.#close(cause);
        }
        return;
      }
      if (!this.#used) {
        logger.info(this.via, `Exchange never used, closing directly`);
        return this.#close(cause);
      }
      {
        var _stack = [];
        try {
          const _emitting = __using(_stack, closing.join("emitting"));
          await this.#closing.emit(true);
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          __callDispose(_stack, _error, _hasError);
        }
      }
      if (this.#closed.value) {
        return;
      }
      if (this.#receivedMessageToAck !== void 0) {
        this.#receivedMessageAckTimer.stop();
        const messageToAck = this.#receivedMessageToAck;
        this.#receivedMessageToAck = void 0;
        try {
          var _stack2 = [];
          try {
            const _acking = __using(_stack2, closing.join("acking"));
            await this.sendStandaloneAckForMessage(messageToAck);
          } catch (_2) {
            var _error2 = _2, _hasError2 = true;
          } finally {
            __callDispose(_stack2, _error2, _hasError2);
          }
        } catch (error) {
          logger.error(this.via, `Unhandled error closing exchange`, error);
        }
        if (cause || this.#sentMessageToAck === void 0) {
          await this.#close(cause);
          return;
        }
      } else if (this.#sentMessageToAck === void 0 || cause) {
        await this.#close(cause);
        return;
      }
      let maxResubmissionTime = import_general.Instant;
      for (let i = this.#retransmissionCounter; i <= import_MRP.MRP.MAX_TRANSMISSIONS; i++) {
        maxResubmissionTime = (0, import_general.Millis)(
          maxResubmissionTime + this.channel.getMrpResubmissionBackOffTime(i, void 0, true)
        );
      }
      this.#closeTimer = import_general.Time.getTimer(
        `Exchange ${this.via} close`,
        maxResubmissionTime,
        async () => await this.#close(cause)
      ).start();
    } catch (_3) {
      var _error3 = _3, _hasError3 = true;
    } finally {
      __callDispose(_stack3, _error3, _hasError3);
    }
  }
  async #close(cause) {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      if (this.#closeCause === void 0) {
        this.#closeCause = cause;
      }
      try {
        this.#retransmissionTimer?.stop();
        this.#receivedMessageAckTimer.stop();
        this.#sentMessageAckSuccess?.(void 0);
        this.#closeTimer?.stop();
        this.#timedInteractionTimer?.stop();
        this.#messagesQueue.close(this.#closeCause);
      } finally {
        await this.#closed.emit(true);
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  get via() {
    if (this.session === void 0) {
      return import_general.Diagnostic.via(`${import_Mark.Mark.EXCHANGE}${this.idStr}`);
    }
    return import_general.Diagnostic.via(`${this.session.via}${import_Mark.Mark.EXCHANGE}${this.idStr}`);
  }
  /**
   * Expose some diagnostics for logging.
   * For simply one message exchanges without retransmissions nothing will be exposed
   */
  get diagnostics() {
    if (this.#totalRetransmissionCounter === 0 && this.#messageSendCounter === 1 && this.#messageReceivedCounter === 1) {
      return "";
    }
    return `${this.#messageReceivedCounter}${import_Mark.Mark.TRANSFERRED}${this.#messageSendCounter}${this.#totalRetransmissionCounter > 0 ? `+${this.#totalRetransmissionCounter}` : ""}`;
  }
  get #mrpResubmissionBackOffTime() {
    let backOff = this.channel.getMrpResubmissionBackOffTime(this.#retransmissionCounter);
    if (this.#sendOptions.initialRetransmissionTime !== void 0) {
      backOff = (0, import_general.Millis)(backOff + this.#sendOptions.initialRetransmissionTime);
    }
    return import_general.Duration.min(backOff, this.#sendOptions.maxRetransmissionTime ?? import_general.Forever);
  }
}
