// @matter/protocol/dist/cjs/protocol/ProtocolMocks.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ProtocolMocks_exports = {};
__export(ProtocolMocks_exports, {
  ProtocolMocks: () => ProtocolMocks
});
module.exports = __toCommonJS(ProtocolMocks_exports);
var import_Fabric = require("#fabric/Fabric.js");
var import_InteractionMessenger = require("#interaction/InteractionMessenger.js");
var import_NodeSession = require("#session/NodeSession.js");
var import_SessionParameters = require("#session/SessionParameters.js");
var import_general = require("@matter/general");
var import_model = require("@matter/model");
var import_types = require("@matter/types");
var import_MessageChannel = require("./MessageChannel.js");
var import_MessageExchange = require("./MessageExchange.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ProtocolMocks;
((ProtocolMocks2) => {
  class Fabric extends import_Fabric.Fabric {
    constructor(config, crypto) {
      if (!crypto) {
        crypto = import_general.Environment.default.maybeGet(import_general.Crypto);
        if (!(crypto instanceof import_general.MockCrypto)) {
          crypto = (0, import_general.MockCrypto)();
        }
      }
      const keyPair = config?.keyPair ?? crypto.createKeyPair();
      if (import_general.MaybePromise.is(keyPair)) {
        throw new import_general.ImplementationError("Must provide key pair with async crypto");
      }
      super(crypto, {
        ...Fabric.defaults,
        ...config,
        keyPair
      });
    }
  }
  ProtocolMocks2.Fabric = Fabric;
  ((Fabric2) => {
    Fabric2.defaults = {
      fabricId: (0, import_types.FabricId)(0x2906c908d115d362n),
      fabricIndex: (0, import_types.FabricIndex)(1),
      identityProtectionKey: import_general.Bytes.empty,
      intermediateCACert: import_general.Bytes.empty,
      keyPair: { publicKey: import_general.Bytes.empty, privateKey: import_general.Bytes.empty },
      label: "test-fabric",
      nodeId: (0, import_types.NodeId)(0xcd5544aa7b13ef14n),
      operationalCert: import_general.Bytes.empty,
      globalId: (0, import_types.GlobalFabricId)(0),
      operationalIdentityProtectionKey: import_general.b$`9bc61cd9c62a2df6d64dfcaa9dc472d4`,
      rootCert: import_general.Bytes.empty,
      rootNodeId: (0, import_types.NodeId)(1),
      rootPublicKey: import_general.b$`044a9f42b1ca4840d37292bbc7f6a7e11e22200c976fc900dbc98a7a383a641cb8254a2e56d4e295a847943b4e3897c4a773e930277b4d9fbede8a052686bfacfa`,
      rootVendorId: (0, import_types.VendorId)(1)
    };
  })(Fabric = ProtocolMocks2.Fabric || (ProtocolMocks2.Fabric = {}));
  class NodeSession extends import_NodeSession.NodeSession {
    constructor(config) {
      const index = config?.index ?? 1;
      const fabricIndex = config?.fabricIndex ?? 1;
      const crypto = config?.crypto ?? import_general.Environment.default.get(import_general.Crypto);
      const fabric = config && "fabric" in config ? config.fabric : new Fabric({ fabricIndex });
      const maxPayloadSize = config?.maxPayloadSize;
      let channel;
      if (config && "channel" in config) {
        channel = config.channel;
      } else {
        channel = new NetworkChannel({ index, maxPayloadSize });
      }
      const fullConfig = {
        id: index,
        peerNodeId: (0, import_types.NodeId)(index),
        attestationKey: import_general.Bytes.empty,
        caseAuthenticatedTags: [],
        peerSessionId: index,
        decryptKey: import_general.Bytes.empty,
        encryptKey: import_general.Bytes.empty,
        isInitiator: true,
        ...config,
        crypto,
        fabric
      };
      delete fullConfig.channel;
      super(fullConfig);
      this.channel = new MessageChannel({ channel, session: this });
    }
    static async create(config) {
      const crypto = config?.crypto ?? config?.manager?.crypto ?? import_general.Environment.default.get(import_general.Crypto);
      const index = config?.index ?? 1;
      return import_NodeSession.NodeSession.create.call(this, {
        id: index,
        crypto,
        peerNodeId: (0, import_types.NodeId)(0),
        peerSessionId: index,
        sharedSecret: import_general.Bytes.empty,
        salt: import_general.Bytes.empty,
        isInitiator: true,
        isResumption: false,
        ...config
      });
    }
  }
  ProtocolMocks2.NodeSession = NodeSession;
  class NetworkChannel {
    maxPayloadSize;
    isReliable = true;
    supportsLargeMessages = false;
    name = "mock-byte-channel";
    type = import_general.ChannelType.UDP;
    #networkAddress;
    networkAddressChanged = new import_general.Observable();
    constructor(config) {
      const index = config.index ?? 1;
      this.maxPayloadSize = config.maxPayloadSize ?? import_general.MAX_UDP_MESSAGE_SIZE;
      this.#networkAddress = { type: "udp", ip: `::${index}`, port: 5540 };
    }
    get networkAddress() {
      return this.#networkAddress;
    }
    set networkAddress(networkAddress) {
      if (import_general.ServerAddress.isEqual(networkAddress, this.#networkAddress)) {
        return;
      }
      this.#networkAddress = networkAddress;
      this.networkAddressChanged.emit(this.networkAddress);
    }
    async send() {
    }
    async close() {
    }
  }
  ProtocolMocks2.NetworkChannel = NetworkChannel;
  class MessageChannel extends import_MessageChannel.MessageChannel {
    #requests = new import_general.DataReadQueue();
    #readReady;
    constructor(config) {
      const index = config?.index ?? 1;
      const fabricIndex = config?.fabricIndex ?? 1;
      const channel = config?.channel ?? new NetworkChannel({ index });
      const session = config?.session ?? new NodeSession({ index, fabricIndex });
      super(channel, session, config?.onClose);
    }
    async send(message) {
      this.#requests.write(message);
      await this.#readReady?.emit();
    }
    read() {
      return this.#requests.read();
    }
    get readReady() {
      return this.#readReady ??= new import_general.Observable();
    }
  }
  ProtocolMocks2.MessageChannel = MessageChannel;
  class Exchange extends import_MessageExchange.MessageExchange {
    #responses = new import_general.DataReadQueue();
    constructor(config) {
      const context = config?.context;
      const index = config?.index ?? 1;
      const fabricIndex = config?.fabricIndex ?? 1;
      const maxPayloadSize = config?.maxPayloadSize;
      super({
        isInitiator: true,
        exchangeId: index,
        peerSessionId: index,
        protocolId: config?.protocolId ?? import_types.SECURE_CHANNEL_PROTOCOL_ID,
        ...config,
        context: {
          session: context?.session ?? new NodeSession({ index, fabricIndex, maxPayloadSize }),
          localSessionParameters: (0, import_SessionParameters.SessionParameters)(
            context?.localSessionParameters ?? import_SessionParameters.SessionParameters.defaults
          ),
          async peerLost() {
          },
          retry() {
          }
        }
      });
    }
    /**
     * Enqueue a mock response.
     */
    async write(message) {
      this.#responses.write(message);
    }
    /**
     * Enqueue a mock status response.
     */
    async writeStatus(status = import_types.Status.Success) {
      await this.write({
        packetHeader: {
          messageId: 123
        },
        payloadHeader: {
          messageType: import_InteractionMessenger.MessageType.StatusResponse
        },
        payload: import_types.TlvStatusResponse.encode({
          status,
          interactionModelRevision: import_model.Specification.INTERACTION_MODEL_REVISION
        })
      });
    }
    /**
     * Wait for a mock request.
     */
    read() {
      return this.#mockChannel.read();
    }
    /**
     * Emits when read() won't block.
     */
    get readReady() {
      return this.#mockChannel.readReady;
    }
    nextMessage() {
      return this.#responses.read();
    }
    get #mockChannel() {
      const channel = this.channel;
      if (!(channel instanceof MessageChannel)) {
        throw new import_general.ImplementationError("Message channel is not mocked");
      }
      return channel;
    }
  }
  ProtocolMocks2.Exchange = Exchange;
})(ProtocolMocks || (ProtocolMocks = {}));
