// @matter/general/dist/cjs/net/dns-sd/DnssdSolicitor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var DnssdSolicitor_exports = {};
__export(DnssdSolicitor_exports, {
  DnssdSolicitor: () => DnssdSolicitor,
  QueryMulticaster: () => QueryMulticaster
});
module.exports = __toCommonJS(DnssdSolicitor_exports);
var import_DnsCodec = require("#codec/DnsCodec.js");
var import_Logger = require("#log/Logger.js");
var import_RetrySchedule = require("#net/RetrySchedule.js");
var import_Time = require("#time/Time.js");
var import_TimeUnit = require("#time/TimeUnit.js");
var import_Abort = require("#util/Abort.js");
var import_Multiplex = require("#util/Multiplex.js");
var import_Observable = require("#util/Observable.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = new import_Logger.Logger("DiscoverySolicitor");
var DnssdSolicitor;
((DnssdSolicitor2) => {
  DnssdSolicitor2.DefaultRetries = {
    initialInterval: (0, import_TimeUnit.Seconds)(1),
    jitterFactor: 0.2,
    backoffFactor: 2,
    maximumInterval: (0, import_TimeUnit.Hours)(1)
  };
})(DnssdSolicitor || (DnssdSolicitor = {}));
class QueryMulticaster {
  #names;
  #schedule;
  #abort = new import_Abort.Abort();
  #toSolicit = /* @__PURE__ */ new Map();
  #discovering = /* @__PURE__ */ new Map();
  #namesReady = new import_Observable.ObservableValue();
  #workers = new import_Multiplex.BasicMultiplex();
  constructor(names, retries) {
    this.#names = names;
    this.#schedule = new import_RetrySchedule.RetrySchedule(
      names.entropy,
      import_RetrySchedule.RetrySchedule.Configuration(DnssdSolicitor.DefaultRetries, retries)
    );
    this.#workers.add(this.#emitSolicitations());
  }
  solicit(solicitation) {
    if (this.#abort.aborted) {
      return;
    }
    let entry = this.#toSolicit.get(solicitation.name);
    if (entry === void 0) {
      entry = {
        name: solicitation.name,
        recordTypes: new Set(solicitation.recordTypes),
        associatedNames: /* @__PURE__ */ new Set()
      };
      this.#toSolicit.set(solicitation.name, entry);
    } else {
      for (const type of solicitation.recordTypes) {
        entry.recordTypes.add(type);
      }
    }
    if (solicitation.associatedNames) {
      entry.associatedNames.add(solicitation.associatedNames);
    }
    this.#namesReady.emit(true);
  }
  async discover(discovery) {
    let active = this.#discovering.get(discovery.name);
    if (active) {
      active.waiting.add(discovery);
    } else {
      const abort = new import_Abort.Abort({ abort: this.#abort });
      active = {
        abort,
        finished: this.#discover(discovery, abort),
        waiting: /* @__PURE__ */ new Set([discovery])
      };
      this.#discovering.set(discovery.name, active);
    }
    try {
      await import_Abort.Abort.race(discovery.abort, active.finished);
    } finally {
      active.waiting.delete(discovery);
      if (active.waiting.size === 0) {
        active.abort();
        this.#discovering.delete(discovery.name);
      }
    }
  }
  async #discover(solicitation, abort) {
    const schedule = solicitation.retries ? new import_RetrySchedule.RetrySchedule(
      this.#names.entropy,
      import_RetrySchedule.RetrySchedule.Configuration(DnssdSolicitor.DefaultRetries, solicitation.retries)
    ) : this.#schedule;
    this.solicit(solicitation);
    for (const nextTimeout of schedule) {
      var _stack = [];
      try {
        const delay = __using(_stack, new import_Abort.Abort({ abort, timeout: nextTimeout }));
        await delay;
        if (abort.aborted) {
          break;
        }
        this.solicit(solicitation);
      } catch (_) {
        var _error = _, _hasError = true;
      } finally {
        __callDispose(_stack, _error, _hasError);
      }
    }
  }
  async close() {
    this.#abort();
    await this.#workers;
  }
  async #emitSolicitations() {
    while (true) {
      await this.#abort.race(this.#namesReady);
      if (this.#abort.aborted) {
        return;
      }
      await this.#abort.race(import_Time.Time.sleep("discovery solicitor delay", 0));
      if (this.#abort.aborted) {
        return;
      }
      const entries = [...this.#toSolicit.values()];
      this.#namesReady.value = false;
      this.#toSolicit.clear();
      const queries = Array();
      const answers = Array();
      for (const { name, recordTypes, associatedNames } of entries) {
        for (const recordType of recordTypes) {
          queries.push({ name: name.qname, recordClass: import_DnsCodec.DnsRecordClass.IN, recordType });
        }
        answers.push(...name.records);
        for (const iterable of associatedNames) {
          for (const assocName of iterable) {
            answers.push(...assocName.records);
          }
        }
      }
      if (queries.length === 0) {
        continue;
      }
      try {
        await this.#abort.race(
          this.#names.socket.send({
            messageType: import_DnsCodec.DnsMessageType.Query,
            queries,
            answers
          })
        );
      } catch (e) {
        logger.error("Unhandled error soliciting DNS-SD names:", e);
      }
    }
  }
}
