// @matter/general/dist/cjs/net/dns-sd/MdnsSocket.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MdnsSocket_exports = {};
__export(MdnsSocket_exports, {
  MdnsSocket: () => MdnsSocket
});
module.exports = __toCommonJS(MdnsSocket_exports);
var import_DnsCodec = require("#codec/DnsCodec.js");
var import_Logger = require("#log/Logger.js");
var import_MatterError = require("#MatterError.js");
var import_UdpMulticastServer = require("#net/udp/UdpMulticastServer.js");
var import_Observable = require("#util/Observable.js");
var import_Promises = require("#util/Promises.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("MdnsListener");
class MdnsSocket {
  #socket;
  #handlers;
  #isClosed = false;
  #receipt = new import_Observable.BasicObservable(
    (error) => logger.error("Unhandled error in MDNS listener", error),
    true
  );
  static async create(network, options) {
    const { enableIpv4 = true, netInterface, lifetime } = options ?? {};
    const socket = new MdnsSocket(
      await import_UdpMulticastServer.UdpMulticastServer.create({
        lifetime,
        network,
        netInterface,
        broadcastAddressIpv4: enableIpv4 ? MdnsSocket.BROADCAST_IPV4 : void 0,
        broadcastAddressIpv6: MdnsSocket.BROADCAST_IPV6,
        listeningPort: MdnsSocket.BROADCAST_PORT
      })
    );
    return socket;
  }
  constructor(socket) {
    this.#socket = socket;
    socket.onMessage(this.#handleMessage.bind(this));
  }
  get network() {
    return this.#socket.network;
  }
  get supportsIpv4() {
    return this.#socket.supportsIpv4;
  }
  get netInterface() {
    return this.#socket.netInterface;
  }
  get receipt() {
    return this.#receipt;
  }
  async send(message, intf, unicastDest) {
    const { messageType, queries = [] } = message;
    const queryChunks = this.#splitQueries(queries);
    if (queryChunks.length > 1) {
      for (const queryChunk of queryChunks) {
        const chunkMessage = {
          transactionId: 0,
          messageType,
          queries: queryChunk,
          answers: [],
          authorities: message.authorities ?? [],
          additionalRecords: []
        };
        await this.#send(chunkMessage, intf, unicastDest);
      }
      return;
    }
    const truncatedMessageType = import_DnsCodec.DnsMessageType.isQuery(messageType) ? messageType | import_DnsCodec.DnsMessageTypeFlag.TC : messageType;
    const chunk = {
      transactionId: 0,
      queries: [],
      authorities: [],
      ...message,
      answers: [],
      additionalRecords: []
    };
    let encodedChunkWithoutAnswers = import_DnsCodec.DnsCodec.encode(chunk);
    let chunkSize = encodedChunkWithoutAnswers.byteLength;
    for (const answer of message.answers ?? []) {
      const answerEncoded = import_DnsCodec.DnsCodec.encodeRecord(answer);
      if (chunkSize + answerEncoded.byteLength > import_DnsCodec.MAX_MDNS_MESSAGE_SIZE) {
        await this.#send(
          {
            ...chunk,
            messageType: truncatedMessageType
          },
          intf,
          unicastDest
        );
        if (chunk.queries.length) {
          chunk.queries.length = 0;
          encodedChunkWithoutAnswers = import_DnsCodec.DnsCodec.encode(chunk);
        }
        chunk.answers.length = 0;
        chunkSize = encodedChunkWithoutAnswers.byteLength + answerEncoded.byteLength;
      } else {
        chunkSize += answerEncoded.byteLength;
      }
      chunk.answers.push(answerEncoded);
    }
    const additionalRecords = message.additionalRecords ?? [];
    for (const additionalRecord of additionalRecords) {
      const additionalRecordEncoded = import_DnsCodec.DnsCodec.encodeRecord(additionalRecord);
      chunkSize += additionalRecordEncoded.byteLength;
      if (chunkSize > import_DnsCodec.MAX_MDNS_MESSAGE_SIZE) {
        break;
      }
      chunk.additionalRecords.push(additionalRecordEncoded);
    }
    await this.#send(chunk, intf, unicastDest);
  }
  /**
   * Split queries into chunks that fit within MAX_MDNS_MESSAGE_SIZE.
   * Returns an array of query arrays - if all queries fit in one message, returns [[...queries]].
   */
  #splitQueries(queries) {
    if (queries.length === 0) {
      return [[]];
    }
    const DNS_HEADER_SIZE = 12;
    const chunks = [];
    let currentChunk = [];
    let currentChunkSize = DNS_HEADER_SIZE;
    for (const query of queries) {
      const querySize = import_DnsCodec.DnsCodec.encodeQuery(query).byteLength;
      if (currentChunkSize + querySize > import_DnsCodec.MAX_MDNS_MESSAGE_SIZE && currentChunk.length > 0) {
        chunks.push(currentChunk);
        currentChunk = [];
        currentChunkSize = DNS_HEADER_SIZE;
      }
      currentChunk.push(query);
      currentChunkSize += querySize;
    }
    if (currentChunk.length > 0) {
      chunks.push(currentChunk);
    }
    return chunks;
  }
  async #send(message, intf, unicastDest) {
    await this.#socket.send(import_DnsCodec.DnsCodec.encode(message), intf, unicastDest);
  }
  async close() {
    this.#isClosed = true;
    await this.#socket.close();
    if (this.#handlers) {
      await import_MatterError.MatterAggregateError.allSettled(this.#handlers);
    }
  }
  #handleMessage(bytes, sourceIp, sourceIntf) {
    if (this.#isClosed) {
      return;
    }
    const parsed = import_DnsCodec.DnsCodec.decode(bytes);
    if (parsed === void 0) {
      return;
    }
    let promise = this.#receipt.emit({
      ...parsed,
      sourceIp,
      sourceIntf
    });
    if (import_Promises.MaybePromise.is(promise)) {
      if (this.#handlers === void 0) {
        this.#handlers = /* @__PURE__ */ new Set();
      }
      promise = Promise.resolve(promise).finally(() => this.#handlers?.delete(promise));
    }
  }
}
((MdnsSocket2) => {
  MdnsSocket2.BROADCAST_IPV4 = "224.0.0.251";
  MdnsSocket2.BROADCAST_IPV6 = "ff02::fb";
  MdnsSocket2.BROADCAST_PORT = 5353;
})(MdnsSocket || (MdnsSocket = {}));
