// @matter/protocol/dist/esm/bdx/flow/OutboundFlow.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { BytesStreamReader, InternalError } from "@matter/general";
import { BdxStatusCode } from "@matter/types";
import { BdxError } from "../BdxError.js";
import { Flow } from "./Flow.js";
class OutboundFlow extends Flow {
  #closeStreams;
  #iterator;
  #streamReader;
  get stream() {
    if (this.#iterator === void 0 || this.#streamReader === void 0) {
      throw new InternalError("Read stream not initialized");
    }
    return { iterator: this.#iterator, streamReader: this.#streamReader };
  }
  /**
   * Returns initialized streams for the transfer and initializes the #closeStream class function
   */
  async initTransfer() {
    const { blockSize } = this.transferParameters;
    const blob = await this.transferParameters.fileDesignator.openBlob();
    const blobSize = blob.size;
    const { startOffset, dataLength = blobSize } = this.transferParameters;
    this.dataLength = dataLength;
    const dataBlob = startOffset > 0 || dataLength !== blobSize ? blob.slice(startOffset, startOffset + dataLength) : blob;
    const stream = dataBlob.stream();
    const reader = stream.getReader();
    this.#closeStreams = async (_error) => {
      if (stream.locked) {
        reader?.releaseLock();
        try {
          await reader?.cancel();
        } catch (error) {
          if (!(error instanceof TypeError)) {
            throw error;
          }
        }
      }
      await stream.cancel();
    };
    const streamReader = new BytesStreamReader(reader);
    this.#iterator = streamReader.read(blockSize);
    this.#streamReader = streamReader;
  }
  /**
   * Reads one data chunk from the reader and does some basic checks.
   */
  async readDataChunk(reader) {
    const { blockSize, dataLength } = this.transferParameters;
    let { value, done = false } = await reader.next();
    if (value === void 0) {
      if (!done) {
        throw new BdxError(
          `Data length too short, expected ${blockSize}bytes, but got less`,
          BdxStatusCode.LengthTooShort
        );
      }
      value = new Uint8Array();
    } else if (value.byteLength < blockSize) {
      ({ done = false } = await reader.next());
      if (!done) {
        throw new BdxError(
          `Data length too short, expected ${blockSize}bytes, but got less`,
          BdxStatusCode.LengthTooShort
        );
      }
    }
    if (this.bytesLeft !== void 0 && dataLength !== void 0) {
      this.bytesLeft -= value.byteLength;
      if (this.bytesLeft < 0) {
        throw new BdxError(
          `Data length exceeded, expected ${dataLength}bytes, but got ${-this.bytesLeft}bytes more`,
          BdxStatusCode.LengthTooLarge
        );
      }
      if (done && this.bytesLeft > 0) {
        throw new BdxError(
          `Data length too short, expected ${dataLength}bytes, but got ${this.bytesLeft}bytes less`,
          BdxStatusCode.LengthTooShort
        );
      }
      if (this.bytesLeft === 0) {
        done = true;
      }
    }
    return { data: value, done };
  }
  async finalizeTransfer() {
    const blockCounter = this.finalBlockCounter;
    const { blockCounter: ackedBlockCounter } = await this.messenger.readBlockAckEof();
    this.validateCounter(ackedBlockCounter, blockCounter);
  }
  async close(error) {
    await super.close(error);
    await this.#closeStreams?.(error);
  }
}
export {
  OutboundFlow
};
