// @matter/general/dist/esm/log/DiagnosticPresentation.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DiagnosticPresentation = /* @__PURE__ */ ((DiagnosticPresentation2) => {
  DiagnosticPresentation2["Message"] = "message";
  DiagnosticPresentation2["List"] = "list";
  DiagnosticPresentation2["Squash"] = "squash";
  DiagnosticPresentation2["Strong"] = "strong";
  DiagnosticPresentation2["Weak"] = "weak";
  DiagnosticPresentation2["Flag"] = "flag";
  DiagnosticPresentation2["Error"] = "error";
  DiagnosticPresentation2["Dictionary"] = "dictionary";
  DiagnosticPresentation2["Via"] = "via";
  DiagnosticPresentation2["Added"] = "added";
  DiagnosticPresentation2["Deleted"] = "deleted";
  return DiagnosticPresentation2;
})(DiagnosticPresentation || {});
((DiagnosticPresentation2) => {
  DiagnosticPresentation2.presentation = /* @__PURE__ */ Symbol.for("matter.diagnostic.presentation");
  DiagnosticPresentation2.value = /* @__PURE__ */ Symbol.for("matter.diagnostic.value");
  DiagnosticPresentation2.name = /* @__PURE__ */ Symbol.for("matter.diagnostic.name");
})(DiagnosticPresentation || (DiagnosticPresentation = {}));
export {
  DiagnosticPresentation
};
