// @matter/nodejs/dist/cjs/storage/fs/DirectoryBlobStorageDriver.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DirectoryBlobStorageDriver_exports = {};
__export(DirectoryBlobStorageDriver_exports, {
  DirectoryBlobStorageDriver: () => DirectoryBlobStorageDriver
});
module.exports = __toCommonJS(DirectoryBlobStorageDriver_exports);
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class DirectoryBlobStorageDriver extends import_general.FilesystemBlobStorageDriver {
  static id = "dir";
  static create(namespace, _descriptor) {
    return new DirectoryBlobStorageDriver(namespace);
  }
  #rootDir;
  #initialized = false;
  constructor(namespace) {
    super(namespace);
    this.#rootDir = this.root.directory;
  }
  get initialized() {
    return this.#initialized;
  }
  async initialize() {
    await super.initialize();
    await this.#rootDir.mkdir();
    this.#initialized = true;
  }
  async close() {
    this.#initialized = false;
    await super.close();
  }
  async openBlob(contexts, key) {
    const file = this.#resolve(contexts, key);
    if (!await file.exists()) {
      return new Blob([]);
    }
    const bytes = await file.readAllBytes();
    return new Blob([import_general.Bytes.exclusive(bytes)]);
  }
  // TODO: writes go directly to the final file via the File abstraction's streaming support.
  // Not atomic — a crash mid-write leaves a partial file.  Hardening with write-to-tmp + rename
  // would require bypassing File.write() to get a temp path in the same directory.
  async writeBlobFromStream(contexts, key, stream) {
    validateKey(key);
    let dir = this.#rootDir;
    for (const ctx of contexts) {
      if (ctx === ".." || ctx === "." || !ctx.length || ctx.includes("\\")) {
        throw new import_general.StorageError(`Context segment "${ctx}" is not allowed`);
      }
      dir = dir.directory(encodeContextSegment(ctx));
      await dir.mkdir();
    }
    const file = dir.file(key);
    await file.write(streamToAsyncIterable(stream));
  }
  async delete(contexts, key) {
    const file = this.#resolve(contexts, key);
    if (await file.exists()) {
      await file.delete();
    }
  }
  async has(contexts, key) {
    return this.#resolve(contexts, key).exists();
  }
  async keys(contexts) {
    const dir = this.#contextDir(contexts);
    if (!await dir.exists()) {
      return [];
    }
    const files = await dir.files();
    return files.filter((f) => !import_general.BaseStorageDriver.RESERVED_FILENAMES.has(f) && !f.endsWith(".tmp"));
  }
  async contexts(contexts) {
    const dir = this.#contextDir(contexts);
    if (!await dir.exists()) {
      return [];
    }
    const encoded = await dir.directories();
    return encoded.map(decodeContextSegment);
  }
  async clearAll(contexts) {
    const dir = this.#contextDir(contexts);
    if (await dir.exists()) {
      await dir.delete();
    }
  }
  #contextDir(contexts) {
    let dir = this.#rootDir;
    for (const ctx of contexts) {
      if (ctx === ".." || ctx === "." || !ctx.length || ctx.includes("\\")) {
        throw new import_general.StorageError(`Context segment "${ctx}" is not allowed`);
      }
      dir = dir.directory(encodeContextSegment(ctx));
    }
    return dir;
  }
  #resolve(contexts, key) {
    validateKey(key);
    return this.#contextDir(contexts).file(key);
  }
}
function validateKey(key) {
  if (!key.length) {
    throw new import_general.StorageError("Key must not be empty");
  }
  if (key.includes("/") || key.includes("\\") || key === ".." || key.startsWith("../") || key.includes("/../")) {
    throw new import_general.StorageError(`Key "${key}" contains path traversal characters`);
  }
  if (import_general.BaseStorageDriver.RESERVED_FILENAMES.has(key)) {
    throw new import_general.StorageError(`Key "${key}" is a reserved filename`);
  }
  if (key.endsWith(".tmp")) {
    throw new import_general.StorageError(`Key "${key}" must not end with ".tmp" (reserved for atomic writes)`);
  }
}
async function* streamToAsyncIterable(stream) {
  const reader = stream.getReader();
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      yield import_general.Bytes.of(value);
    }
  } finally {
    reader.releaseLock();
  }
}
function encodeContextSegment(segment) {
  return segment.replace(/%/g, "%25").replace(/\//g, "%2F");
}
function decodeContextSegment(encoded) {
  return encoded.replace(/%2F/gi, "/").replace(/%25/g, "%");
}
