// @matter/nodejs/dist/cjs/storage/fs/FileStorageDriver.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FileStorageDriver_exports = {};
__export(FileStorageDriver_exports, {
  FileStorageDriver: () => FileStorageDriver
});
module.exports = __toCommonJS(FileStorageDriver_exports);
var import_general = require("@matter/general");
var import_node_fs = require("node:fs");
var import_promises = require("node:fs/promises");
var import_node_path = require("node:path");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = new import_general.Logger("FileStorageDriver");
class FileStorageDriver extends import_general.FilesystemStorageDriver {
  static id = "file";
  static async create(namespace, _descriptor) {
    const storage = new FileStorageDriver(namespace);
    try {
      await storage.initialize();
    } catch (error) {
      await storage.close().catch(() => {
      });
      throw error;
    }
    return storage;
  }
  #path;
  #clear;
  isInitialized = false;
  #writeFileBlocker = /* @__PURE__ */ new Map();
  #index = {};
  constructor(namespaceOrPath, clear = false) {
    super(typeof namespaceOrPath === "string" || namespaceOrPath === void 0 ? void 0 : namespaceOrPath);
    this.#path = typeof namespaceOrPath === "string" ? namespaceOrPath : namespaceOrPath !== void 0 ? this.root.directory.path : "";
    this.#clear = clear;
  }
  get initialized() {
    return this.isInitialized;
  }
  async initialize() {
    if (this.isInitialized) {
      throw new import_general.StorageError("Storage already initialized!");
    }
    await super.initialize();
    if (this.#clear) {
      this.#index = {};
      await (0, import_promises.rm)(this.#path, { recursive: true, force: true });
    }
    await (0, import_promises.mkdir)(this.#path, { recursive: true });
    const files = await (0, import_promises.readdir)(this.#path);
    for (const file of files) {
      if (import_general.StorageDriver.RESERVED_FILENAMES.has(file)) {
        continue;
      }
      if (file.endsWith(".tmp")) {
        logger.info("Deleting orphaned temp file", file);
        await (0, import_promises.rm)((0, import_node_path.join)(this.#path, file), { force: true });
        continue;
      }
      const parts = decodeURIComponent(file).split(".");
      this.#markValue(parts.slice(0, -1), parts[parts.length - 1]);
    }
    this.isInitialized = true;
  }
  #indexFor(contexts) {
    let node = this.#index;
    for (const name of contexts) {
      let child = node.contexts?.get(name);
      if (child === void 0) {
        child = {};
        if (!node.contexts) {
          node.contexts = /* @__PURE__ */ new Map();
        }
        node.contexts.set(name, child);
      }
      node = child;
    }
    return node;
  }
  #markValue(contexts, key) {
    const index = this.#indexFor(contexts);
    if (!index.keys) {
      index.keys = /* @__PURE__ */ new Set();
    }
    index.keys.add(key);
  }
  async #finishAllWrites(filename) {
    if (filename !== void 0 && this.#writeFileBlocker.has(filename) || filename === void 0 && this.#writeFileBlocker.size > 0) {
      for (let i = 0; i < 10; i++) {
        await import_general.MatterAggregateError.allSettled(
          filename !== void 0 ? [this.#writeFileBlocker.get(filename)] : this.#writeFileBlocker.values(),
          "Error on finishing all file system writes to storage"
        );
        if (!this.#writeFileBlocker.size) {
          return;
        }
      }
      await this.#fsyncStorageDir();
    }
  }
  async close() {
    this.isInitialized = false;
    await this.#finishAllWrites();
    await super.close();
  }
  filePath(fileName) {
    return (0, import_node_path.join)(this.#path, fileName);
  }
  getContextBaseKey(contexts) {
    for (const ctx of contexts) {
      if (!ctx.length || ctx.includes(".")) {
        throw new import_general.StorageError("Context must not contain empty segments or leading or trailing dots.");
      }
    }
    return contexts.join(".");
  }
  buildStorageKey(contexts, key) {
    if (!key.length) {
      throw new import_general.StorageError("Key must not be an empty string.");
    }
    if (key === "tmp") {
      throw new import_general.StorageError(`Key "tmp" is reserved for atomic write operations.`);
    }
    const contextKey = this.getContextBaseKey(contexts);
    const rawName = contextKey.length ? `${contextKey}.${key}` : key;
    return encodeURIComponent(rawName).replace(/[!'()]/g, escape).replace(/\*/g, "%2A");
  }
  async has(contexts, key) {
    const index = this.#indexFor(contexts);
    return !!index.keys?.has(key);
  }
  async get(contexts, key) {
    const fileName = this.filePath(this.buildStorageKey(contexts, key));
    await this.#finishAllWrites(fileName);
    let value;
    try {
      value = await (0, import_promises.readFile)(fileName, "utf8");
    } catch (error) {
      if (error.code === "ENOENT") return void 0;
      throw error;
    }
    try {
      return (0, import_general.fromJson)(value);
    } catch (error) {
      logger.error(`Failed to parse storage value for key ${key} in context ${contexts.join(".")}`);
    }
  }
  /**
   * Read a blob from the flat file format.  Retained for legacy migration support — the
   * {@link StorageMigration} cross-type kv→blob strategy duck-types this method to extract
   * blobs from old FileStorageDriver data.
   *
   * @deprecated Only used for migration from legacy flat-file blob format.
   */
  async openBlob(contexts, key) {
    const fileName = this.filePath(this.buildStorageKey(contexts, key));
    if (await this.has(contexts, key)) {
      return await (0, import_node_fs.openAsBlob)(fileName);
    }
    return new Blob();
  }
  /**
   * Write blob data in the flat file format.  Retained for legacy migration support.
   *
   * @deprecated Only used for migration from legacy flat-file blob format.
   */
  async writeBlobFromStream(contexts, key, stream) {
    return this.#writeFile(contexts, key, stream);
  }
  async set(contexts, keyOrValues, value) {
    if (typeof keyOrValues === "string") {
      return this.#writeFile(contexts, keyOrValues, (0, import_general.toJson)(value));
    }
    const promises = new Array();
    for (const [key, value2] of Object.entries(keyOrValues)) {
      promises.push(this.#writeFile(contexts, key, (0, import_general.toJson)(value2)));
    }
    await import_general.MatterAggregateError.allSettled(promises, "Error when writing values into filesystem storage");
  }
  /** According to Node.js documentation, writeFile is not atomic. This method ensures atomicity. */
  async #writeFile(contexts, key, valueOrStream) {
    const fileName = this.buildStorageKey(contexts, key);
    if (import_general.StorageDriver.RESERVED_FILENAMES.has(fileName)) {
      throw new import_general.StorageError(
        `Key "${key}" in context "${contexts.join(".")}" maps to reserved file "${fileName}"`
      );
    }
    const blocker = this.#writeFileBlocker.get(fileName);
    if (blocker !== void 0) {
      await blocker;
      return this.#writeFile(contexts, key, valueOrStream);
    }
    const promise = this.#writeAndMoveFile(this.filePath(fileName), valueOrStream).finally(() => {
      this.#writeFileBlocker.delete(fileName);
      this.#markValue(contexts, key);
    });
    this.#writeFileBlocker.set(fileName, promise);
    return promise;
  }
  async #writeAndMoveFile(filepath, valueOrStream) {
    const tmpName = `${filepath}.tmp`;
    const handle = await (0, import_promises.open)(tmpName, "w");
    const isStream = valueOrStream instanceof ReadableStream;
    const writer = handle.createWriteStream({ encoding: isStream ? null : "utf8", flush: true });
    const { resolver, rejecter, promise: writePromise } = (0, import_general.createPromise)();
    writer.on("finish", resolver);
    writer.on("error", rejecter);
    let reader;
    try {
      if (isStream) {
        reader = valueOrStream.getReader();
        while (true) {
          const { value: chunk, done } = await reader.read();
          if (chunk) {
            if (!writer.write(chunk)) {
              await new Promise((resolve) => writer.once("drain", resolve));
            }
          }
          if (done) {
            break;
          }
        }
      } else {
        writer.write(valueOrStream);
      }
      writer.end();
      await writePromise;
    } finally {
      if (isStream && reader) {
        if (valueOrStream.locked) {
          reader.releaseLock();
        }
        await valueOrStream.cancel();
      }
      await handle.close();
    }
    await (0, import_promises.rename)(tmpName, filepath);
  }
  /**
   * fsync on a directory ensures there are no rename operations etc. which haven't been persisted to disk.
   * We do this as best effort to ensure that all writes are persisted to disk.
   */
  async #fsyncStorageDir() {
    if (process.platform === "win32") {
      return;
    }
    const fd = await (0, import_promises.open)(this.#path, "r");
    try {
      await fd.sync();
    } catch (error) {
      logger.warn(`Failed to fsync storage directory ${this.#path}`, error);
    } finally {
      await fd.close();
    }
  }
  async delete(contexts, key) {
    await this.#rm(this.buildStorageKey(contexts, key), this.#indexFor(contexts), key);
  }
  async #rm(filename, index, key) {
    await this.#finishAllWrites(filename);
    return (0, import_promises.rm)(this.filePath(filename), { force: true }).finally(() => {
      index.keys?.delete(key);
    });
  }
  /** Returns all keys of a storage context without keys of sub-contexts */
  async keys(contexts) {
    const index = this.#indexFor(contexts);
    return index.keys ? [...index.keys] : [];
  }
  async values(contexts) {
    const values = {};
    const promises = new Array();
    for (const key of await this.keys(contexts)) {
      promises.push(
        (async () => {
          const value = await this.get(contexts, key);
          if (value !== void 0) {
            values[key] = value;
          }
        })()
      );
    }
    await import_general.MatterAggregateError.allSettled(promises, "Error when reading values from filesystem storage");
    return values;
  }
  contexts(contexts) {
    const index = this.#indexFor(contexts);
    return index.contexts ? [...index.contexts.keys()] : [];
  }
  async clearAll(contexts) {
    await this.#finishAllWrites();
    const parent = this.#indexFor(contexts.slice(0, -1));
    const name = contexts[contexts.length - 1];
    await this.#clearChildContext(contexts, parent, name);
  }
  async #clearChildContext(contexts, parent, name) {
    const index = parent.contexts?.get(name);
    if (index === void 0) {
      return;
    }
    if (index.contexts) {
      for (const name2 of index.contexts.keys()) {
        await this.#clearChildContext([...contexts, name2], index, name2);
      }
    }
    if (index.keys) {
      await import_general.MatterAggregateError.allSettled(
        [...index.keys].map((key) => this.#rm(this.buildStorageKey(contexts, key), index, key)),
        `Error deleting keys of storage context ${contexts.join(".")}`
      );
    }
  }
}
