// @matter/nodejs/dist/cjs/storage/fs/FlatFileBlobStorageDriver.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FlatFileBlobStorageDriver_exports = {};
__export(FlatFileBlobStorageDriver_exports, {
  FlatFileBlobStorageDriver: () => FlatFileBlobStorageDriver
});
module.exports = __toCommonJS(FlatFileBlobStorageDriver_exports);
var import_general = require("@matter/general");
var import_node_fs = require("node:fs");
var import_promises = require("node:fs/promises");
var import_node_path = require("node:path");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class FlatFileBlobStorageDriver extends import_general.FilesystemBlobStorageDriver {
  static id = "file";
  static create(namespace, _descriptor) {
    return new FlatFileBlobStorageDriver(namespace);
  }
  #path;
  #initialized = false;
  constructor(namespace) {
    super(namespace);
    this.#path = this.root.directory.path;
  }
  get initialized() {
    return this.#initialized;
  }
  async initialize() {
    await super.initialize();
    await (0, import_promises.mkdir)(this.#path, { recursive: true });
    this.#initialized = true;
  }
  async close() {
    this.#initialized = false;
    await super.close();
  }
  async openBlob(contexts, key) {
    const filePath = (0, import_node_path.join)(this.#path, buildStorageKey(contexts, key));
    if (!(0, import_node_fs.existsSync)(filePath)) {
      return new Blob([]);
    }
    return await (0, import_node_fs.openAsBlob)(filePath);
  }
  // Stream chunks to disk via atomic write-to-tmp + rename
  async writeBlobFromStream(contexts, key, stream) {
    const fileName = buildStorageKey(contexts, key);
    const filePath = (0, import_node_path.join)(this.#path, fileName);
    const tmpPath = `${filePath}.tmp`;
    const reader = stream.getReader();
    const writer = (0, import_node_fs.createWriteStream)(tmpPath);
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        const chunk = import_general.Bytes.of(value);
        if (!writer.write(chunk)) {
          await new Promise((resolve) => writer.once("drain", resolve));
        }
      }
      writer.end();
      await new Promise((resolve, reject) => {
        writer.on("finish", resolve);
        writer.on("error", reject);
      });
      await (0, import_promises.rename)(tmpPath, filePath);
    } catch (error) {
      writer.destroy();
      await (0, import_promises.rm)(tmpPath, { force: true });
      throw error;
    } finally {
      reader.releaseLock();
    }
  }
  async delete(contexts, key) {
    const filePath = (0, import_node_path.join)(this.#path, buildStorageKey(contexts, key));
    await (0, import_promises.rm)(filePath, { force: true });
  }
  async has(contexts, key) {
    const filePath = (0, import_node_path.join)(this.#path, buildStorageKey(contexts, key));
    return (0, import_node_fs.existsSync)(filePath);
  }
  // When contexts is empty, prefixDot is "" and startsWith("") is always true — every file
  // matches.  The caller (keys/contexts/clearAll) then applies its own filtering on the remainder.
  async keys(contexts) {
    const prefixDot = contextPrefixDot(contexts);
    const result = new Array();
    for (const { remainder } of await this.#listDecodedFiles(prefixDot)) {
      if (remainder.length && !remainder.includes(".")) {
        result.push(remainder);
      }
    }
    return result;
  }
  async contexts(contexts) {
    const prefixDot = contextPrefixDot(contexts);
    const found = /* @__PURE__ */ new Set();
    for (const { remainder } of await this.#listDecodedFiles(prefixDot)) {
      const dotIndex = remainder.indexOf(".");
      if (dotIndex > 0) {
        found.add(remainder.substring(0, dotIndex));
      }
    }
    return [...found];
  }
  async clearAll(contexts) {
    const prefix = getContextBaseKey(contexts);
    const prefixDot = prefix.length ? `${prefix}.` : "";
    for (const { file, decoded } of await this.#listDecodedFiles("")) {
      if (decoded === prefix || decoded.startsWith(prefixDot)) {
        await (0, import_promises.rm)((0, import_node_path.join)(this.#path, file), { force: true });
      }
    }
  }
  /**
   * List all data files in the directory, decoded and filtered (excludes reserved files and
   * `.tmp` atomicity artifacts).  Returns raw filename + decoded name + remainder after prefix.
   */
  async #listDecodedFiles(prefixDot) {
    let files;
    try {
      files = await (0, import_promises.readdir)(this.#path);
    } catch (e) {
      if (e.code === "ENOENT") {
        return [];
      }
      throw e;
    }
    const result = new Array();
    for (const file of files) {
      if (import_general.BaseStorageDriver.RESERVED_FILENAMES.has(file) || file.endsWith(".tmp")) {
        continue;
      }
      let decoded;
      try {
        decoded = decodeURIComponent(file);
      } catch {
        continue;
      }
      if (!decoded.startsWith(prefixDot)) {
        continue;
      }
      result.push({ file, decoded, remainder: decoded.substring(prefixDot.length) });
    }
    return result;
  }
}
function getContextBaseKey(contexts) {
  for (const ctx of contexts) {
    if (!ctx.length || ctx.includes(".")) {
      throw new import_general.StorageError("Context must not contain empty segments or leading or trailing dots.");
    }
  }
  return contexts.join(".");
}
function contextPrefixDot(contexts) {
  const prefix = getContextBaseKey(contexts);
  return prefix.length ? `${prefix}.` : "";
}
function buildStorageKey(contexts, key) {
  if (!key.length) {
    throw new import_general.StorageError("Key must not be an empty string.");
  }
  if (key.includes(".")) {
    throw new import_general.StorageError(`Key "${key}" must not contain "." (would be misinterpreted as context separator)`);
  }
  if (key.endsWith(".tmp")) {
    throw new import_general.StorageError(`Key "${key}" must not end with ".tmp" (reserved for atomic writes)`);
  }
  if (import_general.BaseStorageDriver.RESERVED_FILENAMES.has(key)) {
    throw new import_general.StorageError(`Key "${key}" is a reserved filename`);
  }
  const contextKey = getContextBaseKey(contexts);
  const rawName = contextKey.length ? `${contextKey}.${key}` : key;
  return encodeURIComponent(rawName).replace(/[!'()]/g, escape).replace(/\*/g, "%2A");
}
