// @matter/protocol/dist/esm/dcl/DclRestApiTypes.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DclErrorCodes = /* @__PURE__ */ ((DclErrorCodes2) => {
  DclErrorCodes2[DclErrorCodes2["Ok"] = 0] = "Ok";
  DclErrorCodes2[DclErrorCodes2["Canceled"] = 1] = "Canceled";
  DclErrorCodes2[DclErrorCodes2["Unknown"] = 2] = "Unknown";
  DclErrorCodes2[DclErrorCodes2["InvalidArgument"] = 3] = "InvalidArgument";
  DclErrorCodes2[DclErrorCodes2["DeadlineExceeded"] = 4] = "DeadlineExceeded";
  DclErrorCodes2[DclErrorCodes2["NotFound"] = 5] = "NotFound";
  DclErrorCodes2[DclErrorCodes2["AlreadyExists"] = 6] = "AlreadyExists";
  DclErrorCodes2[DclErrorCodes2["PermissionDenied"] = 7] = "PermissionDenied";
  DclErrorCodes2[DclErrorCodes2["ResourceExhausted"] = 8] = "ResourceExhausted";
  DclErrorCodes2[DclErrorCodes2["FailedPrecondition"] = 9] = "FailedPrecondition";
  DclErrorCodes2[DclErrorCodes2["Aborted"] = 10] = "Aborted";
  DclErrorCodes2[DclErrorCodes2["OutOfRange"] = 11] = "OutOfRange";
  DclErrorCodes2[DclErrorCodes2["Unimplemented"] = 12] = "Unimplemented";
  DclErrorCodes2[DclErrorCodes2["Internal"] = 13] = "Internal";
  DclErrorCodes2[DclErrorCodes2["Unavailable"] = 14] = "Unavailable";
  DclErrorCodes2[DclErrorCodes2["DataLoss"] = 15] = "DataLoss";
  DclErrorCodes2[DclErrorCodes2["Unauthenticated"] = 16] = "Unauthenticated";
  return DclErrorCodes2;
})(DclErrorCodes || {});
export {
  DclErrorCodes
};
