// @matter/model/dist/esm/common/DeviceClassification.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DeviceClassification = /* @__PURE__ */ ((DeviceClassification2) => {
  DeviceClassification2["Base"] = "base";
  DeviceClassification2["Utility"] = "utility";
  DeviceClassification2["Application"] = "application";
  DeviceClassification2["Simple"] = "simple";
  DeviceClassification2["Dynamic"] = "dynamic";
  DeviceClassification2["Node"] = "node";
  return DeviceClassification2;
})(DeviceClassification || {});
((DeviceClassification2) => {
  function isUtility(classification) {
    switch (classification) {
      case "node" /* Node */:
      case "utility" /* Utility */:
        return true;
      default:
        return false;
    }
  }
  DeviceClassification2.isUtility = isUtility;
  function isApplication(classification) {
    return !isUtility(classification);
  }
  DeviceClassification2.isApplication = isApplication;
})(DeviceClassification || (DeviceClassification = {}));
export {
  DeviceClassification
};
