// @matter/protocol/dist/esm/peer/PeerCommunicationError.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Duration, Instant, MatterError } from "@matter/general";
class PeerCommunicationError extends MatterError {
}
class TransientPeerCommunicationError extends PeerCommunicationError {
}
class PeerUnreachableError extends TransientPeerCommunicationError {
  constructor(timeOffline) {
    super(`Peer has been unreachable for ${Duration.format(timeOffline)}`);
  }
}
class PeerUnresponsiveError extends TransientPeerCommunicationError {
  constructor(timeWaited = Instant) {
    super(
      `Peer is no longer responding to active session${timeWaited > 0 ? ` (timed out after ${Duration.format(timeWaited)})` : ""}`
    );
  }
}
class PeerShutdownError extends TransientPeerCommunicationError {
  constructor(message = "Peer reports shutdown") {
    super(message);
  }
}
class PeerInitiatedCloseError extends TransientPeerCommunicationError {
  constructor(message = "Session ended by peer") {
    super(message);
  }
}
class PeerLeftError extends PeerCommunicationError {
  constructor(message = "Peer has left the fabric") {
    super(message);
  }
}
class FabricRemovedError extends PeerCommunicationError {
  constructor(message = "This node is no longer a member of the fabric") {
    super(message);
  }
}
class FabricChangedError extends PeerCommunicationError {
  constructor(message = "The fabric associated with this session has changed") {
    super(message);
  }
}
class FailsafeExpiredError extends PeerCommunicationError {
  constructor(message = "Failsafe timer has expired") {
    super(message);
  }
}
class CommissioningTransitionError extends PeerCommunicationError {
}
class OperationalAddressChangedError extends PeerCommunicationError {
  constructor(message = "Peer's operational address has changed") {
    super(message);
  }
}
export {
  CommissioningTransitionError,
  FabricChangedError,
  FabricRemovedError,
  FailsafeExpiredError,
  OperationalAddressChangedError,
  PeerCommunicationError,
  PeerInitiatedCloseError,
  PeerLeftError,
  PeerShutdownError,
  PeerUnreachableError,
  PeerUnresponsiveError,
  TransientPeerCommunicationError
};
