// @matter/protocol/dist/esm/peer/PeerExchangeProvider.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { PeerAddress } from "#peer/PeerAddress.js";
import { ExchangeProvider } from "#protocol/ExchangeProvider.js";
import { MRP } from "#protocol/MRP.js";
import { ChannelType, InternalError } from "@matter/general";
import { INTERACTION_PROTOCOL_ID } from "@matter/types";
import { PeerConnection } from "./PeerConnection.js";
class PeerExchangeProvider extends ExchangeProvider {
  #peer;
  #context;
  constructor(peer, context) {
    super(context.exchanges);
    this.#peer = peer;
    this.#context = context;
  }
  get maxPathsPerInvoke() {
    return this.#peer.sessionParameters.maxPathsPerInvoke;
  }
  get peerAddress() {
    return this.#peer.address;
  }
  /**
   * The transport type of the current session. Returns the actual transport (TCP/UDP) of the
   * newest active session, or UDP as default when no session is active.
   */
  get channelType() {
    const session = this.#peer.newestSession();
    if (session && !session.isClosed) {
      return session.channel.transportChannel.type;
    }
    return ChannelType.UDP;
  }
  async connect(options) {
    await this.#peer.connect({
      abort: options?.abort,
      network: options?.network,
      connectionTimeout: options?.connectionTimeout,
      requiredTransport: options?.requiredTransport,
      preferredTransport: options?.preferredTransport
    });
  }
  async initiateExchange(options) {
    const abort = options?.abort;
    const isGroup = PeerAddress.isGroup(this.#peer.address);
    while (true) {
      if (!isGroup && !options?.requireExistingSession) {
        await this.connect(options);
        abort?.throwIfAborted();
      }
      const network = this.#context.networks.select(this.#peer, options?.network);
      const slot = await network.semaphore.obtainSlot(abort);
      try {
        abort?.throwIfAborted();
        let session;
        if (isGroup) {
          session = await this.#context.sessions.groupSessionForAddress(
            this.#peer.address,
            this.#context.exchanges
          );
        } else {
          const transports = this.#peer.resolveTransports(
            options?.requiredTransport,
            options?.preferredTransport
          );
          if (transports === void 0) {
            session = this.#peer.newestSession();
          } else {
            for (const t of transports) {
              session = this.#peer.newestSession(t);
              if (session) break;
            }
          }
        }
        if (session === void 0) {
          if (options?.requireExistingSession) {
            throw new InternalError("No existing session available for probe");
          }
          slot.close();
          continue;
        }
        const exchange = PeerConnection.createExchange(
          this.#peer,
          this.#context.exchanges,
          session,
          network,
          options?.protocol ?? INTERACTION_PROTOCOL_ID,
          options?.addressOverride
        );
        exchange.closing.on(() => {
          slot.close();
        });
        return exchange;
      } catch (e) {
        slot.close();
        throw e;
      }
    }
  }
  maximumPeerResponseTime(expectedProcessingTime, includeMaximumSendingTime) {
    return MRP.maxPeerResponseTimeOf({
      channelType: this.channelType,
      isPeerActive: true,
      localSessionParameters: this.#context.sessions.sessionParameters,
      peerSessionParameters: includeMaximumSendingTime ? this.#peer.sessionParameters : void 0,
      expectedProcessingTime
    });
  }
}
export {
  PeerExchangeProvider
};
