// @matter/protocol/dist/esm/peer/PeerSet.js
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MdnsService } from "#mdns/MdnsService.js";
import { PeerAddress } from "#peer/PeerAddress.js";
import { SessionManager } from "#session/SessionManager.js";
import {
  Abort,
  AsyncObservable,
  BasicSet,
  ChannelType,
  Environmental,
  ImplementationError,
  isIpNetworkChannel,
  isIPv6,
  MatterAggregateError,
  NetworkUnreachableError,
  ObserverGroup
} from "@matter/general";
import { FabricIndex } from "@matter/types";
import { NetworkProfiles } from "./NetworkProfile.js";
import { Peer } from "./Peer.js";
import { PeerTimingParameters } from "./PeerTimingParameters.js";
class PeerSet {
  #lifetime;
  #sessions;
  #peers = new BasicSet();
  #disconnected = AsyncObservable();
  #peerContext;
  #networks;
  #observers = new ObserverGroup();
  #exchanges;
  #transportPreference;
  constructor(context) {
    const { lifetime, sessions, names, networks, timing, handleError } = context;
    this.#lifetime = lifetime.join("peers");
    this.#sessions = sessions;
    this.#networks = networks;
    const self = this;
    this.#peerContext = {
      sessions,
      names,
      networks,
      timing: PeerTimingParameters(timing),
      handleError,
      get exchanges() {
        if (self.#exchanges === void 0) {
          throw new ImplementationError("Client networking is not initialized");
        }
        return self.#exchanges;
      },
      openSocket: (address, abort) => this.#openSocket(address, abort),
      closed: (peer) => this.#peers.delete(peer),
      join: (name) => this.#lifetime.join(name)
    };
    this.#peers.added.on((peer) => {
      peer.sessions.deleted.on(() => {
        if (!peer.sessions.size) {
          this.#disconnected.emit(peer);
        }
      });
    });
    this.#sessions.sessions.added.on((session) => {
      if (session.peerAddress.fabricIndex === FabricIndex.NO_FABRIC || session.isClosed) {
        return;
      }
      this.addKnownPeer({
        address: session.peerAddress,
        operationalAddress: operationalAddressOf(session)
      });
    });
    this.#observers.on(this.#sessions.sessions.added, (session) => {
      if (session.fabric === void 0) {
        return;
      }
      this.for(session.peerAddress).sessions.add(session);
    });
  }
  set exchanges(exchanges) {
    this.#exchanges = exchanges;
  }
  get added() {
    return this.#peers.added;
  }
  get deleted() {
    return this.#peers.deleted;
  }
  get empty() {
    return this.#peers.empty;
  }
  get disconnected() {
    return this.#disconnected;
  }
  /**
   * Default outgoing transport preference applied to peers created by this set when they have
   * no per-peer override. Honored as a soft preference at connect time.
   */
  get transportPreference() {
    return this.#transportPreference;
  }
  set transportPreference(pref) {
    this.#transportPreference = pref;
  }
  /**
   * Unconditional get.
   *
   * Creates the peer if not already present.
   */
  for(address) {
    let peer = this.get(address);
    if (peer) {
      return peer;
    }
    peer = new Peer({ address }, this.#peerContext);
    this.#applyDefaultPreference(peer);
    this.#peers.add(peer);
    return peer;
  }
  #applyDefaultPreference(peer) {
    if (peer.transportPreference === void 0 && this.#transportPreference !== void 0) {
      peer.transportPreference = this.#transportPreference;
    }
  }
  has(item) {
    return !!this.get(item);
  }
  get size() {
    return this.#peers.size;
  }
  find(predicate) {
    return this.#peers.find(predicate);
  }
  filter(predicate) {
    return this.#peers.filter(predicate);
  }
  map(mapper) {
    return this.#peers.map(mapper);
  }
  [Symbol.iterator]() {
    return this.#peers[Symbol.iterator]();
  }
  static [Environmental.create](env) {
    const instance = new PeerSet({
      lifetime: env,
      sessions: env.get(SessionManager),
      names: env.get(MdnsService).names,
      networks: env.get(NetworkProfiles)
    });
    env.set(PeerSet, instance);
    return instance;
  }
  get peers() {
    return this.#peers;
  }
  get networks() {
    return this.#networks;
  }
  get timing() {
    return this.#peerContext.timing;
  }
  set timing(timing) {
    this.#peerContext.timing = PeerTimingParameters(timing);
  }
  set handleError(handler) {
    this.#peerContext.handleError = handler;
  }
  /**
   * Retrieve a peer by address.
   */
  get(peer) {
    if ("address" in peer) {
      return this.#peers.get("address", PeerAddress(peer.address));
    }
    return this.#peers.get("address", PeerAddress(peer));
  }
  /**
   * Terminate any active peer networking operations.
   *
   * Also sets {@link exchanges} to undefined to prevent future connections.
   */
  async disconnect() {
    var _stack = [];
    try {
      const _disconnecting = __using(_stack, this.#lifetime.join("disconnecting"));
      this.#exchanges = void 0;
      await MatterAggregateError.allSettled(
        this.#peers.map((peer) => peer.disconnect()),
        "Error disconnecting peers"
      );
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async close() {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      for (const peer of this.#peers) {
        await peer.close();
      }
      this.#observers.close();
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async #openSocket(address, abort) {
    const channelType = "type" in address && address.type === "tcp" ? ChannelType.TCP : ChannelType.UDP;
    let lookupAddress;
    if (channelType === ChannelType.UDP) {
      lookupAddress = isIPv6(address.ip) ? "::" : "0.0.0.0";
    }
    const operationalInterface = this.#peerContext.exchanges.interfaceFor(channelType, lookupAddress);
    if (operationalInterface === void 0) {
      throw new NetworkUnreachableError(
        `No ${channelType.toUpperCase()} interface available for address ${address.ip}`
      );
    }
    return await Abort.race(abort, operationalInterface.openChannel(address));
  }
  addKnownPeer(descriptor) {
    let peer = this.get(descriptor.address);
    if (peer === void 0) {
      peer = new Peer(descriptor, this.#peerContext);
      this.#applyDefaultPreference(peer);
      this.#peers.add(peer);
    }
    if (descriptor.operationalAddress !== void 0) {
      peer.descriptor.operationalAddress = descriptor.operationalAddress;
    }
    if (descriptor.discoveryData !== void 0) {
      peer.descriptor.discoveryData = {
        ...peer.descriptor.discoveryData,
        ...descriptor.discoveryData
      };
    }
    if (descriptor.caseAuthenticatedTags !== void 0) {
      peer.descriptor.caseAuthenticatedTags = descriptor.caseAuthenticatedTags;
    }
    if (descriptor.sessionParameters !== void 0) {
      peer.descriptor.sessionParameters = {
        ...peer.descriptor.sessionParameters,
        ...descriptor.sessionParameters
      };
    }
    return peer;
  }
}
function operationalAddressOf(session) {
  if (session.isClosed || !isIpNetworkChannel(session.channel)) {
    return;
  }
  if (session.channel.transportChannel.type === ChannelType.TCP) {
    return;
  }
  return { ...session.channel.networkAddress, type: "udp" };
}
export {
  PeerSet
};
