// @matter/protocol/dist/esm/peer/PeerAddress.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DiagnosticPresentation } from "@matter/general";
import { GroupId } from "@matter/types";
const interned = /* @__PURE__ */ Symbol("interned-logical-address");
const internedAddresses = /* @__PURE__ */ new Map();
function PeerAddress(address) {
  if (address === void 0) {
    return void 0;
  }
  if (interned in address) {
    return address;
  }
  let internedFabric = internedAddresses.get(address.fabricIndex);
  if (internedFabric === void 0) {
    internedAddresses.set(address.fabricIndex, internedFabric = /* @__PURE__ */ new Map());
  }
  let internedAddress = internedFabric.get(address.nodeId);
  if (internedAddress) {
    return internedAddress;
  }
  internedAddress = InternedAddress(address);
  internedFabric.set(address.nodeId, internedAddress);
  return internedAddress;
}
const InternedAddressPrototoype = {};
Object.defineProperties(InternedAddressPrototoype, {
  [interned]: { value: true },
  toString: {
    value() {
      return `@${this.fabricIndex.toString(16)}:${this.nodeId.toString(16)}`;
    }
  },
  [DiagnosticPresentation.value]: {
    get() {
      return this.toString();
    }
  }
});
function InternedAddress(address) {
  const interned2 = Object.create(InternedAddressPrototoype, {
    fabricIndex: { value: address.fabricIndex, enumerable: true },
    nodeId: { value: address.nodeId, enumerable: true }
  });
  Object.freeze(interned2);
  return interned2;
}
((PeerAddress2) => {
  function is(addr1, addr2) {
    if (addr1 === void 0 || addr2 === void 0) {
      return false;
    }
    return addr1.fabricIndex === addr2.fabricIndex && addr1.nodeId === addr2.nodeId;
  }
  PeerAddress2.is = is;
  function isGroup(address) {
    return GroupId.isGroupNodeId(address.nodeId);
  }
  PeerAddress2.isGroup = isGroup;
})(PeerAddress || (PeerAddress = {}));
class PeerAddressMap extends Map {
  delete(key) {
    return super.delete(PeerAddress(key));
  }
  has(key) {
    return super.has(PeerAddress(key));
  }
  set(key, value) {
    return super.set(PeerAddress(key), value);
  }
  get(key) {
    return super.get(PeerAddress(key));
  }
}
class PeerAddressSet extends Set {
  add(value) {
    return super.add(PeerAddress(value));
  }
  has(value) {
    return super.has(PeerAddress(value));
  }
  delete(value) {
    return super.delete(PeerAddress(value));
  }
}
export {
  PeerAddress,
  PeerAddressMap,
  PeerAddressSet
};
