// @matter/model/dist/esm/standard/elements/root-node.element.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  DeviceTypeElement as DeviceType,
  RequirementElement as Requirement,
  ConditionElement as Condition
} from "../../elements/index.js";
const RootNodeDt = DeviceType(
  { name: "RootNode", id: 22, classification: "node" },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 22, revision: 4 }], element: "attribute" })
  ),
  Requirement(
    { name: "AccessControl", id: 31, conformance: "M", element: "serverCluster", quality: "I" },
    Requirement({ name: "MNGD", conformance: "[ManagedAclAllowed]", constraint: "desc", element: "feature" }),
    Requirement({ name: "Extension", conformance: "ACLExtensionCond", element: "attribute" })
  ),
  Requirement({ name: "BasicInformation", id: 40, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({
    name: "LocalizationConfiguration",
    id: 43,
    conformance: "LanguageLocale",
    element: "serverCluster",
    quality: "I"
  }),
  Requirement(
    { name: "TimeFormatLocalization", id: 44, conformance: "TimeLocale", element: "serverCluster", quality: "I" }
  ),
  Requirement({ name: "UnitLocalization", id: 45, conformance: "UnitLocale", element: "serverCluster", quality: "I" }),
  Requirement(
    { name: "PowerSourceConfiguration", id: 46, conformance: "O, D", element: "serverCluster", quality: "I" }
  ),
  Requirement({ name: "GeneralCommissioning", id: 48, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "NetworkCommissioning", id: 49, conformance: "!CustomNetworkConfig", element: "serverCluster" }),
  Requirement({ name: "DiagnosticLogs", id: 50, conformance: "O", element: "serverCluster", quality: "I" }),
  Requirement({ name: "GeneralDiagnostics", id: 51, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "SoftwareDiagnostics", id: 52, conformance: "O", element: "serverCluster", quality: "I" }),
  Requirement({ name: "ThreadNetworkDiagnostics", id: 53, conformance: "[Thread]", element: "serverCluster" }),
  Requirement({ name: "WiFiNetworkDiagnostics", id: 54, conformance: "[WiFi]", element: "serverCluster" }),
  Requirement({ name: "EthernetNetworkDiagnostics", id: 55, conformance: "[Ethernet]", element: "serverCluster" }),
  Requirement(
    {
      name: "TimeSynchronization",
      id: 56,
      conformance: "TimeSyncCond, TimeSyncWithClientCond, TimeSyncWithNTPCCond, TimeSyncWithTZCond, TLSClientCond, TLSCertificatesCond, O",
      element: "serverCluster",
      quality: "I"
    },
    Requirement({
      name: "TIMESYNCCLIENT",
      conformance: "TimeSyncWithClientCond, [TLSCertificatesCond | TLSClientCond].a+, O",
      element: "feature"
    }),
    Requirement({
      name: "NTPCLIENT",
      conformance: "TimeSyncWithNTPCCond, [TLSCertificatesCond | TLSClientCond].a+, O",
      element: "feature"
    }),
    Requirement({ name: "TIMEZONE", conformance: "TimeSyncWithTZCond, O", element: "feature" })
  ),
  Requirement(
    {
      name: "TimeSynchronization",
      id: 56,
      conformance: "TimeSyncWithClientCond, O",
      element: "clientCluster",
      quality: "I"
    },
    Requirement({
      name: "TIMESYNCCLIENT",
      conformance: "TimeSyncWithClientCond, [TLSCertificatesCond | TLSClientCond].a+, O",
      element: "feature"
    }),
    Requirement({
      name: "NTPCLIENT",
      conformance: "TimeSyncWithNTPCCond, [TLSCertificatesCond | TLSClientCond].a+, O",
      element: "feature"
    }),
    Requirement({ name: "TIMEZONE", conformance: "TimeSyncWithTZCond, O", element: "feature" })
  ),
  Requirement({ name: "AdministratorCommissioning", id: 60, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "OperationalCredentials", id: 62, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement({ name: "GroupKeyManagement", id: 63, conformance: "M", element: "serverCluster", quality: "I" }),
  Requirement(
    { name: "IcdManagement", id: 70, conformance: "SIT | LIT", element: "serverCluster", quality: "I" },
    Requirement({ name: "LONGIDLETIMESUPPORT", conformance: "LIT", element: "feature" })
  ),
  Requirement({
    name: "TlsCertificateManagement",
    id: 2049,
    conformance: "TLSCertificatesCond, O",
    element: "serverCluster",
    quality: "I"
  }),
  Requirement({ name: "TlsClientManagement", id: 2050, conformance: "TLSClientCond, O", element: "serverCluster", quality: "I" }),
  Requirement({ name: "PowerSource", id: 17, conformance: "PowerSourceCond, O", element: "deviceType" }),
  Condition({ name: "CustomNetworkConfig" }),
  Condition({ name: "ManagedAclAllowed" }),
  Condition({ name: "TimeSyncCond" }),
  Condition({ name: "TimeSyncWithClientCond" }),
  Condition({ name: "TimeSyncWithNtpcCond" }),
  Condition({ name: "TimeSyncWithTzCond" }),
  Condition({ name: "TlsCertificatesCond" }),
  Condition({ name: "TlsClientCond" }),
  Condition({ name: "PowerSourceCond" }),
  Condition({ name: "AclExtensionCond" })
);
MatterDefinition.children.push(RootNodeDt);
export {
  RootNodeDt
};
