// @matter/model/dist/esm/standard/elements/unit-localization.element.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const UnitLocalization = Cluster(
  { name: "UnitLocalization", id: 45, classification: "node" },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap" },
    Field({ name: "TEMP", constraint: "0", title: "TemperatureUnit" })
  ),
  Attribute(
    { name: "TemperatureUnit", id: 0, type: "TempUnitEnum", access: "RW VM", conformance: "TEMP", quality: "N" }
  ),
  Attribute(
    {
      name: "SupportedTemperatureUnits",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "TEMP",
      constraint: "2 to 3",
      quality: "F"
    },
    Field({ name: "entry", type: "TempUnitEnum" })
  ),
  Datatype(
    { name: "TempUnitEnum", type: "enum8" },
    Field({ name: "Fahrenheit", id: 0, conformance: "O.a2+" }),
    Field({ name: "Celsius", id: 1, conformance: "O.a2+" }),
    Field({ name: "Kelvin", id: 2, conformance: "O.a2+" })
  )
);
MatterDefinition.children.push(UnitLocalization);
export {
  UnitLocalization
};
