// @matter/model/dist/cjs/standard/elements/joint-fabric-datastore.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var joint_fabric_datastore_element_exports = {};
__export(joint_fabric_datastore_element_exports, {
  JointFabricDatastore: () => JointFabricDatastore
});
module.exports = __toCommonJS(joint_fabric_datastore_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const JointFabricDatastore = (0, import_elements.ClusterElement)(
  { name: "JointFabricDatastore", id: 1874, classification: "node" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({ name: "AnchorRootCa", id: 0, type: "octstr", access: "R A", conformance: "P, M" }),
  (0, import_elements.AttributeElement)({ name: "AnchorNodeId", id: 1, type: "node-id", access: "R A", conformance: "P, M" }),
  (0, import_elements.AttributeElement)({ name: "AnchorVendorId", id: 2, type: "vendor-id", access: "R A", conformance: "P, M" }),
  (0, import_elements.AttributeElement)({ name: "FriendlyName", id: 3, type: "string", access: "R A", conformance: "P, M", constraint: "max 32" }),
  (0, import_elements.AttributeElement)(
    { name: "GroupKeySetList", id: 4, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreGroupKeySetStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "GroupList", id: 5, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreGroupInformationEntryStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "NodeList", id: 6, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreNodeInformationEntryStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "AdminList", id: 7, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreAdministratorInformationEntryStruct" })
  ),
  (0, import_elements.AttributeElement)({ name: "Status", id: 8, type: "DatastoreStatusEntryStruct", access: "R A", conformance: "P, M" }),
  (0, import_elements.AttributeElement)(
    { name: "EndpointGroupIdList", id: 9, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreEndpointGroupIDEntryStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "EndpointBindingList", id: 10, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreEndpointBindingEntryStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "NodeKeySetList", id: 11, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreNodeKeySetEntryStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "NodeAclList", id: 12, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreACLEntryStruct" })
  ),
  (0, import_elements.AttributeElement)(
    { name: "NodeEndpointList", id: 13, type: "list", access: "R A", conformance: "P, M" },
    (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreEndpointEntryStruct" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddKeySet", id: 0, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupKeySet", id: 0, type: "DatastoreGroupKeySetStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "UpdateKeySet", id: 1, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupKeySet", id: 0, type: "DatastoreGroupKeySetStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveKeySet", id: 2, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 0, type: "uint16", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddGroup", id: 3, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 1, type: "string", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 2, type: "uint16", conformance: "M", constraint: "1 to 65534", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupCat", id: 3, type: "uint16", conformance: "M", constraint: "desc", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupCatVersion", id: 4, type: "uint16", conformance: "M", constraint: "1 to 65534", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupPermission", id: 5, type: "DatastoreAccessControlEntryPrivilegeEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "UpdateGroup", id: 4, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 1, type: "string", conformance: "M", constraint: "max 32", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 2, type: "uint16", conformance: "M", constraint: "1 to 65535", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupCat", id: 3, type: "uint16", conformance: "M", constraint: "desc", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupCatVersion", id: 4, type: "uint16", conformance: "M", constraint: "1 to 65535", quality: "X" }),
    (0, import_elements.FieldElement)({
      name: "GroupPermission",
      id: 5,
      type: "DatastoreAccessControlEntryPrivilegeEnum",
      conformance: "M",
      constraint: "x"
    })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveGroup", id: 5, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "group-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddAdmin", id: 6, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 1, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 2, type: "string", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({ name: "VendorId", id: 3, type: "vendor-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Icac", id: 4, type: "octstr", conformance: "M", constraint: "max 400" })
  ),
  (0, import_elements.CommandElement)(
    { name: "UpdateAdmin", id: 7, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 1, type: "string", conformance: "M", constraint: "max 32", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "Icac", id: 2, type: "octstr", conformance: "M", constraint: "max 400", quality: "X" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveAdmin", id: 8, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddPendingNode", id: 9, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 1, type: "string", conformance: "M", constraint: "max 32" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RefreshNode", id: 10, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "UpdateNode", id: 11, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 1, type: "string", conformance: "M", constraint: "max 32" })
  ),
  (0, import_elements.CommandElement)(
    { name: "RemoveNode", id: 12, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "UpdateEndpointForNode",
      id: 13,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 0, type: "endpoint-no", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NodeId", id: 1, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 2, type: "string", conformance: "M", constraint: "max 32" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "AddGroupIdToEndpointForNode",
      id: 14,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 1, type: "endpoint-no", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 2, type: "group-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveGroupIdFromEndpointForNode",
      id: 15,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 1, type: "endpoint-no", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 2, type: "group-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "AddBindingToEndpointForNode",
      id: 16,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 1, type: "endpoint-no", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Binding", id: 2, type: "DatastoreBindingTargetStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveBindingFromEndpointForNode",
      id: 17,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ListId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 1, type: "endpoint-no", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NodeId", id: 2, type: "node-id", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddAclToNode", id: 18, access: "A", conformance: "P, M", direction: "request", response: "status" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AclEntry", id: 1, type: "DatastoreAccessControlEntryStruct", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveAclFromNode",
      id: 19,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "ListId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NodeId", id: 1, type: "node-id", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreStateEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Pending", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Committed", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "DeletePending", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "CommitFailed", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreStatusEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "State", id: 0, type: "DatastoreStateEnum", access: "R V", conformance: "M", default: 0 }),
    (0, import_elements.FieldElement)({ name: "UpdateTimestamp", id: 1, type: "epoch-s", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FailureCode", id: 2, type: "status", access: "R V", conformance: "M", default: 0 })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreNodeKeySetEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 1, type: "uint16", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StatusEntry", id: 2, type: "DatastoreStatusEntryStruct", access: "R V", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreAccessControlEntryPrivilegeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "View", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ProxyView", id: 2, conformance: "D" }),
    (0, import_elements.FieldElement)({ name: "Operate", id: 3, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Manage", id: 4, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Administer", id: 5, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreGroupInformationEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "GroupId", id: 0, type: "uint64", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 1, type: "string", access: "R V", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({
      name: "GroupKeySetId",
      id: 2,
      type: "uint16",
      access: "R V",
      conformance: "M",
      constraint: "1 to 65534",
      quality: "X"
    }),
    (0, import_elements.FieldElement)({
      name: "GroupCat",
      id: 3,
      type: "uint16",
      access: "R V",
      conformance: "M",
      constraint: "desc",
      quality: "X"
    }),
    (0, import_elements.FieldElement)({
      name: "GroupCatVersion",
      id: 4,
      type: "uint16",
      access: "R V",
      conformance: "M",
      constraint: "1 to 65534",
      quality: "X"
    }),
    (0, import_elements.FieldElement)({
      name: "GroupPermission",
      id: 5,
      type: "DatastoreAccessControlEntryPrivilegeEnum",
      access: "R V",
      conformance: "M"
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreBindingTargetStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Node", id: 1, type: "node-id", conformance: "Endpoint" }),
    (0, import_elements.FieldElement)({ name: "Group", id: 2, type: "group-id", conformance: "!Endpoint", constraint: "min 1" }),
    (0, import_elements.FieldElement)({ name: "Endpoint", id: 3, type: "endpoint-no", conformance: "!Group" }),
    (0, import_elements.FieldElement)({ name: "Cluster", id: 4, type: "cluster-id", conformance: "O" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreEndpointBindingEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 1, type: "endpoint-no", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ListId", id: 2, type: "uint16", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({
      name: "Binding",
      id: 3,
      type: "DatastoreBindingTargetStruct",
      access: "R V",
      conformance: "M",
      constraint: "desc"
    }),
    (0, import_elements.FieldElement)({ name: "StatusEntry", id: 4, type: "DatastoreStatusEntryStruct", access: "R V", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreEndpointGroupIDEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 1, type: "endpoint-no", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupId", id: 2, type: "group-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StatusEntry", id: 3, type: "DatastoreStatusEntryStruct", access: "R V", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreEndpointEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 0, type: "endpoint-no", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "NodeId", id: 1, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 2, type: "string", access: "R V", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({ name: "StatusEntry", id: 3, type: "DatastoreStatusEntryStruct", access: "R V", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreAccessControlEntryAuthModeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Pase", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Case", id: 2, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Group", id: 3, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreAccessControlTargetStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Cluster", id: 0, type: "cluster-id", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "Endpoint", id: 1, type: "endpoint-no", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "DeviceType", id: 2, type: "devtype-id", conformance: "M", quality: "X" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreAccessControlEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "Privilege", id: 1, type: "DatastoreAccessControlEntryPrivilegeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AuthMode", id: 2, type: "DatastoreAccessControlEntryAuthModeEnum", conformance: "M" }),
    (0, import_elements.FieldElement)(
      {
        name: "Subjects",
        id: 3,
        type: "list",
        conformance: "M",
        constraint: "max subjectsPerAccessControlEntry",
        quality: "X"
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "subject-id" })
    ),
    (0, import_elements.FieldElement)(
      {
        name: "Targets",
        id: 4,
        type: "list",
        conformance: "M",
        constraint: "max targetsPerAccessControlEntry",
        quality: "X"
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "DatastoreAccessControlTargetStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreACLEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 0, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "ListId", id: 1, type: "uint16", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AclEntry", id: 2, type: "DatastoreAccessControlEntryStruct", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "StatusEntry", id: 3, type: "DatastoreStatusEntryStruct", access: "R V", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreNodeInformationEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 1, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 2, type: "string", access: "R V", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({
      name: "CommissioningStatusEntry",
      id: 3,
      type: "DatastoreStatusEntryStruct",
      access: "R V",
      conformance: "M"
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreAdministratorInformationEntryStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "NodeId", id: 1, type: "node-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "FriendlyName", id: 2, type: "string", access: "R V", conformance: "M", constraint: "max 32" }),
    (0, import_elements.FieldElement)({ name: "VendorId", id: 3, type: "vendor-id", access: "R V", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "Icac", id: 4, type: "octstr", access: "R V", conformance: "M", constraint: "max 400" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreGroupKeySecurityPolicyEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "TrustFirst", id: 0, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreGroupKeyMulticastPolicyEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "PerGroupId", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "AllNodes", id: 1, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "DatastoreGroupKeySetStruct", type: "struct" },
    (0, import_elements.FieldElement)({ name: "GroupKeySetId", id: 0, type: "uint16", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "GroupKeySecurityPolicy", id: 1, type: "DatastoreGroupKeySecurityPolicyEnum", conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "EpochKey0", id: 2, type: "octstr", conformance: "M", constraint: "16", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochStartTime0", id: 3, type: "epoch-us", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochKey1", id: 4, type: "octstr", conformance: "M", constraint: "16", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochStartTime1", id: 5, type: "epoch-us", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochKey2", id: 6, type: "octstr", conformance: "M", constraint: "16", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "EpochStartTime2", id: 7, type: "epoch-us", conformance: "M", quality: "X" }),
    (0, import_elements.FieldElement)({ name: "GroupKeyMulticastPolicy", id: 8, type: "DatastoreGroupKeyMulticastPolicyEnum", conformance: "P, M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(JointFabricDatastore);
