// @matter/model/dist/cjs/standard/elements/joint-fabric-administrator-cluster.element.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var joint_fabric_administrator_cluster_element_exports = {};
__export(joint_fabric_administrator_cluster_element_exports, {
  JointFabricAdministrator: () => JointFabricAdministrator
});
module.exports = __toCommonJS(joint_fabric_administrator_cluster_element_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const JointFabricAdministrator = (0, import_elements.ClusterElement)(
  { name: "JointFabricAdministrator", id: 1875, classification: "node" },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "AdministratorFabricIndex",
    id: 0,
    type: "fabric-idx",
    access: "A",
    conformance: "P, M",
    constraint: "1 to 254",
    quality: "X"
  }),
  (0, import_elements.CommandElement)({
    name: "IcaccsrRequest",
    id: 0,
    access: "A",
    conformance: "P, M",
    direction: "request",
    response: "IcaccsrResponse"
  }),
  (0, import_elements.CommandElement)(
    { name: "IcaccsrResponse", id: 1, conformance: "P, M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "Icaccsr", id: 0, type: "octstr", conformance: "M", constraint: "max 600" })
  ),
  (0, import_elements.CommandElement)(
    { name: "AddIcac", id: 2, access: "A", conformance: "P, M", direction: "request", response: "IcacResponse" },
    (0, import_elements.FieldElement)({ name: "IcacValue", id: 1, type: "octstr", conformance: "M", constraint: "max 400" })
  ),
  (0, import_elements.CommandElement)(
    { name: "IcacResponse", id: 3, conformance: "P, M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "StatusCode", id: 0, type: "ICACResponseStatusEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "OpenJointCommissioningWindow",
      id: 4,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "CommissioningTimeout", id: 0, type: "uint16", conformance: "M", constraint: "desc" }),
    (0, import_elements.FieldElement)({ name: "PakePasscodeVerifier", id: 1, type: "octstr", conformance: "M", constraint: "97" }),
    (0, import_elements.FieldElement)({ name: "Discriminator", id: 2, type: "uint16", conformance: "M", constraint: "max 4095" }),
    (0, import_elements.FieldElement)({ name: "Iterations", id: 3, type: "uint32", conformance: "M", constraint: "1000 to 100000" }),
    (0, import_elements.FieldElement)({ name: "Salt", id: 4, type: "octstr", conformance: "M", constraint: "16 to 32" })
  ),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorRequest",
    id: 5,
    access: "A",
    conformance: "P, M",
    direction: "request",
    response: "TransferAnchorResponse"
  }),
  (0, import_elements.CommandElement)(
    { name: "TransferAnchorResponse", id: 6, access: "A", conformance: "P, M", direction: "response" },
    (0, import_elements.FieldElement)({ name: "StatusCode", id: 0, type: "TransferAnchorResponseStatusEnum", conformance: "M" })
  ),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorComplete",
    id: 7,
    access: "A",
    conformance: "P, M",
    direction: "request",
    response: "status"
  }),
  (0, import_elements.CommandElement)(
    {
      name: "AnnounceJointFabricAdministrator",
      id: 8,
      access: "A",
      conformance: "P, M",
      direction: "request",
      response: "status"
    },
    (0, import_elements.FieldElement)({ name: "EndpointId", id: 0, type: "endpoint-no", conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ICACResponseStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvalidPublicKey", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "InvalidIcac", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "TransferAnchorResponseStatusEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransferAnchorStatusDatastoreBusy", id: 1, conformance: "M" }),
    (0, import_elements.FieldElement)({ name: "TransferAnchorStatusNoUserConsent", id: 2, conformance: "M" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusCodeEnum", type: "enum8" },
    (0, import_elements.FieldElement)({ name: "Busy", id: 2, conformance: "P, M" }),
    (0, import_elements.FieldElement)({ name: "PakeParameterError", id: 3, conformance: "P, M" }),
    (0, import_elements.FieldElement)({ name: "WindowNotOpen", id: 4, conformance: "P, M" }),
    (0, import_elements.FieldElement)({ name: "VidNotVerified", id: 5, conformance: "P, M" }),
    (0, import_elements.FieldElement)({ name: "InvalidAdministratorFabricIndex", id: 6, conformance: "P, M" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(JointFabricAdministrator);
