// @matter/nodejs/dist/cjs/storage/sqlite/platform/BunSqlite.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BunSqlite_exports = {};
__export(BunSqlite_exports, {
  createBunDatabase: () => createBunDatabase
});
module.exports = __toCommonJS(BunSqlite_exports);
var import_general = require("@matter/general");
var import_bun_sqlite = require("bun:sqlite");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("BunSqlite");
function createBunDatabase(path) {
  const db = new import_bun_sqlite.Database(path, {
    strict: true,
    create: true
  });
  if (path === ":memory:") {
    return db;
  }
  db.exec("PRAGMA journal_mode = WAL");
  db.exec("PRAGMA synchronous = NORMAL");
  return {
    prepare: db.prepare.bind(db),
    exec: db.exec.bind(db),
    close: () => {
      try {
        db.fileControl(import_bun_sqlite.constants.SQLITE_FCNTL_PERSIST_WAL, 0);
      } catch (error) {
        logger.warn("Failed to clear WAL persistence flag:", error);
      }
      try {
        db.exec("PRAGMA wal_checkpoint(TRUNCATE)");
      } catch (error) {
        logger.warn("WAL checkpoint failed, WAL will be replayed on next open:", error);
      } finally {
        db.close();
      }
    }
  };
}
