// @matter/model/dist/cjs/common/DataModelPath.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataModelPath_exports = {};
__export(DataModelPath_exports, {
  DataModelPath: () => DataModelPath
});
module.exports = __toCommonJS(DataModelPath_exports);
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class DataModelPath {
  parent;
  id;
  type;
  constructor(id, type, parent) {
    this.id = id;
    this.type = type;
    this.parent = parent;
  }
  #identity(includeType) {
    if (this.type && includeType) {
      return `${this.type}#${this.id}`;
    }
    return this.id;
  }
  toString(includeType) {
    if (this.parent) {
      const sep = this.type === "marker" ? "" : ".";
      return `${this.parent.toString(includeType)}${sep}${this.#identity(includeType)}`;
    }
    return this.#identity(includeType).toString();
  }
  toArray() {
    if (this.parent) {
      return [...this.parent.toArray(), this.id];
    }
    return [this.id];
  }
  get [import_general.Diagnostic.presentation]() {
    return "squash";
  }
  get [import_general.Diagnostic.value]() {
    const result = Array();
    this.#appendDiagnostic(result);
    return result;
  }
  #appendDiagnostic(result) {
    if (this.parent) {
      this.parent.#appendDiagnostic(result);
      if (this.type !== "marker") {
        result.push(".");
      }
    }
    result.push(import_general.Diagnostic.strong(this.id));
  }
  at(id, type) {
    return new DataModelPath(id, type, this);
  }
  static none = new DataModelPath("none");
}
