// @matter/model/dist/cjs/common/DeviceClassification.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DeviceClassification_exports = {};
__export(DeviceClassification_exports, {
  DeviceClassification: () => DeviceClassification
});
module.exports = __toCommonJS(DeviceClassification_exports);
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DeviceClassification = /* @__PURE__ */ ((DeviceClassification2) => {
  DeviceClassification2["Base"] = "base";
  DeviceClassification2["Utility"] = "utility";
  DeviceClassification2["Application"] = "application";
  DeviceClassification2["Simple"] = "simple";
  DeviceClassification2["Dynamic"] = "dynamic";
  DeviceClassification2["Node"] = "node";
  return DeviceClassification2;
})(DeviceClassification || {});
((DeviceClassification2) => {
  function isUtility(classification) {
    switch (classification) {
      case "node" /* Node */:
      case "utility" /* Utility */:
        return true;
      default:
        return false;
    }
  }
  DeviceClassification2.isUtility = isUtility;
  function isApplication(classification) {
    return !isUtility(classification);
  }
  DeviceClassification2.isApplication = isApplication;
})(DeviceClassification || (DeviceClassification = {}));
