// @matter/node/dist/esm/node/remote/WebSocketConnection.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Abort, asJson, Bytes, Logger, Stream, WebSocketClient } from "@matter/general";
const logger = Logger.get("WebSocketConnection");
class WebSocketConnection {
  #ws;
  #url;
  #env;
  #abort = new Abort();
  #nextId = 1;
  #pending = /* @__PURE__ */ new Map();
  #subscriptionListeners = /* @__PURE__ */ new Map();
  #readLoop;
  constructor(url, env) {
    this.#url = url;
    this.#env = env;
  }
  get url() {
    return this.#url;
  }
  /**
   * Open the WebSocket connection and start the read loop.
   */
  async open() {
    this.#ws = await this.#env.get(WebSocketClient).connect(this.#url);
    this.#readLoop = this.#runReadLoop();
  }
  /**
   * Close the connection.
   */
  async close() {
    this.#abort();
    for (const pending of this.#pending.values()) {
      pending.reject(new Error("Connection closed"));
    }
    this.#pending.clear();
    this.#subscriptionListeners.clear();
    await this.#readLoop;
    this.#ws = void 0;
  }
  /**
   * Send a request and wait for a single response.
   */
  async request(request) {
    const id = String(this.#nextId++);
    const message = { ...request, id };
    return new Promise((resolve, reject) => {
      this.#pending.set(id, { resolve, reject });
      this.#send(message).catch(reject);
    });
  }
  /**
   * Send a subscribe request.  Returns the subscription ID and installs a listener for streaming responses.
   *
   * The caller receives the "ok" confirmation, then subsequent update/delete responses arrive via the listener.
   */
  async subscribe(request, listener) {
    const id = String(this.#nextId++);
    const message = { ...request, id };
    return new Promise((resolve, reject) => {
      this.#pending.set(id, {
        resolve: (response) => {
          if (response.kind === "ok") {
            this.#subscriptionListeners.set(id, listener);
          }
          resolve(response);
        },
        reject
      });
      this.#send(message).catch(reject);
    });
  }
  /**
   * Remove a subscription listener.
   */
  unsubscribe(id) {
    this.#subscriptionListeners.delete(id);
  }
  async #send(message) {
    if (!this.#ws) {
      throw new Error("WebSocket not connected");
    }
    const writer = this.#ws.writable.getWriter();
    try {
      await writer.write(asJson(message));
    } finally {
      writer.releaseLock();
    }
  }
  async #runReadLoop() {
    if (!this.#ws) {
      return;
    }
    try {
      for await (let message of Stream.iterable(this.#ws.readable)) {
        if (Abort.is(this.#abort)) {
          break;
        }
        try {
          const text = typeof message === "string" ? message : Bytes.toString(message);
          const response = JSON.parse(text);
          this.#dispatch(response);
        } catch (e) {
          logger.error("Error processing WebSocket message:", e);
        }
      }
    } catch (e) {
      if (!Abort.is(this.#abort)) {
        logger.error("WebSocket read loop error:", e);
      }
    }
  }
  #dispatch(response) {
    const id = response.id;
    if (id !== void 0) {
      const subscriptionListener = this.#subscriptionListeners.get(id);
      if (subscriptionListener) {
        subscriptionListener(response);
        return;
      }
      const pending = this.#pending.get(id);
      if (pending) {
        this.#pending.delete(id);
        if (response.kind === "error") {
          pending.reject(new Error(response.message));
        } else {
          pending.resolve(response);
        }
        return;
      }
    }
    logger.debug("Received unmatched response:", response.kind, id ?? "(no id)");
  }
}
export {
  WebSocketConnection
};
