// @matter/protocol/dist/cjs/action/request/Read.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Read_exports = {};
__export(Read_exports, {
  Read: () => Read
});
module.exports = __toCommonJS(Read_exports);
var import_general = require("@matter/general");
var import_model = require("@matter/model");
var import_MalformedRequestError = require("./MalformedRequestError.js");
var import_Specifier = require("./Specifier.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function Read(optionsOrSelector, ...selectors) {
  if (optionsOrSelector === void 0) {
    throw new import_MalformedRequestError.MalformedRequestError(`Read action designates no attributes or events`);
  }
  let options;
  if ("kind" in optionsOrSelector) {
    selectors = [optionsOrSelector, ...selectors];
    options = {};
  } else {
    options = optionsOrSelector;
  }
  const { fabricFilter = true, interactionModelRevision = import_model.Specification.INTERACTION_MODEL_REVISION } = options;
  let { attributes: attributeRequests, versionFilters, events: eventRequests, eventFilters } = options;
  const result = {
    isFabricFiltered: fabricFilter,
    interactionModelRevision,
    [import_general.Diagnostic.value]: () => import_general.Diagnostic.dict({
      attributes: attributeRequests?.length ? selectors.length ? selectors.filter(({ kind }) => kind === "attribute").map((path) => (0, import_Specifier.resolvePathForSpecifier)(path)).join(", ") : attributeRequests.length : void 0,
      events: eventRequests?.length ? selectors.length ? selectors.filter(({ kind }) => kind === "event").map((path) => (0, import_Specifier.resolvePathForSpecifier)(path)).join(", ") : eventRequests.length : void 0,
      dataVersionFilters: versionFilters?.length ?? result.dataVersionFilters?.length,
      eventFilters: eventFilters?.length ? eventFilters.map(({ nodeId, eventMin }) => `${nodeId}=${eventMin}`).join(", ") : void 0
    })
  };
  for (const selector of selectors) {
    reifySelector(selector);
  }
  if (attributeRequests) {
    result.attributeRequests = attributeRequests;
  }
  if (versionFilters) {
    result.dataVersionFilters = versionFilters;
  }
  if (eventRequests) {
    result.eventRequests = eventRequests;
  }
  if (eventFilters !== void 0) {
    result.eventFilters = eventFilters;
  }
  return result;
  function reifySelector(selector) {
    switch (selector.kind) {
      case "attribute":
        reifyAttributeSelector(selector);
        break;
      case "event":
        reifyEventSelector(selector);
        break;
      default:
        throw new import_MalformedRequestError.MalformedRequestError(`Invalid selector kind "${selector.kind}"`);
    }
  }
  function reifyAttributeSelector(selector) {
    const cluster = import_Specifier.Specifier.clusterOf(selector);
    const { endpoint } = selector;
    if (typeof endpoint === "object" && endpoint?.versions && typeof cluster === "object" && cluster !== null) {
      const version = endpoint.versions?.[(0, import_general.camelize)(cluster.name)];
      if (version !== void 0) {
        const filter = {
          path: { endpointId: endpoint.number, clusterId: cluster.id },
          dataVersion: version
        };
        if (versionFilters === void 0) {
          versionFilters = [filter];
        } else if (versionFilters.find(
          ({ path: { endpointId, clusterId } }) => endpointId === endpoint.number && clusterId === cluster.id
        ) === void 0) {
          versionFilters.push(filter);
        }
      }
    }
    if (attributeRequests === void 0) {
      attributeRequests = [];
    }
    const prototype = {};
    if (endpoint !== void 0) {
      prototype.endpointId = import_Specifier.Specifier.endpointIdOf(selector);
    }
    if (cluster !== void 0) {
      prototype.clusterId = cluster.id;
    }
    let { attributes } = selector;
    if (attributes === void 0) {
      attributeRequests.push(prototype);
      return;
    }
    if (!Array.isArray(attributes)) {
      attributes = [attributes];
    }
    for (const specifier of attributes) {
      attributeRequests.push({ ...prototype, attributeId: import_Specifier.Specifier.attributeFor(cluster, specifier).id });
    }
  }
  function reifyEventSelector(selector) {
    const cluster = import_Specifier.Specifier.clusterOf(selector);
    const { endpoint } = selector;
    if (typeof endpoint === "object" && endpoint.minEvent !== void 0) {
      if (eventFilters === void 0) {
        eventFilters = [{ eventMin: endpoint.minEvent }];
      }
    }
    if (eventRequests === void 0) {
      eventRequests = [];
    }
    const prototype = {};
    if (endpoint !== void 0) {
      prototype.endpointId = import_Specifier.Specifier.endpointIdOf(selector);
    }
    if (cluster !== void 0) {
      prototype.clusterId = cluster.id;
    }
    let { events } = selector;
    if (events === void 0) {
      eventRequests.push(prototype);
      return;
    }
    if (!Array.isArray(events)) {
      events = [events];
    }
    for (const specifier of events) {
      eventRequests.push({ ...prototype, eventId: import_Specifier.Specifier.eventFor(cluster, specifier).id });
    }
  }
}
((Read2) => {
  function containsAttribute(request) {
    return !!request.attributeRequests?.length;
  }
  Read2.containsAttribute = containsAttribute;
  function containsEvent(request) {
    return !!request.eventRequests?.length;
  }
  Read2.containsEvent = containsEvent;
  function Attribute(selector) {
    return {
      kind: "attribute",
      ...selector
    };
  }
  Read2.Attribute = Attribute;
  function Event(selector) {
    return {
      kind: "event",
      ...selector
    };
  }
  Read2.Event = Event;
})(Read || (Read = {}));
