// @matter/protocol/dist/cjs/action/request/Write.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Write_exports = {};
__export(Write_exports, {
  Write: () => Write
});
module.exports = __toCommonJS(Write_exports);
var import_general = require("@matter/general");
var import_model = require("@matter/model");
var import_types = require("@matter/types");
var import_access_control = require("@matter/types/clusters/access-control");
var import_MalformedRequestError = require("./MalformedRequestError.js");
var import_Specifier = require("./Specifier.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AclClusterId = import_access_control.AccessControl.id;
const AclAttributeId = import_access_control.AccessControl.attributes.acl.id;
const AclExtensionAttributeId = import_access_control.AccessControl.attributes.extension.id;
function isAclOrExtensionPath(path) {
  const { clusterId, attributeId } = path;
  return clusterId === AclClusterId && (attributeId === AclAttributeId || attributeId === AclExtensionAttributeId);
}
function Write(optionsOrData, ...data) {
  if (optionsOrData === void 0) {
    throw new import_MalformedRequestError.MalformedRequestError(`Write action must have options or data`);
  }
  let options;
  if ("kind" in optionsOrData) {
    data = [optionsOrData, ...data];
    options = {};
  } else {
    options = optionsOrData;
  }
  const {
    writes: writeRequests = [],
    timed,
    timeout,
    chunkLists = true,
    // Should be true currently to stay in sync with chip sdk, see below
    suppressResponse,
    interactionModelRevision = import_model.Specification.INTERACTION_MODEL_REVISION
  } = options;
  const result = {
    timedRequest: !!timed || !!timeout,
    timeout,
    writeRequests,
    moreChunkedMessages: false,
    suppressResponse,
    interactionModelRevision,
    [import_general.Diagnostic.value]: () => import_general.Diagnostic.list(
      data.map((entry) => {
        const { version, value } = entry;
        return `${(0, import_Specifier.resolvePathForSpecifier)(entry)} = ${import_general.Diagnostic.json(
          value
        )}${version !== void 0 ? `(version=${version})` : ""}`;
      })
    )
  };
  for (const entry of data) {
    reifyData(entry);
  }
  if (!writeRequests.length) {
    throw new import_MalformedRequestError.MalformedRequestError(`Write action contains no attributes to write`);
  }
  return result;
  function reifyData(data2) {
    const cluster = import_Specifier.Specifier.clusterOf(data2);
    if (cluster === void 0) {
      throw new import_MalformedRequestError.MalformedRequestError(`Write action must specify a cluster`);
    }
    let { attributes } = data2;
    if (attributes === void 0) {
      throw new import_MalformedRequestError.MalformedRequestError(`Write action must specify an attribute`);
    }
    if (!Array.isArray(attributes)) {
      attributes = [attributes];
    }
    const { endpoint, value, version: dataVersion } = data2;
    const prototype = {
      path: {
        endpointId: endpoint !== void 0 ? import_Specifier.Specifier.endpointIdOf(data2) : void 0,
        clusterId: cluster.id,
        attributeId: void 0
      },
      dataVersion
    };
    for (const specifier of attributes) {
      const clusterId = cluster.id;
      const attribute = import_Specifier.Specifier.attributeFor(cluster, specifier);
      const { id: attributeId } = attribute;
      const tlv = (0, import_types.TlvOfModel)(attribute);
      if (chunkLists && // As implemented for Matter 1.4.2 in https://github.com/project-chip/connectedhomeip/pull/38263
      // Acl writes will no longer be chunked by default, all others still
      // Will be streamlined later ... see https://github.com/project-chip/connectedhomeip/issues/38270
      Array.isArray(value) && tlv instanceof import_types.ArraySchema && !isAclOrExtensionPath({ clusterId, attributeId })) {
        writeRequests.push(
          ...tlv.encodeAsChunkedArray(value, { forWriteInteraction: true }).map(({ element: data3, listIndex }) => ({
            path: {
              ...prototype.path,
              attributeId: attribute.id,
              listIndex
            },
            data: data3,
            dataVersion
          }))
        );
      } else {
        writeRequests.push({
          ...prototype,
          path: {
            ...prototype.path,
            attributeId: attribute.id
          },
          data: tlv?.encodeTlv(value, { forWriteInteraction: true })
        });
      }
      result.timedRequest ||= attribute.schema.effectiveAccess.timed === true;
    }
  }
}
((Write2) => {
  function Attribute(data) {
    return {
      kind: "attribute",
      ...data
    };
  }
  Write2.Attribute = Attribute;
})(Write || (Write = {}));
