// @matter/general/dist/cjs/net/Network.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Network_exports = {};
__export(Network_exports, {
  AddressInUseError: () => AddressInUseError,
  AddressUnreachableError: () => AddressUnreachableError,
  BindError: () => BindError,
  InterfaceType: () => InterfaceType,
  Network: () => Network,
  NetworkError: () => NetworkError,
  NetworkUnreachableError: () => NetworkUnreachableError,
  NoAddressAvailableError: () => NoAddressAvailableError,
  STANDARD_MATTER_PORT: () => STANDARD_MATTER_PORT
});
module.exports = __toCommonJS(Network_exports);
var import_MatterError = require("../MatterError.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NetworkError extends import_MatterError.MatterError {
}
class NoAddressAvailableError extends NetworkError {
}
class BindError extends NetworkError {
}
class AddressInUseError extends BindError {
}
class AddressUnreachableError extends NetworkError {
}
class NetworkUnreachableError extends NetworkError {
}
const STANDARD_MATTER_PORT = 5540;
var InterfaceType = /* @__PURE__ */ ((InterfaceType2) => {
  InterfaceType2[InterfaceType2["Unspecified"] = 0] = "Unspecified";
  InterfaceType2[InterfaceType2["WiFi"] = 1] = "WiFi";
  InterfaceType2[InterfaceType2["Ethernet"] = 2] = "Ethernet";
  InterfaceType2[InterfaceType2["Cellular"] = 3] = "Cellular";
  InterfaceType2[InterfaceType2["Thread"] = 4] = "Thread";
  return InterfaceType2;
})(InterfaceType || {});
class Network {
  /** Create a TCP server socket. Override in platform implementations that support TCP. */
  createTcpListener(_options) {
    throw new NetworkError("TCP server not supported on this platform");
  }
  /** Connect to a remote TCP endpoint. Override in platform implementations that support TCP. */
  connectTcp(_host, _port, _options) {
    throw new NetworkError("TCP client not supported on this platform");
  }
  async close() {
  }
}
