// @matter/node/dist/cjs/storage/NodeStore.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NodeStore_exports = {};
__export(NodeStore_exports, {
  NodeStore: () => NodeStore
});
module.exports = __toCommonJS(NodeStore_exports);
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NodeStore {
  #storageFactory;
  #construction;
  /**
   * Obtain the BDX blob storage driver for this node.  Lazily opened on first call.
   * Override in subclasses that support blob storage.
   */
  async bdxStore() {
    throw new import_general.ImplementationError("Blob storage is not available for this node store");
  }
  get construction() {
    return this.#construction;
  }
  constructor(storageFactory) {
    this.#storageFactory = storageFactory;
    this.#construction = (0, import_general.Construction)(this);
  }
  toString() {
    return "node store";
  }
  [import_general.Construction.construct]() {
    return this.load();
  }
  createStorageContext(name) {
    return this.#storageFactory.createContext(name);
  }
  get storageFactory() {
    return this.#storageFactory;
  }
}
