// @matter/types/dist/esm/globals/Namespace.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Namespace = /* @__PURE__ */ ((Namespace2) => {
  Namespace2[Namespace2["CommonClosure"] = 1] = "CommonClosure";
  Namespace2[Namespace2["CommonCompassDirection"] = 2] = "CommonCompassDirection";
  Namespace2[Namespace2["CommonCompassLocation"] = 3] = "CommonCompassLocation";
  Namespace2[Namespace2["CommonDirection"] = 4] = "CommonDirection";
  Namespace2[Namespace2["CommonLevel"] = 5] = "CommonLevel";
  Namespace2[Namespace2["CommonLocation"] = 6] = "CommonLocation";
  Namespace2[Namespace2["CommonNumber"] = 7] = "CommonNumber";
  Namespace2[Namespace2["CommonPosition"] = 8] = "CommonPosition";
  Namespace2[Namespace2["ElectricalMeasurement"] = 10] = "ElectricalMeasurement";
  Namespace2[Namespace2["CommodityTariffChronology"] = 11] = "CommodityTariffChronology";
  Namespace2[Namespace2["CommodityTariffCommodity"] = 13] = "CommodityTariffCommodity";
  Namespace2[Namespace2["Laundry"] = 14] = "Laundry";
  Namespace2[Namespace2["PowerSource"] = 15] = "PowerSource";
  Namespace2[Namespace2["CommonAreaNamespace"] = 16] = "CommonAreaNamespace";
  Namespace2[Namespace2["CommonLandmarkNamespace"] = 17] = "CommonLandmarkNamespace";
  Namespace2[Namespace2["CommonRelativePosition"] = 18] = "CommonRelativePosition";
  Namespace2[Namespace2["CommodityTariffFlow"] = 19] = "CommodityTariffFlow";
  Namespace2[Namespace2["Refrigerator"] = 65] = "Refrigerator";
  Namespace2[Namespace2["RoomAirConditioner"] = 66] = "RoomAirConditioner";
  Namespace2[Namespace2["Switches"] = 67] = "Switches";
  Namespace2[Namespace2["Closure"] = 68] = "Closure";
  Namespace2[Namespace2["ClosurePanel"] = 69] = "ClosurePanel";
  Namespace2[Namespace2["ClosureCovering"] = 70] = "ClosureCovering";
  Namespace2[Namespace2["ClosureWindow"] = 71] = "ClosureWindow";
  Namespace2[Namespace2["ClosureCabinet"] = 72] = "ClosureCabinet";
  return Namespace2;
})(Namespace || {});
export {
  Namespace
};
