// @matter/nodejs/dist/esm/storage/sqlite/SqliteBlobStorageDriver.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DatafileRoot, FilesystemBlobStorageDriver } from "@matter/general";
import { resolve } from "node:path";
import { platformDatabaseCreator } from "./SqlitePlatform.js";
import { SqliteStorageDriverError } from "./SqliteStorageDriverError.js";
import { buildContextKeyPair, buildContextPath, escapeGlob } from "./SqliteUtil.js";
class SqliteBlobStorageDriver extends FilesystemBlobStorageDriver {
  static id = "sqlite";
  static memoryPath = ":memory:";
  static defaultTableName = "blobstore";
  /**
   * Create a SqliteBlobStorageDriver for the given namespace using the platform-appropriate database.
   */
  static async create(namespace) {
    const storage = new SqliteBlobStorageDriver({ namespaceOrPath: namespace });
    await storage.initialize();
    return storage;
  }
  #isInitialized = false;
  #database;
  #databaseCreator;
  #dbPath;
  #tableName;
  // Prepared queries
  #queryOpenBlob;
  #queryWriteBlob;
  #queryDeleteBlob;
  #queryHasBlob;
  #queryKeysBlob;
  #queryContextsSub;
  #queryClearAll;
  /**
   * Create sqlite-based blob storage.
   *
   * @param args.databaseCreator Optional database instance creator. If omitted, resolved automatically during
   *   {@link initialize} via platform detection.
   * @param args.namespaceOrPath DataNamespace (derives path from root directory), string (direct path), or
   *   null/undefined for in-memory database
   * @param args.tableName table name
   */
  constructor(args) {
    const namespaceOrPath = args?.namespaceOrPath;
    super(namespaceOrPath instanceof DatafileRoot ? namespaceOrPath : void 0);
    this.#dbPath = typeof namespaceOrPath === "string" ? namespaceOrPath : namespaceOrPath instanceof DatafileRoot ? resolve(namespaceOrPath.directory.path, "storage.db") : SqliteBlobStorageDriver.memoryPath;
    this.#tableName = args?.tableName ?? SqliteBlobStorageDriver.defaultTableName;
    if (args?.databaseCreator) {
      this.#openDatabase(args.databaseCreator);
    }
  }
  #openDatabase(databaseCreator) {
    this.#databaseCreator = databaseCreator;
    this.#database = databaseCreator(this.#dbPath);
    const initQuery = this.#database.prepare(`
      CREATE TABLE IF NOT EXISTS ${this.#tableName} (
        context TEXT NOT NULL,
        key TEXT NOT NULL,
        data BLOB,
        PRIMARY KEY (context, key)
      ) STRICT
    `);
    initQuery.run();
    this.#queryOpenBlob = this.#database.prepare(`
      SELECT data FROM ${this.#tableName} WHERE
        context=$context AND
        key=$key
    `);
    this.#queryWriteBlob = this.#database.prepare(`
      INSERT INTO ${this.#tableName}
        (context, key, data)
      VALUES($context, $key, $data)
      ON CONFLICT(context, key)
      DO UPDATE SET data = excluded.data
    `);
    this.#queryDeleteBlob = this.#database.prepare(`
      DELETE FROM ${this.#tableName} WHERE
        context=$context AND
        key=$key
    `);
    this.#queryHasBlob = this.#database.prepare(`
      SELECT EXISTS(
        SELECT 1 FROM ${this.#tableName}
        WHERE context=$context AND key=$key
      ) as has_record
    `);
    this.#queryKeysBlob = this.#database.prepare(`
      SELECT DISTINCT key FROM ${this.#tableName} WHERE
        context=$context
    `);
    this.#queryContextsSub = this.#database.prepare(`
      SELECT DISTINCT context FROM ${this.#tableName} WHERE
        context GLOB $contextGlob
    `);
    this.#queryClearAll = this.#database.prepare(`
      DELETE FROM ${this.#tableName} WHERE
        context=$context OR context GLOB $contextGlob
    `);
  }
  get initialized() {
    return this.#isInitialized;
  }
  async initialize() {
    if (this.#isInitialized) {
      throw new SqliteStorageDriverError("initialize", this.#tableName, "Storage already initialized!");
    }
    await super.initialize();
    if (!this.#databaseCreator) {
      this.#openDatabase(await platformDatabaseCreator());
    }
    this.#isInitialized = true;
  }
  async close() {
    this.#isInitialized = false;
    this.#database?.close();
    await super.close();
  }
  openBlob(contexts, key) {
    const queryResult = this.#queryOpenBlob.get(buildContextKeyPair(contexts, key));
    if (queryResult == null || queryResult.data == null) {
      return new Blob();
    }
    return new Blob([new Uint8Array(queryResult.data)]);
  }
  // SQLite does not support incremental blob writes — the full blob must be buffered before INSERT.
  async writeBlobFromStream(contexts, key, stream) {
    const arrayBuffer = await new Response(stream).arrayBuffer();
    const bytes = new Uint8Array(arrayBuffer);
    const queryResult = this.#queryWriteBlob.run({
      ...buildContextKeyPair(contexts, key),
      data: bytes
    });
    if (Number(queryResult.changes) <= 0) {
      throw new SqliteStorageDriverError(
        "writeBlob",
        `${buildContextPath(contexts)}$${key}`,
        `Something went wrong! Value wasn't changed.`
      );
    }
  }
  delete(contexts, key) {
    this.#queryDeleteBlob.run(buildContextKeyPair(contexts, key));
  }
  has(contexts, key) {
    const result = this.#queryHasBlob.get(buildContextKeyPair(contexts, key));
    return result?.has_record === 1;
  }
  keys(contexts) {
    const queryResults = this.#queryKeysBlob.all({
      context: buildContextPath(contexts)
    }).filter((v) => v != null);
    return queryResults.map((v) => v.key);
  }
  contexts(contexts) {
    const parentCtx = contexts.length > 0 ? buildContextPath(contexts) : "";
    let subContexts;
    if (contexts.length === 0) {
      const allContexts = this.#queryContextsSub.all({ contextGlob: "*" }).filter((v) => v != null);
      subContexts = allContexts.map((v) => {
        const firstDotIndex = v.context.indexOf(".");
        if (firstDotIndex < 0) {
          return v.context;
        }
        return v.context.substring(0, firstDotIndex);
      });
    } else {
      const allSubContexts = this.#queryContextsSub.all({
        contextGlob: escapeGlob(parentCtx) + ".*"
      }).filter((v) => v != null);
      subContexts = allSubContexts.map((v) => {
        const subKey = v.context.substring(parentCtx.length + 1);
        const dotIndex = subKey.indexOf(".");
        if (dotIndex < 0) {
          return subKey;
        }
        return subKey.substring(0, dotIndex);
      });
    }
    return [...new Set(subContexts.filter((c) => c != null && c.trim().length > 0))];
  }
  clearAll(contexts) {
    if (contexts.length === 0) {
      return;
    }
    const contextPath = buildContextPath(contexts);
    this.#queryClearAll.run({
      context: contextPath,
      contextGlob: escapeGlob(contextPath) + ".*"
    });
  }
}
export {
  SqliteBlobStorageDriver
};
