// @matter/nodejs/dist/esm/storage/sqlite/SqliteUtil.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SqliteStorageDriverError } from "./SqliteStorageDriverError.js";
function buildContextPath(contexts) {
  for (const ctx of contexts) {
    if (ctx.trim().length <= 0 || ctx.includes(".")) {
      throw new SqliteStorageDriverError(
        "build",
        `{${contexts.join(",")}}`,
        "Context must not contain empty segments or leading or trailing dots."
      );
    }
  }
  return contexts.join(".");
}
function buildContextKeyPair(contexts, key) {
  return {
    context: buildContextPath(contexts),
    key: buildKey(key)
  };
}
function buildContextKeyLog(contexts, key) {
  return `${buildContextPath(contexts)}$${key}`;
}
function buildKey(key) {
  if (key.trim().length <= 0) {
    throw new SqliteStorageDriverError("build", key, "Key must not be an empty string.");
  }
  return key;
}
function escapeGlob(value) {
  return value.replace(/[*[?]/g, "[$&]");
}
export {
  buildContextKeyLog,
  buildContextKeyPair,
  buildContextPath,
  buildKey,
  escapeGlob
};
