// @matter/protocol/dist/esm/mdns/MdnsConsts.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { GlobalFabricId, NodeId } from "@matter/types";
const SERVICE_DISCOVERY_QNAME = "_services._dns-sd._udp.local";
const MATTER_COMMISSION_SERVICE_QNAME = "_matterc._udp.local";
const MATTER_COMMISSIONER_SERVICE_QNAME = "_matterd._udp.local";
const MATTER_SERVICE_QNAME = "_matter._tcp.local";
const DEFAULT_PAIRING_HINT = {
  powerCycle: true,
  deviceManual: true
};
const PAIRING_HINTS_REQUIRING_INSTRUCTION = Array(
  "customInstruction",
  "pressResetButtonForNumberOfSeconds",
  "pressResetButtonUntilLightBlinks",
  "pressResetButtonForNumberOfSecondsWithApplicationOfPower",
  "pressResetButtonUntilLightBlinksWithApplicationOfPower",
  "pressResetButtonNumberOfTimes",
  "pressSetupButtonForNumberOfSeconds",
  "pressSetupButtonUntilLightBlinks",
  "pressSetupButtonForNumberOfSecondsWithApplicationOfPower",
  "pressSetupButtonUntilLightBlinksWithApplicationOfPower",
  "pressSetupButtonNumberOfTimes"
);
const getFabricQname = (fabric) => `_I${GlobalFabricId.strOf(fabric).toUpperCase()}._sub.${MATTER_SERVICE_QNAME}`;
const getOperationalDeviceQname = (fabric, node) => `${GlobalFabricId.strOf(fabric).toUpperCase()}-${NodeId.strOf(node).toUpperCase()}.${MATTER_SERVICE_QNAME}`;
const getVendorQname = (vendorId) => `_V${vendorId}._sub.${MATTER_COMMISSION_SERVICE_QNAME}`;
const getDeviceTypeQname = (deviceType) => `_T${deviceType}._sub.${MATTER_COMMISSION_SERVICE_QNAME}`;
const getShortDiscriminatorQname = (shortDiscriminator) => `_S${shortDiscriminator}._sub.${MATTER_COMMISSION_SERVICE_QNAME}`;
const getLongDiscriminatorQname = (longDiscriminator) => `_L${longDiscriminator}._sub.${MATTER_COMMISSION_SERVICE_QNAME}`;
const getCommissioningModeQname = () => `_CM._sub.${MATTER_COMMISSION_SERVICE_QNAME}`;
const getCommissionableDeviceQname = (instanceId) => `${instanceId}.${MATTER_COMMISSION_SERVICE_QNAME}`;
const getCommissionerDeviceQname = (instanceId) => `${instanceId}.${MATTER_COMMISSIONER_SERVICE_QNAME}`;
export {
  DEFAULT_PAIRING_HINT,
  MATTER_COMMISSIONER_SERVICE_QNAME,
  MATTER_COMMISSION_SERVICE_QNAME,
  MATTER_SERVICE_QNAME,
  PAIRING_HINTS_REQUIRING_INSTRUCTION,
  SERVICE_DISCOVERY_QNAME,
  getCommissionableDeviceQname,
  getCommissionerDeviceQname,
  getCommissioningModeQname,
  getDeviceTypeQname,
  getFabricQname,
  getLongDiscriminatorQname,
  getOperationalDeviceQname,
  getShortDiscriminatorQname,
  getVendorQname
};
