// @matter/nodejs/dist/cjs/fs/fs-utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fs_utils_exports = {};
__export(fs_utils_exports, {
  isBytes: () => isBytes,
  nodeCopy: () => nodeCopy,
  nodeDelete: () => nodeDelete,
  nodeEntries: () => nodeEntries,
  nodeExists: () => nodeExists,
  nodeStat: () => nodeStat,
  resolveCopyArg: () => resolveCopyArg,
  toBytes: () => toBytes,
  writeData: () => writeData
});
module.exports = __toCommonJS(fs_utils_exports);
var import_general = require("@matter/general");
var import_node_fs = require("node:fs");
var import_promises = require("node:fs/promises");
var import_node_path = require("node:path");
var import_promises2 = require("node:stream/promises");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
async function nodeExists(path) {
  try {
    await (0, import_promises.stat)(path);
    return true;
  } catch (e) {
    if (e.code === "ENOENT") {
      return false;
    }
    throw e;
  }
}
async function nodeStat(path) {
  let s;
  try {
    s = await (0, import_promises.stat)(path);
  } catch (e) {
    if (e.code === "ENOENT") {
      throw new import_general.FileNotFoundError(`Not found: ${path}`);
    }
    throw e;
  }
  return {
    size: s.size,
    mtime: s.mtime,
    type: s.isDirectory() ? "directory" : "file"
  };
}
async function* nodeEntries(fs, dirPath, FileClass, DirClass) {
  const dirents = await (0, import_promises.readdir)(dirPath, { withFileTypes: true });
  for (const dirent of dirents) {
    const fullPath = (0, import_node_path.resolve)(dirPath, dirent.name);
    const s = await (0, import_promises.stat)(fullPath);
    const cached = {
      size: s.size,
      mtime: s.mtime,
      type: dirent.isDirectory() ? "directory" : "file"
    };
    if (dirent.isDirectory()) {
      yield new DirClass(fs, fullPath, dirent.name, cached);
    } else {
      yield new FileClass(fs, fullPath, dirent.name, cached);
    }
  }
}
function resolveCopyArg(basePath, arg) {
  if (typeof arg === "string") {
    return (0, import_node_path.resolve)(basePath, arg);
  }
  if ("path" in arg && typeof arg.path === "string") {
    return arg.path;
  }
  throw new import_general.FileTypeError("Cannot resolve path for copy argument");
}
async function nodeCopy(basePath, source, target) {
  const srcPath = resolveCopyArg(basePath, source);
  const dstPath = resolveCopyArg(basePath, target);
  await (0, import_promises.cp)(srcPath, dstPath, { recursive: true });
}
async function nodeDelete(path) {
  await (0, import_promises.rm)(path, { recursive: true, force: true });
}
function isBytes(value) {
  return ArrayBuffer.isView(value) || value instanceof ArrayBuffer || value instanceof SharedArrayBuffer;
}
function toBytes(value) {
  return value instanceof Uint8Array ? value : new Uint8Array(ArrayBuffer.isView(value) ? value.buffer : value);
}
async function streamWrite(ws, chunk) {
  if (!ws.write(chunk)) {
    await new Promise((resolve2) => ws.once("drain", resolve2));
  }
}
async function writeData(path, data) {
  if (typeof data === "string") {
    await (0, import_promises.writeFile)(path, data, "utf8");
    return;
  }
  if (isBytes(data)) {
    await (0, import_promises.writeFile)(path, toBytes(data));
    return;
  }
  const iter = Symbol.asyncIterator in data ? data[Symbol.asyncIterator]() : data[Symbol.iterator]();
  const first = await iter.next();
  if (first.done) {
    await (0, import_promises.writeFile)(path, new Uint8Array(0));
    return;
  }
  const isText = typeof first.value === "string";
  const ws = (0, import_node_fs.createWriteStream)(path, isText ? { encoding: "utf8" } : void 0);
  try {
    if (isText) {
      await streamWrite(ws, first.value);
      while (true) {
        const next = await iter.next();
        if (next.done) break;
        await streamWrite(ws, "\n" + next.value);
      }
    } else {
      await streamWrite(ws, toBytes(first.value));
      while (true) {
        const next = await iter.next();
        if (next.done) break;
        await streamWrite(ws, toBytes(next.value));
      }
    }
    ws.end();
    await (0, import_promises2.finished)(ws);
  } catch (e) {
    ws.destroy();
    throw e;
  }
}
