// @matter/nodejs/dist/cjs/fs/NodeJsFile.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NodeJsFile_exports = {};
__export(NodeJsFile_exports, {
  NodeJsFile: () => NodeJsFile
});
module.exports = __toCommonJS(NodeJsFile_exports);
var import_general = require("@matter/general");
var import_node_fs = require("node:fs");
var import_promises = require("node:fs/promises");
var import_node_path = require("node:path");
var import_fs_utils = require("./fs-utils.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NodeJsFile extends import_general.File {
  #fs;
  #path;
  #name;
  #cachedStat;
  constructor(fs, path, name, cachedStat) {
    super();
    this.#fs = fs;
    this.#path = path;
    this.#name = name;
    this.#cachedStat = cachedStat;
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#name;
  }
  get path() {
    return this.#path;
  }
  async exists() {
    return (0, import_fs_utils.nodeExists)(this.#path);
  }
  stat() {
    if (this.#cachedStat) {
      return this.#cachedStat;
    }
    return (0, import_fs_utils.nodeStat)(this.#path);
  }
  async *readBytes() {
    const s = await (0, import_fs_utils.nodeStat)(this.#path);
    if (s.type === "directory") {
      throw new import_general.FileTypeError("Cannot read bytes from a directory");
    }
    const stream = (0, import_node_fs.createReadStream)(this.#path);
    try {
      for await (const chunk of stream) {
        yield chunk;
      }
    } finally {
      stream.destroy();
    }
  }
  async *readText(options) {
    const s = await (0, import_fs_utils.nodeStat)(this.#path);
    if (s.type === "directory") {
      throw new import_general.FileTypeError("Cannot read text from a directory");
    }
    if (options?.lines) {
      const stream = (0, import_node_fs.createReadStream)(this.#path, { encoding: "utf8" });
      try {
        let carry = "";
        for await (const chunk of stream) {
          const parts = (carry + chunk).split("\n");
          carry = parts.pop();
          yield* parts;
        }
        yield carry;
      } finally {
        stream.destroy();
      }
    } else {
      const stream = (0, import_node_fs.createReadStream)(this.#path, { encoding: "utf8" });
      try {
        for await (const chunk of stream) {
          yield chunk;
        }
      } finally {
        stream.destroy();
      }
    }
  }
  async write(data) {
    await (0, import_fs_utils.writeData)(this.#path, data);
  }
  async open(purpose, mode) {
    const m = mode ?? "r";
    const flags = m === "w" ? "w" : m === "a" ? "a" : "r";
    if (flags === "w" || flags === "a") {
      await (0, import_promises.mkdir)((0, import_node_path.dirname)(this.#path), { recursive: true });
    }
    const fh = await (0, import_promises.open)(this.#path, flags);
    return new NodeJsFileHandle({ fs: this.#fs, path: this.#path, name: this.#name, purpose, fh });
  }
  async rename(newName) {
    const newPath = (0, import_node_path.resolve)((0, import_node_path.dirname)(this.#path), newName);
    await (0, import_promises.rename)(this.#path, newPath);
    this.#path = newPath;
    this.#name = (0, import_node_path.basename)(newPath);
  }
  async delete() {
    await (0, import_promises.rm)(this.#path, { recursive: true, force: true });
  }
}
class NodeJsFileHandle extends import_general.File.Handle {
  #fs;
  #path;
  #name;
  purpose;
  #fh;
  constructor(options) {
    super();
    this.#fs = options.fs;
    this.#path = options.path;
    this.#name = options.name;
    this.purpose = options.purpose;
    this.#fh = options.fh;
    import_general.FileHandleTracker.register(this);
  }
  get fs() {
    return this.#fs;
  }
  get name() {
    return this.#name;
  }
  get path() {
    return this.#path;
  }
  async exists() {
    return (0, import_fs_utils.nodeExists)(this.#path);
  }
  stat() {
    return (0, import_fs_utils.nodeStat)(this.#path);
  }
  async *readBytes() {
    const stream = (0, import_node_fs.createReadStream)(this.#path);
    try {
      for await (const chunk of stream) {
        yield chunk;
      }
    } finally {
      stream.destroy();
    }
  }
  async *readText(options) {
    if (options?.lines) {
      const stream = (0, import_node_fs.createReadStream)(this.#path, { encoding: "utf8" });
      try {
        let carry = "";
        for await (const chunk of stream) {
          const parts = (carry + chunk).split("\n");
          carry = parts.pop();
          yield* parts;
        }
        yield carry;
      } finally {
        stream.destroy();
      }
    } else {
      const stream = (0, import_node_fs.createReadStream)(this.#path, { encoding: "utf8" });
      try {
        for await (const chunk of stream) {
          yield chunk;
        }
      } finally {
        stream.destroy();
      }
    }
  }
  async write(data) {
    await (0, import_fs_utils.writeData)(this.#path, data);
  }
  async open(_purpose) {
    return this;
  }
  async writeHandle(data) {
    if (typeof data === "string") {
      await this.#fh.write(data, void 0, "utf8");
    } else {
      await this.#fh.write((0, import_fs_utils.toBytes)(data));
    }
  }
  cursor(max, bufferSize) {
    return new NodeJsCursor(this.#fh, max, bufferSize);
  }
  async truncate(size) {
    await this.#fh.truncate(size ?? 0);
  }
  async fsync() {
    await this.#fh.datasync();
  }
  async close() {
    import_general.FileHandleTracker.unregister(this);
    await this.#fh.close();
  }
  async rename(newName) {
    const newPath = (0, import_node_path.resolve)((0, import_node_path.dirname)(this.#path), newName);
    await (0, import_promises.rename)(this.#path, newPath);
    this.#path = newPath;
    this.#name = (0, import_node_path.basename)(newPath);
  }
  async delete() {
    await (0, import_promises.rm)(this.#path, { recursive: true, force: true });
  }
}
class NodeJsCursor extends import_general.File.Cursor {
  #fh;
  #shared;
  constructor(fh, max, bufferSize) {
    super(max);
    this.#fh = fh;
    this.#shared = Buffer.alloc(bufferSize ?? 8192);
  }
  async readAt(position, length, copy) {
    if (copy || length > this.#shared.length) {
      const buf = Buffer.alloc(length);
      const { bytesRead: bytesRead2 } = await this.#fh.read(buf, 0, length, position);
      return bytesRead2 < length ? buf.subarray(0, bytesRead2) : buf;
    }
    const { bytesRead } = await this.#fh.read(this.#shared, 0, length, position);
    return this.#shared.subarray(0, bytesRead);
  }
}
