// @matter/nodejs/dist/cjs/fs/NodeJsFilesystem.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NodeJsFilesystem_exports = {};
__export(NodeJsFilesystem_exports, {
  NodeJsFilesystem: () => NodeJsFilesystem
});
module.exports = __toCommonJS(NodeJsFilesystem_exports);
var import_general = require("@matter/general");
var import_promises = require("node:fs/promises");
var import_node_os = require("node:os");
var import_node_path = require("node:path");
var import_NodeJsDirectory = require("./NodeJsDirectory.js");
var import_NodeJsFile = require("./NodeJsFile.js");
var import_fs_utils = require("./fs-utils.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NodeJsFilesystem extends import_general.Filesystem {
  #rootPath;
  #tempCounter = 0;
  constructor(workingDirectory) {
    super();
    this.#rootPath = workingDirectory;
  }
  get name() {
    return "";
  }
  get path() {
    return typeof this.#rootPath === "function" ? this.#rootPath() : this.#rootPath;
  }
  async exists() {
    return (0, import_fs_utils.nodeExists)(this.path);
  }
  stat() {
    return (0, import_fs_utils.nodeStat)(this.path);
  }
  rename() {
    throw new import_general.FileTypeError("Cannot rename root");
  }
  async delete() {
    await (0, import_promises.rm)(this.path, { recursive: true, force: true });
  }
  async *entries() {
    yield* (0, import_fs_utils.nodeEntries)(this, this.path, import_NodeJsFile.NodeJsFile, import_NodeJsDirectory.NodeJsDirectory);
  }
  file(name) {
    return new import_NodeJsFile.NodeJsFile(this, (0, import_node_path.resolve)(this.path, name), name);
  }
  directory(name) {
    return new import_NodeJsDirectory.NodeJsDirectory(this, (0, import_node_path.resolve)(this.path, name), name);
  }
  async mkdir() {
    await (0, import_promises.mkdir)(this.path, { recursive: true });
  }
  async copy(source, target) {
    await (0, import_fs_utils.nodeCopy)(this.path, source, target);
  }
  tempFilename() {
    return (0, import_node_path.resolve)((0, import_node_os.tmpdir)(), `matter-${process.pid}-${Date.now()}-${this.#tempCounter++}`);
  }
  tempDirectory() {
    return new import_NodeJsDirectory.NodeJsDirectory(this, this.tempFilename(), "");
  }
}
