// @matter/protocol/dist/cjs/common/BleScanner.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BleScanner_exports = {};
__export(BleScanner_exports, {
  BleScanner: () => BleScanner
});
module.exports = __toCommonJS(BleScanner_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_Ble = require("../ble/Ble.js");
var import_BtpCodec = require("../codec/BtpCodec.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("BleScanner");
const STALE_ENTRY_AGE = (0, import_general.Seconds)(60);
class BleScanner {
  type = import_general.ChannelType.BLE;
  #client;
  #recordWaiters = /* @__PURE__ */ new Map();
  #discoveredMatterDevices = /* @__PURE__ */ new Map();
  constructor(client) {
    this.#client = client;
    this.#client.setDiscoveryCallback(
      (peripheral, manufacturerData) => this.#handleDiscoveredDevice(peripheral, manufacturerData)
    );
  }
  getDiscoveredDevice(address) {
    const device = this.#discoveredMatterDevices.get(address);
    if (device === void 0) {
      throw new import_Ble.BleError(`No device found for address ${address}`);
    }
    return device;
  }
  /**
   * Registers a deferred promise for a specific queryId together with a timeout and return the promise.
   * The promise will be resolved when the timer runs out latest.
   */
  async #registerWaiterPromise(queryId, timeout, resolveOnUpdatedRecords = true, cancelResolver) {
    const { promise, resolver } = (0, import_general.createPromise)();
    let timer;
    if (timeout) {
      timer = import_general.Time.getTimer("BLE query timeout", timeout, () => {
        cancelResolver?.();
        this.#finishWaiter(queryId, true);
      }).start();
    }
    this.#recordWaiters.set(queryId, { resolver, timer, resolveOnUpdatedRecords, cancelResolver });
    logger.debug(
      `Registered waiter for query ${queryId} with timeout ${timeout === void 0 ? "(none)" : import_general.Duration.format(timeout)} ${resolveOnUpdatedRecords ? "" : " (not resolving on updated records)"}`
    );
    await promise;
  }
  /**
   * Remove a waiter promise for a specific queryId and stop the connected timer. If required also resolve the
   * promise.
   */
  #finishWaiter(queryId, resolvePromise, isUpdatedRecord = false) {
    const waiter = this.#recordWaiters.get(queryId);
    if (waiter === void 0) return;
    const { timer, resolver, resolveOnUpdatedRecords } = waiter;
    if (isUpdatedRecord && !resolveOnUpdatedRecords) return;
    logger.debug(`Finishing waiter for query ${queryId}, resolving: ${resolvePromise}`);
    timer?.stop();
    if (resolvePromise) {
      resolver();
    }
    this.#recordWaiters.delete(queryId);
  }
  cancelCommissionableDeviceDiscovery(identifier, resolvePromise = true) {
    const queryKey = this.#buildCommissionableQueryIdentifier(identifier);
    if (queryKey === void 0) return;
    const { cancelResolver } = this.#recordWaiters.get(queryKey) ?? {};
    cancelResolver?.();
    this.#finishWaiter(queryKey, resolvePromise);
  }
  #handleDiscoveredDevice(peripheral, manufacturerServiceData) {
    const address = peripheral.address;
    try {
      const { discriminator, vendorId, productId, hasAdditionalAdvertisementData } = import_BtpCodec.BtpCodec.decodeBleAdvertisementServiceData(manufacturerServiceData);
      const deviceData = {
        deviceIdentifier: address,
        D: discriminator,
        SD: discriminator >> 8 & 15,
        VP: `${vendorId}+${productId}`,
        CM: 1,
        // Can be no other mode,
        addresses: [{ type: "ble", peripheralAddress: address }]
      };
      const deviceExisting = this.#discoveredMatterDevices.has(address);
      const serviceDataHex = import_general.Bytes.toHex(manufacturerServiceData);
      const now = import_general.Time.nowMs;
      for (const [otherAddress, otherEntry] of this.#discoveredMatterDevices) {
        if (otherAddress === address) continue;
        if (otherEntry.serviceDataHex !== serviceDataHex) continue;
        if (now - otherEntry.lastSeen <= STALE_ENTRY_AGE) continue;
        logger.debug(
          `Dropping stale BLE entry ${otherAddress} \u2014 matching service data arrived from ${address} and prior entry is ${import_general.Duration.format((0, import_general.Millis)(now - otherEntry.lastSeen))} old`
        );
        this.#discoveredMatterDevices.delete(otherAddress);
      }
      logger.debug(
        `${deviceExisting ? "Re-" : ""}Discovered device ${address} data: ${import_general.Diagnostic.json(deviceData)}`
      );
      this.#discoveredMatterDevices.set(address, {
        deviceData,
        peripheral,
        hasAdditionalAdvertisementData,
        serviceDataHex,
        lastSeen: now
      });
      const queryKey = this.#findCommissionableQueryIdentifier(deviceData);
      if (queryKey !== void 0) {
        this.#finishWaiter(queryKey, true, deviceExisting);
      }
    } catch (error) {
      logger.debug(
        `Discovered device ${address} ${manufacturerServiceData === void 0 ? void 0 : import_general.Bytes.toHex(manufacturerServiceData)} does not seem to be a valid Matter device: ${error}`
      );
    }
  }
  #findCommissionableQueryIdentifier(record) {
    const longDiscriminatorQueryId = this.#buildCommissionableQueryIdentifier({ longDiscriminator: record.D });
    if (longDiscriminatorQueryId !== void 0 && this.#recordWaiters.has(longDiscriminatorQueryId)) {
      return longDiscriminatorQueryId;
    }
    const shortDiscriminatorQueryId = this.#buildCommissionableQueryIdentifier({ shortDiscriminator: record.SD });
    if (shortDiscriminatorQueryId !== void 0 && this.#recordWaiters.has(shortDiscriminatorQueryId)) {
      return shortDiscriminatorQueryId;
    }
    if (record.VP !== void 0) {
      const vpParts = record.VP.split("+");
      const vendorId = (0, import_types.VendorId)(parseInt(vpParts[0]));
      const productId = vpParts[1] !== void 0 ? parseInt(vpParts[1]) : void 0;
      if (productId !== void 0) {
        const vendorProductQueryId = this.#buildCommissionableQueryIdentifier({
          vendorId,
          productId
        });
        if (vendorProductQueryId !== void 0 && this.#recordWaiters.has(vendorProductQueryId)) {
          return vendorProductQueryId;
        }
      }
      const vendorIdQueryId = this.#buildCommissionableQueryIdentifier({ vendorId });
      if (vendorIdQueryId !== void 0 && this.#recordWaiters.has(vendorIdQueryId)) {
        return vendorIdQueryId;
      }
      if (productId !== void 0) {
        const productIdQueryId = this.#buildCommissionableQueryIdentifier({ productId });
        if (productIdQueryId !== void 0 && this.#recordWaiters.has(productIdQueryId)) {
          return productIdQueryId;
        }
      }
    }
    if (this.#recordWaiters.has("*")) {
      return "*";
    }
    return void 0;
  }
  /**
   * Builds an identifier string for commissionable queries based on the given identifier object.
   * Some identifiers are identical to the official DNS-SD identifiers, others are custom.
   */
  #buildCommissionableQueryIdentifier(identifier) {
    if ("instanceId" in identifier) {
      return void 0;
    } else if ("longDiscriminator" in identifier) {
      return `D:${identifier.longDiscriminator}`;
    } else if ("shortDiscriminator" in identifier) {
      return `SD:${identifier.shortDiscriminator}`;
    } else if ("vendorId" in identifier && "productId" in identifier) {
      return `VP:${identifier.vendorId}+${identifier.productId}`;
    } else if ("vendorId" in identifier) {
      return `V:${identifier.vendorId}`;
    } else if ("deviceType" in identifier) {
      return void 0;
    } else if ("productId" in identifier) {
      return `P:${identifier.productId}`;
    } else return "*";
  }
  #getCommissionableDevices(identifier) {
    const storedRecords = Array.from(this.#discoveredMatterDevices.values()).sort(
      (a, b) => b.lastSeen - a.lastSeen
    );
    const foundRecords = new Array();
    if ("instanceId" in identifier || "deviceType" in identifier) {
      return foundRecords;
    } else if ("longDiscriminator" in identifier) {
      foundRecords.push(...storedRecords.filter(({ deviceData: { D } }) => D === identifier.longDiscriminator));
    } else if ("shortDiscriminator" in identifier) {
      foundRecords.push(
        ...storedRecords.filter(({ deviceData: { SD } }) => SD === identifier.shortDiscriminator)
      );
    } else if ("vendorId" in identifier && "productId" in identifier) {
      foundRecords.push(
        ...storedRecords.filter(
          ({ deviceData: { VP } }) => VP === `${identifier.vendorId}+${identifier.productId}`
        )
      );
    } else if ("vendorId" in identifier) {
      foundRecords.push(
        ...storedRecords.filter(
          ({ deviceData: { VP } }) => VP === `${identifier.vendorId}` || VP?.startsWith(`${identifier.vendorId}+`)
        )
      );
    } else if ("productId" in identifier) {
      foundRecords.push(
        ...storedRecords.filter(({ deviceData: { VP } }) => VP?.endsWith(`+${identifier.productId}`))
      );
    } else {
      foundRecords.push(...storedRecords.filter(({ deviceData: { CM } }) => CM === 1 || CM === 2));
    }
    return foundRecords;
  }
  async findCommissionableDevices(identifier, timeout = (0, import_general.Seconds)(10), ignoreExistingRecords = false) {
    const queryKey = this.#buildCommissionableQueryIdentifier(identifier);
    if (queryKey === void 0) {
      return [];
    }
    let storedRecords = this.#getCommissionableDevices(identifier);
    if (ignoreExistingRecords) {
      for (const record of storedRecords) {
        this.#discoveredMatterDevices.delete(record.peripheral.address);
      }
      storedRecords = [];
    }
    if (storedRecords.length === 0) {
      await this.#client.startScanning();
      await this.#registerWaiterPromise(queryKey, timeout);
      storedRecords = this.#getCommissionableDevices(identifier);
      await this.#client.stopScanning();
    }
    return storedRecords.map(({ deviceData }) => deviceData);
  }
  async findCommissionableDevicesContinuously(identifier, callback, timeout, cancelSignal) {
    const queryKey = this.#buildCommissionableQueryIdentifier(identifier);
    if (queryKey === void 0) {
      return [];
    }
    const discoveredDevices = /* @__PURE__ */ new Set();
    const discoveryEndTime = timeout ? import_general.Time.nowMs + timeout : void 0;
    await this.#client.startScanning();
    let queryResolver;
    if (cancelSignal === void 0) {
      const { promise, resolver } = (0, import_general.createPromise)();
      cancelSignal = promise;
      queryResolver = resolver;
    }
    let canceled = false;
    cancelSignal?.then(
      () => {
        canceled = true;
        this.#finishWaiter(queryKey, true);
      },
      (cause) => {
        logger.error("Unexpected error canceling commissioning", cause);
      }
    );
    while (!canceled) {
      this.#getCommissionableDevices(identifier).forEach(({ deviceData }) => {
        const { deviceIdentifier } = deviceData;
        if (!discoveredDevices.has(deviceIdentifier)) {
          discoveredDevices.add(deviceIdentifier);
          callback(deviceData);
        }
      });
      let remainingTime;
      if (discoveryEndTime !== void 0) {
        remainingTime = import_general.Millis.ceil((0, import_general.Timespan)(import_general.Time.nowMs, discoveryEndTime).duration);
        if (remainingTime <= 0) {
          break;
        }
      }
      await this.#registerWaiterPromise(queryKey, remainingTime, false, queryResolver);
    }
    await this.#client.stopScanning();
    return this.#getCommissionableDevices(identifier).map(({ deviceData }) => deviceData);
  }
  getDiscoveredCommissionableDevices(identifier) {
    return this.#getCommissionableDevices(identifier).map(({ deviceData }) => deviceData);
  }
  closeClient() {
    return this.#client.stopScanning();
  }
  async close() {
    await this.closeClient();
    [...this.#recordWaiters.keys()].forEach(
      (queryId) => this.#finishWaiter(queryId, !!this.#recordWaiters.get(queryId)?.timer)
    );
  }
}
