// @matter/protocol/dist/cjs/common/FailsafeContext.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FailsafeContext_exports = {};
__export(FailsafeContext_exports, {
  FailsafeContext: () => FailsafeContext,
  MatterFabricInvalidAdminSubjectError: () => MatterFabricInvalidAdminSubjectError
});
module.exports = __toCommonJS(FailsafeContext_exports);
var import_PeerCommunicationError = require("#peer/PeerCommunicationError.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_Fabric = require("../fabric/Fabric.js");
var import_FailsafeTimer = require("./FailsafeTimer.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("FailsafeContext");
class MatterFabricInvalidAdminSubjectError extends import_general.MatterFlowError {
}
class FailsafeContext {
  #sessions;
  #fabrics;
  #failsafe;
  #construction;
  #associatedFabric;
  #csrSessionId;
  #forUpdateNoc;
  #fabricBuilder;
  #rootCertSet = false;
  #commissioned = (0, import_general.AsyncObservable)();
  constructor(options) {
    const { sessions, fabrics, expiryLength, session, maxCumulativeFailsafe } = options;
    this.#sessions = sessions;
    this.#fabrics = fabrics;
    this.#associatedFabric = session.fabric;
    this.#construction = (0, import_general.Construction)(this, async () => {
      this.#fabricBuilder = await import_Fabric.FabricBuilder.create(this.#fabrics.crypto);
      await Promise.resolve();
      await this.storeEndpointState();
      this.#failsafe = new import_FailsafeTimer.FailsafeTimer(
        this.#associatedFabric,
        expiryLength,
        maxCumulativeFailsafe,
        this.#failSafeExpired.bind(this)
      );
      logger.debug(`Arm failSafe timer for ${import_general.Duration.format(expiryLength)}`);
      session.closedByPeer.on(() => {
        if (!this.#failsafe?.completed) {
          return this.#failSafeExpired();
        }
      });
    });
  }
  async extend(fabric, expiryLength, currentExchange) {
    await this.#construction;
    await this.#failsafe?.reArm(fabric, expiryLength, currentExchange);
    if (expiryLength > 0) {
      logger.debug(`Extend failSafe timer for ${import_general.Duration.format(expiryLength)}`);
    }
  }
  get fabricIndex() {
    return this.#builder.fabricIndex;
  }
  get construction() {
    return this.#construction;
  }
  get commissioned() {
    return this.#commissioned;
  }
  get associatedFabric() {
    return this.#associatedFabric;
  }
  get csrSessionId() {
    return this.#csrSessionId;
  }
  get forUpdateNoc() {
    return this.#forUpdateNoc;
  }
  get rootCertSet() {
    return this.#rootCertSet;
  }
  get hasRootCert() {
    return this.#builder.rootCert !== void 0;
  }
  get rootCert() {
    return this.#builder.rootCert;
  }
  async completeCommission() {
    if (this.#failsafe === void 0) {
      throw new import_general.MatterFlowError("armFailSafe should be called first!");
    }
    this.#failsafe.complete();
    if (this.fabricIndex !== void 0) {
      await this.#fabrics.persistFabrics();
    }
    this.#failsafe = void 0;
    await this.commissioned.emit();
    await this.closePaseSession({ cause: new import_PeerCommunicationError.FailsafeExpiredError() });
    await this.close();
  }
  getFailSafeContext() {
    if (this.#failsafe === void 0) throw new import_general.MatterFlowError("armFailSafe should be called first!");
    return this.#failsafe;
  }
  getNextFabricIndex() {
    return this.#fabrics.allocateFabricIndex();
  }
  async replaceFabric(fabric) {
    await this.#fabrics.replaceFabric(fabric);
    await this.#sessions.deleteResumptionRecordsForFabric(fabric);
  }
  /**
   * Handles a CSR from OperationalCredentials cluster and stores additional internal information for further
   * validity checks.
   */
  createCertificateSigningRequest(isForUpdateNoc, sessionId) {
    if (this.#fabrics.findByKeypair(this.#builder.keyPair)) {
      throw new import_general.MatterFlowError("Key pair already exists.");
    }
    const result = this.#builder.createCertificateSigningRequest();
    this.#csrSessionId = sessionId;
    this.#forUpdateNoc = isForUpdateNoc;
    return result;
  }
  async closePaseSession(context) {
    const session = this.#sessions.getPaseSession();
    if (session) {
      await session.initiateForceClose(context);
    }
  }
  async close(currentExchange) {
    await this.#construction.close(async () => {
      if (this.#failsafe) {
        await this.#failsafe.close();
        this.#failsafe = void 0;
        await this.rollback(currentExchange);
      }
    });
  }
  /** Handles adding a trusted root certificate from Operational Credentials cluster. */
  async setRootCert(rootCert) {
    await this.#builder.setRootCert(rootCert);
    this.#rootCertSet = true;
  }
  /**
   * Build a new Fabric object based on an existing fabric for the "UpdateNoc" case of the Operational Credentials
   * cluster.
   */
  async buildUpdatedFabric(nocValue, icacValue) {
    if (this.associatedFabric === void 0) {
      throw new import_general.MatterFlowError("No fabric associated with failsafe context, but we prepare an Fabric update.");
    }
    this.#builder.initializeFromFabricForUpdate(this.associatedFabric);
    await this.#builder.setOperationalCert(nocValue, icacValue);
    return await this.#builder.build(this.associatedFabric.fabricIndex);
  }
  /** Build a new Fabric object for a new fabric for the "AddNoc" case of the Operational Credentials cluster. */
  async buildFabric(nocData) {
    const builder = this.#builder;
    const { nocValue, icacValue, adminVendorId, ipkValue, caseAdminSubject } = nocData;
    if (!import_types.NodeId.isOperationalNodeId(caseAdminSubject) && !import_types.NodeId.isCaseAuthenticatedTag(caseAdminSubject)) {
      try {
        if (import_types.CaseAuthenticatedTag.getVersion(import_types.NodeId.extractAsCaseAuthenticatedTag(caseAdminSubject)) === 0) {
          throw new MatterFabricInvalidAdminSubjectError();
        }
      } catch (error) {
        if (error instanceof import_types.ValidationError || error instanceof import_general.UnexpectedDataError) {
          throw new MatterFabricInvalidAdminSubjectError();
        } else {
          throw error;
        }
      }
    }
    await builder.setOperationalCert(nocValue, icacValue);
    const newGlobalId = await builder.globalId;
    if (this.#fabrics.has(newGlobalId)) {
      throw new import_FailsafeTimer.MatterFabricConflictError(
        `Fabric with Id ${builder.fabricId} and Node Id ${builder.nodeId} already exists.`
      );
    }
    this.#associatedFabric = await builder.setRootVendorId(adminVendorId).setIdentityProtectionKey(ipkValue).setRootNodeId(caseAdminSubject).build(this.#fabrics.allocateFabricIndex());
    this.#fabrics.addFabric(this.#associatedFabric);
    if (this.#failsafe) {
      this.#failsafe.associatedFabric = this.#associatedFabric;
    }
    return this.#associatedFabric;
  }
  async #failSafeExpired(currentExchange) {
    logger.info("Failsafe timer expired; resetting fabric builder");
    await this.close(currentExchange);
  }
  async rollback(currentExchange) {
    if (this.fabricIndex !== void 0 && !this.#forUpdateNoc) {
      logger.debug(`Revoking fabric index ${this.fabricIndex}`);
      await this.#associatedFabric?.delete(currentExchange);
    }
    await this.closePaseSession({ cause: new import_PeerCommunicationError.FailsafeExpiredError(), currentExchange });
    let fabric = void 0;
    if (this.fabricIndex !== void 0) {
      const fabricIndex = this.fabricIndex;
      if (this.#fabrics.has(fabricIndex)) {
        fabric = this.#fabrics.for(fabricIndex);
        for (const session of this.#sessions.sessionsForFabricIndex(fabricIndex)) {
          await session.initiateForceClose({ cause: new import_PeerCommunicationError.FabricChangedError(), currentExchange });
        }
      }
    }
    await this.restoreNetworkState();
    if (this.#forUpdateNoc && this.associatedFabric !== void 0) {
      await this.restoreFabric(this.associatedFabric);
    }
    if (!this.#forUpdateNoc && fabric !== void 0) {
      await this.#associatedFabric?.delete();
    }
    await this.restoreBreadcrumb();
  }
  async restoreFabric(fabric) {
    await this.replaceFabric(fabric);
  }
  get #builder() {
    if (this.#fabricBuilder === void 0) {
      throw new import_general.UninitializedDependencyError("FailsafeContext", "Fabric builder has not been initialized");
    }
    return this.#fabricBuilder;
  }
}
