// @matter/protocol/dist/cjs/common/Scanner.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Scanner_exports = {};
__export(Scanner_exports, {
  DiscoveryData: () => DiscoveryData,
  DiscoveryDataDiagnostics: () => DiscoveryDataDiagnostics,
  ScannerSet: () => ScannerSet
});
module.exports = __toCommonJS(Scanner_exports);
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function DiscoveryData(kvs) {
  const dd = {};
  for (const key of kvs.keys()) {
    switch (key) {
      case "VP":
      case "DN":
      case "RI":
      case "PI":
        dd[key] = `${kvs.get(key)}`;
        break;
      case "DT":
      case "PH":
      case "T":
      case "ICD": {
        const num = Number(kvs.get(key));
        if (isFinite(num)) {
          dd[key] = num;
        }
        break;
      }
      case "SII":
      case "SAI":
      case "SAT": {
        const num = Number(kvs.get(key));
        if (isFinite(num)) {
          dd[key] = (0, import_general.Millis)(num);
        }
        break;
      }
    }
  }
  return dd;
}
function DiscoveryDataDiagnostics(data, kind) {
  return import_general.Diagnostic.dict({
    kind,
    DN: data.DN,
    SII: data.SII !== void 0 ? import_general.Duration.format(data.SII) : void 0,
    SAI: data.SAI !== void 0 ? import_general.Duration.format(data.SAI) : void 0,
    SAT: data.SAT !== void 0 ? import_general.Duration.format(data.SAT) : void 0,
    T: data.T,
    DT: data.DT,
    PH: data.PH,
    ICD: data.ICD,
    VP: data.VP,
    RI: data.RI,
    PI: data.PI
  });
}
class ScannerSet extends import_general.BasicSet {
  scannerFor(type) {
    return this.find((scanner) => scanner.type === type);
  }
  hasScannerFor(type) {
    return this.scannerFor(type) !== void 0;
  }
  /**
   * Select a set of scanners based on discovery capabilities.
   */
  select(discoveryCapabilities) {
    return this.filter(
      (scanner) => scanner.type === import_general.ChannelType.UDP || discoveryCapabilities?.ble && scanner.type === import_general.ChannelType.BLE
    );
  }
  static [import_general.Environmental.create](env) {
    const instance = new ScannerSet();
    env.set(ScannerSet, instance);
    return instance;
  }
}
