// @matter/protocol/dist/cjs/bdx/BdxMessenger.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BdxMessenger_exports = {};
__export(BdxMessenger_exports, {
  BDX_TRANSFER_IDLE_TIMEOUT: () => BDX_TRANSFER_IDLE_TIMEOUT,
  BdxMessenger: () => BdxMessenger
});
module.exports = __toCommonJS(BdxMessenger_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_BdxError = require("./BdxError.js");
var import_BdxMessage = require("./schema/BdxMessage.js");
var import_BdxStatusMessageSchema = require("./schema/BdxStatusMessageSchema.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("BdxMessenger");
const BDX_TRANSFER_IDLE_TIMEOUT = (0, import_general.Minutes)(5);
class BdxMessenger {
  #exchange;
  #messageTimeout;
  /** Creates a new BdxMessenger instance by initiating a new MessageExchange from the given ExchangeProvider. */
  static async create(exchangeProvider, messageTimeout = BDX_TRANSFER_IDLE_TIMEOUT) {
    const exchange = await exchangeProvider.initiateExchange({ protocol: import_types.BDX_PROTOCOL_ID });
    return new this(exchange, messageTimeout);
  }
  /**
   * Creates a new BdxMessenger instance.
   * @param exchange Exchange to use for the messaging
   * @param messageTimeout Communication Timeout for the Bdx Messages, defaults to 5 minutes as defined for Matter OTA transfers
   */
  constructor(exchange, messageTimeout = BDX_TRANSFER_IDLE_TIMEOUT) {
    if (!exchange.session.isSecure) {
      throw new import_general.ImplementationError("Bdx Protocol requires a secure session");
    }
    this.#messageTimeout = messageTimeout;
    this.#exchange = exchange;
  }
  get channel() {
    return this.#exchange.channel;
  }
  get exchange() {
    return this.#exchange;
  }
  get maxPayloadSize() {
    return this.#exchange.maxPayloadSize;
  }
  /**
   * Waits for the next message and returns it.
   * A List of allowed expected message types can be provided.
   * If the message type is not in the list, an error will be thrown.
   */
  async nextMessage(expectedMessageTypes, timeout = this.#messageTimeout, expectedMessageInfo) {
    logger.debug(
      `Waiting for Bdx ${expectedMessageTypes.map((t) => import_types.BdxMessageType[t]).join("/")} message with timeout ${import_general.Duration.format(timeout)}`
    );
    const message = await this.exchange.nextMessage({ timeout });
    const messageType = message.payloadHeader.messageType;
    if (expectedMessageInfo === void 0) {
      expectedMessageInfo = expectedMessageTypes.map((t) => `${import_types.BdxMessageType[t]}#${t}`).join(",");
    }
    this.throwIfErrorStatusReport(message, expectedMessageInfo);
    if (!expectedMessageTypes.includes(messageType))
      throw new import_general.UnexpectedDataError(
        `Received unexpected message type: ${import_types.BdxMessageType[messageType] ?? "unknown"}#${messageType}, expected: ${expectedMessageInfo}`
      );
    return import_BdxMessage.BdxMessage.decode(messageType, message.payload);
  }
  async send(bdxMessage) {
    await this.exchange.send(bdxMessage.kind, import_BdxMessage.BdxMessage.encode(bdxMessage), {
      expectedProcessingTime: this.#messageTimeout
    });
  }
  /** Sends a Bdx SendInit message and waits for the SendAccept message as a response and returns it decoded. */
  async sendSendInit(message) {
    await this.send({ kind: import_types.BdxMessageType.SendInit, message });
    const response = await this.nextMessage([import_types.BdxMessageType.SendAccept]);
    import_BdxMessage.BdxMessage.assert(response, import_types.BdxMessageType.SendAccept);
    return response.message;
  }
  /** Sends a ReceiveInit message and waits for the ReceiveAccept message as a response and returns it decoded. */
  async sendReceiveInit(message) {
    await this.send({ kind: import_types.BdxMessageType.ReceiveInit, message });
    const response = await this.nextMessage([import_types.BdxMessageType.ReceiveAccept]);
    import_BdxMessage.BdxMessage.assert(response, import_types.BdxMessageType.ReceiveAccept);
    return response.message;
  }
  /** Encodes and sends a Bdx SendAccept message. */
  async sendSendAccept(message) {
    await this.send({ kind: import_types.BdxMessageType.SendAccept, message });
  }
  /** Encodes and sends a Bdx ReceiveAccept message. */
  async sendReceiveAccept(message) {
    await this.send({ kind: import_types.BdxMessageType.ReceiveAccept, message });
  }
  /** Encodes and sends a Bdx Block message. */
  async sendBlock(message) {
    await this.send({ kind: import_types.BdxMessageType.Block, message });
  }
  /** Encodes and sends a Bdx BlockQuery message. */
  async sendBlockQuery(message) {
    await this.send({ kind: import_types.BdxMessageType.BlockQuery, message });
  }
  /** Encodes and sends a Bdx BlockQueryWithSkip message. */
  async sendBlockQueryWithSkip(message) {
    await this.send({ kind: import_types.BdxMessageType.BlockQueryWithSkip, message });
  }
  /** Encodes and sends a Bdx BlockEof message. */
  async sendBlockEof(message) {
    await this.send({ kind: import_types.BdxMessageType.BlockEof, message });
  }
  /** Encodes and sends a Bdx BlockAck message. */
  async sendBlockAck(message) {
    await this.send({ kind: import_types.BdxMessageType.BlockAck, message });
  }
  /** Encodes and sends a Bdx BlockAckEof message */
  async sendBlockAckEof(message) {
    await this.send({ kind: import_types.BdxMessageType.BlockAckEof, message });
  }
  /** Read the next Block message, accepts Block and BlockEof messages. Returns the decoded message and it's type. */
  async readBlock() {
    const block = await this.nextMessage([import_types.BdxMessageType.Block, import_types.BdxMessageType.BlockEof]);
    if (import_BdxMessage.BdxMessage.is(block, import_types.BdxMessageType.Block) && block.message.data.byteLength === 0) {
      throw new import_BdxError.BdxError("Received empty data in Block message", import_types.BdxStatusCode.BadMessageContent);
    }
    return block;
  }
  /**
   * Read the next BlockQuery message, accepts BlockQuery and BlockQueryWithSkip and BlockAck messages.
   * When a BlockAck is received, it will be validated and the next BlockQuery message will be read.
   * Returns the decoded message and it's type.
   */
  async readBlockQuery() {
    let response = await this.nextMessage([
      import_types.BdxMessageType.BlockQuery,
      import_types.BdxMessageType.BlockQueryWithSkip,
      import_types.BdxMessageType.BlockAck
    ]);
    let expectedBlockMessageCounter = void 0;
    if (import_BdxMessage.BdxMessage.is(response, import_types.BdxMessageType.BlockAck)) {
      expectedBlockMessageCounter = (response.message.blockCounter + 1) % 4294967296;
      response = await this.nextMessage([import_types.BdxMessageType.BlockQuery, import_types.BdxMessageType.BlockQueryWithSkip]);
    }
    if (expectedBlockMessageCounter !== void 0 && response.message.blockCounter !== expectedBlockMessageCounter) {
      throw new import_BdxError.BdxError(
        `Received BlockQuery with unexpected block counter: ${response.message.blockCounter}, expected: ${expectedBlockMessageCounter}`,
        import_types.BdxStatusCode.BadBlockCounter
      );
    }
    return response;
  }
  /** Reads the next BlockAckEof message and returns the decoded message. */
  async readBlockAckEof() {
    const response = await this.nextMessage([import_types.BdxMessageType.BlockAckEof]);
    import_BdxMessage.BdxMessage.assert(response, import_types.BdxMessageType.BlockAckEof);
    return response.message;
  }
  /** Reads the next BlockAck message and returns the decoded message. */
  async readBlockAck() {
    const response = await this.nextMessage([import_types.BdxMessageType.BlockAck]);
    import_BdxMessage.BdxMessage.assert(response, import_types.BdxMessageType.BlockAck);
    return response.message;
  }
  /** Sends a Bdx Error StatusReport message with the given protocol status. */
  sendError(code) {
    return this.#sendStatusReport(import_types.GeneralStatusCode.Failure, code);
  }
  /** Encodes and sends a Bdx StatusReport message with the given general and protocol status. */
  async #sendStatusReport(generalStatus, protocolStatus, requiresAck) {
    await this.#exchange.send(
      import_types.SecureMessageType.StatusReport,
      import_BdxStatusMessageSchema.BdxStatusMessage.encode({
        generalStatus,
        protocolStatus
      }),
      {
        requiresAck,
        logContext: {
          generalStatus: import_types.GeneralStatusCode[generalStatus] ?? import_general.Diagnostic.hex(generalStatus),
          protocolStatus: import_types.BdxStatusCode[protocolStatus] ?? import_general.Diagnostic.hex(protocolStatus)
        }
      }
    );
  }
  throwIfErrorStatusReport(message, logHint) {
    const {
      payloadHeader: { messageType },
      payload
    } = message;
    if (messageType !== import_types.SecureMessageType.StatusReport) return;
    const { generalStatus, protocolId, protocolStatus } = import_BdxStatusMessageSchema.BdxStatusMessage.decode(payload);
    if (generalStatus !== import_types.GeneralStatusCode.Success) {
      throw new import_BdxError.BdxStatusResponseError(
        `Received general error status for protocol ${protocolId}${logHint ? ` (${logHint})` : ""}`,
        generalStatus,
        protocolStatus
      );
    }
    if (protocolStatus !== import_types.BdxStatusCode.Success) {
      throw new import_BdxError.BdxStatusResponseError(
        `Received general success status, but protocol status is not Success${logHint ? ` (${logHint})` : ""}`,
        generalStatus,
        protocolStatus
      );
    }
  }
  close() {
    return this.#exchange.close();
  }
  [Symbol.asyncDispose]() {
    return this.close();
  }
  /**
   * Ensure that the value is a safe JavaScript "number" type and that it is not too large. Matter spec allows also
   * 64bit values, but they make little sense for now, so make sure we handle them as too large. MAX_SAFE_INTEGER is
   * 2^53-1 and is enough for now.
   */
  static asSafeNumber(value, context = "", bdxErrorCode = import_types.BdxStatusCode.Unknown) {
    if (typeof value !== "number" && typeof value !== "bigint") {
      throw new import_general.InternalError(`${context} ${value} is not a number`);
    }
    if (value > Number.MAX_SAFE_INTEGER) {
      throw new import_BdxError.BdxError(`${context} ${value} exceeds maximum safe integer value`, bdxErrorCode);
    }
    return Number(value);
  }
}
