// @matter/protocol/dist/cjs/bdx/BdxSessionConfiguration.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BdxSessionConfiguration_exports = {};
__export(BdxSessionConfiguration_exports, {
  BdxSessionConfiguration: () => BdxSessionConfiguration
});
module.exports = __toCommonJS(BdxSessionConfiguration_exports);
var import_general = require("@matter/general");
var import_Flow = require("./flow/Flow.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class BdxSessionConfiguration {
  #isSender;
  #initMessage;
  #transferConfig;
  #fileDesignator;
  #transferFileDesignator;
  constructor(options) {
    const { isSender, fileDesignator, transferFileDesignator, initMessage } = options;
    this.#fileDesignator = fileDesignator;
    this.#transferFileDesignator = transferFileDesignator ?? fileDesignator;
    this.#isSender = isSender;
    this.#initMessage = initMessage;
    this.#transferConfig = { ...BdxSessionConfiguration.DefaultConfig, ...options };
    const {
      preferredDriverModes = [],
      maxBlockSize,
      maxTransferSize,
      asynchronousTransferAllowed,
      senderStartOffset,
      senderMaxLength
    } = this.#transferConfig;
    if (preferredDriverModes.length === 0) {
      throw new import_general.ImplementationError("At least one preferred driver mode must be set.");
    }
    if (maxBlockSize !== void 0 && maxBlockSize <= 0) {
      throw new import_general.ImplementationError("Max block size must be greater than 0");
    }
    if (maxTransferSize !== void 0 && maxTransferSize <= 0) {
      throw new import_general.ImplementationError("Max transfer size must be greater than 0");
    }
    if (asynchronousTransferAllowed) {
      throw new import_general.ImplementationError("Asynchronous transfer is not supported");
    }
    if (!isSender && (senderStartOffset !== void 0 || senderMaxLength !== void 0)) {
      throw new import_general.ImplementationError("Sender start offset and sender max length are only supported for senders");
    }
  }
  get isSender() {
    return this.#isSender;
  }
  get initMessage() {
    return this.#initMessage;
  }
  get isInitiator() {
    return this.#initMessage === void 0;
  }
  get fileDesignator() {
    return this.#fileDesignator;
  }
  get transferFileDesignator() {
    return this.#transferFileDesignator;
  }
  get transferConfig() {
    return this.#transferConfig;
  }
}
((BdxSessionConfiguration2) => {
  BdxSessionConfiguration2.DefaultConfig = {
    preferredDriverModes: [
      import_Flow.Flow.DriverMode.SenderDrive,
      // Default if multiple is supported, so lets use this
      import_Flow.Flow.DriverMode.ReceiverDrive
    ],
    asynchronousTransferAllowed: false,
    // Provisional, not supported
    maxTransferSize: 1024 * 1e5
    // 100 MB, lets use that as maximum transfer filesize for now just to protect
  };
})(BdxSessionConfiguration || (BdxSessionConfiguration = {}));
