// @matter/protocol/dist/cjs/bdx/PersistedFileDesignator.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PersistedFileDesignator_exports = {};
__export(PersistedFileDesignator_exports, {
  PersistedFileDesignator: () => PersistedFileDesignator
});
module.exports = __toCommonJS(PersistedFileDesignator_exports);
var import_ScopedStorage = require("#bdx/ScopedStorage.js");
var import_general = require("@matter/general");
var import_FileDesignator = require("./FileDesignator.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("PersistedFileDesignator");
class PersistedFileDesignator extends import_FileDesignator.FileDesignator {
  #blobDriver;
  #contexts;
  #blob;
  #blobName;
  constructor(fd, storage, contexts) {
    if (fd instanceof import_FileDesignator.FileDesignator) {
      fd = fd.bytes;
    }
    const blobName = import_general.Bytes.isBytes(fd) ? import_general.Bytes.toString(fd) : fd;
    let blobDriver;
    let baseContexts;
    if (storage instanceof import_ScopedStorage.ScopedStorage) {
      fd = import_general.Bytes.fromString(`${storage.scope}/${blobName}`);
      blobDriver = storage.blobDriver;
      baseContexts = storage.baseContexts;
    } else {
      blobDriver = storage;
      baseContexts = contexts ?? [];
    }
    super(fd);
    const subContext = blobName.split(".");
    this.#blobName = subContext.pop();
    this.#contexts = [...baseContexts, ...subContext];
    this.#blobDriver = blobDriver;
  }
  exists() {
    return this.#blobDriver.has(this.#contexts, this.#blobName);
  }
  get blobName() {
    return this.#blobName;
  }
  async openBlob() {
    if (this.#blob === void 0) {
      if (!await this.#blobDriver.has(this.#contexts, this.#blobName)) {
        throw new import_general.ImplementationError(
          `File designator must point to an existing blob "${this.#blobName}" in the storage to send data`
        );
      }
      this.#blob = await this.#blobDriver.openBlob(this.#contexts, this.#blobName);
    }
    return this.#blob;
  }
  writeFromStream(stream) {
    logger.debug(`Writing blob "${this.text}" (${this.#blobName}) to storage`);
    return this.#blobDriver.writeBlobFromStream(this.#contexts, this.#blobName, stream);
  }
  delete() {
    logger.debug(`Deleting blob "${this.text}" (${this.#blobName}) from storage`);
    return this.#blobDriver.delete(this.#contexts, this.#blobName);
  }
}
