// @matter/protocol/dist/cjs/ble/Ble.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Ble_exports = {};
__export(Ble_exports, {
  Ble: () => Ble,
  BleChannel: () => BleChannel,
  BleChannelClosedError: () => BleChannelClosedError,
  BleDisconnectedError: () => BleDisconnectedError,
  BleError: () => BleError
});
module.exports = __toCommonJS(Ble_exports);
var import_general = require("@matter/general");
var import_PeerCommunicationError = require("../peer/PeerCommunicationError.js");
var import_BleConsts = require("./BleConsts.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class BleError extends import_PeerCommunicationError.PeerCommunicationError {
}
class BleDisconnectedError extends BleError {
}
class BleChannelClosedError extends BleDisconnectedError {
}
class Ble {
}
class BleChannel {
  maxPayloadSize = import_BleConsts.MatterBle.MAX_MATTER_PAYLOAD_SIZE;
  isReliable = true;
  supportsLargeMessages = false;
  type = import_general.ChannelType.BLE;
  #closed = (0, import_general.Observable)();
  #closedFired = false;
  /**
   * Emitted exactly once when the channel is lost (peripheral disconnect, BTP session close,
   * or explicit {@link close}).  Consumers that hold a secure session over this channel use
   * this to force-close the session and reject pending exchanges without waiting for MRP
   * timeouts.
   *
   * Subclasses must call {@link emitClosed} from each of their termination paths.  The
   * latch guarantees exactly-once semantics regardless of which triggers fire.
   */
  get closed() {
    return this.#closed;
  }
  emitClosed() {
    if (this.#closedFired) {
      return;
    }
    this.#closedFired = true;
    this.#closed.emit();
  }
}
