// @matter/node/dist/cjs/behavior/Events.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Events_exports = {};
__export(Events_exports, {
  ElementEvent: () => ElementEvent,
  Events: () => Events,
  OfflineEvent: () => OfflineEvent,
  OnlineEvent: () => OnlineEvent,
  QuietEvent: () => QuietEvent
});
module.exports = __toCommonJS(Events_exports);
var import_ChangeNotificationService = require("#node/integration/ChangeNotificationService.js");
var import_general = require("@matter/general");
var import_model = require("@matter/model");
var import_protocol = require("@matter/protocol");
var import_types = require("@matter/types");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("Events");
class Events extends import_general.EventEmitter {
  #endpoint;
  #behavior;
  #changes;
  setContext(endpoint, behavior) {
    this.#endpoint = endpoint;
    this.#behavior = behavior;
  }
  static {
    for (const name of ["interactionBegin", "interactionEnd", "stateChanged"]) {
      Object.defineProperty(this.prototype, name, {
        get() {
          if (this.hasEvent(name, true)) {
            return this.getEvent(name);
          }
          const event = (0, import_general.Observable)();
          this.addEvent(name, event);
          return event;
        },
        enumerable: true
      });
    }
  }
  get endpoint() {
    return this.#endpoint;
  }
  get behavior() {
    return this.#behavior;
  }
  get changes() {
    if (this.#changes === void 0 && this.#endpoint !== void 0 && this.#endpoint.env.has(import_ChangeNotificationService.ChangeNotificationService)) {
      this.#changes = this.#endpoint.env.get(import_ChangeNotificationService.ChangeNotificationService);
    }
    return this.#changes;
  }
  toString() {
    return `${this.#endpoint ?? "?"}.${this.#behavior?.id ?? "?"}.events`;
  }
}
class ElementEvent extends import_general.BasicObservable {
  #schema;
  #owner;
  constructor(schema, owner) {
    super();
    this.#schema = schema;
    this.#owner = owner;
  }
  /**
   * The element that triggers this event.
   */
  get schema() {
    return this.#schema;
  }
  get owner() {
    return this.#owner;
  }
}
class OfflineEvent extends ElementEvent {
  constructor(schema, owner) {
    super(schema, owner);
  }
}
class OnlineEvent extends ElementEvent {
  isQuieter = false;
  #baseOccurrence;
  #occurrenceTrigger;
  handleError(error, observer) {
    logger.error(`Error in ${descriptionOf(this, observer)}`, error);
  }
  constructor(schema, owner) {
    super(schema, owner);
    const eventSchema = this.schema;
    if (this.schema.tag === import_model.ElementTag.Event && this.schema.id !== void 0 && this.owner.endpoint !== void 0 && this.owner.behavior !== void 0 && eventSchema.priority !== void 0) {
      this.#baseOccurrence = {
        eventId: (0, import_types.EventId)(this.schema.id),
        clusterId: (0, import_types.ClusterId)(this.owner.behavior.schema.id),
        endpointId: this.owner.endpoint.number,
        priority: import_model.EventElement.PriorityId[eventSchema.priority]
      };
      this.#connectWithOccurenceManager();
    }
  }
  #connectWithOccurenceManager() {
    if (this.owner.endpoint === void 0) {
      throw new import_general.InternalError("Events initialized with no assigned endpoint");
    }
    const occurrenceManager = this.owner.endpoint.env.maybeGet(import_protocol.OccurrenceManager);
    if (occurrenceManager === void 0) {
      return;
    }
    const trigger = (payload) => {
      const occurrence = occurrenceManager.add({
        ...this.#baseOccurrence,
        epochTimestamp: import_general.Time.nowMs,
        payload
      });
      if (import_general.MaybePromise.is(occurrence)) {
        this.owner.endpoint.env.runtime.add(occurrence.then(this.#broadcast.bind(this)));
      } else {
        this.#broadcast(occurrence);
      }
    };
    this.online.on(trigger);
    this.#occurrenceTrigger = trigger;
  }
  #broadcast({ number, epochTimestamp: timestamp, priority, payload }) {
    this.owner.changes?.broadcastEvent(this.owner.endpoint, this.owner.behavior, this.schema, {
      number,
      timestamp,
      priority,
      payload
    });
  }
  /**
   * An {@link Observable} that emits only those events conveyed between nodes.
   *
   * Normally this is the {@link OnlineEvent}, but in the case of server-side elements that are
   * {@link Quality.quieter} this is {@link quiet}.
   */
  get online() {
    return this;
  }
  /**
   * A quieter version of the {@link OnlineEvent}.
   *
   * Throws if the node is not a server or the associated element is not {@link Quality.quieter}.
   *
   * By default emits latest changes once per second but you can reconfigure via {@link QuietObservable} properties
   * and/or trigger emits using {@link QuietObservable.emitNow} and {@link QuietObservable.emitSoon}.
   */
  get quiet() {
    throw new import_general.ImplementationError(`Matter does not define ${this} with "quieter" (Q) quality`);
  }
  toString() {
    const base = `${this.owner.toString()}.${this.schema.propertyName}`;
    if (this.schema.tag === import_model.ElementTag.Attribute || this.schema.tag === import_model.ElementTag.Field) {
      return `${base}$Changed`;
    }
    return base;
  }
  [Symbol.dispose]() {
    if (this.#occurrenceTrigger) {
      this.online.off(this.#occurrenceTrigger);
      this.#occurrenceTrigger = void 0;
    }
  }
}
class QuietEvent extends OnlineEvent {
  isQuieter = true;
  #quiet;
  constructor(schema, owner, config) {
    super(schema, owner);
    this.#quiet = new import_general.QuietObservable({
      shouldEmit(...args) {
        const [newValue, oldValue] = args;
        return oldValue === null || newValue === null && oldValue !== newValue ? "now" : true;
      },
      ...config,
      source: this
    });
  }
  get online() {
    return this.#quiet;
  }
  get quiet() {
    return this.#quiet;
  }
  [Symbol.dispose]() {
    this.#quiet[Symbol.dispose]();
    super[Symbol.dispose]();
  }
}
function descriptionOf(observable, observer) {
  let desc = `${observable} observer`;
  if (observer.name) {
    desc = `${desc} ${observer.name}`;
  }
  return desc;
}
