// @matter/protocol/dist/cjs/bdx/schema/BdxBlockMessagesSchema.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BdxBlockMessagesSchema_exports = {};
__export(BdxBlockMessagesSchema_exports, {
  BdxBlockAckEofMessage: () => BdxBlockAckEofMessage,
  BdxBlockAckMessage: () => BdxBlockAckMessage,
  BdxBlockEofMessage: () => BdxBlockEofMessage,
  BdxBlockMessage: () => BdxBlockMessage,
  BdxBlockMessageSchema: () => BdxBlockMessageSchema,
  BdxBlockQueryMessage: () => BdxBlockQueryMessage,
  BdxBlockQueryWithSkipMessage: () => BdxBlockQueryWithSkipMessage,
  BdxBlockQueryWithSkipMessageSchema: () => BdxBlockQueryWithSkipMessageSchema,
  BdxCounterOnlyMessageSchema: () => BdxCounterOnlyMessageSchema
});
module.exports = __toCommonJS(BdxBlockMessagesSchema_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class BdxCounterOnlyMessageSchema extends import_types.Schema {
  encodeInternal(message) {
    const { blockCounter } = message;
    const writer = new import_general.DataWriter(import_general.Endian.Little);
    writer.writeUInt32(blockCounter);
    return writer.toByteArray();
  }
  decodeInternal(bytes) {
    const reader = new import_general.DataReader(bytes, import_general.Endian.Little);
    return {
      blockCounter: reader.readUInt32()
    };
  }
}
const BdxCounterOnlyMessage = new BdxCounterOnlyMessageSchema();
const BdxBlockQueryMessage = BdxCounterOnlyMessage;
const BdxBlockAckMessage = BdxCounterOnlyMessage;
const BdxBlockAckEofMessage = BdxCounterOnlyMessage;
class BdxBlockQueryWithSkipMessageSchema extends import_types.Schema {
  encodeInternal(message) {
    const { blockCounter, bytesToSkip } = message;
    const writer = new import_general.DataWriter(import_general.Endian.Little);
    writer.writeUInt32(blockCounter);
    writer.writeUInt64(bytesToSkip);
    return writer.toByteArray();
  }
  decodeInternal(bytes) {
    const reader = new import_general.DataReader(bytes, import_general.Endian.Little);
    return {
      blockCounter: reader.readUInt32(),
      bytesToSkip: reader.readUInt64()
    };
  }
}
const BdxBlockQueryWithSkipMessage = new BdxBlockQueryWithSkipMessageSchema();
class BdxBlockMessageSchema extends import_types.Schema {
  encodeInternal(message) {
    const { blockCounter, data } = message;
    const writer = new import_general.DataWriter(import_general.Endian.Little);
    writer.writeUInt32(blockCounter);
    writer.writeByteArray(data);
    return writer.toByteArray();
  }
  decodeInternal(bytes) {
    const reader = new import_general.DataReader(bytes, import_general.Endian.Little);
    return {
      blockCounter: reader.readUInt32(),
      data: reader.remainingBytes
    };
  }
}
const BdxBlockMessage = new BdxBlockMessageSchema();
const BdxBlockEofMessage = BdxBlockMessage;
