// @matter/protocol/dist/cjs/fabric/FabricManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FabricManager_exports = {};
__export(FabricManager_exports, {
  FabricManager: () => FabricManager,
  FabricNotFoundError: () => FabricNotFoundError,
  FabricTableFullError: () => FabricTableFullError
});
module.exports = __toCommonJS(FabricManager_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_Fabric = require("./Fabric.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("FabricManager");
class FabricNotFoundError extends import_general.MatterError {
}
class FabricTableFullError extends import_general.MatterError {
}
class FabricManager {
  #crypto;
  #nextFabricIndex = 1;
  #fabrics = new import_general.BasicSet();
  #initializationDone = false;
  #storage;
  #events = {
    added: (0, import_general.Observable)(),
    replaced: (0, import_general.Observable)(unhandled("replacing")),
    leaving: (0, import_general.Observable)(unhandled("leaving")),
    deleting: (0, import_general.AsyncObservable)(unhandled("deleting")),
    deleted: (0, import_general.AsyncObservable)(unhandled("deleted")),
    failsafeClosed: (0, import_general.Observable)()
  };
  #construction;
  constructor(crypto, storage) {
    this.#crypto = crypto;
    this.#storage = storage;
    let construct;
    if (this.#storage === void 0) {
      construct = () => {
      };
    } else {
      construct = async () => {
        if (this.#storage === void 0) {
          return;
        }
        const fabrics = await this.#storage.get("fabrics", []);
        for (const fabricConfig of fabrics) {
          this.#addNewFabric(await import_Fabric.Fabric.create(crypto, fabricConfig));
        }
        this.#nextFabricIndex = await this.#storage.get("nextFabricIndex", this.#nextFabricIndex);
        this.#initializationDone = true;
      };
    }
    this.#construction = (0, import_general.Construction)(this, construct);
  }
  get crypto() {
    return this.#crypto;
  }
  get construction() {
    return this.#construction;
  }
  async [import_general.Construction.construct]() {
    await this.construction;
  }
  static [import_general.Environmental.create](env) {
    const instance = new FabricManager(env.get(import_general.Crypto), env.get(import_general.StorageManager).createContext("fabrics"));
    env.set(FabricManager, instance);
    return instance;
  }
  get events() {
    return this.#events;
  }
  async clear() {
    await this.#construction;
    this.#nextFabricIndex = 1;
    this.#fabrics.clear();
    await this.#storage?.clearAll();
  }
  /**
   * Test whether the fabric identified by {@link identifier} is present.
   */
  has(address) {
    let fabric;
    if (typeof address === "bigint") {
      fabric = this.#fabrics.get("globalId", address);
    } else {
      if (typeof address === "object") {
        address = address.fabricIndex;
      }
      fabric = this.#fabrics.get("fabricIndex", address);
    }
    return fabric !== void 0 && !fabric.isDeleting;
  }
  /**
   * Obtain the fabric identified by {@link identifier}.
   *
   * Throws if the fabric is not found.
   */
  for(identifier) {
    let fabric;
    if (typeof identifier === "bigint") {
      fabric = this.#fabrics.get("globalId", identifier);
    } else {
      if (typeof identifier === "object") {
        identifier = identifier.fabricIndex;
      }
      fabric = this.#fabrics.get("fabricIndex", identifier);
    }
    if (fabric === void 0 || fabric.isDeleting) {
      const str = typeof identifier === "bigint" ? import_types.GlobalFabricId.strOf(identifier) : `#${identifier}`;
      throw new FabricNotFoundError(
        `Fabric index ${str} ${fabric === void 0 ? "does not exist" : "is deleted"}`
      );
    }
    return fabric;
  }
  /**
   * Get the fabric identified by {@link identifier} if present.
   */
  maybeFor(identifier) {
    if (this.has(identifier)) {
      return this.for(identifier);
    }
  }
  allocateFabricIndex() {
    this.#construction.assert();
    for (let i = 0; i < 254; i++) {
      const fabricIndex = this.#nextFabricIndex++;
      if (this.#nextFabricIndex > 254) this.#nextFabricIndex = 1;
      if (!this.#fabrics.get("fabricIndex", (0, import_types.FabricIndex)(fabricIndex))) {
        return (0, import_types.FabricIndex)(fabricIndex);
      }
    }
    throw new FabricTableFullError("No free fabric index available.");
  }
  persistFabrics() {
    if (this.#storage === void 0) {
      throw new import_general.ImplementationError(
        "Fabric persistence is disabled because FabricManager constructed without storage"
      );
    }
    this.#construction.assert();
    const storeResult = this.#storage.set(
      "fabrics",
      this.fabrics.map((fabric) => fabric.config)
    );
    if (import_general.MaybePromise.is(storeResult)) {
      return storeResult.then(() => this.#storage.set("nextFabricIndex", this.#nextFabricIndex));
    }
    return this.#storage.set("nextFabricIndex", this.#nextFabricIndex);
  }
  addFabric(fabric) {
    this.#construction.assert();
    this.#addNewFabric(fabric);
  }
  #addNewFabric(fabric) {
    const { fabricIndex } = fabric;
    if (this.#fabrics.get("fabricIndex", fabricIndex)) {
      throw new import_general.MatterFlowError(`Fabric with index ${fabricIndex} already exists.`);
    }
    this.#addOrUpdateFabricEntry(fabric);
    if (this.#initializationDone) {
      this.#events.added.emit(fabric);
    }
  }
  /** Insert Fabric into the manager without emitting events */
  #addOrUpdateFabricEntry(fabric) {
    const existing = this.#fabrics.get("fabricIndex", fabric.fabricIndex);
    if (existing) {
      this.#fabrics.delete(existing);
    }
    this.#fabrics.add(fabric);
    fabric.leaving.on(() => this.events.leaving.emit(fabric));
    fabric.deleting.on(() => this.events.deleting.emit(fabric));
    fabric.deleted.on(() => this.#handleFabricDeleted(fabric));
    fabric.persistCallback = (isUpdate = true) => {
      if (!this.#storage) {
        if (isUpdate) {
          logger.warn(`Fabric ${fabric.fabricIndex} cannot persist because FabricManager has no storage`);
        }
        return;
      }
      const persistResult = this.persistFabrics();
      return import_general.MaybePromise.then(persistResult, () => {
        if (isUpdate) {
          this.#events.replaced.emit(fabric);
        }
      });
    };
    if (this.#storage !== void 0 && fabric.storage === void 0) {
      fabric.storage = this.#storage.createContext(`fabric-${fabric.fabricIndex}`);
    }
  }
  async #handleFabricDeleted(fabric) {
    await this.#construction;
    this.#fabrics.delete(fabric);
    if (this.#storage) {
      await this.persistFabrics();
    }
    await fabric.storage?.clearAll();
  }
  [Symbol.iterator]() {
    this.#construction.assert();
    return this.fabrics[Symbol.iterator]();
  }
  get fabrics() {
    this.#construction.assert();
    return this.#fabrics.filter((fabric) => !fabric.isDeleting);
  }
  get length() {
    return this.fabrics.length;
  }
  find(predicate) {
    return this.fabrics.find(predicate);
  }
  map(translator) {
    return this.fabrics.map(translator);
  }
  async findFabricFromDestinationId(destinationId, initiatorRandom) {
    this.#construction.assert();
    for (const fabric of this.#fabrics) {
      const candidateDestinationIds = await fabric.destinationIdsFor(fabric.nodeId, initiatorRandom);
      if (candidateDestinationIds.some((candidate) => import_general.Bytes.areEqual(candidate, destinationId))) {
        if (fabric.isDeleting) {
          throw new FabricNotFoundError("Fabric is deleting for CASE sigma2");
        }
        return fabric;
      }
    }
    const fabrics = this.#fabrics.map(
      (fabric) => `#${fabric.fabricIndex} (node ID ${import_types.NodeId.strOf(fabric.nodeId)}) keys ${fabric.groups.keySets.allKeysForId(0).map(({ key }) => import_general.Bytes.toHex(key)).join(" & ")}`
    );
    logger.debug(
      `No match for destination ID`,
      import_general.Diagnostic.dict({ destId: destinationId, random: initiatorRandom, ...fabrics })
    );
    throw new FabricNotFoundError("Fabric not found for CASE sigma2");
  }
  findByKeypair(keypair) {
    this.#construction.assert();
    for (const fabric of this.fabrics) {
      if (fabric.matchesKeyPair(keypair)) {
        return fabric;
      }
    }
    return void 0;
  }
  forDescriptor(descriptor) {
    this.#construction.assert();
    return this.find(
      (fabric) => import_general.Bytes.areEqual(fabric.rootPublicKey, descriptor.rootPublicKey) && fabric.fabricId === descriptor.fabricId
    );
  }
  async replaceFabric(fabric) {
    await this.#construction;
    const { fabricIndex } = fabric;
    const existingFabric = this.for(fabricIndex);
    if (existingFabric === void 0) {
      throw new FabricNotFoundError(
        `Fabric with index ${fabricIndex} cannot be replaced because it does not exist.`
      );
    }
    if (existingFabric === fabric) {
      return;
    }
    this.#addOrUpdateFabricEntry(fabric);
    if (this.#storage) {
      await this.persistFabrics();
    }
    this.#events.replaced.emit(fabric);
  }
}
function unhandled(what) {
  return (e, observer) => {
    logger.error(`Unhandled error in fabric ${what} observer ${observer.name || "(anon)"}`, e);
  };
}
