// @matter/nodejs/dist/cjs/storage/sqlite/SqliteUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SqliteUtil_exports = {};
__export(SqliteUtil_exports, {
  buildContextKeyLog: () => buildContextKeyLog,
  buildContextKeyPair: () => buildContextKeyPair,
  buildContextPath: () => buildContextPath,
  buildKey: () => buildKey,
  escapeGlob: () => escapeGlob
});
module.exports = __toCommonJS(SqliteUtil_exports);
var import_SqliteStorageDriverError = require("./SqliteStorageDriverError.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function buildContextPath(contexts) {
  for (const ctx of contexts) {
    if (ctx.trim().length <= 0 || ctx.includes(".")) {
      throw new import_SqliteStorageDriverError.SqliteStorageDriverError(
        "build",
        `{${contexts.join(",")}}`,
        "Context must not contain empty segments or leading or trailing dots."
      );
    }
  }
  return contexts.join(".");
}
function buildContextKeyPair(contexts, key) {
  return {
    context: buildContextPath(contexts),
    key: buildKey(key)
  };
}
function buildContextKeyLog(contexts, key) {
  return `${buildContextPath(contexts)}$${key}`;
}
function buildKey(key) {
  if (key.trim().length <= 0) {
    throw new import_SqliteStorageDriverError.SqliteStorageDriverError("build", key, "Key must not be an empty string.");
  }
  return key;
}
function escapeGlob(value) {
  return value.replace(/[*[?]/g, "[$&]");
}
