// @matter/nodejs/dist/cjs/storage/sqlite/SqliteBlobStorageDriver.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SqliteBlobStorageDriver_exports = {};
__export(SqliteBlobStorageDriver_exports, {
  SqliteBlobStorageDriver: () => SqliteBlobStorageDriver
});
module.exports = __toCommonJS(SqliteBlobStorageDriver_exports);
var import_general = require("@matter/general");
var import_node_path = require("node:path");
var import_SqlitePlatform = require("./SqlitePlatform.js");
var import_SqliteStorageDriverError = require("./SqliteStorageDriverError.js");
var import_SqliteUtil = require("./SqliteUtil.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class SqliteBlobStorageDriver extends import_general.FilesystemBlobStorageDriver {
  static id = "sqlite";
  static memoryPath = ":memory:";
  static defaultTableName = "blobstore";
  /**
   * Create a SqliteBlobStorageDriver for the given namespace using the platform-appropriate database.
   */
  static async create(namespace) {
    const storage = new SqliteBlobStorageDriver({ namespaceOrPath: namespace });
    await storage.initialize();
    return storage;
  }
  #isInitialized = false;
  #database;
  #databaseCreator;
  #dbPath;
  #tableName;
  // Prepared queries
  #queryOpenBlob;
  #queryWriteBlob;
  #queryDeleteBlob;
  #queryHasBlob;
  #queryKeysBlob;
  #queryContextsSub;
  #queryClearAll;
  /**
   * Create sqlite-based blob storage.
   *
   * @param args.databaseCreator Optional database instance creator. If omitted, resolved automatically during
   *   {@link initialize} via platform detection.
   * @param args.namespaceOrPath DataNamespace (derives path from root directory), string (direct path), or
   *   null/undefined for in-memory database
   * @param args.tableName table name
   */
  constructor(args) {
    const namespaceOrPath = args?.namespaceOrPath;
    super(namespaceOrPath instanceof import_general.DatafileRoot ? namespaceOrPath : void 0);
    this.#dbPath = typeof namespaceOrPath === "string" ? namespaceOrPath : namespaceOrPath instanceof import_general.DatafileRoot ? (0, import_node_path.resolve)(namespaceOrPath.directory.path, "storage.db") : SqliteBlobStorageDriver.memoryPath;
    this.#tableName = args?.tableName ?? SqliteBlobStorageDriver.defaultTableName;
    if (args?.databaseCreator) {
      this.#openDatabase(args.databaseCreator);
    }
  }
  #openDatabase(databaseCreator) {
    this.#databaseCreator = databaseCreator;
    this.#database = databaseCreator(this.#dbPath);
    const initQuery = this.#database.prepare(`
      CREATE TABLE IF NOT EXISTS ${this.#tableName} (
        context TEXT NOT NULL,
        key TEXT NOT NULL,
        data BLOB,
        PRIMARY KEY (context, key)
      ) STRICT
    `);
    initQuery.run();
    this.#queryOpenBlob = this.#database.prepare(`
      SELECT data FROM ${this.#tableName} WHERE
        context=$context AND
        key=$key
    `);
    this.#queryWriteBlob = this.#database.prepare(`
      INSERT INTO ${this.#tableName}
        (context, key, data)
      VALUES($context, $key, $data)
      ON CONFLICT(context, key)
      DO UPDATE SET data = excluded.data
    `);
    this.#queryDeleteBlob = this.#database.prepare(`
      DELETE FROM ${this.#tableName} WHERE
        context=$context AND
        key=$key
    `);
    this.#queryHasBlob = this.#database.prepare(`
      SELECT EXISTS(
        SELECT 1 FROM ${this.#tableName}
        WHERE context=$context AND key=$key
      ) as has_record
    `);
    this.#queryKeysBlob = this.#database.prepare(`
      SELECT DISTINCT key FROM ${this.#tableName} WHERE
        context=$context
    `);
    this.#queryContextsSub = this.#database.prepare(`
      SELECT DISTINCT context FROM ${this.#tableName} WHERE
        context GLOB $contextGlob
    `);
    this.#queryClearAll = this.#database.prepare(`
      DELETE FROM ${this.#tableName} WHERE
        context=$context OR context GLOB $contextGlob
    `);
  }
  get initialized() {
    return this.#isInitialized;
  }
  async initialize() {
    if (this.#isInitialized) {
      throw new import_SqliteStorageDriverError.SqliteStorageDriverError("initialize", this.#tableName, "Storage already initialized!");
    }
    await super.initialize();
    if (!this.#databaseCreator) {
      this.#openDatabase(await (0, import_SqlitePlatform.platformDatabaseCreator)());
    }
    this.#isInitialized = true;
  }
  async close() {
    this.#isInitialized = false;
    this.#database?.close();
    await super.close();
  }
  openBlob(contexts, key) {
    const queryResult = this.#queryOpenBlob.get((0, import_SqliteUtil.buildContextKeyPair)(contexts, key));
    if (queryResult == null || queryResult.data == null) {
      return new Blob();
    }
    return new Blob([new Uint8Array(queryResult.data)]);
  }
  // SQLite does not support incremental blob writes — the full blob must be buffered before INSERT.
  async writeBlobFromStream(contexts, key, stream) {
    const arrayBuffer = await new Response(stream).arrayBuffer();
    const bytes = new Uint8Array(arrayBuffer);
    const queryResult = this.#queryWriteBlob.run({
      ...(0, import_SqliteUtil.buildContextKeyPair)(contexts, key),
      data: bytes
    });
    if (Number(queryResult.changes) <= 0) {
      throw new import_SqliteStorageDriverError.SqliteStorageDriverError(
        "writeBlob",
        `${(0, import_SqliteUtil.buildContextPath)(contexts)}$${key}`,
        `Something went wrong! Value wasn't changed.`
      );
    }
  }
  delete(contexts, key) {
    this.#queryDeleteBlob.run((0, import_SqliteUtil.buildContextKeyPair)(contexts, key));
  }
  has(contexts, key) {
    const result = this.#queryHasBlob.get((0, import_SqliteUtil.buildContextKeyPair)(contexts, key));
    return result?.has_record === 1;
  }
  keys(contexts) {
    const queryResults = this.#queryKeysBlob.all({
      context: (0, import_SqliteUtil.buildContextPath)(contexts)
    }).filter((v) => v != null);
    return queryResults.map((v) => v.key);
  }
  contexts(contexts) {
    const parentCtx = contexts.length > 0 ? (0, import_SqliteUtil.buildContextPath)(contexts) : "";
    let subContexts;
    if (contexts.length === 0) {
      const allContexts = this.#queryContextsSub.all({ contextGlob: "*" }).filter((v) => v != null);
      subContexts = allContexts.map((v) => {
        const firstDotIndex = v.context.indexOf(".");
        if (firstDotIndex < 0) {
          return v.context;
        }
        return v.context.substring(0, firstDotIndex);
      });
    } else {
      const allSubContexts = this.#queryContextsSub.all({
        contextGlob: (0, import_SqliteUtil.escapeGlob)(parentCtx) + ".*"
      }).filter((v) => v != null);
      subContexts = allSubContexts.map((v) => {
        const subKey = v.context.substring(parentCtx.length + 1);
        const dotIndex = subKey.indexOf(".");
        if (dotIndex < 0) {
          return subKey;
        }
        return subKey.substring(0, dotIndex);
      });
    }
    return [...new Set(subContexts.filter((c) => c != null && c.trim().length > 0))];
  }
  clearAll(contexts) {
    if (contexts.length === 0) {
      return;
    }
    const contextPath = (0, import_SqliteUtil.buildContextPath)(contexts);
    this.#queryClearAll.run({
      context: contextPath,
      contextGlob: (0, import_SqliteUtil.escapeGlob)(contextPath) + ".*"
    });
  }
}
