// @matter/general/dist/esm/net/ServerAddress.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Diagnostic } from "#log/Diagnostic.js";
import { Duration } from "#time/Duration.js";
function ServerAddress(definition) {
  return {
    ttl: void 0,
    discoveredAt: void 0,
    healthyAt: void 0,
    unhealthyAt: void 0,
    priority: void 0,
    weight: void 0,
    ...definition
  };
}
((ServerAddress2) => {
  function isIp(address) {
    return address.ip !== void 0;
  }
  ServerAddress2.isIp = isIp;
  function isBle(address) {
    return address.peripheralAddress !== void 0;
  }
  ServerAddress2.isBle = isBle;
  function protocolOf(address) {
    if (isBle(address)) {
      return "ble";
    }
    if ("type" in address && typeof address.type === "string") {
      return address.type;
    }
    return "ip";
  }
  ServerAddress2.protocolOf = protocolOf;
  function urlFor(address) {
    if (isIp(address)) {
      const proto = protocolOf(address);
      const host = address.ip.includes(":") ? `[${address.ip}]` : address.ip;
      return `${proto}://${host}:${address.port}`;
    }
    if (isBle(address)) {
      return `ble://${address.peripheralAddress}`;
    }
    return `unknown://`;
  }
  ServerAddress2.urlFor = urlFor;
  function diagnosticFor(address) {
    const diagnostic = Array();
    if (isIp(address)) {
      diagnostic.push(`${protocolOf(address)}://`, Diagnostic.strong(address.ip), ":", address.port);
    } else if (isBle(address)) {
      diagnostic.push("ble://", Diagnostic.strong(address.peripheralAddress));
    } else {
      diagnostic.push("unknown://");
    }
    if ("ttl" in address && typeof address.ttl === "number") {
      diagnostic.push(" ttl ", Duration.format(address.ttl));
    }
    return Diagnostic.squash(...diagnostic);
  }
  ServerAddress2.diagnosticFor = diagnosticFor;
  function isEqual(a, b) {
    if (isIp(a) && isIp(b)) {
      return a.ip === b.ip && a.port === b.port;
    }
    if (isBle(a) && isBle(b)) {
      return a.peripheralAddress === b.peripheralAddress;
    }
    return false;
  }
  ServerAddress2.isEqual = isEqual;
  function healthOf(health) {
    if (health.unhealthyAt === void 0) {
      return health;
    }
    if (health.healthyAt !== void 0 && health.healthyAt > health.unhealthyAt) {
      return {
        healthyAt: health.healthyAt
      };
    }
    return health;
  }
  ServerAddress2.healthOf = healthOf;
  let SelectionPreference;
  ((SelectionPreference2) => {
    SelectionPreference2[SelectionPreference2["IPV6_ULA"] = 0] = "IPV6_ULA";
    SelectionPreference2[SelectionPreference2["IPV6_LINK_LOCAL"] = 1] = "IPV6_LINK_LOCAL";
    SelectionPreference2[SelectionPreference2["IPV6"] = 2] = "IPV6";
    SelectionPreference2[SelectionPreference2["IPV4"] = 3] = "IPV4";
    SelectionPreference2[SelectionPreference2["NOT_IP"] = 3] = "NOT_IP";
  })(SelectionPreference = ServerAddress2.SelectionPreference || (ServerAddress2.SelectionPreference = {}));
  const IPV6_LINK_LOCAL_PATTERN = /^fe[89ab]/;
  function isIpv6LinkLocal(ip) {
    return IPV6_LINK_LOCAL_PATTERN.test(ip);
  }
  ServerAddress2.isIpv6LinkLocal = isIpv6LinkLocal;
  function selectionPreferenceOf(address) {
    if (!isIp(address)) {
      return 3 /* NOT_IP */;
    }
    const ip = address.ip;
    if (ip.startsWith("fd")) {
      return 0 /* IPV6_ULA */;
    }
    if (isIpv6LinkLocal(ip)) {
      return 1 /* IPV6_LINK_LOCAL */;
    }
    if (ip.includes(":")) {
      return 2 /* IPV6 */;
    }
    return 3 /* IPV4 */;
  }
  ServerAddress2.selectionPreferenceOf = selectionPreferenceOf;
})(ServerAddress || (ServerAddress = {}));
export {
  ServerAddress
};
