// @matter/node/dist/cjs/storage/server/ServerNodeStore.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ServerNodeStore_exports = {};
__export(ServerNodeStore_exports, {
  ServerNodeStore: () => ServerNodeStore
});
module.exports = __toCommonJS(ServerNodeStore_exports);
var import_general = require("@matter/general");
var import_NodeStore = require("../NodeStore.js");
var import_ClientNodeStores = require("../client/ClientNodeStores.js");
var import_ServerEndpointStores = require("./ServerEndpointStores.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("ServerNodeStore");
class ServerNodeStore extends import_NodeStore.NodeStore {
  #env;
  #nodeId;
  #endpointStores;
  #storageManager;
  #clientStores;
  #bdxHandle;
  constructor(environment, nodeId) {
    super({
      createContext: (name) => {
        if (!this.#storageManager) {
          throw new import_general.ImplementationError(
            `Cannot create storage context ${name} because store is not initialized`
          );
        }
        return this.#storageManager.createContext(name);
      }
    });
    this.#endpointStores = new import_ServerEndpointStores.ServerEndpointStores();
    this.#env = environment;
    this.#nodeId = nodeId;
    this.construction.start();
  }
  static async create(environment, nodeId) {
    return await (0, import_general.asyncNew)(this, environment, nodeId);
  }
  async close() {
    await this.construction.close(async () => {
      await this.#clientStores?.close();
      await this.#bdxHandle?.close();
      await this.#storageManager?.close();
      await this.#env.get(import_general.StorageService).close(this.#nodeId);
      this.#logChange("Closed");
    });
  }
  /**
   * Stores associated with server endpoints supported by this node.
   */
  get endpointStores() {
    return this.construction.assert("endpoint stores", this.#endpointStores);
  }
  /**
   * Stores associated with remote nodes known by this node.
   */
  get clientStores() {
    return this.construction.assert("client stores", this.#clientStores);
  }
  /**
   * The underlying {@link StorageManager} that provides node data.
   */
  get storage() {
    return this.construction.assert("storage manager", this.#storageManager);
  }
  #logChange(what) {
    const root = this.#env.has(import_general.DatafileRoot) ? this.#env.get(import_general.DatafileRoot) : void 0;
    logger.info(
      what,
      import_general.Diagnostic.strong(this.#nodeId ?? "node"),
      "storage",
      import_general.Diagnostic.dict({
        location: root?.path ?? "(unknown location)",
        driver: this.#storageManager?.driverId ?? "unknown"
      })
    );
  }
  storeForEndpoint(endpoint) {
    return this.#endpointStores.storeForEndpoint(endpoint);
  }
  /**
   * Lazily opens and returns the BDX blob storage driver using a dedicated namespace.
   */
  async bdxStore() {
    if (!this.#bdxHandle) {
      const root = this.#env.has(import_general.DatafileRoot) ? this.#env.get(import_general.DatafileRoot) : void 0;
      const blobNamespace = root ? new import_general.DatafileRoot(root.directory.directory(`${this.#nodeId}-bdx`)) : `${this.#nodeId}-bdx`;
      this.#bdxHandle = await this.#env.get(import_general.StorageService).openBlobStorage(blobNamespace);
    }
    return this.#bdxHandle.driver;
  }
  erase() {
    return this.#endpointStores.erase();
  }
  async load() {
    const root = this.#env.has(import_general.DatafileRoot) ? this.#env.get(import_general.DatafileRoot) : void 0;
    this.#storageManager = await this.#env.get(import_general.StorageService).open(root ?? this.#nodeId);
    this.#env.set(import_general.StorageManager, this.#storageManager);
    this.#clientStores = await (0, import_general.asyncNew)(import_ClientNodeStores.ClientNodeStores, this.#storageManager.createContext("nodes"));
    const rootContext = this.storageFactory.createContext("root");
    await this.#endpointStores.load(rootContext);
    this.#logChange("Opened");
  }
}
