// @matter/protocol/dist/cjs/action/server/DataResponse.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataResponse_exports = {};
__export(DataResponse_exports, {
  DataResponse: () => DataResponse,
  FallbackLimits: () => FallbackLimits,
  WildcardPathFlagsCodec: () => WildcardPathFlagsCodec
});
module.exports = __toCommonJS(DataResponse_exports);
var import_model = require("@matter/model");
var import_types = require("@matter/types");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WildcardPathFlagsCodec = (0, import_types.BitmapSchema)(import_types.WildcardPathFlagsBitmap);
const FallbackLimits = {
  fabricScoped: false,
  fabricSensitive: false,
  readable: true,
  readLevel: import_model.AccessLevel.View,
  timed: false,
  largeMessage: false,
  writable: true,
  writeLevel: import_model.AccessLevel.Administer
};
class DataResponse {
  // Configuration
  #session;
  #node;
  // The node ID may be expensive to retrieve and is invariant, so we cache it here
  #cachedNodeId;
  constructor(node, session) {
    this.#node = node;
    this.#session = session;
  }
  get node() {
    return this.#node;
  }
  get session() {
    return this.#session;
  }
  /**
   * Update the session for processing subsequent chunks.
   * This allows reusing the same response instance while maintaining state across chunks.
   */
  updateSession(session) {
    this.#session = session;
  }
  /**
   * The node ID used to filter paths with node ID specified.  Unsure if this is ever actually used.
   */
  get nodeId() {
    if (this.#cachedNodeId === void 0) {
      this.#cachedNodeId = (this.#session.fabric && this.#node.nodeIdFor(this.#session.fabric)) ?? import_types.NodeId.UNSPECIFIED_NODE_ID;
    }
    return this.#cachedNodeId;
  }
  join(...name) {
    return this.#session.transaction?.join(...name);
  }
}
