// @matter/general/dist/cjs/util/DataReadQueue.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var DataReadQueue_exports = {};
__export(DataReadQueue_exports, {
  DataReadQueue: () => DataReadQueue
});
module.exports = __toCommonJS(DataReadQueue_exports);
var import_TimeUnit = require("#time/TimeUnit.js");
var import_MatterError = require("../MatterError.js");
var import_Abort = require("./Abort.js");
var import_Promises = require("./Promises.js");
var import_Streams = require("./Streams.js");
/**
 * Promise-based blocking queue.
 *
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class DataReadQueue {
  #queue = new Array();
  #pendingRead;
  #closeCause;
  async read(abort) {
    var _stack = [];
    try {
      const { promise, resolver } = (0, import_Promises.createPromise)();
      if (this.#closeCause) {
        throw new import_MatterError.ClosedError("Channel is closed", { cause: this.#closeCause });
      }
      const data = this.#queue.shift();
      if (data !== void 0) {
        return data;
      }
      if (this.#pendingRead !== void 0) throw new import_MatterError.InternalError("Only one pending read is supported");
      const localAbort = __using(_stack, new import_Abort.Abort({
        timeout: abort ? void 0 : import_TimeUnit.Minutes.one,
        abort
      }));
      this.#pendingRead = {
        resolver,
        rejecter(cause) {
          localAbort.abort(cause);
        }
      };
      try {
        return await localAbort.attempt(promise);
      } catch (e) {
        if (e instanceof import_MatterError.AbortedError) {
          throw e;
        }
        throw e;
      } finally {
        this.#pendingRead = void 0;
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  write(data) {
    if (this.#closeCause) {
      throw new import_MatterError.ClosedError("Channel is closed", { cause: this.#closeCause });
    }
    const pendingRead = this.#pendingRead;
    this.#pendingRead = void 0;
    if (pendingRead) {
      pendingRead.resolver(data);
      return;
    }
    this.#queue.push(data);
  }
  get size() {
    return this.#queue.length;
  }
  close(cause = new import_Streams.EndOfStreamError()) {
    if (this.#closeCause) {
      return;
    }
    this.#closeCause = cause;
    const pendingRead = this.#pendingRead;
    this.#pendingRead = void 0;
    if (pendingRead) {
      pendingRead.rejecter(cause);
    }
  }
}
