// @matter/general/dist/cjs/util/Entropy.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Entropy_exports = {};
__export(Entropy_exports, {
  Entropy: () => Entropy
});
module.exports = __toCommonJS(Entropy_exports);
var import_Bytes = require("#util/Bytes.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Entropy {
  get randomUint8() {
    return import_Bytes.Bytes.of(this.randomBytes(1))[0];
  }
  get randomUint16() {
    return import_Bytes.Bytes.dataViewOf(this.randomBytes(2)).getUint16(0);
  }
  get randomUint32() {
    return import_Bytes.Bytes.dataViewOf(this.randomBytes(4)).getUint32(0);
  }
  get randomBigUint64() {
    return import_Bytes.Bytes.dataViewOf(this.randomBytes(8)).getBigUint64(0);
  }
  randomBigInt(size, maxValue) {
    if (maxValue === void 0) {
      return import_Bytes.Bytes.asBigInt(this.randomBytes(size));
    }
    while (true) {
      const random = import_Bytes.Bytes.asBigInt(this.randomBytes(size));
      if (random < maxValue) return random;
    }
  }
}
