// @matter/general/dist/cjs/util/Heap.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Heap_exports = {};
__export(Heap_exports, {
  Heap: () => Heap
});
module.exports = __toCommonJS(Heap_exports);
var import_MatterError = require("#MatterError.js");
var import_Abort = require("./Abort.js");
var import_Observable = require("./Observable.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Heap {
  // We encode the heap as an array for efficiency.  See #leftChildOf and #rightChildOf for how to navigate the tree
  // when encoded this way
  #buffer = Array();
  #compare;
  #normalize;
  #firstChanged;
  #deleted;
  #added;
  #positions;
  /**
   * Create new heap.
   *
   * @param comparator performs ordering of items in the heap
   * @param normalizer optionally converts items to normal form on insert
   */
  constructor(comparator, normalizer) {
    this.#compare = comparator;
    this.#normalize = normalizer;
  }
  /**
   * Return lowest-ranked item.
   */
  shift() {
    if (!this.#buffer?.length) {
      return void 0;
    }
    const result = this.first;
    this.#deleteAt(0);
    return result;
  }
  /**
   * The lowest-ranked item.
   */
  get first() {
    return this.#buffer?.[0];
  }
  /**
   * The queue length.
   */
  get size() {
    return this.#buffer.length;
  }
  /**
   * Is the heap empty?
   */
  get isEmpty() {
    return !!this.#buffer.length;
  }
  /**
   * Emits when the head of the queue changes.
   */
  get firstChanged() {
    if (!this.#firstChanged) {
      this.#firstChanged = new import_Observable.Observable();
    }
    return this.#firstChanged;
  }
  /**
   * Emits when an item is added to the heap.
   */
  get added() {
    if (!this.#added) {
      this.#added = new import_Observable.Observable();
    }
    return this.#added;
  }
  /**
   * Emits when an item is deleted.
   */
  get deleted() {
    if (!this.#deleted) {
      this.#deleted = new import_Observable.Observable();
    }
    return this.#deleted;
  }
  /**
   * Add an item.
   */
  add(...items) {
    for (let item of items) {
      if (this.#normalize) {
        item = this.#normalize(item);
      }
      this.#buffer.push(item);
      this.#bubbleUp(this.#buffer.length - 1);
      this.#added?.emit(item);
    }
  }
  /**
   * Delete an item.
   *
   * The first delete is O(n); subsequent deletes are O(1) but insertions and deletions will have additional cost
   * associated.
   */
  delete(item) {
    if (this.#buffer === void 0) {
      return false;
    }
    if (!this.#positions) {
      this.#positions = /* @__PURE__ */ new Map();
      for (let i = 0; i < this.#buffer.length; i++) {
        this.#positions.set(this.#buffer[i], i);
      }
    }
    if (this.#normalize) {
      item = this.#normalize(item);
    }
    const pos = this.#positions.get(item);
    if (pos === void 0) {
      return false;
    }
    this.#deleteAt(pos);
    return true;
  }
  /**
   * Remove all entries.
   */
  clear() {
    if (this.#buffer.length) this.#buffer.length = 0;
  }
  /**
   * Stream the first value from the heap until aborted.
   */
  async *stream(abort) {
    if (import_Abort.Abort.is(abort)) {
      return;
    }
    while (true) {
      while (this.size) {
        yield this.shift();
        if (import_Abort.Abort.is(abort)) {
          return;
        }
      }
      let resolve;
      const ready = new Promise((r) => resolve = r);
      try {
        this.added.once(resolve);
        await import_Abort.Abort.race(abort, ready);
        if (import_Abort.Abort.is(abort)) {
          return;
        }
      } finally {
        this.added.off(resolve);
      }
    }
  }
  /**
   * Perform internal validation of queue order.
   */
  validate() {
    for (let i = 0; i < this.#buffer.length; i++) {
      const leftChild = this.#leftChildOf(i);
      if (leftChild < this.#buffer.length) {
        if (this.#compare(this.#buffer[i], this.#buffer[leftChild]) > 0) {
          throw new import_MatterError.InternalError(
            `Heap error: buffer #${i} (${this.#buffer[i]}) is greater than left child #${leftChild} (${this.#buffer[leftChild]})`
          );
        }
        const rightChild = this.#rightChildOf(i);
        if (rightChild < this.#buffer.length) {
          if (this.#compare(this.#buffer[i], this.#buffer[rightChild]) > 0) {
            throw new import_MatterError.InternalError(
              `Heap error: buffer #${i} (${this.#buffer[i]}) is greater than right child #${rightChild} (${this.#buffer[rightChild]})`
            );
          }
          if (this.#compare(this.#buffer[leftChild], this.#buffer[rightChild]) > 0) {
            throw new import_MatterError.InternalError(
              `Heap error: buffer #${leftChild} (${this.#buffer[leftChild]}) is greater than right sibling #${rightChild} (${this.#buffer[rightChild]})`
            );
          }
        }
      }
    }
  }
  #deleteAt(index) {
    if (index >= this.#buffer.length) {
      return;
    }
    const item = this.#buffer[index];
    if (this.#buffer.length === 1) {
      this.#buffer.length = 0;
    } else {
      const lastIndex = this.#buffer.length - 1;
      this.#buffer[index] = this.#buffer[lastIndex];
      this.#positions?.set(this.#buffer[index], index);
      this.#buffer.length = lastIndex;
      if (index < this.#buffer.length) {
        if (index) {
          index = this.#bubbleUp(index);
        }
        this.#sinkDown(index);
      }
    }
    this.#positions?.delete(item);
    this.#deleted?.emit(item);
    if (!index) {
      this.#firstChanged?.emit(this.first);
    }
  }
  #sinkDown(index) {
    while (true) {
      let moveTo;
      const leftChild = this.#leftChildOf(index);
      if (leftChild < this.#buffer.length) {
        if (this.#compare(this.#buffer[index], this.#buffer[leftChild]) > 0) {
          moveTo = leftChild;
        }
        const rightChild = this.#rightChildOf(index);
        if (rightChild < this.#buffer.length) {
          if (this.#compare(this.#buffer[moveTo ?? index], this.#buffer[rightChild]) > 0) {
            moveTo = rightChild;
          }
        }
      }
      if (moveTo === void 0) {
        break;
      }
      this.#swap(index, moveTo);
      index = moveTo;
    }
  }
  #bubbleUp(index) {
    while (index) {
      const parent = this.#parentOf(index);
      if (this.#compare(this.#buffer[parent], this.#buffer[index]) <= 0) {
        break;
      }
      this.#swap(index, parent);
      index = parent;
    }
    if (!index) {
      this.#firstChanged?.emit(this.first);
    }
    return index;
  }
  #swap(index1, index2) {
    [this.#buffer[index1], this.#buffer[index2]] = [this.#buffer[index2], this.#buffer[index1]];
    if (this.#positions) {
      this.#positions.set(this.#buffer[index1], index1);
      this.#positions.set(this.#buffer[index2], index2);
    }
  }
  #leftChildOf(index) {
    return 2 * index + 1;
  }
  #rightChildOf(index) {
    return 2 * index + 2;
  }
  #parentOf(index) {
    return Math.floor((index - 1) / 2);
  }
}
