// @matter/types/dist/cjs/globals/Namespace.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Namespace_exports = {};
__export(Namespace_exports, {
  Namespace: () => Namespace
});
module.exports = __toCommonJS(Namespace_exports);
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Namespace = /* @__PURE__ */ ((Namespace2) => {
  Namespace2[Namespace2["CommonClosure"] = 1] = "CommonClosure";
  Namespace2[Namespace2["CommonCompassDirection"] = 2] = "CommonCompassDirection";
  Namespace2[Namespace2["CommonCompassLocation"] = 3] = "CommonCompassLocation";
  Namespace2[Namespace2["CommonDirection"] = 4] = "CommonDirection";
  Namespace2[Namespace2["CommonLevel"] = 5] = "CommonLevel";
  Namespace2[Namespace2["CommonLocation"] = 6] = "CommonLocation";
  Namespace2[Namespace2["CommonNumber"] = 7] = "CommonNumber";
  Namespace2[Namespace2["CommonPosition"] = 8] = "CommonPosition";
  Namespace2[Namespace2["ElectricalMeasurement"] = 10] = "ElectricalMeasurement";
  Namespace2[Namespace2["CommodityTariffChronology"] = 11] = "CommodityTariffChronology";
  Namespace2[Namespace2["CommodityTariffCommodity"] = 13] = "CommodityTariffCommodity";
  Namespace2[Namespace2["Laundry"] = 14] = "Laundry";
  Namespace2[Namespace2["PowerSource"] = 15] = "PowerSource";
  Namespace2[Namespace2["CommonAreaNamespace"] = 16] = "CommonAreaNamespace";
  Namespace2[Namespace2["CommonLandmarkNamespace"] = 17] = "CommonLandmarkNamespace";
  Namespace2[Namespace2["CommonRelativePosition"] = 18] = "CommonRelativePosition";
  Namespace2[Namespace2["CommodityTariffFlow"] = 19] = "CommodityTariffFlow";
  Namespace2[Namespace2["Refrigerator"] = 65] = "Refrigerator";
  Namespace2[Namespace2["RoomAirConditioner"] = 66] = "RoomAirConditioner";
  Namespace2[Namespace2["Switches"] = 67] = "Switches";
  Namespace2[Namespace2["Closure"] = 68] = "Closure";
  Namespace2[Namespace2["ClosurePanel"] = 69] = "ClosurePanel";
  Namespace2[Namespace2["ClosureCovering"] = 70] = "ClosureCovering";
  Namespace2[Namespace2["ClosureWindow"] = 71] = "ClosureWindow";
  Namespace2[Namespace2["ClosureCabinet"] = 72] = "ClosureCabinet";
  return Namespace2;
})(Namespace || {});
