// @matter/general/dist/esm/util/Cache.js
/**
 * Cache computed values or resources for a specified duration to improve performances.
 *
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Diagnostic } from "../log/Diagnostic.js";
import { Time } from "../time/Time.js";
const KEY_SEPARATOR = "";
class GenericCache {
  constructor(name, expiration, expireCallback) {
    this.expiration = expiration;
    this.expireCallback = expireCallback;
    this.periodicTimer = Time.getPeriodicTimer(
      Diagnostic.upgrade(`${name} cache expiration`, [Diagnostic.strong(name), "cache expiration"]),
      expiration,
      () => this.expire()
    ).start();
    this.periodicTimer.utility = true;
  }
  expiration;
  expireCallback;
  knownKeys = /* @__PURE__ */ new Set();
  values = /* @__PURE__ */ new Map();
  timestamps = /* @__PURE__ */ new Map();
  periodicTimer;
  // Single-param keys must stay identical to the stringified raw value so `keys()` → `get()` round-trips.
  keyFor(params) {
    return params.map((p) => String(p)).join(KEY_SEPARATOR);
  }
  keys() {
    return Array.from(this.knownKeys.values());
  }
  async delete(key) {
    if (this.expireCallback !== void 0 && this.values.has(key)) {
      await this.expireCallback(key, this.values.get(key));
    }
    this.values.delete(key);
    this.timestamps.delete(key);
  }
  async clear() {
    for (const key of this.values.keys()) {
      await this.delete(key);
    }
    this.values.clear();
    this.timestamps.clear();
  }
  async close() {
    await this.clear();
    this.knownKeys.clear();
    this.periodicTimer.stop();
  }
  async expire() {
    const now = Time.nowMs;
    for (const [key, timestamp] of this.timestamps.entries()) {
      if (now - timestamp < this.expiration) continue;
      await this.delete(key);
    }
  }
}
class Cache extends GenericCache {
  constructor(name, generator, expiration, expireCallback) {
    super(name, expiration, expireCallback);
    this.generator = generator;
  }
  generator;
  get(...params) {
    const key = this.keyFor(params);
    if (!this.values.has(key)) {
      this.values.set(key, this.generator(...params));
      this.knownKeys.add(key);
    }
    this.timestamps.set(key, Time.nowMs);
    return this.values.get(key);
  }
}
class AsyncCache extends GenericCache {
  constructor(name, generator, expiration, expireCallback) {
    super(name, expiration, expireCallback);
    this.generator = generator;
  }
  generator;
  #inflight = /* @__PURE__ */ new Map();
  get(...params) {
    const key = this.keyFor(params);
    if (this.values.has(key)) {
      this.timestamps.set(key, Time.nowMs);
      return Promise.resolve(this.values.get(key));
    }
    let pending = this.#inflight.get(key);
    if (pending === void 0) {
      pending = this.#fill(key, params);
      this.#inflight.set(key, pending);
      const drop = () => this.#inflight.delete(key);
      pending.then(drop, drop);
    }
    return pending;
  }
  async #fill(key, params) {
    const value = await this.generator(...params);
    this.values.set(key, value);
    this.knownKeys.add(key);
    this.timestamps.set(key, Time.nowMs);
    return value;
  }
  // Sync point for `expireCallback` to see the settled value.
  async delete(key) {
    const pending = this.#inflight.get(key);
    if (pending !== void 0) {
      try {
        await pending;
      } catch {
      }
    }
    await super.delete(key);
  }
  async clear() {
    await Promise.allSettled(this.#inflight.values());
    await super.clear();
  }
}
export {
  AsyncCache,
  Cache
};
