// @matter/general/dist/esm/util/Cache.js
/**
 * Cache computed values or resources for a specified duration to improve performances.
 *
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Diagnostic } from "../log/Diagnostic.js";
import { Time } from "../time/Time.js";
class GenericCache {
  constructor(name, expiration, expireCallback) {
    this.expiration = expiration;
    this.expireCallback = expireCallback;
    this.periodicTimer = Time.getPeriodicTimer(
      Diagnostic.upgrade(`${name} cache expiration`, [Diagnostic.strong(name), "cache expiration"]),
      expiration,
      () => this.expire()
    ).start();
    this.periodicTimer.utility = true;
  }
  expiration;
  expireCallback;
  knownKeys = /* @__PURE__ */ new Set();
  values = /* @__PURE__ */ new Map();
  timestamps = /* @__PURE__ */ new Map();
  periodicTimer;
  keys() {
    return Array.from(this.knownKeys.values());
  }
  async delete(key) {
    const value = this.values.get(key);
    if (this.expireCallback !== void 0 && value !== void 0) {
      await this.expireCallback(key, value);
    }
    this.values.delete(key);
    this.timestamps.delete(key);
  }
  async clear() {
    for (const key of this.values.keys()) {
      await this.delete(key);
    }
    this.values.clear();
    this.timestamps.clear();
  }
  async close() {
    await this.clear();
    this.knownKeys.clear();
    this.periodicTimer.stop();
  }
  async expire() {
    const now = Time.nowMs;
    for (const [key, timestamp] of this.timestamps.entries()) {
      if (now - timestamp < this.expiration) continue;
      await this.delete(key);
    }
  }
}
class Cache extends GenericCache {
  constructor(name, generator, expiration, expireCallback) {
    super(name, expiration, expireCallback);
    this.generator = generator;
  }
  generator;
  get(...params) {
    const key = params.join(",");
    let value = this.values.get(key);
    if (value === void 0) {
      value = this.generator(...params);
      this.values.set(key, value);
      this.knownKeys.add(key);
    }
    this.timestamps.set(key, Time.nowMs);
    return value;
  }
}
class AsyncCache extends GenericCache {
  constructor(name, generator, expiration, expireCallback) {
    super(name, expiration, expireCallback);
    this.generator = generator;
  }
  generator;
  async get(...params) {
    const key = params.join(",");
    let value = this.values.get(key);
    if (value === void 0) {
      value = await this.generator(...params);
      this.values.set(key, value);
      this.knownKeys.add(key);
    }
    this.timestamps.set(key, Time.nowMs);
    return value;
  }
}
export {
  AsyncCache,
  Cache
};
