// @matter/general/dist/esm/util/DataReadQueue.js
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
/**
 * Promise-based blocking queue.
 *
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Minutes } from "#time/TimeUnit.js";
import { AbortedError, ClosedError, InternalError } from "../MatterError.js";
import { Abort } from "./Abort.js";
import { createPromise } from "./Promises.js";
import { EndOfStreamError } from "./Streams.js";
class DataReadQueue {
  #queue = new Array();
  #pendingRead;
  #closeCause;
  async read(abort) {
    var _stack = [];
    try {
      const { promise, resolver } = createPromise();
      if (this.#closeCause) {
        throw new ClosedError("Channel is closed", { cause: this.#closeCause });
      }
      const data = this.#queue.shift();
      if (data !== void 0) {
        return data;
      }
      if (this.#pendingRead !== void 0) throw new InternalError("Only one pending read is supported");
      const localAbort = __using(_stack, new Abort({
        timeout: abort ? void 0 : Minutes.one,
        abort
      }));
      this.#pendingRead = {
        resolver,
        rejecter(cause) {
          localAbort.abort(cause);
        }
      };
      try {
        return await localAbort.attempt(promise);
      } catch (e) {
        if (e instanceof AbortedError) {
          throw e;
        }
        throw e;
      } finally {
        this.#pendingRead = void 0;
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  write(data) {
    if (this.#closeCause) {
      throw new ClosedError("Channel is closed", { cause: this.#closeCause });
    }
    const pendingRead = this.#pendingRead;
    this.#pendingRead = void 0;
    if (pendingRead) {
      pendingRead.resolver(data);
      return;
    }
    this.#queue.push(data);
  }
  get size() {
    return this.#queue.length;
  }
  close(cause = new EndOfStreamError()) {
    if (this.#closeCause) {
      return;
    }
    this.#closeCause = cause;
    const pendingRead = this.#pendingRead;
    this.#pendingRead = void 0;
    if (pendingRead) {
      pendingRead.rejecter(cause);
    }
  }
}
export {
  DataReadQueue
};
