// @matter/general/dist/esm/util/Scheduler.js
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "#log/Logger.js";
import { ImplementationError } from "#MatterError.js";
import { Time } from "#time/Time.js";
import { Timestamp } from "#time/Timestamp.js";
import { Abort } from "./Abort.js";
import { asError } from "./Error.js";
import { Heap } from "./Heap.js";
import { Lifetime } from "./Lifetime.js";
import { BasicMultiplex } from "./Multiplex.js";
const logger = Logger.get("Scheduler");
class Scheduler {
  #name;
  #lifetime;
  #timeOf;
  #run;
  #queue;
  #abort = new Abort();
  #workers = new BasicMultiplex();
  #semaphore;
  constructor({
    name = "scheduler",
    lifetime = Lifetime.process,
    timeOf,
    run,
    semaphore
  }) {
    this.#name = name;
    this.#lifetime = lifetime.join(this.#name);
    this.#timeOf = timeOf;
    this.#run = run;
    this.#semaphore = semaphore;
    this.#queue = new Heap(comparatorFor(timeOf));
    this.#workers.add(this.#schedule());
  }
  /**
   * Schedule a worker.
   */
  add(worker) {
    if (this.#lifetime.isClosing) {
      throw new ImplementationError(`Cannot schedule new work because ${this.#name} is closed`);
    }
    this.#queue.add(worker);
  }
  /**
   * Unschedule a worker.
   */
  delete(worker) {
    this.#queue.delete(worker);
  }
  /**
   * Update a worker's scheduled run time.
   */
  reschedule(worker) {
    this.#queue.delete(worker);
    this.#queue.add(worker);
  }
  /**
   * Close.
   *
   * Stops accepting workers.  Returns when all workers are complete.
   */
  async close() {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      this.#abort();
      await this.#workers;
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  /**
   * Performs actual scheduling of workers.
   */
  async #schedule() {
    while (!this.#abort.aborted) {
      var _stack = [];
      try {
        const now = Time.nowMs;
        let first = this.#queue.first;
        let nextAt = first ? this.#timeOf(first) : void 0;
        while (nextAt !== void 0 && nextAt <= now) {
          if (this.#abort.aborted) {
            return;
          }
          let slot;
          if (this.#semaphore) {
            slot = await this.#semaphore.obtainSlot(this.#abort);
            if (this.#abort.aborted) {
              return;
            }
          }
          try {
            this.#queue.shift();
            const promise = this.#run(first, this.#abort.signal);
            if (promise) {
              let p = Promise.resolve(promise).catch(this.#unhandled.bind(this));
              if (slot) {
                p = p.finally(() => slot[Symbol.dispose]());
              }
              this.#workers.add(p);
            } else {
              slot?.[Symbol.dispose]();
            }
          } catch (e) {
            slot?.[Symbol.dispose]();
            this.#unhandled(e);
          }
          first = this.#queue.first;
          nextAt = first ? this.#timeOf(first) : void 0;
        }
        const abort = __using(_stack, new Abort({
          abort: this.#abort,
          timeout: nextAt === void 0 ? void 0 : Timestamp.delta(now, nextAt)
        }));
        const _abortOnChange = __using(_stack, this.#queue.firstChanged.use((newFirst) => abort.if(newFirst !== first)));
        await abort;
      } catch (_) {
        var _error = _, _hasError = true;
      } finally {
        __callDispose(_stack, _error, _hasError);
      }
    }
  }
  #unhandled(e) {
    logger.error(`Unhandled error in ${this.#name} worker:`, asError(e));
  }
}
function comparatorFor(timeOf) {
  return (a, b) => timeOf(a) - timeOf(b);
}
export {
  Scheduler
};
