// @matter/general/dist/esm/util/String.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ImplementationError } from "#MatterError.js";
import { Bytes } from "./Bytes.js";
import { isObject } from "./Type.js";
function asJson(value, space) {
  return JSON.stringify(
    value,
    (_key, value2) => {
      if (typeof value2 === "bigint") {
        return value2.toString();
      }
      if (Bytes.isBytes(value2)) {
        return Bytes.toHex(value2);
      }
      return value2;
    },
    space
  );
}
function describeList(setType, ...entries) {
  const text = Array();
  if (entries.length === 1) {
    return entries[0];
  }
  for (let i = 0; i < entries.length; i++) {
    if (i === entries.length - 1) {
      text.push(setType, entries[i]);
    } else if (i === entries.length - 2) {
      text.push(entries[i]);
    } else {
      text.push(`${entries[i]},`);
    }
  }
  return text.join(" ");
}
function serializeToJs(value) {
  switch (typeof value) {
    case "bigint":
      return `0x${value.toString(16)}n`;
    case "boolean":
      return `${value}`;
    case "number":
      if (Number.isSafeInteger(value) && value > 10) {
        return `0x${value.toString(16)}`;
      }
      return `${value}`;
    case "string":
      return JSON.stringify(value);
    case "undefined":
      return "undefined";
    case "function":
    case "symbol":
      throw new ImplementationError(`Cannot serialize a ${typeof value}`);
  }
  if (value instanceof Date) {
    return `new Date(${JSON.stringify(value.toISOString)})`;
  }
  if (Bytes.isBytes(value)) {
    return `b$\`${Bytes.toHex(value)}\``;
  }
  if (value === null) {
    return "null";
  }
  if (Array.isArray(value)) {
    const parts2 = ["["];
    for (const item of value) {
      parts2.push(serializeToJs(item), ",");
    }
    parts2.push("]");
    return parts2.join("");
  }
  if (!isObject(value)) {
    throw new ImplementationError(`Cannot serialize a ${value.constructor.name}`);
  }
  const parts = ["{"];
  for (const key in value) {
    parts.push(key, ":", serializeToJs(value[key]), ",");
  }
  parts.push("}");
  return parts.join("");
}
function hex(number) {
  return number.toString(16);
}
((hex2) => {
  function fixed(value, width) {
    return hex2(value).padStart(width, "0");
  }
  hex2.fixed = fixed;
  function byte(value) {
    return fixed(value, 2);
  }
  hex2.byte = byte;
  function word(value) {
    return fixed(value, 4);
  }
  hex2.word = word;
})(hex || (hex = {}));
export {
  asJson,
  describeList,
  hex,
  serializeToJs
};
