// @matter/nodejs/dist/esm/storage/fs/FlatFileBlobStorageDriver.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  BaseStorageDriver,
  Bytes,
  FilesystemBlobStorageDriver,
  StorageError
} from "@matter/general";
import { createWriteStream, existsSync, openAsBlob } from "node:fs";
import { mkdir, readdir, rename, rm } from "node:fs/promises";
import { join } from "node:path";
class FlatFileBlobStorageDriver extends FilesystemBlobStorageDriver {
  static id = "file";
  static create(namespace, _descriptor) {
    return new FlatFileBlobStorageDriver(namespace);
  }
  #path;
  #initialized = false;
  constructor(namespace) {
    super(namespace);
    this.#path = this.root.directory.path;
  }
  get initialized() {
    return this.#initialized;
  }
  async initialize() {
    await super.initialize();
    await mkdir(this.#path, { recursive: true });
    this.#initialized = true;
  }
  async close() {
    this.#initialized = false;
    await super.close();
  }
  async openBlob(contexts, key) {
    const filePath = join(this.#path, buildStorageKey(contexts, key));
    if (!existsSync(filePath)) {
      return new Blob([]);
    }
    return await openAsBlob(filePath);
  }
  // Stream chunks to disk via atomic write-to-tmp + rename
  async writeBlobFromStream(contexts, key, stream) {
    const fileName = buildStorageKey(contexts, key);
    const filePath = join(this.#path, fileName);
    const tmpPath = `${filePath}.tmp`;
    const reader = stream.getReader();
    const writer = createWriteStream(tmpPath);
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        const chunk = Bytes.of(value);
        if (!writer.write(chunk)) {
          await new Promise((resolve) => writer.once("drain", resolve));
        }
      }
      writer.end();
      await new Promise((resolve, reject) => {
        writer.on("finish", resolve);
        writer.on("error", reject);
      });
      await rename(tmpPath, filePath);
    } catch (error) {
      writer.destroy();
      await rm(tmpPath, { force: true });
      throw error;
    } finally {
      reader.releaseLock();
    }
  }
  async delete(contexts, key) {
    const filePath = join(this.#path, buildStorageKey(contexts, key));
    await rm(filePath, { force: true });
  }
  async has(contexts, key) {
    const filePath = join(this.#path, buildStorageKey(contexts, key));
    return existsSync(filePath);
  }
  // When contexts is empty, prefixDot is "" and startsWith("") is always true — every file
  // matches.  The caller (keys/contexts/clearAll) then applies its own filtering on the remainder.
  async keys(contexts) {
    const prefixDot = contextPrefixDot(contexts);
    const result = new Array();
    for (const { remainder } of await this.#listDecodedFiles(prefixDot)) {
      if (remainder.length && !remainder.includes(".")) {
        result.push(remainder);
      }
    }
    return result;
  }
  async contexts(contexts) {
    const prefixDot = contextPrefixDot(contexts);
    const found = /* @__PURE__ */ new Set();
    for (const { remainder } of await this.#listDecodedFiles(prefixDot)) {
      const dotIndex = remainder.indexOf(".");
      if (dotIndex > 0) {
        found.add(remainder.substring(0, dotIndex));
      }
    }
    return [...found];
  }
  async clearAll(contexts) {
    const prefix = getContextBaseKey(contexts);
    const prefixDot = prefix.length ? `${prefix}.` : "";
    for (const { file, decoded } of await this.#listDecodedFiles("")) {
      if (decoded === prefix || decoded.startsWith(prefixDot)) {
        await rm(join(this.#path, file), { force: true });
      }
    }
  }
  /**
   * List all data files in the directory, decoded and filtered (excludes reserved files and
   * `.tmp` atomicity artifacts).  Returns raw filename + decoded name + remainder after prefix.
   */
  async #listDecodedFiles(prefixDot) {
    let files;
    try {
      files = await readdir(this.#path);
    } catch (e) {
      if (e.code === "ENOENT") {
        return [];
      }
      throw e;
    }
    const result = new Array();
    for (const file of files) {
      if (BaseStorageDriver.RESERVED_FILENAMES.has(file) || file.endsWith(".tmp")) {
        continue;
      }
      let decoded;
      try {
        decoded = decodeURIComponent(file);
      } catch {
        continue;
      }
      if (!decoded.startsWith(prefixDot)) {
        continue;
      }
      result.push({ file, decoded, remainder: decoded.substring(prefixDot.length) });
    }
    return result;
  }
}
function getContextBaseKey(contexts) {
  for (const ctx of contexts) {
    if (!ctx.length || ctx.includes(".")) {
      throw new StorageError("Context must not contain empty segments or leading or trailing dots.");
    }
  }
  return contexts.join(".");
}
function contextPrefixDot(contexts) {
  const prefix = getContextBaseKey(contexts);
  return prefix.length ? `${prefix}.` : "";
}
function buildStorageKey(contexts, key) {
  if (!key.length) {
    throw new StorageError("Key must not be an empty string.");
  }
  if (key.includes(".")) {
    throw new StorageError(`Key "${key}" must not contain "." (would be misinterpreted as context separator)`);
  }
  if (key.endsWith(".tmp")) {
    throw new StorageError(`Key "${key}" must not end with ".tmp" (reserved for atomic writes)`);
  }
  if (BaseStorageDriver.RESERVED_FILENAMES.has(key)) {
    throw new StorageError(`Key "${key}" is a reserved filename`);
  }
  const contextKey = getContextBaseKey(contexts);
  const rawName = contextKey.length ? `${contextKey}.${key}` : key;
  return encodeURIComponent(rawName).replace(/[!'()]/g, escape).replace(/\*/g, "%2A");
}
export {
  FlatFileBlobStorageDriver
};
