// @matter/node/dist/cjs/behavior/state/managed/ManagedReference.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ManagedReference_exports = {};
__export(ManagedReference_exports, {
  ManagedReference: () => ManagedReference
});
module.exports = __toCommonJS(ManagedReference_exports);
var import_protocol = require("@matter/protocol");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ManagedReference {
  primaryKey;
  parent;
  subrefs;
  owner;
  supervisionConfig;
  #key;
  #altKey;
  #assertWriteOk;
  #clone;
  #session;
  #expired = false;
  #location;
  #value;
  #dynamicContainer;
  /**
   * @param parent a reference to the container we reference
   * @param primaryKey the preferred key for lookup
   * @param name the name (in the case of structs) or index (in case of lists)
   * @param id the lookup ID in the case of structs
   * @param assertWriteOk enforces ACLs and read-only
   * @param clone clones the container prior to write; undefined if not transactional
   * @param session the access control session
   */
  constructor(parent, primaryKey, name, id, assertWriteOk, clone, session) {
    this.primaryKey = primaryKey;
    this.parent = parent;
    this.#assertWriteOk = assertWriteOk;
    this.#clone = clone;
    this.#session = session;
    this.#location = {
      ...parent.location,
      path: parent.location.path.at(name)
    };
    const key = primaryKey === "id" ? id ?? name : name;
    const altKey = primaryKey === "id" ? key === name ? void 0 : name : id;
    this.#key = key;
    this.#altKey = altKey;
    let dynamicContainer;
    if (parent.value[import_protocol.Val.properties]) {
      dynamicContainer = parent.value[import_protocol.Val.properties](parent.rootOwner, session);
      if (key in dynamicContainer) {
        this.#value = dynamicContainer[key];
      } else if (altKey !== void 0 && altKey in dynamicContainer) {
        this.#value = dynamicContainer[altKey];
      } else {
        dynamicContainer = void 0;
      }
    }
    this.#dynamicContainer = dynamicContainer;
    if (dynamicContainer === void 0) {
      if (key in parent.value) {
        this.#value = parent.value[key];
      } else if (altKey !== void 0) {
        this.#value = parent.value[altKey];
      }
    }
    if (parent.supervisionConfig) {
      this.supervisionConfig = parent.supervisionConfig.readonlyChild(key);
    }
    if (!parent.subrefs) {
      parent.subrefs = {};
    }
    parent.subrefs[key] = this;
  }
  get rootOwner() {
    return this.parent.rootOwner;
  }
  get value() {
    return this.#value;
  }
  set value(newValue) {
    if (this.#value === newValue) {
      return;
    }
    this.#assertWriteOk(newValue);
    this.#replaceValue(newValue);
    this.change(() => {
      if (this.#dynamicContainer) {
        this.#dynamicContainer[this.#key] = newValue;
        if (this.#altKey !== void 0 && this.#altKey in this.#dynamicContainer) {
          delete this.#dynamicContainer[this.#altKey];
        }
      } else {
        this.parent.value[this.#key] = newValue;
        if (this.#altKey !== void 0 && this.#altKey in this.parent.value) {
          delete this.parent.value[this.#altKey];
        }
      }
    });
  }
  get expired() {
    return this.#expired;
  }
  get location() {
    return this.#location;
  }
  set location(loc) {
    this.#location = loc;
  }
  get original() {
    if (!this.parent.original) {
      return void 0;
    }
    if (this.#dynamicContainer !== void 0) {
      const origProperties = this.parent.original[import_protocol.Val.properties](
        this.parent.rootOwner,
        this.#session
      );
      if (this.#key in origProperties) {
        return origProperties[this.#key];
      }
      if (this.#altKey !== void 0) {
        return origProperties[this.#altKey];
      }
    } else {
      if (this.#key in this.parent.original) {
        return this.parent.original[this.#key];
      }
      if (this.#altKey !== void 0) {
        return this.parent.original[this.#altKey];
      }
    }
  }
  change(mutator) {
    if (this.#expired) {
      throw new import_protocol.ExpiredReferenceError(this.location);
    }
    this.parent.change(() => {
      if (this.#clone && this.#value === this.original) {
        const newValue = this.#clone(this.#value);
        if (this.#dynamicContainer !== void 0) {
          this.#dynamicContainer[this.#key] = newValue;
          if (this.#altKey !== void 0 && this.#altKey in this.#dynamicContainer) {
            delete this.#dynamicContainer[this.#altKey];
          }
        } else {
          this.parent.value[this.#key] = newValue;
          if (this.#altKey !== void 0 && this.#altKey in this.parent.value) {
            delete this.parent.value[this.#altKey];
          }
        }
        this.#replaceValue(newValue);
      }
      mutator();
    });
  }
  refresh() {
    if (this.parent.expired) {
      this.#expired = true;
      return;
    }
    if (this.parent.value === void 0 || this.parent.value === null) {
      this.#expired = true;
      this.#replaceValue(void 0);
      return;
    }
    let value;
    if (this.#dynamicContainer !== void 0) {
      if (this.#key in this.#dynamicContainer) {
        value = this.#dynamicContainer[this.#key];
      } else if (this.#altKey !== void 0 && this.#altKey in this.#dynamicContainer) {
        value = this.#dynamicContainer[this.#altKey];
      }
    } else {
      if (this.#key in this.parent.value) {
        value = this.parent.value[this.#key];
      } else if (this.#altKey !== void 0 && this.#altKey in this.parent.value) {
        value = this.parent.value[this.#altKey];
      }
    }
    this.#replaceValue(value);
  }
  #replaceValue(newValue) {
    this.#value = newValue;
    const subrefs = this.subrefs;
    if (subrefs) {
      for (const key in subrefs) {
        subrefs[key].refresh();
      }
    }
  }
}
