// @matter/protocol/dist/cjs/ota/OtaImageReader.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OtaImageReader_exports = {};
__export(OtaImageReader_exports, {
  OtaImageError: () => OtaImageError,
  OtaImageReader: () => OtaImageReader
});
module.exports = __toCommonJS(OtaImageReader_exports);
var import_general = require("@matter/general");
var import_OtaImageHeader = require("./OtaImageHeader.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("OtaImageReader");
class OtaImageError extends import_general.MatterError {
  constructor(message, options) {
    super(message, options);
  }
}
class OtaImageReader {
  #crypto;
  #streamReader;
  #totalSize;
  #headerSize;
  #headerData;
  #fullFileChecksum;
  #fullFileChecksumType = "SHA-256";
  /** Read only the OTA image header from the stream and returns the  parsed header data. */
  static async header(streamReader) {
    const reader = new OtaImageReader(streamReader);
    await reader.#processHeader(false);
    if (reader.#headerData === void 0) {
      throw new import_general.InternalError("OTA header not read");
    }
    return reader.#headerData;
  }
  /** Read and validate the full OTA image file from the stream and returns the header data on success. */
  static async file(streamReader, crypto, expectedTotalSize, options) {
    const reader = new OtaImageReader(streamReader, crypto);
    if (options?.checksumType !== void 0) {
      reader.#fullFileChecksumType = options.checksumType;
    }
    const { headerBytes, remainingData } = await reader.#processHeader(options?.calculateFullChecksum ?? true);
    if (reader.#headerData === void 0) {
      throw new import_general.InternalError("OTA header not read");
    }
    await reader.#processPayloadDigest(remainingData, headerBytes, options?.calculateFullChecksum ?? true);
    if (options?.expectedChecksum && reader.#fullFileChecksum !== options.expectedChecksum) {
      throw new OtaImageError(
        `OTA full file checksum mismatch: expected "${options.expectedChecksum}", got "${reader.#fullFileChecksum}"`
      );
    }
    if (expectedTotalSize !== void 0) {
      expectedTotalSize = BigInt(expectedTotalSize);
      if (reader.totalSize !== expectedTotalSize) {
        if (options?.expectedChecksum) {
          logger.warn(
            `OTA file size mismatch, but checksum matched: expected ${expectedTotalSize}, got ${reader.totalSize}`
          );
        } else {
          throw new OtaImageError(
            `OTA file size mismatch: expected ${expectedTotalSize}, got ${reader.totalSize}`
          );
        }
      }
    }
    return reader.#headerData;
  }
  /**
   * Read and validate OTA file, extracting payload to a writable stream.
   * Returns the header information after successful validation and extraction.
   */
  static async extractPayload(streamReader, payloadWriter, crypto, expectedTotalSize) {
    const reader = new OtaImageReader(streamReader, crypto);
    const { remainingData } = await reader.#processHeader(false);
    if (reader.#headerData === void 0) {
      throw new import_general.InternalError("OTA header not read");
    }
    await reader.#extractAndValidatePayload(remainingData, payloadWriter);
    if (expectedTotalSize !== void 0) {
      expectedTotalSize = BigInt(expectedTotalSize);
      if (reader.totalSize !== expectedTotalSize) {
        throw new OtaImageError(
          `OTA file size mismatch: expected ${expectedTotalSize}, got ${reader.totalSize}`
        );
      }
    }
    return reader.#headerData;
  }
  constructor(streamReader, crypto) {
    this.#streamReader = streamReader;
    this.#crypto = crypto;
  }
  get totalSize() {
    return this.#totalSize;
  }
  /** Process and read the OTA image header from the stream. */
  async #processHeader(collectHeaderBytes = false) {
    if (this.#headerData !== void 0) {
      throw new import_general.InternalError("OTA header already read.");
    }
    let headerReader = void 0;
    const allHeaderBytes = [];
    while (true) {
      const { value, done } = await this.#streamReader.read();
      if (value === void 0 || done) {
        throw new OtaImageError("OTA file ended unexpectedly");
      }
      if (collectHeaderBytes) {
        allHeaderBytes.push(value);
      }
      if (headerReader !== void 0) {
        headerReader = new import_general.DataReader(import_general.Bytes.concat(headerReader.remainingBytes, value), import_general.Endian.Little);
      } else {
        headerReader = new import_general.DataReader(value, import_general.Endian.Little);
      }
      if (this.#headerSize === void 0 && headerReader.remainingBytesCount >= 16) {
        const fileIdentifier = headerReader.readUInt32();
        if (fileIdentifier !== 468644126) {
          throw new OtaImageError("Invalid OTA file identifier");
        }
        this.#totalSize = headerReader.readUInt64();
        this.#headerSize = headerReader.readUInt32();
      }
      if (this.#headerData === void 0 && this.#headerSize !== void 0 && headerReader.remainingBytesCount >= this.#headerSize) {
        const data = headerReader.readByteArray(this.#headerSize);
        this.#headerData = import_OtaImageHeader.TlvOtaImageHeader.decode(data);
        const remainingBytes = headerReader.remainingBytes;
        let headerBytes = void 0;
        if (collectHeaderBytes && allHeaderBytes.length > 0) {
          const totalHeaderAndRemaining = import_general.Bytes.of(import_general.Bytes.concat(...allHeaderBytes));
          const headerOnlyLength = totalHeaderAndRemaining.byteLength - remainingBytes.byteLength;
          headerBytes = totalHeaderAndRemaining.subarray(0, headerOnlyLength);
        }
        return { headerBytes, remainingData: remainingBytes };
      }
    }
  }
  /** Process and validate the OTA image payload digest from the stream. */
  async #processPayloadDigest(initialPayloadData, headerBytes, calculateFullChecksum = false) {
    if (this.#headerData === void 0) {
      throw new import_general.InternalError("OTA header not read");
    }
    if (this.#crypto === void 0) {
      throw new import_general.InternalError("No crypto implementation provided");
    }
    const { imageDigestType, imageDigest, payloadSize } = this.#headerData;
    let readPayloadSize = 0n;
    const streamReader = this.#streamReader;
    const payloadIterator = async function* (headerBytes2) {
      if (headerBytes2 !== void 0) {
        yield headerBytes2;
      }
      if (initialPayloadData !== void 0) {
        readPayloadSize += BigInt(initialPayloadData.byteLength);
        yield initialPayloadData;
      }
      while (true) {
        const { value, done } = await streamReader.read();
        if (value === void 0 || done) {
          break;
        }
        readPayloadSize += BigInt(value.byteLength);
        yield value;
      }
    };
    if (calculateFullChecksum && headerBytes !== void 0) {
      const fullChecksum = await this.#crypto.computeHash(
        payloadIterator(headerBytes),
        this.#fullFileChecksumType
      );
      this.#fullFileChecksum = import_general.Bytes.toBase64(import_general.Bytes.of(fullChecksum));
      if (readPayloadSize !== BigInt(payloadSize)) {
        throw new OtaImageError(`OTA payload size mismatch: expected ${payloadSize}, got ${readPayloadSize}`);
      }
    } else {
      const hashBytes = await this.#crypto.computeHash(
        payloadIterator(),
        import_general.HashFipsAlgorithmId[imageDigestType]
      );
      if (readPayloadSize !== BigInt(payloadSize)) {
        throw new OtaImageError(`OTA payload size mismatch: expected ${payloadSize}, got ${readPayloadSize}`);
      }
      if (!import_general.Bytes.areEqual(hashBytes, imageDigest)) {
        throw new OtaImageError(
          `OTA file digest mismatch "${import_general.Bytes.toHex(hashBytes)}" vs. "${import_general.Bytes.toHex(imageDigest)}"`
        );
      }
    }
  }
  /** Extract the OTA image payload to a writable stream while validating its digest. */
  async #extractAndValidatePayload(initialPayloadData, payloadWriter) {
    if (this.#headerData === void 0) {
      throw new import_general.InternalError("OTA header not read");
    }
    if (this.#crypto === void 0) {
      throw new import_general.InternalError("No crypto implementation provided");
    }
    const { imageDigestType, imageDigest, payloadSize } = this.#headerData;
    if (imageDigestType !== 1) {
      throw new import_general.InternalError(`Unsupported image digest type ${imageDigestType}`);
    }
    let readPayloadSize = 0n;
    const streamReader = this.#streamReader;
    const iterator = async function* () {
      if (initialPayloadData !== void 0) {
        readPayloadSize += BigInt(initialPayloadData.byteLength);
        await payloadWriter.write(initialPayloadData);
        yield initialPayloadData;
      }
      while (true) {
        const { value, done } = await streamReader.read();
        if (value === void 0 || done) {
          break;
        }
        readPayloadSize += BigInt(value.byteLength);
        await payloadWriter.write(value);
        yield value;
      }
    };
    const hashBytes = await this.#crypto.computeHash(
      iterator(),
      import_general.HashFipsAlgorithmId[imageDigestType]
    );
    if (readPayloadSize !== BigInt(payloadSize)) {
      throw new OtaImageError(`OTA payload size mismatch: expected ${payloadSize}, got ${readPayloadSize}`);
    }
    if (!import_general.Bytes.areEqual(hashBytes, imageDigest)) {
      throw new OtaImageError(
        `OTA file digest mismatch "${import_general.Bytes.toHex(hashBytes)}" vs. "${import_general.Bytes.toHex(imageDigest)}"`
      );
    }
    await payloadWriter.close();
  }
}
