// homebridge/dist/api.js
import { EventEmitter } from 'node:events';
import hapNodeJs from '@homebridge/hap-nodejs';
import semver from 'semver';
import { Logger } from './logger.js';
import { PlatformAccessory } from './platformAccessory.js';
import { PluginManager } from './pluginManager.js';
import { User } from './user.js';
import getVersion from './version.js';
const log = Logger.internal;
export var PluginType;
(function (PluginType) {
    PluginType["ACCESSORY"] = "accessory";
    PluginType["PLATFORM"] = "platform";
})(PluginType || (PluginType = {}));
export var APIEvent;
(function (APIEvent) {
    APIEvent["DID_FINISH_LAUNCHING"] = "didFinishLaunching";
    APIEvent["SHUTDOWN"] = "shutdown";
})(APIEvent || (APIEvent = {}));
export var InternalAPIEvent;
(function (InternalAPIEvent) {
    InternalAPIEvent["REGISTER_ACCESSORY"] = "registerAccessory";
    InternalAPIEvent["REGISTER_PLATFORM"] = "registerPlatform";
    InternalAPIEvent["PUBLISH_EXTERNAL_ACCESSORIES"] = "publishExternalAccessories";
    InternalAPIEvent["REGISTER_PLATFORM_ACCESSORIES"] = "registerPlatformAccessories";
    InternalAPIEvent["UPDATE_PLATFORM_ACCESSORIES"] = "updatePlatformAccessories";
    InternalAPIEvent["UNREGISTER_PLATFORM_ACCESSORIES"] = "unregisterPlatformAccessories";
    InternalAPIEvent["PUBLISH_EXTERNAL_MATTER_ACCESSORIES"] = "publishExternalMatterAccessories";
    InternalAPIEvent["REGISTER_MATTER_PLATFORM_ACCESSORIES"] = "registerMatterPlatformAccessories";
    InternalAPIEvent["UPDATE_MATTER_PLATFORM_ACCESSORIES"] = "updateMatterPlatformAccessories";
    InternalAPIEvent["UNREGISTER_MATTER_PLATFORM_ACCESSORIES"] = "unregisterMatterPlatformAccessories";
    InternalAPIEvent["UNREGISTER_EXTERNAL_MATTER_ACCESSORIES"] = "unregisterExternalMatterAccessories";
    InternalAPIEvent["UPDATE_MATTER_ACCESSORY_STATE"] = "updateMatterAccessoryState";
})(InternalAPIEvent || (InternalAPIEvent = {}));
export class HomebridgeAPI extends EventEmitter {
    version = 2.7;
    serverVersion = getVersion();
    user = User;
    hap = hapNodeJs;
    hapLegacyTypes = hapNodeJs.LegacyTypes;
    platformAccessory = PlatformAccessory;
    matterEnabled = false;
    _matterServer = null;
    _matterManager = null;
    _pendingExternalRegistrations = new Map();
    _matterAPI;
    get matter() {
        if (!this._matterAPI) {
            throw new Error('Matter API must be pre-loaded before access. '
                + 'Call await api.loadMatterAPI() in your plugin constructor or access api.matter in didFinishLaunching().');
        }
        return this._matterAPI;
    }
    async loadMatterAPI() {
        if (!this._matterAPI) {
            const { MatterAPIImpl } = await import('./matter/MatterAPIImpl.js');
            this._matterAPI = new MatterAPIImpl(this);
        }
    }
    constructor() {
        super();
    }
    _setMatterEnabled(enabled) {
        this.matterEnabled = enabled;
    }
    _setMatterServer(server) {
        this._matterServer = server;
    }
    _setMatterManager(manager) {
        this._matterManager = manager;
    }
    _resolveExternalRegistration(registrationId) {
        const resolve = this._pendingExternalRegistrations.get(registrationId);
        if (resolve) {
            resolve();
            this._pendingExternalRegistrations.delete(registrationId);
        }
    }
    versionGreaterOrEqual(version) {
        return semver.gte(this.serverVersion, version);
    }
    static isDynamicPlatformPlugin(platformPlugin) {
        return 'configureAccessory' in platformPlugin;
    }
    static isStaticPlatformPlugin(platformPlugin) {
        return 'accessories' in platformPlugin;
    }
    signalFinished() {
        this.emit("didFinishLaunching");
    }
    signalShutdown() {
        this.emit("shutdown");
    }
    registerAccessory(pluginIdentifier, accessoryName, constructor) {
        if (typeof accessoryName === 'function') {
            constructor = accessoryName;
            accessoryName = pluginIdentifier;
            this.emit("registerAccessory", accessoryName, constructor);
        }
        else {
            this.emit("registerAccessory", accessoryName, constructor, pluginIdentifier);
        }
    }
    registerPlatform(pluginIdentifier, platformName, constructor) {
        if (typeof platformName === 'function') {
            constructor = platformName;
            platformName = pluginIdentifier;
            this.emit("registerPlatform", platformName, constructor);
        }
        else {
            this.emit("registerPlatform", platformName, constructor, pluginIdentifier);
        }
    }
    publishCameraAccessories(pluginIdentifier, accessories) {
        this.publishExternalAccessories(pluginIdentifier, accessories);
    }
    publishExternalAccessories(pluginIdentifier, accessories) {
        if (!PluginManager.isQualifiedPluginIdentifier(pluginIdentifier)) {
            log.info(`One of your plugins incorrectly registered an external accessory using the platform name (${pluginIdentifier}) and not the plugin identifier. Please report this to the developer!`);
        }
        accessories.forEach((accessory) => {
            if (!(accessory instanceof PlatformAccessory)) {
                throw new TypeError(`${pluginIdentifier} attempt to register an accessory that isn't PlatformAccessory!`);
            }
            accessory._associatedPlugin = pluginIdentifier;
        });
        this.emit("publishExternalAccessories", accessories);
    }
    registerPlatformAccessories(pluginIdentifier, platformName, accessories) {
        accessories.forEach((accessory) => {
            if (!(accessory instanceof PlatformAccessory)) {
                throw new TypeError(`${pluginIdentifier} - ${platformName} attempt to register an accessory that isn't PlatformAccessory!`);
            }
            accessory._associatedPlugin = pluginIdentifier;
            accessory._associatedPlatform = platformName;
        });
        this.emit("registerPlatformAccessories", accessories);
    }
    updatePlatformAccessories(accessories) {
        this.emit("updatePlatformAccessories", accessories);
    }
    unregisterPlatformAccessories(pluginIdentifier, platformName, accessories) {
        accessories.forEach((accessory) => {
            if (!(accessory instanceof PlatformAccessory)) {
                throw new TypeError(`${pluginIdentifier} - ${platformName} attempt to unregister an accessory that isn't PlatformAccessory!`);
            }
        });
        this.emit("unregisterPlatformAccessories", accessories);
    }
    isMatterAvailable() {
        return semver.gte(this.serverVersion, '2.0.0-alpha.0');
    }
    isMatterEnabled() {
        return this.matterEnabled;
    }
}
