// homebridge/dist/bridgeService.js
import { Accessory, Bridge, Characteristic, HAPLibraryVersion, once, Service, uuid, } from '@homebridge/hap-nodejs';
import { getLogPrefix, Logger } from './logger.js';
import { PlatformAccessory } from './platformAccessory.js';
import { PluginManager } from './pluginManager.js';
import { StorageService } from './storageService.js';
import { generate } from './util/mac.js';
import getVersion from './version.js';
export const DEFAULT_BRIDGE_DEFAULTS = {
    vendorName: 'Homebridge',
    manufacturer: 'homebridge.io',
    model: 'homebridge',
};
const log = Logger.internal;
export class BridgeService {
    api;
    pluginManager;
    externalPortService;
    bridgeOptions;
    bridgeConfig;
    bridge;
    storageService;
    allowInsecureAccess;
    cachedPlatformAccessories = [];
    cachedAccessoriesFileLoaded = false;
    publishedExternalAccessories = new Map();
    constructor(api, pluginManager, externalPortService, bridgeOptions, bridgeConfig) {
        this.api = api;
        this.pluginManager = pluginManager;
        this.externalPortService = externalPortService;
        this.bridgeOptions = bridgeOptions;
        this.bridgeConfig = bridgeConfig;
        this.storageService = new StorageService(this.bridgeOptions.cachedAccessoriesDir);
        this.storageService.initSync();
        this.allowInsecureAccess = this.bridgeOptions.insecureAccess || false;
        this.api.on("registerPlatformAccessories", this.handleRegisterPlatformAccessories.bind(this));
        this.api.on("updatePlatformAccessories", this.handleUpdatePlatformAccessories.bind(this));
        this.api.on("unregisterPlatformAccessories", this.handleUnregisterPlatformAccessories.bind(this));
        this.api.on("publishExternalAccessories", this.handlePublishExternalAccessories.bind(this));
        this.bridge = new Bridge(bridgeConfig.name, uuid.generate('HomeBridge'));
        this.bridge.on("characteristic-warning", () => {
        });
    }
    static printCharacteristicWriteWarning(plugin, accessory, opts, warning) {
        const wikiInfo = 'See https://homebridge.io/w/JtMGR for more info.';
        switch (warning.type) {
            case "slow-read":
            case "slow-write":
                if (!opts.ignoreSlow) {
                    log.info(getLogPrefix(plugin.getPluginIdentifier()), 'This plugin slows down Homebridge.', warning.message, wikiInfo);
                }
                break;
            case "timeout-read":
            case "timeout-write":
                log.error(getLogPrefix(plugin.getPluginIdentifier()), 'This plugin slows down Homebridge.', warning.message, wikiInfo);
                break;
            case "warn-message":
                log.info(getLogPrefix(plugin.getPluginIdentifier()), `This plugin generated a warning from the characteristic '${warning.characteristic.displayName}':`, `${warning.message}.`, wikiInfo);
                break;
            case "error-message":
                log.error(getLogPrefix(plugin.getPluginIdentifier()), `This plugin threw an error from the characteristic '${warning.characteristic.displayName}':`, `${warning.message}.`, wikiInfo);
                break;
            case "debug-message":
                log.debug(getLogPrefix(plugin.getPluginIdentifier()), `Characteristic '${warning.characteristic.displayName}':`, `${warning.message}.`, wikiInfo);
                break;
            default:
                log.info(getLogPrefix(plugin.getPluginIdentifier()), `This plugin generated a warning from the characteristic '${warning.characteristic.displayName}':`, `${warning.message}.`, wikiInfo);
                break;
        }
        if (warning.stack) {
            log.debug(getLogPrefix(plugin.getPluginIdentifier()), warning.stack);
        }
    }
    publishBridge() {
        const bridgeConfig = this.bridgeConfig;
        const info = this.bridge.getService(Service.AccessoryInformation);
        info.setCharacteristic(Characteristic.Manufacturer, bridgeConfig.manufacturer || DEFAULT_BRIDGE_DEFAULTS.manufacturer);
        info.setCharacteristic(Characteristic.Model, bridgeConfig.model || DEFAULT_BRIDGE_DEFAULTS.model);
        info.setCharacteristic(Characteristic.SerialNumber, bridgeConfig.serialNumber || bridgeConfig.username);
        info.setCharacteristic(Characteristic.FirmwareRevision, bridgeConfig.firmwareRevision || getVersion());
        this.bridge.on("listening", (port) => {
            log.success('Homebridge v%s (HAP v%s) (%s) is running on port %s.', getVersion(), HAPLibraryVersion(), bridgeConfig.name, port);
        });
        const publishInfo = {
            username: bridgeConfig.username,
            port: bridgeConfig.port,
            pincode: bridgeConfig.pin,
            category: 2,
            bind: bridgeConfig.bind,
            addIdentifyingMaterial: true,
            advertiser: bridgeConfig.advertiser,
        };
        if (bridgeConfig.setupID && bridgeConfig.setupID.length === 4) {
            publishInfo.setupID = bridgeConfig.setupID;
        }
        log.debug('Publishing bridge accessory (name: %s, publishInfo: %o).', this.bridge.displayName, BridgeService.strippingPinCode(publishInfo));
        void this.bridge.publish(publishInfo, this.allowInsecureAccess);
    }
    async loadCachedPlatformAccessoriesFromDisk() {
        let cachedAccessories = null;
        try {
            cachedAccessories = await this.storageService.getItem(this.bridgeOptions.cachedAccessoriesItemName);
        }
        catch (error) {
            log.error('Failed to load cached accessories from disk:', error.message);
            if (error instanceof SyntaxError) {
                cachedAccessories = await this.restoreCachedAccessoriesBackup();
            }
            else {
                log.error('Not restoring cached accessories - some accessories may be reset.');
            }
        }
        if (cachedAccessories) {
            log.info(`Loaded ${cachedAccessories.length} cached accessories from ${this.bridgeOptions.cachedAccessoriesItemName}.`);
            this.cachedPlatformAccessories = cachedAccessories.map((serialized) => {
                return PlatformAccessory.deserialize(serialized);
            });
            if (cachedAccessories.length) {
                await this.createCachedAccessoriesBackup();
            }
        }
        this.cachedAccessoriesFileLoaded = true;
    }
    get backupCacheFileName() {
        return `.${this.bridgeOptions.cachedAccessoriesItemName}.bak`;
    }
    async createCachedAccessoriesBackup() {
        try {
            await this.storageService.copyItem(this.bridgeOptions.cachedAccessoriesItemName, this.backupCacheFileName);
        }
        catch (error) {
            log.warn(`Failed to create a backup of the ${this.bridgeOptions.cachedAccessoriesItemName} cached accessories file:`, error.message);
        }
    }
    async restoreCachedAccessoriesBackup() {
        try {
            const cachedAccessories = await this.storageService.getItem(this.backupCacheFileName);
            if (cachedAccessories && cachedAccessories.length) {
                log.warn(`Recovered ${cachedAccessories.length} accessories from ${this.bridgeOptions.cachedAccessoriesItemName} cache backup.`);
            }
            return cachedAccessories;
        }
        catch (error) {
            return null;
        }
    }
    restoreCachedPlatformAccessories() {
        this.cachedPlatformAccessories = this.cachedPlatformAccessories.filter((accessory) => {
            let plugin = this.pluginManager.getPlugin(accessory._associatedPlugin);
            if (!plugin) {
                try {
                    plugin = this.pluginManager.getPluginByActiveDynamicPlatform(accessory._associatedPlatform);
                    if (plugin) {
                        log.info(`When searching for the associated plugin of the accessory '${accessory.displayName}' `
                            + `it seems like the plugin name changed from '${accessory._associatedPlugin}' to '${plugin.getPluginIdentifier()}'. Plugin association is now being transformed!`);
                        accessory._associatedPlugin = plugin.getPluginIdentifier();
                    }
                }
                catch (error) {
                    log.info(`Could not find the associated plugin for the accessory '${accessory.displayName}'. `
                        + `Tried to find the plugin by the platform name but ${error.message}`);
                }
            }
            const platformPlugins = plugin && plugin.getActiveDynamicPlatform(accessory._associatedPlatform);
            if (plugin) {
                accessory._associatedHAPAccessory.on("characteristic-warning", BridgeService.printCharacteristicWriteWarning.bind(this, plugin, accessory._associatedHAPAccessory, {}));
            }
            if (!platformPlugins) {
                log.info(`Failed to find plugin to handle accessory ${accessory._associatedHAPAccessory.displayName}`);
                if (!this.bridgeOptions.keepOrphanedCachedAccessories) {
                    log.info(`Removing orphaned accessory ${accessory._associatedHAPAccessory.displayName}`);
                    return false;
                }
            }
            else {
                accessory.getService(Service.AccessoryInformation)?.setCharacteristic(Characteristic.FirmwareRevision, '0');
                platformPlugins.configureAccessory(accessory);
            }
            try {
                this.bridge.addBridgedAccessory(accessory._associatedHAPAccessory);
            }
            catch (error) {
                log.warn(`${accessory._associatedPlugin ? getLogPrefix(accessory._associatedPlugin) : ''} Could not restore cached accessory '${accessory._associatedHAPAccessory.displayName}':`, error.message);
                return false;
            }
            return true;
        });
    }
    saveCachedPlatformAccessoriesOnDisk() {
        try {
            if (this.cachedAccessoriesFileLoaded) {
                const serializedAccessories = this.cachedPlatformAccessories.map(accessory => PlatformAccessory.serialize(accessory));
                this.storageService.setItemSync(this.bridgeOptions.cachedAccessoriesItemName, serializedAccessories);
            }
        }
        catch (error) {
            log.error('Failed to save cached accessories to disk:', error.message);
            log.error('Your accessories will not persist between restarts until this issue is resolved.');
        }
    }
    handleRegisterPlatformAccessories(accessories) {
        const hapAccessories = accessories.map((accessory) => {
            const existingAccessory = this.cachedPlatformAccessories.find(cached => cached._associatedHAPAccessory.UUID === accessory._associatedHAPAccessory.UUID);
            if (existingAccessory) {
                log.warn('Accessory \'%s\' has the same UUID as existing accessory \'%s\' (UUID: %s). Skipping duplicate.', accessory.displayName, existingAccessory.displayName, accessory._associatedHAPAccessory.UUID);
                return undefined;
            }
            this.cachedPlatformAccessories.push(accessory);
            const plugin = this.pluginManager.getPlugin(accessory._associatedPlugin);
            if (plugin) {
                const platforms = plugin.getActiveDynamicPlatform(accessory._associatedPlatform);
                if (!platforms) {
                    log.warn('The plugin \'%s\' registered a new accessory for the platform \'%s\'. The platform couldn\'t be found though!', accessory._associatedPlugin, accessory._associatedPlatform);
                }
                accessory._associatedHAPAccessory.on("characteristic-warning", BridgeService.printCharacteristicWriteWarning.bind(this, plugin, accessory._associatedHAPAccessory, {}));
            }
            else {
                log.warn('A platform configured a new accessory under the plugin name \'%s\'. However no loaded plugin could be found for the name!', accessory._associatedPlugin);
            }
            return accessory._associatedHAPAccessory;
        }).filter((hapAccessory) => hapAccessory !== undefined);
        this.bridge.addBridgedAccessories(hapAccessories);
        this.saveCachedPlatformAccessoriesOnDisk();
    }
    handleUpdatePlatformAccessories(accessories) {
        if (!Array.isArray(accessories)) {
            return;
        }
        const nonUpdatedPlugins = this.cachedPlatformAccessories.filter(cachedPlatformAccessory => (!accessories.some(accessory => accessory.UUID === cachedPlatformAccessory._associatedHAPAccessory.UUID)));
        this.cachedPlatformAccessories = [...nonUpdatedPlugins, ...accessories];
        this.saveCachedPlatformAccessoriesOnDisk();
    }
    handleUnregisterPlatformAccessories(accessories) {
        const hapAccessories = accessories.map((accessory) => {
            const index = this.cachedPlatformAccessories.indexOf(accessory);
            if (index >= 0) {
                this.cachedPlatformAccessories.splice(index, 1);
            }
            return accessory._associatedHAPAccessory;
        });
        this.bridge.removeBridgedAccessories(hapAccessories);
        this.saveCachedPlatformAccessoriesOnDisk();
    }
    async handlePublishExternalAccessories(accessories) {
        const accessoryPin = this.bridgeConfig.pin;
        for (const accessory of accessories) {
            const hapAccessory = accessory._associatedHAPAccessory;
            const advertiseAddress = generate(hapAccessory.UUID);
            const accessoryPort = await this.externalPortService.requestPort(advertiseAddress);
            if (this.publishedExternalAccessories.has(advertiseAddress)) {
                throw new Error(`Accessory ${hapAccessory.displayName} experienced an address collision.`);
            }
            else {
                this.publishedExternalAccessories.set(advertiseAddress, accessory);
            }
            const plugin = this.pluginManager.getPlugin(accessory._associatedPlugin);
            if (plugin) {
                hapAccessory.on("characteristic-warning", BridgeService.printCharacteristicWriteWarning.bind(this, plugin, hapAccessory, { ignoreSlow: true }));
            }
            else if (PluginManager.isQualifiedPluginIdentifier(accessory._associatedPlugin)) {
                log.warn('A platform configured a external accessory under the plugin name \'%s\'. However no loaded plugin could be found for the name!', accessory._associatedPlugin);
            }
            hapAccessory.on("listening", (port) => {
                log.success('%s is running on port %s.', hapAccessory.displayName, port);
                log.info('Please add [%s] manually in Home app. Setup Code: %s', hapAccessory.displayName, accessoryPin);
            });
            const publishInfo = {
                username: advertiseAddress,
                pincode: accessoryPin,
                category: accessory.category,
                port: accessoryPort,
                bind: this.bridgeConfig.bind,
                addIdentifyingMaterial: true,
                advertiser: this.bridgeConfig.advertiser,
            };
            log.debug('Publishing external accessory (name: %s, publishInfo: %o).', hapAccessory.displayName, BridgeService.strippingPinCode(publishInfo));
            void hapAccessory.publish(publishInfo, this.allowInsecureAccess);
        }
    }
    createHAPAccessory(plugin, accessoryInstance, displayName, accessoryType, uuidBase) {
        const services = (accessoryInstance.getServices() || [])
            .filter(service => !!service);
        const controllers = ((accessoryInstance.getControllers && accessoryInstance.getControllers()) || [])
            .filter(controller => !!controller);
        if (services.length === 0 && controllers.length === 0) {
            return undefined;
        }
        const accessoryUUID = uuid.generate(`${accessoryType}:${uuidBase || displayName}`);
        const accessory = new Accessory(displayName, accessoryUUID);
        if (accessoryInstance.identify) {
            accessory.on("identify", (paired, callback) => {
                accessoryInstance.identify(() => { });
                callback();
            });
        }
        const informationService = accessory.getService(Service.AccessoryInformation);
        services.forEach((service) => {
            if (service instanceof Service.AccessoryInformation) {
                service.setCharacteristic(Characteristic.Name, displayName);
                service.getCharacteristic(Characteristic.Identify).removeAllListeners("set");
                informationService.replaceCharacteristicsFromService(service);
            }
            else {
                accessory.addService(service);
            }
        });
        accessory.on("characteristic-warning", BridgeService.printCharacteristicWriteWarning.bind(this, plugin, accessory, {}));
        controllers.forEach((controller) => {
            accessory.configureController(controller);
        });
        return accessory;
    }
    async loadPlatformAccessories(plugin, platformInstance, platformType, logger) {
        return new Promise((resolve) => {
            const loadDelayWarningInterval = setInterval(() => {
                log.warn(getLogPrefix(plugin.getPluginIdentifier()), 'This plugin is taking long time to load and preventing Homebridge from starting. See https://homebridge.io/w/JtMGR for more info.');
            }, 20000);
            platformInstance.accessories(once((accessories) => {
                clearInterval(loadDelayWarningInterval);
                accessories.forEach((accessoryInstance, index) => {
                    const accessoryName = accessoryInstance.name;
                    const uuidBase = accessoryInstance.uuid_base;
                    log.info('Initializing platform accessory \'%s\'...', accessoryName);
                    const accessory = this.createHAPAccessory(plugin, accessoryInstance, accessoryName, platformType, uuidBase);
                    if (accessory) {
                        this.bridge.addBridgedAccessory(accessory);
                    }
                    else {
                        logger('Platform %s returned an accessory at index %d with an empty set of services. Won\'t adding it to the bridge!', platformType, index);
                    }
                });
                resolve();
            }));
        });
    }
    teardown() {
        void this.bridge.unpublish();
        for (const accessory of this.publishedExternalAccessories.values()) {
            void accessory._associatedHAPAccessory.unpublish();
        }
        this.saveCachedPlatformAccessoriesOnDisk();
        this.api.signalShutdown();
    }
    static strippingPinCode(publishInfo) {
        const info = {
            ...publishInfo,
        };
        info.pincode = '***-**-***';
        return info;
    }
}
