// homebridge/dist/childBridgeService.js
import { fork } from 'node:child_process';
import { dirname, resolve } from 'node:path';
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import fs from 'fs-extra';
import { Logger } from './logger.js';
import { User } from './user.js';
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const COLON_RE = /:/g;
export var ChildProcessMessageEventType;
(function (ChildProcessMessageEventType) {
    ChildProcessMessageEventType["READY"] = "ready";
    ChildProcessMessageEventType["LOAD"] = "load";
    ChildProcessMessageEventType["LOADED"] = "loaded";
    ChildProcessMessageEventType["START"] = "start";
    ChildProcessMessageEventType["ONLINE"] = "online";
    ChildProcessMessageEventType["PORT_REQUEST"] = "portRequest";
    ChildProcessMessageEventType["PORT_ALLOCATED"] = "portAllocated";
    ChildProcessMessageEventType["STATUS_UPDATE"] = "status";
    ChildProcessMessageEventType["START_MATTER_MONITORING"] = "startMatterMonitoring";
    ChildProcessMessageEventType["STOP_MATTER_MONITORING"] = "stopMatterMonitoring";
    ChildProcessMessageEventType["GET_MATTER_ACCESSORIES"] = "getMatterAccessories";
    ChildProcessMessageEventType["GET_MATTER_ACCESSORY_INFO"] = "getMatterAccessoryInfo";
    ChildProcessMessageEventType["MATTER_ACCESSORY_CONTROL"] = "matterAccessoryControl";
    ChildProcessMessageEventType["MATTER_EVENT"] = "matterEvent";
})(ChildProcessMessageEventType || (ChildProcessMessageEventType = {}));
export var ChildBridgeStatus;
(function (ChildBridgeStatus) {
    ChildBridgeStatus["PENDING"] = "pending";
    ChildBridgeStatus["OK"] = "ok";
    ChildBridgeStatus["DOWN"] = "down";
})(ChildBridgeStatus || (ChildBridgeStatus = {}));
export class ChildBridgeService {
    type;
    identifier;
    plugin;
    bridgeConfig;
    homebridgeConfig;
    homebridgeOptions;
    api;
    ipcService;
    externalPortService;
    child;
    args = [];
    processEnv = {};
    shuttingDown = false;
    lastBridgeStatus = "pending";
    pairedStatus = null;
    manuallyStopped = false;
    setupUri = null;
    matterCommissioningInfo;
    pluginConfig = [];
    log;
    displayName;
    restartCount = 0;
    maxRestarts = 4;
    scheduledRestartTimeout;
    matterAccessoriesResolve;
    onExternalBridgeRegistered;
    constructor(type, identifier, plugin, bridgeConfig, homebridgeConfig, homebridgeOptions, api, ipcService, externalPortService) {
        this.type = type;
        this.identifier = identifier;
        this.plugin = plugin;
        this.bridgeConfig = bridgeConfig;
        this.homebridgeConfig = homebridgeConfig;
        this.homebridgeOptions = homebridgeOptions;
        this.api = api;
        this.ipcService = ipcService;
        this.externalPortService = externalPortService;
        this.log = Logger.withPrefix(this.plugin.getPluginIdentifier());
        this.api.on('shutdown', () => {
            this.shuttingDown = true;
            this.teardown();
        });
        this.api.setMaxListeners(this.api.getMaxListeners() + 1);
    }
    start() {
        this.setProcessFlags();
        this.setProcessEnv();
        this.startChildProcess();
        if (this.pluginConfig.length !== 1) {
            this.displayName = this.plugin.getPluginIdentifier();
        }
        else {
            this.displayName = this.pluginConfig[0]?.name || this.plugin.getPluginIdentifier();
        }
        this.log = Logger.withPrefix(this.displayName);
    }
    addConfig(config) {
        this.pluginConfig.push(config);
    }
    startMatterMonitoring() {
        this.sendMessage("startMatterMonitoring");
    }
    stopMatterMonitoring() {
        this.sendMessage("stopMatterMonitoring");
    }
    requestMatterAccessories(timeoutMs = 500) {
        return new Promise((resolve) => {
            const timeout = setTimeout(() => {
                this.matterAccessoriesResolve = undefined;
                resolve(undefined);
            }, timeoutMs);
            this.matterAccessoriesResolve = (data) => {
                clearTimeout(timeout);
                this.matterAccessoriesResolve = undefined;
                resolve(data);
            };
            this.sendMessage("getMatterAccessories");
        });
    }
    getMatterAccessoryInfo(uuid) {
        this.sendMessage("getMatterAccessoryInfo", { uuid });
    }
    controlMatterAccessory(data) {
        this.sendMessage("matterAccessoryControl", data);
    }
    get bridgeStatus() {
        return this.lastBridgeStatus;
    }
    set bridgeStatus(value) {
        this.lastBridgeStatus = value;
        this.sendStatusUpdate();
    }
    startChildProcess() {
        this.bridgeStatus = "pending";
        this.child = fork(resolve(__dirname, 'childBridgeFork.js'), this.args, this.processEnv);
        this.child.stdout?.on('data', (data) => {
            process.stdout.write(data);
        });
        this.child.stderr?.on('data', (data) => {
            process.stderr.write(data);
        });
        this.child.on('error', (e) => {
            this.bridgeStatus = "down";
            this.log.error('Child bridge process error', e);
        });
        this.child.once('close', (code, signal) => {
            this.handleProcessClose(code, signal);
        });
        this.child.on('message', (message) => {
            if (typeof message !== 'object' || !message.id) {
                return;
            }
            switch (message.id) {
                case "ready": {
                    this.log(`Child bridge starting${this.child?.pid ? ` (pid ${this.child.pid})` : ''}...`);
                    this.loadPlugin();
                    break;
                }
                case "loaded": {
                    const version = message.data.version;
                    if (this.pluginConfig.length > 1) {
                        this.log.success(`Child bridge started successfully with ${this.pluginConfig.length} accessories (plugin v${version}).`);
                    }
                    else {
                        this.log.success(`Child bridge started successfully (plugin v${version}).`);
                    }
                    this.startBridge();
                    break;
                }
                case "online": {
                    this.bridgeStatus = "ok";
                    break;
                }
                case "portRequest": {
                    void this.handlePortRequest(message.data);
                    break;
                }
                case "status": {
                    const statusData = message.data;
                    this.pairedStatus = statusData.paired;
                    this.setupUri = statusData.setupUri;
                    if (statusData.matter) {
                        this.matterCommissioningInfo = {
                            qrCode: statusData.matter.qrCode,
                            manualPairingCode: statusData.matter.manualPairingCode,
                            serialNumber: statusData.matter.serialNumber,
                            commissioned: statusData.matter.commissioned || false,
                            deviceCount: statusData.matter.deviceCount,
                        };
                    }
                    this.sendStatusUpdate();
                    break;
                }
                case "matterEvent": {
                    const matterEvent = message.data;
                    if (matterEvent.type === 'accessoriesData') {
                        this.matterAccessoriesResolve?.(matterEvent.data);
                    }
                    else if (matterEvent.type === 'externalBridgeRegistration') {
                        const data = matterEvent.data;
                        if (this.onExternalBridgeRegistered) {
                            this.onExternalBridgeRegistered(data.externalBridgeUsername, this.bridgeConfig.username);
                        }
                    }
                    else {
                        this.ipcService.sendMessage("matterEvent", matterEvent);
                    }
                    break;
                }
            }
        });
    }
    handleProcessClose(code, signal) {
        const isLikelyPluginCrash = code === 1 && signal === null;
        this.log.warn(`Child bridge ended (code ${code}, signal ${signal}).${isLikelyPluginCrash
            ? ' The child bridge ended unexpectedly, which is normally due to the plugin not catching its errors properly. Please report this to the plugin developer by clicking on the'
                + ' \'Report An Issue\' option in the plugin menu dropdown from the Homebridge UI. If there are related logs shown above, please include them in your report.'
            : ''}`);
        if (isLikelyPluginCrash) {
            if (this.restartCount < this.maxRestarts) {
                this.bridgeStatus = "pending";
                this.restartCount += 1;
                const delay = this.restartCount * 10;
                this.log(`Child bridge will automatically restart in ${delay} seconds (restart attempt ${this.restartCount} of ${this.maxRestarts}).`);
                this.scheduledRestartTimeout = setTimeout(() => {
                    this.scheduledRestartTimeout = undefined;
                    if (!this.shuttingDown && !this.manuallyStopped) {
                        this.startChildProcess();
                    }
                }, delay * 1000);
            }
            else {
                this.bridgeStatus = "down";
                this.manuallyStopped = true;
                this.log.error(`Child bridge will no longer restart after failing ${this.maxRestarts + 1} times, you will need to manually start this child bridge from the Homebridge UI.`);
            }
            return;
        }
        if (!this.shuttingDown) {
            this.bridgeStatus = "down";
            this.restartCount = 0;
            this.startChildProcess();
        }
    }
    sendMessage(type, data) {
        if (this.child && this.child.connected) {
            this.child.send({
                id: type,
                data,
            });
        }
    }
    setProcessFlags() {
        if (this.bridgeConfig.debugModeEnabled) {
            this.args.push('-D');
        }
        if (this.homebridgeOptions.forceColourLogging) {
            this.args.push('-C');
        }
        if (this.homebridgeOptions.insecureAccess) {
            this.args.push('-I');
        }
        if (this.homebridgeOptions.noLogTimestamps) {
            this.args.push('-T');
        }
        if (this.homebridgeOptions.keepOrphanedCachedAccessories) {
            this.args.push('-K');
        }
        if (this.homebridgeOptions.customStoragePath) {
            this.args.push('-U', this.homebridgeOptions.customStoragePath);
        }
        if (this.homebridgeOptions.customPluginPath) {
            this.args.push('-P', this.homebridgeOptions.customPluginPath);
        }
    }
    setProcessEnv() {
        this.processEnv = {
            env: {
                ...process.env,
                DEBUG: `${process.env.DEBUG || ''} ${this.bridgeConfig.env?.DEBUG || ''}`.trim(),
                NODE_OPTIONS: `${process.env.NODE_OPTIONS || ''} ${this.bridgeConfig.env?.NODE_OPTIONS || ''}`.trim(),
            },
            silent: true,
        };
    }
    loadPlugin() {
        const bridgeConfig = {
            name: this.bridgeConfig.name || this.displayName || this.plugin.getPluginIdentifier(),
            port: this.bridgeConfig.port,
            username: this.bridgeConfig.username,
            advertiser: this.homebridgeConfig.bridge.advertiser,
            pin: this.bridgeConfig.pin || this.homebridgeConfig.bridge.pin,
            bind: this.homebridgeConfig.bridge.bind,
            setupID: this.bridgeConfig.setupID,
            manufacturer: this.bridgeConfig.manufacturer || this.homebridgeConfig.bridge.manufacturer,
            model: this.bridgeConfig.model || this.homebridgeConfig.bridge.model,
            firmwareRevision: this.bridgeConfig.firmwareRevision || this.homebridgeConfig.bridge.firmwareRevision,
            serialNumber: this.bridgeConfig.serialNumber || this.bridgeConfig.username,
            matter: this.bridgeConfig.matter,
        };
        const bridgeOptions = {
            cachedAccessoriesDir: User.cachedAccessoryPath(),
            cachedAccessoriesItemName: `cachedAccessories.${this.bridgeConfig.username.replace(COLON_RE, '').toUpperCase()}`,
        };
        Object.assign(bridgeOptions, this.homebridgeOptions);
        if (this.bridgeConfig.debugModeEnabled !== undefined) {
            bridgeOptions.debugModeEnabled = this.bridgeConfig.debugModeEnabled;
        }
        this.sendMessage("load", {
            type: this.type,
            identifier: this.identifier,
            pluginPath: this.plugin.getPluginPath(),
            pluginConfig: this.pluginConfig,
            bridgeConfig,
            bridgeOptions,
            homebridgeConfig: {
                bridge: this.homebridgeConfig.bridge,
                ports: this.homebridgeConfig.ports,
                disabledPlugins: [],
                accessories: [],
                platforms: [],
            },
        });
    }
    startBridge() {
        this.sendMessage("start");
    }
    async handlePortRequest(request) {
        let port;
        if (request.portType === 'matter') {
            port = await this.externalPortService.requestMatterPort(request.username);
        }
        else {
            port = await this.externalPortService.requestPort(request.username);
        }
        this.sendMessage("portAllocated", {
            username: request.username,
            port,
        });
    }
    teardown() {
        if (this.child && this.child.connected) {
            this.bridgeStatus = "down";
            this.child.kill('SIGTERM');
        }
    }
    sendStatusUpdate() {
        this.ipcService.sendMessage("childBridgeStatusUpdate", this.getMetadata());
    }
    restartChildBridge() {
        if (this.manuallyStopped) {
            this.restartCount = 0;
            this.startChildBridge();
        }
        else {
            this.log.warn('Child bridge restarting...');
            void this.refreshConfig();
            this.teardown();
        }
    }
    stopChildBridge() {
        if (!this.shuttingDown) {
            this.log.warn('Child bridge stopping, will not restart.');
            this.shuttingDown = true;
            this.manuallyStopped = true;
            this.restartCount = 0;
            if (this.scheduledRestartTimeout) {
                clearTimeout(this.scheduledRestartTimeout);
                this.scheduledRestartTimeout = undefined;
            }
            this.bridgeStatus = "down";
            this.child?.removeAllListeners();
            this.teardown();
        }
        else {
            this.log.warn('Child bridge already shutting down or stopped.');
        }
    }
    startChildBridge() {
        if (this.manuallyStopped && this.bridgeStatus === "down" && (!this.child || !this.child.connected)) {
            void this.refreshConfig();
            this.startChildProcess();
            this.shuttingDown = false;
            this.manuallyStopped = false;
        }
        else {
            this.log.warn('Child bridge cannot be started, it is still running or was not manually stopped.');
        }
    }
    async refreshConfig() {
        try {
            const homebridgeConfig = await fs.readJson(User.configPath());
            if (this.type === "platform") {
                const config = homebridgeConfig.platforms?.filter(x => x.platform === this.identifier && x._bridge?.username === this.bridgeConfig.username);
                if (config.length) {
                    this.pluginConfig = config;
                    this.bridgeConfig = this.pluginConfig[0]._bridge || this.bridgeConfig;
                }
                else {
                    this.log.warn('Platform config could not be found, using existing config.');
                }
            }
            else if (this.type === "accessory") {
                const config = homebridgeConfig.accessories?.filter(x => x.accessory === this.identifier && x._bridge?.username === this.bridgeConfig.username);
                if (config.length) {
                    this.pluginConfig = config;
                    this.bridgeConfig = this.pluginConfig[0]._bridge || this.bridgeConfig;
                }
                else {
                    this.log.warn('Accessory config could not be found, using existing config.');
                }
            }
        }
        catch (error) {
            this.log.error('Failed to refresh plugin config:', error.message);
        }
    }
    getMetadata() {
        return {
            status: this.bridgeStatus,
            paired: this.pairedStatus,
            setupUri: this.setupUri,
            username: this.bridgeConfig.username,
            port: this.bridgeConfig.port,
            pin: this.bridgeConfig.pin || this.homebridgeConfig.bridge.pin,
            name: this.bridgeConfig.name || this.displayName || this.plugin.getPluginIdentifier(),
            plugin: this.plugin.getPluginIdentifier(),
            identifier: this.identifier,
            pid: this.child?.pid,
            manuallyStopped: this.manuallyStopped,
            matterConfig: this.bridgeConfig.matter,
            matterIdentifier: this.bridgeConfig.matter ? this.bridgeConfig.username : undefined,
            matterSetupUri: this.matterCommissioningInfo?.qrCode,
            matterPin: this.matterCommissioningInfo?.manualPairingCode,
            matterSerialNumber: this.matterCommissioningInfo?.serialNumber,
            matterCommissioned: this.matterCommissioningInfo?.commissioned,
            matterDeviceCount: this.matterCommissioningInfo?.deviceCount,
        };
    }
}
