// homebridge/dist/childBridgeFork.js
import process from 'node:process';
import { HAPStorage } from '@homebridge/hap-nodejs';
import { HomebridgeAPI } from './api.js';
import { BridgeService } from './bridgeService.js';
import { ChildBridgeExternalPortService } from './externalPortService.js';
import { Logger } from './logger.js';
import { ChildBridgeMatterMessageHandler } from './matter/ChildBridgeMatterMessageHandler.js';
import { PluginManager } from './pluginManager.js';
import { User } from './user.js';
import 'source-map-support/register.js';
process.title = 'homebridge: child bridge';
const matterLogger = Logger.withPrefix('Matter/ChildManager');
export class ChildBridgeFork {
    bridgeService;
    api;
    pluginManager;
    externalPortService;
    matterManager;
    matterMessageHandler;
    type;
    plugin;
    identifier;
    pluginConfig;
    bridgeConfig;
    bridgeOptions;
    portRequestCallback = new Map();
    constructor() {
        this.sendMessage("ready");
    }
    sendMessage(type, data) {
        if (process.send) {
            process.send({
                id: type,
                data,
            });
        }
    }
    async loadPlugin(data) {
        this.type = data.type;
        this.identifier = data.identifier;
        this.pluginConfig = data.pluginConfig;
        this.bridgeConfig = data.bridgeConfig;
        this.bridgeOptions = data.bridgeOptions;
        for (const config of this.pluginConfig) {
            delete config._bridge;
        }
        if (this.bridgeOptions.noLogTimestamps) {
            Logger.setTimestampEnabled(false);
        }
        if (this.bridgeOptions.debugModeEnabled) {
            Logger.setDebugEnabled(true);
        }
        if (this.bridgeOptions.forceColourLogging) {
            Logger.forceColor();
        }
        if (this.bridgeOptions.customStoragePath) {
            User.setStoragePath(this.bridgeOptions.customStoragePath);
        }
        HAPStorage.setCustomStoragePath(User.persistPath());
        this.api = new HomebridgeAPI();
        this.pluginManager = new PluginManager(this.api);
        this.externalPortService = new ChildBridgeExternalPortService(this);
        this.plugin = this.pluginManager.loadPlugin(data.pluginPath);
        await this.plugin.load();
        await this.pluginManager.initializePlugin(this.plugin, data.identifier);
        process.title = `homebridge: ${this.plugin.getPluginIdentifier()}`;
        this.sendMessage("loaded", {
            version: this.plugin.version,
        });
    }
    async startBridge() {
        if (this.bridgeConfig.matter && this.type === "accessory") {
            matterLogger.warn('Matter is not supported on accessory child bridges. Ignoring matter configuration.');
        }
        if (this.bridgeConfig.matter && this.type !== "accessory") {
            matterLogger.info('Loading Matter support for child bridge...');
            await this.api.loadMatterAPI();
            const { ChildBridgeMatterManager } = await import('./matter/index.js');
            this.matterManager = new ChildBridgeMatterManager(this.bridgeConfig, this.bridgeOptions, this.api, this.externalPortService, this.pluginManager);
            this.api._setMatterManager(this.matterManager);
            await this.matterManager.initialize(() => {
                this.sendPairedStatusEvent();
            });
            matterLogger.debug('Creating ChildBridgeMatterMessageHandler...');
            this.matterMessageHandler = new ChildBridgeMatterMessageHandler(this.matterManager, this.bridgeConfig.username, (type, data) => this.sendMessage(type, data));
            matterLogger.debug(`Matter message handler created for child bridge ${this.bridgeConfig.username}`);
        }
        else {
            matterLogger.debug('Matter not configured for this child bridge, skipping Matter setup');
        }
        this.bridgeService = new BridgeService(this.api, this.pluginManager, this.externalPortService, this.bridgeOptions, this.bridgeConfig);
        this.bridgeService.bridge.on("advertised", () => {
            this.sendPairedStatusEvent();
        });
        this.bridgeService.bridge.on("paired", () => {
            this.sendPairedStatusEvent();
        });
        this.bridgeService.bridge.on("unpaired", () => {
            this.sendPairedStatusEvent();
        });
        await this.bridgeService.loadCachedPlatformAccessoriesFromDisk();
        for (const config of this.pluginConfig) {
            if (this.type === "platform") {
                const plugin = this.pluginManager.getPluginForPlatform(this.identifier);
                const displayName = config.name || plugin.getPluginIdentifier();
                const logger = Logger.withPrefix(displayName);
                const constructor = plugin.getPlatformConstructor(this.identifier);
                const platform = new constructor(logger, config, this.api);
                if (HomebridgeAPI.isDynamicPlatformPlugin(platform)) {
                    plugin.assignDynamicPlatform(this.identifier, platform);
                }
                else if (HomebridgeAPI.isStaticPlatformPlugin(platform)) {
                    await this.bridgeService.loadPlatformAccessories(plugin, platform, this.identifier, logger);
                }
                else {
                }
            }
            else if (this.type === "accessory") {
                const plugin = this.pluginManager.getPluginForAccessory(this.identifier);
                const displayName = config.name;
                if (!displayName) {
                    Logger.internal.warn('Could not load accessory %s as it is missing the required \'name\' property!', this.identifier);
                    return;
                }
                const logger = Logger.withPrefix(displayName);
                const constructor = plugin.getAccessoryConstructor(this.identifier);
                const accessoryInstance = new constructor(logger, config, this.api);
                const accessory = this.bridgeService.createHAPAccessory(plugin, accessoryInstance, displayName, this.identifier, config.uuid_base);
                if (accessory) {
                    this.bridgeService.bridge.addBridgedAccessory(accessory);
                }
                else {
                    logger('Accessory %s returned empty set of services. Won\'t adding it to the bridge!', this.identifier);
                }
            }
        }
        this.bridgeService.restoreCachedPlatformAccessories();
        if (this.matterManager) {
            this.matterManager.restoreCachedAccessories(this.bridgeOptions.keepOrphanedCachedAccessories ?? false);
        }
        this.bridgeService.publishBridge();
        this.api.signalFinished();
        this.sendPairedStatusEvent();
        this.sendMessage("online");
    }
    async requestExternalPort(username) {
        return new Promise((resolve) => {
            const requestTimeout = setTimeout(() => {
                Logger.internal.warn('Parent process did not respond to port allocation request within 5 seconds - assigning random port.');
                this.portRequestCallback.delete(username);
                resolve(undefined);
            }, 5000);
            const callback = (port) => {
                clearTimeout(requestTimeout);
                this.portRequestCallback.delete(username);
                resolve(port);
            };
            this.portRequestCallback.set(username, callback);
            this.sendMessage("portRequest", { username });
        });
    }
    async requestMatterPort(uniqueId) {
        return new Promise((resolve) => {
            const mac = uniqueId;
            const requestTimeout = setTimeout(() => {
                matterLogger.warn('Parent process did not respond to Matter port allocation request within 5 seconds - assigning random port.');
                this.portRequestCallback.delete(mac);
                resolve(undefined);
            }, 5000);
            const callback = (port) => {
                clearTimeout(requestTimeout);
                this.portRequestCallback.delete(mac);
                resolve(port);
            };
            this.portRequestCallback.set(mac, callback);
            this.sendMessage("portRequest", {
                username: mac,
                portType: 'matter',
            });
        });
    }
    handleExternalResponse(data) {
        const callback = this.portRequestCallback.get(data.username);
        if (callback) {
            callback(data.port);
        }
    }
    sendPairedStatusEvent() {
        const matterInfo = this.matterManager?.getMatterStatusInfo();
        this.sendMessage("status", {
            paired: this.bridgeService?.bridge?._accessoryInfo?.paired() ?? null,
            setupUri: this.bridgeService?.bridge?.setupURI() ?? null,
            ...(matterInfo && { matter: matterInfo }),
        });
    }
    handleStartMatterMonitoring() {
        this.matterMessageHandler?.handleStartMatterMonitoring();
    }
    handleStopMatterMonitoring() {
        this.matterMessageHandler?.handleStopMatterMonitoring();
    }
    handleGetMatterAccessories() {
        if (!this.matterMessageHandler) {
            if (!this.bridgeConfig) {
                return;
            }
            const event = {
                type: 'accessoriesData',
                data: {
                    bridgeUsername: this.bridgeConfig.username,
                    accessories: [],
                },
            };
            this.sendMessage("matterEvent", event);
            return;
        }
        this.matterMessageHandler.handleGetMatterAccessories();
    }
    handleGetMatterAccessoryInfo(data) {
        this.matterMessageHandler?.handleGetMatterAccessoryInfo(data);
    }
    handleMatterAccessoryControl(data) {
        this.matterMessageHandler?.handleMatterAccessoryControl(data);
    }
    shutdown() {
        this.bridgeService.teardown();
        if (this.matterManager) {
            this.matterManager.teardown().catch((error) => {
                matterLogger.error('Error tearing down Matter manager:', error);
            });
        }
    }
}
const childPluginFork = new ChildBridgeFork();
process.on('message', (message) => {
    if (typeof message !== 'object' || !message.id) {
        return;
    }
    switch (message.id) {
        case "load": {
            childPluginFork.loadPlugin(message.data);
            break;
        }
        case "start": {
            childPluginFork.startBridge();
            break;
        }
        case "portAllocated": {
            childPluginFork.handleExternalResponse(message.data);
            break;
        }
        case "startMatterMonitoring": {
            childPluginFork.handleStartMatterMonitoring();
            break;
        }
        case "stopMatterMonitoring": {
            childPluginFork.handleStopMatterMonitoring();
            break;
        }
        case "getMatterAccessories": {
            childPluginFork.handleGetMatterAccessories();
            break;
        }
        case "getMatterAccessoryInfo": {
            childPluginFork.handleGetMatterAccessoryInfo(message.data);
            break;
        }
        case "matterAccessoryControl": {
            childPluginFork.handleMatterAccessoryControl(message.data);
            break;
        }
        default: {
            Logger.internal.warn(`Received unknown message type from parent process: ${message.id}`);
            break;
        }
    }
});
let shuttingDown = false;
function signalHandler(signal, signalNum) {
    if (shuttingDown) {
        return;
    }
    shuttingDown = true;
    Logger.internal.info('Got %s, shutting down child bridge process...', signal);
    try {
        childPluginFork.shutdown();
    }
    catch (error) {
    }
    setTimeout(() => process.exit(128 + signalNum), 5000);
}
process.on('SIGINT', signalHandler.bind(undefined, 'SIGINT', 2));
process.on('SIGTERM', signalHandler.bind(undefined, 'SIGTERM', 15));
setInterval(() => {
    if (!process.connected) {
        Logger.internal.info('Parent process not connected, terminating process...');
        process.exit(1);
    }
}, 5000);
