// homebridge/dist/externalPortService.js
import { Logger } from './logger.js';
import { MatterPortAllocator } from './matter/MatterPortAllocator.js';
export class ExternalPortService {
    externalPorts;
    nextExternalPort;
    allocatedPorts = new Map();
    matterPortAllocator;
    constructor(externalPorts, matterPorts, configuredMatterPorts) {
        this.externalPorts = externalPorts;
        this.matterPortAllocator = new MatterPortAllocator(matterPorts, configuredMatterPorts);
    }
    async requestPort(username) {
        const existingPortAllocation = this.allocatedPorts.get(username);
        if (existingPortAllocation) {
            return existingPortAllocation;
        }
        const port = this.getNextFreePort();
        this.allocatedPorts.set(username, port);
        return port;
    }
    async requestMatterPort(uuid) {
        return await this.matterPortAllocator.requestPort(uuid);
    }
    getNextFreePort() {
        if (!this.externalPorts) {
            return undefined;
        }
        if (this.nextExternalPort === undefined) {
            this.nextExternalPort = this.externalPorts.start;
            return this.nextExternalPort;
        }
        this.nextExternalPort++;
        if (this.nextExternalPort <= this.externalPorts.end) {
            return this.nextExternalPort;
        }
        Logger.internal.warn('External HAP port pool ran out of ports. Falling back to random port assignment.');
        return undefined;
    }
}
export class ChildBridgeExternalPortService extends ExternalPortService {
    childBridge;
    constructor(childBridge) {
        super();
        this.childBridge = childBridge;
    }
    async requestPort(username) {
        return await this.childBridge.requestExternalPort(username);
    }
    async requestMatterPort(uniqueId) {
        return await this.childBridge.requestMatterPort(uniqueId);
    }
}
