// homebridge/dist/index.js
/**
 * Export API const enums
 */
export { APIEvent, PluginType } from './api.js';
/**
 * Export Logger const enums
 */
export { LogLevel } from './logger.js';
/**
 * Matter protocol status errors for plugin handlers
 * @example
 * ```typescript
 * import { MatterStatus } from 'homebridge'
 *
 * handlers: {
 *   onOff: {
 *     on: async () => {
 *       if (deviceIsBusy) {
 *         throw new MatterStatus.Busy('Device is processing another command')
 *       }
 *       // ... control device
 *     }
 *   }
 * }
 * ```
 */
export { MatterStatus } from './matter/index.js';
export { ChildMatterMessageType, MatterBridgeStatus } from './matter/sharedTypes.js';
/**
 * ═══════════════════════════════════════════════════════════════════════
 * Matter Protocol - UI Integration Types
 * ═══════════════════════════════════════════════════════════════════════
 */
/**
 * Matter device types, clusters, and cluster names
 * Access via api.matter?.deviceTypes, api.matter?.clusters, api.matter?.clusterNames
 */
export { clusterNames, clusters, devices, deviceTypes, MatterAccessoryEventTypes } from './matter/types.js';
/**
 * Matter error types for error handling
 */
export { MatterCommissioningError, MatterDeviceError, MatterError, MatterErrorType, MatterNetworkError, MatterStorageError, } from './matter/types.js';
/**
 * Export Platform Accessory const enums
 */
export { PlatformAccessoryEvent } from './platformAccessory.js';
/**
 * Export the CONST ENUMS from hap-nodejs
 * These get converted to their string value at compile time
 * and can be safely used directly.
 */
export { Access, AccessControlEvent, AccessLevel, AccessoryEventTypes, AdaptiveLightingControllerEvents, AdaptiveLightingControllerMode, AudioBitrate, AudioCodecTypes, AudioRecordingCodecType, AudioRecordingSamplerate, AudioSamplerate, AudioStreamingCodecType, AudioStreamingSamplerate, ButtonState, ButtonType, CameraControllerEvents, Categories, ChangeReason, CharacteristicEventTypes, DataFormatTags, DataStreamConnectionEvent, DataStreamServerEvent, DataStreamStatus, DefaultControllerType, EventTriggerOption, Formats, H264Level, H264Profile, HAPServerEventTypes, HAPStatus, HDSProtocolSpecificErrorReason, HDSStatus, MediaContainerType, PacketDataType, Perms, Protocols, RemoteControllerEvents, ResourceRequestReason, ServiceEventTypes, SiriAudioSessionEvents, SRTPCryptoSuites, StreamRequestTypes, TargetCategory, TargetUpdates, Topics, Units, } from '@homebridge/hap-nodejs';
