// homebridge/dist/ipcService.js
import { EventEmitter } from 'node:events';
import process from 'node:process';
export var IpcIncomingEvent;
(function (IpcIncomingEvent) {
    IpcIncomingEvent["RESTART_CHILD_BRIDGE"] = "restartChildBridge";
    IpcIncomingEvent["STOP_CHILD_BRIDGE"] = "stopChildBridge";
    IpcIncomingEvent["START_CHILD_BRIDGE"] = "startChildBridge";
    IpcIncomingEvent["CHILD_BRIDGE_METADATA_REQUEST"] = "childBridgeMetadataRequest";
    IpcIncomingEvent["START_MATTER_MONITORING"] = "startMatterMonitoring";
    IpcIncomingEvent["STOP_MATTER_MONITORING"] = "stopMatterMonitoring";
    IpcIncomingEvent["GET_MATTER_ACCESSORIES"] = "getMatterAccessories";
    IpcIncomingEvent["GET_MATTER_ACCESSORY_INFO"] = "getMatterAccessoryInfo";
    IpcIncomingEvent["MATTER_ACCESSORY_CONTROL"] = "matterAccessoryControl";
})(IpcIncomingEvent || (IpcIncomingEvent = {}));
export var IpcOutgoingEvent;
(function (IpcOutgoingEvent) {
    IpcOutgoingEvent["SERVER_STATUS_UPDATE"] = "serverStatusUpdate";
    IpcOutgoingEvent["CHILD_BRIDGE_METADATA_RESPONSE"] = "childBridgeMetadataResponse";
    IpcOutgoingEvent["CHILD_BRIDGE_STATUS_UPDATE"] = "childBridgeStatusUpdate";
    IpcOutgoingEvent["MATTER_EVENT"] = "matterEvent";
})(IpcOutgoingEvent || (IpcOutgoingEvent = {}));
export class IpcService extends EventEmitter {
    constructor() {
        super();
    }
    start() {
        process.on('message', (message) => {
            if (!message || typeof message !== 'object' || !message.id) {
                return;
            }
            this.emit(message.id, message.data);
        });
    }
    sendMessage(id, data) {
        if (process.send) {
            process.send({
                id,
                data,
            });
        }
    }
}
