// homebridge/dist/logger.js
import util from 'node:util';
import chalk from 'chalk';
export var LogLevel;
(function (LogLevel) {
    LogLevel["INFO"] = "info";
    LogLevel["SUCCESS"] = "success";
    LogLevel["WARN"] = "warn";
    LogLevel["ERROR"] = "error";
    LogLevel["DEBUG"] = "debug";
})(LogLevel || (LogLevel = {}));
export class Logger {
    static internal = new Logger();
    static loggerCache = new Map();
    static debugEnabled = false;
    static timestampEnabled = true;
    prefix;
    constructor(prefix) {
        this.prefix = prefix;
    }
    static withPrefix(prefix) {
        const loggerStuff = Logger.loggerCache.get(prefix);
        if (loggerStuff) {
            return loggerStuff;
        }
        else {
            const logger = new Logger(prefix);
            const log = logger.info.bind(logger);
            log.info = logger.info;
            log.success = logger.success;
            log.warn = logger.warn;
            log.error = logger.error;
            log.debug = logger.debug;
            log.log = logger.log;
            log.prefix = logger.prefix;
            const logging = log;
            Logger.loggerCache.set(prefix, logging);
            return logging;
        }
    }
    static setDebugEnabled(enabled = true) {
        Logger.debugEnabled = enabled;
    }
    static setTimestampEnabled(enabled = true) {
        Logger.timestampEnabled = enabled;
    }
    static isTimestampEnabled() {
        return Logger.timestampEnabled;
    }
    static forceColor() {
        chalk.level = 1;
    }
    info(message, ...parameters) {
        this.log("info", message, ...parameters);
    }
    success(message, ...parameters) {
        this.log("success", message, ...parameters);
    }
    warn(message, ...parameters) {
        this.log("warn", message, ...parameters);
    }
    error(message, ...parameters) {
        this.log("error", message, ...parameters);
    }
    debug(message, ...parameters) {
        this.log("debug", message, ...parameters);
    }
    log(level, message, ...parameters) {
        if (level === "debug" && !Logger.debugEnabled) {
            return;
        }
        message = util.format(message, ...parameters);
        let loggingFunction = console.log;
        switch (level) {
            case "success":
                message = chalk.green(message);
                break;
            case "warn":
                message = chalk.yellow(message);
                loggingFunction = console.error;
                break;
            case "error":
                message = chalk.red(message);
                loggingFunction = console.error;
                break;
            case "debug":
                message = chalk.gray(message);
                break;
        }
        if (this.prefix) {
            message = `${getLogPrefix(this.prefix)} ${message}`;
        }
        if (Logger.timestampEnabled) {
            const date = new Date();
            message = chalk.white(`[${date.toLocaleString()}] `) + message;
        }
        loggingFunction(message);
    }
}
export function getLogPrefix(prefix) {
    return chalk.cyan(`[${prefix}]`);
}
