// homebridge/dist/platformAccessory.js
import { EventEmitter } from 'node:events';
import { Accessory } from '@homebridge/hap-nodejs';
export var PlatformAccessoryEvent;
(function (PlatformAccessoryEvent) {
    PlatformAccessoryEvent["IDENTIFY"] = "identify";
})(PlatformAccessoryEvent || (PlatformAccessoryEvent = {}));
export class PlatformAccessory extends EventEmitter {
    static injectedAccessory;
    _associatedPlugin;
    _associatedPlatform;
    _associatedHAPAccessory;
    displayName;
    UUID;
    category;
    services = [];
    context = {};
    constructor(displayName, uuid, category) {
        super();
        this._associatedHAPAccessory = PlatformAccessory.injectedAccessory
            ? PlatformAccessory.injectedAccessory
            : new Accessory(displayName, uuid);
        if (category) {
            this._associatedHAPAccessory.category = category;
        }
        this.displayName = this._associatedHAPAccessory.displayName;
        this.UUID = this._associatedHAPAccessory.UUID;
        this.category = category || 1;
        this.services = this._associatedHAPAccessory.services;
        this._associatedHAPAccessory.on("identify", (paired, callback) => {
            this.emit("identify", paired, () => { });
            callback();
        });
    }
    updateDisplayName(name) {
        if (name) {
            this.displayName = name;
            this._associatedHAPAccessory.displayName = name;
        }
    }
    addService(service, ...constructorArgs) {
        return this._associatedHAPAccessory.addService(service, ...constructorArgs);
    }
    removeService(service) {
        this._associatedHAPAccessory.removeService(service);
    }
    getService(name) {
        return this._associatedHAPAccessory.getService(name);
    }
    getServiceById(uuid, subType) {
        return this._associatedHAPAccessory.getServiceById(uuid, subType);
    }
    configureController(controller) {
        this._associatedHAPAccessory.configureController(controller);
    }
    removeController(controller) {
        this._associatedHAPAccessory.removeController(controller);
    }
    static serialize(accessory) {
        if (!accessory._associatedPlugin) {
            throw new Error(`Cannot serialize accessory '${accessory.displayName}' - missing associated plugin`);
        }
        if (!accessory._associatedPlatform) {
            throw new Error(`Cannot serialize accessory '${accessory.displayName}' - missing associated platform`);
        }
        accessory._associatedHAPAccessory.displayName = accessory.displayName;
        return {
            plugin: accessory._associatedPlugin,
            platform: accessory._associatedPlatform,
            context: accessory.context,
            ...Accessory.serialize(accessory._associatedHAPAccessory),
        };
    }
    static deserialize(json) {
        const accessory = Accessory.deserialize(json);
        PlatformAccessory.injectedAccessory = accessory;
        const platformAccessory = new PlatformAccessory(accessory.displayName, accessory.UUID);
        PlatformAccessory.injectedAccessory = undefined;
        platformAccessory._associatedPlugin = json.plugin;
        platformAccessory._associatedPlatform = json.platform;
        platformAccessory.context = json.context;
        platformAccessory.category = json.category;
        return platformAccessory;
    }
}
