// homebridge/dist/plugin.js
import assert from 'node:assert';
import { join } from 'node:path';
import process from 'node:process';
import { pathToFileURL } from 'node:url';
import { satisfies } from 'semver';
import { Logger } from './logger.js';
import { PluginManager } from './pluginManager.js';
import getVersion from './version.js';
const log = Logger.internal;
export class Plugin {
    pluginName;
    scope;
    pluginPath;
    disabled = false;
    version;
    main;
    loadContext;
    pluginInitializer;
    registeredAccessories = new Map();
    registeredPlatforms = new Map();
    activeDynamicPlatforms = new Map();
    constructor(name, path, packageJSON, scope) {
        this.pluginName = name;
        this.scope = scope;
        this.pluginPath = path;
        this.version = packageJSON.version || '0.0.0';
        this.main = '';
        if (packageJSON.exports) {
            if (typeof packageJSON.exports === 'string') {
                this.main = packageJSON.exports;
            }
            else {
                const exports = packageJSON.exports.import || packageJSON.exports.require || packageJSON.exports.node || packageJSON.exports.default || packageJSON.exports['.'];
                if (typeof exports !== 'string') {
                    if (exports.import) {
                        this.main = exports.import;
                    }
                    else {
                        this.main = exports.require || exports.node || exports.default;
                    }
                }
                else {
                    this.main = exports;
                }
            }
        }
        if (!this.main) {
            this.main = packageJSON.main || './index.js';
        }
        if (packageJSON.peerDependencies && (!packageJSON.engines || !packageJSON.engines.homebridge)) {
            packageJSON.engines = packageJSON.engines || {};
            packageJSON.engines.homebridge = packageJSON.peerDependencies.homebridge;
        }
        this.loadContext = {
            engines: packageJSON.engines,
            dependencies: packageJSON.dependencies,
        };
    }
    getPluginIdentifier() {
        return (this.scope ? `${this.scope}/` : '') + this.pluginName;
    }
    getPluginPath() {
        return this.pluginPath;
    }
    registerAccessory(name, constructor) {
        if (this.registeredAccessories.has(name)) {
            throw new Error(`Plugin '${this.getPluginIdentifier()}' tried to register an accessory '${name}' which has already been registered!`);
        }
        if (!this.disabled) {
            log.info('Registering accessory \'%s\'', `${this.getPluginIdentifier()}.${name}`);
        }
        this.registeredAccessories.set(name, constructor);
    }
    registerPlatform(name, constructor) {
        if (this.registeredPlatforms.has(name)) {
            throw new Error(`Plugin '${this.getPluginIdentifier()}' tried to register a platform '${name}' which has already been registered!`);
        }
        if (!this.disabled) {
            log.info('Registering platform \'%s\'', `${this.getPluginIdentifier()}.${name}`);
        }
        this.registeredPlatforms.set(name, constructor);
    }
    getAccessoryConstructor(accessoryIdentifier) {
        const name = PluginManager.getAccessoryName(accessoryIdentifier);
        const constructor = this.registeredAccessories.get(name);
        if (!constructor) {
            throw new Error(`The requested accessory '${name}' was not registered by the plugin '${this.getPluginIdentifier()}'.`);
        }
        return constructor;
    }
    getPlatformConstructor(platformIdentifier) {
        const name = PluginManager.getPlatformName(platformIdentifier);
        const constructor = this.registeredPlatforms.get(name);
        if (!constructor) {
            throw new Error(`The requested platform '${name}' was not registered by the plugin '${this.getPluginIdentifier()}'.`);
        }
        if (this.activeDynamicPlatforms.has(name)) {
            throw new Error(`The dynamic platform ${name} from the plugin ${this.getPluginIdentifier()} is configured `
                + 'multiple times in your config.json.');
        }
        return constructor;
    }
    assignDynamicPlatform(platformIdentifier, platformPlugin) {
        const name = PluginManager.getPlatformName(platformIdentifier);
        let platforms = this.activeDynamicPlatforms.get(name);
        if (!platforms) {
            platforms = [];
            this.activeDynamicPlatforms.set(name, platforms);
        }
        platforms.unshift(platformPlugin);
    }
    getActiveDynamicPlatform(platformName) {
        const platforms = this.activeDynamicPlatforms.get(platformName);
        return platforms && platforms[0];
    }
    async load() {
        const context = this.loadContext;
        assert(context, 'Reached illegal state. Plugin state is undefined!');
        this.loadContext = undefined;
        if (!context.engines || !context.engines.homebridge) {
            throw new Error(`Plugin ${this.pluginPath} does not contain the 'homebridge' package in 'engines'.`);
        }
        const versionRequired = context.engines.homebridge;
        const nodeVersionRequired = context.engines.node;
        if (!satisfies(getVersion(), versionRequired, { includePrerelease: true })) {
            log.warn(`The plugin "${this.pluginName}" requires a Homebridge version of ${versionRequired} which does \
not satisfy the current Homebridge version of v${getVersion()}. You may need to update this plugin (or Homebridge) to a newer version. \
You may face unexpected issues or stability problems running this plugin.`);
        }
        if (nodeVersionRequired && !satisfies(process.version, nodeVersionRequired)) {
            log.warn(`The plugin "${this.pluginName}" requires a Node.js version of ${nodeVersionRequired} which does \
not satisfy the current Node.js version of ${process.version}. You may need to upgrade your installation of Node.js - see https://homebridge.io/w/JTKEF`);
        }
        const dependencies = context.dependencies || {};
        if (dependencies.homebridge || dependencies['hap-nodejs'] || dependencies['@homebridge/hap-nodejs']) {
            log.error(`The plugin "${this.pluginName}" defines 'homebridge' and/or 'hap-nodejs' in their 'dependencies' section, \
meaning they carry an additional copy of homebridge and hap-nodejs. This not only wastes disk space, but also can cause \
major incompatibility issues and thus is considered bad practice. Please inform the developer to update their plugin!`);
        }
        const mainPath = join(this.pluginPath, this.main);
        const pluginModules = (await import(pathToFileURL(mainPath).href)).default;
        if (typeof pluginModules === 'function') {
            this.pluginInitializer = pluginModules;
        }
        else if (pluginModules && typeof pluginModules.default === 'function') {
            this.pluginInitializer = pluginModules.default;
        }
        else {
            throw new Error(`Plugin ${this.pluginPath} does not export a initializer function from main.`);
        }
    }
    initialize(api) {
        if (!this.pluginInitializer) {
            throw new Error('Tried to initialize a plugin which hasn\'t been loaded yet!');
        }
        return this.pluginInitializer(api);
    }
}
