// homebridge/dist/pluginManager.js
import { execSync } from 'node:child_process';
import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { createRequire } from 'node:module';
import { delimiter, join, resolve } from 'node:path';
import process from 'node:process';
import { Logger } from './logger.js';
import { Plugin } from './plugin.js';
const log = Logger.internal;
const require = createRequire(import.meta.url);
const paths = require.resolve.paths('');
export class PluginManager {
    static PLUGIN_IDENTIFIER_PATTERN = /^((@[\w-]+(\.[\w-]+)*)\/)?(homebridge-[\w-]+)$/;
    api;
    searchPaths = new Set();
    strictPluginResolution = false;
    activePlugins;
    disabledPlugins;
    plugins = new Map();
    pluginIdentifierTranslation = new Map();
    accessoryToPluginMap = new Map();
    platformToPluginMap = new Map();
    currentInitializingPlugin;
    constructor(api, options) {
        this.api = api;
        if (options) {
            if (options.customPluginPath) {
                this.searchPaths.add(resolve(process.cwd(), options.customPluginPath));
            }
            this.strictPluginResolution = options.strictPluginResolution || false;
            this.activePlugins = options.activePlugins;
            this.disabledPlugins = Array.isArray(options.disabledPlugins) ? options.disabledPlugins : undefined;
        }
        this.api.on("registerAccessory", this.handleRegisterAccessory.bind(this));
        this.api.on("registerPlatform", this.handleRegisterPlatform.bind(this));
    }
    static isQualifiedPluginIdentifier(identifier) {
        return PluginManager.PLUGIN_IDENTIFIER_PATTERN.test(identifier);
    }
    static extractPluginName(name) {
        const match = name.match(PluginManager.PLUGIN_IDENTIFIER_PATTERN);
        if (!match) {
            throw new Error(`Cannot extract plugin name from invalid identifier: '${name}'`);
        }
        return match[4];
    }
    static extractPluginScope(name) {
        const match = name.match(PluginManager.PLUGIN_IDENTIFIER_PATTERN);
        if (!match) {
            throw new Error(`Cannot extract plugin scope from invalid identifier: '${name}'`);
        }
        return match[2];
    }
    static getAccessoryName(identifier) {
        if (!identifier.includes('.')) {
            return identifier;
        }
        return identifier.split('.')[1];
    }
    static getPlatformName(identifier) {
        if (!identifier.includes('.')) {
            return identifier;
        }
        return identifier.split('.')[1];
    }
    static getPluginIdentifier(identifier) {
        return identifier.split('.')[0];
    }
    async initializeInstalledPlugins() {
        log.info('---');
        this.loadInstalledPlugins();
        for (const [identifier, plugin] of this.plugins) {
            try {
                await plugin.load();
            }
            catch (error) {
                log.error('====================');
                log.error(`ERROR LOADING PLUGIN ${identifier}:`);
                log.error(error.stack);
                log.error('====================');
                this.plugins.delete(identifier);
                continue;
            }
            if (this.disabledPlugins && this.disabledPlugins.includes(plugin.getPluginIdentifier())) {
                plugin.disabled = true;
            }
            if (plugin.disabled) {
                log.warn(`Disabled plugin: ${identifier}@${plugin.version}`);
            }
            else {
                log.info(`Loaded plugin: ${identifier}@${plugin.version}`);
            }
            await this.initializePlugin(plugin, identifier);
            log.info('---');
        }
        this.currentInitializingPlugin = undefined;
    }
    async initializePlugin(plugin, identifier) {
        try {
            this.currentInitializingPlugin = plugin;
            await plugin.initialize(this.api);
        }
        catch (error) {
            log.error('====================');
            log.error(`ERROR INITIALIZING PLUGIN ${identifier}:`);
            log.error(error.stack);
            log.error('====================');
            this.plugins.delete(identifier);
        }
    }
    handleRegisterAccessory(name, constructor, pluginIdentifier) {
        if (!this.currentInitializingPlugin) {
            throw new Error(`Unexpected accessory registration. Plugin ${pluginIdentifier ? `'${pluginIdentifier}' ` : ''}tried to register outside the initializer function!`);
        }
        if (pluginIdentifier && pluginIdentifier !== this.currentInitializingPlugin.getPluginIdentifier()) {
            log.info(`Plugin '${this.currentInitializingPlugin.getPluginIdentifier()}' tried to register with an incorrect plugin identifier: '${pluginIdentifier}'. Please report this to the developer!`);
            this.pluginIdentifierTranslation.set(pluginIdentifier, this.currentInitializingPlugin.getPluginIdentifier());
        }
        this.currentInitializingPlugin.registerAccessory(name, constructor);
        let plugins = this.accessoryToPluginMap.get(name);
        if (!plugins) {
            plugins = [];
            this.accessoryToPluginMap.set(name, plugins);
        }
        plugins.push(this.currentInitializingPlugin);
    }
    handleRegisterPlatform(name, constructor, pluginIdentifier) {
        if (!this.currentInitializingPlugin) {
            throw new Error(`Unexpected platform registration. Plugin ${pluginIdentifier ? `'${pluginIdentifier}' ` : ''}tried to register outside the initializer function!`);
        }
        if (pluginIdentifier && pluginIdentifier !== this.currentInitializingPlugin.getPluginIdentifier()) {
            log.debug(`Plugin '${this.currentInitializingPlugin.getPluginIdentifier()}' tried to register with an incorrect plugin identifier: '${pluginIdentifier}'. Please report this to the developer!`);
            this.pluginIdentifierTranslation.set(pluginIdentifier, this.currentInitializingPlugin.getPluginIdentifier());
        }
        this.currentInitializingPlugin.registerPlatform(name, constructor);
        let plugins = this.platformToPluginMap.get(name);
        if (!plugins) {
            plugins = [];
            this.platformToPluginMap.set(name, plugins);
        }
        plugins.push(this.currentInitializingPlugin);
    }
    getPluginForAccessory(accessoryIdentifier) {
        let plugin;
        if (!accessoryIdentifier.includes('.')) {
            let found = this.accessoryToPluginMap.get(accessoryIdentifier);
            if (!found) {
                throw new Error(`No plugin was found for the accessory "${accessoryIdentifier}" in your config.json. Please make sure the corresponding plugin is installed correctly.`);
            }
            if (found.length > 1) {
                const options = found.map(plugin => `${plugin.getPluginIdentifier()}.${accessoryIdentifier}`).join(', ');
                found = found.filter(plugin => !plugin.disabled);
                if (found.length !== 1) {
                    throw new Error(`The requested accessory '${accessoryIdentifier}' has been registered multiple times. Please be more specific by writing one of: ${options}`);
                }
            }
            plugin = found[0];
            accessoryIdentifier = `${plugin.getPluginIdentifier()}.${accessoryIdentifier}`;
        }
        else {
            const pluginIdentifier = PluginManager.getPluginIdentifier(accessoryIdentifier);
            if (!this.hasPluginRegistered(pluginIdentifier)) {
                throw new Error(`The requested plugin '${pluginIdentifier}' was not registered.`);
            }
            plugin = this.getPlugin(pluginIdentifier);
        }
        return plugin;
    }
    getPluginForPlatform(platformIdentifier) {
        let plugin;
        if (!platformIdentifier.includes('.')) {
            let found = this.platformToPluginMap.get(platformIdentifier);
            if (!found) {
                throw new Error(`No plugin was found for the platform "${platformIdentifier}" in your config.json. Please make sure the corresponding plugin is installed correctly.`);
            }
            if (found.length > 1) {
                const options = found.map(plugin => `${plugin.getPluginIdentifier()}.${platformIdentifier}`).join(', ');
                found = found.filter(plugin => !plugin.disabled);
                if (found.length !== 1) {
                    throw new Error(`The requested platform '${platformIdentifier}' has been registered multiple times. Please be more specific by writing one of: ${options}`);
                }
            }
            plugin = found[0];
            platformIdentifier = `${plugin.getPluginIdentifier()}.${platformIdentifier}`;
        }
        else {
            const pluginIdentifier = PluginManager.getPluginIdentifier(platformIdentifier);
            if (!this.hasPluginRegistered(pluginIdentifier)) {
                throw new Error(`The requested plugin '${pluginIdentifier}' was not registered.`);
            }
            plugin = this.getPlugin(pluginIdentifier);
        }
        return plugin;
    }
    hasPluginRegistered(pluginIdentifier) {
        return this.plugins.has(pluginIdentifier) || this.pluginIdentifierTranslation.has(pluginIdentifier);
    }
    getPlugin(pluginIdentifier) {
        const plugin = this.plugins.get(pluginIdentifier);
        if (plugin) {
            return plugin;
        }
        else {
            const translation = this.pluginIdentifierTranslation.get(pluginIdentifier);
            if (translation) {
                return this.plugins.get(translation);
            }
        }
        return undefined;
    }
    getPluginByActiveDynamicPlatform(platformName) {
        const found = (this.platformToPluginMap.get(platformName) || [])
            .filter(plugin => !!plugin.getActiveDynamicPlatform(platformName));
        if (found.length === 0) {
            return undefined;
        }
        else if (found.length > 1) {
            const plugins = found.map(plugin => plugin.getPluginIdentifier()).join(', ');
            throw new Error(`'${platformName}' is an ambiguous platform name. It was registered by multiple plugins: ${plugins}`);
        }
        else {
            return found[0];
        }
    }
    loadInstalledPlugins() {
        this.loadDefaultPaths();
        this.searchPaths.forEach((searchPath) => {
            if (!existsSync(searchPath)) {
                return;
            }
            if (existsSync(join(searchPath, 'package.json'))) {
                try {
                    this.loadPlugin(searchPath);
                }
                catch (error) {
                    log.warn(error.message);
                }
            }
            else {
                let relativePluginPaths = readdirSync(searchPath)
                    .filter((relativePath) => {
                    try {
                        return statSync(resolve(searchPath, relativePath)).isDirectory();
                    }
                    catch (error) {
                        log.debug(`Ignoring path ${resolve(searchPath, relativePath)} - ${error.message}`);
                        return false;
                    }
                });
                const scopeDirectories = relativePluginPaths.filter(path => path.startsWith('@'));
                relativePluginPaths = relativePluginPaths.filter(path => !path.startsWith('@'));
                for (const scopeDirectory of scopeDirectories) {
                    const absolutePath = join(searchPath, scopeDirectory);
                    readdirSync(absolutePath)
                        .filter(name => PluginManager.isQualifiedPluginIdentifier(name))
                        .filter((name) => {
                        try {
                            return statSync(resolve(absolutePath, name)).isDirectory();
                        }
                        catch (error) {
                            log.debug(`Ignoring path ${resolve(absolutePath, name)} - ${error.message}`);
                            return false;
                        }
                    })
                        .forEach(name => relativePluginPaths.push(`${scopeDirectory}/${name}`));
                }
                relativePluginPaths
                    .filter((pluginIdentifier) => {
                    return PluginManager.isQualifiedPluginIdentifier(pluginIdentifier)
                        && (!this.activePlugins || this.activePlugins.includes(pluginIdentifier));
                })
                    .forEach((pluginIdentifier) => {
                    try {
                        const absolutePath = resolve(searchPath, pluginIdentifier);
                        this.loadPlugin(absolutePath);
                    }
                    catch (error) {
                        log.warn(error.message);
                    }
                });
            }
        });
        if (this.plugins.size === 0) {
            log.warn('No plugins found.');
        }
    }
    loadPlugin(absolutePath) {
        const packageJson = PluginManager.loadPackageJSON(absolutePath);
        const identifier = packageJson.name;
        const name = PluginManager.extractPluginName(identifier);
        const scope = PluginManager.extractPluginScope(identifier);
        const alreadyInstalled = this.plugins.get(identifier);
        if (alreadyInstalled) {
            throw new Error(`Warning: skipping plugin found at '${absolutePath}' since we already loaded the same plugin from '${alreadyInstalled.getPluginPath()}'.`);
        }
        const plugin = new Plugin(name, absolutePath, packageJson, scope);
        this.plugins.set(identifier, plugin);
        return plugin;
    }
    static loadPackageJSON(pluginPath) {
        const packageJsonPath = join(pluginPath, 'package.json');
        let packageJson;
        if (!existsSync(packageJsonPath)) {
            throw new Error(`Plugin ${pluginPath} does not contain a package.json.`);
        }
        try {
            packageJson = JSON.parse(readFileSync(packageJsonPath, { encoding: 'utf8' }));
        }
        catch (error) {
            throw new Error(`Plugin ${pluginPath} contains an invalid package.json. Error: ${error}`);
        }
        if (!packageJson.name || !PluginManager.isQualifiedPluginIdentifier(packageJson.name)) {
            throw new Error(`Plugin ${pluginPath} does not have a package name that begins with 'homebridge-' or '@scope/homebridge-.`);
        }
        if (!packageJson.keywords || !packageJson.keywords.includes('homebridge-plugin')) {
            throw new Error(`Plugin ${pluginPath} package.json does not contain the keyword 'homebridge-plugin'.`);
        }
        return packageJson;
    }
    loadDefaultPaths() {
        if (this.strictPluginResolution) {
            if (this.searchPaths.size === 0) {
                this.addNpmPrefixToSearchPaths();
            }
            return;
        }
        if (paths) {
            paths.forEach(path => this.searchPaths.add(path));
        }
        if (process.env.NODE_PATH) {
            process.env
                .NODE_PATH
                .split(delimiter)
                .filter(path => !!path)
                .forEach(path => this.searchPaths.add(path));
        }
        else {
            if (process.platform !== 'win32') {
                this.searchPaths.add('/usr/local/lib/node_modules');
                this.searchPaths.add('/usr/lib/node_modules');
            }
            this.addNpmPrefixToSearchPaths();
        }
    }
    addNpmPrefixToSearchPaths() {
        if (process.platform === 'win32') {
            this.searchPaths.add(join(process.env.APPDATA, 'npm/node_modules'));
        }
        else {
            this.searchPaths.add(execSync('/bin/echo -n "$(npm -g prefix)/lib/node_modules"', {
                env: {
                    npm_config_loglevel: 'silent',
                    npm_update_notifier: 'false',
                    ...process.env,
                },
            }).toString('utf8'));
        }
    }
}
