// @matter/protocol/dist/cjs/peer/ControllerCommissioningFlow.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ControllerCommissioningFlow_exports = {};
__export(ControllerCommissioningFlow_exports, {
  CommissioningTimeoutError: () => CommissioningTimeoutError,
  ControllerCommissioningFlow: () => ControllerCommissioningFlow,
  DeviceAlreadyCommissionedToThisFabricError: () => DeviceAlreadyCommissionedToThisFabricError,
  FabricLabelConflictError: () => FabricLabelConflictError,
  MaximumCommissionedFabricsReachedError: () => MaximumCommissionedFabricsReachedError,
  NodeIdConflictError: () => NodeIdConflictError,
  OperativeConnectionFailedError: () => OperativeConnectionFailedError,
  ThreadNetworkSetupFailedError: () => ThreadNetworkSetupFailedError,
  WifiNetworkSetupFailedError: () => WifiNetworkSetupFailedError,
  WifiOrThreadNetworkCredentialsNotConfiguredError: () => WifiOrThreadNetworkCredentialsNotConfiguredError
});
module.exports = __toCommonJS(ControllerCommissioningFlow_exports);
var import_action = require("#action/index.js");
var import_Invoke = require("#action/request/Invoke.js");
var import_Read = require("#action/request/Read.js");
var import_Write = require("#action/request/Write.js");
var import_Ble = require("#ble/Ble.js");
var import_Certificate = require("#certificate/kinds/Certificate.js");
var import_PeerCommunicationError = require("#peer/PeerCommunicationError.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_basic_information = require("@matter/types/clusters/basic-information");
var import_descriptor = require("@matter/types/clusters/descriptor");
var import_general_commissioning = require("@matter/types/clusters/general-commissioning");
var import_network_commissioning = require("@matter/types/clusters/network-commissioning");
var import_operational_credentials = require("@matter/types/clusters/operational-credentials");
var import_ota_software_update_requestor = require("@matter/types/clusters/ota-software-update-requestor");
var import_time_synchronization = require("@matter/types/clusters/time-synchronization");
var import_DeviceAttestationValidator = require("../certificate/DeviceAttestationValidator.js");
var import_AttestationCertificates = require("../certificate/kinds/AttestationCertificates.js");
var import_OperationalCredentialsTypes = require("../common/OperationalCredentialsTypes.js");
var import_CommissioningError = require("./CommissioningError.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("ControllerCommissioner");
var CommissioningStepResultCode = /* @__PURE__ */ ((CommissioningStepResultCode2) => {
  CommissioningStepResultCode2[CommissioningStepResultCode2["Success"] = 0] = "Success";
  CommissioningStepResultCode2[CommissioningStepResultCode2["Failure"] = 1] = "Failure";
  CommissioningStepResultCode2[CommissioningStepResultCode2["Skipped"] = 2] = "Skipped";
  CommissioningStepResultCode2[CommissioningStepResultCode2["Stop"] = 3] = "Stop";
  return CommissioningStepResultCode2;
})(CommissioningStepResultCode || {});
class MaximumCommissionedFabricsReachedError extends import_CommissioningError.CommissioningError {
}
class CommissioningTimeoutError extends import_CommissioningError.CommissioningError {
}
class DeviceAlreadyCommissionedToThisFabricError extends import_CommissioningError.CommissioningError {
}
class FabricLabelConflictError extends import_CommissioningError.CommissioningError {
}
class WifiOrThreadNetworkCredentialsNotConfiguredError extends import_CommissioningError.CommissioningError {
}
class WifiNetworkSetupFailedError extends import_CommissioningError.CommissioningError {
}
class ThreadNetworkSetupFailedError extends import_CommissioningError.CommissioningError {
}
class NodeIdConflictError extends import_CommissioningError.CommissioningError {
}
class OperativeConnectionFailedError extends import_CommissioningError.CommissioningError {
}
class RecoverableCommissioningError extends import_CommissioningError.CommissioningError {
}
const DEFAULT_FAILSAFE_TIME = import_general.Minutes.one;
const BTP_IDLE_ALIVE_INTERVAL = (0, import_general.Seconds)(25);
const CASE_RECONNECT_TIMEOUT = (0, import_general.Minutes)(5);
const MIN_NETWORK_SCAN_TIMEOUT_SECONDS = 60;
const MIN_NETWORK_CONNECT_TIMEOUT_SECONDS = 60;
const RootEndpointNumber = (0, import_types.EndpointNumber)(0);
class ControllerCommissioningFlow {
  interaction;
  ca;
  fabric;
  transitionToCase;
  commissioningOptions;
  commissioningSteps = new Array();
  commissioningStepResults = /* @__PURE__ */ new Map();
  #clusterClients = /* @__PURE__ */ new Map();
  #commissioningStartedTime;
  #commissioningExpiryTime;
  #currentFailSafeEndTime;
  #dacPublicKey;
  lastBreadcrumb = 1;
  collectedCommissioningData = {};
  #defaultFailSafeTime = DEFAULT_FAILSAFE_TIME;
  #armFailsafeInterval;
  constructor(interaction, ca, fabric, commissioningOptions, transitionToCase) {
    this.interaction = interaction;
    this.ca = ca;
    this.fabric = fabric;
    this.transitionToCase = transitionToCase;
    this.commissioningOptions = commissioningOptions;
    logger.debug(`Commissioning options: ${import_general.Diagnostic.json(commissioningOptions)}`);
    this.#initializeCommissioningSteps();
  }
  async [Symbol.asyncDispose]() {
    await this.interaction.close();
  }
  /**
   * Execute the commissioning process in the defined order. The steps are sorted before execution based on the step
   * number and sub step number.
   * If >50% of the failsafe time has passed, the failsafe timer is re-armed (50% of 60s default are 30s and each
   * action is allowed to take 30s at minimum based on specs).
   */
  async executeCommissioning() {
    this.#sortSteps();
    let failSafeTimerReArmedAfterPreviousStep = false;
    for (const step of this.commissioningSteps) {
      logger.info(`Executing commissioning step ${step.stepNumber}.${step.subStepNumber}: ${step.name}`);
      try {
        if (step.reArmFailsafe && !failSafeTimerReArmedAfterPreviousStep && !this.#armFailsafeInterval?.isRunning) {
          logger.debug(`Re-Arming failsafe timer before executing step`);
          await this.#armFailsafe();
        }
        failSafeTimerReArmedAfterPreviousStep = false;
        const result = await step.stepLogic();
        this.#setCommissioningStepResult(step, result);
        if (this.#currentFailSafeEndTime !== void 0) {
          if (this.#commissioningExpiryTime !== void 0 && import_general.Time.nowMs > this.#commissioningExpiryTime) {
            logger.error(
              `Commissioning step ${step.stepNumber}.${step.subStepNumber}: ${step.name} succeeded, but commissioning took too long in general!`
            );
            const maxTimeS = this.collectedCommissioningData.basicCommissioningInfo?.maxCumulativeFailsafeSeconds;
            throw new CommissioningTimeoutError(
              `Commissioning time exceeds the maximum timeframe${maxTimeS ? ` of ${maxTimeS}s` : ""}`
            );
          }
          const timeLeft = (0, import_general.Timespan)(import_general.Time.nowMs, this.#currentFailSafeEndTime).duration;
          if (timeLeft < this.#defaultFailSafeTime / 2) {
            logger.info(
              `After Commissioning step ${step.stepNumber}.${step.subStepNumber}: ${step.name} succeeded, ${import_general.Duration.format(timeLeft)} left for failsafe timer, re-arming failsafe`
            );
            await this.#armFailsafe();
            failSafeTimerReArmedAfterPreviousStep = true;
          }
        }
        if (result.code === 3 /* Stop */) {
          break;
        }
      } catch (e) {
        const error = (0, import_general.asError)(e);
        if (error instanceof RecoverableCommissioningError) {
          logger.warn(
            `Commissioning step ${step.stepNumber}.${step.subStepNumber}: ${step.name} failed with recoverable error:`,
            import_general.Diagnostic.errorMessage(error),
            import_general.Diagnostic.weak("(continuing commissioning)")
          );
        } else if ((0, import_general.causedBy)(error, import_CommissioningError.CommissioningError, import_types.StatusResponseError)) {
          logger.error(
            `Commissioning step ${step.stepNumber}.${step.subStepNumber}: ${step.name} failed with error:`,
            import_general.Diagnostic.errorMessage(error),
            import_general.Diagnostic.weak("(terminating commissioning)")
          );
          await this.#resetFailsafeTimer();
          import_types.StatusResponseError.accept(error);
          throw (0, import_general.repackErrorAs)(error, import_CommissioningError.CommissioningError);
        } else {
          throw error;
        }
      }
    }
  }
  /**
   * Convenience method to get a number of attributes in a read and return exactly these values in the order as
   * defined in the request
   */
  async #readConcreteAttributeValues(request) {
    const attributeMap = /* @__PURE__ */ new Map();
    if (request.attributeRequests === void 0) {
      throw new import_general.ImplementationError("Can only handle Attribute reads");
    }
    for (const { endpointId, clusterId, attributeId } of request.attributeRequests) {
      if (endpointId === void 0 || clusterId === void 0) {
        continue;
      }
      attributeMap.set(`${endpointId}-${clusterId}-${attributeId}`, void 0);
    }
    for await (const data of this.interaction.read(request)) {
      for (const entry of data) {
        if (entry.kind !== "attr-value") {
          continue;
        }
        const {
          path: { endpointId, clusterId, attributeId },
          value
        } = entry;
        const key = `${endpointId}-${clusterId}-${attributeId}`;
        if (!attributeMap.has(key)) {
          continue;
        }
        attributeMap.set(key, value);
      }
    }
    return [...attributeMap.values()];
  }
  // TODO improve response typing
  async #invokeCommand(request, options = {}) {
    for await (const data of this.interaction.invoke(
      (0, import_Invoke.Invoke)({
        commands: [request],
        ...options
      })
    )) {
      for (const entry of data) {
        switch (entry.kind) {
          case "cmd-status":
            if (entry.status !== import_types.Status.Success) {
              throw import_types.StatusResponseError.create(entry.status, void 0, entry.clusterStatus);
            }
            return;
          case "cmd-response":
            return entry.data;
        }
      }
    }
  }
  /**
   * Initialize commissioning steps and add them in the default order as defined by
   * @see {@link MatterSpecification.v13.Core} § 5.5
   */
  #initializeCommissioningSteps() {
    this.commissioningSteps.push({
      stepNumber: 0,
      // Preparation
      subStepNumber: 1,
      name: "GetInitialData",
      stepLogic: () => this.#getInitialData()
    });
    this.commissioningSteps.push({
      stepNumber: 7,
      subStepNumber: 1,
      name: "GeneralCommissioning.ArmFailsafe",
      stepLogic: () => this.#armFailsafe()
    });
    this.commissioningSteps.push({
      stepNumber: 8,
      subStepNumber: 1,
      name: "GeneralCommissioning.ConfigureRegulatoryInformation",
      stepLogic: () => this.#configureRegulatoryInformation()
    });
    this.commissioningSteps.push({
      stepNumber: 8,
      subStepNumber: 2,
      name: "TimeSynchronization.SynchronizeTime",
      stepLogic: () => this.#synchronizeTime()
    });
    this.commissioningSteps.push({
      stepNumber: 10,
      subStepNumber: 1,
      name: "OperationalCredentials.DeviceAttestation",
      stepLogic: () => this.#deviceAttestation()
    });
    this.commissioningSteps.push({
      stepNumber: 11,
      // includes 11-13
      subStepNumber: 1,
      name: "OperationalCredentials.Certificates",
      stepLogic: () => this.#certificates()
    });
    this.commissioningSteps.push({
      stepNumber: 15,
      subStepNumber: 1,
      name: "AccessControl",
      stepLogic: () => this.#configureAccessControlLists()
    });
    if (this.interaction.channelType === import_general.ChannelType.BLE) {
      this.commissioningSteps.push({
        stepNumber: 16,
        subStepNumber: 1,
        name: "NetworkCommissioning.Validate",
        stepLogic: () => this.#validateNetwork()
      });
      if (this.commissioningOptions.wifiNetwork !== void 0) {
        this.commissioningSteps.push({
          stepNumber: 16,
          // includes step 17
          subStepNumber: 2,
          name: "NetworkCommissioning.Wifi",
          reArmFailsafe: true,
          stepLogic: () => this.#configureNetworkWifi()
        });
      }
      if (this.commissioningOptions.threadNetwork !== void 0) {
        this.commissioningSteps.push({
          stepNumber: 16,
          // includes step 17
          subStepNumber: 3,
          name: "NetworkCommissioning.Thread",
          reArmFailsafe: true,
          stepLogic: () => this.#configureNetworkThread()
        });
      }
    } else {
      logger.info(
        `Skipping NetworkCommissioning steps because the device is already on IP network (${this.interaction.channelType})`
      );
    }
    this.commissioningSteps.push({
      stepNumber: 18,
      // includes step 19 (CASE connection)
      subStepNumber: 1,
      name: "Reconnect",
      stepLogic: () => this.#reconnectWithDevice()
    });
    this.commissioningSteps.push({
      stepNumber: 20,
      subStepNumber: 1,
      name: "GeneralCommissioning.Complete",
      stepLogic: () => this.#completeCommissioning()
    });
    this.commissioningSteps.push({
      stepNumber: 98,
      // Should be allowed in Step 13, but Tasmota is not supporting this
      subStepNumber: 1,
      name: "OperationalCredentials.UpdateFabricLabel",
      stepLogic: () => this.#updateFabricLabel()
    });
    if (this.commissioningOptions.otaUpdateProviderLocation !== void 0) {
      this.commissioningSteps.push({
        stepNumber: 99,
        subStepNumber: 1,
        name: "AdditionalLogic.AddDefaultOtaProvider",
        stepLogic: () => this.#addDefaultOtaProvider()
      });
    }
  }
  #sortSteps() {
    this.commissioningSteps.sort((a, b) => {
      if (a.stepNumber !== b.stepNumber) return a.stepNumber - b.stepNumber;
      return a.subStepNumber - b.subStepNumber;
    });
  }
  #setCommissioningStepResult(step, result) {
    this.commissioningStepResults.set(`${step.stepNumber}-${step.subStepNumber}`, result);
  }
  getCommissioningStepResult(stepNumber, subStepNumber) {
    return this.commissioningStepResults.get(`${stepNumber}-${subStepNumber}`);
  }
  /** Helper method to check for errorCode/debugTest responses and throw error on failure */
  #ensureOperationalCredentialsSuccess(context, { statusCode, debugText, fabricIndex }) {
    logger.debug(
      `Commissioning step ${context} returned ${import_operational_credentials.OperationalCredentials.NodeOperationalCertStatus[statusCode]} (${statusCode}), ${debugText}${fabricIndex !== void 0 ? `, fabricIndex: ${fabricIndex}` : ""}`
    );
    if (statusCode === import_operational_credentials.OperationalCredentials.NodeOperationalCertStatus.Ok) return;
    if (context === "addNoc") {
      if (statusCode === import_operational_credentials.OperationalCredentials.NodeOperationalCertStatus.FabricConflict) {
        throw new DeviceAlreadyCommissionedToThisFabricError(
          `Commission error: This device is already commissioned into this fabric. You cannot commission it again.`
        );
      } else if (statusCode === import_operational_credentials.OperationalCredentials.NodeOperationalCertStatus.TableFull) {
        throw new MaximumCommissionedFabricsReachedError(
          `Commission error: This device reached the maximum number of fabrics it can be part of. Please remove a fabric before trying to add another one.`
        );
      }
    } else if (context === "updateFabricLabel" && statusCode === import_operational_credentials.OperationalCredentials.NodeOperationalCertStatus.LabelConflict) {
      throw new FabricLabelConflictError(
        `Cannot set fabric label to "${this.fabric.label}" because another fabric on this device already uses this name. Please adjust fabric labels to be unique.`
      );
    }
    throw new import_CommissioningError.CommissioningError(
      `Commission error for "${context}": ${import_operational_credentials.OperationalCredentials.NodeOperationalCertStatus[statusCode]} (${statusCode}), ${debugText}${fabricIndex !== void 0 ? `, fabricIndex: ${fabricIndex}` : ""}`
    );
  }
  /** Helper method to check for errorCode/debugTest responses and throw error on failure */
  #ensureGeneralCommissioningSuccess(context, { errorCode, debugText }) {
    logger.debug(`Commissioning step ${context} returned ${errorCode}, ${debugText}`);
    if (errorCode === import_general_commissioning.GeneralCommissioning.CommissioningError.Ok) return;
    throw new import_CommissioningError.CommissioningError(
      `Commission error for "${context}": ${import_general_commissioning.GeneralCommissioning.CommissioningError[errorCode]} (${errorCode})${debugText ? `, ${debugText}` : ""}`
    );
  }
  /**
   * Initial Step to receive some common data used by other steps
   */
  async #getInitialData() {
    const [
      supportedFabrics,
      commissionedFabrics,
      rootPartsList,
      rootServerList,
      vendorId,
      productId,
      productName,
      supportsConcurrentConnection
    ] = await this.#readConcreteAttributeValues(
      (0, import_Read.Read)(
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_operational_credentials.OperationalCredentials,
          attributes: ["supportedFabrics", "commissionedFabrics"]
        }),
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_descriptor.Descriptor,
          attributes: ["partsList", "serverList"]
        }),
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_basic_information.BasicInformation,
          attributes: ["vendorId", "productId", "productName"]
        }),
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_general_commissioning.GeneralCommissioning,
          attributes: ["supportsConcurrentConnection"]
        })
      )
    );
    if (commissionedFabrics >= supportedFabrics) {
      throw new MaximumCommissionedFabricsReachedError(
        `Commissioned fabrics (${commissionedFabrics}) exceed supported fabrics (${supportedFabrics}). Please remove some fabrics before commissioning.`
      );
    }
    this.collectedCommissioningData.rootPartsList = rootPartsList;
    this.collectedCommissioningData.rootServerList = rootServerList;
    this.collectedCommissioningData.vendorId = vendorId;
    this.collectedCommissioningData.productId = productId;
    this.collectedCommissioningData.productName = productName;
    this.collectedCommissioningData.supportsConcurrentConnection = supportsConcurrentConnection;
    const networkData = this.interaction.read(
      (0, import_Read.Read)(
        import_Read.Read.Attribute({
          cluster: import_network_commissioning.NetworkCommissioning,
          attributes: ["featureMap", "networks"]
        }),
        import_Read.Read.Attribute({
          cluster: import_ota_software_update_requestor.OtaSoftwareUpdateRequestor,
          attributes: ["defaultOtaProviders"]
        })
      )
    );
    const networkFeatures = new Array();
    const networkStatus = new Array();
    const otaRequestors = new Array();
    for await (const data of networkData) {
      for (const entry of data) {
        if (entry.kind !== "attr-value") {
          continue;
        }
        const {
          path: { endpointId, attributeId },
          value
        } = entry;
        switch (attributeId) {
          case import_types.Global.attributes.featureMap.id:
            networkFeatures.push({
              endpointId,
              value
            });
            break;
          case import_network_commissioning.NetworkCommissioning.attributes.networks.id:
            networkStatus.push({
              endpointId,
              value
            });
            break;
          case import_ota_software_update_requestor.OtaSoftwareUpdateRequestor.attributes.defaultOtaProviders.id:
            otaRequestors.push(endpointId);
            break;
        }
      }
    }
    this.collectedCommissioningData.networkFeatures = networkFeatures;
    this.collectedCommissioningData.networkStatus = networkStatus;
    this.collectedCommissioningData.otaRequestorList = otaRequestors;
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 7
   * Commissioner SHALL re-arm the Fail-safe timer on the Commissionee to the desired commissioning
   * timeout within 60 seconds of the completion of a PASE session establishment, using the
   * ArmFailSafe command (see Section 11.10.6.2, “ArmFailSafe Command”). A Commissioner MAY
   * collect device information including guidance on the fail-safe value from the Commissionee by
   * reading BasicCommissioningInfo attribute (see Section 11.10.5.2, “BasicCommissioningInfo
   * Attribute”) before invoking the ArmFailSafe command.
   */
  async #armFailsafe(time) {
    if (this.collectedCommissioningData.basicCommissioningInfo === void 0) {
      const [basicCommissioningInfo] = await this.#readConcreteAttributeValues(
        (0, import_Read.Read)(
          import_Read.Read.Attribute({
            endpoint: RootEndpointNumber,
            cluster: import_general_commissioning.GeneralCommissioning,
            attributes: ["basicCommissioningInfo"]
          })
        )
      );
      this.collectedCommissioningData.basicCommissioningInfo = basicCommissioningInfo;
      this.#defaultFailSafeTime = (0, import_general.Seconds)(basicCommissioningInfo.failSafeExpiryLengthSeconds);
      this.#commissioningStartedTime = import_general.Time.nowMs;
      this.#commissioningExpiryTime = (0, import_general.Timestamp)(
        this.#commissioningStartedTime + (0, import_general.Seconds)(basicCommissioningInfo.maxCumulativeFailsafeSeconds)
      );
    }
    const expiryLength = time ?? this.#defaultFailSafeTime;
    this.#ensureGeneralCommissioningSuccess(
      "armFailSafe",
      await this.#invokeCommand({
        endpoint: RootEndpointNumber,
        cluster: import_general_commissioning.GeneralCommissioning,
        command: "armFailSafe",
        fields: {
          breadcrumb: this.lastBreadcrumb,
          expiryLengthSeconds: import_general.Seconds.of(expiryLength)
        }
      })
    );
    this.#currentFailSafeEndTime = (0, import_general.Timestamp)(import_general.Time.nowMs + expiryLength);
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  get #failSafeTimeLeft() {
    if (this.#currentFailSafeEndTime === void 0) {
      return import_general.Instant;
    }
    return import_general.Duration.max((0, import_general.Timespan)(import_general.Time.nowMs, this.#currentFailSafeEndTime).duration, 0);
  }
  async #ensureFailsafeTimerFor(maxProcessingTime) {
    const minFailsafeTime = this.interaction.maximumPeerResponseTime(maxProcessingTime, true);
    if (this.interaction.channelType === import_general.ChannelType.BLE) {
      this.#armFailsafeInterval?.stop();
      this.#armFailsafeInterval = import_general.Time.getPeriodicTimer(
        "Re-Arm Failsafe during longer interactions",
        BTP_IDLE_ALIVE_INTERVAL,
        () => {
          if (this.interaction.session?.isClosed) {
            logger.debug("Skipping periodic armFailSafe: session is closed");
            this.#armFailsafeInterval?.stop();
            this.#armFailsafeInterval = void 0;
            return;
          }
          const now = import_general.Time.nowMs;
          if (this.#commissioningExpiryTime !== void 0 && now < this.#commissioningExpiryTime) {
            logger.debug(
              `Re-Arm Failsafe Timer during longer actions with device. Time left: ${import_general.Duration.format((0, import_general.Timespan)(now, this.#commissioningExpiryTime).duration)}`
            );
            this.#armFailsafe().catch((error) => {
              logger.info("Error while re-arming failsafe during reconnect", error);
              this.#armFailsafeInterval?.stop();
              this.#armFailsafeInterval = void 0;
            });
          } else {
            this.#armFailsafeInterval?.stop();
            this.#armFailsafeInterval = void 0;
          }
        }
      ).start();
    }
    const timeLeft = this.#failSafeTimeLeft;
    if (timeLeft < minFailsafeTime) {
      logger.debug(
        `Failsafe timer has only ${import_general.Duration.format(timeLeft)} left, re-arming for at least ${import_general.Duration.format(minFailsafeTime)}`
      );
      await this.#armFailsafe(import_general.Duration.max(minFailsafeTime, this.#defaultFailSafeTime));
    } else {
      logger.debug(`Failsafe timer is already set for at least ${import_general.Duration.format(timeLeft)}`);
    }
  }
  /**
   * On BLE the device may drop the commissioning channel when `connectNetwork` is sent, so the
   * failsafe must cover both the connect time and the subsequent CASE reconnection.  We arm
   * pessimistically whenever BLE is in play — not just when the device has declared
   * `supportsConcurrentConnection=false` — because some devices report concurrent support but
   * behave non-concurrent in practice.  Once BLE is gone we cannot extend the failsafe, and if
   * it expires on the device before CASE reconnect lands, the device rolls back addNOC.
   *
   * On IP transports, BLE drop is not a concern and the declared flag can be trusted.
   */
  #connectNetworkFailsafeTime(connectMaxTimeSeconds) {
    const pessimistic = this.interaction.channelType === import_general.ChannelType.BLE || this.collectedCommissioningData.supportsConcurrentConnection === false;
    if (pessimistic) {
      return import_general.Duration.max(
        (0, import_general.Millis)((0, import_general.Seconds)(connectMaxTimeSeconds) + CASE_RECONNECT_TIMEOUT),
        this.#defaultFailSafeTime
      );
    }
    return (0, import_general.Seconds)(connectMaxTimeSeconds);
  }
  /**
   * Recognise transport-level failures of `connectNetwork` that indicate the device has stopped
   * responding on the commissioning channel — typically because it moved to the operational network.
   *
   * Device-returned `NetworkingStatus != Success` is NOT a transport error; those are real
   * spec-level failures (bad credentials, unknown network, etc.) and must surface normally.
   *
   * Bare `AbortedError` is deliberately excluded: a user-initiated cancel or shutdown abort must
   * propagate cleanly, not be swallowed as a "non-concurrent signal".  Transport aborts surface
   * with a more specific cause ({@link PeerUnresponsiveError} from MRP timeouts,
   * {@link BleDisconnectedError} from BLE loss) that `causedBy` already matches.
   */
  #isConnectNetworkTransportError(error) {
    return (0, import_general.causedBy)(
      error,
      import_Ble.BleDisconnectedError,
      // also matches BleChannelClosedError subclass
      import_PeerCommunicationError.PeerUnresponsiveError,
      import_general.NoResponseTimeoutError
    );
  }
  /**
   * Respond to a transport-level `connectNetwork` failure by treating the device as non-concurrent:
   * stop the periodic failsafe re-arm timer (BLE is gone anyway) and flip the flag so
   * `#reconnectWithDevice` takes the non-concurrent path.
   */
  #handleConnectNetworkTransportError(error) {
    logger.warn(
      "connectNetwork did not complete on the commissioning channel \u2014 assuming device is non-concurrent, proceeding to CASE",
      import_general.Diagnostic.errorMessage((0, import_general.asError)(error))
    );
    this.collectedCommissioningData.supportsConcurrentConnection = false;
    this.#armFailsafeInterval?.stop();
    this.#armFailsafeInterval = void 0;
  }
  /**
   * Issue `NetworkCommissioning.connectNetwork` for the given network and classify the outcome.
   *
   * - `transportFailure: true` means the invoke threw a transport-level error (BLE/MRP gone).
   *   The caller should return {@link CommissioningStepResultCode.Success} so the flow falls
   *   through into the non-concurrent CASE reconnect path — the failsafe flag has already
   *   been flipped by {@link #handleConnectNetworkTransportError}.
   * - `transportFailure: false` carries the device's actual response and is the caller's job
   *   to validate ({@link NetworkCommissioning.NetworkCommissioningStatus}, debug text, etc).
   *
   * Shared between the WiFi and Thread paths so the try/catch + ensureFailsafe dance lives in
   * one place and both media stay in lockstep.
   */
  async #invokeConnectNetwork(networkId, connectMaxTimeSeconds) {
    await this.#ensureFailsafeTimerFor(this.#connectNetworkFailsafeTime(connectMaxTimeSeconds));
    try {
      const { networkingStatus, debugText } = await this.#invokeCommand(
        {
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          command: "connectNetwork",
          fields: {
            networkId,
            breadcrumb: this.lastBreadcrumb++
          }
        },
        {
          expectedProcessingTime: (0, import_general.Seconds)(connectMaxTimeSeconds)
        }
      );
      return { transportFailure: false, networkingStatus, debugText };
    } catch (error) {
      if (!this.#isConnectNetworkTransportError(error)) throw error;
      this.#handleConnectNetworkTransportError(error);
      return { transportFailure: true };
    }
  }
  async #resetFailsafeTimer() {
    if (this.#currentFailSafeEndTime === void 0) return;
    try {
      await this.#invokeCommand({
        endpoint: RootEndpointNumber,
        cluster: import_general_commissioning.GeneralCommissioning,
        command: "armFailSafe",
        fields: {
          breadcrumb: this.lastBreadcrumb,
          expiryLengthSeconds: 0
        }
      });
      this.#currentFailSafeEndTime = void 0;
    } catch (error) {
      logger.error(`Error while resetting failsafe timer`, error);
    }
  }
  /**
   * Step 8 - 1
   * Commissioner SHALL configure regulatory information if the Commissionee has at least one instance of
   * the Network Commissioning cluster on any endpoint with either the WI (i.e. Wi-Fi) or TH (i.e. Thread)
   * feature flags set in its FeatureMap, Commissioner SHALL configure regulatory information in the
   * Commissionee using the SetRegulatoryConfig command.
   */
  async #configureRegulatoryInformation() {
    if (this.collectedCommissioningData.networkFeatures === void 0) {
      throw new import_CommissioningError.CommissioningError(
        "Initial information collection failed. No network features collected. This should never happen."
      );
    }
    const hasRadioNetwork = this.collectedCommissioningData.networkFeatures.some(
      ({ value: { wiFiNetworkInterface, threadNetworkInterface } }) => wiFiNetworkInterface || threadNetworkInterface
    );
    if (hasRadioNetwork) {
      let [locationCapability] = await this.#readConcreteAttributeValues(
        (0, import_Read.Read)(
          import_Read.Read.Attribute({
            endpoint: RootEndpointNumber,
            cluster: import_general_commissioning.GeneralCommissioning,
            attributes: ["locationCapability"]
          })
        )
      );
      if (locationCapability === import_general_commissioning.GeneralCommissioning.RegulatoryLocationType.IndoorOutdoor) {
        locationCapability = this.commissioningOptions.regulatoryLocation;
      } else {
        logger.debug(
          `Device does not support a configurable regulatory location type. Location is set to "${locationCapability === import_general_commissioning.GeneralCommissioning.RegulatoryLocationType.Indoor ? "Indoor" : "Outdoor"}"`
        );
      }
      let countryCode = this.commissioningOptions.regulatoryCountryCode;
      const regulatoryResult = await this.#invokeCommand(
        {
          endpoint: RootEndpointNumber,
          cluster: import_general_commissioning.GeneralCommissioning,
          command: "setRegulatoryConfig",
          fields: {
            breadcrumb: this.lastBreadcrumb++,
            newRegulatoryConfig: locationCapability,
            countryCode
          }
        },
        {
          useExtendedFailSafeMessageResponseTimeout: true
        }
      );
      if (regulatoryResult.errorCode === import_general_commissioning.GeneralCommissioning.CommissioningError.ValueOutsideRange && countryCode !== "XX") {
        logger.debug(
          `Device does not support a configurable country code. Use "XX" instead of "${countryCode}"`
        );
        countryCode = "XX";
        this.#ensureGeneralCommissioningSuccess(
          "setRegulatoryConfig",
          await this.#invokeCommand(
            {
              endpoint: RootEndpointNumber,
              cluster: import_general_commissioning.GeneralCommissioning,
              command: "setRegulatoryConfig",
              fields: {
                breadcrumb: this.lastBreadcrumb,
                newRegulatoryConfig: locationCapability,
                countryCode
              }
            },
            {
              useExtendedFailSafeMessageResponseTimeout: true
            }
          )
        );
      } else {
        this.#ensureGeneralCommissioningSuccess("setRegulatoryConfig", regulatoryResult);
      }
      return {
        code: 0 /* Success */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    return {
      code: 2 /* Skipped */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 8 - 2
   * Commissioner SHOULD configure UTC time, timezone, and DST offset, if the Commissionee supports the
   * time synchronization cluster.
   * ▪ The Commissioner SHOULD configure UTC time using the SetUTCTime command.
   * ▪ The Commissioner SHOULD set the time zone using the SetTimeZone command, if the TimeZone Feature is supported.
   * ▪ The Commissioner SHOULD set the DST offsets using the SetDSTOffset command if the TimeZone Feature is supported, and the SetTimeZoneResponse from the Commissionee had the DSTOffsetsRequired field set to True.
   * ▪ The Commissioner SHOULD set a Default NTP server using the SetDefaultNTP command if the NTPClient Feature is supported and the DefaultNTP attribute is null. If the current value is non-null, Commissioners MAY opt to overwrite the current value.
   */
  async #synchronizeTime() {
    if (this.collectedCommissioningData.rootServerList !== void 0 && this.collectedCommissioningData.rootServerList.find((clusterId) => clusterId === import_time_synchronization.TimeSynchronization.id)) {
      logger.debug("TimeSynchronization cluster is supported");
    }
    return {
      code: 2 /* Skipped */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 10
   * Commissioner SHALL establish the authenticity of the Commissionee as a certified Matter device
   * (see Section 6.2.3, “Device Attestation Procedure”).
   */
  async #deviceAttestation() {
    const { certificate: deviceAttestation } = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_operational_credentials.OperationalCredentials,
        command: "certificateChainRequest",
        fields: {
          certificateType: import_operational_credentials.OperationalCredentials.CertificateChainType.DacCertificate
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    const { certificate: productAttestation } = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_operational_credentials.OperationalCredentials,
        command: "certificateChainRequest",
        fields: {
          certificateType: import_operational_credentials.OperationalCredentials.CertificateChainType.PaiCertificate
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    const attestationNonce = this.fabric.crypto.randomBytes(32);
    const { attestationElements, attestationSignature } = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_operational_credentials.OperationalCredentials,
        command: "attestationRequest",
        fields: {
          attestationNonce
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    const { attestation } = this.commissioningOptions;
    try {
      const result = await import_DeviceAttestationValidator.DeviceAttestationValidator.validate(
        {
          crypto: this.fabric.crypto,
          dclCertificateService: attestation.dclCertificateService,
          attestationChallenge: attestation.challengeKey
        },
        {
          dac: deviceAttestation,
          pai: productAttestation,
          attestationElements,
          attestationSignature,
          attestationNonce,
          vendorId: this.collectedCommissioningData.vendorId,
          productId: this.collectedCommissioningData.productId
        }
      );
      if (result.findings.length > 0) {
        const proceed = await this.#resolveAttestationFindings(result.findings);
        if (!proceed) {
          throw new import_CommissioningError.CommissioningError(
            `Device attestation produced ${result.findings.length} finding(s) and was rejected by policy`
          );
        }
        logger.info(
          `Device attestation successfully verified with ${result.findings.length} accepted finding(s)`
        );
      } else {
        logger.info("Device attestation successfully verified");
      }
    } catch (error) {
      if (error instanceof import_DeviceAttestationValidator.DeviceAttestationError) {
        const errorFinding = {
          level: "error",
          type: error.failure,
          message: error.message
        };
        const proceed = await this.#resolveAttestationFindings([errorFinding]);
        if (!proceed) {
          throw error;
        }
      } else {
        throw error;
      }
    }
    this.#dacPublicKey = (0, import_general.PublicKey)(import_AttestationCertificates.Dac.fromAsn1(deviceAttestation).cert.ellipticCurvePublicKey);
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /** Resolve attestation findings according to the configured policy. */
  #resolveAttestationFindings(findings) {
    const policy = this.commissioningOptions.attestation.onFailure;
    if (typeof policy === "function") {
      return policy(findings);
    }
    for (const f of findings) {
      switch (policy) {
        case void 0:
          logger.warn("Attestation finding accepted for backward compatibility:", f.type, f.message);
          break;
        case true:
          logger.info("Attestation finding accepted by policy:", f.type, f.message);
          break;
        case false:
          logger.info("Attestation finding, rejecting:", f.type, f.message);
          break;
      }
    }
    return policy !== false;
  }
  // TODO consider Distributed Compliance Ledger Info about Commissioning Flow
  /**
   * Step 11-13
   * 11: Following the Device Attestation Procedure yielding a decision to proceed with commissioning, the Commissioner
   *     SHALL request operational CSR from Commissionee using the CSRRequest command (see Section 11.17.6.5,
   *     “CSRRequest Command”). The CSRRequest command will cause the generation of a new operational key pair at the
   *     Commissionee.
   * 12: Commissioner SHALL generate or otherwise obtain an Operational Certificate containing Operational ID after
   *     receiving the CSRResponse command from the Commissionee (see Section 11.17.6.5, “CSRRequest Command”), using
   *     implementation-specific means.
   * 13: Commissioner SHALL install operational credentials (see Figure 40, “Node Operational Credentials
   *     flow”) on the Commissionee using the AddTrustedRootCertificate and AddNOC commands,
   *     and SHALL use the UpdateFabricLabel command to set a string that the user can recognize and
   *     relate to this Commissioner/Administrator.
   *     The AdminVendorId field of the AddNOC command SHALL be set to a value for which the Vendor Schema in
   *     DCL contains the name and other information of the Commissioner’s manufacturer.
   */
  async #certificates() {
    const { nocsrElements, attestationSignature: csrSignature } = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_operational_credentials.OperationalCredentials,
        command: "csrRequest",
        fields: { csrNonce: this.fabric.crypto.randomBytes(32) }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    if (nocsrElements.byteLength === 0 || csrSignature.byteLength === 0) {
      throw new import_general.UnexpectedDataError("Invalid response from device");
    }
    if (this.#dacPublicKey !== void 0) {
      try {
        await this.fabric.crypto.verifyEcdsa(
          this.#dacPublicKey,
          import_general.Bytes.concat(nocsrElements, this.commissioningOptions.attestation.challengeKey),
          new import_general.EcdsaSignature(csrSignature)
        );
      } catch {
        throw new import_CommissioningError.CommissioningError("CSR signature verification failed against DAC public key");
      }
    }
    const { certSigningRequest } = import_OperationalCredentialsTypes.TlvCertSigningRequest.decode(nocsrElements);
    const operationalPublicKey = await import_Certificate.Certificate.getPublicKeyFromCsr(this.ca.crypto, certSigningRequest);
    await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_operational_credentials.OperationalCredentials,
        command: "addTrustedRootCertificate",
        fields: {
          rootCaCertificate: this.ca.rootCert
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    const peerOperationalCert = await this.ca.generateNoc(
      operationalPublicKey,
      this.fabric.fabricId,
      this.interaction.address.nodeId
    );
    const addNocResponse = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_operational_credentials.OperationalCredentials,
        command: "addNoc",
        fields: {
          nocValue: peerOperationalCert,
          icacValue: this.ca.icacCert ?? new Uint8Array(0),
          ipkValue: this.fabric.identityProtectionKey,
          adminVendorId: this.fabric.rootVendorId,
          caseAdminSubject: this.fabric.rootNodeId
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    this.#ensureOperationalCredentialsSuccess("addNoc", addNocResponse);
    const { fabricIndex } = addNocResponse;
    this.collectedCommissioningData.fabricIndex = fabricIndex;
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 13-2 (we do as 99 at the end because)
   * The Administrator having established a CASE session with the Commissionee over the operational network in the
   * previous steps SHALL invoke the CommissioningComplete command (see Section 11.9.6.6,
   * “CommissioningComplete Command”). A success response after invocation of the CommissioningComplete command ends
   * the commissioning process.
   */
  async #updateFabricLabel() {
    const { fabricIndex } = this.collectedCommissioningData;
    if (fabricIndex === void 0) {
      logger.error("No fabric index available after addNoc. This should never happen.");
      return {
        code: 1 /* Failure */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    try {
      this.#ensureOperationalCredentialsSuccess(
        "updateFabricLabel",
        await this.#invokeCommand({
          endpoint: RootEndpointNumber,
          cluster: import_operational_credentials.OperationalCredentials,
          command: "updateFabricLabel",
          fields: {
            label: this.fabric.label,
            fabricIndex
          }
        })
      );
    } catch (error) {
      throw (0, import_general.repackErrorAs)(error, RecoverableCommissioningError);
    }
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 15
   * Commissioner MAY configure the Access Control List (see Access Control Cluster) on the Commissionee in any way
   * it sees fit, if the singular entry added by the AddNOC command in the previous step granting Administer
   * privilege over CASE authentication type for the Node ID provided with the command is not sufficient to express
   * its desired access control policies.
   */
  async #configureAccessControlLists() {
    return {
      code: 2 /* Skipped */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 16-17
   * 16: If the Commissionee both supports it and requires it, the Commissioner SHALL configure the operational network
   *     at the Commissionee using commands such as AddOrUpdateWiFiNetwork (see Section 11.8.7.3, “AddOrUpdateWiFiNetwork
   *     Command”) and AddOrUpdateThreadNetwork (see Section 11.8.7.4, “AddOrUpdateThreadNetwork Command”).
   *     A Commissionee requires network commissioning if it is not already on the desired operational network.
   *     A Commissionee supports network commissioning if it has any NetworkCommissioning cluster instances.
   *     A Commissioner MAY learn about the networks visible to the Commissionee using ScanNetworks command
   *     (see Section 11.8.7.1, “ScanNetworks Command”).
   * 17: The Commissioner SHALL trigger the Commissionee to connect to the operational network using ConnectNetwork
   *     command (see Section 11.8.7.9, “ConnectNetwork Command”) unless the Commissionee is already on the desired
   *     operational network.
   */
  async #validateNetwork() {
    if (this.collectedCommissioningData.networkFeatures === void 0 || this.collectedCommissioningData.networkStatus === void 0) {
      throw new import_CommissioningError.CommissioningError(
        "Initial information collection failed. No network features or status collected. This should never happen."
      );
    }
    if ((this.commissioningOptions.wifiNetwork === void 0 || !this.commissioningOptions.wifiNetwork.wifiSsid || !this.commissioningOptions.wifiNetwork.wifiCredentials) && (this.commissioningOptions.threadNetwork === void 0 || !this.commissioningOptions.threadNetwork.operationalDataset)) {
      const anyEthernetInterface = this.collectedCommissioningData.networkFeatures.length === 0 || this.collectedCommissioningData.networkFeatures.some(
        ({ value: { ethernetNetworkInterface } }) => ethernetNetworkInterface === true
      );
      const anyInterfaceConnected = this.collectedCommissioningData.networkStatus.length === 0 || this.collectedCommissioningData.networkStatus.some(
        ({ value }) => value.some(({ connected }) => connected)
      );
      if (!anyEthernetInterface && !anyInterfaceConnected) {
        throw new WifiOrThreadNetworkCredentialsNotConfiguredError(
          "No Wi-Fi/Thread network credentials are configured for commissioning and no Ethernet interface is available on the device and no interface already connected."
        );
      }
    }
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  async #configureNetworkWifi() {
    if (this.commissioningOptions.wifiNetwork === void 0) {
      logger.debug("WiFi network is not configured");
      return {
        code: 2 /* Skipped */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    if (this.collectedCommissioningData.networkFeatures !== void 0 && this.collectedCommissioningData.networkStatus !== void 0) {
      const rootNetworkFeatures = this.collectedCommissioningData.networkFeatures.find(
        ({ endpointId }) => endpointId === 0
      )?.value;
      const rootNetworkStatus = this.collectedCommissioningData.networkStatus.find(
        ({ endpointId }) => endpointId === 0
      )?.value;
      logger.debug(
        `Root Networks found: ${import_general.Diagnostic.json(rootNetworkFeatures)} - ${import_general.Diagnostic.json(rootNetworkStatus)}`
      );
      if (rootNetworkFeatures?.wiFiNetworkInterface !== true) {
        logger.debug("Commissionee does not support any WiFi network interface");
        return {
          code: 2 /* Skipped */,
          breadcrumb: this.lastBreadcrumb
        };
      }
      if (rootNetworkStatus !== void 0 && rootNetworkStatus.length > 0 && rootNetworkStatus[0].connected) {
        logger.debug("Commissionee is already connected to the WiFi network");
        this.collectedCommissioningData.successfullyConnectedToNetwork = true;
        return {
          code: 2 /* Skipped */,
          breadcrumb: this.lastBreadcrumb
        };
      }
    }
    logger.debug("Configuring WiFi network ...");
    const ssid = import_general.Bytes.fromString(this.commissioningOptions.wifiNetwork.wifiSsid);
    const credentials = import_general.Bytes.fromString(this.commissioningOptions.wifiNetwork.wifiCredentials);
    const [rawScanMaxTimeSeconds, rawConnectMaxTimeSeconds] = await this.#readConcreteAttributeValues(
      (0, import_Read.Read)(
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          attributes: ["scanMaxTimeSeconds", "connectMaxTimeSeconds"]
        })
      )
    );
    const scanMaxTimeSeconds = Math.max(rawScanMaxTimeSeconds ?? 0, MIN_NETWORK_SCAN_TIMEOUT_SECONDS);
    const connectMaxTimeSeconds = Math.max(rawConnectMaxTimeSeconds ?? 0, MIN_NETWORK_CONNECT_TIMEOUT_SECONDS);
    if (this.collectedCommissioningData.supportsConcurrentConnection !== false) {
      await this.#ensureFailsafeTimerFor((0, import_general.Seconds)(scanMaxTimeSeconds));
      const { networkingStatus, wiFiScanResults, debugText } = await this.#invokeCommand(
        {
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          command: "scanNetworks",
          fields: {
            ssid,
            breadcrumb: this.lastBreadcrumb++
          }
        },
        {
          expectedProcessingTime: (0, import_general.Seconds)(scanMaxTimeSeconds)
        }
      );
      if (networkingStatus !== import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus.Success) {
        throw new WifiNetworkSetupFailedError(`Commissionee failed to scan for WiFi networks: ${debugText}`);
      }
      if (wiFiScanResults === void 0 || wiFiScanResults.length === 0) {
        throw new WifiNetworkSetupFailedError(
          `Commissionee did not return any WiFi networks for the requested SSID ${this.commissioningOptions.wifiNetwork.wifiSsid}`
        );
      }
    }
    const {
      networkingStatus: addNetworkingStatus,
      debugText: addDebugText,
      networkIndex
    } = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_network_commissioning.NetworkCommissioning,
        command: "addOrUpdateWiFiNetwork",
        fields: {
          ssid,
          credentials,
          breadcrumb: this.lastBreadcrumb++
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    if (addNetworkingStatus !== import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus.Success) {
      throw new WifiNetworkSetupFailedError(`Commissionee failed to add WiFi network: ${addDebugText}`);
    }
    if (networkIndex === void 0) {
      throw new WifiNetworkSetupFailedError(`Commissionee did not return network index`);
    }
    logger.debug(
      `Commissionee added WiFi network ${this.commissioningOptions.wifiNetwork.wifiSsid} with network index ${networkIndex}`
    );
    const [updatedNetworks] = await this.#readConcreteAttributeValues(
      (0, import_Read.Read)(
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          attributes: ["networks"]
        })
      )
    );
    if (updatedNetworks[networkIndex] === void 0) {
      throw new WifiNetworkSetupFailedError(`Commissionee did not return network with index ${networkIndex}`);
    }
    const { networkId, connected } = updatedNetworks[networkIndex];
    if (connected) {
      this.collectedCommissioningData.successfullyConnectedToNetwork = true;
      logger.debug(
        `Commissionee is already connected to WiFi network ${this.commissioningOptions.wifiNetwork.wifiSsid} (networkId ${import_general.Bytes.toHex(networkId)})`
      );
      return {
        code: 0 /* Success */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    const connectResult = await this.#invokeConnectNetwork(networkId, connectMaxTimeSeconds);
    if (connectResult.transportFailure) {
      return {
        code: 0 /* Success */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    if (connectResult.networkingStatus !== import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus.Success) {
      throw new WifiNetworkSetupFailedError(
        `Commissionee failed to connect to WiFi network: ${connectResult.debugText}`
      );
    }
    this.collectedCommissioningData.successfullyConnectedToNetwork = true;
    logger.debug(
      `Commissionee successfully connected to WiFi network ${this.commissioningOptions.wifiNetwork.wifiSsid} (networkId ${import_general.Bytes.toHex(networkId)})`
    );
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  async #configureNetworkThread() {
    if (this.collectedCommissioningData.successfullyConnectedToNetwork) {
      logger.info("Node is already connected to a network. Skipping Thread configuration.");
      return {
        code: 2 /* Skipped */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    if (this.commissioningOptions.threadNetwork === void 0) {
      logger.debug("Thread network is not configured");
      return {
        code: 2 /* Skipped */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    if (this.collectedCommissioningData.networkFeatures !== void 0 && this.collectedCommissioningData.networkStatus !== void 0) {
      const rootNetworkFeatures = this.collectedCommissioningData.networkFeatures.find(
        ({ endpointId }) => endpointId === 0
      )?.value;
      const rootNetworkStatus = this.collectedCommissioningData.networkStatus.find(
        ({ endpointId }) => endpointId === 0
      )?.value;
      logger.debug(
        `Root Networks found: ${import_general.Diagnostic.json(rootNetworkFeatures)} - ${import_general.Diagnostic.json(rootNetworkStatus)}`
      );
      if (rootNetworkFeatures?.threadNetworkInterface !== true) {
        logger.debug("Commissionee does not support any Thread network interface");
        return {
          code: 2 /* Skipped */,
          breadcrumb: this.lastBreadcrumb
        };
      }
      if (rootNetworkStatus !== void 0 && rootNetworkStatus.length > 0 && rootNetworkStatus[0].connected) {
        logger.debug("Commissionee is already connected to the Thread network");
        return {
          code: 2 /* Skipped */,
          breadcrumb: this.lastBreadcrumb
        };
      }
    }
    logger.debug("Configuring Thread network ...");
    const [rawScanMaxTimeSeconds, rawConnectMaxTimeSeconds] = await this.#readConcreteAttributeValues(
      (0, import_Read.Read)(
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          attributes: ["scanMaxTimeSeconds", "connectMaxTimeSeconds"]
        })
      )
    );
    const scanMaxTimeSeconds = Math.max(rawScanMaxTimeSeconds ?? 0, MIN_NETWORK_SCAN_TIMEOUT_SECONDS);
    const connectMaxTimeSeconds = Math.max(rawConnectMaxTimeSeconds ?? 0, MIN_NETWORK_CONNECT_TIMEOUT_SECONDS);
    if (!this.commissioningOptions.threadNetwork?.networkName) {
      logger.info("Thread network name is not configured. Skip scanning for it.");
    } else if (this.collectedCommissioningData.supportsConcurrentConnection !== false) {
      await this.#ensureFailsafeTimerFor((0, import_general.Seconds)(scanMaxTimeSeconds));
      const { networkingStatus: networkingStatus2, threadScanResults, debugText: debugText2 } = await this.#invokeCommand(
        {
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          command: "scanNetworks",
          fields: { breadcrumb: this.lastBreadcrumb++ }
        },
        {
          expectedProcessingTime: (0, import_general.Seconds)(scanMaxTimeSeconds)
        }
      );
      if (networkingStatus2 !== import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus.Success) {
        throw new ThreadNetworkSetupFailedError(
          `Commissionee failed to scan for Thread networks: ${debugText2}`
        );
      }
      if (threadScanResults === void 0 || threadScanResults.length === 0) {
        throw new ThreadNetworkSetupFailedError(
          `Commissionee did not return any Thread networks for the requested Network ${this.commissioningOptions.threadNetwork.networkName}`
        );
      }
      const wantedNetworkFound = threadScanResults.find(
        ({ networkName }) => networkName === this.commissioningOptions.threadNetwork?.networkName
      );
      if (wantedNetworkFound === void 0) {
        throw new ThreadNetworkSetupFailedError(
          `Commissionee did not return the requested Network ${this.commissioningOptions.threadNetwork.networkName}: ${import_general.Diagnostic.json(threadScanResults)}`
        );
      }
      logger.debug(
        `Commissionee found wanted Thread network ${this.commissioningOptions.threadNetwork.networkName}: ${import_general.Diagnostic.json(wantedNetworkFound)}`
      );
    }
    const {
      networkingStatus: addNetworkingStatus,
      debugText: addDebugText,
      networkIndex
    } = await this.#invokeCommand(
      {
        endpoint: RootEndpointNumber,
        cluster: import_network_commissioning.NetworkCommissioning,
        command: "addOrUpdateThreadNetwork",
        fields: {
          operationalDataset: import_general.Bytes.fromHex(this.commissioningOptions.threadNetwork.operationalDataset),
          breadcrumb: this.lastBreadcrumb++
        }
      },
      {
        useExtendedFailSafeMessageResponseTimeout: true
      }
    );
    if (addNetworkingStatus !== import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus.Success) {
      throw new ThreadNetworkSetupFailedError(
        `Commissionee failed to add Thread network: status=${import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus[addNetworkingStatus]} (${addNetworkingStatus})${addDebugText ? ` ${addDebugText}` : ""}`
      );
    }
    if (networkIndex === void 0) {
      throw new ThreadNetworkSetupFailedError(`Commissionee did not return network index`);
    }
    logger.debug(
      `Commissionee added Thread network ${this.commissioningOptions.threadNetwork.networkName ?? "via operational dataset"} with network index ${networkIndex}`
    );
    const [updatedNetworks] = await this.#readConcreteAttributeValues(
      (0, import_Read.Read)(
        import_Read.Read.Attribute({
          endpoint: RootEndpointNumber,
          cluster: import_network_commissioning.NetworkCommissioning,
          attributes: ["networks"]
        })
      )
    );
    if (updatedNetworks[networkIndex] === void 0) {
      throw new ThreadNetworkSetupFailedError(`Commissionee did not return network with index ${networkIndex}`);
    }
    const { networkId, connected } = updatedNetworks[networkIndex];
    if (connected) {
      logger.debug(
        `Commissionee is already connected to Thread network ${this.commissioningOptions.threadNetwork.networkName ?? "via operational dataset"} (networkId ${import_general.Bytes.toHex(networkId)})`
      );
      return {
        code: 0 /* Success */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    const connectResult = await this.#invokeConnectNetwork(networkId, connectMaxTimeSeconds);
    if (connectResult.transportFailure) {
      return {
        code: 0 /* Success */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    const { networkingStatus, debugText } = connectResult;
    if (networkingStatus !== import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus.Success) {
      throw new ThreadNetworkSetupFailedError(
        `Commissionee failed to connect to Thread network: status=${import_network_commissioning.NetworkCommissioning.NetworkCommissioningStatus[networkingStatus]} (${networkingStatus})${debugText ? ` ${debugText}` : ""}`
      );
    }
    logger.debug(
      `Commissionee successfully connected to Thread network ${this.commissioningOptions.threadNetwork.networkName ?? "via operational dataset"} (networkId ${import_general.Bytes.toHex(networkId)})`
    );
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 18-19
   * 18: Finalization of the Commissioning process begins. An Administrator configured in the ACL of the Commissionee
   *     by the Commissioner SHALL use Operational Discovery to discover the Commissionee. This Administrator MAY be
   *     the Commissioner itself, or another Node to which the Commissioner has delegated the task.
   * 19: The Administrator SHALL open a CASE (see Section 4.13.2, “Certificate Authenticated Session Establishment
   *     (CASE)”) session with the Commissionee over the operational network.
   */
  async #reconnectWithDevice() {
    const isConcurrentFlow = this.collectedCommissioningData.supportsConcurrentConnection !== false;
    logger.debug(`Reconnecting with device with ${isConcurrentFlow ? "concurrent" : "non-concurrent"} flow ...`);
    if (isConcurrentFlow) {
      try {
        await this.#ensureFailsafeTimerFor(CASE_RECONNECT_TIMEOUT);
      } catch (error) {
        logger.info("Failed to re-arm failsafe before operational reconnect, proceeding anyway", error);
      }
    } else {
      this.#armFailsafeInterval?.stop();
    }
    let transitionResult;
    try {
      transitionResult = await this.transitionToCase(
        this.interaction.address,
        // Assume concurrent connections are supported if not know (which should not be the case when we came here)
        isConcurrentFlow
      );
    } catch (cause) {
      throw new OperativeConnectionFailedError("Operative reconnection with device failed", { cause });
    } finally {
      this.#armFailsafeInterval?.stop();
    }
    if (transitionResult === void 0) {
      logger.debug("CASE commissioning handled externally, terminating commissioning flow");
      return {
        code: 3 /* Stop */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    await this.interaction.close();
    this.interaction = transitionResult;
    this.#clusterClients.clear();
    logger.debug("Successfully reconnected with device ...");
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  /**
   * Step 20
   * The Administrator having established a CASE session with the Commissionee over the operational network in the
   * previous steps SHALL invoke the CommissioningComplete command (see Section 11.9.6.6,
   * “CommissioningComplete Command”). A success response after invocation of the CommissioningComplete command ends
   * the commissioning process.
   */
  async #completeCommissioning() {
    this.#ensureGeneralCommissioningSuccess(
      "commissioningComplete",
      await this.#invokeCommand(
        {
          endpoint: RootEndpointNumber,
          cluster: import_general_commissioning.GeneralCommissioning,
          command: "commissioningComplete"
        },
        {
          useExtendedFailSafeMessageResponseTimeout: true
        }
      )
    );
    this.#currentFailSafeEndTime = void 0;
    return {
      code: 0 /* Success */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  async #addDefaultOtaProvider() {
    if (this.commissioningOptions.otaUpdateProviderLocation === void 0) {
      return {
        code: 2 /* Skipped */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    if (!this.collectedCommissioningData?.otaRequestorList?.length) {
      logger.debug("No OTA Requestor found, skipping adding default OTA provider");
      return {
        code: 2 /* Skipped */,
        breadcrumb: this.lastBreadcrumb
      };
    }
    const { nodeId: providerNodeId, endpoint } = this.commissioningOptions.otaUpdateProviderLocation;
    let success = false;
    for (const requestorEndpoint of this.collectedCommissioningData.otaRequestorList) {
      try {
        const writeResult = await this.interaction.write(
          (0, import_Write.Write)(
            import_Write.Write.Attribute({
              endpoint: requestorEndpoint,
              cluster: import_ota_software_update_requestor.OtaSoftwareUpdateRequestor,
              attributes: ["defaultOtaProviders"],
              value: [
                {
                  providerNodeId,
                  endpoint,
                  fabricIndex: this.fabric.fabricIndex
                }
              ]
            })
          )
        );
        import_action.WriteResult.assertSuccess(writeResult);
        success = true;
        logger.debug(`Added default OTA provider on endpoint ${endpoint}`);
      } catch (error) {
        logger.info("Could not set default OTA provider", error);
      }
    }
    return {
      code: success ? 0 /* Success */ : 1 /* Failure */,
      breadcrumb: this.lastBreadcrumb
    };
  }
  close() {
    this.#armFailsafeInterval?.stop();
  }
}
