// @matter/protocol/dist/cjs/peer/PeerTimingParameters.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PeerTimingParameters_exports = {};
__export(PeerTimingParameters_exports, {
  PeerTimingParameters: () => PeerTimingParameters
});
module.exports = __toCommonJS(PeerTimingParameters_exports);
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const complete = /* @__PURE__ */ Symbol("complete-timing-parameters");
function PeerTimingParameters(options) {
  if (options && options[complete]) {
    return options;
  }
  const result = { ...PeerTimingParameters.defaults };
  if (options) {
    for (const key of Object.keys(options)) {
      const value = options[key];
      if (value !== void 0) {
        result[key] = value;
      }
    }
  }
  result[complete] = true;
  return result;
}
((PeerTimingParameters2) => {
  function merge(base, overrides) {
    const result = (0, import_general.merge)(base, overrides);
    result[complete] = true;
    return result;
  }
  PeerTimingParameters2.merge = merge;
  PeerTimingParameters2.defaults = {
    defaultConnectionTimeout: (0, import_general.Seconds)(90),
    maxDelayBetweenInitialContactRetries: (0, import_general.Minutes)(2),
    // We assume 30s processing time on peer for single Sigma actions, so give one IP a bit of time
    // to have a chance before potentially adding a load with a second try
    delayBeforeNextAddress: (0, import_general.Seconds)(45),
    delayAfterNetworkError: (0, import_general.Seconds)(15),
    delayAfterPeerError: (0, import_general.Minutes)(5),
    delayAfterUnhandledError: (0, import_general.Minutes)(2),
    kickThrottleInterval: (0, import_general.Seconds)(3),
    kickMinRetransmissions: 2,
    kickRestartCooldown: {
      addressChange: (0, import_general.Minutes)(30),
      connect: (0, import_general.Minutes)(10)
    },
    addressChangeStabilizationDelay: (0, import_general.Seconds)(10),
    addressChangeProbeCooldown: {
      minimum: (0, import_general.Minutes)(2),
      maximum: (0, import_general.Minutes)(60)
    }
  };
})(PeerTimingParameters || (PeerTimingParameters = {}));
