// @matter/protocol/dist/cjs/peer/PeerExchangeProvider.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PeerExchangeProvider_exports = {};
__export(PeerExchangeProvider_exports, {
  PeerExchangeProvider: () => PeerExchangeProvider
});
module.exports = __toCommonJS(PeerExchangeProvider_exports);
var import_PeerAddress = require("#peer/PeerAddress.js");
var import_ExchangeProvider = require("#protocol/ExchangeProvider.js");
var import_MRP = require("#protocol/MRP.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_PeerConnection = require("./PeerConnection.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class PeerExchangeProvider extends import_ExchangeProvider.ExchangeProvider {
  #peer;
  #context;
  constructor(peer, context) {
    super(context.exchanges);
    this.#peer = peer;
    this.#context = context;
  }
  get maxPathsPerInvoke() {
    return this.#peer.sessionParameters.maxPathsPerInvoke;
  }
  get peerAddress() {
    return this.#peer.address;
  }
  // TODO - TCP support
  channelType = import_general.ChannelType.UDP;
  async connect(options) {
    await this.#peer.connect(options);
  }
  async initiateExchange(options) {
    const abort = options?.abort;
    const isGroup = import_PeerAddress.PeerAddress.isGroup(this.#peer.address);
    while (true) {
      if (!isGroup && !options?.requireExistingSession) {
        await this.#peer.connect(options);
        abort?.throwIfAborted();
      }
      const network = this.#context.networks.select(this.#peer, options?.network);
      const slot = await network.semaphore.obtainSlot(abort);
      try {
        abort?.throwIfAborted();
        const session = isGroup ? await this.#context.sessions.groupSessionForAddress(this.#peer.address, this.#context.exchanges) : this.#peer.newestSession;
        if (session === void 0) {
          if (options?.requireExistingSession) {
            throw new import_general.InternalError("No existing session available for probe");
          }
          slot.close();
          continue;
        }
        const exchange = import_PeerConnection.PeerConnection.createExchange(
          this.#peer,
          this.#context.exchanges,
          session,
          network,
          options?.protocol ?? import_types.INTERACTION_PROTOCOL_ID,
          options?.addressOverride
        );
        exchange.closing.on(() => {
          slot.close();
        });
        return exchange;
      } catch (e) {
        slot.close();
        throw e;
      }
    }
  }
  maximumPeerResponseTime(expectedProcessingTime, includeMaximumSendingTime) {
    return import_MRP.MRP.maxPeerResponseTimeOf({
      channelType: this.channelType,
      isPeerActive: true,
      localSessionParameters: this.#context.sessions.sessionParameters,
      peerSessionParameters: includeMaximumSendingTime ? this.#peer.sessionParameters : void 0,
      expectedProcessingTime
    });
  }
}
