// @matter/protocol/dist/cjs/peer/PeerSet.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var PeerSet_exports = {};
__export(PeerSet_exports, {
  PeerSet: () => PeerSet
});
module.exports = __toCommonJS(PeerSet_exports);
var import_MdnsService = require("#mdns/MdnsService.js");
var import_PeerAddress = require("#peer/PeerAddress.js");
var import_SessionManager = require("#session/SessionManager.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_NetworkProfile = require("./NetworkProfile.js");
var import_Peer = require("./Peer.js");
var import_PeerTimingParameters = require("./PeerTimingParameters.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class PeerSet {
  #lifetime;
  #sessions;
  #peers = new import_general.BasicSet();
  #disconnected = (0, import_general.AsyncObservable)();
  #peerContext;
  #networks;
  #observers = new import_general.ObserverGroup();
  #exchanges;
  constructor(context) {
    const { lifetime, sessions, names, networks, timing, handleError } = context;
    this.#lifetime = lifetime.join("peers");
    this.#sessions = sessions;
    this.#networks = networks;
    const self = this;
    this.#peerContext = {
      sessions,
      names,
      networks,
      timing: (0, import_PeerTimingParameters.PeerTimingParameters)(timing),
      handleError,
      get exchanges() {
        if (self.#exchanges === void 0) {
          throw new import_general.ImplementationError("Client networking is not initialized");
        }
        return self.#exchanges;
      },
      openSocket: (address, abort) => this.#openSocket(address, abort),
      closed: (peer) => this.#peers.delete(peer),
      join: (name) => this.#lifetime.join(name)
    };
    this.#peers.added.on((peer) => {
      peer.sessions.deleted.on(() => {
        if (!peer.sessions.size) {
          this.#disconnected.emit(peer);
        }
      });
    });
    this.#sessions.sessions.added.on((session) => {
      if (session.peerAddress.fabricIndex === import_types.FabricIndex.NO_FABRIC || session.isClosed) {
        return;
      }
      this.addKnownPeer({
        address: session.peerAddress,
        operationalAddress: operationalAddressOf(session)
      });
    });
    this.#observers.on(this.#sessions.sessions.added, (session) => {
      if (session.fabric === void 0) {
        return;
      }
      this.for(session.peerAddress).sessions.add(session);
    });
  }
  set exchanges(exchanges) {
    this.#exchanges = exchanges;
  }
  get added() {
    return this.#peers.added;
  }
  get deleted() {
    return this.#peers.deleted;
  }
  get empty() {
    return this.#peers.empty;
  }
  get disconnected() {
    return this.#disconnected;
  }
  /**
   * Unconditional get.
   *
   * Creates the peer if not already present.
   */
  for(address) {
    let peer = this.get(address);
    if (peer) {
      return peer;
    }
    peer = new import_Peer.Peer({ address }, this.#peerContext);
    this.#peers.add(peer);
    return peer;
  }
  has(item) {
    return !!this.get(item);
  }
  get size() {
    return this.#peers.size;
  }
  find(predicate) {
    return this.#peers.find(predicate);
  }
  filter(predicate) {
    return this.#peers.filter(predicate);
  }
  map(mapper) {
    return this.#peers.map(mapper);
  }
  [Symbol.iterator]() {
    return this.#peers[Symbol.iterator]();
  }
  static [import_general.Environmental.create](env) {
    const instance = new PeerSet({
      lifetime: env,
      sessions: env.get(import_SessionManager.SessionManager),
      names: env.get(import_MdnsService.MdnsService).names,
      networks: env.get(import_NetworkProfile.NetworkProfiles)
    });
    env.set(PeerSet, instance);
    return instance;
  }
  get peers() {
    return this.#peers;
  }
  get networks() {
    return this.#networks;
  }
  get timing() {
    return this.#peerContext.timing;
  }
  set timing(timing) {
    this.#peerContext.timing = (0, import_PeerTimingParameters.PeerTimingParameters)(timing);
  }
  set handleError(handler) {
    this.#peerContext.handleError = handler;
  }
  /**
   * Retrieve a peer by address.
   */
  get(peer) {
    if ("address" in peer) {
      return this.#peers.get("address", (0, import_PeerAddress.PeerAddress)(peer.address));
    }
    return this.#peers.get("address", (0, import_PeerAddress.PeerAddress)(peer));
  }
  /**
   * Terminate any active peer networking operations.
   *
   * Also sets {@link exchanges} to undefined to prevent future connections.
   */
  async disconnect() {
    var _stack = [];
    try {
      const _disconnecting = __using(_stack, this.#lifetime.join("disconnecting"));
      this.#exchanges = void 0;
      await import_general.MatterAggregateError.allSettled(
        this.#peers.map((peer) => peer.disconnect()),
        "Error disconnecting peers"
      );
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async close() {
    var _stack = [];
    try {
      const _closing = __using(_stack, this.#lifetime.closing());
      for (const peer of this.#peers) {
        await peer.close();
      }
      this.#observers.close();
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async #openSocket(address, abort) {
    const isIpv6Address = (0, import_general.isIPv6)(address.ip);
    const operationalInterface = this.#peerContext.exchanges.interfaceFor(
      import_general.ChannelType.UDP,
      isIpv6Address ? "::" : "0.0.0.0"
    );
    if (operationalInterface === void 0) {
      throw new import_general.NetworkUnreachableError(`No interface available for IP address ${address.ip}`);
    }
    return await import_general.Abort.race(abort, operationalInterface.openChannel(address));
  }
  addKnownPeer(descriptor) {
    let peer = this.get(descriptor.address);
    if (peer === void 0) {
      peer = new import_Peer.Peer(descriptor, this.#peerContext);
      this.#peers.add(peer);
    }
    if (descriptor.operationalAddress !== void 0) {
      peer.descriptor.operationalAddress = descriptor.operationalAddress;
    }
    if (descriptor.discoveryData !== void 0) {
      peer.descriptor.discoveryData = {
        ...peer.descriptor.discoveryData,
        ...descriptor.discoveryData
      };
    }
    if (descriptor.caseAuthenticatedTags !== void 0) {
      peer.descriptor.caseAuthenticatedTags = descriptor.caseAuthenticatedTags;
    }
    if (descriptor.sessionParameters !== void 0) {
      peer.descriptor.sessionParameters = {
        ...peer.descriptor.sessionParameters,
        ...descriptor.sessionParameters
      };
    }
    return peer;
  }
}
function operationalAddressOf(session) {
  if (session.isClosed || !(0, import_general.isIpNetworkChannel)(session.channel)) {
    return;
  }
  return session.channel.networkAddress;
}
