// homebridge/dist/matter/behaviors/BehaviorRegistry.js
import { Logger } from '../../logger.js';
const log = Logger.withPrefix('Matter/Behaviors');
export class BehaviorRegistry {
    handlers = new Map();
    partEndpoints = new Map();
    accessoriesMap;
    server;
    constructor(accessoriesMap, server) {
        this.accessoriesMap = accessoriesMap;
        this.server = server;
    }
    setServer(server) {
        this.server = server;
    }
    registerHandler(endpointId, clusterName, commandName, handler) {
        if (!this.handlers.has(endpointId)) {
            this.handlers.set(endpointId, new Map());
        }
        const endpointHandlers = this.handlers.get(endpointId);
        if (!endpointHandlers.has(clusterName)) {
            endpointHandlers.set(clusterName, new Map());
        }
        const clusterHandlers = endpointHandlers.get(clusterName);
        clusterHandlers.set(commandName, handler);
        log.debug(`Registered handler: ${endpointId}.${clusterName}.${commandName}`);
    }
    getHandler(endpointId, clusterName, commandName) {
        return this.handlers.get(endpointId)?.get(clusterName)?.get(commandName);
    }
    async executeHandler(endpointId, clusterName, commandName, args, context) {
        const handler = this.getHandler(endpointId, clusterName, commandName);
        if (!handler) {
            throw new Error(`No handler registered for ${endpointId}.${clusterName}.${commandName}`);
        }
        try {
            await handler(args, context);
            return true;
        }
        catch (error) {
            log.error(`Handler error for ${endpointId}.${clusterName}.${commandName}:`, error);
            throw error;
        }
    }
    registerPartEndpoint(endpointId, parentUuid, partId) {
        this.partEndpoints.set(endpointId, { parentUuid, partId });
        log.debug(`Registered part endpoint: ${endpointId} -> ${parentUuid}.${partId}`);
    }
    getPartEndpointInfo(endpointId) {
        return this.partEndpoints.get(endpointId);
    }
    syncStateToCache(endpointId, clusterName, attributes) {
        const partInfo = this.partEndpoints.get(endpointId);
        if (partInfo) {
            const accessory = this.accessoriesMap.get(partInfo.parentUuid);
            if (!accessory?._parts) {
                return;
            }
            const part = accessory._parts.find(p => p.id === partInfo.partId);
            if (!part?.clusters) {
                return;
            }
            if (!part.clusters[clusterName]) {
                part.clusters[clusterName] = {};
            }
            part.clusters[clusterName] = {
                ...part.clusters[clusterName],
                ...attributes,
            };
            log.debug(`Synced ${clusterName} state to cache for part ${partInfo.partId}:`, attributes);
            if (this.server) {
                this.server.notifyStateChange(partInfo.parentUuid, clusterName, attributes, partInfo.partId);
            }
        }
        else {
            const accessory = this.accessoriesMap.get(endpointId);
            if (!accessory?.clusters) {
                return;
            }
            if (!accessory.clusters[clusterName]) {
                accessory.clusters[clusterName] = {};
            }
            accessory.clusters[clusterName] = {
                ...accessory.clusters[clusterName],
                ...attributes,
            };
            log.debug(`Synced ${clusterName} state to cache for ${endpointId}:`, attributes);
            if (this.server) {
                this.server.notifyStateChange(endpointId, clusterName, attributes);
            }
        }
    }
    clear() {
        this.handlers.clear();
        this.partEndpoints.clear();
    }
    getStats() {
        let handlerCount = 0;
        for (const endpointHandlers of this.handlers.values()) {
            for (const clusterHandlers of endpointHandlers.values()) {
                handlerCount += clusterHandlers.size;
            }
        }
        return {
            handlerCount,
            partCount: this.partEndpoints.size,
        };
    }
}
