// homebridge/dist/matter/behaviors/ColorControlBehavior.js
import { ColorControlServer } from '@matter/main/behaviors/color-control';
import { Status, StatusResponseError } from '@matter/main/types';
import { MatterStatus } from '../errors.js';
import { getRegistryManager } from './EndpointContext.js';
export class HomebridgeColorControlServer extends ColorControlServer {
    getRegistry() {
        return getRegistryManager(this.endpoint).getRegistry(this.endpoint.id);
    }
    async moveToColorTemperatureLogic(colorTemperatureMireds, transitionTime) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'colorControl', 'moveToColorTemperatureLogic', { colorTemperatureMireds, transitionTime });
            await super.moveToColorTemperatureLogic(colorTemperatureMireds, transitionTime);
            const currentState = this.state;
            if (currentState.colorTemperatureMireds !== undefined) {
                registry.syncStateToCache(endpointId, 'colorControl', {
                    colorTemperatureMireds: currentState.colorTemperatureMireds,
                });
            }
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set color temperature: ${message}`, Status.Failure);
        }
    }
    async moveToHueAndSaturationLogic(hue, saturation, transitionTime) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'colorControl', 'moveToHueAndSaturationLogic', { hue, saturation, transitionTime });
            await super.moveToHueAndSaturationLogic(hue, saturation, transitionTime);
            const currentState = this.state;
            const stateUpdate = {};
            if (currentState.currentHue !== undefined) {
                stateUpdate.currentHue = currentState.currentHue;
            }
            if (currentState.currentSaturation !== undefined) {
                stateUpdate.currentSaturation = currentState.currentSaturation;
            }
            registry.syncStateToCache(endpointId, 'colorControl', stateUpdate);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set hue and saturation: ${message}`, Status.Failure);
        }
    }
    async moveToColorLogic(targetX, targetY, transitionTime) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'colorControl', 'moveToColorLogic', { targetX, targetY, transitionTime });
            await super.moveToColorLogic(targetX, targetY, transitionTime);
            const currentState = this.state;
            const stateUpdate = {};
            if (currentState.currentX !== undefined) {
                stateUpdate.currentX = currentState.currentX;
            }
            if (currentState.currentY !== undefined) {
                stateUpdate.currentY = currentState.currentY;
            }
            registry.syncStateToCache(endpointId, 'colorControl', stateUpdate);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set XY color: ${message}`, Status.Failure);
        }
    }
    async moveToHueLogic(targetHue, direction, transitionTime, isEnhancedHue = false) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'colorControl', 'moveToHueLogic', { targetHue, direction, transitionTime, isEnhancedHue });
            await super.moveToHueLogic(targetHue, direction, transitionTime, isEnhancedHue);
            const currentState = this.state;
            const stateUpdate = {};
            if (isEnhancedHue && currentState.enhancedCurrentHue !== undefined) {
                stateUpdate.enhancedCurrentHue = currentState.enhancedCurrentHue;
            }
            else if (currentState.currentHue !== undefined) {
                stateUpdate.currentHue = currentState.currentHue;
            }
            registry.syncStateToCache(endpointId, 'colorControl', stateUpdate);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set hue: ${message}`, Status.Failure);
        }
    }
    async moveToSaturationLogic(targetSaturation, transitionTime) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'colorControl', 'moveToSaturationLogic', { targetSaturation, transitionTime });
            await super.moveToSaturationLogic(targetSaturation, transitionTime);
            const currentState = this.state;
            if (currentState.currentSaturation !== undefined) {
                registry.syncStateToCache(endpointId, 'colorControl', {
                    currentSaturation: currentState.currentSaturation,
                });
            }
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set saturation: ${message}`, Status.Failure);
        }
    }
    async stopAllColorMovement() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'colorControl', 'stopAllColorMovement');
            await super.stopAllColorMovement();
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to stop color movement: ${message}`, Status.Failure);
        }
    }
}
