// homebridge/dist/matter/behaviors/LevelControlBehavior.js
import { LevelControlServer } from '@matter/main/behaviors/level-control';
import { Status, StatusResponseError } from '@matter/main/types';
import { MatterStatus } from '../errors.js';
import { getRegistryManager } from './EndpointContext.js';
export class HomebridgeLevelControlServer extends LevelControlServer {
    getRegistry() {
        return getRegistryManager(this.endpoint).getRegistry(this.endpoint.id);
    }
    async moveToLevel(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'levelControl', 'moveToLevel', request);
            await super.moveToLevel(request);
            registry.syncStateToCache(endpointId, 'levelControl', {
                currentLevel: request.level,
            });
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set level: ${message}`, Status.Failure);
        }
    }
    async move(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'levelControl', 'move', request);
            await super.move(request);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to move level: ${message}`, Status.Failure);
        }
    }
    async step(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'levelControl', 'step', request);
            await super.step(request);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to step level: ${message}`, Status.Failure);
        }
    }
    async stop(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'levelControl', 'stop', request);
            await super.stop(request);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to stop level change: ${message}`, Status.Failure);
        }
    }
    async moveToLevelWithOnOff(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'levelControl', 'moveToLevelWithOnOff', request);
            await super.moveToLevelWithOnOff(request);
            registry.syncStateToCache(endpointId, 'levelControl', {
                currentLevel: request.level,
            });
            const targetOnOff = (request.level ?? 0) > 0;
            await this.endpoint.set({
                onOff: {
                    onOff: targetOnOff,
                },
            });
            registry.syncStateToCache(endpointId, 'onOff', {
                onOff: targetOnOff,
            });
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set level with on/off: ${message}`, Status.Failure);
        }
    }
}
