// homebridge/dist/matter/behaviors/RegistryManager.js
import { Logger } from '../../logger.js';
const log = Logger.withPrefix('Matter/RegistryManager');
export class RegistryManager {
    endpointToRegistry = new Map();
    registerEndpoint(endpointId, registry) {
        this.endpointToRegistry.set(endpointId, registry);
        log.debug(`Registered registry for endpoint: ${endpointId}`);
    }
    getRegistry(endpointId) {
        const registry = this.endpointToRegistry.get(endpointId);
        if (!registry) {
            throw new Error(`No registry found for endpoint ${endpointId}. Available endpoints: ${[...this.endpointToRegistry.keys()].join(', ')}`);
        }
        return registry;
    }
    unregisterEndpoint(endpointId) {
        this.endpointToRegistry.delete(endpointId);
        log.debug(`Unregistered endpoint: ${endpointId}`);
    }
    clear() {
        this.endpointToRegistry.clear();
    }
    getStats() {
        return {
            endpointCount: this.endpointToRegistry.size,
            endpoints: [...this.endpointToRegistry.keys()],
        };
    }
}
