// homebridge/dist/matter/behaviors/RvcOperationalStateBehavior.js
import { RvcOperationalStateServer } from '@matter/main/behaviors/rvc-operational-state';
import { RvcOperationalState } from '@matter/main/clusters/rvc-operational-state';
import { Status, StatusResponseError } from '@matter/main/types';
import { MatterStatus } from '../errors.js';
import { getRegistryManager } from './EndpointContext.js';
export class HomebridgeRvcOperationalStateServer extends RvcOperationalStateServer {
    getRegistry() {
        return getRegistryManager(this.endpoint).getRegistry(this.endpoint.id);
    }
    syncOperationalStateToCache() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        const currentState = this.state;
        if (currentState.operationalState !== undefined) {
            registry.syncStateToCache(endpointId, 'rvcOperationalState', {
                operationalState: currentState.operationalState,
            });
        }
    }
    async pause() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'rvcOperationalState', 'pause');
            this.syncOperationalStateToCache();
            return {
                commandResponseState: {
                    errorStateId: RvcOperationalState.ErrorState.NoError,
                },
            };
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to pause: ${message}`, Status.Failure);
        }
    }
    async resume() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'rvcOperationalState', 'resume');
            this.syncOperationalStateToCache();
            return {
                commandResponseState: {
                    errorStateId: RvcOperationalState.ErrorState.NoError,
                },
            };
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to resume: ${message}`, Status.Failure);
        }
    }
    async goHome() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'rvcOperationalState', 'goHome');
            this.syncOperationalStateToCache();
            return {
                commandResponseState: {
                    errorStateId: RvcOperationalState.ErrorState.NoError,
                },
            };
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to go home: ${message}`, Status.Failure);
        }
    }
}
