// homebridge/dist/matter/behaviors/ValveConfigurationAndControlBehavior.js
import { ValveConfigurationAndControlServer } from '@matter/main/behaviors/valve-configuration-and-control';
import { ValveConfigurationAndControl } from '@matter/main/clusters/valve-configuration-and-control';
import { Status, StatusResponseError } from '@matter/main/types';
import { MatterStatus } from '../errors.js';
import { getRegistryManager } from './EndpointContext.js';
export class HomebridgeValveConfigurationAndControlServer extends ValveConfigurationAndControlServer {
    getRegistry() {
        return getRegistryManager(this.endpoint).getRegistry(this.endpoint.id);
    }
    async open(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'valveConfigurationAndControl', 'open', request);
            await super.open(request);
            registry.syncStateToCache(endpointId, 'valveConfigurationAndControl', {
                currentState: ValveConfigurationAndControl.ValveState.Open,
                targetState: ValveConfigurationAndControl.ValveState.Open,
            });
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to open valve: ${message}`, Status.Failure);
        }
    }
    async close() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'valveConfigurationAndControl', 'close');
            await super.close();
            registry.syncStateToCache(endpointId, 'valveConfigurationAndControl', {
                currentState: ValveConfigurationAndControl.ValveState.Closed,
                targetState: ValveConfigurationAndControl.ValveState.Closed,
            });
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to close valve: ${message}`, Status.Failure);
        }
    }
}
