// homebridge/dist/matter/behaviors/WindowCoveringBehavior.js
import { WindowCoveringBaseServer } from '@matter/main/behaviors/window-covering';
import { Status, StatusResponseError } from '@matter/main/types';
import { MatterStatus } from '../errors.js';
import { getRegistryManager } from './EndpointContext.js';
const WindowCoveringStateProps = {
    targetPositionLiftPercent100ths: 'targetPositionLiftPercent100ths',
    currentPositionLiftPercent100ths: 'currentPositionLiftPercent100ths',
    targetPositionTiltPercent100ths: 'targetPositionTiltPercent100ths',
    currentPositionTiltPercent100ths: 'currentPositionTiltPercent100ths',
};
export class HomebridgeWindowCoveringServer extends WindowCoveringBaseServer {
    getRegistry() {
        return getRegistryManager(this.endpoint).getRegistry(this.endpoint.id);
    }
    syncPositionStateToCache(endpointId, targetProperty, currentProperty) {
        const registry = this.getRegistry();
        const currentState = this.state;
        const stateUpdate = {};
        if (currentState[targetProperty] !== undefined) {
            stateUpdate[targetProperty] = currentState[targetProperty];
        }
        if (currentState[currentProperty] !== undefined) {
            stateUpdate[currentProperty] = currentState[currentProperty];
        }
        registry.syncStateToCache(endpointId, 'windowCovering', stateUpdate);
    }
    async upOrOpen() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'windowCovering', 'upOrOpen');
            await super.upOrOpen();
            this.syncPositionStateToCache(endpointId, WindowCoveringStateProps.targetPositionLiftPercent100ths, WindowCoveringStateProps.currentPositionLiftPercent100ths);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to open window covering: ${message}`, Status.Failure);
        }
    }
    async downOrClose() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'windowCovering', 'downOrClose');
            await super.downOrClose();
            this.syncPositionStateToCache(endpointId, WindowCoveringStateProps.targetPositionLiftPercent100ths, WindowCoveringStateProps.currentPositionLiftPercent100ths);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to close window covering: ${message}`, Status.Failure);
        }
    }
    async stopMotion() {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'windowCovering', 'stopMotion');
            await super.stopMotion();
            this.syncPositionStateToCache(endpointId, WindowCoveringStateProps.targetPositionLiftPercent100ths, WindowCoveringStateProps.currentPositionLiftPercent100ths);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to stop window covering: ${message}`, Status.Failure);
        }
    }
    async goToLiftPercentage(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'windowCovering', 'goToLiftPercentage', request);
            await super.goToLiftPercentage(request);
            this.syncPositionStateToCache(endpointId, WindowCoveringStateProps.targetPositionLiftPercent100ths, WindowCoveringStateProps.currentPositionLiftPercent100ths);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set window covering position: ${message}`, Status.Failure);
        }
    }
    async goToTiltPercentage(request) {
        const endpointId = this.endpoint.id;
        const registry = this.getRegistry();
        try {
            await registry.executeHandler(endpointId, 'windowCovering', 'goToTiltPercentage', request);
            await super.goToTiltPercentage(request);
            this.syncPositionStateToCache(endpointId, WindowCoveringStateProps.targetPositionTiltPercent100ths, WindowCoveringStateProps.currentPositionTiltPercent100ths);
        }
        catch (error) {
            if (MatterStatus.isMatterProtocolError(error)) {
                throw error;
            }
            const message = error instanceof Error ? error.message : String(error);
            throw new StatusResponseError(`Failed to set window covering tilt: ${message}`, Status.Failure);
        }
    }
}
