// @matter/general/dist/esm/storage/MockStorageService.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "#MatterError.js";
import { MemoryBlobStorageDriver } from "./MemoryBlobStorageDriver.js";
import { MemoryStorageDriver } from "./MemoryStorageDriver.js";
import { StorageManager } from "./StorageManager.js";
import { StorageService } from "./StorageService.js";
class MockStorageService extends StorageService {
  #opener;
  #stores = /* @__PURE__ */ new Map();
  constructor(environment, opener) {
    super(environment);
    this.#opener = opener;
    this.registerBlobDriver({
      id: MemoryBlobStorageDriver.id,
      create: () => new MemoryBlobStorageDriver()
    });
    this.defaultBlobDriver = MemoryBlobStorageDriver.id;
  }
  /**
   * Get or create the in-memory store for a namespace.
   *
   * Only available when using the default (no custom opener) mode.  Useful for tests that need to pre-populate or
   * inspect storage directly.
   */
  store(namespace) {
    if (this.#opener !== void 0) {
      throw new InternalError("store() is only available without a custom opener");
    }
    let store = this.#stores.get(namespace);
    if (!store) {
      store = MemoryStorageDriver.create();
      this.#stores.set(namespace, store);
    }
    return store;
  }
  get isConfigured() {
    return true;
  }
  async open(namespace) {
    const name = typeof namespace === "string" ? namespace : namespace.namespace;
    let storage;
    if (this.#opener !== void 0) {
      storage = await this.#opener(name);
    } else {
      storage = this.store(name);
    }
    const manager = new StorageManager(storage);
    await manager.initialize();
    return manager;
  }
}
export {
  MockStorageService
};
