// @matter/general/dist/esm/storage/StorageContext.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MaybePromise } from "../util/Promises.js";
import { StorageError } from "./StorageDriver.js";
class StorageContext {
  constructor(storage, thisContexts) {
    this.storage = storage;
    this.thisContexts = thisContexts;
  }
  storage;
  thisContexts;
  get(key, defaultValue) {
    const value = this.storage.get(this.thisContexts, key);
    if (value !== void 0) {
      if (MaybePromise.is(value)) {
        return value.then((v) => {
          if (v !== void 0) return v;
          if (defaultValue === void 0) {
            throw new StorageError(
              `No value found for key ${key} in context ${this.thisContexts} and no default value specified`
            );
          }
          return defaultValue;
        });
      }
      return value;
    }
    if (defaultValue === void 0) {
      throw new StorageError(
        `No value found for key ${key} in context ${this.thisContexts} and no default value specified`
      );
    }
    return defaultValue;
  }
  has(key) {
    return this.storage.has(this.thisContexts, key);
  }
  set(keyOrValues, value) {
    if (typeof keyOrValues === "string") {
      if (keyOrValues.includes(".")) {
        throw new StorageError("Storage keys must not contain dots");
      }
      return this.storage.set(this.thisContexts, keyOrValues, value);
    } else if (keyOrValues === null || typeof keyOrValues !== "object") {
      throw new StorageError("Storage keys must be string or object ok string-value pairs");
    }
    for (const key of Object.keys(keyOrValues)) {
      if (!key) {
        throw new StorageError("Storage keys can not be empty");
      }
      if (key.includes(".")) {
        throw new StorageError("Storage keys must not contain dots");
      }
    }
    return this.storage.set(this.thisContexts, keyOrValues);
  }
  delete(key) {
    return this.storage.delete(this.thisContexts, key);
  }
  begin() {
    return this.storage.begin();
  }
  createContext(context) {
    if (context.length === 0) throw new StorageError("Context must not be an empty string");
    if (context.includes(".")) throw new StorageError("Context must not contain dots");
    return new StorageContext(this.storage, [...this.thisContexts, context]);
  }
  keys() {
    return this.storage.keys(this.thisContexts);
  }
  values() {
    return this.storage.values(this.thisContexts);
  }
  contexts() {
    return this.storage.contexts(this.thisContexts);
  }
  /** @deprecated Use {@link clearAll} instead. */
  clear() {
    throw new StorageError("clear() is deprecated; use clearAll() instead");
  }
  /** Clears all keys in this context and all created sub-contexts. */
  clearAll() {
    return this.storage.clearAll(this.thisContexts);
  }
}
export {
  StorageContext
};
