// @matter/general/dist/esm/storage/StorageMigration.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "#log/Logger.js";
import { MatterAggregateError } from "#MatterError.js";
import { StorageError } from "./StorageDriver.js";
const logger = new Logger("StorageMigrator");
var StorageMigration;
((StorageMigration2) => {
  async function migrate(source, target) {
    const strategy = resolveStrategy(source.type, target.type);
    const result = {
      success: true,
      migratedCount: 0,
      skippedCount: 0,
      otherTypeKeysSkipped: 0,
      skippedItems: []
    };
    await migrateContext({ source, target, strategy, contexts: [], result });
    if (result.otherTypeKeysSkipped > 0) {
      logger.info(`[migrate] Skipped ${result.otherTypeKeysSkipped} keys not matching target type`);
    }
    result.success = result.skippedCount === 0;
    return result;
  }
  StorageMigration2.migrate = migrate;
  function resultToLog(result) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    let logContent = `
    # Migration Log
    - Timestamp: ${timestamp}
    - Success: ${result.success}
    - Migrated: ${result.migratedCount}
    - Skipped: ${result.skippedCount}
    `.replace(/^\s+/gm, "");
    if (result.skippedCount > 0 && result.skippedItems.length > 0) {
      logContent += `
## Skipped Items
`;
      for (const item of result.skippedItems) {
        logContent += `- ${buildContextKeyLog(item.contexts, item.key)}: ${item.error}
`;
      }
    }
    return logContent;
  }
  StorageMigration2.resultToLog = resultToLog;
})(StorageMigration || (StorageMigration = {}));
function resolveStrategy(sourceType, targetType) {
  if (sourceType === targetType) {
    return sourceType;
  }
  if (sourceType === "kv" && targetType === "blob") {
    return "kv-to-blob";
  }
  throw new StorageError(`Cannot migrate from "${sourceType}" storage to "${targetType}" storage`);
}
async function migrateContext(args) {
  const { source, target, strategy, contexts, result } = args;
  const keys = await source.keys(contexts);
  for (const key of keys) {
    try {
      switch (strategy) {
        case "kv":
          await migrateKvKey(source, target, contexts, key, result);
          break;
        case "blob":
          await migrateBlobKey(
            source,
            target,
            contexts,
            key,
            result
          );
          break;
        case "kv-to-blob":
          await extractBlobFromKv(
            source,
            target,
            contexts,
            key,
            result
          );
          break;
      }
    } catch (err) {
      result.skippedCount += 1;
      result.skippedItems.push({
        contexts,
        key,
        error: err instanceof Error ? err.message : String(err)
      });
      logger.warn(`[migrate] Skipped '${buildContextKeyLog(contexts, key)}' entity: ${err}`);
    }
  }
  const subContexts = await source.contexts(contexts);
  const promises = [];
  subContexts.forEach((subCon) => {
    promises.push(
      migrateContext({
        ...args,
        contexts: [...contexts, subCon]
      })
    );
  });
  await MatterAggregateError.allSettled(promises, "[migrate] Error migrating sub-contexts!");
}
async function migrateKvKey(source, target, contexts, key, result) {
  const value = await source.get(contexts, key);
  if (value === void 0) {
    result.otherTypeKeysSkipped++;
    return;
  }
  await target.set(contexts, key, value);
  result.migratedCount++;
}
async function migrateBlobKey(source, target, contexts, key, result) {
  const blob = await source.openBlob(contexts, key);
  if (blob.size === 0) {
    return;
  }
  await target.writeBlobFromStream(contexts, key, blob.stream());
  result.migratedCount++;
}
async function extractBlobFromKv(source, target, contexts, key, result) {
  const value = await source.get(contexts, key);
  if (value !== void 0) {
    result.otherTypeKeysSkipped++;
    return;
  }
  if (!isLegacyBlobSource(source)) {
    result.otherTypeKeysSkipped++;
    return;
  }
  await migrateBlobKey(source, target, contexts, key, result);
}
function isLegacyBlobSource(driver) {
  return "openBlob" in driver && typeof driver.openBlob === "function";
}
function buildContextKeyLog(contexts, key) {
  return `${contexts.join(".")}$${key}`;
}
export {
  StorageMigration
};
