// @matter/protocol/dist/cjs/action/client/subscription/ClientSubscriptionHandler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var ClientSubscriptionHandler_exports = {};
__export(ClientSubscriptionHandler_exports, {
  ClientSubscriptionHandler: () => ClientSubscriptionHandler
});
module.exports = __toCommonJS(ClientSubscriptionHandler_exports);
var import_InteractionMessenger = require("#interaction/InteractionMessenger.js");
var import_Subscription = require("#interaction/Subscription.js");
var import_SecureSession = require("#session/SecureSession.js");
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_InputChunk = require("../InputChunk.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("ClientSubscriptionHandler");
class ClientSubscriptionHandler {
  id = import_types.INTERACTION_PROTOCOL_ID;
  requiresSecureSession = true;
  #subscriptions;
  constructor(subscriptions) {
    this.#subscriptions = subscriptions;
  }
  async onNewExchange(exchange) {
    var _stack = [];
    try {
      if (this.#subscriptions.isBlocked) {
        const messenger = new import_InteractionMessenger.IncomingInteractionClientMessenger(exchange);
        try {
          await sendInvalid(messenger);
        } finally {
          await messenger.close();
        }
        return;
      }
      const _reading = __using(_stack, this.#subscriptions.beginReading());
      try {
        await this.#handleExchange(exchange);
      } catch (error) {
        if (this.#subscriptions.readingAbortSignal.aborted && (0, import_general.causedBy)(error, import_general.AbortedError)) {
          logger.debug(exchange.via, "Data report processing aborted during shutdown");
          return;
        }
        throw error;
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      __callDispose(_stack, _error, _hasError);
    }
  }
  async #handleExchange(exchange) {
    const messenger = new import_InteractionMessenger.IncomingInteractionClientMessenger(exchange);
    const abort = this.#subscriptions.readingAbortSignal;
    const reports = messenger.readDataReports({ abort });
    const initialIteration = await reports.next();
    if (initialIteration.done) {
      throw new import_general.InternalError("Exchange initiated with no initial message");
    }
    const initialReport = initialIteration.value;
    const { subscriptionId } = initialReport;
    if (subscriptionId === void 0) {
      logger.debug(exchange.via, "Ignoring unsolicited data report with no subscription ID");
      try {
        await sendInvalid(messenger);
      } finally {
        await messenger.close();
      }
      return;
    }
    const { session } = exchange.channel;
    import_SecureSession.SecureSession.assert(session);
    const subscription = this.#subscriptions.getPeer(session.peerAddress, subscriptionId);
    if (subscription === void 0) {
      logger.info(
        exchange.via,
        "Ignoring data report for unknown subscription ID",
        import_general.Diagnostic.strong(import_Subscription.Subscription.idStrOf(subscriptionId))
      );
      try {
        await sendInvalid(messenger, subscriptionId);
      } finally {
        await messenger.close();
      }
      return;
    }
    try {
      subscription.isReading = true;
      if (!initialReport.attributeReports?.length && !initialReport.eventReports?.length) {
        const ending = await reports.next();
        if (!ending.done) {
          logger.warn(
            exchange.via,
            "Unexpected data reports after empty report",
            import_general.Diagnostic.strong(import_Subscription.Subscription.idStrOf(subscriptionId))
          );
          for await (const _chunk of reports) ;
        }
      } else {
        if (subscription.request.updated) {
          await subscription.request.updated(processReports(initialReport, reports, messenger));
        } else {
          for await (const _chunk of reports) ;
        }
      }
    } finally {
      subscription.isReading = false;
      subscription.timeoutAt = void 0;
      this.#subscriptions.resetTimer();
      await messenger.close();
    }
  }
  async close() {
  }
}
async function sendInvalid(messenger, subscriptionId) {
  await messenger.sendStatus(import_types.Status.InvalidSubscription, {
    multipleMessageInteraction: true,
    logContext: import_Subscription.Subscription.diagnosticOf(subscriptionId)
  });
}
async function* processReports(initialReport, otherReports, messenger) {
  const leftOverData = new Array();
  yield (0, import_InputChunk.InputChunk)(initialReport, leftOverData);
  const { subscriptionId } = initialReport;
  for await (const report of otherReports) {
    const { subscriptionId: reportSubscriptionId } = report;
    if (reportSubscriptionId === void 0) {
      logger.debug(messenger.exchange.via, "Ignoring data report with missing subscription id");
      await sendInvalid(messenger, reportSubscriptionId);
      return;
    }
    if (reportSubscriptionId !== subscriptionId) {
      logger.debug(
        messenger.exchange.via,
        "Ignoring data report for incorrect subscription id",
        import_general.Diagnostic.strong(import_Subscription.Subscription.idStrOf(reportSubscriptionId)),
        "expected",
        import_Subscription.Subscription.idStrOf(subscriptionId)
      );
      await sendInvalid(messenger, reportSubscriptionId);
      return;
    }
    yield (0, import_InputChunk.InputChunk)(report, leftOverData);
  }
}
