// @matter/general/dist/esm/time/TimeUnit.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { NodeJsStyleInspectable } from "#log/NodeJsStyleInspectable.js";
function TimeUnit(kind, abbrev, one, props = {}) {
  const unit = {
    [kind]: (scale) => {
      if (scale === void 0) {
        return void 0;
      }
      if (typeof scale === "bigint") {
        scale = Number(scale);
      }
      return scale * one;
    }
  }[kind];
  Object.assign(unit, {
    ...props,
    kind,
    abbrev,
    one,
    of,
    fractionalOf,
    ceil,
    floor,
    round,
    toString: kindOf
  });
  NodeJsStyleInspectable(unit, kindOf);
  return unit;
}
function of(duration) {
  return Math.floor(duration / this.one);
}
function fractionalOf(duration) {
  return duration / this.one;
}
function ceil(duration) {
  return Math.ceil(duration / this.one) * this.one;
}
function floor(duration) {
  return Math.floor(duration / this.one) * this.one;
}
function round(duration) {
  return Math.round(duration / this.one) * this.one;
}
function kindOf() {
  return this.kind;
}
const Microseconds = TimeUnit("microsecond", "\u03BCs", 1e-3);
const Millis = TimeUnit("millisecond", "ms", 1);
const Seconds = TimeUnit("second", "s", 1e3);
const Minutes = TimeUnit("minute", "m", 6e4);
const Hours = TimeUnit("hour", "h", 36e5);
const Days = TimeUnit("day", "d", 864e5);
const Instant = Millis(0);
const Forever = Millis(Infinity);
export {
  Days,
  Forever,
  Hours,
  Instant,
  Microseconds,
  Millis,
  Minutes,
  Seconds,
  TimeUnit
};
