// @matter/node/dist/cjs/storage/client/ClientCacheBuffer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var ClientCacheBuffer_exports = {};
__export(ClientCacheBuffer_exports, {
  ClientCacheBuffer: () => ClientCacheBuffer
});
module.exports = __toCommonJS(ClientCacheBuffer_exports);
var import_NetworkServer = require("#behavior/system/network/NetworkServer.js");
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("ClientCacheBuffer");
class ClientCacheBuffer {
  #dirty = /* @__PURE__ */ new Set();
  #abort = new import_general.Abort();
  #mutex = new import_general.Mutex(this);
  #storageDriver;
  #flushInterval;
  #timerDone;
  constructor(storageDriver, flushInterval) {
    this.#storageDriver = storageDriver;
    this.#flushInterval = flushInterval;
  }
  /**
   * If buffering is configured on the parent ServerNode, set up the buffer on the store.  The first call creates
   * the buffer and starts the timer; subsequent calls reuse it.
   */
  static configure(node, store) {
    const flushInterval = node.owner.stateOf(import_NetworkServer.NetworkServer).clientCacheFlushInterval;
    if (flushInterval === void 0) {
      return;
    }
    const parentEnv = node.owner.env;
    if (parentEnv.has(ClientCacheBuffer)) {
      store.buffer = parentEnv.get(ClientCacheBuffer);
      return;
    }
    const buffer = new ClientCacheBuffer(parentEnv.get(import_general.StorageManager).driver, flushInterval);
    parentEnv.set(ClientCacheBuffer, buffer);
    buffer.#start();
    store.buffer = buffer;
  }
  /**
   * Mark a cache as having dirty data that needs flushing.
   */
  markDirty(cache) {
    this.#dirty.add(cache);
  }
  /**
   * Remove a cache from the dirty set (e.g. on erase or reclaim).
   */
  removeDirty(cache) {
    this.#dirty.delete(cache);
  }
  /**
   * Schedule a flush.  The flush runs serialized through the mutex; the promise is tracked internally.  Safe to
   * call from sync contexts.
   */
  initiateFlush() {
    this.#mutex.run(() => this.#doFlush());
  }
  /**
   * Flush all dirty caches to storage in a single transaction.  Awaits completion.
   */
  async flush() {
    this.initiateFlush();
    await this.#mutex;
  }
  /**
   * Stop the timer and flush any remaining dirty data.
   */
  async close() {
    this.#abort();
    await this.#timerDone;
    this.initiateFlush();
    await this.#mutex.close();
  }
  async #doFlush() {
    var _stack = [];
    try {
      const snapshot = [...this.#dirty];
      this.#dirty.clear();
      if (!snapshot.length) {
        return;
      }
      const tx = __using(_stack, await this.#storageDriver.begin(), true);
      const flushedKeys = new Array();
      for (const cache of snapshot) {
        try {
          const keys = await cache.flush(tx);
          if (keys) {
            flushedKeys.push({ cache, keys });
          }
        } catch (e) {
          if (cache.erased) {
            logger.debug("Dropping erased cache from flush buffer");
          } else {
            this.#dirty.add(cache);
            logger.warn("Cache flush failed, will retry:", e);
          }
        }
      }
      if (!flushedKeys.length) {
        return;
      }
      try {
        await tx.commit();
      } catch (e) {
        for (const { cache, keys } of flushedKeys) {
          if (!cache.erased) {
            cache.restoreDirtyKeys(keys);
            this.#dirty.add(cache);
          }
        }
        throw e;
      }
      logger.debug("Flushed", flushedKeys.length, "dirty caches");
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      var _promise = __callDispose(_stack, _error, _hasError);
      _promise && await _promise;
    }
  }
  #start() {
    this.#timerDone = this.#timerLoop();
  }
  async #timerLoop() {
    while (!this.#abort.aborted) {
      await this.#abort.sleep("client-cache-flush", this.#flushInterval);
      if (this.#abort.aborted) {
        break;
      }
      try {
        await this.#mutex.produce(() => this.#doFlush());
      } catch (e) {
        logger.error("Periodic cache flush failed:", e);
      }
    }
  }
}
