// @matter/node/dist/esm/behavior/cluster/ClusterBehaviorType.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { OfflineEvent, OnlineEvent, QuietEvent } from "#behavior/Events.js";
import { camelize, GeneratedClass, ImplementationError } from "@matter/general";
import {
  ClassSemantics,
  DecodedBitmap,
  DefaultValue,
  ElementTag,
  FeatureMap,
  FeatureSet,
  Matter,
  Metatype,
  Scope
} from "@matter/model";
import { ClusterType } from "@matter/types";
import { Behavior } from "../Behavior.js";
import { DerivedState } from "../state/StateType.js";
import { ClusterBehaviorCache } from "./ClusterBehaviorCache.js";
import { introspectionInstanceOf } from "./cluster-behavior-utils.js";
function ClusterBehaviorType({
  namespace,
  base,
  features,
  schema,
  name,
  forClient,
  commandFactory
}) {
  if (schema === void 0) {
    if (base.schema?.tag === ElementTag.Cluster) {
      schema = base.schema;
    }
    if (!schema && namespace) {
      schema = namespace.schema;
    }
    if (!schema && namespace) {
      const nsId = namespace.id;
      if (nsId !== void 0) {
        schema = schemaForId(nsId);
      }
    }
  }
  if (!schema) {
    throw new ImplementationError("ClusterBehaviorType: no schema available");
  }
  if (features !== void 0) {
    schema = applyFeatureSelection(schema, features);
  } else {
    schema = syncFeatures(schema, namespace);
  }
  if (!namespace) {
    namespace = ClusterType(schema);
  }
  const useCache = name === void 0;
  if (useCache) {
    const cached = ClusterBehaviorCache.get(base, schema, forClient);
    if (cached) {
      return cached;
    }
    if (base.name.startsWith(schema.name)) {
      name = base.name;
    } else {
      name = `${schema.name}Behavior`;
    }
  }
  const context = {
    scope: Scope(schema),
    namespace,
    base,
    newProps: {},
    forClient,
    commandFactory
  };
  const supportedFeatures = computeFeatureFlags(schema);
  const type = GeneratedClass({
    name,
    base,
    staticProperties: {
      State: createDerivedState(context),
      Events: createDerivedEvents(context)
    },
    staticDescriptors: {
      id: {
        value: schema.propertyName,
        enumerable: true
      },
      cluster: {
        value: namespace,
        enumerable: true
      },
      features: {
        value: supportedFeatures,
        enumerable: true
      }
    },
    instanceDescriptors: createDefaultCommandDescriptors(context)
  });
  ClassSemantics.of(type).mutableModel = schema;
  if (type.schema !== schema) {
    Object.defineProperty(type, "schema", { value: schema });
  }
  schema.finalize();
  if (useCache) {
    ClusterBehaviorCache.set(base, schema, type);
  }
  return type;
}
const KNOWN_DEFAULTS = /* @__PURE__ */ Symbol("knownDefaults");
function createDerivedState({ scope, base, newProps }) {
  const BaseState = base["State"];
  if (BaseState === void 0) {
    throw new ImplementationError(`No state class defined for behavior class ${base.name}`);
  }
  const schema = scope.owner;
  const oldDefaults = new BaseState();
  let knownDefaults = BaseState[KNOWN_DEFAULTS];
  const props = {};
  for (const member of scope.membersOf(scope.owner, { conformance: "deconflicted" })) {
    const name = member.propertyName;
    if (props[name]) {
      props[name].push(member);
    } else {
      props[name] = [member];
    }
  }
  const defaults = {};
  for (const name in props) {
    const attrs = props[name];
    let propSchema;
    for (const attr of attrs) {
      if (!attr.effectiveConformance.applicabilityFor(scope)) {
        continue;
      }
      propSchema = attr;
      break;
    }
    if (!propSchema) {
      if (oldDefaults[name] !== void 0) {
        if (!knownDefaults) {
          knownDefaults = {};
        } else if (knownDefaults === BaseState[KNOWN_DEFAULTS]) {
          knownDefaults = { ...knownDefaults };
        }
        knownDefaults[name] = oldDefaults[name];
        defaults[name] = void 0;
      }
      continue;
    }
    newProps[name] = propSchema;
    if (propSchema.id === FeatureMap.id) {
      defaults[name] = computeFeatureFlags(schema);
      continue;
    }
    defaults[name] = selectDefaultValue(
      scope,
      oldDefaults[name] === void 0 ? knownDefaults?.[name] : oldDefaults[name],
      propSchema
    );
  }
  const StateType = DerivedState({
    name: `${schema.name}$State`,
    base: base.State,
    values: defaults
  });
  if (knownDefaults) {
    StateType[KNOWN_DEFAULTS] = knownDefaults;
  }
  return StateType;
}
function createDerivedEvents({ scope, base, newProps, forClient }) {
  const instanceDescriptors = {};
  const baseInstance = new base.Events();
  const eventNames = /* @__PURE__ */ new Set();
  const quieterImplementation = forClient ? OnlineEvent : QuietEvent;
  const applicableClusterEvents = /* @__PURE__ */ new Set();
  for (const event of scope.membersOf(scope.owner, {
    conformance: "conformant",
    tags: [ElementTag.Event]
  })) {
    const name = event.propertyName;
    applicableClusterEvents.add(name);
    if (baseInstance[name] !== void 0) {
      continue;
    }
    if (!forClient && !scope.hasOperationalSupport(event)) {
      continue;
    }
    eventNames.add(name);
    instanceDescriptors[name] = createEventDescriptor(
      name,
      event,
      event.quality.quieter ? quieterImplementation : OnlineEvent
    );
  }
  for (const attrName in newProps) {
    const changing = `${attrName}$Changing`;
    const prop = newProps[attrName];
    if (baseInstance[changing] === void 0) {
      eventNames.add(changing);
      instanceDescriptors[changing] = createEventDescriptor(changing, prop, OfflineEvent);
    }
    const changed = `${attrName}$Changed`;
    if (baseInstance[changed] === void 0) {
      eventNames.add(changed);
      instanceDescriptors[changed] = createEventDescriptor(
        changed,
        prop,
        prop.quality.quieter ? quieterImplementation : OnlineEvent
      );
    }
  }
  return GeneratedClass({
    name: `${base.name}$Events`,
    base: base.Events,
    instanceDescriptors,
    initialize() {
      for (const name of eventNames) {
        this.addEvent(name);
      }
    }
  });
}
function schemaForId(id) {
  for (const child of Matter.children) {
    if (child.tag === ElementTag.Cluster && child.id === id) {
      return child;
    }
  }
  return void 0;
}
const configuredSchemaCache = /* @__PURE__ */ new Map();
function syncFeatures(schema, namespace) {
  let incomingFeatures;
  const nsSupportedFeatures = namespace?.supportedFeatures;
  if (nsSupportedFeatures) {
    incomingFeatures = new FeatureSet(nsSupportedFeatures);
  } else {
    return schema;
  }
  if (incomingFeatures.is(schema.supportedFeatures)) {
    return schema;
  }
  const featureKey = [...incomingFeatures].sort().join(",");
  let schemaBucket = configuredSchemaCache.get(schema);
  if (schemaBucket === void 0) {
    schemaBucket = {};
    configuredSchemaCache.set(schema, schemaBucket);
  } else {
    if (featureKey in schemaBucket) {
      return schemaBucket[featureKey];
    }
  }
  schema = schema.clone();
  schema.supportedFeatures = incomingFeatures;
  schemaBucket[featureKey] = schema;
  return schema;
}
function applyFeatureSelection(schema, features) {
  if (features === true) {
    const featureSet2 = new FeatureSet();
    for (const feature of schema.features) {
      featureSet2.add(feature.name);
    }
    return syncFeaturesFromSet(schema, featureSet2);
  }
  const featureSet = new FeatureSet();
  const model = schema;
  for (const name of features) {
    for (const feature of model.features) {
      if ((feature.title ?? feature.name) === name || feature.name === name) {
        featureSet.add(feature.name);
        break;
      }
    }
  }
  return syncFeaturesFromSet(schema, featureSet);
}
function syncFeaturesFromSet(schema, featureSet) {
  if (featureSet.is(schema.supportedFeatures)) {
    return schema;
  }
  const featureKey = [...featureSet].sort().join(",");
  let schemaBucket = configuredSchemaCache.get(schema);
  if (schemaBucket === void 0) {
    schemaBucket = {};
    configuredSchemaCache.set(schema, schemaBucket);
  } else if (featureKey in schemaBucket) {
    return schemaBucket[featureKey];
  }
  schema = schema.clone();
  schema.supportedFeatures = featureSet;
  schemaBucket[featureKey] = schema;
  return schema;
}
function computeFeatureFlags(schema) {
  const flags = {};
  for (const child of schema.featureMap.children) {
    const key = camelize(child.title ?? child.name);
    flags[key] = schema.supportedFeatures.has(child.name);
  }
  return flags;
}
const sourceFactory = /* @__PURE__ */ Symbol("source-factory");
function createDefaultCommandDescriptors({ scope, base, commandFactory }) {
  const result = {};
  const instance = introspectionInstanceOf(base);
  const names = new Set(
    scope.membersOf(scope.owner, { tags: [ElementTag.Command] }).map((command) => command.propertyName)
  );
  const conformantNames = new Set(
    scope.membersOf(scope.owner, { tags: [ElementTag.Command], conformance: "conformant" }).map((command) => command.propertyName)
  );
  for (const name of names) {
    let implementation;
    if (!conformantNames.has(name)) {
      if (instance[name] && instance[name] !== Behavior.unimplemented) {
        continue;
      }
    }
    if (commandFactory) {
      if (instance[name]?.[sourceFactory] === commandFactory) {
        continue;
      }
      implementation = commandFactory(name);
      implementation[sourceFactory] = commandFactory;
    } else {
      if (instance[name]) {
        continue;
      }
      implementation = Behavior.unimplemented;
    }
    result[name] = {
      value: implementation,
      writable: true
    };
  }
  return result;
}
function selectDefaultValue(scope, oldDefault, member) {
  if (oldDefault !== void 0) {
    return oldDefault;
  }
  if (!scope.hasOperationalSupport(member)) {
    return;
  }
  const effectiveDefault = DefaultValue(scope, member);
  if (effectiveDefault !== void 0) {
    if (member.effectiveMetatype === "bitmap") {
      return DecodedBitmap(member, effectiveDefault);
    }
    return effectiveDefault;
  }
  if (member.nullable) {
    return null;
  }
  switch (member.effectiveMetatype) {
    case Metatype.integer:
    case Metatype.float:
      return 0;
    case Metatype.boolean:
      return false;
    case Metatype.bitmap:
    case Metatype.object:
      return {};
    case Metatype.array:
      return [];
  }
}
function createEventDescriptor(name, schema, constructor) {
  return {
    get() {
      if (this.hasEvent(name, true)) {
        return this.getEvent(name);
      }
      const event = new constructor(schema, this);
      this.addEvent(name, event);
      return event;
    },
    enumerable: true
  };
}
export {
  ClusterBehaviorType
};
