// @matter/node/dist/esm/behavior/cluster/ClusterBehavior.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Events } from "#behavior/Events.js";
import { ImplementationError } from "@matter/general";
import { ClusterModifier } from "@matter/model";
import { ClusterType } from "@matter/types";
import { Behavior } from "../Behavior.js";
import { NetworkBehavior } from "../system/network/NetworkBehavior.js";
import { isClientBehavior } from "./cluster-behavior-utils.js";
import { ClusterBehaviorType } from "./ClusterBehaviorType.js";
import { ClusterInterface } from "./ClusterInterface.js";
class ClusterBehavior extends Behavior {
  /**
   * The ID of ClusterBehavior implementations is the uncapitalized cluster name.
   */
  static id;
  /**
   * The cluster namespace associated with this behavior.
   */
  get cluster() {
    return this.constructor.cluster;
  }
  /**
   * Supported features as a flag object.
   */
  get features() {
    return this.constructor.features;
  }
  get type() {
    return this.constructor;
  }
  /**
   * Every cluster behavior has an associated cluster namespace defined statically.
   */
  static cluster = ClusterType.Unknown;
  /**
   * Supported features as a flag object.
   */
  static features = {};
  /**
   * Method definitions.
   */
  static Interface = ClusterInterface.Empty;
  /**
   * All ClusterBehavior initialization currently runs as part of {@link Endpoint} initialization.
   */
  static early = true;
  /**
   * Keep networking alive until I'm destroyed.
   */
  static dependencies = [NetworkBehavior];
  /**
   * Automatically lock state on command invoke.
   */
  static lockOnInvoke = true;
  constructor(agent, backing) {
    super(agent, backing);
    const cluster = this.constructor.cluster;
    if (!cluster) {
      throw new ImplementationError("ClusterBehavior class has no cluster defined");
    }
  }
  static for(ns, schema, name) {
    return ClusterBehaviorType({
      namespace: ns,
      base: this,
      schema,
      name
    });
  }
  /**
   * Create a new behavior with different cluster features.
   */
  static withFeatures(...features) {
    return ClusterBehaviorType({
      namespace: this.cluster,
      base: this,
      features
    });
  }
  /**
   * Alias for {@link withFeatures}.
   */
  static with(...features) {
    return this.withFeatures(...features);
  }
  /**
   * Create a new behavior with modified cluster elements.
   */
  static alter(alterations) {
    const schema = ClusterModifier.applyRequirements(this.schema, alterations);
    return this.for(this.cluster, schema);
  }
  /**
   * Create a new behavior with additional cluster features marked "mandatory".
   *
   * This informs matter.js that an application supports these elements.
   */
  static enable(flags) {
    const schema = ClusterModifier.applyPresence(this.schema, flags);
    return this.for(this.cluster, schema);
  }
  /**
   * Returns this behavior type with all cluster elements present regardless of feature selection.
   *
   * This is type-only; at runtime the deconflicted conformance mode already includes all elements.
   */
  static get complete() {
    return this;
  }
  static isType(type) {
    return typeof type?.withFeatures === "function";
  }
  static supports(other) {
    if (!ClusterBehavior.isType(other)) {
      return false;
    }
    if (isClientBehavior(other) && other.schema.id === this.schema.id) {
      return true;
    }
    if (!Behavior.supports.call(this, other)) {
      return false;
    }
    const otherFeatures = other.schema.supportedFeatures;
    for (const name of otherFeatures) {
      if (!this.schema.supportedFeatures.has(name)) {
        return false;
      }
    }
    return true;
  }
  requireAttributeEnabled(attributeName) {
    if (this.state[attributeName] === void 0) {
      throw new ImplementationError(
        `To use this feature, please enable attribute ${String(attributeName)} by setting the value during initialization`
      );
    }
    return this.state[attributeName];
  }
  assertAttributeEnabled(attributeName) {
    if (this.state[attributeName] === void 0) {
      throw new ImplementationError(
        `To use this feature, please enable attribute ${String(attributeName)} by setting the value during initialization`
      );
    }
  }
  static Events = Events;
}
((ClusterBehavior2) => {
  function is(type) {
    return "cluster" in type;
  }
  ClusterBehavior2.is = is;
})(ClusterBehavior || (ClusterBehavior = {}));
export {
  ClusterBehavior
};
