// @matter/model/dist/cjs/logic/GeneratorScope.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GeneratorScope_exports = {};
__export(GeneratorScope_exports, {
  GeneratorScope: () => GeneratorScope
});
module.exports = __toCommonJS(GeneratorScope_exports);
var import_general = require("@matter/general");
var import_ElementTag = require("../common/ElementTag.js");
var import_Metatype = require("../common/Metatype.js");
var import_ClusterModel = require("../models/ClusterModel.js");
var import_CommandModel = require("../models/CommandModel.js");
var import_DatatypeModel = require("../models/DatatypeModel.js");
var import_EventModel = require("../models/EventModel.js");
var import_FieldModel = require("../models/FieldModel.js");
var import_MatterModel = require("../models/MatterModel.js");
var import_Model = require("../models/Model.js");
var import_ValueModel = require("../models/ValueModel.js");
var import_Scope = require("./Scope.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const cache = /* @__PURE__ */ new WeakMap();
function GeneratorScope(model) {
  if (!(model instanceof import_Model.Model)) {
    return model;
  }
  let rootModel = model;
  while (rootModel.parent && rootModel.parent.tag !== import_ElementTag.ElementTag.Matter) {
    rootModel = rootModel.parent;
  }
  const owner = (0, import_Scope.Scope)(rootModel, { forceOwner: true, forceCache: true });
  let result = cache.get(owner);
  if (result === void 0) {
    cache.set(owner, result = allocateScope(owner));
  }
  return result;
}
function allocateScope(scope) {
  const { locations, names } = assignNames(scope);
  const { owner } = scope;
  return {
    owner,
    scope,
    nameFor(model, tlv, specific = false) {
      const { definition: definer } = this.locationOf(model, specific);
      const name = names.get(definer);
      if (name === void 0) {
        throw new import_general.InternalError(`No name assigned to ${model} in scope ${owner}`);
      }
      if (tlv) {
        return `Tlv${name}`;
      }
      return name;
    },
    locationOf(model, specific = false) {
      if (!specific) {
        model = scope.modelFor(model);
      }
      const location = locations.get(model);
      if (location === void 0) {
        throw new import_general.InternalError(`No location identified for ${model} from scope ${owner}`);
      }
      return location;
    },
    canonicalModelFor(model) {
      return scope.extensionOf(model) ?? model;
    },
    definingModelFor(model) {
      let result;
      model.forEachAncestor((model2) => {
        if (model2 instanceof import_ValueModel.ValueModel) {
          result = this.canonicalModelFor(model2).definingModel;
          if (result) {
            result = this.canonicalModelFor(result);
            return false;
          }
        }
      });
      return result;
    },
    *namedModels() {
      for (const [model, location] of locations) {
        const name = names.get(model);
        if (name !== void 0) {
          yield [model, name, location];
        }
      }
    }
  };
}
var Priority = /* @__PURE__ */ ((Priority2) => {
  Priority2[Priority2["LocalNamespace"] = 0] = "LocalNamespace";
  Priority2[Priority2["Enum"] = 1] = "Enum";
  Priority2[Priority2["Struct"] = 2] = "Struct";
  Priority2[Priority2["Bitmap"] = 3] = "Bitmap";
  Priority2[Priority2["OtherNamespace"] = 4] = "OtherNamespace";
  Priority2[Priority2["Global"] = 5] = "Global";
  Priority2[Priority2["Attribute"] = 6] = "Attribute";
  Priority2[Priority2["Command"] = 7] = "Command";
  Priority2[Priority2["Event"] = 8] = "Event";
  return Priority2;
})(Priority || {});
class Locations extends Map {
}
function assignNames(scope) {
  const locations = identifyNamedModels(scope);
  const priorities = assignPriorities(locations);
  const names = assignNamesByPriority(priorities);
  return { names, locations };
}
function identifyNamedModels(scope) {
  const locations = new Locations();
  const { owner } = scope;
  locations.set(scope.owner, {
    definition: owner,
    scope: owner,
    isLocal: true,
    isGlobal: true
  });
  const base = owner.base;
  if (base instanceof import_ClusterModel.ClusterModel || base instanceof import_ValueModel.ValueModel && !base.metatype) {
    locations.set(base, {
      definition: base,
      scope: base,
      isLocal: false,
      isGlobal: true
    });
  }
  owner.visit(define);
  if (owner instanceof import_ClusterModel.ClusterModel) {
    const aces = owner.allAces;
    for (const ace of aces) {
      if (ace.parent === owner) {
        continue;
      }
      ace.visit(define);
    }
  }
  return locations;
  function define(model, specific = false) {
    if (!(model instanceof import_ValueModel.ValueModel)) {
      return;
    }
    const metatype = model.effectiveMetatype;
    switch (metatype) {
      case import_Metatype.Metatype.array:
        const entry = model.listEntry;
        if (entry) {
          define(entry);
        }
        break;
      case import_Metatype.Metatype.object:
      case import_Metatype.Metatype.enum:
      case import_Metatype.Metatype.bitmap:
        break;
      default:
        return;
    }
    let definer = model.definingModel;
    if (!definer) {
      return;
    }
    if (!specific) {
      const extension = scope.extensionOf(definer);
      if (extension === model) {
        definer = model;
      } else if (extension) {
        define(extension);
        return;
      }
    }
    if (locations.has(definer)) {
      return;
    }
    let modelScope = definer.owner(import_ClusterModel.ClusterModel) ?? definer.owner(import_MatterModel.MatterModel);
    if (modelScope === void 0) {
      throw new import_general.InternalError(`Model ${definer} does not appear to be part of a proper hierarchy`);
    }
    const location = {
      definition: definer,
      scope: modelScope,
      isLocal: modelScope === owner,
      isGlobal: modelScope.tag === import_ElementTag.ElementTag.Matter
    };
    locations.set(definer, location);
    if (definer instanceof import_ValueModel.ValueModel && !specific && referencesShadows(definer)) {
      location.isLocal = true;
      modelScope = owner;
      definer.members.forEach((m) => define(m));
    }
    if (modelScope instanceof import_ClusterModel.ClusterModel && modelScope !== owner) {
      locations.set(modelScope, {
        definition: modelScope,
        scope: modelScope,
        isLocal: false,
        isGlobal: true
      });
    }
    if (metatype === import_Metatype.Metatype.enum) {
      const { base: base2 } = definer;
      if (base2 && base2.metatype === void 0) {
        define(base2, true);
      }
    }
  }
  function referencesShadows(model) {
    const visited = /* @__PURE__ */ new Set();
    return model.visit((descendent) => {
      if (visited.has(descendent)) {
        return;
      }
      visited.add(descendent);
      if (scope.isShadow(descendent)) {
        return false;
      }
      const base2 = descendent.base;
      if (!base2) {
        return;
      }
      if (referencesShadows(base2)) {
        return false;
      }
    }) === false;
  }
}
function assignPriorities(locations) {
  const priorities = Array();
  function collect(priority, model) {
    let slot = priorities[priority];
    if (slot === void 0) {
      slot = priorities[priority] = /* @__PURE__ */ new Set();
    }
    slot.add(model);
  }
  for (const [model, { definition: definer, isLocal, isGlobal }] of locations) {
    if (definer instanceof import_DatatypeModel.DatatypeModel || definer instanceof import_FieldModel.FieldModel) {
      const definingNamespace = definer.owner(import_ClusterModel.ClusterModel) ?? definer.owner(import_MatterModel.MatterModel);
      if (definingNamespace === void 0) {
        throw new import_general.InternalError(`Model ${definer} does not appear to be part of a proper hierarchy`);
      }
      if (isLocal) {
        switch (model.effectiveMetatype) {
          case import_Metatype.Metatype.object:
            collect(2 /* Struct */, definer);
            break;
          case import_Metatype.Metatype.enum:
            collect(1 /* Enum */, definer);
            break;
          case import_Metatype.Metatype.bitmap:
            collect(3 /* Bitmap */, definer);
            break;
        }
      } else {
        if (isGlobal) {
          collect(5 /* Global */, definer);
        } else {
          collect(4 /* OtherNamespace */, definingNamespace);
        }
      }
    } else {
      switch (model.tag) {
        case import_ElementTag.ElementTag.Attribute:
          collect(6 /* Attribute */, definer);
          break;
        case import_ElementTag.ElementTag.Event:
          collect(8 /* Event */, definer);
          break;
        case import_ElementTag.ElementTag.Command:
          collect(7 /* Command */, definer);
          break;
        case import_ElementTag.ElementTag.Cluster:
          collect(isLocal ? 0 /* LocalNamespace */ : 4 /* OtherNamespace */, definer);
      }
    }
  }
  return priorities;
}
function assignNamesByPriority(priorities) {
  const names = /* @__PURE__ */ new Map();
  const assigned = /* @__PURE__ */ new Set();
  function assign(model) {
    let name = (0, import_general.camelize)(model.name, true);
    if (model instanceof import_DatatypeModel.DatatypeModel) {
      name = name.replace(/(?:Enum|Struct|Bitmap)$/, "");
    } else if (model instanceof import_CommandModel.CommandModel) {
      name = name.replace(/(?:Request|Response)$/, "");
      if (model.isRequest) {
        name += "Request";
      } else {
        name += "Response";
      }
    } else if (model instanceof import_EventModel.EventModel) {
      if (!name.endsWith("Event")) {
        name += "Event";
      }
    } else if (model instanceof import_ClusterModel.ClusterModel) {
      name = name.replace(/Cluster$/, "");
    }
    if (name === "Type") {
      name = `${model.parent?.name}Type`;
    }
    if (assigned.has(name)) {
      switch (model.tag) {
        case import_ElementTag.ElementTag.Attribute:
          name += "Attribute";
          break;
        case import_ElementTag.ElementTag.Field:
          if (model.parent && model.name !== "Type") {
            name = `${model.parent.name}${name}`;
          }
          break;
        case import_ElementTag.ElementTag.Cluster:
          name += "Namespace";
          break;
        case import_ElementTag.ElementTag.Datatype:
          switch (model.effectiveMetatype) {
            case import_Metatype.Metatype.object:
              name += "Struct";
              break;
            case import_Metatype.Metatype.enum:
              name += "Enum";
              break;
            case import_Metatype.Metatype.bitmap:
              name += "Bitmap";
              break;
          }
      }
    }
    if (assigned.has(name)) {
      let next = 2;
      while (true) {
        const candidate = `${name}${next}`;
        if (assigned.has(candidate)) {
          next++;
        } else {
          name = candidate;
          break;
        }
      }
    }
    names.set(model, name);
    assigned.add(name);
  }
  for (const models of priorities) {
    if (!models) {
      continue;
    }
    for (const model of models) {
      assign(model);
    }
  }
  return names;
}
