// @matter/types/dist/esm/dcl/device.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { BitFlag, BitmapSchema } from "../schema/BitmapSchema.js";
const EnhancedSetupFlowOptionsBitmap = {
  /**
   * A commissioner is required to show Terms and Conditions content if this is the initial
   * commissioning for the device out of a factory-reset state.
   */
  tcOnInitialCommission: BitFlag(0),
  /**
   * The commissioner SHALL NOT reuse user's responses to the required Terms and Conditions
   * from commissioning a previous product (same VID) with the same T&C.
   */
  disallowTcReuseForRequiredTc: BitFlag(1),
  /**
   * The commissioner SHALL NOT reuse user's responses to the non-required Terms and Conditions
   * from commissioning a previous product (same VID) with the same T&C.
   */
  disallowTcReuseForNonRequiredTc: BitFlag(2),
  /**
   * The commissioner SHALL NOT reuse user's responses to the Terms and Conditions from
   * commissioning a previous product (same VID, other PID) with the same T&C.
   */
  disallowTcReuseBetweenPids: BitFlag(3)
};
const EnhancedSetupFlowOptionsSchema = BitmapSchema(EnhancedSetupFlowOptionsBitmap);
export {
  EnhancedSetupFlowOptionsBitmap,
  EnhancedSetupFlowOptionsSchema
};
