// @matter/general/dist/cjs/net/http/MockWsConnection.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MockWsConnection_exports = {};
__export(MockWsConnection_exports, {
  MockWsConnection: () => MockWsConnection
});
module.exports = __toCommonJS(MockWsConnection_exports);
var import_Time = require("#time/Time.js");
var import_Bytes = require("#util/Bytes.js");
var import_Promises = require("#util/Promises.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function MockWsConnection() {
  const clientToServer = createPipe();
  const serverToClient = createPipe();
  return {
    client: {
      readable: serverToClient.readable,
      writable: clientToServer.writable
    },
    server: {
      readable: clientToServer.readable,
      writable: serverToClient.writable
    }
  };
}
((MockWsConnection2) => {
  async function send(connection, message) {
    const writer = connection.writable.getWriter();
    try {
      await writer.write(JSON.stringify(message));
    } finally {
      writer.releaseLock();
    }
  }
  MockWsConnection2.send = send;
  async function receive(connection) {
    const reader = connection.readable.getReader();
    try {
      const result = await reader.read();
      if (result.done) {
        throw new Error("Expected message but stream ended");
      }
      return JSON.parse(import_Bytes.Bytes.toString(result.value));
    } finally {
      reader.releaseLock();
    }
  }
  MockWsConnection2.receive = receive;
})(MockWsConnection || (MockWsConnection = {}));
function createPipe() {
  const queue = Array();
  const ready = new import_Promises.Gate();
  let isClosed = false;
  let deliveryScheduled = false;
  function scheduleDelivery() {
    if (!deliveryScheduled) {
      deliveryScheduled = true;
      void import_Time.Time.macrotask.then(() => {
        deliveryScheduled = false;
        ready.open();
      });
    }
  }
  const readable = new ReadableStream({
    async pull(controller) {
      if (!isClosed && !queue.length) {
        await ready;
      }
      const next = queue.shift();
      if (!queue.length && !isClosed) {
        ready.close();
      }
      if (next === void 0) {
        controller.close();
        return;
      }
      controller.enqueue(next);
    },
    cancel() {
      isClosed = true;
      queue.length = 0;
      ready.open();
    }
  });
  const writable = new WritableStream({
    write(chunk) {
      if (isClosed) {
        return;
      }
      queue.push(chunk);
      scheduleDelivery();
    },
    close() {
      isClosed = true;
      scheduleDelivery();
    },
    abort() {
      isClosed = true;
      queue.length = 0;
      ready.open();
    }
  });
  return { readable, writable };
}
