// @matter/general/dist/cjs/net/http/MockHttpEndpointFactory.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MockHttpEndpointFactory_exports = {};
__export(MockHttpEndpointFactory_exports, {
  MockHttpEndpointFactory: () => MockHttpEndpointFactory
});
module.exports = __toCommonJS(MockHttpEndpointFactory_exports);
var import_HttpEndpointFactory = require("./HttpEndpointFactory.js");
var import_MockWsConnection = require("./MockWsConnection.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MockHttpEndpointFactory extends import_HttpEndpointFactory.HttpEndpointFactory {
  #target;
  async create(_options) {
    const endpoint = {
      http: void 0,
      ws: void 0,
      async close() {
      }
    };
    this.#target = endpoint;
    return endpoint;
  }
  /**
   * Create a new mock WebSocket connection pair.  Invokes the captured WS handler with the server side and returns
   * the client side.
   */
  async connect() {
    const target = this.#target;
    if (!target?.ws) {
      throw new Error("No WebSocket handler registered; is WebSocketServer started?");
    }
    const pair = (0, import_MockWsConnection.MockWsConnection)();
    target.ws(new Request(MockHttpEndpointFactory.WS_URL), () => Promise.resolve(pair.server));
    return pair.client;
  }
}
((MockHttpEndpointFactory2) => {
  MockHttpEndpointFactory2.WS_URL = "ws+unix://matter.sock/";
})(MockHttpEndpointFactory || (MockHttpEndpointFactory = {}));
