// @matter/node/dist/esm/behaviors/door-lock/UserDoorLockServer.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Supervision } from "#behavior/supervision/Supervision.js";
import { StatusCode, StatusResponseError } from "@matter/types";
import { DoorLock } from "@matter/types/clusters/door-lock";
import { LockAuth } from "./LockAuth.js";
import { NonUserDoorLockServer } from "./NonUserDoorLockServer.js";
const CredentialRule = DoorLock.CredentialRule;
const CredentialType = DoorLock.CredentialType;
const DataOperationType = DoorLock.DataOperationType;
const LockDataType = DoorLock.LockDataType;
const OperationSource = DoorLock.OperationSource;
const UserStatus = DoorLock.UserStatus;
const UserType = DoorLock.UserType;
const UserBase = NonUserDoorLockServer.with(
  "PinCredential",
  "RfidCredential",
  "FingerCredentials",
  "FaceCredentials",
  "User",
  "WeekDayAccessSchedules",
  "YearDayAccessSchedules",
  "HolidaySchedules"
);
class UserDoorLockServer extends UserBase {
  // ── User Database (USR) ──────────────────────────────────────────────────
  setUser(request) {
    const { operationType, userIndex } = request;
    const maxUsers = this.state.numberOfTotalUsersSupported;
    const fabricIndex = this.#fabricIndex;
    if (userIndex < 1 || userIndex > maxUsers) {
      throw new StatusResponseError("Invalid user index", StatusCode.InvalidCommand);
    }
    const auth = this.auth;
    const existing = auth.findUser(userIndex);
    if (operationType === DataOperationType.Add) {
      if (existing) {
        throw new DoorLock.OccupiedError("User slot is occupied");
      }
      auth.addUser({
        userIndex,
        userName: request.userName ?? "",
        userUniqueId: request.userUniqueId ?? null,
        userStatus: request.userStatus ?? UserStatus.OccupiedEnabled,
        userType: request.userType ?? UserType.UnrestrictedUser,
        credentialRule: request.credentialRule ?? CredentialRule.Single,
        credentials: [],
        creatorFabricIndex: fabricIndex,
        lastModifiedFabricIndex: fabricIndex
      });
      this.#emitLockUserChange(LockDataType.UserIndex, DataOperationType.Add, userIndex, fabricIndex, null);
    } else if (operationType === DataOperationType.Modify) {
      if (!existing) {
        throw new StatusResponseError("User slot is available", StatusCode.InvalidCommand);
      }
      if (request.userName !== null && fabricIndex !== existing.creatorFabricIndex) {
        throw new StatusResponseError(
          "Cannot modify userName from different fabric",
          StatusCode.InvalidCommand
        );
      }
      if (request.userUniqueId !== null && fabricIndex !== existing.creatorFabricIndex) {
        throw new StatusResponseError(
          "Cannot modify userUniqueId from different fabric",
          StatusCode.InvalidCommand
        );
      }
      auth.replaceUser(userIndex, {
        ...existing,
        userName: request.userName ?? existing.userName,
        userUniqueId: request.userUniqueId ?? existing.userUniqueId,
        userStatus: request.userStatus ?? existing.userStatus,
        userType: request.userType ?? existing.userType,
        credentialRule: request.credentialRule ?? existing.credentialRule,
        lastModifiedFabricIndex: fabricIndex
      });
      this.#emitLockUserChange(LockDataType.UserIndex, DataOperationType.Modify, userIndex, fabricIndex, null);
    } else {
      throw new StatusResponseError("Invalid operation type", StatusCode.InvalidCommand);
    }
  }
  getUser(request) {
    const maxUsers = this.state.numberOfTotalUsersSupported;
    if (request.userIndex < 1 || request.userIndex > maxUsers) {
      throw new StatusResponseError("Invalid user index", StatusCode.InvalidCommand);
    }
    const auth = this.auth;
    const user = auth.findUser(request.userIndex);
    const nextUserIndex = auth.findNextOccupiedUserIndex(request.userIndex);
    if (!user) {
      return {
        userIndex: request.userIndex,
        userName: null,
        userUniqueId: null,
        userStatus: null,
        userType: null,
        credentialRule: null,
        credentials: null,
        creatorFabricIndex: null,
        lastModifiedFabricIndex: null,
        nextUserIndex
      };
    }
    return {
      userIndex: user.userIndex,
      userName: user.userName,
      userUniqueId: user.userUniqueId,
      userStatus: user.userStatus,
      userType: user.userType,
      credentialRule: user.credentialRule,
      credentials: user.credentials.length > 0 ? [...user.credentials] : [],
      creatorFabricIndex: user.creatorFabricIndex,
      lastModifiedFabricIndex: user.lastModifiedFabricIndex,
      nextUserIndex
    };
  }
  clearUser(request) {
    const { userIndex } = request;
    const maxUsers = this.state.numberOfTotalUsersSupported;
    const fabricIndex = this.#fabricIndex;
    const auth = this.auth;
    if (userIndex === 65534) {
      for (const user2 of auth.users) {
        this.#clearCredentialsForUser(auth, user2.userIndex);
        this.#clearSchedulesForUser(user2.userIndex);
      }
      auth.clearUsers();
      this.#emitLockUserChange(LockDataType.UserIndex, DataOperationType.Clear, 65534, fabricIndex, null);
      return;
    }
    if (userIndex < 1 || userIndex > maxUsers) {
      throw new StatusResponseError("Invalid user index", StatusCode.InvalidCommand);
    }
    const user = auth.findUser(userIndex);
    if (!user) {
      return;
    }
    this.#clearCredentialsForUser(auth, userIndex);
    this.#clearSchedulesForUser(userIndex);
    auth.removeUser(userIndex);
    this.#emitLockUserChange(LockDataType.UserIndex, DataOperationType.Clear, userIndex, fabricIndex, null);
  }
  // ── Credential Management (USR) ──────────────────────────────────────────
  setCredential(request) {
    const { operationType, credential, credentialData, userIndex, userStatus, userType } = request;
    const fabricIndex = this.#fabricIndex;
    const maxCredentials = this.#maxCredentialsForType(credential.credentialType);
    const auth = this.auth;
    if (credential.credentialIndex < 0 || credential.credentialIndex > maxCredentials) {
      return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex: null };
    }
    if (!this.#validateCredentialDataLength(credential.credentialType, credentialData)) {
      return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex: null };
    }
    if (auth.isDuplicateCredential(credential.credentialType, credentialData, credential.credentialIndex)) {
      return { status: StatusCode.Failure, userIndex: null, nextCredentialIndex: null };
    }
    const nextCredentialIndex = auth.findNextAvailableCredentialIndex(
      credential.credentialType,
      credential.credentialIndex,
      maxCredentials
    );
    if (operationType === DataOperationType.Add) {
      const existingCred = auth.findCredential(credential.credentialType, credential.credentialIndex);
      if (existingCred) {
        return { status: StatusCode.Failure, userIndex: null, nextCredentialIndex };
      }
      auth.addCredential(credential.credentialType, credential.credentialIndex, credentialData, fabricIndex);
      let createdUserIndex = null;
      if (userIndex === null) {
        const newUserIndex = auth.findAvailableUserIndex(this.state.numberOfTotalUsersSupported);
        if (newUserIndex === null) {
          auth.removeCredential(credential.credentialType, credential.credentialIndex);
          return { status: StatusCode.ResourceExhausted, userIndex: null, nextCredentialIndex };
        }
        auth.addUser({
          userIndex: newUserIndex,
          userName: "",
          userUniqueId: 4294967295,
          userStatus: userStatus ?? UserStatus.OccupiedEnabled,
          userType: userType ?? UserType.UnrestrictedUser,
          credentialRule: CredentialRule.Single,
          credentials: [
            { credentialType: credential.credentialType, credentialIndex: credential.credentialIndex }
          ],
          creatorFabricIndex: fabricIndex,
          lastModifiedFabricIndex: fabricIndex
        });
        createdUserIndex = newUserIndex;
        this.#emitLockUserChange(
          LockDataType.UserIndex,
          DataOperationType.Add,
          newUserIndex,
          fabricIndex,
          credential.credentialIndex
        );
      } else {
        const user = auth.findUser(userIndex);
        if (!user) {
          auth.removeCredential(credential.credentialType, credential.credentialIndex);
          return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
        }
        if (fabricIndex !== user.creatorFabricIndex) {
          auth.removeCredential(credential.credentialType, credential.credentialIndex);
          return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
        }
        if (user.credentials.length >= this.state.numberOfCredentialsSupportedPerUser) {
          auth.removeCredential(credential.credentialType, credential.credentialIndex);
          return { status: StatusCode.ResourceExhausted, userIndex: null, nextCredentialIndex };
        }
        auth.replaceUser(userIndex, {
          ...user,
          credentials: [
            ...user.credentials,
            { credentialType: credential.credentialType, credentialIndex: credential.credentialIndex }
          ],
          lastModifiedFabricIndex: fabricIndex
        });
      }
      this.#emitLockUserChange(
        LockAuth.credentialTypeToLockDataType(credential.credentialType),
        DataOperationType.Add,
        userIndex ?? createdUserIndex,
        fabricIndex,
        credential.credentialIndex
      );
      return { status: StatusCode.Success, userIndex: createdUserIndex, nextCredentialIndex };
    } else if (operationType === DataOperationType.Modify) {
      const existingCred = auth.findCredential(credential.credentialType, credential.credentialIndex);
      if (!existingCred) {
        return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
      }
      if (fabricIndex !== existingCred.creatorFabricIndex) {
        return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
      }
      if (userIndex !== null) {
        const user = auth.findUser(userIndex);
        if (!user) {
          return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
        }
        if (fabricIndex !== user.creatorFabricIndex) {
          return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
        }
        const hasCredential = user.credentials.some(
          (c) => c.credentialType === credential.credentialType && c.credentialIndex === credential.credentialIndex
        );
        if (!hasCredential) {
          return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex };
        }
      }
      auth.updateCredentialData(
        credential.credentialType,
        credential.credentialIndex,
        credentialData,
        fabricIndex
      );
      this.#emitLockUserChange(
        LockAuth.credentialTypeToLockDataType(credential.credentialType),
        DataOperationType.Modify,
        userIndex,
        fabricIndex,
        credential.credentialIndex
      );
      return { status: StatusCode.Success, userIndex: null, nextCredentialIndex };
    }
    return { status: StatusCode.InvalidCommand, userIndex: null, nextCredentialIndex: null };
  }
  getCredentialStatus(request) {
    const { credential } = request;
    const auth = this.auth;
    const cred = auth.findCredential(credential.credentialType, credential.credentialIndex);
    const nextCredentialIndex = auth.findNextOccupiedCredentialIndex(
      credential.credentialType,
      credential.credentialIndex
    );
    if (!cred) {
      return {
        credentialExists: false,
        userIndex: null,
        creatorFabricIndex: null,
        lastModifiedFabricIndex: null,
        nextCredentialIndex
      };
    }
    const userIndex = auth.findUserIndexForCredential(credential.credentialType, credential.credentialIndex);
    return {
      credentialExists: true,
      userIndex,
      creatorFabricIndex: cred.creatorFabricIndex,
      lastModifiedFabricIndex: cred.lastModifiedFabricIndex,
      nextCredentialIndex
    };
  }
  clearCredential(request) {
    const { credential } = request;
    const fabricIndex = this.#fabricIndex;
    const auth = this.auth;
    if (credential === null) {
      const typesToClear = [
        CredentialType.Pin,
        CredentialType.Rfid,
        CredentialType.Fingerprint,
        CredentialType.FingerVein,
        CredentialType.Face
      ];
      for (const type of typesToClear) {
        this.#clearAllCredentialsOfType(auth, type, fabricIndex);
      }
      return;
    }
    if (credential.credentialType === CredentialType.ProgrammingPin) {
      throw new StatusResponseError("Cannot clear ProgrammingPIN", StatusCode.InvalidCommand);
    }
    if (credential.credentialIndex === 65534) {
      this.#clearAllCredentialsOfType(auth, credential.credentialType, fabricIndex);
      return;
    }
    const cred = auth.findCredential(credential.credentialType, credential.credentialIndex);
    if (!cred) {
      throw new StatusResponseError("Credential not found", StatusCode.InvalidCommand);
    }
    this.#removeCredentialFromUsers(auth, credential.credentialType, credential.credentialIndex, fabricIndex);
    auth.removeCredential(credential.credentialType, credential.credentialIndex);
    this.#emitLockUserChange(
      LockAuth.credentialTypeToLockDataType(credential.credentialType),
      DataOperationType.Clear,
      null,
      fabricIndex,
      credential.credentialIndex
    );
  }
  // ── Schedules (WDSCH, YDSCH, HDSCH) ─────────────────────────────────────
  setWeekDaySchedule(request) {
    const maxSchedules = this.state.numberOfWeekDaySchedulesSupportedPerUser;
    const { weekDayIndex, userIndex } = request;
    if (weekDayIndex < 1 || weekDayIndex > maxSchedules) {
      throw new StatusResponseError("Invalid schedule index", StatusCode.InvalidCommand);
    }
    this.#requireValidUser(userIndex);
    if (request.endHour < request.startHour) {
      throw new StatusResponseError("End hour must be >= start hour", StatusCode.InvalidCommand);
    }
    if (request.endHour === request.startHour && request.endMinute <= request.startMinute) {
      throw new StatusResponseError("End time must be after start time", StatusCode.InvalidCommand);
    }
    const schedules = this.state.weekDaySchedules.filter(
      (s) => !(s.weekDayIndex === weekDayIndex && s.userIndex === userIndex)
    );
    const schedule = {
      weekDayIndex,
      userIndex,
      daysMask: request.daysMask,
      startHour: request.startHour,
      startMinute: request.startMinute,
      endHour: request.endHour,
      endMinute: request.endMinute
    };
    this.state.weekDaySchedules = [...schedules, schedule];
  }
  getWeekDaySchedule(request) {
    const maxSchedules = this.state.numberOfWeekDaySchedulesSupportedPerUser;
    const { weekDayIndex, userIndex } = request;
    if (weekDayIndex < 1 || weekDayIndex > maxSchedules || userIndex < 1 || userIndex > this.state.numberOfTotalUsersSupported) {
      const response = { weekDayIndex, userIndex, status: StatusCode.InvalidCommand };
      Supervision(response, "weekDayIndex").constraint = false;
      Supervision(response, "userIndex").constraint = false;
      return response;
    }
    if (!this.auth.findUser(userIndex)) {
      return { weekDayIndex, userIndex, status: StatusCode.NotFound };
    }
    const schedule = this.state.weekDaySchedules.find(
      (s) => s.weekDayIndex === weekDayIndex && s.userIndex === userIndex
    );
    if (!schedule) {
      return { weekDayIndex, userIndex, status: StatusCode.NotFound };
    }
    return {
      weekDayIndex,
      userIndex,
      status: StatusCode.Success,
      daysMask: schedule.daysMask,
      startHour: schedule.startHour,
      startMinute: schedule.startMinute,
      endHour: schedule.endHour,
      endMinute: schedule.endMinute
    };
  }
  clearWeekDaySchedule(request) {
    const { weekDayIndex, userIndex } = request;
    this.#requireValidUser(userIndex);
    if (weekDayIndex === 254) {
      this.state.weekDaySchedules = this.state.weekDaySchedules.filter((s) => s.userIndex !== userIndex);
    } else {
      const maxSchedules = this.state.numberOfWeekDaySchedulesSupportedPerUser;
      if (weekDayIndex < 1 || weekDayIndex > maxSchedules) {
        throw new StatusResponseError("Invalid schedule index", StatusCode.InvalidCommand);
      }
      this.state.weekDaySchedules = this.state.weekDaySchedules.filter(
        (s) => !(s.weekDayIndex === weekDayIndex && s.userIndex === userIndex)
      );
    }
  }
  setYearDaySchedule(request) {
    const maxSchedules = this.state.numberOfYearDaySchedulesSupportedPerUser;
    const { yearDayIndex, userIndex } = request;
    if (yearDayIndex < 1 || yearDayIndex > maxSchedules) {
      throw new StatusResponseError("Invalid schedule index", StatusCode.InvalidCommand);
    }
    this.#requireValidUser(userIndex);
    if (request.localEndTime <= request.localStartTime) {
      throw new StatusResponseError("End time must be after start time", StatusCode.InvalidCommand);
    }
    const schedules = this.state.yearDaySchedules.filter(
      (s) => !(s.yearDayIndex === yearDayIndex && s.userIndex === userIndex)
    );
    const schedule = {
      yearDayIndex,
      userIndex,
      localStartTime: request.localStartTime,
      localEndTime: request.localEndTime
    };
    this.state.yearDaySchedules = [...schedules, schedule];
  }
  getYearDaySchedule(request) {
    const maxSchedules = this.state.numberOfYearDaySchedulesSupportedPerUser;
    const { yearDayIndex, userIndex } = request;
    if (yearDayIndex < 1 || yearDayIndex > maxSchedules || userIndex < 1 || userIndex > this.state.numberOfTotalUsersSupported) {
      const response = { yearDayIndex, userIndex, status: StatusCode.InvalidCommand };
      Supervision(response, "yearDayIndex").constraint = false;
      Supervision(response, "userIndex").constraint = false;
      return response;
    }
    if (!this.auth.findUser(userIndex)) {
      return { yearDayIndex, userIndex, status: StatusCode.NotFound };
    }
    const schedule = this.state.yearDaySchedules.find(
      (s) => s.yearDayIndex === yearDayIndex && s.userIndex === userIndex
    );
    if (!schedule) {
      return { yearDayIndex, userIndex, status: StatusCode.NotFound };
    }
    return {
      yearDayIndex,
      userIndex,
      status: StatusCode.Success,
      localStartTime: schedule.localStartTime,
      localEndTime: schedule.localEndTime
    };
  }
  clearYearDaySchedule(request) {
    const { yearDayIndex, userIndex } = request;
    this.#requireValidUser(userIndex);
    if (yearDayIndex === 254) {
      this.state.yearDaySchedules = this.state.yearDaySchedules.filter((s) => s.userIndex !== userIndex);
    } else {
      const maxSchedules = this.state.numberOfYearDaySchedulesSupportedPerUser;
      if (yearDayIndex < 1 || yearDayIndex > maxSchedules) {
        throw new StatusResponseError("Invalid schedule index", StatusCode.InvalidCommand);
      }
      this.state.yearDaySchedules = this.state.yearDaySchedules.filter(
        (s) => !(s.yearDayIndex === yearDayIndex && s.userIndex === userIndex)
      );
    }
  }
  setHolidaySchedule(request) {
    const maxSchedules = this.state.numberOfHolidaySchedulesSupported;
    const { holidayIndex } = request;
    if (holidayIndex < 1 || holidayIndex > maxSchedules) {
      throw new StatusResponseError("Invalid schedule index", StatusCode.InvalidCommand);
    }
    if (request.localEndTime <= request.localStartTime) {
      throw new StatusResponseError("End time must be after start time", StatusCode.InvalidCommand);
    }
    if (request.operatingMode !== DoorLock.OperatingMode.Normal && request.operatingMode !== DoorLock.OperatingMode.Vacation && request.operatingMode !== DoorLock.OperatingMode.Privacy && request.operatingMode !== DoorLock.OperatingMode.NoRemoteLockUnlock && request.operatingMode !== DoorLock.OperatingMode.Passage) {
      throw new StatusResponseError("Invalid operating mode", StatusCode.InvalidCommand);
    }
    const schedules = this.state.holidaySchedules.filter((s) => s.holidayIndex !== holidayIndex);
    const schedule = {
      holidayIndex,
      localStartTime: request.localStartTime,
      localEndTime: request.localEndTime,
      operatingMode: request.operatingMode
    };
    this.state.holidaySchedules = [...schedules, schedule];
  }
  getHolidaySchedule(request) {
    const maxSchedules = this.state.numberOfHolidaySchedulesSupported;
    const { holidayIndex } = request;
    if (holidayIndex < 1 || holidayIndex > maxSchedules) {
      const response = { holidayIndex, status: StatusCode.InvalidCommand };
      Supervision(response, "holidayIndex").constraint = false;
      return response;
    }
    const schedule = this.state.holidaySchedules.find((s) => s.holidayIndex === holidayIndex);
    if (!schedule) {
      return { holidayIndex, status: StatusCode.NotFound };
    }
    return {
      holidayIndex,
      status: StatusCode.Success,
      localStartTime: schedule.localStartTime,
      localEndTime: schedule.localEndTime,
      operatingMode: schedule.operatingMode
    };
  }
  clearHolidaySchedule(request) {
    const { holidayIndex } = request;
    if (holidayIndex === 254) {
      this.state.holidaySchedules = [];
    } else {
      const maxSchedules = this.state.numberOfHolidaySchedulesSupported;
      if (holidayIndex < 1 || holidayIndex > maxSchedules) {
        throw new StatusResponseError("Invalid schedule index", StatusCode.InvalidCommand);
      }
      this.state.holidaySchedules = this.state.holidaySchedules.filter((s) => s.holidayIndex !== holidayIndex);
    }
  }
  // ── Private Helpers ────────────────────────────────────────────────────────
  get #fabricIndex() {
    const fabric = this.context.fabric;
    if (fabric === void 0) {
      throw new StatusResponseError("Fabric required", StatusCode.UnsupportedAccess);
    }
    return fabric;
  }
  get #sourceNode() {
    const context = this.context;
    if ("subject" in context && context.subject?.kind === "node") {
      return context.subject.id;
    }
    return null;
  }
  #emitLockUserChange(lockDataType, dataOperationType, userIndex, fabricIndex, dataIndex) {
    this.events.lockUserChange.emit(
      {
        lockDataType,
        dataOperationType,
        operationSource: OperationSource.Remote,
        userIndex,
        fabricIndex,
        sourceNode: this.#sourceNode,
        dataIndex
      },
      this.context
    );
  }
  #requireValidUser(userIndex) {
    const maxUsers = this.state.numberOfTotalUsersSupported;
    if (userIndex < 1 || userIndex > maxUsers) {
      throw new StatusResponseError("Invalid user index", StatusCode.InvalidCommand);
    }
    if (!this.auth.findUser(userIndex)) {
      throw new StatusResponseError("User not found", StatusCode.Failure);
    }
  }
  #maxCredentialsForType(type) {
    switch (type) {
      case CredentialType.Pin:
      case CredentialType.ProgrammingPin:
        return this.state.numberOfPinUsersSupported;
      case CredentialType.Rfid:
        return this.state.numberOfRfidUsersSupported;
      default:
        return this.state.numberOfTotalUsersSupported;
    }
  }
  #validateCredentialDataLength(type, data) {
    switch (type) {
      case CredentialType.Pin:
      case CredentialType.ProgrammingPin: {
        const min = this.state.minPinCodeLength;
        const max = this.state.maxPinCodeLength;
        return data.byteLength >= min && data.byteLength <= max;
      }
      case CredentialType.Rfid: {
        const min = this.state.minRfidCodeLength;
        const max = this.state.maxRfidCodeLength;
        return data.byteLength >= min && data.byteLength <= max;
      }
      default:
        return true;
    }
  }
  #clearCredentialsForUser(auth, userIndex) {
    const user = auth.findUser(userIndex);
    if (!user) return;
    for (const cred of user.credentials) {
      auth.removeCredential(cred.credentialType, cred.credentialIndex);
    }
  }
  #clearAllCredentialsOfType(auth, type, fabricIndex) {
    const credsOfType = auth.credentials.filter((c) => c.credentialType === type);
    for (const cred of credsOfType) {
      this.#removeCredentialFromUsers(auth, type, cred.credentialIndex, fabricIndex);
    }
    auth.removeAllCredentialsOfType(type);
    this.#emitLockUserChange(
      LockAuth.credentialTypeToLockDataType(type),
      DataOperationType.Clear,
      null,
      fabricIndex,
      65534
    );
  }
  #removeCredentialFromUsers(auth, type, index, _fabricIndex) {
    const userIndex = auth.findUserIndexForCredential(type, index);
    if (userIndex === null) return;
    const user = auth.findUser(userIndex);
    if (!user) return;
    auth.replaceUser(userIndex, {
      ...user,
      credentials: user.credentials.filter((c) => !(c.credentialType === type && c.credentialIndex === index))
    });
  }
  #clearSchedulesForUser(userIndex) {
    this.state.weekDaySchedules = this.state.weekDaySchedules.filter((s) => s.userIndex !== userIndex);
    this.state.yearDaySchedules = this.state.yearDaySchedules.filter((s) => s.userIndex !== userIndex);
  }
}
const throwInvalidCommand = () => {
  throw new StatusResponseError("Invalid field value", StatusCode.InvalidCommand);
};
for (const method of [
  "setUser",
  "getUser",
  "clearUser",
  "setCredential",
  "setWeekDaySchedule",
  "clearWeekDaySchedule",
  "setYearDaySchedule",
  "clearYearDaySchedule",
  "setHolidaySchedule",
  "clearHolidaySchedule"
]) {
  Supervision(UserDoorLockServer, method).onValidationError = throwInvalidCommand;
}
for (const field of ["weekDayIndex", "userIndex"]) {
  Supervision(UserDoorLockServer, "getWeekDaySchedule", field).constraint = false;
}
for (const field of ["yearDayIndex", "userIndex"]) {
  Supervision(UserDoorLockServer, "getYearDaySchedule", field).constraint = false;
}
Supervision(UserDoorLockServer, "getHolidaySchedule", "holidayIndex").constraint = false;
export {
  UserDoorLockServer
};
