// @matter/node/dist/cjs/devices/variance-test.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var variance_test_exports = {};
__export(variance_test_exports, {
  TestColorControl: () => TestColorControl,
  TestDoubleWith: () => TestDoubleWith,
  TestIdentify: () => TestIdentify,
  TestLevelControl: () => TestLevelControl,
  TestOnOff: () => TestOnOff,
  TestThermostat: () => TestThermostat
});
module.exports = __toCommonJS(variance_test_exports);
var import_LevelControlServer = require("../behaviors/level-control/LevelControlServer.js");
var import_ColorControlServer = require("../behaviors/color-control/ColorControlServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_ThermostatServer = require("../behaviors/thermostat/ThermostatServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var TestLevelControl;
((TestLevelControl2) => {
  TestLevelControl2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("Lighting", "OnOff").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 }
    }
  });
})(TestLevelControl || (TestLevelControl = {}));
var TestColorControl;
((TestColorControl2) => {
  TestColorControl2.ColorControlServer = import_ColorControlServer.ColorControlServer.with("ColorTemperature").alter({
    attributes: {
      startUpColorTemperatureMireds: { default: 0 }
    }
  });
})(TestColorControl || (TestColorControl = {}));
var TestOnOff;
((TestOnOff2) => {
  TestOnOff2.OnOffServer = import_OnOffServer.OnOffServer.with("Lighting").alter({ attributes: {} });
})(TestOnOff || (TestOnOff = {}));
var TestThermostat;
((TestThermostat2) => {
  TestThermostat2.ThermostatServer = import_ThermostatServer.ThermostatServer.with("Heating").alter({ attributes: {} });
})(TestThermostat || (TestThermostat = {}));
var TestIdentify;
((TestIdentify2) => {
  TestIdentify2.IdentifyServer = import_IdentifyServer.IdentifyServer.with().alter({ attributes: {} });
})(TestIdentify || (TestIdentify = {}));
var TestDoubleWith;
((TestDoubleWith2) => {
  TestDoubleWith2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("Lighting", "OnOff").with();
})(TestDoubleWith || (TestDoubleWith = {}));
