// @matter/protocol/dist/cjs/certificate/CertificateAuthority.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CertificateAuthority_exports = {};
__export(CertificateAuthority_exports, {
  CertificateAuthority: () => CertificateAuthority
});
module.exports = __toCommonJS(CertificateAuthority_exports);
var import_general = require("@matter/general");
var import_asn = require("./kinds/definitions/asn.js");
var import_Icac = require("./kinds/Icac.js");
var import_Noc = require("./kinds/Noc.js");
var import_Rcac = require("./kinds/Rcac.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("CertificateAuthority");
class CertificateAuthority {
  #crypto;
  #rootCertId = BigInt(0);
  #rootKeyPair;
  #rootKeyIdentifier;
  #rootCertBytes;
  #nextCertificateId = BigInt(1);
  #construction;
  #icacProps;
  get crypto() {
    return this.#crypto;
  }
  get construction() {
    return this.#construction;
  }
  close() {
    return this.#construction.close();
  }
  static async create(crypto, options, generateIntermediateCert) {
    return (0, import_general.asyncNew)(CertificateAuthority, crypto, options, generateIntermediateCert);
  }
  constructor(crypto, options, generateIntermediateCert) {
    this.#crypto = crypto;
    this.#construction = (0, import_general.Construction)(this, async () => {
      if (typeof options === "boolean") {
        generateIntermediateCert = options;
        options = void 0;
      }
      const certValues = options instanceof import_general.StorageContext ? await options.values() : options ?? {};
      const requireIcac = generateIntermediateCert ?? this.#isValidStoredIcacCertificate(certValues);
      if (this.#isValidStoredRootCertificate(certValues)) {
        this.#loadFromStorage(certValues, requireIcac);
        logger.info(
          `Loaded stored credentials with ID ${this.#rootCertId}${this.#icacProps !== void 0 ? ` and ICAC with ID ${this.#icacProps.certId}` : ""}`
        );
        return;
      }
      this.#rootKeyPair = await this.#crypto.createKeyPair();
      this.#rootKeyIdentifier = import_general.Bytes.of(await this.#crypto.computeHash(this.#rootKeyPair.publicKey)).slice(
        0,
        20
      );
      this.#rootCertBytes = await this.#generateRootCert();
      if (requireIcac) {
        this.#icacProps = await this.#generateIcacProps();
      }
      logger.info(
        `Created new credentials with ID ${this.#rootCertId}${this.#icacProps !== void 0 ? ` and ICAC with ID ${this.#icacProps.certId}` : ""}`
      );
      if (options instanceof import_general.StorageContext) {
        await options.set(this.#buildStorageData());
      }
    });
  }
  static [import_general.Environmental.create](env) {
    const storage = env.get(import_general.StorageManager).createContext("certificates");
    const instance = new CertificateAuthority(env.get(import_general.Crypto), storage);
    env.set(CertificateAuthority, instance);
    return instance;
  }
  get rootCert() {
    return this.#construction.assert("root cert", this.#rootCertBytes);
  }
  get icacCert() {
    if (!this.#icacProps) {
      return void 0;
    }
    return this.#construction.assert("icac cert", this.#icacProps.certBytes);
  }
  get config() {
    return {
      rootCertId: this.#rootCertId,
      rootKeyIdentifier: this.construction.assert("root key identifier", this.#rootKeyIdentifier),
      rootCertBytes: this.construction.assert("root cert bytes", this.#rootCertBytes),
      nextCertificateId: this.#nextCertificateId,
      ...this.#icacProps !== void 0 ? {
        rootKeyPair: this.#rootKeyPair?.keyPair,
        // rootKeyPair is optional when using ICAC
        icacCertId: this.#icacProps.certId,
        icacKeyPair: this.construction.assert("icac key pair", this.#icacProps.keyPair).keyPair,
        icacKeyIdentifier: this.construction.assert("icac key identifier", this.#icacProps.keyIdentifier),
        icacCertBytes: this.construction.assert("icac cert bytes", this.#icacProps.certBytes)
      } : {
        rootKeyPair: this.construction.assert("root key pair", this.#rootKeyPair).keyPair
      }
    };
  }
  async #generateRootCert() {
    const now = import_general.Time.now;
    const cert = new import_Rcac.Rcac({
      serialNumber: import_general.Bytes.fromHex((0, import_general.toHex)(this.#rootCertId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: { rcacId: this.#rootCertId },
      notBefore: (0, import_asn.jsToMatterDate)(now, -1),
      notAfter: (0, import_asn.jsToMatterDate)(now, 10),
      subject: { rcacId: this.#rootCertId },
      ellipticCurvePublicKey: this.#initializedRootKeyPair.publicKey,
      extensions: {
        basicConstraints: { isCa: true },
        keyUsage: {
          keyCertSign: true,
          cRLSign: true
        },
        subjectKeyIdentifier: this.#initializedRootKeyIdentifier,
        authorityKeyIdentifier: this.#initializedRootKeyIdentifier
      }
    });
    await cert.sign(this.#crypto, this.#initializedRootKeyPair);
    return cert.asSignedTlv();
  }
  async #generateIcacProps() {
    const certId = this.#nextCertificateId++;
    const keyPair = await this.#crypto.createKeyPair();
    const keyIdentifier = import_general.Bytes.of(await this.#crypto.computeHash(keyPair.publicKey)).slice(0, 20);
    const now = import_general.Time.now;
    const cert = new import_Icac.Icac({
      serialNumber: import_general.Bytes.fromHex((0, import_general.toHex)(certId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: { rcacId: this.#rootCertId },
      notBefore: (0, import_asn.jsToMatterDate)(now, -1),
      notAfter: (0, import_asn.jsToMatterDate)(now, 10),
      subject: { icacId: certId },
      ellipticCurvePublicKey: keyPair.publicKey,
      extensions: {
        basicConstraints: { isCa: true },
        keyUsage: {
          keyCertSign: true,
          cRLSign: true
        },
        subjectKeyIdentifier: keyIdentifier,
        authorityKeyIdentifier: this.#initializedRootKeyIdentifier
      }
    });
    await cert.sign(this.#crypto, this.#initializedRootKeyPair);
    return {
      certId,
      keyPair,
      keyIdentifier,
      certBytes: cert.asSignedTlv()
    };
  }
  async generateNoc(publicKey, fabricId, nodeId, caseAuthenticatedTags) {
    const now = import_general.Time.now;
    const certId = this.#nextCertificateId++;
    const { issuer, signingKey, authorityKeyId } = this.#getSigningParameters();
    const cert = new import_Noc.Noc({
      serialNumber: import_general.Bytes.fromHex((0, import_general.toHex)(certId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer,
      notBefore: (0, import_asn.jsToMatterDate)(now, -1),
      notAfter: (0, import_asn.jsToMatterDate)(now, 10),
      subject: { fabricId, nodeId, caseAuthenticatedTags },
      ellipticCurvePublicKey: publicKey,
      extensions: {
        basicConstraints: { isCa: false },
        keyUsage: {
          digitalSignature: true
        },
        extendedKeyUsage: [2, 1],
        subjectKeyIdentifier: import_general.Bytes.of(await this.#crypto.computeHash(publicKey)).slice(0, 20),
        authorityKeyIdentifier: authorityKeyId
      }
    });
    await cert.sign(this.#crypto, signingKey);
    return cert.asSignedTlv();
  }
  get #initializedRootKeyPair() {
    if (this.#rootKeyPair === void 0) {
      throw new import_general.InternalError("CA private key is not installed");
    }
    return this.#rootKeyPair;
  }
  get #initializedRootKeyIdentifier() {
    if (this.#rootKeyIdentifier === void 0) {
      throw new import_general.InternalError("CA key identifier is not installed");
    }
    return this.#rootKeyIdentifier;
  }
  #isValidStoredRootCertificate(certValues) {
    return (typeof certValues.rootCertId === "number" || typeof certValues.rootCertId === "bigint") && (certValues.rootKeyPair === void 0 || import_general.Bytes.isBytes(certValues.rootKeyPair) || typeof certValues.rootKeyPair === "object") && import_general.Bytes.isBytes(certValues.rootKeyIdentifier) && import_general.Bytes.isBytes(certValues.rootCertBytes) && (typeof certValues.nextCertificateId === "number" || typeof certValues.nextCertificateId === "bigint");
  }
  #isValidStoredIcacCertificate(certValues) {
    return (typeof certValues.icacCertId === "number" || typeof certValues.icacCertId === "bigint") && (import_general.Bytes.isBytes(certValues.icacKeyPair) || typeof certValues.icacKeyPair === "object") && import_general.Bytes.isBytes(certValues.icacKeyIdentifier) && import_general.Bytes.isBytes(certValues.icacCertBytes);
  }
  #loadFromStorage(certValues, requireIcac) {
    this.#rootCertId = BigInt(certValues.rootCertId);
    if (certValues.rootKeyPair !== void 0) {
      this.#rootKeyPair = (0, import_general.PrivateKey)(certValues.rootKeyPair);
    }
    this.#rootKeyIdentifier = certValues.rootKeyIdentifier;
    this.#rootCertBytes = certValues.rootCertBytes;
    this.#nextCertificateId = BigInt(certValues.nextCertificateId);
    const hasIcac = this.#isValidStoredIcacCertificate(certValues);
    if (requireIcac !== void 0 && requireIcac !== hasIcac) {
      throw new import_general.ImplementationError(
        `Stored credentials contain ICAC certificate: ${hasIcac}, but configuration expected it to be ${requireIcac}`
      );
    }
    if (hasIcac) {
      this.#icacProps = {
        certId: BigInt(certValues.icacCertId),
        keyPair: (0, import_general.PrivateKey)(certValues.icacKeyPair),
        keyIdentifier: certValues.icacKeyIdentifier,
        certBytes: certValues.icacCertBytes
      };
    } else {
      if (this.#rootKeyPair === void 0) {
        throw new import_general.ImplementationError(
          "rootKeyPair is required when not using ICAC (2-tier PKI requires RCAC private key to sign NOCs)"
        );
      }
    }
  }
  #buildStorageData() {
    return {
      rootCertId: this.#rootCertId,
      rootKeyIdentifier: this.#initializedRootKeyIdentifier,
      rootCertBytes: this.#initializedRootCertBytes,
      nextCertificateId: this.#nextCertificateId,
      ...this.#icacProps ? {
        rootKeyPair: this.#rootKeyPair?.keyPair,
        // rootKeyPair is optional when using ICAC
        icacCertId: this.#icacProps.certId,
        icacKeyPair: this.#icacProps.keyPair.keyPair,
        icacKeyIdentifier: this.#icacProps.keyIdentifier,
        icacCertBytes: this.#icacProps.certBytes
      } : {
        rootKeyPair: this.#initializedRootKeyPair.keyPair
      }
    };
  }
  #getSigningParameters() {
    if (this.#icacProps) {
      return {
        issuer: { icacId: this.#icacProps.certId },
        signingKey: this.#icacProps.keyPair,
        authorityKeyId: this.#icacProps.keyIdentifier
      };
    }
    return {
      issuer: { rcacId: this.#rootCertId },
      signingKey: this.#initializedRootKeyPair,
      authorityKeyId: this.#initializedRootKeyIdentifier
    };
  }
  get #initializedRootCertBytes() {
    if (this.#rootCertBytes === void 0) {
      throw new import_general.InternalError("Root certificate is not installed");
    }
    return this.#rootCertBytes;
  }
}
