// @matter/general/dist/cjs/storage/wal/WalTransaction.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WalTransaction_exports = {};
__export(WalTransaction_exports, {
  WalTransaction: () => WalTransaction,
  applyCommit: () => applyCommit
});
module.exports = __toCommonJS(WalTransaction_exports);
var import_StorageTransaction = require("../StorageTransaction.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class WalTransaction extends import_StorageTransaction.StorageTransaction {
  #ops = [];
  #writer;
  #onCommit;
  constructor(storage, writer, onCommit) {
    super(storage);
    this.#writer = writer;
    this.#onCommit = onCommit;
  }
  get(contexts, key) {
    const contextKey = contexts.join(".");
    for (let i = this.#ops.length - 1; i >= 0; i--) {
      const op = this.#ops[i];
      if (op.key !== contextKey) continue;
      if (op.op === "upd") {
        if (key in op.values) {
          return op.values[key];
        }
      } else if (op.op === "del") {
        if (!op.values || op.values.includes(key)) {
          return void 0;
        }
      }
    }
    return this.storage.get(contexts, key);
  }
  set(contexts, keyOrValues, value) {
    this.assertActive();
    const contextKey = contexts.join(".");
    const values = typeof keyOrValues === "string" ? { [keyOrValues]: value } : keyOrValues;
    this.#ops.push({ op: "upd", key: contextKey, values });
  }
  delete(contexts, key) {
    this.assertActive();
    const contextKey = contexts.join(".");
    this.#ops.push({ op: "del", key: contextKey, values: [key] });
  }
  clearAll(contexts) {
    this.assertActive();
    const contextKey = contexts.join(".");
    this.#ops.push({ op: "del", key: contextKey });
  }
  keys(contexts) {
    const contextKey = contexts.join(".");
    const result = /* @__PURE__ */ new Set();
    const baseKeys = this.storage.keys(contexts);
    const applyBuffered = (keys) => {
      for (const k of keys) {
        result.add(k);
      }
      for (const op of this.#ops) {
        if (op.key !== contextKey) continue;
        if (op.op === "upd") {
          for (const k of Object.keys(op.values)) {
            result.add(k);
          }
        } else if (op.op === "del") {
          if (op.values) {
            for (const k of op.values) {
              result.delete(k);
            }
          } else {
            result.clear();
          }
        }
      }
      return [...result];
    };
    if (baseKeys instanceof Promise) {
      return baseKeys.then(applyBuffered);
    }
    return applyBuffered(baseKeys);
  }
  values(contexts) {
    const baseValues = this.storage.values(contexts);
    const contextKey = contexts.join(".");
    const applyBuffered = (vals) => {
      const result = { ...vals };
      for (const op of this.#ops) {
        if (op.key !== contextKey) continue;
        if (op.op === "upd") {
          Object.assign(result, op.values);
        } else if (op.op === "del") {
          if (op.values) {
            for (const k of op.values) {
              delete result[k];
            }
          } else {
            for (const k of Object.keys(result)) {
              delete result[k];
            }
          }
        }
      }
      return result;
    };
    if (baseValues instanceof Promise) {
      return baseValues.then(applyBuffered);
    }
    return applyBuffered(baseValues);
  }
  contexts(contexts) {
    const baseContexts = this.storage.contexts(contexts);
    const contextKey = contexts.length ? contexts.join(".") : "";
    const prefix = contextKey.length ? `${contextKey}.` : "";
    const applyBuffered = (ctxs) => {
      const result = new Set(ctxs);
      for (const op of this.#ops) {
        if (op.op === "upd") {
          if (op.key.startsWith(prefix)) {
            const sub = op.key.substring(prefix.length).split(".");
            if (sub.length >= 1 && sub[0].length > 0) {
              result.add(sub[0]);
            }
          }
        } else if (op.op === "del" && !op.values) {
          if (op.key === contextKey) {
            result.clear();
          } else if (op.key.startsWith(prefix)) {
            const sub = op.key.substring(prefix.length).split(".");
            if (sub.length >= 1 && sub[0].length > 0) {
              result.delete(sub[0]);
            }
          }
        }
      }
      return [...result];
    };
    if (baseContexts instanceof Promise) {
      return baseContexts.then(applyBuffered);
    }
    return applyBuffered(baseContexts);
  }
  async commit() {
    this.assertActive();
    if (this.#ops.length > 0) {
      const ops = [...this.#ops];
      const { id, ts } = await this.#writer.write(ops);
      this.#onCommit(id, ts, ops);
    }
    super.commit();
  }
  rollback() {
    this.#ops.length = 0;
  }
}
function applyCommit(store, commit) {
  for (const op of commit.ops) {
    if (op.op === "upd") {
      if (!store[op.key]) {
        store[op.key] = {};
      }
      Object.assign(store[op.key], op.values);
    } else if (op.op === "del") {
      if (op.values) {
        if (store[op.key]) {
          for (const k of op.values) {
            delete store[op.key][k];
          }
        }
      } else if (op.key === "") {
        for (const k of Object.keys(store)) {
          delete store[k];
        }
      } else {
        delete store[op.key];
        const prefix = `${op.key}.`;
        for (const k of Object.keys(store)) {
          if (k.startsWith(prefix)) {
            delete store[k];
          }
        }
      }
    }
  }
}
