// @matter/protocol/dist/esm/advertisement/ServiceDescription.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ServiceDescription;
((ServiceDescription2) => {
  function isCommissioning(description) {
    return description.kind === "commissionable" || description.kind === "commissioner";
  }
  ServiceDescription2.isCommissioning = isCommissioning;
  function isOperational(description) {
    return description.kind === "operational";
  }
  ServiceDescription2.isOperational = isOperational;
  function Commissionable(definition) {
    return {
      ...definition,
      kind: "commissionable"
    };
  }
  ServiceDescription2.Commissionable = Commissionable;
  function Operational(definition) {
    return {
      ...definition,
      kind: "operational"
    };
  }
  ServiceDescription2.Operational = Operational;
  function Commissioner(definition) {
    return {
      ...definition,
      kind: "commissioner"
    };
  }
  ServiceDescription2.Commissioner = Commissioner;
})(ServiceDescription || (ServiceDescription = {}));
export {
  ServiceDescription
};
