// @matter/protocol/dist/cjs/advertisement/mdns/MdnsAdvertisement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MdnsAdvertisement_exports = {};
__export(MdnsAdvertisement_exports, {
  MdnsAdvertisement: () => MdnsAdvertisement
});
module.exports = __toCommonJS(MdnsAdvertisement_exports);
var import_Advertisement = require("#advertisement/Advertisement.js");
var import_SupportedTransportsBitmap = require("#common/SupportedTransportsBitmap.js");
var import_SessionIntervals = require("#session/SessionIntervals.js");
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_general.Logger.get("MdnsAdvertisement");
class MdnsAdvertisement extends import_Advertisement.Advertisement {
  /**
   * The device qname.
   */
  qname;
  #isPrivacyMasked;
  #stopAt;
  #broadcastPromise;
  constructor(advertiser, qname, description) {
    description = {
      ...description,
      ...(0, import_SessionIntervals.SessionIntervals)(description)
    };
    super(advertiser, `mdns:${qname}`, description, { omitPrivateDetails: advertiser.omitPrivateDetails });
    this.qname = qname;
    this.#isPrivacyMasked = this.isPrivacyMasked;
  }
  get txtValues() {
    return {};
  }
  async run(context, event = "startup") {
    this.#stopAt = void 0;
    let number = 0;
    for (const retryInterval of this.advertiser.broadcastScheduleFor(this, event)) {
      if (this.#stopAt !== void 0 && this.#stopAt <= import_general.Time.nowMs) {
        break;
      }
      number++;
      logger.debug("Broadcast", this.dict({ number, next: import_general.Duration.format(retryInterval) }));
      await this.broadcast();
      await context.sleep("MDNS repeat", retryInterval);
      if (context.cancelled) {
        break;
      }
    }
  }
  /**
   * Broadcast a single announcement immediately.
   */
  async broadcast() {
    if (this.#broadcastPromise !== void 0) {
      return this.#broadcastPromise;
    }
    this.#broadcastPromise = this.#broadcast().then(() => {
      this.#broadcastPromise = void 0;
    });
    return this.#broadcastPromise;
  }
  async #broadcast() {
    if (!this.#isPrivacyMasked && this.isPrivacyMasked) {
      this.#isPrivacyMasked = true;
      await this.advertiser.server.setRecordsGenerator(this.service, this.#recordsGenerator);
    }
    await this.advertiser.server.broadcast(this.service);
  }
  /**
   * Broadcast expiration announcement immediately.
   */
  async expire() {
    if (this.#broadcastPromise !== void 0) {
      await this.#broadcastPromise;
    }
    logger.info("Unpublishing", this.dict({ time: import_general.Duration.format(this.duration) }));
    await this.advertiser.server.expireAnnouncements(this.service);
  }
  serviceConnected() {
    const { broadcastAfterConnection } = this.advertiser.broadcastConfigFor(this);
    if (broadcastAfterConnection === void 0) {
      return;
    }
    if (broadcastAfterConnection <= 0) {
      this.stop();
      return;
    }
    this.#stopAt = (0, import_general.Timestamp)(import_general.Time.nowMs + broadcastAfterConnection);
  }
  serviceDisconnected() {
    this.start("reconnect");
  }
  async onCreate() {
    await Promise.resolve().then(() => {
      logger.info("Publishing", this.dict());
      return this.advertiser.server.setRecordsGenerator(this.service, this.#recordsGenerator);
    });
  }
  async onClose() {
    await this.expire();
  }
  get #recordsGenerator() {
    return (_intf, addrs) => this.#recordsFor(addrs);
  }
  #recordsFor(addrs) {
    const hostname = addrs.mac.replace(/:/g, "").toUpperCase() + "0000.local";
    const records = [
      ...this.ptrRecords,
      (0, import_general.SrvRecord)(this.qname, { priority: 0, weight: 0, port: this.advertiser.port, target: hostname }),
      (0, import_general.TxtRecord)(
        this.qname,
        Object.entries(this.#txtValues).filter(([, v]) => v !== void 0).map(([k, v]) => `${k}=${v}`)
      )
    ];
    for (const addr of addrs.ipV6) {
      records.push((0, import_general.AAAARecord)(hostname, addr));
    }
    if (this.advertiser.server.supportsIpv4) {
      for (const addr of addrs.ipV4) {
        records.push((0, import_general.ARecord)(hostname, addr));
      }
    }
    return records;
  }
  get #txtValues() {
    const values = {
      SII: this.description.idleInterval,
      SAI: this.description.activeInterval,
      SAT: this.description.activeThreshold,
      ...this.txtValues
    };
    if (this.description.tcp !== void 0) {
      values.T = import_SupportedTransportsBitmap.SupportedTransportsSchema.encode(this.description.tcp);
    }
    if (this.description.icd !== void 0) {
      values.ICD = this.description.icd;
    }
    return values;
  }
}
