// @matter/protocol/dist/esm/bdx/schema/BdxInitMessagesSchema.js
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { DataReader, DataWriter, Endian } from "@matter/general";
import { BdxStatusCode, BitField, BitFlag, BitmapSchema, Schema } from "@matter/types";
import { BdxError } from "../BdxError.js";
const BDX_VERSION = 0;
const BdxTransferControlBitmap = {
  /** Protocol Version, Only allowed version is 0.0 for now */
  version: BitField(0, 4),
  /** Indicates if Sender drive is supported by Initiator */
  senderDrive: BitFlag(4),
  /** Indicates if Receiver drive is supported by Initiator */
  receiverDrive: BitFlag(5),
  /**
   * Indicates if Asynchronous Transfer is supported by Initiator
   * Synchronous mode is always implicitly supported.
   */
  asynchronousTransfer: BitFlag(6)
};
const BdxTransferControlSchema = BitmapSchema(BdxTransferControlBitmap);
const BdxRangeControlBitmap = {
  /** Indicates if the transfer has a definite length */
  definiteLength: BitFlag(0),
  /** Indicates if the transfer has a start offset */
  startOffset: BitFlag(1),
  /**
   * Indicates if wide (64-bit) range enabled for values
   * Offset (STARTFOFS) and length (DEFLEN) are 4 octets (32-bit) little-endian unsigned quantities.
   */
  wideRange: BitFlag(4)
};
const BdxRangeControlSchema = BitmapSchema(BdxRangeControlBitmap);
class BdxInitMessageSchema extends Schema {
  encodeInternal(message) {
    this.validate(message);
    const { transferProtocol, maxBlockSize, startOffset, maxLength, fileDesignator, metaData } = message;
    const rangeControl = {
      startOffset: startOffset !== void 0,
      definiteLength: maxLength !== void 0 && maxLength !== 0,
      wideRange: typeof startOffset === "bigint" || typeof maxLength === "bigint"
    };
    const writer = new DataWriter(Endian.Little);
    writer.writeUInt8(BdxTransferControlSchema.encode(transferProtocol));
    writer.writeUInt8(BdxRangeControlSchema.encode(rangeControl));
    writer.writeUInt16(maxBlockSize);
    if (startOffset !== void 0) {
      if (rangeControl.wideRange) {
        writer.writeUInt64(startOffset);
      } else {
        writer.writeUInt32(startOffset);
      }
    }
    if (maxLength !== void 0 && maxLength !== 0) {
      if (rangeControl.wideRange) {
        writer.writeUInt64(maxLength);
      } else {
        writer.writeUInt32(maxLength);
      }
    }
    writer.writeUInt16(fileDesignator.byteLength);
    writer.writeByteArray(fileDesignator);
    if (metaData && metaData.byteLength > 0) {
      writer.writeByteArray(metaData);
    }
    return writer.toByteArray();
  }
  decodeInternal(bytes) {
    const reader = new DataReader(bytes, Endian.Little);
    const transferProtocol = BdxTransferControlSchema.decode(reader.readUInt8());
    const rangeControl = BdxRangeControlSchema.decode(reader.readUInt8());
    const maxBlockSize = reader.readUInt16();
    const startOffset = rangeControl.startOffset ? rangeControl.wideRange ? reader.readUInt64() : reader.readUInt32() : void 0;
    let maxLength = rangeControl.definiteLength ? rangeControl.wideRange ? reader.readUInt64() : reader.readUInt32() : void 0;
    if (maxLength === 0) {
      maxLength = void 0;
    }
    const fileDesignatorLength = reader.readUInt16();
    const fileDesignator = reader.readByteArray(fileDesignatorLength);
    const metaData = reader.remainingBytesCount > 0 ? reader.remainingBytes : void 0;
    const result = {
      transferProtocol,
      maxBlockSize,
      startOffset,
      maxLength,
      fileDesignator,
      metaData
    };
    this.validate(result);
    return result;
  }
  validate({ transferProtocol, maxBlockSize, fileDesignator }) {
    if (transferProtocol.version !== BDX_VERSION) {
      throw new BdxError(
        `Unsupported BDX version: ${transferProtocol.version}`,
        BdxStatusCode.VersionNotSupported
      );
    }
    if (!transferProtocol.senderDrive && !transferProtocol.receiverDrive) {
      throw new BdxError(
        "BDX transfer protocol needs to allow either sender or receiver drive",
        BdxStatusCode.BadMessageContent
      );
    }
    if (fileDesignator.byteLength === 0) {
      throw new BdxError("File Designator cannot be empty", BdxStatusCode.BadMessageContent);
    }
    if (fileDesignator.byteLength > 65535) {
      throw new BdxError(
        "File Designator length exceeds maximum allowed size of 65535bytes",
        BdxStatusCode.BadMessageContent
      );
    }
    if (maxBlockSize < 1 || maxBlockSize > 65535) {
      throw new BdxError("Max Block Size must be between 1 and 65535bytes", BdxStatusCode.BadMessageContent);
    }
  }
}
const BdxSendInitMessage = new BdxInitMessageSchema();
const BdxReceiveInitMessage = BdxSendInitMessage;
export {
  BDX_VERSION,
  BdxInitMessageSchema,
  BdxRangeControlBitmap,
  BdxRangeControlSchema,
  BdxReceiveInitMessage,
  BdxSendInitMessage,
  BdxTransferControlBitmap,
  BdxTransferControlSchema
};
