// @matter/protocol/dist/cjs/bdx/flow/Flow.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Flow_exports = {};
__export(Flow_exports, {
  Flow: () => Flow
});
module.exports = __toCommonJS(Flow_exports);
var import_general = require("@matter/general");
var import_types = require("@matter/types");
var import_BdxError = require("../BdxError.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Flow {
  #transferParameters;
  #messenger;
  #isClosed = false;
  #blockCounter = -1;
  #bytesLeft;
  #finalBlockCounter;
  progressInfo = (0, import_general.Observable)();
  progressFinished = (0, import_general.Observable)();
  progressCancelled = (0, import_general.Observable)();
  #dataLength;
  #transferredBytes = 0;
  constructor(messenger, transferParameters) {
    this.#messenger = messenger;
    this.#transferParameters = transferParameters;
    this.#bytesLeft = transferParameters.dataLength;
    this.#dataLength = transferParameters.dataLength;
  }
  set dataLength(value) {
    this.#dataLength = value;
  }
  get dataLength() {
    return this.#dataLength;
  }
  set transferredBytes(value) {
    this.#transferredBytes = value;
  }
  get transferredBytes() {
    return this.#transferredBytes;
  }
  get transferParameters() {
    return this.#transferParameters;
  }
  get isClosed() {
    return this.#isClosed;
  }
  get messenger() {
    return this.#messenger;
  }
  get bytesLeft() {
    return this.#bytesLeft;
  }
  set bytesLeft(value) {
    this.#bytesLeft = value;
  }
  set finalBlockCounter(blockCounter) {
    if (this.#finalBlockCounter !== void 0) {
      throw new import_general.InternalError("Transfer already finalized. finalizeTransfer() should only be called once.");
    }
    this.#finalBlockCounter = blockCounter;
  }
  get finalBlockCounter() {
    if (this.#finalBlockCounter === void 0) {
      throw new import_general.InternalError("Transfer not finalized. Call finalizeTransfer() after completing the transfer.");
    }
    return this.#finalBlockCounter;
  }
  /** Determines the next message counter to use for the next message, also handles wrapping around at 2^32. */
  get nextMessageCounter() {
    this.#blockCounter = (this.#blockCounter + 1) % 4294967296;
    return this.#blockCounter;
  }
  /** Utility function to check if a block counter is valid */
  validateCounter(messageBlockCounter, expectedCounter = this.nextMessageCounter) {
    if (messageBlockCounter !== expectedCounter) {
      throw new import_BdxError.BdxError(
        `Received Block with unexpected block counter: ${messageBlockCounter}, expected: ${expectedCounter}`,
        import_types.BdxStatusCode.BadBlockCounter
      );
    }
    return messageBlockCounter;
  }
  async close(_error) {
    this.#isClosed = true;
  }
  /**
   * Main logic method to execute the flow.
   * The promise resolves when the flow is complete, or rejects on any error or unexpected conditions.
   */
  async processTransfer() {
    await this.initTransfer();
    this.progressInfo.emit(0, this.dataLength);
    try {
      while (!this.isClosed) {
        if (await this.transferNextChunk()) {
          break;
        }
        this.progressInfo.emit(this.transferredBytes, this.dataLength);
      }
      this.progressInfo.emit(this.transferredBytes, this.dataLength);
      if (this.isClosed) {
        this.progressCancelled.emit();
      } else {
        await this.finalizeTransfer();
        this.progressFinished.emit(this.transferredBytes);
      }
    } catch (error) {
      this.progressCancelled.emit();
      throw error;
    }
  }
}
((Flow2) => {
  let DriverMode;
  ((DriverMode2) => {
    DriverMode2["SenderDrive"] = "senderDrive";
    DriverMode2["ReceiverDrive"] = "receiverDrive";
  })(DriverMode = Flow2.DriverMode || (Flow2.DriverMode = {}));
})(Flow || (Flow = {}));
