// homebridge/dist/matter/accessoryCache.js
import { mkdir, readFile, rename, rm, stat, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { Logger } from '../logger.js';
const log = Logger.withPrefix('Matter/Accessories');
export class MatterAccessoryCache {
    cacheFilePath;
    cachedAccessories = new Map();
    cacheLoaded = false;
    saveQueue = Promise.resolve();
    directoryEnsured = false;
    saveDebounceTimer = null;
    SAVE_DEBOUNCE_MS = 2000;
    constructor(storagePath, bridgeId) {
        this.cacheFilePath = join(storagePath, bridgeId, 'accessories.json');
    }
    async load() {
        if (this.cacheLoaded) {
            return this.cachedAccessories;
        }
        try {
            try {
                await stat(this.cacheFilePath);
            }
            catch {
                log.info('No cached Matter accessories found (first run)');
                this.cacheLoaded = true;
                return this.cachedAccessories;
            }
            const cacheData = JSON.parse(await readFile(this.cacheFilePath, 'utf-8'));
            if (!Array.isArray(cacheData)) {
                throw new TypeError('Cache file does not contain an array');
            }
            for (const serialized of cacheData) {
                if (serialized.uuid) {
                    this.cachedAccessories.set(serialized.uuid, serialized);
                }
            }
            log.info(`Loaded ${this.cachedAccessories.size} cached Matter accessories`);
            this.directoryEnsured = true;
            this.cacheLoaded = true;
            return this.cachedAccessories;
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            log.error(`Failed to load Matter accessory cache from ${this.cacheFilePath}: ${errorMessage}`);
            log.warn('Deleting corrupted cache file and starting fresh');
            try {
                await rm(this.cacheFilePath, { force: true });
            }
            catch (removeError) {
                log.debug('Could not delete corrupted cache file:', removeError);
            }
            this.cacheLoaded = true;
            return this.cachedAccessories;
        }
    }
    requestSave(accessories) {
        if (this.saveDebounceTimer) {
            clearTimeout(this.saveDebounceTimer);
        }
        this.saveDebounceTimer = setTimeout(() => {
            this.save(accessories).catch((error) => {
                log.error('Debounced cache save failed:', error);
            });
        }, this.SAVE_DEBOUNCE_MS);
    }
    async save(accessories) {
        if (this.saveDebounceTimer) {
            clearTimeout(this.saveDebounceTimer);
            this.saveDebounceTimer = null;
        }
        this.saveQueue = this.saveQueue.then(() => this.performSave(accessories));
        await this.saveQueue;
    }
    async performSave(accessories) {
        const tempFilePath = `${this.cacheFilePath}.tmp`;
        try {
            const serialized = Array.from(accessories.values(), accessory => this.serializeAccessory(accessory));
            if (!this.directoryEnsured) {
                const directory = dirname(this.cacheFilePath);
                await mkdir(directory, { recursive: true });
                this.directoryEnsured = true;
                log.debug(`Cache directory ensured: ${directory}`);
            }
            await writeFile(tempFilePath, JSON.stringify(serialized, null, 2), 'utf-8');
            await rename(tempFilePath, this.cacheFilePath);
            log.debug(`Saved ${serialized.length} Matter accessor${serialized.length === 1 ? 'y' : 'ies'} to cache`);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            log.error(`Failed to save Matter accessory cache: ${errorMessage}`);
            try {
                await rm(tempFilePath, { force: true });
            }
            catch (cleanupError) {
                log.debug('Could not clean up temporary cache file:', cleanupError);
            }
        }
    }
    getCached(uuid) {
        return this.cachedAccessories.get(uuid);
    }
    hasCached(uuid) {
        return this.cachedAccessories.has(uuid);
    }
    removeCached(uuid) {
        this.cachedAccessories.delete(uuid);
    }
    getAllCached() {
        return new Map(this.cachedAccessories);
    }
    serializeAccessory(accessory) {
        const deviceType = accessory.deviceType;
        const deviceTypeInfo = {
            name: deviceType?.name,
            code: deviceType?.code,
        };
        let serializedParts;
        if (accessory.parts && accessory.parts.length > 0) {
            serializedParts = accessory.parts.map((part) => {
                const partDeviceType = part.deviceType;
                return {
                    id: part.id,
                    displayName: part.displayName,
                    deviceType: {
                        name: partDeviceType?.name,
                        code: partDeviceType?.code,
                    },
                    clusters: structuredClone(part.clusters),
                };
            });
        }
        return {
            plugin: accessory._associatedPlugin || '',
            platform: accessory._associatedPlatform || '',
            uuid: accessory.UUID,
            displayName: accessory.displayName,
            deviceType: deviceTypeInfo,
            serialNumber: accessory.serialNumber,
            manufacturer: accessory.manufacturer,
            model: accessory.model,
            firmwareRevision: accessory.firmwareRevision,
            hardwareRevision: accessory.hardwareRevision,
            softwareVersion: accessory.softwareVersion,
            context: accessory.context,
            clusters: structuredClone(accessory.clusters),
            parts: serializedParts,
        };
    }
}
