// homebridge/dist/matter/BaseMatterManager.js
import { rmSync } from 'node:fs';
import path from 'node:path';
import { Logger } from '../logger.js';
import { User } from '../user.js';
import { generate } from '../util/mac.js';
import { mapAttributesToCommand } from './ClusterCommandMapper.js';
const log = Logger.withPrefix('Matter/BaseManager');
const COLON_RE = /:/g;
export class BaseMatterManager {
    matterServer;
    externalMatterServers = new Map();
    pluginManager;
    constructor(pluginManager) {
        this.pluginManager = pluginManager;
    }
    getExternalServer(uuid) {
        return this.externalMatterServers.get(uuid);
    }
    async handleTriggerCommand(uuid, cluster, attributes, partId) {
        const commandMapping = mapAttributesToCommand(cluster, attributes);
        log.debug(`handleTriggerCommand: uuid=${uuid}, cluster=${cluster}, partId=${partId}`);
        log.debug(`Attributes: ${JSON.stringify(attributes)}`);
        log.debug(`Command mapping: ${commandMapping ? JSON.stringify(commandMapping) : 'null (state-only update)'}`);
        log.debug(`External servers count: ${this.externalMatterServers.size}`);
        if (this.externalMatterServers.size > 0) {
            log.debug(`External server UUIDs: ${[...this.externalMatterServers.keys()].join(', ')}`);
        }
        const externalServer = this.externalMatterServers.get(uuid);
        if (externalServer) {
            log.debug(`Found external server for ${uuid}`);
            if (commandMapping) {
                await externalServer.triggerCommand(uuid, cluster, commandMapping.command, commandMapping.args, partId);
                this.notifyCurrentState(externalServer, uuid, cluster, partId);
            }
            else {
                await externalServer.updateAccessoryState(uuid, cluster, attributes, partId);
            }
            return;
        }
        if (!this.matterServer) {
            log.debug(`No matterServer and external server not found for ${uuid}`);
            throw new Error(`Accessory ${uuid} not found on this bridge`);
        }
        log.debug(`Trying matterServer for ${uuid}`);
        if (commandMapping) {
            await this.matterServer.triggerCommand(uuid, cluster, commandMapping.command, commandMapping.args, partId);
            this.notifyCurrentState(this.matterServer, uuid, cluster, partId);
        }
        else {
            await this.matterServer.updateAccessoryState(uuid, cluster, attributes, partId);
        }
    }
    async handleUpdateAccessoryState(uuid, cluster, attributes, partId) {
        const externalServer = this.externalMatterServers.get(uuid);
        if (externalServer) {
            await externalServer.updateAccessoryState(uuid, cluster, attributes, partId);
            return;
        }
        if (!this.matterServer) {
            throw new Error(`Accessory ${uuid} not found on this bridge`);
        }
        await this.matterServer.updateAccessoryState(uuid, cluster, attributes, partId);
    }
    enableStateMonitoring() {
        this.matterServer?.enableStateMonitoring();
        for (const externalServer of this.externalMatterServers.values()) {
            externalServer.enableStateMonitoring();
        }
    }
    notifyCurrentState(server, uuid, cluster, partId) {
        const currentState = server.getAccessoryState(uuid, cluster, partId);
        if (currentState) {
            server.notifyStateChange(uuid, cluster, currentState, partId);
        }
    }
    disableStateMonitoring() {
        this.matterServer?.disableStateMonitoring();
        for (const externalServer of this.externalMatterServers.values()) {
            externalServer.disableStateMonitoring();
        }
    }
    restoreCachedAccessories(keepOrphaned) {
        if (!this.matterServer) {
            return;
        }
        const cachedAccessories = this.matterServer.getAllCachedAccessories();
        log.debug(`Restoring ${cachedAccessories.length} cached Matter accessories`);
        for (const cachedAccessory of cachedAccessories) {
            let plugin = this.pluginManager.getPlugin(cachedAccessory.plugin);
            if (!plugin) {
                try {
                    plugin = this.pluginManager.getPluginByActiveDynamicPlatform(cachedAccessory.platform);
                    if (plugin) {
                        log.info(`When searching for the associated plugin of the Matter accessory '${cachedAccessory.displayName}' `
                            + `it seems like the plugin name changed from '${cachedAccessory.plugin}' to '${plugin.getPluginIdentifier()}'. Plugin association is now being transformed!`);
                    }
                }
                catch (error) {
                    const errorMessage = error instanceof Error ? error.message : String(error);
                    log.warn(`Could not find the associated plugin for the Matter accessory '${cachedAccessory.displayName}'. `
                        + `Tried to find the plugin by the platform name but ${errorMessage}`);
                }
            }
            const platformPlugin = plugin && plugin.getActiveDynamicPlatform(cachedAccessory.platform);
            if (!platformPlugin) {
                log.warn(`Failed to find plugin to handle Matter accessory ${cachedAccessory.displayName} (plugin: ${cachedAccessory.plugin}, platform: ${cachedAccessory.platform})`);
                if (!keepOrphaned) {
                    log.info(`Removing orphaned Matter accessory ${cachedAccessory.displayName}`);
                    this.matterServer.unregisterAccessory(cachedAccessory.uuid).catch((error) => {
                        log.warn(`Failed to unregister orphaned Matter accessory ${cachedAccessory.displayName}:`, error);
                    });
                }
            }
            else {
                if (platformPlugin.configureMatterAccessory) {
                    log.debug(`Calling configureMatterAccessory for ${cachedAccessory.displayName}`);
                    const accessory = this.deserializeMatterAccessory(cachedAccessory);
                    platformPlugin.configureMatterAccessory(accessory);
                }
                else {
                    log.debug(`Platform ${cachedAccessory.platform} does not implement configureMatterAccessory`);
                }
            }
        }
    }
    async handleRegisterPlatformAccessories(pluginIdentifier, platformName, accessories) {
        if (!this.matterServer) {
            log.warn('Cannot register Matter accessories - Matter server is not running');
            return;
        }
        await this.matterServer.registerPlatformAccessories(pluginIdentifier, platformName, accessories);
    }
    async handleUpdatePlatformAccessories(accessories) {
        const bridgeAccessories = [];
        for (const accessory of accessories) {
            const externalServer = this.externalMatterServers.get(accessory.UUID);
            if (externalServer) {
                await externalServer.updatePlatformAccessories([accessory]);
            }
            else {
                bridgeAccessories.push(accessory);
            }
        }
        if (bridgeAccessories.length > 0) {
            if (!this.matterServer) {
                log.warn('Cannot update Matter platform accessories - Matter server is not running');
                return;
            }
            await this.matterServer.updatePlatformAccessories(bridgeAccessories);
        }
    }
    async handleUnregisterPlatformAccessories(pluginIdentifier, platformName, accessories) {
        if (!this.matterServer) {
            log.warn('Cannot unregister Matter accessories - Matter server is not running');
            return;
        }
        await this.matterServer.unregisterPlatformAccessories(pluginIdentifier, platformName, accessories);
    }
    async handleUnregisterExternalAccessories(accessories) {
        log.info(`Unregistering ${accessories.length} external Matter accessor${accessories.length === 1 ? 'y' : 'ies'}`);
        for (const accessory of accessories) {
            try {
                const matterServer = this.externalMatterServers.get(accessory.UUID);
                if (!matterServer) {
                    log.warn(`External Matter accessory ${accessory.displayName} (${accessory.UUID}) is not registered`);
                    continue;
                }
                log.info(`Stopping external Matter server for ${accessory.displayName}`);
                await matterServer.stop();
                this.externalMatterServers.delete(accessory.UUID);
                const advertiseAddress = generate(accessory.UUID);
                const uniqueId = advertiseAddress.replace(COLON_RE, '');
                const storagePath = path.join(User.matterPath(), uniqueId);
                try {
                    log.debug(`Removing Matter storage for external accessory at: ${storagePath}`);
                    rmSync(storagePath, { recursive: true, force: true });
                    log.info(`✓ Cleaned up storage for external Matter accessory: ${accessory.displayName}`);
                }
                catch (error) {
                    log.error(`Failed to clean up storage for external Matter accessory ${accessory.displayName}:`, error);
                }
                log.info(`✓ External Matter accessory unregistered: ${accessory.displayName}`);
            }
            catch (error) {
                log.error(`Failed to unregister external Matter accessory ${accessory.displayName}:`, error);
            }
        }
    }
    deserializeMatterAccessory(serialized) {
        return {
            UUID: serialized.uuid,
            displayName: serialized.displayName,
            deviceType: serialized.deviceType,
            serialNumber: serialized.serialNumber,
            manufacturer: serialized.manufacturer,
            model: serialized.model,
            firmwareRevision: serialized.firmwareRevision,
            hardwareRevision: serialized.hardwareRevision,
            softwareVersion: serialized.softwareVersion,
            context: serialized.context ?? {},
            clusters: serialized.clusters,
            parts: serialized.parts,
        };
    }
}
