// homebridge/dist/matter/ChildBridgeMatterManager.js
import process from 'node:process';
import { DEFAULT_BRIDGE_DEFAULTS } from '../bridgeService.js';
import { Logger } from '../logger.js';
import { User } from '../user.js';
import getVersion from '../version.js';
import { BaseMatterManager } from './BaseMatterManager.js';
import { publishExternalMatterAccessory } from './ExternalMatterAccessoryPublisher.js';
import { MatterServer } from './server.js';
import { appendUsernameSuffix, normalizeBindConfig } from './utils.js';
const log = Logger.withPrefix('Matter/ChildManager');
const COLON_RE = /:/g;
export class ChildBridgeMatterManager extends BaseMatterManager {
    bridgeConfig;
    bridgeOptions;
    api;
    externalPortService;
    matterConfig;
    matterSerialNumber;
    constructor(bridgeConfig, bridgeOptions, api, externalPortService, pluginManager) {
        super(pluginManager);
        this.bridgeConfig = bridgeConfig;
        this.bridgeOptions = bridgeOptions;
        this.api = api;
        this.externalPortService = externalPortService;
        this.matterConfig = bridgeConfig.matter;
    }
    async initialize(onCommissioningChanged) {
        if (!this.matterConfig) {
            return;
        }
        log.debug(`Child bridge ${this.bridgeConfig.username} has Matter config (Combined HAP+Matter), starting Matter server`);
        if (!this.matterConfig.port) {
            const matterUsername = appendUsernameSuffix(this.bridgeConfig.username, 'MATTER');
            const matterPort = await this.externalPortService.requestPort(matterUsername);
            if (!matterPort) {
                throw new Error('Failed to allocate Matter port for child bridge. '
                    + 'Please specify a port manually in the _bridge.matter configuration, or free up ports in the configured range.');
            }
            this.matterConfig.port = matterPort;
            log.debug(`Allocated Matter port: ${this.matterConfig.port} (HAP port: ${this.bridgeConfig.port})`);
        }
        await this.startMatterServer(this.matterConfig);
        if (onCommissioningChanged && this.matterServer) {
            this.matterServer.on('commissioning-status-changed', (commissioned, fabricCount) => {
                log.info(`Matter commissioning status changed for child bridge ${this.bridgeConfig.username}: commissioned=${commissioned}, fabricCount=${fabricCount}`);
                onCommissioningChanged();
            });
        }
        if (this.matterServer) {
            this.matterServer.on('stateChange', ({ uuid, cluster, state, partId }) => {
                if (process.send) {
                    process.send({
                        id: 'matterEvent',
                        data: {
                            type: 'accessoryUpdate',
                            data: {
                                uuid,
                                cluster,
                                state,
                                partId,
                            },
                        },
                    });
                }
            });
        }
    }
    async startMatterServer(matterConfig) {
        log.info(`Starting Matter server for child bridge ${this.bridgeConfig.username}`);
        const serialNumber = this.bridgeConfig.username.replace(COLON_RE, '');
        const networkInterfaces = normalizeBindConfig(this.bridgeConfig.bind);
        this.matterServer = new MatterServer({
            port: matterConfig.port || 5540,
            uniqueId: serialNumber,
            storagePath: User.matterPath(),
            displayName: this.bridgeConfig.name || 'Child Bridge',
            debugModeEnabled: this.bridgeOptions.debugModeEnabled,
            manufacturer: this.bridgeConfig.manufacturer || DEFAULT_BRIDGE_DEFAULTS.manufacturer,
            model: this.bridgeConfig.model || DEFAULT_BRIDGE_DEFAULTS.model,
            firmwareRevision: this.bridgeConfig.firmwareRevision || getVersion(),
            serialNumber,
            networkInterfaces,
        });
        await this.matterServer.start();
        this.api._setMatterEnabled(true);
        this.api._setMatterServer(this.matterServer);
        const commissioningInfo = this.matterServer.getCommissioningInfo();
        log.info(`Matter server started for child bridge ${this.bridgeConfig.username} with commissioning info:`, commissioningInfo);
        this.matterSerialNumber = commissioningInfo.serialNumber;
        this.setupEventListeners();
    }
    setupEventListeners() {
        this.api.on("publishExternalMatterAccessories", (accessories, registrationId) => {
            this.handlePublishExternalAccessories(accessories, registrationId).catch((error) => {
                log.error('Failed to publish external Matter accessories:', error);
                this.api._resolveExternalRegistration(registrationId);
            });
        });
        this.api.on("registerMatterPlatformAccessories", (pluginIdentifier, platformName, accessories) => {
            this.handleRegisterPlatformAccessories(pluginIdentifier, platformName, accessories).catch((error) => {
                log.error(`Failed to register Matter accessories for ${pluginIdentifier}:`, error);
            });
        });
        this.api.on("updateMatterPlatformAccessories", (accessories) => {
            this.handleUpdatePlatformAccessories(accessories).catch((error) => {
                log.error('Failed to update Matter platform accessories:', error);
            });
        });
        this.api.on("unregisterMatterPlatformAccessories", (pluginIdentifier, platformName, accessories) => {
            this.handleUnregisterPlatformAccessories(pluginIdentifier, platformName, accessories).catch((error) => {
                log.error(`Failed to unregister Matter accessories for ${pluginIdentifier}:`, error);
            });
        });
        this.api.on("unregisterExternalMatterAccessories", (accessories) => {
            this.handleUnregisterExternalAccessories(accessories).catch((error) => {
                log.error('Failed to unregister external Matter accessories:', error);
            });
        });
        this.api.on("updateMatterAccessoryState", (uuid, cluster, attributes, partId) => {
            this.handleUpdateAccessoryState(uuid, cluster, attributes, partId).catch((error) => {
                log.error(`Failed to update Matter accessory state for ${uuid}:`, error);
            });
        });
    }
    async handlePublishExternalAccessories(accessories, registrationId) {
        log.info(`Publishing ${accessories.length} external Matter accessor${accessories.length === 1 ? 'y' : 'ies'} from child bridge ${this.bridgeConfig.username}`);
        try {
            const networkInterfaces = normalizeBindConfig(this.bridgeConfig.bind);
            for (const accessory of accessories) {
                try {
                    if (this.externalMatterServers.has(accessory.UUID)) {
                        log.warn(`External Matter accessory ${accessory.displayName} (${accessory.UUID}) is already published`);
                        continue;
                    }
                    const result = await publishExternalMatterAccessory(accessory, {
                        portService: this.externalPortService,
                        networkInterfaces,
                        debugModeEnabled: this.bridgeOptions.debugModeEnabled,
                    });
                    if (!result) {
                        continue;
                    }
                    this.externalMatterServers.set(accessory.UUID, result.server);
                    result.server.on('stateChange', ({ uuid, cluster, state, partId }) => {
                        if (process.send) {
                            process.send({
                                id: 'matterEvent',
                                data: {
                                    type: 'accessoryUpdate',
                                    data: {
                                        uuid,
                                        cluster,
                                        state,
                                        partId,
                                    },
                                },
                            });
                        }
                    });
                    if (process.send) {
                        process.send({
                            id: 'matterEvent',
                            data: {
                                type: 'externalBridgeRegistration',
                                data: {
                                    externalBridgeUsername: result.username,
                                    childBridgeUsername: this.bridgeConfig.username,
                                },
                            },
                        });
                    }
                    if (result.commissioningInfo.qrCode && result.commissioningInfo.manualPairingCode) {
                        log.info(`📱 Commissioning codes for ${accessory.displayName}:`);
                        log.info(`   QR Code: ${result.commissioningInfo.qrCode}`);
                        log.info(`   Manual Code: ${result.commissioningInfo.manualPairingCode}`);
                    }
                }
                catch (error) {
                    log.error(`Failed to publish external Matter accessory ${accessory.displayName}:`, error);
                }
            }
        }
        finally {
            this.api._resolveExternalRegistration(registrationId);
        }
    }
    getMatterStatusInfo() {
        if (!this.matterConfig || !this.matterServer) {
            return undefined;
        }
        const commissioningInfo = this.matterServer.getCommissioningInfo();
        return {
            qrCode: commissioningInfo.qrCode,
            manualPairingCode: commissioningInfo.manualPairingCode,
            serialNumber: this.matterSerialNumber || commissioningInfo.serialNumber,
            commissioned: commissioningInfo.commissioned || false,
            deviceCount: this.matterServer.getAccessories().length,
        };
    }
    isMatterEnabled() {
        return this.matterServer !== undefined;
    }
    enableStateMonitoring() {
        log.debug(`Enabling Matter state monitoring for child bridge ${this.bridgeConfig.username}`);
        super.enableStateMonitoring();
    }
    disableStateMonitoring() {
        log.debug(`Disabling Matter state monitoring for child bridge ${this.bridgeConfig.username}`);
        super.disableStateMonitoring();
    }
    collectAllAccessories() {
        const accessories = [];
        if (this.matterServer) {
            const serverAccessories = this.matterServer.collectAccessories(this.bridgeConfig.username, 'child', this.bridgeConfig.name || 'Child Bridge');
            accessories.push(...serverAccessories);
        }
        for (const server of this.externalMatterServers.values()) {
            const externalAccessories = server.collectAccessories(server.username, 'external', server.bridgeName);
            accessories.push(...externalAccessories);
        }
        return accessories;
    }
    getAccessoryInfo(uuid) {
        if (this.matterServer) {
            const info = this.matterServer.getAccessoryInfo(uuid);
            if (info) {
                return info;
            }
        }
        for (const server of this.externalMatterServers.values()) {
            const info = server.getAccessoryInfo(uuid);
            if (info) {
                return info;
            }
        }
        return undefined;
    }
    async teardown() {
        if (this.matterServer) {
            log.debug(`Stopping Matter server for child bridge ${this.bridgeConfig.username}`);
            try {
                await this.matterServer.stop();
            }
            catch (error) {
                log.error('Error stopping Matter server:', error);
            }
        }
        for (const [uuid, matterServer] of this.externalMatterServers) {
            log.debug(`Stopping external Matter server for ${uuid}`);
            try {
                await matterServer.stop();
            }
            catch (error) {
                log.error(`Error stopping external Matter server for ${uuid}:`, error);
            }
        }
        this.externalMatterServers.clear();
    }
}
