// homebridge/dist/matter/ChildBridgeMatterMessageHandler.js
import { Logger } from '../logger.js';
const log = Logger.withPrefix('Matter/ChildMessageHandler');
export class ChildBridgeMatterMessageHandler {
    matterManager;
    bridgeUsername;
    sendMessage;
    constructor(matterManager, bridgeUsername, sendMessage) {
        this.matterManager = matterManager;
        this.bridgeUsername = bridgeUsername;
        this.sendMessage = sendMessage;
    }
    handleStartMatterMonitoring() {
        if (this.matterManager?.isMatterEnabled()) {
            this.matterManager.enableStateMonitoring();
        }
    }
    handleStopMatterMonitoring() {
        if (this.matterManager?.isMatterEnabled()) {
            this.matterManager.disableStateMonitoring();
        }
    }
    handleGetMatterAccessories(data) {
        log.debug(`handleGetMatterAccessories called for bridge ${this.bridgeUsername}`);
        const correlationId = data?.correlationId;
        try {
            if (!this.matterManager?.isMatterEnabled()) {
                log.debug('Matter not enabled, returning empty accessories list');
                const event = {
                    type: 'accessoriesData',
                    correlationId,
                    data: {
                        bridgeUsername: this.bridgeUsername,
                        accessories: [],
                    },
                };
                this.sendMessage('matterEvent', event);
                return;
            }
            const accessories = this.matterManager.collectAllAccessories();
            log.debug(`Collected ${accessories.length} accessories from child bridge`);
            const event = {
                type: 'accessoriesData',
                correlationId,
                data: {
                    bridgeUsername: this.bridgeUsername,
                    accessories,
                },
            };
            this.sendMessage('matterEvent', event);
        }
        catch (error) {
            log.error('Failed to get Matter accessories:', error);
            const event = {
                type: 'accessoriesData',
                correlationId,
                data: {
                    bridgeUsername: this.bridgeUsername,
                    error: error instanceof Error ? error.message : 'Unknown error',
                },
            };
            this.sendMessage('matterEvent', event);
        }
    }
    handleGetMatterAccessoryInfo(data) {
        const correlationId = data?.correlationId;
        try {
            if (!this.matterManager?.isMatterEnabled()) {
                return;
            }
            const accessoryInfo = this.matterManager.getAccessoryInfo(data.uuid);
            if (accessoryInfo) {
                const event = {
                    type: 'accessoryInfoData',
                    correlationId,
                    data: accessoryInfo,
                };
                this.sendMessage('matterEvent', event);
            }
        }
        catch (error) {
            log.error('Failed to get Matter accessory info:', error);
            const event = {
                type: 'accessoryInfoData',
                correlationId,
                data: {
                    error: error instanceof Error ? error.message : 'Unknown error',
                },
            };
            this.sendMessage('matterEvent', event);
        }
    }
    handleMatterAccessoryControl(data) {
        const correlationId = data?.correlationId;
        if (!this.matterManager?.isMatterEnabled()) {
            log.debug(`Ignoring Matter control for ${data.uuid} - Matter not enabled on child bridge ${this.bridgeUsername}`);
            return;
        }
        log.debug(`Matter control request for child bridge ${this.bridgeUsername}: uuid=${data.uuid}, cluster=${data.cluster}, part=${data.partId || 'main'}`);
        this.matterManager.handleTriggerCommand(data.uuid, data.cluster, data.attributes, data.partId)
            .then(() => {
            log.debug(`Successfully controlled accessory ${data.uuid} on child bridge ${this.bridgeUsername}`);
            const controlResponse = {
                type: 'accessoryControlResponse',
                correlationId,
                data: {
                    success: true,
                    uuid: data.uuid,
                },
            };
            this.sendMessage('matterEvent', controlResponse);
        })
            .catch((error) => {
            if (error.message.includes('not found on this bridge')) {
                log.debug(`Accessory ${data.uuid} not on child bridge ${this.bridgeUsername}, ignoring`);
                return;
            }
            log.error(`Failed to control ${data.uuid} on child bridge ${this.bridgeUsername}: ${error.message}`);
            const event = {
                type: 'accessoryControlResponse',
                correlationId,
                data: {
                    success: false,
                    error: error instanceof Error ? error.message : 'Unknown error',
                    uuid: data.uuid,
                },
            };
            this.sendMessage('matterEvent', event);
        });
    }
}
