// homebridge/dist/matter/ClusterCommandMapper.js
const CLUSTER_COMMAND_MAPPINGS = {
    onOff: {
        map: (attributes) => {
            if ('onOff' in attributes) {
                return {
                    command: attributes.onOff ? 'on' : 'off',
                };
            }
            return null;
        },
    },
    levelControl: {
        map: (attributes) => {
            if ('currentLevel' in attributes) {
                return {
                    command: 'moveToLevelWithOnOff',
                    args: { level: attributes.currentLevel },
                };
            }
            return null;
        },
    },
    colorControl: {
        map: (attributes) => {
            if ('colorTemperatureMireds' in attributes) {
                return {
                    command: 'moveToColorTemperature',
                    args: {
                        colorTemperatureMireds: attributes.colorTemperatureMireds,
                        transitionTime: attributes.transitionTime ?? 0,
                        optionsMask: 1,
                        optionsOverride: 1,
                    },
                };
            }
            if ('colorX' in attributes && 'colorY' in attributes) {
                return {
                    command: 'moveToColor',
                    args: {
                        colorX: attributes.colorX,
                        colorY: attributes.colorY,
                        transitionTime: attributes.transitionTime ?? 0,
                        optionsMask: 1,
                        optionsOverride: 1,
                    },
                };
            }
            if ('currentHue' in attributes || 'currentSaturation' in attributes) {
                return {
                    command: 'moveToHueAndSaturation',
                    args: {
                        hue: attributes.currentHue ?? 0,
                        saturation: attributes.currentSaturation ?? 0,
                        transitionTime: attributes.transitionTime ?? 0,
                        optionsMask: 1,
                        optionsOverride: 1,
                    },
                };
            }
            return null;
        },
    },
    fanControl: {
        map: (attributes) => {
            if ('fanMode' in attributes) {
                return null;
            }
            if ('percentSetting' in attributes) {
                return null;
            }
            return null;
        },
    },
    windowCovering: {
        map: (attributes) => {
            if ('_command' in attributes) {
                const cmd = attributes._command;
                if (cmd === 'upOrOpen' || cmd === 'downOrClose' || cmd === 'stopMotion') {
                    return { command: cmd };
                }
            }
            if ('targetPositionLiftPercent100ths' in attributes) {
                return {
                    command: 'goToLiftPercentage',
                    args: {
                        liftPercent100thsValue: attributes.targetPositionLiftPercent100ths,
                    },
                };
            }
            if ('targetPositionTiltPercent100ths' in attributes) {
                return {
                    command: 'goToTiltPercentage',
                    args: {
                        tiltPercent100thsValue: attributes.targetPositionTiltPercent100ths,
                    },
                };
            }
            return null;
        },
    },
    doorLock: {
        map: (attributes) => {
            if ('_command' in attributes) {
                const cmd = attributes._command;
                if (cmd === 'lockDoor' || cmd === 'unlockDoor') {
                    return { command: cmd };
                }
            }
            if ('lockState' in attributes) {
                return {
                    command: attributes.lockState === 1 ? 'lockDoor' : 'unlockDoor',
                };
            }
            return null;
        },
    },
    thermostat: {
        map: (attributes) => {
            if ('_command' in attributes && attributes._command === 'setpointRaiseLower') {
                return {
                    command: 'setpointRaiseLower',
                    args: {
                        mode: attributes.mode ?? 0,
                        amount: attributes.amount ?? 10,
                    },
                };
            }
            if ('systemMode' in attributes) {
                return null;
            }
            if ('occupiedHeatingSetpoint' in attributes) {
                return null;
            }
            if ('occupiedCoolingSetpoint' in attributes) {
                return null;
            }
            return null;
        },
    },
    rvcOperationalState: {
        map: (attributes) => {
            if ('_command' in attributes) {
                const cmd = attributes._command;
                if (cmd === 'pause' || cmd === 'resume' || cmd === 'goHome') {
                    return { command: cmd };
                }
                return null;
            }
            if ('operationalState' in attributes) {
                const state = attributes.operationalState;
                switch (state) {
                    case 1:
                        return { command: 'resume' };
                    case 2:
                        return { command: 'pause' };
                    default:
                        return null;
                }
            }
            return null;
        },
    },
    rvcRunMode: {
        map: (attributes) => {
            if ('currentMode' in attributes) {
                return {
                    command: 'changeToMode',
                    args: { newMode: attributes.currentMode },
                };
            }
            return null;
        },
    },
    rvcCleanMode: {
        map: (attributes) => {
            if ('currentMode' in attributes) {
                return {
                    command: 'changeToMode',
                    args: { newMode: attributes.currentMode },
                };
            }
            return null;
        },
    },
    serviceArea: {
        map: (attributes) => {
            if ('selectedAreas' in attributes && Array.isArray(attributes.selectedAreas)) {
                return {
                    command: 'selectAreas',
                    args: { newAreas: attributes.selectedAreas },
                };
            }
            if ('skipArea' in attributes) {
                return {
                    command: 'skipArea',
                    args: { skippedArea: attributes.skipArea },
                };
            }
            return null;
        },
    },
};
export function mapAttributesToCommand(cluster, attributes) {
    const mapping = CLUSTER_COMMAND_MAPPINGS[cluster];
    if (!mapping) {
        throw new Error(`Command mapping not implemented for cluster: ${cluster}`);
    }
    const result = mapping.map(attributes);
    if (result === null) {
        return null;
    }
    return result;
}
