// homebridge/dist/matter/configValidator.js
import { Logger } from '../logger.js';
const log = Logger.withPrefix('Matter/Config');
const COLON_RE = /:/g;
export function validatePort(port, checkConflicts = false) {
    if (!Number.isInteger(port)) {
        return { valid: false, error: `Port must be an integer (got: ${port})` };
    }
    if (port < 1025 || port > 65534) {
        return { valid: false, error: `Port must be between 1025 and 65534 (got: ${port})` };
    }
    if (checkConflicts) {
        const conflictPorts = [5353, 8080, 8443];
        if (conflictPorts.includes(port)) {
            return {
                valid: true,
                warning: `Port ${port} may conflict with other services. Consider using a different port.`,
            };
        }
    }
    return { valid: true };
}
export function sanitizeUniqueId(uniqueId) {
    const warnings = [];
    const original = uniqueId;
    if (uniqueId.trim().length === 0) {
        return {
            value: '',
            warnings: ['uniqueId is empty after trimming'],
        };
    }
    const sanitized = uniqueId.replace(COLON_RE, '').toUpperCase();
    if (sanitized !== original) {
        warnings.push(`uniqueId was sanitized from "${original}" to "${sanitized}"`);
    }
    if (sanitized.length === 0) {
        warnings.push('uniqueId resulted in empty string after sanitization');
    }
    return { value: sanitized, warnings };
}
export function truncateString(value, maxLength, fieldName) {
    const warnings = [];
    if (value.length > maxLength) {
        warnings.push(`${fieldName} exceeds ${maxLength} characters, truncating: ${value}`);
        log.warn(`${fieldName} exceeds ${maxLength} characters, truncating: ${value}`);
        return {
            value: value.slice(0, maxLength),
            warnings,
        };
    }
    return { value, warnings };
}
export function checkPortProximity(hapPort, matterPort) {
    const MIN_PORT_SEPARATION = 10;
    if (Math.abs(hapPort - matterPort) < MIN_PORT_SEPARATION) {
        return `HAP port ${hapPort} and Matter port ${matterPort} are very close. Consider spacing them further apart.`;
    }
    return undefined;
}
export class MatterConfigValidator {
    static validate(config) {
        const result = {
            isValid: true,
            errors: [],
            warnings: [],
        };
        this.validatePort(config, result);
        result.isValid = result.errors.length === 0;
        if (result.warnings.length > 0) {
            log.warn('Matter configuration warnings:');
            result.warnings.forEach(warning => log.warn(`  - ${warning}`));
        }
        if (result.errors.length > 0) {
            log.error('Matter configuration errors:');
            result.errors.forEach(error => log.error(`  - ${error}`));
        }
        return result;
    }
    static validatePort(config, result) {
        const port = config.port;
        if (port !== undefined && port !== null) {
            if (typeof port === 'number') {
                const validation = validatePort(port, true);
                if (!validation.valid) {
                    result.errors.push(validation.error);
                }
                else if (validation.warning) {
                    result.warnings.push(validation.warning);
                }
            }
            else {
                result.errors.push(`Port must be a number, got ${typeof port}.`);
            }
        }
    }
    static validateChildMatterConfig(config, configType, identifier) {
        const result = {
            isValid: true,
            errors: [],
            warnings: [],
        };
        if (!config._bridge?.matter) {
            return result;
        }
        const matterConfig = config._bridge.matter;
        const prefix = `Child Matter bridge for ${configType} "${identifier}"`;
        if (matterConfig.port !== undefined) {
            const validation = validatePort(matterConfig.port, false);
            if (!validation.valid) {
                result.errors.push(`${prefix}: ${validation.error}`);
                result.isValid = false;
            }
        }
        else {
            result.warnings.push(`${prefix}: No port specified. Port will be auto-allocated.`);
        }
        if (config._bridge && config._bridge.matter) {
            if (config._bridge.port && matterConfig.port) {
                const proximityWarning = checkPortProximity(config._bridge.port, matterConfig.port);
                if (proximityWarning) {
                    result.warnings.push(`${prefix}: ${proximityWarning}`);
                }
            }
        }
        if (result.errors.length > 0) {
            log.error(`${prefix} validation errors:`);
            result.errors.forEach(error => log.error(`  - ${error}`));
        }
        if (result.warnings.length > 0) {
            log.warn(`${prefix} validation warnings:`);
            result.warnings.forEach(warning => log.warn(`  - ${warning}`));
        }
        return result;
    }
    static validateAllChildMatterConfigs(platforms, accessories) {
        const result = {
            isValid: true,
            errors: [],
            warnings: [],
        };
        const usedPorts = new Set();
        for (const platform of platforms) {
            if (platform._bridge?.matter) {
                const validation = this.validateChildMatterConfig(platform, 'platform', platform.platform || 'unknown');
                result.errors.push(...validation.errors);
                result.warnings.push(...validation.warnings);
                result.isValid = result.isValid && validation.isValid;
                if (platform._bridge.matter.port) {
                    if (usedPorts.has(platform._bridge.matter.port)) {
                        result.errors.push(`Duplicate Matter port ${platform._bridge.matter.port} detected. Each Matter bridge must use a unique port.`);
                        result.isValid = false;
                    }
                    usedPorts.add(platform._bridge.matter.port);
                }
            }
        }
        for (const accessory of accessories) {
            if (accessory._bridge?.matter) {
                const validation = this.validateChildMatterConfig(accessory, 'accessory', accessory.accessory || 'unknown');
                result.errors.push(...validation.errors);
                result.warnings.push(...validation.warnings);
                result.isValid = result.isValid && validation.isValid;
                if (accessory._bridge.matter.port) {
                    if (usedPorts.has(accessory._bridge.matter.port)) {
                        result.errors.push(`Duplicate Matter port ${accessory._bridge.matter.port} detected. Each Matter bridge must use a unique port.`);
                        result.isValid = false;
                    }
                    usedPorts.add(accessory._bridge.matter.port);
                }
            }
        }
        return result;
    }
}
