// homebridge/dist/matter/errors.js
import { Status, StatusResponseError } from '@matter/main/types';
export class MatterProtocolError extends StatusResponseError {
    constructor(message, code, clusterCode) {
        super(message, code, clusterCode);
    }
}
export class Busy extends MatterProtocolError {
    constructor(message = 'Device is busy') {
        super(message, Status.Busy);
    }
}
export class Timeout extends MatterProtocolError {
    constructor(message = 'Operation timed out') {
        super(message, Status.Timeout);
    }
}
export class ConstraintError extends MatterProtocolError {
    constructor(message = 'Value out of range or invalid') {
        super(message, Status.ConstraintError);
    }
}
export class InvalidAction extends MatterProtocolError {
    constructor(message = 'Invalid action or values') {
        super(message, Status.InvalidAction);
    }
}
export class InvalidInState extends MatterProtocolError {
    constructor(message = 'Operation not allowed in current state') {
        super(message, Status.InvalidInState);
    }
}
export class Failure extends MatterProtocolError {
    constructor(message = 'Operation failed') {
        super(message, Status.Failure);
    }
}
export class ResourceExhausted extends MatterProtocolError {
    constructor(message = 'Insufficient resources') {
        super(message, Status.ResourceExhausted);
    }
}
export class PermissionDenied extends MatterProtocolError {
    constructor(message = 'Permission denied') {
        super(message, Status.UnsupportedAccess);
    }
}
export class NotFound extends MatterProtocolError {
    constructor(message = 'Entity not found') {
        super(message, Status.NotFound);
    }
}
export function isMatterProtocolError(error) {
    return error instanceof StatusResponseError;
}
export const MatterStatus = {
    MatterProtocolError,
    Busy,
    Timeout,
    ConstraintError,
    InvalidAction,
    InvalidInState,
    Failure,
    ResourceExhausted,
    PermissionDenied,
    NotFound,
    isMatterProtocolError,
};
