// homebridge/dist/matter/ExternalMatterAccessoryPublisher.js
import { Logger } from '../logger.js';
import { User } from '../user.js';
import { generate } from '../util/mac.js';
import { MatterServer } from './server.js';
const log = Logger.withPrefix('Matter/External');
const COLON_RE = /:/g;
export async function publishExternalMatterAccessory(accessory, context) {
    if (!accessory.UUID) {
        log.error('External Matter accessory missing UUID - skipping');
        return null;
    }
    if (!accessory.displayName) {
        log.error(`External Matter accessory ${accessory.UUID} missing displayName - skipping`);
        return null;
    }
    const advertiseAddress = generate(accessory.UUID);
    const uniqueId = advertiseAddress.replace(COLON_RE, '');
    const port = await context.portService.requestMatterPort(uniqueId);
    if (!port) {
        log.error(`Failed to allocate Matter port for external Matter accessory ${accessory.displayName}`);
        log.error('Please configure matterPorts in config.json or free up ports in the default range (5530-5541)');
        return null;
    }
    log.info(`Allocated port ${port} for external Matter accessory: ${accessory.displayName}`);
    const matterServer = new MatterServer({
        port,
        uniqueId,
        storagePath: User.matterPath(),
        displayName: accessory.displayName || 'Matter Device',
        manufacturer: accessory.manufacturer,
        model: accessory.model,
        firmwareRevision: accessory.firmwareRevision,
        serialNumber: accessory.serialNumber || uniqueId,
        debugModeEnabled: context.debugModeEnabled,
        externalAccessory: true,
        networkInterfaces: context.networkInterfaces,
    });
    await matterServer.start();
    const pluginIdentifier = accessory._associatedPlugin || 'unknown';
    await matterServer.registerPlatformAccessories(pluginIdentifier, 'ExternalMatter', [accessory]);
    await matterServer.runServer();
    log.info(`✓ External Matter accessory published: ${accessory.displayName} on port ${port} (bridge ${advertiseAddress})`);
    const commissioningInfo = matterServer.getCommissioningInfo();
    return {
        server: matterServer,
        port,
        username: advertiseAddress,
        commissioningInfo,
    };
}
