// homebridge/dist/matter/index.js
/**
 * Matter Protocol Support for Homebridge
 *
 * This module provides Matter protocol support through a Plugin API,
 * allowing plugin developers to explicitly register Matter accessories.
 */
export { ChildBridgeMatterManager } from './ChildBridgeMatterManager.js';
export { MatterConfigCollector } from './config.js';
export { MatterConfigValidator } from './configValidator.js';
export { MatterStatus } from './errors.js';
export { MatterBridgeManager } from './MatterBridgeManager.js';
export { MatterServer } from './server.js';
export { clusterNames, clusters, devices, deviceTypes, MatterAccessoryEventTypes, MatterCommissioningError, MatterDeviceError, MatterError, MatterErrorType, MatterNetworkError, MatterStorageError, } from './types.js';
/**
 * Matter Cluster Types & Enums
 * Import these to access type-safe enum values for cluster attributes.
 *
 * Only the clusters that are actually used are exported here to minimize startup time.
 * If you need additional clusters, they must be added to the imports in types.ts.
 */
export { clusters as MatterTypes } from './types.js';
