// homebridge/dist/matter/logFormatter.js
import { Diagnostic } from '@matter/general';
import chalk from 'chalk';
import { Logger } from '../logger.js';
export function createHomebridgeLogFormatter() {
    const timestampEnabled = Logger.isTimestampEnabled();
    return (diagnostic) => {
        if (typeof diagnostic === 'object' && diagnostic !== null) {
            const msg = diagnostic;
            if (msg.now && msg.facility && msg.values) {
                if (msg.facility === 'ValidatedElements' || msg.facility === 'Commissioning') {
                    return '';
                }
                const facility = formatCyan(`[Matter/${msg.facility}]`);
                let messageText = formatMessageValues(msg.values);
                if (msg.facility === 'MessageChannel' && messageText.length > 200) {
                    messageText = `${messageText.substring(0, 200)} [trimmed...]`;
                }
                let coloredMessage = messageText;
                if (msg.level !== undefined) {
                    switch (msg.level) {
                        case 5:
                        case 4:
                            coloredMessage = chalk.red(messageText);
                            break;
                        case 3:
                            coloredMessage = chalk.yellow(messageText);
                            break;
                        case 0:
                            coloredMessage = chalk.gray(messageText);
                            break;
                    }
                }
                if (timestampEnabled) {
                    const timestamp = formatHomebridgeTimestamp(msg.now);
                    return `${timestamp} ${facility} ${coloredMessage}`;
                }
                else {
                    return `${facility} ${coloredMessage}`;
                }
            }
        }
        return String(diagnostic);
    };
}
function formatHomebridgeTimestamp(date) {
    const timestamp = `[${date.toLocaleString()}]`;
    return chalk.white(timestamp);
}
function formatCyan(text) {
    return chalk.cyan(text);
}
function formatMessageValues(values) {
    if (!Array.isArray(values)) {
        return String(values);
    }
    return values
        .map((value) => {
        if (value === null) {
            return 'null';
        }
        if (value === undefined) {
            return 'undefined';
        }
        if (typeof value === 'function') {
            try {
                return String(value());
            }
            catch {
                return String(value);
            }
        }
        if (typeof value === 'object') {
            const diagnosticValue = Diagnostic.valueOf(value);
            if (diagnosticValue !== undefined && diagnosticValue !== value) {
                if (Array.isArray(diagnosticValue)) {
                    return formatMessageValues(diagnosticValue);
                }
                return formatMessageValues([diagnosticValue]);
            }
            if (value instanceof Error) {
                const errorDetails = {
                    ...value,
                    name: value.name,
                    message: value.message,
                    stack: value.stack,
                };
                return JSON.stringify(errorDetails, null, 2);
            }
            try {
                return JSON.stringify(value);
            }
            catch {
                return String(value);
            }
        }
        return String(value).trim();
    })
        .join(' ');
}
