// homebridge/dist/matter/MatterAPIImpl.js
import { Logger } from '../logger.js';
import { clusterNames, clusters, deviceTypes, MatterTypes } from './index.js';
const log = Logger.withPrefix('Matter/API');
const EXTERNAL_DEVICE_TYPES = [
    deviceTypes.RoboticVacuumCleaner,
];
function requiresExternalBridge(deviceType) {
    return EXTERNAL_DEVICE_TYPES.some(externalType => externalType.deviceType === deviceType.deviceType);
}
class MatterAccessoryValidationError extends Error {
    accessory;
    constructor(message, accessory) {
        super(message);
        this.accessory = accessory;
        this.name = 'MatterAccessoryValidationError';
    }
}
export class MatterAPIImpl {
    api;
    constructor(api) {
        this.api = api;
    }
    validateAccessory(accessory, context) {
        if (!accessory.UUID) {
            throw new MatterAccessoryValidationError(`${context}: Matter accessory missing required 'UUID' field`, accessory);
        }
        if (!accessory.displayName) {
            throw new MatterAccessoryValidationError(`${context}: Matter accessory '${accessory.UUID}' missing required 'displayName' field`, accessory);
        }
        if (!accessory.deviceType) {
            throw new MatterAccessoryValidationError(`${context}: Matter accessory '${accessory.displayName}' (${accessory.UUID}) missing required 'deviceType' field`, accessory);
        }
        if (!accessory.manufacturer) {
            throw new MatterAccessoryValidationError(`${context}: Matter accessory '${accessory.displayName}' (${accessory.UUID}) missing required 'manufacturer' field`, accessory);
        }
        if (!accessory.model) {
            throw new MatterAccessoryValidationError(`${context}: Matter accessory '${accessory.displayName}' (${accessory.UUID}) missing required 'model' field`, accessory);
        }
        if (!accessory.serialNumber) {
            throw new MatterAccessoryValidationError(`${context}: Matter accessory '${accessory.displayName}' (${accessory.UUID}) missing required 'serialNumber' field`, accessory);
        }
    }
    validateAccessories(accessories, context) {
        const validAccessories = [];
        for (const accessory of accessories) {
            try {
                this.validateAccessory(accessory, context);
                validAccessories.push(accessory);
            }
            catch (error) {
                if (error instanceof MatterAccessoryValidationError) {
                    log.error(error.message);
                    log.error('This accessory will not be registered. Please fix the issue in your plugin.');
                }
                else {
                    log.error(`${context}: Unexpected error validating accessory:`, error);
                }
            }
        }
        return validAccessories;
    }
    validateClusterName(clusterName, context) {
        const validClusterNames = Object.values(clusterNames);
        if (!validClusterNames.includes(clusterName)) {
            log.warn(`${context}: Unknown cluster name '${clusterName}'. This might cause issues. `
                + `Valid clusters: ${validClusterNames.join(', ')}`);
        }
    }
    get uuid() {
        return this.api.hap.uuid;
    }
    get deviceTypes() {
        return deviceTypes;
    }
    get clusters() {
        return clusters;
    }
    get clusterNames() {
        return clusterNames;
    }
    get types() {
        return MatterTypes;
    }
    async registerPlatformAccessories(pluginIdentifier, platformName, accessories) {
        if (accessories.length === 0) {
            log.warn(`${pluginIdentifier}: Attempted to register 0 Matter accessories`);
            return;
        }
        const validAccessories = this.validateAccessories(accessories, `registerPlatformAccessories (${pluginIdentifier}/${platformName})`);
        if (validAccessories.length === 0) {
            log.error(`${pluginIdentifier}: All ${accessories.length} Matter accessories failed validation`);
            return;
        }
        if (validAccessories.length < accessories.length) {
            log.warn(`${pluginIdentifier}: ${accessories.length - validAccessories.length} of ${accessories.length} Matter accessories failed validation`);
        }
        const normalAccessories = [];
        const externalAccessories = [];
        for (const accessory of validAccessories) {
            if (requiresExternalBridge(accessory.deviceType)) {
                externalAccessories.push(accessory);
            }
            else {
                normalAccessories.push(accessory);
            }
        }
        if (normalAccessories.length > 0) {
            normalAccessories.forEach((accessory) => {
                const internal = accessory;
                internal._associatedPlugin = pluginIdentifier;
                internal._associatedPlatform = platformName;
            });
            log.debug(`${pluginIdentifier}: Registering ${normalAccessories.length} Matter accessor${normalAccessories.length === 1 ? 'y' : 'ies'} for platform '${platformName}'`);
            this.api.emit("registerMatterPlatformAccessories", pluginIdentifier, platformName, normalAccessories);
        }
        if (externalAccessories.length > 0) {
            externalAccessories.forEach((accessory) => {
                const internal = accessory;
                internal._associatedPlugin = pluginIdentifier;
            });
            log.debug(`${pluginIdentifier}: Publishing ${externalAccessories.length} external Matter accessor${externalAccessories.length === 1 ? 'y' : 'ies'} (${externalAccessories.map(a => a.displayName).join(', ')})`);
            const registrationId = `${pluginIdentifier}-${Date.now()}-${Math.random()}`;
            const registrationPromise = new Promise((resolve) => {
                const internalApi = this.api;
                if (!internalApi._pendingExternalRegistrations) {
                    internalApi._pendingExternalRegistrations = new Map();
                }
                internalApi._pendingExternalRegistrations.set(registrationId, resolve);
            });
            this.api.emit("publishExternalMatterAccessories", externalAccessories, registrationId);
            await registrationPromise;
        }
    }
    async updatePlatformAccessories(accessories) {
        if (accessories.length === 0) {
            log.warn('Attempted to update 0 Matter platform accessories');
            return;
        }
        log.debug(`Updating ${accessories.length} Matter platform accessor${accessories.length === 1 ? 'y' : 'ies'} in cache`);
        this.api.emit("updateMatterPlatformAccessories", accessories);
    }
    async unregisterPlatformAccessories(pluginIdentifier, platformName, accessories) {
        if (accessories.length === 0) {
            log.warn(`${pluginIdentifier}: Attempted to unregister 0 Matter accessories`);
            return;
        }
        const normalAccessories = [];
        const externalAccessories = [];
        for (const accessory of accessories) {
            if (requiresExternalBridge(accessory.deviceType)) {
                externalAccessories.push(accessory);
            }
            else {
                normalAccessories.push(accessory);
            }
        }
        if (normalAccessories.length > 0) {
            log.debug(`${pluginIdentifier}: Unregistering ${normalAccessories.length} Matter accessor${normalAccessories.length === 1 ? 'y' : 'ies'} from platform '${platformName}'`);
            this.api.emit("unregisterMatterPlatformAccessories", pluginIdentifier, platformName, normalAccessories);
        }
        if (externalAccessories.length > 0) {
            log.debug(`${pluginIdentifier}: Unregistering ${externalAccessories.length} external Matter accessor${externalAccessories.length === 1 ? 'y' : 'ies'} (${externalAccessories.map(a => a.displayName).join(', ')})`);
            this.api.emit("unregisterExternalMatterAccessories", externalAccessories);
        }
    }
    async updateAccessoryState(uuid, cluster, attributes, partId) {
        if (!uuid) {
            log.error('updateAccessoryState: uuid parameter is required');
            return;
        }
        if (!cluster) {
            log.error(`updateAccessoryState: cluster parameter is required for accessory ${uuid}`);
            return;
        }
        if (!attributes || Object.keys(attributes).length === 0) {
            log.warn(`updateAccessoryState: No attributes provided for accessory ${uuid}, cluster ${cluster}`);
            return;
        }
        this.validateClusterName(cluster, `updateAccessoryState (${uuid})`);
        log.debug(`Updating Matter accessory state: uuid=${uuid}, cluster=${cluster}, attributes=${Object.keys(attributes).join(', ')}${partId ? `, partId=${partId}` : ''}`);
        this.api.emit("updateMatterAccessoryState", uuid, cluster, attributes, partId);
    }
    async getAccessoryState(uuid, cluster, partId) {
        if (!uuid) {
            log.error('getAccessoryState: uuid parameter is required');
            return undefined;
        }
        if (!cluster) {
            log.error(`getAccessoryState: cluster parameter is required for accessory ${uuid}`);
            return undefined;
        }
        this.validateClusterName(cluster, `getAccessoryState (${uuid})`);
        log.debug(`Getting Matter accessory state: uuid=${uuid}, cluster=${cluster}${partId ? `, partId=${partId}` : ''}`);
        const internalApi = this.api;
        const matterManager = internalApi._matterManager;
        if (matterManager) {
            const externalServer = matterManager.getExternalServer(uuid);
            if (externalServer) {
                return externalServer.getAccessoryState(uuid, cluster, partId);
            }
        }
        const matterServer = internalApi._matterServer;
        if (!matterServer) {
            log.debug(`getAccessoryState: Matter server not available for accessory ${uuid}`);
            return undefined;
        }
        return matterServer.getAccessoryState(uuid, cluster, partId);
    }
}
