// homebridge/dist/matter/MatterPortAllocator.js
import { Logger } from '../logger.js';
const log = Logger.withPrefix('Matter/PortAllocator');
export class MatterPortAllocator {
    matterPorts;
    allocatedPorts = new Map();
    configuredPorts = new Set();
    constructor(matterPorts, configuredMatterPorts) {
        this.matterPorts = matterPorts;
        if (configuredMatterPorts) {
            this.configuredPorts = new Set(configuredMatterPorts);
        }
    }
    async requestPort(uuid) {
        const existingPortAllocation = this.allocatedPorts.get(uuid);
        if (existingPortAllocation !== undefined) {
            return existingPortAllocation;
        }
        const port = this.getNextFreePort();
        this.allocatedPorts.set(uuid, port);
        return port;
    }
    getNextFreePort() {
        const rangeStart = this.matterPorts?.start || 5530;
        const rangeEnd = this.matterPorts?.end || 5541;
        if (rangeStart > rangeEnd) {
            log.error(`Invalid Matter port range: start (${rangeStart}) is greater than end (${rangeEnd}).`);
            return undefined;
        }
        const allocatedPortValues = new Set([
            ...this.configuredPorts,
            ...[...this.allocatedPorts.values()].filter((p) => p !== undefined),
        ]);
        for (let port = rangeStart; port <= rangeEnd; port += 1) {
            if (!allocatedPortValues.has(port)) {
                return port;
            }
        }
        log.warn(`Matter port range ${rangeStart}-${rangeEnd} exhausted, allocating from extended range.`);
        for (let port = rangeEnd + 1; port <= 65535; port += 1) {
            if (!allocatedPortValues.has(port)) {
                return port;
            }
        }
        log.error('No available ports remaining for Matter allocation.');
        return undefined;
    }
    getStats() {
        return {
            allocatedCount: this.allocatedPorts.size,
            configuredPortsCount: this.configuredPorts.size,
        };
    }
}
