// homebridge/dist/matter/MatterBridgeManager.js
import { DEFAULT_BRIDGE_DEFAULTS } from '../bridgeService.js';
import { Logger } from '../logger.js';
import { User } from '../user.js';
import getVersion from '../version.js';
import { BaseMatterManager } from './BaseMatterManager.js';
import { publishExternalMatterAccessory } from './ExternalMatterAccessoryPublisher.js';
import { MatterServer } from './server.js';
import { getErrorCode, normalizeBindConfig } from './utils.js';
const log = Logger.withPrefix('Matter/MainManager');
const COLON_RE = /:/g;
export class MatterBridgeManager extends BaseMatterManager {
    config;
    api;
    externalPortService;
    options;
    server;
    constructor(config, api, externalPortService, pluginManager, options, server) {
        super(pluginManager);
        this.config = config;
        this.api = api;
        this.externalPortService = externalPortService;
        this.options = options;
        this.server = server;
        this.setupEventListeners();
    }
    setupEventListeners() {
        this.api.on("publishExternalMatterAccessories", (accessories, registrationId) => {
            this.handlePublishExternalAccessories(accessories, registrationId).catch((error) => {
                log.error('Failed to publish external Matter accessories:', error);
                this.api._resolveExternalRegistration(registrationId);
            });
        });
        this.api.on("registerMatterPlatformAccessories", (pluginIdentifier, platformName, accessories) => {
            this.handleRegisterPlatformAccessories(pluginIdentifier, platformName, accessories).catch((error) => {
                log.error(`Failed to register Matter accessories for ${pluginIdentifier}:`, error);
            });
        });
        this.api.on("updateMatterPlatformAccessories", (accessories) => {
            this.handleUpdatePlatformAccessories(accessories).catch((error) => {
                log.error('Failed to update Matter platform accessories:', error);
            });
        });
        this.api.on("unregisterMatterPlatformAccessories", (pluginIdentifier, platformName, accessories) => {
            this.handleUnregisterPlatformAccessories(pluginIdentifier, platformName, accessories).catch((error) => {
                log.error(`Failed to unregister Matter accessories for ${pluginIdentifier}:`, error);
            });
        });
        this.api.on("unregisterExternalMatterAccessories", (accessories) => {
            this.handleUnregisterExternalAccessories(accessories).catch((error) => {
                log.error('Failed to unregister external Matter accessories:', error);
            });
        });
        this.api.on("updateMatterAccessoryState", (uuid, cluster, attributes, partId) => {
            this.handleUpdateAccessoryState(uuid, cluster, attributes, partId).catch((error) => {
                log.error('Failed to update Matter accessory state:', error);
            });
        });
    }
    async initialize() {
        if (!this.config.bridge.matter) {
            return;
        }
        let matterPort;
        try {
            log.info('Initializing Matter server for main bridge...');
            matterPort = this.config.bridge.matter.port;
            if (!matterPort) {
                matterPort = await this.externalPortService.requestPort(`${this.config.bridge.username}:MATTER`);
                if (!matterPort) {
                    matterPort = 5540;
                    log.warn('No port available from pool for main Matter bridge, using default port 5540');
                }
                else {
                    log.info(`Allocated port ${matterPort} from pool for main Matter bridge`);
                }
            }
            const serialNumber = this.config.bridge.username.replace(COLON_RE, '');
            const networkInterfaces = normalizeBindConfig(this.config.bridge.bind);
            this.matterServer = new MatterServer({
                storagePath: User.matterPath(),
                port: matterPort,
                uniqueId: serialNumber,
                displayName: this.config.bridge.name || 'Main Bridge',
                manufacturer: this.config.bridge.manufacturer || DEFAULT_BRIDGE_DEFAULTS.manufacturer,
                model: this.config.bridge.model || DEFAULT_BRIDGE_DEFAULTS.model,
                firmwareRevision: this.config.bridge.firmwareRevision || getVersion(),
                serialNumber,
                debugModeEnabled: this.options.debugModeEnabled,
                networkInterfaces,
            });
            await this.matterServer.start();
            log.info('Matter server initialized for main bridge');
            this.api._setMatterEnabled(true);
            this.api._setMatterServer(this.matterServer);
            this.matterServer.on('stateChange', ({ uuid, cluster, state, partId }) => {
                const event = {
                    type: 'accessoryUpdate',
                    data: {
                        uuid,
                        cluster,
                        state,
                        partId,
                    },
                };
                this.server.ipcService.sendMessage("matterEvent", event);
            });
        }
        catch (error) {
            log.error('Failed to initialize Matter server for main bridge:', error);
            const errorMessage = error instanceof Error ? error.message : String(error);
            const errorCode = getErrorCode(error);
            if (errorMessage.includes('corrupted')) {
                log.error('');
                log.error('╔════════════════════════════════════════════════════════════════════════════╗');
                log.error('║  MATTER STORAGE CORRUPTED                                                  ║');
                log.error('╠════════════════════════════════════════════════════════════════════════════╣');
                log.error('║  Your Matter storage has become corrupted. This can happen when:          ║');
                log.error('║  • Matter.js library version changes                                       ║');
                log.error('║  • Storage format upgrades occur                                           ║');
                log.error('║  • Incomplete writes during shutdown                                       ║');
                log.error('║                                                                            ║');
                log.error('║  To fix this, delete the corrupted storage directory:                     ║');
                log.error(`║  rm -rf ~/.homebridge/matter/${this.config.bridge.username}                                   ║`);
                log.error('║                                                                            ║');
                log.error('║  Note: You will need to re-pair your Matter devices after deletion.       ║');
                log.error('╚════════════════════════════════════════════════════════════════════════════╝');
                log.error('');
            }
            else if (errorCode === 'EADDRINUSE' || errorMessage.includes('address already in use')) {
                log.error('');
                log.error('╔════════════════════════════════════════════════════════════════════════════╗');
                log.error('║  MATTER PORT ALREADY IN USE                                                ║');
                log.error('╠════════════════════════════════════════════════════════════════════════════╣');
                log.error(`║  Port ${matterPort} is already in use by another application.                    ║`);
                log.error('║                                                                            ║');
                log.error('║  To fix this:                                                              ║');
                log.error('║  1. Stop the application using this port, or                              ║');
                log.error('║  2. Configure a different port in your config.json:                       ║');
                log.error('║     "bridge": {                                                            ║');
                log.error('║       "matter": {                                                          ║');
                log.error('║         "port": <different-port>                                           ║');
                log.error('║       }                                                                    ║');
                log.error('║     }                                                                      ║');
                log.error('╚════════════════════════════════════════════════════════════════════════════╝');
                log.error('');
            }
        }
    }
    async handlePublishExternalAccessories(accessories, registrationId) {
        log.info(`Publishing ${accessories.length} external Matter accessor${accessories.length === 1 ? 'y' : 'ies'}`);
        const networkInterfaces = normalizeBindConfig(this.config.bridge.bind);
        try {
            for (const accessory of accessories) {
                try {
                    if (this.externalMatterServers.has(accessory.UUID)) {
                        log.warn(`External Matter accessory ${accessory.displayName} (${accessory.UUID}) is already published`);
                        continue;
                    }
                    const result = await publishExternalMatterAccessory(accessory, {
                        portService: this.externalPortService,
                        networkInterfaces,
                        debugModeEnabled: this.options.debugModeEnabled,
                    });
                    if (!result) {
                        continue;
                    }
                    this.externalMatterServers.set(accessory.UUID, result.server);
                    result.server.on('stateChange', ({ uuid, cluster, state, partId }) => {
                        const event = {
                            type: 'accessoryUpdate',
                            data: {
                                uuid,
                                cluster,
                                state,
                                partId,
                            },
                        };
                        this.server.ipcService.sendMessage("matterEvent", event);
                    });
                    this.server.registerExternalMatterBridge(result.username, this.config.bridge.username);
                    if (accessory._eventEmitter) {
                        accessory._eventEmitter.emit('ready', result.port);
                    }
                    if (result.commissioningInfo.qrCode && result.commissioningInfo.manualPairingCode) {
                        log.info(`📱 Commissioning codes for ${accessory.displayName}:`);
                        log.info(`   QR Code: ${result.commissioningInfo.qrCode}`);
                        log.info(`   Manual Code: ${result.commissioningInfo.manualPairingCode}`);
                    }
                }
                catch (error) {
                    log.error(`Failed to publish external Matter accessory ${accessory.displayName}:`, error);
                }
            }
        }
        finally {
            this.api._resolveExternalRegistration(registrationId);
        }
    }
    getMatterStatus() {
        if (this.matterServer) {
            const commissioningInfo = this.matterServer.getCommissioningInfo();
            return {
                enabled: true,
                port: this.config.bridge.matter?.port,
                setupUri: commissioningInfo.qrCode,
                pin: commissioningInfo.manualPairingCode,
                serialNumber: commissioningInfo.serialNumber,
                commissioned: commissioningInfo.commissioned || false,
                deviceCount: this.matterServer.getAccessories().length,
            };
        }
        else if (this.config.bridge.matter) {
            return {
                enabled: false,
                port: this.config.bridge.matter?.port,
            };
        }
        return {
            enabled: false,
        };
    }
    collectAllAccessories(bridgeUsername) {
        const accessories = [];
        if (!bridgeUsername || bridgeUsername === this.config.bridge.username) {
            if (this.matterServer) {
                const mainAccessories = this.collectAccessoriesFromServer(this.matterServer, this.config.bridge.username, 'main', this.config.bridge.name || 'Homebridge');
                accessories.push(...mainAccessories);
                for (const server of this.externalMatterServers.values()) {
                    const externalAccessories = this.collectAccessoriesFromServer(server, server.username, 'external', server.bridgeName);
                    accessories.push(...externalAccessories);
                }
            }
        }
        return accessories;
    }
    getAccessoryInfo(uuid) {
        if (this.matterServer) {
            const accessory = this.getAccessoryDetailFromServer(this.matterServer, uuid, this.config.bridge.username, 'main');
            if (accessory) {
                return accessory;
            }
        }
        for (const server of this.externalMatterServers.values()) {
            const accessory = this.getAccessoryDetailFromServer(server, uuid, server.username, 'external');
            if (accessory) {
                return accessory;
            }
        }
        return undefined;
    }
    collectAccessoriesFromServer(server, bridgeUsername, bridgeType, bridgeName) {
        const cached = server.getAllCachedAccessories();
        const accessories = [];
        for (const acc of cached) {
            const accessory = this.transformAccessoryData(acc, server, bridgeUsername, bridgeType, bridgeName);
            accessories.push(accessory);
        }
        return accessories;
    }
    transformAccessoryData(acc, server, bridgeUsername, bridgeType, bridgeName) {
        const currentState = this.getCurrentStateFromServer(server, acc.uuid);
        const deviceTypeStr = acc.deviceType.name || `Device Code ${acc.deviceType.code || 'unknown'}`;
        return {
            uuid: acc.uuid,
            displayName: acc.displayName,
            serialNumber: acc.serialNumber,
            manufacturer: acc.manufacturer,
            model: acc.model,
            firmwareRevision: acc.firmwareRevision,
            deviceType: deviceTypeStr,
            clusters: currentState,
            parts: acc.parts?.map(part => ({
                id: part.id,
                displayName: part.displayName,
                deviceType: part.deviceType.name || `Device Code ${part.deviceType.code || 'unknown'}`,
                clusters: this.getCurrentStateFromServer(server, acc.uuid, part.id),
            })),
            bridge: {
                username: bridgeUsername,
                type: bridgeType,
                name: bridgeName,
            },
            plugin: acc.plugin,
            platform: acc.platform,
            context: acc.context,
            commissioned: server.isCommissioned(),
            fabricCount: server.getCommissionedFabricCount(),
            fabrics: server.getFabricInfo().map(fabric => ({
                fabricIndex: fabric.fabricIndex,
                fabricId: BigInt(fabric.fabricId),
                nodeId: BigInt(fabric.nodeId),
                vendorId: fabric.rootVendorId,
                label: fabric.label,
            })),
        };
    }
    getAccessoryDetailFromServer(server, uuid, bridgeUsername, bridgeType) {
        const accessory = server.getAccessory(uuid);
        if (!accessory) {
            return undefined;
        }
        const cached = server.getAllCachedAccessories().find((a) => a.uuid === uuid);
        if (!cached) {
            return undefined;
        }
        return this.transformAccessoryData(cached, server, bridgeUsername, bridgeType, server.bridgeName || 'Matter Bridge');
    }
    getCurrentStateFromServer(server, uuid, partId) {
        const accessory = server.getAccessory(uuid);
        if (!accessory) {
            return {};
        }
        const endpoint = partId
            ? accessory._parts?.find((p) => p.id === partId)?.endpoint
            : accessory.endpoint;
        if (!endpoint) {
            return {};
        }
        const state = {};
        for (const [clusterName, clusterState] of Object.entries(endpoint.state)) {
            state[clusterName] = {};
            for (const [key, value] of Object.entries(clusterState)) {
                if (!key.startsWith('_') && !key.startsWith('$')) {
                    state[clusterName][key] = value;
                }
            }
        }
        return state;
    }
    async teardown() {
        if (this.matterServer) {
            try {
                await this.matterServer.stop();
            }
            catch (error) {
                log.error('Failed to stop Matter server:', error);
            }
        }
        for (const [uuid, matterServer] of this.externalMatterServers) {
            try {
                await matterServer.stop();
                log.debug(`Stopped external Matter server for ${uuid}`);
            }
            catch (error) {
                log.error(`Failed to stop external Matter server for ${uuid}:`, error);
            }
        }
        this.externalMatterServers.clear();
    }
}
