// homebridge/dist/matter/serverHelpers.js
import { Logger } from '../logger.js';
import { HomebridgeRvcCleanModeServer, HomebridgeServiceAreaServer, HomebridgeWindowCoveringServer, } from './behaviors/index.js';
import { clusters, devices, MatterDeviceError } from './types.js';
const log = Logger.withPrefix('Matter/Server');
export const CLUSTER_IDS = {
    AIR_QUALITY: clusters.AirQuality.Cluster.id,
    CARBON_MONOXIDE_CONCENTRATION: clusters.CarbonMonoxideConcentrationMeasurement.Cluster.id,
    COLOR_CONTROL: clusters.ColorControl.Cluster.id,
    DOOR_LOCK: clusters.DoorLock.Cluster.id,
    LEVEL_CONTROL: clusters.LevelControl.Cluster.id,
    NITROGEN_DIOXIDE_CONCENTRATION: clusters.NitrogenDioxideConcentrationMeasurement.Cluster.id,
    ON_OFF: clusters.OnOff.Cluster.id,
    OZONE_CONCENTRATION: clusters.OzoneConcentrationMeasurement.Cluster.id,
    PM10_CONCENTRATION: clusters.Pm10ConcentrationMeasurement.Cluster.id,
    PM25_CONCENTRATION: clusters.Pm25ConcentrationMeasurement.Cluster.id,
    THERMOSTAT: clusters.Thermostat.Cluster.id,
    WINDOW_COVERING: clusters.WindowCovering.Cluster.id,
};
export function validateAccessoryRequiredFields(accessory) {
    if (!accessory.deviceType) {
        throw new MatterDeviceError(`Matter accessory "${accessory.displayName || 'unknown'}" is missing required field 'deviceType'. `
            + 'Example: deviceType: api.matter.deviceTypes.OnOffLight\n'
            + 'Available device types: OnOffLight, DimmableLight, TemperatureSensor, etc.\n'
            + 'See the Matter types documentation for the full list.');
    }
    if (!accessory.UUID) {
        throw new MatterDeviceError('Matter accessory is missing required field \'UUID\'.\n'
            + 'Generate a unique UUID for your accessory:\n'
            + '  const UUID = api.matter.uuid.generate(\'my-unique-id\')');
    }
    if (!accessory.displayName) {
        throw new MatterDeviceError(`Matter accessory (${accessory.UUID}) is missing required field 'displayName'.\n`
            + 'Example: displayName: \'Living Room Light\'');
    }
    if (!accessory.serialNumber) {
        throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" is missing required field 'serialNumber'.\n`
            + 'Example: serialNumber: \'ABC123\' or serialNumber: accessory.UUID');
    }
    if (!accessory.manufacturer) {
        throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" is missing required field 'manufacturer'.\n`
            + 'Example: manufacturer: \'Homebridge\' or manufacturer: \'My Plugin Name\'');
    }
    if (!accessory.model) {
        throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" is missing required field 'model'.\n`
            + 'Example: model: \'v1.0\' or model: \'Smart Light\'');
    }
    if (!accessory.parts || accessory.parts.length === 0) {
        if (!accessory.clusters || typeof accessory.clusters !== 'object') {
            throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" is missing or has invalid 'clusters' field.\n`
                + 'Clusters define the functionality of your device. Example:\n'
                + '  clusters: {\n'
                + '    onOff: { onOff: false },\n'
                + '    levelControl: { currentLevel: 0, minLevel: 0, maxLevel: 254 }\n'
                + '  }\n'
                + 'Alternatively, use "parts" array for composed devices with multiple endpoints.');
        }
    }
    if (accessory.parts && accessory.parts.length > 0) {
        for (const part of accessory.parts) {
            if (!part.id) {
                throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" has a part missing required field 'id'`);
            }
            if (!part.deviceType) {
                throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" part "${part.id}" is missing required field 'deviceType'`);
            }
            if (!part.clusters || typeof part.clusters !== 'object') {
                throw new MatterDeviceError(`Matter accessory "${accessory.displayName}" part "${part.id}" is missing or has invalid 'clusters' field`);
            }
        }
    }
}
function convertBehaviorsToArray(behaviors) {
    if (Array.isArray(behaviors)) {
        return behaviors;
    }
    if (typeof behaviors === 'object' && behaviors !== null) {
        const values = Object.values(behaviors);
        if (values.length > 0) {
            return values;
        }
    }
    try {
        return [...behaviors];
    }
    catch {
        return [];
    }
}
function findBehaviorByCluster(behaviors, clusterIdOrName) {
    return behaviors.find((behavior) => {
        if (typeof clusterIdOrName === 'number') {
            return behavior.cluster?.id === clusterIdOrName;
        }
        return behavior.id === clusterIdOrName;
    });
}
export function detectBehaviorFeatures(deviceType, clusterIdOrName, featureExtractor) {
    const deviceTypeDef = deviceType;
    const existingBehaviors = deviceTypeDef.behaviors;
    if (!existingBehaviors) {
        return null;
    }
    const behaviorsArray = convertBehaviorsToArray(existingBehaviors);
    const behavior = findBehaviorByCluster(behaviorsArray, clusterIdOrName);
    if (!behavior?.cluster?.supportedFeatures) {
        return null;
    }
    return featureExtractor(behavior.cluster.supportedFeatures);
}
export function extractColorControlFeatures(supportedFeatures) {
    const features = [];
    if (supportedFeatures.hueSaturation) {
        features.push('HueSaturation');
    }
    if (supportedFeatures.xy) {
        features.push('Xy');
    }
    if (supportedFeatures.colorTemperature) {
        features.push('ColorTemperature');
    }
    return features;
}
export function extractThermostatFeatures(supportedFeatures) {
    const features = [];
    if (supportedFeatures.heating) {
        features.push('Heating');
    }
    if (supportedFeatures.cooling) {
        features.push('Cooling');
    }
    if (supportedFeatures.occupancy) {
        features.push('Occupancy');
    }
    if (supportedFeatures.autoMode) {
        features.push('AutoMode');
    }
    return features;
}
export function determineColorControlFeaturesFromHandlers(handlers) {
    const features = [];
    if ('moveToHueAndSaturationLogic' in handlers) {
        features.push('HueSaturation');
    }
    if ('moveToColorLogic' in handlers) {
        features.push('Xy');
    }
    if ('moveToColorTemperatureLogic' in handlers) {
        features.push('ColorTemperature');
    }
    return features;
}
export function detectWindowCoveringFeatures(accessory) {
    const features = [];
    const wcCluster = accessory.clusters?.windowCovering;
    if (!wcCluster) {
        return features;
    }
    const hasLiftAttrs = 'targetPositionLiftPercent100ths' in wcCluster
        || 'currentPositionLiftPercent100ths' in wcCluster;
    const configStatus = wcCluster.configStatus;
    const hasConfigLift = configStatus?.liftPositionAware === true;
    const hasTiltAttrs = 'targetPositionTiltPercent100ths' in wcCluster
        || 'currentPositionTiltPercent100ths' in wcCluster;
    const hasConfigTilt = configStatus?.tiltPositionAware === true;
    log.debug(`[${accessory.displayName}] WindowCovering detection: `
        + `hasLiftAttrs=${hasLiftAttrs}, hasConfigLift=${hasConfigLift}, `
        + `hasTiltAttrs=${hasTiltAttrs}, hasConfigTilt=${hasConfigTilt}`);
    if (hasLiftAttrs) {
        features.push('Lift');
        if (hasConfigLift) {
            features.push('PositionAwareLift');
        }
    }
    if (hasTiltAttrs) {
        features.push('Tilt');
        if (hasConfigTilt) {
            features.push('PositionAwareTilt');
        }
    }
    return features;
}
export function detectServiceAreaFeatures(serviceAreaCluster) {
    const features = [];
    if (!serviceAreaCluster) {
        return features;
    }
    if ('supportedMaps' in serviceAreaCluster) {
        features.push('Maps');
    }
    if ('progress' in serviceAreaCluster) {
        features.push('ProgressReporting');
    }
    return features;
}
export function applyWindowCoveringFeatures(deviceType, accessory, features) {
    if (features.length === 0) {
        log.warn(`⚠️  No WindowCovering features detected for ${accessory.displayName}!`);
        return deviceType;
    }
    log.info(`Auto-detected WindowCovering features for ${accessory.displayName}: ${features.join(', ')}`);
    const windowCoveringWithFeatures = HomebridgeWindowCoveringServer.with(...features);
    const modifiedDeviceType = deviceType.with(windowCoveringWithFeatures);
    const hasTiltFeatures = features.includes('Tilt');
    if (hasTiltFeatures && accessory.clusters) {
        const wcCluster = accessory.clusters.windowCovering;
        wcCluster.type = 8;
        log.debug('Set WindowCovering type to 8 (TiltBlindLift) for tilt-capable device');
    }
    if (!accessory.context) {
        accessory.context = {};
    }
    accessory.context._skipWindowCoveringBehavior = true;
    return modifiedDeviceType;
}
export function buildRvcCustomBehaviors(accessory, serviceAreaFeatures) {
    const customBehaviors = [];
    const { RvcCleanModeServer, ServiceAreaServer } = devices.RoboticVacuumCleanerRequirements;
    if (accessory.clusters?.rvcCleanMode) {
        if (accessory.handlers?.rvcCleanMode) {
            customBehaviors.push(HomebridgeRvcCleanModeServer);
            log.info('Adding custom RvcCleanMode behavior with handlers');
        }
        else {
            customBehaviors.push(RvcCleanModeServer);
            log.info('Adding base RvcCleanMode server');
        }
    }
    if (accessory.clusters?.serviceArea) {
        let behaviorClass = accessory.handlers?.serviceArea
            ? HomebridgeServiceAreaServer
            : ServiceAreaServer;
        if (serviceAreaFeatures && serviceAreaFeatures.length > 0) {
            behaviorClass = behaviorClass.with(...serviceAreaFeatures);
            log.info(`ServiceArea ${accessory.handlers?.serviceArea ? 'custom behavior' : 'base server'} will have features: ${serviceAreaFeatures.join(', ')}`);
        }
        customBehaviors.push(behaviorClass);
    }
    return customBehaviors;
}
export function applyFeaturesToBehavior(behaviorClass, features, clusterName) {
    if (!features || features.length === 0) {
        return behaviorClass;
    }
    const modifiedBehavior = behaviorClass.with(...features);
    log.info(`${clusterName} custom behavior will preserve features: ${features.join(', ')}`);
    return modifiedBehavior;
}
