// homebridge/dist/matter/types.js
import { AirQuality } from '@matter/main/clusters/air-quality';
import { BooleanState } from '@matter/main/clusters/boolean-state';
import { CarbonMonoxideConcentrationMeasurement } from '@matter/main/clusters/carbon-monoxide-concentration-measurement';
import { ColorControl } from '@matter/main/clusters/color-control';
import { DoorLock } from '@matter/main/clusters/door-lock';
import { FanControl } from '@matter/main/clusters/fan-control';
import { LevelControl } from '@matter/main/clusters/level-control';
import { NitrogenDioxideConcentrationMeasurement } from '@matter/main/clusters/nitrogen-dioxide-concentration-measurement';
import { OnOff } from '@matter/main/clusters/on-off';
import { OzoneConcentrationMeasurement } from '@matter/main/clusters/ozone-concentration-measurement';
import { Pm10ConcentrationMeasurement } from '@matter/main/clusters/pm10-concentration-measurement';
import { Pm25ConcentrationMeasurement } from '@matter/main/clusters/pm25-concentration-measurement';
import { RvcOperationalState } from '@matter/main/clusters/rvc-operational-state';
import { Thermostat } from '@matter/main/clusters/thermostat';
import { ValveConfigurationAndControl } from '@matter/main/clusters/valve-configuration-and-control';
import { WindowCovering } from '@matter/main/clusters/window-covering';
import { AirQualitySensorDevice } from '@matter/main/devices/air-quality-sensor';
import { ColorTemperatureLightDevice } from '@matter/main/devices/color-temperature-light';
import { ContactSensorDevice } from '@matter/main/devices/contact-sensor';
import { DimmableLightDevice } from '@matter/main/devices/dimmable-light';
import { DimmablePlugInUnitDevice } from '@matter/main/devices/dimmable-plug-in-unit';
import { DoorLockDevice } from '@matter/main/devices/door-lock';
import { ExtendedColorLightDevice } from '@matter/main/devices/extended-color-light';
import { FanDevice } from '@matter/main/devices/fan';
import { GenericSwitchDevice } from '@matter/main/devices/generic-switch';
import { HumiditySensorDevice } from '@matter/main/devices/humidity-sensor';
import { LightSensorDevice } from '@matter/main/devices/light-sensor';
import { OccupancySensorDevice } from '@matter/main/devices/occupancy-sensor';
import { OnOffLightDevice } from '@matter/main/devices/on-off-light';
import { OnOffLightSwitchDevice } from '@matter/main/devices/on-off-light-switch';
import { OnOffPlugInUnitDevice } from '@matter/main/devices/on-off-plug-in-unit';
import { PumpDevice } from '@matter/main/devices/pump';
import { RoboticVacuumCleanerDevice, RoboticVacuumCleanerRequirements } from '@matter/main/devices/robotic-vacuum-cleaner';
import { RoomAirConditionerDevice } from '@matter/main/devices/room-air-conditioner';
import { SmokeCoAlarmDevice } from '@matter/main/devices/smoke-co-alarm';
import { TemperatureSensorDevice } from '@matter/main/devices/temperature-sensor';
import { ThermostatDevice, ThermostatRequirements } from '@matter/main/devices/thermostat';
import { WaterLeakDetectorDevice } from '@matter/main/devices/water-leak-detector';
import { WaterValveDevice, WaterValveRequirements } from '@matter/main/devices/water-valve';
import { WindowCoveringDevice } from '@matter/main/devices/window-covering';
import { BridgedNodeEndpoint } from '@matter/main/endpoints/bridged-node';
export var MatterAccessoryEventTypes;
(function (MatterAccessoryEventTypes) {
    MatterAccessoryEventTypes["READY"] = "ready";
})(MatterAccessoryEventTypes || (MatterAccessoryEventTypes = {}));
export var MatterErrorType;
(function (MatterErrorType) {
    MatterErrorType["INITIALIZATION"] = "INITIALIZATION";
    MatterErrorType["NETWORK"] = "NETWORK";
    MatterErrorType["COMMISSIONING"] = "COMMISSIONING";
    MatterErrorType["DEVICE_SYNC"] = "DEVICE_SYNC";
    MatterErrorType["SERVER"] = "SERVER";
    MatterErrorType["STORAGE"] = "STORAGE";
    MatterErrorType["CONFIGURATION"] = "CONFIGURATION";
    MatterErrorType["DEVICE_ERROR"] = "DEVICE_ERROR";
    MatterErrorType["UNKNOWN"] = "UNKNOWN";
})(MatterErrorType || (MatterErrorType = {}));
export class MatterError extends Error {
    code;
    details;
    type;
    timestamp;
    recoverable;
    constructor(message, code, details) {
        super(message);
        this.code = code;
        this.details = details;
        this.name = 'MatterError';
        this.type = details?.type ?? MatterErrorType.UNKNOWN;
        this.timestamp = new Date();
        this.recoverable = details?.recoverable ?? true;
    }
}
export class MatterCommissioningError extends MatterError {
    constructor(message, details) {
        super(message, 'COMMISSIONING_ERROR', { ...details, type: MatterErrorType.COMMISSIONING });
        this.name = 'MatterCommissioningError';
    }
}
export class MatterStorageError extends MatterError {
    constructor(message, details) {
        super(message, 'STORAGE_ERROR', { ...details, type: MatterErrorType.STORAGE });
        this.name = 'MatterStorageError';
    }
}
export class MatterDeviceError extends MatterError {
    constructor(message, details) {
        super(message, 'DEVICE_ERROR', { ...details, type: MatterErrorType.DEVICE_ERROR });
        this.name = 'MatterDeviceError';
    }
}
export class MatterNetworkError extends MatterError {
    constructor(message, details) {
        super(message, 'NETWORK_ERROR', { ...details, type: MatterErrorType.NETWORK });
        this.name = 'MatterNetworkError';
    }
}
const devices = {
    AirQualitySensorDevice,
    ColorTemperatureLightDevice,
    ContactSensorDevice,
    DimmableLightDevice,
    DimmablePlugInUnitDevice,
    DoorLockDevice,
    ExtendedColorLightDevice,
    FanDevice,
    GenericSwitchDevice,
    HumiditySensorDevice,
    LightSensorDevice,
    OccupancySensorDevice,
    OnOffLightDevice,
    OnOffLightSwitchDevice,
    OnOffPlugInUnitDevice,
    PumpDevice,
    RoboticVacuumCleanerDevice,
    RoboticVacuumCleanerRequirements,
    RoomAirConditionerDevice,
    SmokeCoAlarmDevice,
    TemperatureSensorDevice,
    ThermostatDevice,
    ThermostatRequirements,
    WaterLeakDetectorDevice,
    WaterValveDevice,
    WaterValveRequirements,
    WindowCoveringDevice,
};
const clusters = {
    AirQuality,
    BooleanState,
    CarbonMonoxideConcentrationMeasurement,
    ColorControl,
    DoorLock,
    FanControl,
    LevelControl,
    NitrogenDioxideConcentrationMeasurement,
    OnOff,
    OzoneConcentrationMeasurement,
    Pm10ConcentrationMeasurement,
    Pm25ConcentrationMeasurement,
    RvcOperationalState,
    Thermostat,
    ValveConfigurationAndControl,
    WindowCovering,
};
export { clusters, devices };
export const deviceTypes = {
    OnOffLight: devices.OnOffLightDevice,
    DimmableLight: devices.DimmableLightDevice,
    ColorTemperatureLight: devices.ColorTemperatureLightDevice,
    ExtendedColorLight: devices.ExtendedColorLightDevice,
    OnOffSwitch: devices.OnOffLightSwitchDevice,
    OnOffOutlet: devices.OnOffPlugInUnitDevice,
    DimmableOutlet: devices.DimmablePlugInUnitDevice,
    AirQualitySensor: devices.AirQualitySensorDevice,
    TemperatureSensor: devices.TemperatureSensorDevice,
    HumiditySensor: devices.HumiditySensorDevice,
    LightSensor: devices.LightSensorDevice,
    MotionSensor: devices.OccupancySensorDevice,
    ContactSensor: devices.ContactSensorDevice,
    LeakSensor: devices.WaterLeakDetectorDevice,
    SmokeSensor: devices.SmokeCoAlarmDevice,
    Thermostat: devices.ThermostatDevice.with(devices.ThermostatRequirements.ThermostatServer.with('Heating', 'Cooling', 'AutoMode', 'Occupancy')),
    Fan: devices.FanDevice,
    DoorLock: devices.DoorLockDevice,
    WindowCovering: devices.WindowCoveringDevice,
    RoboticVacuumCleaner: devices.RoboticVacuumCleanerDevice,
    WaterValve: devices.WaterValveDevice.with(devices.WaterValveRequirements.ValveConfigurationAndControlServer),
    GenericSwitch: devices.GenericSwitchDevice,
    Pump: devices.PumpDevice,
    RoomAirConditioner: devices.RoomAirConditionerDevice,
    BridgedNode: BridgedNodeEndpoint,
};
export const clusterNames = {
    OnOff: 'onOff',
    LevelControl: 'levelControl',
    ColorControl: 'colorControl',
    DoorLock: 'doorLock',
    WindowCovering: 'windowCovering',
    Thermostat: 'thermostat',
    FanControl: 'fanControl',
    AirQuality: 'airQuality',
    CarbonMonoxideConcentrationMeasurement: 'carbonMonoxideConcentrationMeasurement',
    NitrogenDioxideConcentrationMeasurement: 'nitrogenDioxideConcentrationMeasurement',
    OzoneConcentrationMeasurement: 'ozoneConcentrationMeasurement',
    Pm10ConcentrationMeasurement: 'pm10ConcentrationMeasurement',
    Pm25ConcentrationMeasurement: 'pm25ConcentrationMeasurement',
    TemperatureMeasurement: 'temperatureMeasurement',
    RelativeHumidityMeasurement: 'relativeHumidityMeasurement',
    IlluminanceMeasurement: 'illuminanceMeasurement',
    OccupancySensing: 'occupancySensing',
    BooleanState: 'booleanState',
    SmokeCoAlarm: 'smokeCoAlarm',
    RvcRunMode: 'rvcRunMode',
    RvcOperationalState: 'rvcOperationalState',
    RvcCleanMode: 'rvcCleanMode',
    ServiceArea: 'serviceArea',
    PumpConfigurationAndControl: 'pumpConfigurationAndControl',
    ValveConfigurationAndControl: 'valveConfigurationAndControl',
    Identify: 'identify',
    BasicInformation: 'basicInformation',
    BridgedDeviceBasicInformation: 'bridgedDeviceBasicInformation',
};
export function hasEndpointState(endpoint) {
    return 'state' in endpoint
        && typeof endpoint.state === 'object'
        && endpoint.state !== null
        && 'set' in endpoint
        && typeof endpoint.set === 'function';
}
export async function updateEndpointState(endpoint, cluster, attributes) {
    if (!hasEndpointState(endpoint)) {
        throw new Error('Endpoint does not support state updates');
    }
    const updateObject = { [cluster]: attributes };
    await endpoint.set(updateObject);
}
export function getWindowCoveringCluster(accessory) {
    return accessory.clusters?.windowCovering;
}
