// @matter/node/dist/cjs/behaviors/door-lock/NonUserDoorLockServer.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : /* @__PURE__ */ Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decoratorStart = (base) => [, , , __create(base?.[__knownSymbol("metadata")] ?? null)];
var __decoratorStrings = ["class", "method", "getter", "setter", "accessor", "field", "value", "get", "set"];
var __expectFn = (fn) => fn !== void 0 && typeof fn !== "function" ? __typeError("Function expected") : fn;
var __decoratorContext = (kind, name, done, metadata, fns) => ({ kind: __decoratorStrings[kind], name, metadata, addInitializer: (fn) => done._ ? __typeError("Already initialized") : fns.push(__expectFn(fn || null)) });
var __decoratorMetadata = (array, target) => __defNormalProp(target, __knownSymbol("metadata"), array[3]);
var __runInitializers = (array, flags, self, value) => {
  for (var i = 0, fns = array[flags >> 1], n = fns && fns.length; i < n; i++) flags & 1 ? fns[i].call(self) : value = fns[i].call(self, value);
  return value;
};
var __decorateElement = (array, flags, name, decorators, target, extra) => {
  var fn, it, done, ctx, access, k = flags & 7, s = !!(flags & 8), p = !!(flags & 16);
  var j = k > 3 ? array.length + 1 : k ? s ? 1 : 2 : 0, key = __decoratorStrings[k + 5];
  var initializers = k > 3 && (array[j - 1] = []), extraInitializers = array[j] || (array[j] = []);
  var desc = k && (!p && !s && (target = target.prototype), k < 5 && (k > 3 || !p) && __getOwnPropDesc(k < 4 ? target : { get [name]() {
    return __privateGet(this, extra);
  }, set [name](x) {
    return __privateSet(this, extra, x);
  } }, name));
  k ? p && k < 4 && __name(extra, (k > 2 ? "set " : k > 1 ? "get " : "") + name) : __name(target, name);
  for (var i = decorators.length - 1; i >= 0; i--) {
    ctx = __decoratorContext(k, name, done = {}, array[3], extraInitializers);
    if (k) {
      ctx.static = s, ctx.private = p, access = ctx.access = { has: p ? (x) => __privateIn(target, x) : (x) => name in x };
      if (k ^ 3) access.get = p ? (x) => (k ^ 1 ? __privateGet : __privateMethod)(x, target, k ^ 4 ? extra : desc.get) : (x) => x[name];
      if (k > 2) access.set = p ? (x, y) => __privateSet(x, target, y, k ^ 4 ? extra : desc.set) : (x, y) => x[name] = y;
    }
    it = (0, decorators[i])(k ? k < 4 ? p ? extra : desc[key] : k > 4 ? void 0 : { get: desc.get, set: desc.set } : target, ctx), done._ = 1;
    if (k ^ 4 || it === void 0) __expectFn(it) && (k > 4 ? initializers.unshift(it) : k ? p ? extra = it : desc[key] = it : target = it);
    else if (typeof it !== "object" || it === null) __typeError("Object expected");
    else __expectFn(fn = it.get) && (desc.get = fn), __expectFn(fn = it.set) && (desc.set = fn), __expectFn(fn = it.init) && initializers.unshift(fn);
  }
  return k || __decoratorMetadata(array, target), desc && __defProp(target, name, desc), p ? k ^ 4 ? extra : desc : target;
};
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateIn = (member, obj) => Object(obj) !== obj ? __typeError('Cannot use the "in" operator on this value') : member.has(obj);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var NonUserDoorLockServer_exports = {};
__export(NonUserDoorLockServer_exports, {
  NonUserDoorLockServer: () => NonUserDoorLockServer
});
module.exports = __toCommonJS(NonUserDoorLockServer_exports);
var import_Supervision = require("#behavior/supervision/Supervision.js");
var import_general = require("@matter/general");
var import_model = require("@matter/model");
var import_types = require("@matter/types");
var import_door_lock = require("@matter/types/clusters/door-lock");
var import_DoorLockBehavior = require("./DoorLockBehavior.js");
var import_LockAuth = require("./LockAuth.js");
var import_LockSchedule = require("./LockSchedule.js");
/**
 * @license
 * Copyright 2022-2026 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AlarmCode = import_door_lock.DoorLock.AlarmCode;
const CredentialRule = import_door_lock.DoorLock.CredentialRule;
const CredentialType = import_door_lock.DoorLock.CredentialType;
const DoorState = import_door_lock.DoorLock.DoorState;
const LockOperationType = import_door_lock.DoorLock.LockOperationType;
const LockState = import_door_lock.DoorLock.LockState;
const OperationError = import_door_lock.DoorLock.OperationError;
const OperationSource = import_door_lock.DoorLock.OperationSource;
const UserStatus = import_door_lock.DoorLock.UserStatus;
const UserType = import_door_lock.DoorLock.UserType;
const NonUserBase = import_DoorLockBehavior.DoorLockBehavior.with(
  "PinCredential",
  "RfidCredential",
  "FingerCredentials",
  "FaceCredentials",
  "DoorPositionSensor",
  "CredentialOverTheAirAccess",
  "Unbolting"
);
class NonUserDoorLockServer extends NonUserBase {
  initialize() {
    const state = this.state;
    if (!Object.values(state.supportedOperatingModes).some((v) => v)) {
      state.supportedOperatingModes = { vacation: true, privacy: true, passage: true, alwaysSet: 2047 };
    } else if (state.supportedOperatingModes.alwaysSet !== 2047) {
      throw new import_general.ImplementationError(
        `DoorLockServer: The "alwaysSet" bit-range in supportedOperatingModes must be set. Please check the specification about the meaning of this field because bits are inverted here!`
      );
    }
    if (!state.users) {
      state.users = [];
    }
    if (!state.credentials) {
      state.credentials = [];
    }
    if (!state.weekDaySchedules) {
      state.weekDaySchedules = [];
    }
    if (!state.yearDaySchedules) {
      state.yearDaySchedules = [];
    }
    if (!state.holidaySchedules) {
      state.holidaySchedules = [];
    }
    if (!state.credentialKey) {
      state.credentialKey = import_general.Bytes.of(this.env.get(import_general.Crypto).randomBytes(16));
      if (state.credentials.length > 0) {
        const auth = this.auth;
        state.credentials = state.credentials.map((c) => ({
          ...c,
          credentialData: auth.encrypt(c.credentialData)
        }));
      }
    }
    this.reactTo(this.events.doorState$Changed, this.#handleDoorStateChange);
  }
  [Symbol.asyncDispose]() {
    this.#stopAutoRelockTimer();
    return super[Symbol.asyncDispose]();
  }
  // ── Core Lock Operations ─────────────────────────────────────────────────
  lockDoor(request) {
    this.#validatePinIfRequired(request.pinCode, LockOperationType.Lock);
    this.state.lockState = LockState.Locked;
    this.#stopAutoRelockTimer();
    this.#emitLockOperation(LockOperationType.Lock, request.pinCode);
  }
  unlockDoor(request) {
    this.#validatePinIfRequired(request.pinCode, LockOperationType.Unlock);
    this.state.lockState = LockState.Unlocked;
    this.#emitLockOperation(LockOperationType.Unlock, request.pinCode);
    this.#scheduleAutoRelock();
  }
  unlockWithTimeout(request) {
    this.#validatePinIfRequired(request.pinCode, LockOperationType.Unlock);
    this.state.lockState = LockState.Unlocked;
    this.#emitLockOperation(LockOperationType.Unlock, request.pinCode);
    this.#scheduleAutoRelock(request.timeout);
  }
  unboltDoor(request) {
    this.#validatePinIfRequired(request.pinCode, LockOperationType.Unlatch);
    this.state.lockState = LockState.Unlatched;
    this.#emitLockOperation(LockOperationType.Unlatch, request.pinCode);
    this.#scheduleAutoRelock();
  }
  // ── Non-User PIN/RFID Commands (PIN & !USR, RID & !USR) ─────────────────
  setPinCode(request) {
    this.#setNonUserCode(CredentialType.Pin, request.userId, request.pin, request.userStatus, request.userType);
  }
  getPinCode(request) {
    return this.#getNonUserPinCode(request.userId);
  }
  clearPinCode(request) {
    this.#clearNonUserCode(CredentialType.Pin, request.pinSlotIndex);
  }
  clearAllPinCodes() {
    this.#clearAllNonUserCodes(CredentialType.Pin);
  }
  setRfidCode(request) {
    this.#setNonUserCode(
      CredentialType.Rfid,
      request.userId,
      request.rfidCode,
      request.userStatus,
      request.userType
    );
  }
  getRfidCode(request) {
    return this.#getNonUserRfidCode(request.userId);
  }
  clearRfidCode(request) {
    this.#clearNonUserCode(CredentialType.Rfid, request.rfidSlotIndex);
  }
  clearAllRfidCodes() {
    this.#clearAllNonUserCodes(CredentialType.Rfid);
  }
  // ── Non-User Status Commands (!USR) ──────────────────────────────────────
  setUserStatus(request) {
    const auth = this.auth;
    const user = auth.findUser(request.userId);
    if (!user) {
      throw new import_types.StatusResponseError("User not found", import_types.StatusCode.InvalidCommand);
    }
    auth.replaceUser(request.userId, { ...user, userStatus: request.userStatus });
  }
  getUserStatus(request) {
    const user = this.auth.findUser(request.userId);
    return { userId: request.userId, userStatus: user?.userStatus ?? UserStatus.Available };
  }
  setUserType(request) {
    const auth = this.auth;
    const user = auth.findUser(request.userId);
    if (!user) {
      throw new import_types.StatusResponseError("User not found", import_types.StatusCode.InvalidCommand);
    }
    auth.replaceUser(request.userId, { ...user, userType: request.userType });
  }
  getUserType(request) {
    const user = this.auth.findUser(request.userId);
    return { userId: request.userId, userType: user?.userType ?? UserType.UnrestrictedUser };
  }
  // ── Credential Storage (overridable) ─────────────────────────────────────
  //
  // The default implementation uses reversible encryption (AES-128-CCM) rather than hashing because the legacy
  // non-User commands (getPinCode, getRfidCode) must return plaintext credential data. Deployments where credential
  // secrecy is critical should override `cipher` (for HSM-backed encryption) or `auth` (for an external credential
  // store).
  /**
   * The cipher used for credential encryption. Override to integrate with an HSM or other secure storage.
   */
  get cipher() {
    return new import_general.AesCipher(this.env.get(import_general.Crypto), this.state.credentialKey);
  }
  /**
   * The credential/user store with integrated encryption. Override to replace the entire storage backend.
   */
  get auth() {
    return new import_LockAuth.LockAuth.Store(this.state, this.cipher);
  }
  // ── Private Helpers ────────────────────────────────────────────────────────
  get #fabricIndex() {
    const fabric = this.context.fabric;
    if (fabric === void 0) {
      throw new import_types.StatusResponseError("Fabric required", import_types.StatusCode.UnsupportedAccess);
    }
    return fabric;
  }
  #stopAutoRelockTimer() {
    this.internal.autoRelockTimer?.stop();
    this.internal.autoRelockTimer = void 0;
  }
  #scheduleAutoRelock(timeoutSeconds) {
    this.#stopAutoRelockTimer();
    const timeout = timeoutSeconds ?? this.state.autoRelockTime;
    if (timeout === void 0 || timeout === 0) {
      return;
    }
    this.internal.autoRelockTimer = import_general.Time.getTimer(
      "auto-relock",
      (0, import_general.Seconds)(timeout),
      this.callback(this.#autoRelock, { lock: true })
    ).start();
  }
  #autoRelock() {
    if (this.state.lockState !== LockState.Locked) {
      this.state.lockState = LockState.Locked;
      this.events.lockOperation.emit(
        {
          lockOperationType: LockOperationType.Lock,
          operationSource: OperationSource.Auto,
          userIndex: null,
          fabricIndex: null,
          sourceNode: null,
          credentials: null
        },
        this.context
      );
    }
  }
  // ── PIN Validation ─────────────────────────────────────────────────────────
  #validatePinIfRequired(pinCode, operationType) {
    const requirePin = this.state.requirePinForRemoteOperation;
    if (pinCode !== void 0) {
      this.#validatePin(pinCode, operationType);
      return;
    }
    if (requirePin) {
      this.#emitLockOperationError(operationType, OperationError.InvalidCredential);
      throw new import_types.StatusResponseError("PIN required for remote operation", import_types.StatusCode.Failure);
    }
  }
  #validatePin(pinCode, operationType) {
    if (this.state.wrongCodeEntryLimit !== void 0 && this.state.userCodeTemporaryDisableTime !== void 0) {
      if (this.internal.wrongCodeCount >= this.state.wrongCodeEntryLimit) {
        this.#emitLockOperationError(operationType, OperationError.InvalidCredential);
        throw new import_types.StatusResponseError("User code temporarily disabled", import_types.StatusCode.Failure);
      }
    }
    const auth = this.auth;
    for (const cred of auth.credentials) {
      if (cred.credentialType !== CredentialType.Pin) {
        continue;
      }
      const decrypted = auth.decrypt(cred.credentialData);
      if (import_general.Bytes.areEqual(decrypted, pinCode)) {
        const userIndex = auth.findUserIndexForCredential(CredentialType.Pin, cred.credentialIndex);
        if (userIndex !== null) {
          const user = auth.findUser(userIndex);
          if (user && user.userStatus === UserStatus.OccupiedDisabled) {
            this.#emitLockOperationError(operationType, OperationError.DisabledUserDenied);
            throw new import_types.StatusResponseError("User is disabled", import_types.StatusCode.Failure);
          }
        }
        this.internal.wrongCodeCount = 0;
        return;
      }
    }
    const wrongCount = ++this.internal.wrongCodeCount;
    if (this.state.wrongCodeEntryLimit !== void 0 && wrongCount >= this.state.wrongCodeEntryLimit) {
      this.events.doorLockAlarm.emit({ alarmCode: AlarmCode.WrongCodeEntryLimit }, this.context);
      if (this.state.userCodeTemporaryDisableTime !== void 0) {
        import_general.Time.getTimer(
          "wrong-code-disable",
          (0, import_general.Seconds)(this.state.userCodeTemporaryDisableTime),
          this.callback(
            () => {
              this.internal.wrongCodeCount = 0;
            },
            { lock: true }
          )
        ).start();
      }
    }
    this.#emitLockOperationError(operationType, OperationError.InvalidCredential);
    throw new import_types.StatusResponseError("Invalid PIN code", import_types.StatusCode.Failure);
  }
  // ── Event Emission ─────────────────────────────────────────────────────────
  #emitLockOperation(operationType, pinCode) {
    let userIndex = null;
    let credentials = null;
    if (pinCode !== void 0) {
      const auth = this.auth;
      for (const cred of auth.credentials) {
        if (cred.credentialType !== CredentialType.Pin) continue;
        const decrypted = auth.decrypt(cred.credentialData);
        if (import_general.Bytes.areEqual(decrypted, pinCode)) {
          userIndex = auth.findUserIndexForCredential(CredentialType.Pin, cred.credentialIndex);
          credentials = [{ credentialType: CredentialType.Pin, credentialIndex: cred.credentialIndex }];
          break;
        }
      }
    }
    this.events.lockOperation.emit(
      {
        lockOperationType: operationType,
        operationSource: OperationSource.Remote,
        userIndex,
        fabricIndex: this.#fabricIndex,
        sourceNode: this.#sourceNode,
        credentials
      },
      this.context
    );
  }
  #emitLockOperationError(operationType, error) {
    this.events.lockOperationError.emit(
      {
        lockOperationType: operationType,
        operationSource: OperationSource.Remote,
        operationError: error,
        userIndex: null,
        fabricIndex: this.#fabricIndex,
        sourceNode: this.#sourceNode,
        credentials: null
      },
      this.context
    );
  }
  get #sourceNode() {
    const context = this.context;
    if ("subject" in context && context.subject?.kind === "node") {
      return context.subject.id;
    }
    return null;
  }
  // ── DPS ────────────────────────────────────────────────────────────────────
  #handleDoorStateChange(value) {
    if (value === null) {
      return;
    }
    this.events.doorStateChange.emit({ doorState: value }, this.context);
    if (value === DoorState.DoorOpen) {
      if (this.state.doorOpenEvents !== void 0) {
        this.state.doorOpenEvents++;
      }
    } else if (value === DoorState.DoorClosed) {
      if (this.state.doorClosedEvents !== void 0) {
        this.state.doorClosedEvents++;
      }
    }
  }
  // ── Credential Helpers ─────────────────────────────────────────────────────
  #maxCredentialsForType(type) {
    switch (type) {
      case CredentialType.Pin:
      case CredentialType.ProgrammingPin:
        return this.state.numberOfPinUsersSupported;
      case CredentialType.Rfid:
        return this.state.numberOfRfidUsersSupported;
      default:
        return 0;
    }
  }
  #validateCredentialDataLength(type, data) {
    switch (type) {
      case CredentialType.Pin:
      case CredentialType.ProgrammingPin: {
        const min = this.state.minPinCodeLength;
        const max = this.state.maxPinCodeLength;
        return data.byteLength >= min && data.byteLength <= max;
      }
      case CredentialType.Rfid: {
        const min = this.state.minRfidCodeLength;
        const max = this.state.maxRfidCodeLength;
        return data.byteLength >= min && data.byteLength <= max;
      }
      default:
        return true;
    }
  }
  #clearAllCredentialsOfType(type) {
    const auth = this.auth;
    const credsOfType = auth.credentials.filter((c) => c.credentialType === type);
    for (const cred of credsOfType) {
      this.#removeCredentialFromUsers(auth, type, cred.credentialIndex);
    }
    auth.removeAllCredentialsOfType(type);
  }
  #removeCredentialFromUsers(auth, type, index) {
    const userIndex = auth.findUserIndexForCredential(type, index);
    if (userIndex === null) return;
    const user = auth.findUser(userIndex);
    if (!user) return;
    const remainingCreds = user.credentials.filter(
      (c) => !(c.credentialType === type && c.credentialIndex === index)
    );
    if (remainingCreds.length === 0) {
      auth.removeUser(userIndex);
      this.state.weekDaySchedules = this.state.weekDaySchedules.filter((s) => s.userIndex !== userIndex);
      this.state.yearDaySchedules = this.state.yearDaySchedules.filter((s) => s.userIndex !== userIndex);
    } else {
      auth.replaceUser(userIndex, { ...user, credentials: remainingCreds });
    }
  }
  // ── NonUser Code Helpers ────────────────────────────────────────────────────
  #setNonUserCode(type, userId, code, userStatus, userType) {
    const max = this.#maxCredentialsForType(type);
    if (userId < 0 || userId > max) {
      throw new import_types.StatusResponseError("Invalid user ID", import_types.StatusCode.InvalidCommand);
    }
    if (!this.#validateCredentialDataLength(type, code)) {
      throw new import_types.StatusResponseError("Invalid code length", import_types.StatusCode.InvalidCommand);
    }
    const auth = this.auth;
    if (auth.isDuplicateCredential(type, code, userId)) {
      throw new import_door_lock.DoorLock.DuplicateError("Duplicate credential");
    }
    const fabricIndex = this.#fabricIndex;
    const existingCred = auth.findCredential(type, userId);
    const existingUser = auth.findUser(userId);
    if (existingCred) {
      auth.updateCredentialData(type, userId, code, fabricIndex);
      if (existingUser) {
        auth.replaceUser(userId, {
          ...existingUser,
          userStatus: userStatus ?? existingUser.userStatus,
          userType: userType ?? existingUser.userType,
          lastModifiedFabricIndex: fabricIndex
        });
      }
    } else {
      auth.addCredential(type, userId, code, fabricIndex);
      if (existingUser) {
        auth.replaceUser(userId, {
          ...existingUser,
          credentials: [...existingUser.credentials, { credentialType: type, credentialIndex: userId }],
          userStatus: userStatus ?? existingUser.userStatus,
          userType: userType ?? existingUser.userType,
          lastModifiedFabricIndex: fabricIndex
        });
      } else {
        auth.addUser({
          userIndex: userId,
          userName: "",
          userUniqueId: 4294967295,
          userStatus: userStatus ?? UserStatus.OccupiedEnabled,
          userType: userType ?? UserType.UnrestrictedUser,
          credentialRule: CredentialRule.Single,
          credentials: [{ credentialType: type, credentialIndex: userId }],
          creatorFabricIndex: fabricIndex,
          lastModifiedFabricIndex: fabricIndex
        });
      }
    }
  }
  #getNonUserPinCode(userId) {
    const auth = this.auth;
    const user = auth.findUser(userId);
    const cred = auth.findCredential(CredentialType.Pin, userId);
    if (!user || !cred) {
      return {
        userId,
        userStatus: UserStatus.Available,
        userType: UserType.UnrestrictedUser,
        pinCode: null
      };
    }
    return {
      userId,
      userStatus: user.userStatus,
      userType: user.userType,
      pinCode: auth.decrypt(cred.credentialData)
    };
  }
  #getNonUserRfidCode(userId) {
    const auth = this.auth;
    const user = auth.findUser(userId);
    const cred = auth.findCredential(CredentialType.Rfid, userId);
    if (!user || !cred) {
      return {
        userId,
        userStatus: UserStatus.Available,
        userType: UserType.UnrestrictedUser,
        rfidCode: null
      };
    }
    return {
      userId,
      userStatus: user.userStatus,
      userType: user.userType,
      rfidCode: auth.decrypt(cred.credentialData)
    };
  }
  #clearNonUserCode(type, slotIndex) {
    const max = this.#maxCredentialsForType(type);
    if (slotIndex < 0 || slotIndex > max) {
      throw new import_types.StatusResponseError("Invalid slot index", import_types.StatusCode.InvalidCommand);
    }
    const auth = this.auth;
    this.#removeCredentialFromUsers(auth, type, slotIndex);
    auth.removeCredential(type, slotIndex);
  }
  #clearAllNonUserCodes(type) {
    this.#clearAllCredentialsOfType(type);
  }
}
const throwInvalidCommand = () => {
  throw new import_types.StatusResponseError("Invalid field value", import_types.StatusCode.InvalidCommand);
};
for (const method of ["clearPinCode", "clearRfidCode"]) {
  (0, import_Supervision.Supervision)(NonUserDoorLockServer, method).onValidationError = throwInvalidCommand;
}
((NonUserDoorLockServer2) => {
  var _credentialKey_dec, _holidaySchedules_dec, _yearDaySchedules_dec, _weekDaySchedules_dec, _credentials_dec, _users_dec, _a, _init;
  class State extends (_a = NonUserBase.State, _users_dec = [(0, import_model.field)((0, import_model.listOf)(import_LockAuth.LockAuth.User), import_model.nonvolatile)], _credentials_dec = [(0, import_model.field)((0, import_model.listOf)(import_LockAuth.LockAuth.Credential), import_model.nonvolatile)], _weekDaySchedules_dec = [(0, import_model.field)((0, import_model.listOf)(import_LockSchedule.LockSchedule.WeekDay), import_model.nonvolatile)], _yearDaySchedules_dec = [(0, import_model.field)((0, import_model.listOf)(import_LockSchedule.LockSchedule.YearDay), import_model.nonvolatile)], _holidaySchedules_dec = [(0, import_model.field)((0, import_model.listOf)(import_LockSchedule.LockSchedule.Holiday), import_model.nonvolatile)], _credentialKey_dec = [(0, import_model.field)(import_model.octstr, import_model.nonvolatile)], _a) {
    constructor() {
      super(...arguments);
      __publicField(this, "users", __runInitializers(_init, 8, this, [])), __runInitializers(_init, 11, this);
      __publicField(this, "credentials", __runInitializers(_init, 12, this, [])), __runInitializers(_init, 15, this);
      __publicField(this, "weekDaySchedules", __runInitializers(_init, 16, this, [])), __runInitializers(_init, 19, this);
      __publicField(this, "yearDaySchedules", __runInitializers(_init, 20, this, [])), __runInitializers(_init, 23, this);
      __publicField(this, "holidaySchedules", __runInitializers(_init, 24, this, [])), __runInitializers(_init, 27, this);
      __publicField(this, "credentialKey", __runInitializers(_init, 28, this)), __runInitializers(_init, 31, this);
    }
  }
  _init = __decoratorStart(_a);
  __decorateElement(_init, 5, "users", _users_dec, State);
  __decorateElement(_init, 5, "credentials", _credentials_dec, State);
  __decorateElement(_init, 5, "weekDaySchedules", _weekDaySchedules_dec, State);
  __decorateElement(_init, 5, "yearDaySchedules", _yearDaySchedules_dec, State);
  __decorateElement(_init, 5, "holidaySchedules", _holidaySchedules_dec, State);
  __decorateElement(_init, 5, "credentialKey", _credentialKey_dec, State);
  __decoratorMetadata(_init, State);
  NonUserDoorLockServer2.State = State;
  class Internal {
    wrongCodeCount = 0;
    autoRelockTimer;
  }
  NonUserDoorLockServer2.Internal = Internal;
})(NonUserDoorLockServer || (NonUserDoorLockServer = {}));
